{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GxSex joint 2df analysis for UHS GWAS of HIV acquisition\n",
    "**Author**: Jesse Marks\n",
    "\n",
    "See [issue #97](https://github.com/RTIInternational/bioinformatics/issues/97)\n",
    "\n",
    "\n",
    "Data were downloaded from S3 at:\n",
    "`s3://rti-hiv/uhs_data`\n",
    "\n",
    "These data were downloaded to EC2 at:\n",
    "`/shared/data/studies/uhs1-3/`\n",
    "\n",
    "The working directory for this 2 degree of freedom GWAS will be:\n",
    "`/shared/jmarks/hiv/uhs1-3`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Auto-scaling tags**\n",
    "Cfncluster automatically creates an autoscaling group for you. To add a tag to the pre-existing auto-scaling groups, you can use the “Auto-scaling” AWS console. This is found on the AWS console left sidebar. On that page you should select the auto-scaling group for your cluster. I believe these are the ones that contain the keywords `jmarks`. For each one, select the “Tags” tab and there will be an option to add a tag. This information was derived from an email sent by Bryan Quach with the subject title being `AWS charges`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extend EBS volume\n",
    "aws ec2 modify-volume --dry-run --volume-id vol-0c6f35ae300a46947 --size 1700\n",
    "\n",
    "# extend file system to the new volume capacity.\n",
    "sudo resize2fs /dev/xvdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare Files for ProbABEL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phenotype data\n",
    "The phenotype data were processed by Fang. The PCs selected for the AA cohort were PC2, PC6, PC5, & PC9.\n",
    "You can find the details for the data processing [here](https://github.com/RTIInternational/ffang_notebooks/blob/master/uhs_gwas_hiv.ipynb).\n",
    "\n",
    "`head pheno_hiv_aa_probabel`\n",
    "```\n",
    "iid     hiv     gender  age     pc2     pc6     pc5     pc9\n",
    "109@1064714572_109@1064714572   0       1       26      0.0033  0.0027  0.0027  -0.0012\n",
    "202@1064714531_202@1064714531   0       0       27      0.0004  -0.0107 -0.0026 0.0079\n",
    "312@1064714548_312@1064714548   0       1       34      -0.0014 -0.0083 -0.0028 -0.0033\n",
    "```\n",
    "\n",
    "Note the subject IDs are of the form famID_indivID per the MIS output.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EC2 #\n",
    "cd /shared/jmarks/hiv/uhs1-3/association_tests/001"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotype data\n",
    "### Inflate Imputation Results\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email. The password was found in [Fang's notebook](https://github.com/RTIInternational/ffang_notebooks/blob/master/uhs_gwas_hiv.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EC2 console ###\n",
    "cd /shared/data/studies/uhs1-3/aa\n",
    "\n",
    "# inflate chr results\n",
    "for f in {1..22};do\n",
    "echo '#!/bin/bash' > chr_$f.sh\n",
    "echo '' >> chr_$f.sh\n",
    "echo 'unzip -P \"jYdXsUKB55Gcag\" chr_'$f'.zip' >> chr_$f.sh\n",
    "done\n",
    "\n",
    "# might want to adjust nslots to more or less\n",
    "for chr in {1..22}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name inflate_chr${chr} \\\n",
    "    --script_prefix chr${chr}_inflation \\\n",
    "    --mem 15 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 5 \\\n",
    "    --program bash chr_${chr}.sh\n",
    "done\n",
    "\n",
    "mv *gz unencrypted\n",
    "# note that I am deleting the zip files from Michigan Imputation Server because\n",
    "# these data are already saved to S3\n",
    "rm\n",
    "rm *.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert VCF files to MaCH format\n",
    "The ProbABEL software requires the data to be in a format that is different than the VCF format that is the output from MIS. Specifically, we need to convert the VCF files to MaCH format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EC2 ###\n",
    "cd /shared/data/studies/uhs1-3/aa/unencrypted\n",
    "\n",
    "for chr in {1..22};do\n",
    "mkdir -p /shared/data/studies/uhs1-3/aa/unencrypted/mach/chr$chr\n",
    "done\n",
    "\n",
    "imputedir=/shared/data/studies/uhs1-3/aa/imputed\n",
    "# Convert dosage file to mldose format\n",
    "for ethnicity in aa; do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name chr${chr}_dosage_convert \\\n",
    "            --script_prefix $imputedir/mach/chr$chr.mach \\\n",
    "            --mem 10 \\\n",
    "            --nslots 4 \\\n",
    "            --buffer 10000 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/third_party/dosage_converter_v1.0.4/bin/DosageConvertor \\\n",
    "                --vcfDose ${imputedir}/chr$chr.dose.vcf.gz \\\n",
    "                --info ${imputedir}/chr$chr.info \\\n",
    "                --prefix ${imputedir}/mach/chr$chr/chr$chr.mach \\\n",
    "                --type mach \\\n",
    "                --format 1  # contains the expected alternate allele count (one value per sample per marker).\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* To get an idea of the range for the run time of DosageConvertor I list here the quickest and longest run times chromosomes 2 and 21, respectively.\n",
    "\n",
    "```\n",
    "==> chr21.mach.qsub.log <==\n",
    " ------------------------------------------------------------------------------\n",
    "                                END OF PROGRAM\n",
    " ------------------------------------------------------------------------------\n",
    "\n",
    " Program Successfully Implemented...\n",
    "\n",
    " Total Run completed in 1 hours, 53 mins, 27 seconds.\n",
    "\n",
    " Thank You for using DosageConvertor !!!\n",
    "\n",
    "\n",
    "==> chr2.mach.qsub.log <==\n",
    " ------------------------------------------------------------------------------\n",
    "                                END OF PROGRAM\n",
    " ------------------------------------------------------------------------------\n",
    "\n",
    " Program Successfully Implemented...\n",
    "\n",
    " Total Run completed in 11 hours, 18 mins, 14 seconds.\n",
    "\n",
    " Thank You for using DosageConvertor !!!\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start ProbABEL Analysis\n",
    "Perform  2 degree-freedom GWAS (GxSex)\n",
    "See the [manual](http://www.genabel.org/sites/default/files/pdfs/ProbABEL_manual.pdf) for more information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare directory structure and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "mkdir -p /shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001/aa/{final,processing}\n",
    "imputeDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "phenoDir=/shared/data/studies/uhs1-3/aa/phenotype/pheno_hiv_aa_probabel\n",
    "\n",
    "\n",
    "for i in {1..22}; do\n",
    "    mkdir -p $imputeDir/chr${i}/mldose_files\n",
    "done\n",
    "\n",
    "# prepare info files (only 7 fields allowed)\n",
    "for chr in {1..22}; do\n",
    "    cut -f 1-7 ${imputeDir}/chr${chr}/chr${chr}.mach.mach.info > ${imputeDir}/chr${chr}/mldose_files/chr${chr}.final.info &\n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reorder and prune mach.dose files down to subjects in phenotype file\n",
    "Note, the following is a bash script\n",
    "\n",
    "Following instructions from the [ProbABEL manual](http://www.genabel.org/sites/default/files/pdfs/ProbABEL_manual.pdf) on page 7 in section 3.2 - *Genomic predictor file*.\n",
    "\n",
    "The order of SNPs in the SNP information file and DOSE  file must be the same. The total number and the order of the IDs in the phenotype file should be exactly the same as in the genomic predictor (DOSE) file. \n",
    "\n",
    "\n",
    "**Note**: The dosage files (imputed data) may contain data on more subjects than is in the phenotype file. For this reason, we have to prune the dosage files down to only the subjects contained in the phenotype file. If, however, the phenotype file and the dosage files contained exactly the same subjects then it would make more sense to reorder the phenotype file rather than reordering the dosage files because the dosage files are much larger."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "processingDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "\n",
    "mkdir $processingDir/file_prep\n",
    "\n",
    "# first, get the order of the IDs in the mldose_file\n",
    "# use chr22 because it is the smallest\n",
    "zcat $processingDir/chr22*gz | cut -f1 > $processingDir/dose.IDs.all &\n",
    "tail -n +2 /shared/data/studies/uhs1-3/aa/phenotype/pheno_hiv_aa_probabel | cut -f 1 > $processingDir/file_prep/phen.IDs.all\n",
    "\n",
    "#-------------------------------------------------------------------------------- \n",
    "awk ' NR==FNR{ map[FNR]=$1} {gsub(/^.+->/, NR\"->\", $1) ' $processingDir/file_prep/phen.IDs.all $processingDir/file_prep/dose.IDs.all > test\n",
    "\n",
    "\n",
    "#-------------------------------------------------------------------------------- \n",
    "\n",
    "\n",
    "\n",
    "## start bash script titled `mldose_ordering.sh`\n",
    "-------------------------------------------------------------------------------\n",
    "#!/bin/bash\n",
    "\n",
    "zcat $1/chr$2/chr$2.mach.mach.dose.gz | grep -f id_order > chr$2/ordered_mach.dose\n",
    "\n",
    "# manual says that probabel expects a sequential id followed by an arrow followed by study ID\n",
    "awk '{gsub(/^.+->/, NR\"->\", $1); print $0}'  chr$2/mldose_files/ordered_mach.dose > chr$2/mldose_files/chr$2.final.dose\n",
    "rm chr$2/ordered_mach.dose\n",
    "\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "zcat $1/chr$3.mach.mach.dose.gz | awk '{gsub(/^.+->/, NR\"->\", $1); print $0}' > $1/mldose_files/chr$3.final.dose\n",
    "\n",
    "\n",
    "\n",
    "## END bash script\n",
    "--------------------------------------------------------------------------------\n",
    "\n",
    "# get id_order from pheno file\n",
    "tail -n +2 ${phenoDir} | cut -d ' ' -f 1 > id_order\n",
    "\n",
    "imputeDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "# submit as jobs\n",
    "for chr in {1..22};do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name mldose_chr${chr}_prep \\\n",
    "            --script_prefix chr${chr}/mldose.ordering \\\n",
    "            --mem 8 \\\n",
    "            --nslots 3 \\\n",
    "            --program bash prep_mldose.sh $imputeDir ${chr}\n",
    "done\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### had to alter the dose file\n",
    "The dosage files contained one extra subject that was not included in the phenotype data. Therefore, we have to perform the following command to exclude this subject."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "\n",
    "# python script\n",
    "#--------------------------------------------------------------------------------\n",
    "### This script prunes the imputed data (dosage files) based on subject IDs\n",
    "### being present or not in the phenotype file\n",
    "### sample run:\n",
    "### python test.script.py 22\n",
    "### \n",
    "### note that the chr number is required as an argument\n",
    "\n",
    "import gzip,sys\n",
    "chr = sys.argv[1]\n",
    "\n",
    "BASE_DIR = \"/shared/data/studies/uhs1-3/aa/imputed/mach/\"\n",
    "\n",
    "doseFile = gzip.open(BASE_DIR + 'chr' + chr +  '/chr' + chr + '.mach.mach.dose.gz')\n",
    "\n",
    "phenFile = open(BASE_DIR + 'phen.IDs.all', 'r')\n",
    "phenLines = phenFile.read().splitlines()\n",
    "\n",
    "outF = file(BASE_DIR + 'chr' + chr + '/mldose_files/' + 'chr' + chr + '.final.dose', 'w')\n",
    "\n",
    "count = 1\n",
    "for line in doseFile:\n",
    "    inF = line.split()\n",
    "    tmpID = inF[0].split(\"->\")[0]\n",
    "    if (tmpID in phenLines):\n",
    "        inF[0] = str(count) + \"->\" + tmpID\n",
    "        outF.write(\"\\t\".join(inF) + \"\\n\")\n",
    "        count = count + 1\n",
    "#--------------------------------------------------------------------------------\n",
    "processingDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "                --job_name alter_dose_chr$chr \\\n",
    "                --script_prefix $processingDir/chr${chr}/mldose.ordering \\\n",
    "                --mem 8 \\\n",
    "                --nslots 5 \\\n",
    "                --program python mldose.file.prep.py $chr\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## delete subject on line 756 because they are missing in the phenotype file\n",
    "#\n",
    "#sed '756d' *dose > chr15.final.dose.filtered\n",
    "#\n",
    "## note do this for all of the dose files\n",
    "#\n",
    "#for chr in {1..22}; do\n",
    "#    sed '756d' chr$chr/mldose_files/*dose > chr$chr/mldose_files/chr$chr.final.dose.filtered\n",
    "#done &\n",
    "#\n",
    "#processingDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "#for chr in {1..14};do\n",
    "#    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "#            --job_name alter_dose_chr$chr \\\n",
    "#            --script_prefix $processingDir/chr${chr}/mldose.delete.line \\\n",
    "#            --mem 8 \\\n",
    "#            --nslots 3 \\\n",
    "#            --program bash alter.dose.sh $processingDir ${chr}\n",
    "#done\n",
    "#    \n",
    "#processingDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "#for chr in {16..22};do\n",
    "#    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "#            --job_name alter_dose_chr$chr \\\n",
    "#            --script_prefix $processingDir/chr${chr}/mldose.delete.line \\\n",
    "#            --mem 8 \\\n",
    "#            --nslots 3 \\\n",
    "#            --program bash alter.dose.sh $processingDir ${chr}\n",
    "#done\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create legend\n",
    "\n",
    "\n",
    "Pg. 10 of the [ProbABEL](http://www.genabel.org/sites/default/files/pdfs/ProbABEL_manual.pdf) manual reports that the legend should be in HapMap format:\n",
    "\n",
    "`rsID, position, allele1, allele2`\n",
    "\n",
    "Note: We do not have the rsID right now. We have the chromosome and the position. It shouldn't be to difficult to convert this to the actual rsID. We might be able to use Nathan Gladdis' Nyholt script to convert these. This can be converted later, though. Also, this is an optional file. The only column that is actually used is the SNP location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "\n",
    "# HapMap \"legend\" file format\n",
    "# note that our markerIDs are not in rsID format - we have chr:position for the rsID instead\n",
    "for chr in {1..22}; do\n",
    "    # grab the position, allele1 and allele2\n",
    "    awk '{gsub(/^.+:/, \"\", $1); print $1,$2,$3}' chr${chr}/mldose_files/*.info > chr${chr}/position_and_alleles.txt\n",
    "    # grab the SNP ID, which is chr:position\n",
    "    cut -f 1 chr${chr}/mldose_files/*.info > chr${chr}/snp_ids.txt\n",
    "    # combine the columns\n",
    "    paste -d \" \"  chr${chr}/snp_ids.txt chr${chr}/position_and_alleles.txt > chr${chr}/map.chr${chr}.legend\n",
    "    # rename the header\n",
    "    sed '1 s/.*/id position 0 1/' chr${chr}/map.chr${chr}.legend > chr${chr}/mldose_files/chr$chr.final.legend \n",
    "    \n",
    "    rm chr${chr}/{*.txt,*legend}\n",
    "done &"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform 2-df GWAS \n",
    "**Note:** Because performing the GWAS is so memory intensive, we updated the compute node fleet to r4.8xlarge because they have 244G of memory and 32 CPU. After this analysis is finished, revert to a smaller (less expensive) set of compute nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001\n",
    "\n",
    "baseDir=/shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001/aa/processing\n",
    "imputeDir=/shared/data/studies/uhs1-3/aa/imputed/mach\n",
    "phenoDir=/shared/data/studies/uhs1-3/aa/phenotype\n",
    "\n",
    "# note the interaction 1 corresponds to the second covariate (gender)\n",
    "# note that I set the nslots so that each job occupies the entire node.\n",
    "for chr in {22..1}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name probabel_gwas_chr${chr} \\\n",
    "            --script_prefix $baseDir/chr${chr}/chr${chr}_gwas \\\n",
    "            --mem 200 \\\n",
    "            --nslots 17 \\\n",
    "            --program /shared/bioinformatics/software/third_party/probabel-0.5.0/bin/palinear \\\n",
    "                --pheno $phenoDir/pheno_hiv_aa_probabel \\\n",
    "                --dose $imputeDir/chr${chr}/mldose_files/chr${chr}.final.dose \\\n",
    "                --info $imputeDir/chr${chr}/mldose_files/chr${chr}.final.info \\\n",
    "                --map $imputeDir/chr$chr/mldose_files/chr${chr}.final.legend \\\n",
    "                --interaction 1 \\\n",
    "                --chrom ${chr} \\\n",
    "                --out $baseDir/chr${chr}/chr${chr}_results\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Post ProbABEL Processing\n",
    "## Calculate chi, p, and OR \n",
    "First, we test chr15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001/aa/processing/chr15\n",
    "\n",
    "base_dir=/shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001/aa/processing\n",
    "for chr in {10..1}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name uhs123_chr${chr} \\\n",
    "            --script_prefix ${base_dir}/chr${chr}/uhs123_aa.1000G.chr$chr.stats \\\n",
    "            --mem 5 \\\n",
    "            --nslots 4 \\\n",
    "            --program Rscript ${base_dir}/calculate_2df_stats_from_probabel_results.R \\\n",
    "                --in ${base_dir}/chr${chr}/chr${chr}_results_add.out.txt \\\n",
    "                --out ${base_dir}/chr${chr}/uhs123_aa.1000G.chr$chr.stats \\\n",
    "                --interaction_covar gender\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "name A1 A2 Freq1 MAF Quality Rsq n Mean_predictor_allele chrom position beta_SNP_addA1 sebeta_SNP_addA1 beta_SNP_gender sebeta_SNP_gender cov_SNP_int_SNP_gender chi2_SNP_add\n",
    "\n",
    "#### calc.stats script\n",
    "Note, I had to add `A1`  to the end of `beta_SNP_add`. Check with the 1df calc stats rscript to see if this is consistent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fileIn <- \"chr22_results_add.out.txt\"\n",
    "interactionCovar <- \"gender\"\n",
    "fileOut<- \"chr22.interactive.test\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/share/apps/R/bin/Rscript\n",
    "\n",
    "args <- commandArgs(TRUE)\n",
    "\n",
    "loop = TRUE\n",
    "while (loop) {\n",
    "        if (args[1] == \"--in\") {\n",
    "                fileIn = args[2]\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--out\") {\n",
    "                fileOut = args[2]\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--interaction_covar\") {\n",
    "                interactionCovar = args[2]\n",
    "        }\n",
    "\n",
    "        if (length(args)>1) {\n",
    "                args<-args[2:length(args)]\n",
    "        } else {\n",
    "                loop = FALSE\n",
    "        }\n",
    "}\n",
    "\n",
    "cat(\"Reading \", fileIn, \"...\\n\", sep=\"\")\n",
    "dataTable = read.table(fileIn, header = TRUE, na.strings = c(\"nan\",\"NA\", \"NaN\"))\n",
    "cat(\"Calculating...\\n\")\n",
    "dataTable = dataTable[complete.cases(dataTable$beta_SNP_add),]\n",
    "for (i in 1:nrow(dataTable)) {\n",
    "        beta = c(dataTable[i,\"beta_SNP_addA1\"],dataTable[i,paste(\"beta_SNP_\",interactionCovar,sep=\"\")])\n",
    "        sigm = matrix(c(dataTable[i,\"sebeta_SNP_addA1\"]^2,dataTable[i,paste(\"cov_SNP_int_SNP_\",interactionCovar,sep=\"\")],dataTable[i,paste(\"cov_SNP_int_SNP_\",interactionCovar,sep=\"\")],dataTable[i,paste(\"sebeta_SNP_\",interactionCovar,sep=\"\")]^2),2,2)\n",
    "  sigmSolved = try(solve(sigm),silent=TRUE)\n",
    "\n",
    "  if (is.matrix(sigmSolved)) {\n",
    "                dataTable[i,\"chi_2df\"] = t(beta) %*% sigmSolved %*% beta\n",
    "                dataTable[i,\"p_2df\"] = 1 - pchisq(dataTable[i,\"chi_2df\"],2)\n",
    "  } else {\n",
    "                dataTable[i,\"chi_2df\"] = NA\n",
    "                dataTable[i,\"p_2df\"] = NA\n",
    "        }\n",
    "}\n",
    "dataTable = dataTable[complete.cases(dataTable$p_2df),]\n",
    "\n",
    "if (\"chrom\" %in% colnames(dataTable) & \"position\" %in% colnames(dataTable)) {\n",
    "        dataTable = cbind(subset(dataTable,select=c(name,chrom,position)),subset(dataTable,select=-c(name,chrom,position)))\n",
    "} else if (\"chrom\" %in% colnames(dataTable)) {\n",
    "        dataTable = cbind(subset(dataTable,select=c(name,chrom)),subset(dataTable,select=-c(name,chrom)))\n",
    "} else if (\"position\" %in% colnames(dataTable)) {\n",
    "        dataTable = cbind(subset(dataTable,select=c(name,position)),subset(dataTable,select=-c(name,position)))\n",
    "}\n",
    "\n",
    "cat(\"Writing \", fileOut, \"...\\n\", sep=\"\")\n",
    "write.table(dataTable, file = fileOut, row.names = FALSE, quote=FALSE)\n",
    "\n",
    "cat(\"Done\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert SNP ID from chr:position to rsID\n",
    "\n",
    "My results do not have an rsID. I need to convert my IDs - which are of the form chr:position - to rsID. We will do this by using the position information of the variants.\n",
    "\n",
    "Note: In some cases, there will be multiple variants at a single position. This happens if the variant is not biallelic. In this case, they will likely be filtered out because of MAF<0.01 anyways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001/aa/processing\n",
    "\n",
    "processingDir=/shared/bioinformatics/jmarks/hiv/uhs1-3/association_tests/001/aa/processing\n",
    "\n",
    "ancestry=aa\n",
    "for ((chr=1; chr<23; chr++)); do\n",
    "    echo \"Processing chr${chr}_${ancestry}\"\n",
    "    awk 'NR==FNR { a[$2]=$1; next } $3 in a { $1 = a[$3]; print $0 }'\\\n",
    "        <(zcat /shared/data/ref_panels/1000G/2013.12/impute2/ALL/ALL.chr${chr}.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.nomono.unique_ids.legend.gz)\\\n",
    "        <(zcat $processingDir/chr${chr}/uhs123_${ancestry}.1000G.chr$chr.stats.gz) >\\\n",
    "        $processingDir/chr${chr}/uhs123_${ancestry}.1000G.chr$chr.rsID.stats\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter by MAF\n",
    "\n",
    "Filter any variant whose MAF <= 0.01 in both subjects and 1000G population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/bioinformatics/ffang/uhs/processing\n",
    "\n",
    "# Filter MAF <= 0.01 in subjects\n",
    "for ethnicity in ea aa; do\n",
    "    for ((chr=1; chr<23; chr++));do\n",
    "        echo \"Processing chr${chr}_${ethnicity}\"\n",
    "        head -n 1 chr${chr}_${ethnicity}_palogist_results_filtered_rsID.stats > \\\n",
    "            chr${chr}_${ethnicity}.subject_maf_gt_0.01\n",
    "\n",
    "        awk ' NR>=2 {if ($7 >= 0.01) {print $0}}' \\\n",
    "            chr${chr}_${ethnicity}_palogist_results_filtered_rsID.stats \\\n",
    "            >> chr${chr}_${ethnicity}.subject_maf_gt_0.01\n",
    "    done\n",
    "done\n",
    "\n",
    "# transfer data from EC2 to S3 and then to another head node\n",
    "# aws s3 cp /shared/bioinformatics/ffang/uhs/processing/ s3://rti-hiv/uhs_data/results/ --recursive \\\n",
    "#     --exclude \"*\" --include \"chr*.subject_maf_gt_0.01_ea\" --dryrun\n",
    "# aws s3 cp s3://rti-hiv/uhs_data/results/ /shared/bioinformatics/ffang/uhs/processing/ --recursive \\\n",
    "#     --exclude \"*\" --include \"chr*.subject_maf_gt_0.01_ea\" --dryrun\n",
    " \n",
    "# Filter MAF <= 0.01 in 1000G population\n",
    "\n",
    "base_dir=/shared/bioinformatics/ffang/uhs/processing\n",
    "final_dir=/shared/bioinformatics/ffang/uhs/final\n",
    "for ethnicity in ea aa; do\n",
    "    if [ $ethnicity == \"aa\" ]; then\n",
    "        group=afr\n",
    "    elif [ $ethnicity == \"ea\" ]; then\n",
    "        group=eur\n",
    "    fi\n",
    "    for ((chr=1; chr<23; chr++)); do\n",
    "        idList=/shared/data/ref_panels/1000G/2013.12/impute2/ALL/ALL.chr${chr}.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.nomono.unique_ids.maf_gt_0.01_${group} \n",
    "        echo \"Processing chr${chr}_${ethnicity}\"      \n",
    "            /shared/bioinformatics/software/perl/utilities/extract_rows.pl \\\n",
    "            --source $base_dir/chr${chr}_${ethnicity}.subject_maf_gt_0.01 \\\n",
    "            --id_list $idList \\\n",
    "            --out $base_dir/chr${chr}_${ethnicity}.subject_1000G_maf_gt_0.01 \\\n",
    "            --header 1 \\\n",
    "            --id_column 0 \n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rsq filter\n",
    "Filter any variant whose $R^2<0.30$.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/bioinformatics/ffang/uhs/processing\n",
    "for ethnicity in ea aa; do\n",
    "    for ((chr=1; chr<23; chr++));do\n",
    "        echo \"Processing chr${chr}_${ethnicity}\"\n",
    "        head -n 1 chr${chr}_${ethnicity}.subject_1000G_maf_gt_0.01 > chr${chr}_${ethnicity}.maf_gt_0.01_rsq_gt_0.30\n",
    "\n",
    "        awk ' NR>=2 {if ($9 >= 0.30) {print $0}}' \\\n",
    "            chr${chr}_${ethnicity}.subject_1000G_maf_gt_0.01 \\\n",
    "            >> chr${chr}_${ethnicity}.maf_gt_0.01_rsq_gt_0.30\n",
    "    done\n",
    "done\n",
    "\n",
    "mv *maf_gt_0.01_rsq_gt_0.30 ../final/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/bioinformatics/ffang/uhs/processing\n",
    "\n",
    "for ethnicity in ea aa; do\n",
    "    outFile=/shared/bioinformatics/ffang/uhs/processing/uhs_hiv_age_gender_pcs_${ethnicity}.table  \n",
    "\n",
    "    echo -e \"VARIANT_id\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "\n",
    "    for ((chr=1; chr<23; chr++)); do\n",
    "       inFile=/shared/bioinformatics/ffang/uhs/final/chr${chr}_${ethnicity}.maf_gt_0.01_rsq_gt_0.30\n",
    "\n",
    "        echo Processing $inFile\n",
    "\n",
    "        tail -n +2 $inFile |\\\n",
    "        perl -lne '/^(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)(?:\\s+\\S+){10}\\s+(\\S+)/;\n",
    "                      if (($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\") && ($5 eq \"A\" || $5 eq \"C\" || $5 eq \"G\" || $5 eq \"T\")) {\n",
    "                        print join(\"\\t\",$1,$2,$3,$6,\"snp\");\n",
    "                      } else {\n",
    "                        print join(\"\\t\",$1,$2,$3,$6,\"indel\");\n",
    "                      }' >> $outFile\n",
    "    done\n",
    "\n",
    "    Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "        --in /shared/bioinformatics/ffang/uhs/processing/uhs_hiv_age_gender_pcs_${ethnicity}.table \\\n",
    "        --in_chromosomes autosomal_nonPAR \\\n",
    "        --in_header \\\n",
    "        --out /shared/bioinformatics/ffang/uhs/processing/uhs_${ethnicity}_hiv_age_gender_pcs_plots \\\n",
    "        --col_id VARIANT_id \\\n",
    "        --col_chromosome CHR \\\n",
    "        --col_position POSITION \\\n",
    "        --col_p P \\\n",
    "        --col_variant_type TYPE \\\n",
    "        --generate_snp_indel_manhattan_plot \\\n",
    "        --manhattan_odd_chr_color red \\\n",
    "        --manhattan_even_chr_color blue \\\n",
    "        --manhattan_points_cex 1.5 \\\n",
    "        --generate_snp_indel_qq_plot \\\n",
    "        --qq_lines \\\n",
    "        --qq_points_bg black \\\n",
    "        --qq_lambda\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View Q-Q and Manhattan Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter to top SNPs\n",
    "P< 0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
