{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [FTND] Minnesota 1df processing\n",
    "**Author**: Jesse Marks\n",
    "\n",
    "~~**Date**: July 27, 2018~~ <br>\n",
    "**Date**: August 10, 2018\n",
    "\n",
    "**GitHub Issue:** [#105](https://github.com/RTIInternational/bioinformatics/issues/105#issuecomment-412103553)\n",
    "\n",
    "The following email was sent to Jesse Marks from John Guo\n",
    "on 20180717. The email subject line reads:<br>\n",
    "`\"[FTND] - Minnesota 1df processing\"`\n",
    "\n",
    "```css\n",
    "The new Minnesota results are available at \n",
    "\n",
    "https://s3.console.aws.amazon.com/s3/buckets/rti-uploads/hyoung/?region=us-east-1&tab=overview\n",
    "\n",
    "I used script on MIDAS to split the file by chromosome:\n",
    "/share/nas04/bioinformatics_group/data/studies/minnesota_twins/raw_results/process_raw.20180611.py\n",
    "\n",
    "Then performed ID conversion, filtering, and plotting using script here:\n",
    "/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/002/_methods.minnesota_twin.imputed.assoc_tests.001.sh\n",
    "\n",
    "Could you please move the files to an appropriate location on S3 and process it on AWS?\n",
    "\n",
    "Please let me know if you have any questions.\n",
    "```\n",
    "\n",
    "**Update August 10, 2018**: There were issues with processing this Minnesota Twins version 3 GWAS data. Specifically, the RSQ and PVALUE data were missing. The re-processing of these data are described in the paragraphs below. The updated processing steps+details in this notebook begin in the section titled `Re-processing`.\n",
    "\n",
    "```\n",
    "There were three versions of Minnesota Twins results. The first version was\n",
    "fine except that the results did not contain the imputation quality column\n",
    "needed to perform the standard RSQ>0.30 filter that is in our GWAS processing\n",
    "protocol. The second version contained the imputation quality data (RSQ column)\n",
    "however the PVALUE does not always match exactly with the first version. The\n",
    "third version is missing both PVALUE and RSQ data. \n",
    "\n",
    "Considering the details gathered from email correspondence between Dana Hancock\n",
    "and Hannah Young, it was decided that we will use the version 1 summary\n",
    "statistics and append the RSQ column from version 2.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download Data From S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "mkdir -p /shared/data/studies/minnesota_twins/raw_results\n",
    "cd /shared/data/studies/minnesota_twins/raws_results\n",
    "\n",
    "aws s3 sync s3://rti-uploads/hyoung/ ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split up by chromosome\n",
    "These GWAS results are all combined in one file. We will split them apart by chromosome first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip,sys\n",
    "BASE_DIR = '/shared/data/studies/minnesota_twins/'\n",
    "for ethnicity in ['ea']:\n",
    "        inF = gzip.open(BASE_DIR + 'raw_results/MCTFR_FTND.MetaScore.assoc.gz')\n",
    "        # each time this is called, the next line will be returned.\n",
    "        line = inF.readline() \n",
    "        # keep skipping the lines until the line with the headers appears.\n",
    "        # Namely, the headers start with one pound sign then CHROM ...\n",
    "        while(line[:2] == \"##\"):\n",
    "                line = inF.readline()\n",
    "\n",
    "        firstLine = line[1:] # skip that first pount in the header line\n",
    "        # split returns a list of words (headers)\n",
    "        chrIndex = firstLine.split().index('CHROM') # returns 0 because CHROM is first header\n",
    "        lastChr = ''\n",
    "        line = inF.readline() # go to actual chromosome\n",
    "        while(line): # while we are not at the end of the file\n",
    "                # if the new line is not the same chromosome as the last one processed enter loop\n",
    "                # note it will enter this loop right from the gitgo because lastChr= '' by default\n",
    "                if(line.split()[chrIndex] != lastChr): \n",
    "                        fname = 'minnesota_twins.' + ethnicity + '.1000G.chr' + line.split()[chrIndex] + '.' + 'CAT_FTND~1df_add.out.txt'\n",
    "                        dir = '001/' + ethnicity + '/' + 'processing/chr' + line.split()[chrIndex] + '/'\n",
    "                        outF = file(BASE_DIR + 'imputed/v1/association_tests/' + dir + fname, 'w')\n",
    "                        # write to a new file based on the new chr we are processing\n",
    "                        # also add the column Marker to the column header\n",
    "                        outF.write(\"Marker\" + \"\\t\" + \"\\t\".join(firstLine.split()) + \"\\n\")\n",
    "                        lastChr = line.split()[chrIndex] # new last chromosome now\n",
    "                        print('Processing : ' + 'Chr ' + lastChr)\n",
    "                # creating the Markername = CHR:POSITION\n",
    "                tmp = line.split()\n",
    "                outF.write(tmp[0] + \":\" + tmp[1] + \"\\t\" + line)\n",
    "                line = inF.readline() # read the next line\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Then performed ID conversion, filtering, and plotting using script here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ethnicity in ea; do\n",
    "  for (( chr=1; chr<23; chr++ )); do\n",
    "    mkdir -p /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr\n",
    "  done\n",
    "  mkdir /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/final\n",
    "done\n",
    "\n",
    "### Convert ID to phase 3 ID ###\n",
    "for ethnicity in ea; do\n",
    "    if [ $ethnicity == \"aa\" ]\n",
    "    then\n",
    "        group=afr\n",
    "    else\n",
    "        group=eur\n",
    "    fi\n",
    "    for (( chr=1; chr<23; chr++ )); do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name MTC_${chr} \\\n",
    "            --script_prefix /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G.chr$chr.CAT_FTND~1df.phase3ID_add.out.txt \\\n",
    "            --mem 15 \\\n",
    "            --nslots 3 \\\n",
    "            --priority 0 \\\n",
    "            --program perl /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "            --file_in /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G.chr$chr.CAT_FTND~1df_add.out.txt \\\n",
    "            --file_out /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G.chr$chr.CAT_FTND~1df.phase3ID_add.out.txt \\\n",
    "            --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.legend.gz \\\n",
    "            --file_in_header 1 \\\n",
    "            --file_in_id_col 0 \\\n",
    "            --file_in_chr_col 1 \\\n",
    "            --file_in_pos_col 2 \\\n",
    "            --file_in_a1_col 3 \\\n",
    "            --file_in_a2_col 4 \\\n",
    "            --chr $chr\n",
    "    done\n",
    "done\n",
    "\n",
    "# Check for completion\n",
    "for ethnicity in ea; do\n",
    "    if [ $ethnicity == \"aa\" ]\n",
    "    then\n",
    "        group=afr\n",
    "    else\n",
    "        group=eur\n",
    "    fi\n",
    "    for (( chr=1; chr<23; chr++ ))\n",
    "    do\n",
    "        file=/shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G.chr$chr.CAT_FTND~1df.phase3ID_add.out.txt.qsub.log\n",
    "        if [ -f $file ]\n",
    "        then\n",
    "          logLineCount=$(wc -l $file | perl -lane 'print $F[0];')\n",
    "          if [ $logLineCount -ne 4 ]\n",
    "          then\n",
    "            echo $file line count: $logLineCount\n",
    "          else\n",
    "            tail -n 1 $file |\n",
    "              perl -ne 'chomp; if (!/^Done/) { print \"'$file'\\n\".$_.\"\\n\"; }'\n",
    "          fi\n",
    "        else\n",
    "          echo $file missing\n",
    "        fi\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "### START Filter ###\n",
    "# create 1000G_p3 MAF filtered files\n",
    "thouDir=/shared/data/ref_panels/1000G/2014.10\n",
    "for chr in {1..22};do\n",
    "    awk ' { if ( $9>=0.01 ) { print $1 } }' <(zcat $thouDir/1000GP_Phase3_chr$chr.legend.gz) >\\\n",
    "        $thouDir/1000GP_Phase3_chr$chr.legend.maf_lte_0.01_eur\n",
    "done &\n",
    "\n",
    "\n",
    "\n",
    "thouDir=/shared/data/ref_panels/1000G/2014.10\n",
    "# MAF > 0.01 in AFR (AA) or EUR (EA)\n",
    "for ethnicity in ea; do\n",
    "  if [ $ethnicity == \"aa\" ]\n",
    "  then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for (( chr=1; chr<23; chr++ )); do\n",
    "    if [ $chr == \"23\" ]; then\n",
    "      idList=/shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.maf_lte_0.01_$group\n",
    "    else\n",
    "      idList=$thouDir/1000GP_Phase3_chr$chr.legend.maf_lte_0.01_eur\n",
    "    fi\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "      --job_name ${ethnicity}_$chr\\\n",
    "      --script_prefix /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G_p3.chr$chr.CAT_FTND~1df.maf_gt_0.01.$group \\\n",
    "      --mem 3.8 \\\n",
    "      --priority 0 \\\n",
    "      --program /shared/bioinformatics/software/perl/utilities/extract_rows.pl \\\n",
    "      --source /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G.chr$chr.CAT_FTND~1df.phase3ID_add.out.txt \\\n",
    "          --id_list $idList \\\n",
    "          --out /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G_p3.chr$chr.CAT_FTND~1df.maf_gt_0.01.$group \\\n",
    "          --header 1 \\\n",
    "          --id_column 0 \n",
    "  done\n",
    "done\n",
    "\n",
    "       \n",
    "#for ethnicity in ea\n",
    "#do\n",
    "#  mv /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr*/*.maf_gt_0.01_??? \\\n",
    "#   /shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/final\n",
    "#done\n",
    "\n",
    "# MAF > 0.01 in study\n",
    "for ethnicity in ea; do\n",
    "  if [ $ethnicity == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for (( chr=1; chr<23; chr++ )); do\n",
    "    inFile=/shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G_p3.chr$chr.CAT_FTND~1df.maf_gt_0.01.$group\n",
    "    outFile=${inFile}+minnesota_twins\n",
    "    echo Processing $inFile\n",
    "    head -n 1 $inFile > $outFile\n",
    "    tail -n +2 $inFile |\n",
    "      perl -lane 'if ($F[6] >= 0.01 & $F[6] <= 0.99) { print; }' >> $outFile\n",
    "  done\n",
    "done\n",
    "\n",
    "# RSQ > 0.3 in study\n",
    "for ethnicity in ea; do\n",
    "  if [ $ethnicity == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for (( chr=1; chr<23; chr++ )); do\n",
    "    inFile=/shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G_p3.chr$chr.CAT_FTND~1df.maf_gt_0.01.$group+minnesota_twins\n",
    "    outFile=${inFile}.RSQ\n",
    "    echo Processing $inFile\n",
    "    head -n 1 $inFile > $outFile\n",
    "    tail -n +2 $inFile |\n",
    "      perl -lane 'if ($F[17] > 0.3) { print; }' >> $outFile\n",
    "  done\n",
    "done\n",
    "\n",
    "\n",
    "### END Filter ###\n",
    "\n",
    "\n",
    "### START Generate plots ###\n",
    "for ethnicity in ea; do\n",
    "  if [ $ethnicity == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for ext in $group ${group}+minnesota_twins ${group}+minnesota_twins.RSQ; do\n",
    "  outFile=/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/minnesota_twins.$ethnicity.1000G.CAT_FTND~1df.maf_gt_0.01_$ext.table\n",
    "  echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "  for (( chr=1; chr<24; chr++ )); do\n",
    "    inFile=/shared/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/chr$chr/minnesota_twins.$ethnicity.1000G_p3.chr$chr.CAT_FTND~1df.maf_gt_0.01.$ext\n",
    "    echo Processing $inFile\n",
    "    tail -n +2 $inFile |\n",
    "      perl -lne '/^(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)(?:\\s+\\S+){11}\\s+(\\S+)/;\n",
    "                  if (($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\") && ($5 eq \"A\" || $5 eq \"C\" || $5 eq \"G\" || $5 eq \"T\")) {\n",
    "                    print join(\"\\t\",$1,$2,$3,$6,\"snp\");\n",
    "                  } else {\n",
    "                    print join(\"\\t\",$1,$2,$3,$6,\"indel\");\n",
    "                  }' >> $outFile\n",
    "        done\n",
    "  done\n",
    "done\n",
    "for ethnicity in ea; do\n",
    "  if [ $ethnicity == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for ext in $group ${group}_minnesota_twins ${group}_minnesota_twins_RSQ; do\n",
    "  /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots \\\n",
    "    --script_prefix /share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/minnesota_twins.$ethnicity.1000G.CAT_FTND~1df.maf_gt_0.01_$ext.plots \\\n",
    "    --mem 15 \\\n",
    "    --priority 0 \\\n",
    "    --program /share/nas03/bioinformatics_group/software/R/dev/generate_gwas_plots.v6.R \\\n",
    "    --in /share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/minnesota_twins.$ethnicity.1000G.CAT_FTND~1df.maf_gt_0.01_$ext.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out /share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/minnesota_twins.$ethnicity.1000G.CAT_FTND~1df.maf_gt_0.01_$ext \\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda\n",
    "  done\n",
    "done\n",
    "\n",
    "for ethnicity in ea\n",
    "do\n",
    "  mv /share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/*.png \\\n",
    "   /share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/final\n",
    "done\n",
    "\n",
    "### END Generate plots ###\n",
    "\n",
    "\n",
    "\n",
    "### START Filter by p-value ###\n",
    "\n",
    "# MAF > 0.01 in AFR and EUR\n",
    "for ethnicity in ea\n",
    "do\n",
    "  if [ $ethnicity == \"aa\" ]\n",
    "  then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for ext in $group ${group}_minnesota_twins; do\n",
    "  outFile=/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/processing/minnesota_twins.$ethnicity.1000G.CAT_FTND~1df.maf_gt_0.01_$ext.p_lte_0.001\n",
    "  head -n 1 /share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/final/minnesota_twins.$ethnicity.1000G.chr1.CAT_FTND~1df.maf_gt_0.01_$ext > $outFile\n",
    "  for (( chr=1; chr<23; chr++ ))\n",
    "  do\n",
    "    inFile=/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/$ethnicity/final/minnesota_twins.$ethnicity.1000G.chr$chr.CAT_FTND~1df.maf_gt_0.01_$ext\n",
    "    echo Processing $inFile\n",
    "    tail -n +2 $inFile |\n",
    "      perl -lane 'if ($F[16] <= 0.001) { print; }' >> $outFile\n",
    "  done\n",
    "  done\n",
    "done\n",
    "\n",
    "# Sort\n",
    "R\n",
    "inData=read.table(\"/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/ea/processing/minnesota_twins.ea.1000G.CAT_FTND~1df.maf_gt_0.01_eur.p_lte_0.001\", header = TRUE)\n",
    "inData = inData[order(inData$PVALUE),]\n",
    "write.csv(inData, file=\"/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/ea/final/minnesota_twins.ea.1000G.CAT_FTND~1df.maf_gt_0.01_eur.p_lte_0.001.csv\", row.names = FALSE)\n",
    "inData=read.table(\"/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/ea/processing/minnesota_twins.ea.1000G.CAT_FTND~1df.maf_gt_0.01_eur_minnesota_twins.p_lte_0.001\", header = TRUE)\n",
    "inData = inData[order(inData$PVALUE),]\n",
    "write.csv(inData, file=\"/share/nas04/bioinformatics_group/data/studies/minnesota_twins/imputed/v1/association_tests/001/ea/final/minnesota_twins.ea.1000G.CAT_FTND~1df.maf_gt_0.01_eur_minnesota_twins.p_lte_0.001.csv\", row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Re-processing\n",
    "## Download Data From S3\n",
    "\n",
    "The results for version1 & version2 are located on AWS S3 at `rti-nd/Minnesota/FTND_1df_1000G_p3/20180402` & `rti-nd/Minnesota/FTND_1df_1000G_p3/20180611`, respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/bioinformatics/jmarks/nicotine/gwas/minnesota_twins/data/\n",
    "\n",
    "aws s3 sync s3://rti-nd/Minnesota/FTND_1df_1000G_p3/20180402 20180402 \n",
    "aws s3 sync s3://rti-nd/Minnesota/FTND_1df_1000G_p3/20180611 20180611\n",
    "\n",
    "#unzip data\n",
    "gzip -dr *\n",
    "\n",
    "# Check the count of the number of variants in both files\n",
    "tail -n +39 20180402/*c | wc -l\n",
    "\"\"\"9458862\"\"\"\n",
    "\n",
    "tail -n +39 20180611/*c | wc -l\n",
    "\"\"\"9458936\"\"\"\n",
    "\n",
    "# since there is a discrepancy between the two versions of data \n",
    "# (version2 has 74 more variants), I will need to take the intersection\n",
    "# of the two. We will see if we can identify which SNPs are uniq to each version.\n",
    "\n",
    "tail -n +39 20180402/*c | awk  'BEGIN{OFS=\":\"} {print $1,$2,$3,$4}' > chrom+pos+A1+A2.version1 &\n",
    "tail -n +39 20180611/*c | awk  'BEGIN{OFS=\":\"} {print $1,$2,$3,$4}' > chrom+pos+A1+A2.version2 &\n",
    "\n",
    "# lines uniq to version1\n",
    "comm -23 <(sort chrom+pos+A1+A2.version1) <(sort chrom+pos+A1+A2.version2) > 20180402/variants_exclusive_to_version1 &\n",
    "wc -l 20180402/variants_exclusive_to_version1\n",
    "\"\"\"2\"\"\"\n",
    "\n",
    "# lines uniq to version2\n",
    "comm -13 <(sort chrom+pos+A1+A2.version1) <(sort chrom+pos+A1+A2.version2) > 20180611/variants_exclusive_to_version2 &\n",
    "wc -l 20180611/variants_exlusive_to_version2\n",
    "\"\"\"76\"\"\"\n",
    "# lines that are in both\n",
    "mkdir 20180810\n",
    "comm -12 <(sort chrom+pos+A1+A2.version1) <(sort chrom+pos+A1+A2.version2) > 20180810/variants_in_both_versions &\n",
    "wc -l 20180810/variants_in_both_versions\n",
    "\"\"\"9458860 20180810/variants_in_both_versions\"\"\" # note this is what we would expect\n",
    "\n",
    "## now we just need to subset each file by the variants in the intersection\n",
    "#awk ' NR==FNR{map[$1]++;next} map[$1]' variants_in_both_versions test\n",
    "\n",
    "# get headers\n",
    "tail -n +38 20180402/*c | head -n1 > 20180810/version1_subset\n",
    "tail -n +38 20180611/*c | head -n1 > 20180810/version2_subset\n",
    "\n",
    "# subset files\n",
    "awk ' NR==FNR{map[$1]++} { if ( $1\":\"$2\":\"$3\":\"$4 in map) {print $0}}' 20180810/variants_in_both_versions\\\n",
    "    20180402/*c >> 20180810/version1_subset &\n",
    "wc -l 20180810/version1_subset\n",
    "\n",
    "awk ' NR==FNR{map[$1]++} { if ( $1\":\"$2\":\"$3\":\"$4 in map) print $0}' 20180810/variants_in_both_versions\\\n",
    "    20180611/*c  >>20180810/version2_subset &\n",
    "wc -l 20180810/version2_subset\n",
    "# double check that lines match\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
