{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1df FTND GWAS of Nelson Data\n",
    "\n",
    "**Author**: Jesse Marks\n",
    "\n",
    "**GitHub Issue**: [issue 96](https://github.com/RTIInternational/bioinformatics/issues/96#issuecomment-401903152)\n",
    "\n",
    "This document logs the steps taken to process:\n",
    "\n",
    "[*A Genome-Wide Association Study of Heroin Dependence*](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000277.v1.p1&phv=171252&phd=&pha=&pht=2797&phvf=&phdf=&phaf=&phtf=&dssp=1&consent=&temp=1) (Nelson data set) and perform an FTND GWAS to test for association with Nicotine Dependence (ND). This Nelson data set is a collaboration of Australian and American investigators aims to identify genes associated with liability for heroin dependence. There is FTND phenotype data included however, so we will perform a ND GWAS.\n",
    "\n",
    "The [FagerstrÃ¶m Test for Nicotine Dependence (FTND](https://cde.drugabuse.gov/instrument/d7c0b0f5-b865-e4de-e040-bb89ad43202b) is a standard instrument for assessing the intensity of physical addiction to nicotine.\n",
    "\n",
    "The data for this study are located on S3 at:\n",
    "\n",
    "`s3://rti-midas-data/studies/heroin_icghd/`\n",
    "\n",
    "**Note**: the imputed data for this cohort had to be reconverted. Originally, the genotype data were imputed with the impute2 software version 2.3.0. The script used to convert the `.gen` files created by impute2 to `.mldose` files, in order to be compatible with the rvtest GWAS software, was incompatible: the wrong column was reference to convert the MAF. As a result, the GWAS results were erroneous. There are currently two versions of the convert-script: one converts `.gen` files to `.mldose` format when the genotype data were imputed with a version of impute2 prior to version 2.3.2 and the other script converts the data when versions of impute2 subsequent to version 2.3.2 were used. On my EC2 instance on AWS, the scripts are located at:\n",
    "\n",
    "```\n",
    "/shared/bioinformatics/software/perl/file_conversion/convert_post-v2.3.2_impute2_files.pl\n",
    "/shared/bioinformatics/software/perl/file_conversion/convert_pre-v2.3.2_impute2_files.pl\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Nelson Data set has been downloaded to an EC2 instance from S3 to the location:\n",
    "`/shared/data/studies/Nelson`\n",
    "\n",
    "\n",
    "`/shared/data/studies/Nelson/UHS_heroin`\n",
    "\n",
    "The head of the data looks like the follow:\n",
    "\n",
    "`head imputed.aa_pop_pheno_controls_overlap.chr1.heroin_10EVs.palogist.stats`\n",
    "\n",
    "`name position A1 A2 Freq1 MAF Quality Rsq n Mean_predictor_allele beta_SNP_add sebeta_SNP_add loglik chi p or_95_percent_ci       chr\n",
    "rs58108140 10583 G A 0.923 0.077 0 0.471 3662 0.923191 0.00656649 0.133317 -2320.59 0.00242602632930075 0.960716312898352 1.01(0.78-1.31) 1\n",
    "`\n",
    "\n",
    "Where is the phenotype file for these data? What data do I need in order to perform a gwas? What I should do is look at my old notebooks to see how I performed the GWAS and then replicate this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Set autoscaling tags**\n",
    "\n",
    "This can be done in the AWS console."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aws autoscaling create-or-update-tags --tags \"ResourceId=cfncluster-JMspot2-ComputeFleet-IQXOI7AVONI2,ResourceType=auto-scaling-group,Key=project-number,Value=0215457.000.001,PropagateAtLaunch=true\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "aws s3 sync s3://rti-midas-data/studies/heroin_icghd/ /shared/data/studies/nelson --quiet &\n",
    "mkdir -p /shared/jmarks/nicotine/gwas/nelson/phenotype\n",
    "cp /shared/data/studies/nelson/phenotypes/probabel/heroin_icghd.ea.CAT_FTND.AGE.SEX.ALCOHOLDEP4.EVs.v1.gz \\\n",
    "    /shared/jmarks/nicotine/gwas/nelson/phenotype\n",
    "\n",
    "cd /shared/jmarks/nicotine/gwas/nelson/phenotype\n",
    "gunzip heroin_icghd.ea.CAT_FTND.AGE.SEX.ALCOHOLDEP4.EVs.v1.gz\n",
    "\n",
    "# only what to run a baseline gwas and therefore we will remove the alcohol covariate column\n",
    "awk '{print $1,$2,$3,$4,$6,$7,$8, $9}' heroin_icghd.ea.CAT_FTND.AGE.SEX.ALCOHOLDEP4.EVs.v1 > nelson.ea.CAT_FTND.AGE.SEX.EVs.v1\n",
    "\n",
    "mkdir -p ../association_test/001\n",
    "cd association_tests/001\n",
    "\n",
    "cp /shared/data/studies/nelson/imputed/v1/association_tests/003/_methods.heroin_icghd.v1.association_tests.003.sh \\\n",
    "    _methods.nelson.v1.association_tests.sh\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation Conversion\n",
    "\n",
    "There are some apparent issues with the `mldose` files. Specifically, the minor allele frequencies (MAF) are not lining up to what they should be. The issue was found after performing a baseline GWAS and finding that all of the results were NULL. After comparing the MAF of the `mldose` files and the `gen.gz` files, one can see that the MAF are in discordance. As an example, let us compare the MAF of SNP `rs16969968` (Mr. Big). \n",
    "\n",
    "```\n",
    "heroin_icghd.ea.1000G_p3.chr15.21.mach.mlinfo:\n",
    "SNP     Al1     Al2     Freq1   MAF     Quality Rsq\n",
    "rs16969968:78882925:G:A   G       A       0.002   0.002   0       0\n",
    "\n",
    " heroin_icghd.ea.1000G_p3.chr15.21.gen_info\n",
    "snp_id rs_id position exp_freq_a1 info certainty type info_type0 concord_type0 r2_type0\n",
    "--- rs16969968:78882925:G:A 78882925 0.349 0.996 0.998 0 -1 -1 -1\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#perl /shared/bioinformatics/software/perl/file_conversion/convert_post-v2.3.2_impute2_files.pl\\\n",
    "dataDir=/shared/data/studies/nelson/imputed/v1/imputations/ea/chr15/sandbox\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name impute_conversion_test \\\n",
    "    --script_prefix $dataDir/test2/mytest \\\n",
    "    --mem 15 \\\n",
    "    --nslots 4 \\\n",
    "    --priority 0 \\\n",
    "    --program perl /shared/bioinformatics/software/perl/file_conversion/convert_pre-v2.3.2_impute2_files.pl\\\n",
    "        --gen $dataDir/heroin_icghd.ea.1000G_p3.chr15.21.gen.gz \\\n",
    "        --sample $dataDir/heroin_icghd.ea.for_imputation.chr15.phased.sample.gz\\\n",
    "        --gen_info $dataDir/heroin_icghd.ea.1000G_p3.chr15.21.gen_info \\\n",
    "        --remove $dataDir/heroin_icghd.ea.1000G_p3.chr15.gen.remove\\\n",
    "        --out $dataDir/test2/heroin_icghd.ea.1000G_p3.chr15.21\\\n",
    "        --generate_mach_mldose_file\\\n",
    "        --generate_mach_mlinfo_file\\\n",
    "        --generate_legend_file \\\n",
    "        --variant_type all\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Outcome lines up\n",
    "Here are the results to the script ran above. It appears that the issue was in fact the results of an erroneous version of the conversion script initially applied.\n",
    "\n",
    "/shared/data/studies/nelson/imputed/v1/imputations/ea/chr15/sandbox/heroin_icghd.ea.1000G_p3.chr15.21.gen_info\n",
    "```\n",
    "snp_id rs_id position exp_freq_a1 info certainty type info_type0 concord_type0 r2_type0\n",
    "--- rs16969968:78882925:G:A 78882925 0.349 0.996 0.998 0 -1 -1 -1\n",
    "```\n",
    "\n",
    "/shared/data/studies/nelson/imputed/v1/imputations/ea/chr15/sandbox/test2/heroin_icghd.ea.1000G_p3.chr15.21.mach.mlinfo\n",
    "```\n",
    "SNP     Al1     Al2     Freq1   MAF     Quality Rsq\n",
    "rs16969968:78882925:G:A G       A       0.651   0.349   0       0.996\n",
    "```\n",
    "\n",
    "We can see that the MAF column is what we would expect now. We will now proceed in the next section with the reconversion of the imputed data. We also explain in more detail about what caused the issue."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reconvert all imputed data\n",
    "**Note**: the imputed data for this cohort had to be reconverted. Originally, the genotype data were imputed with the impute2 software version 2.3.0. The script used to convert the `.gen` files created by impute2 to `.mldose` files, in order to be compatible with the rvtest GWAS software, was incompatible: the wrong column was reference to convert the MAF. As a result, the GWAS results were erroneous. There are currently two versions of the convert-script: one converts `.gen` files to `.mldose` format when the genotype data were imputed with a version of impute2 prior to version 2.3.2 and the other script converts the data when versions of impute2 subsequent to version 2.3.2 were used. On my EC2 instance on AWS, the scripts are located at:\n",
    "\n",
    "```\n",
    "/shared/bioinformatics/software/perl/file_conversion/convert_post-v2.3.2_impute2_files.pl\n",
    "/shared/bioinformatics/software/perl/file_conversion/convert_pre-v2.3.2_impute2_files.pl\n",
    "```\n",
    "\n",
    "We will apply the conversion-script to the rest of the imputed data in `.gen` format to convert them to `.mach` format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseDir=/shared/data/studies/nelson/imputed/v1/imputations/ea\n",
    "sampleDir=/shared/data/studies/nelson/phased/v1/ea/phased_genotypes\n",
    "\n",
    "# First delete all of the old erroneous mach files\n",
    "rm $baseDir/chr*/*mach*\n",
    "\n",
    "# create directory structure\n",
    "for chr in {1..22};do\n",
    "    mkdir -p $baseDir/chr$chr/convert_to_mach\n",
    "done\n",
    "\n",
    "# submit jobs to convert the imputation data files ( gen --> mach )\n",
    "baseDir=/shared/data/studies/nelson/imputed/v1/imputations/ea\n",
    "sampleDir=/shared/data/studies/nelson/phased/v1/ea/phased_genotypes\n",
    "\n",
    "chr=15\n",
    "#for chr in {1..21};do\n",
    "    dataDir=/shared/data/studies/nelson/imputed/v1/imputations/ea/chr$chr\n",
    "    outDir=$dataDir/convert_to_mach \n",
    "    for chunk in $(awk -v a=$chr ' { if ( $1==a ) { print $2 } }' $baseDir/../final_chunks); do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name impute_conversion_test_chr$chr \\\n",
    "            --script_prefix $outDir/chr$chr.chunk$chunk.conversion \\\n",
    "            --mem 15 \\\n",
    "            --nslots 1 \\\n",
    "            --priority 0 \\\n",
    "            --program perl /shared/bioinformatics/software/perl/file_conversion/convert_pre-v2.3.2_impute2_files.pl \\\n",
    "                --gen $dataDir/heroin_icghd.ea.1000G_p3.chr$chr.$chunk.gen.gz \\\n",
    "                --sample $sampleDir/heroin_icghd.ea.for_imputation.chr$chr.phased.sample.gz \\\n",
    "                --gen_info $dataDir/heroin_icghd.ea.1000G_p3.chr$chr.$chunk.gen_info.gz \\\n",
    "                --remove $dataDir/heroin_icghd.ea.1000G_p3.chr$chr.gen.remove.gz \\\n",
    "                --out $outDir/heroin_icghd.ea.1000G_p3.chr$chr.$chunk \\\n",
    "                --generate_mach_mldose_file \\\n",
    "                --generate_mach_mlinfo_file \\\n",
    "                --generate_legend_file \\\n",
    "                --variant_type all\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start Prepare files for ProbABEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002\n",
    "imputationRoot=/shared/data/studies/nelson/imputed/v1/imputations\n",
    "dataDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/data\n",
    "phenotypePath=$dataDir/phenotype\n",
    "\n",
    "for (( chr=1; chr<23; chr++ )); do\n",
    "    mkdir -p $baseDir/ea/processing/chr$chr\n",
    "done\n",
    "mkdir $baseDir/ea/final\n",
    "\n",
    "\n",
    "# Phenotype file\n",
    "/shared/bioinformatics/software/perl/utilities/prepare_probabel_files.pl \\\n",
    "    --in_mldose $imputationRoot/ea/chr1/heroin_icghd.ea.1000G_p3.chr1.0.mach.mldose.gz \\\n",
    "    --in_pheno $phenotypePath/nelson.ea.CAT_FTND.AGE.SEX.EVs.v1 \\\n",
    "    --out_pheno $baseDir/phenotype/nelson.ea.CAT_FTND.AGE.SEX.EVs.v1\n",
    "\n",
    "# MLDOSE files\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=15\n",
    "    for chunk in $(grep -P \"^$chr\\s+\" $imputationRoot/final_chunks | perl -lane 'print $F[1];'); do\n",
    "        baseName=heroin_icghd.ea.1000G_p3.chr$chr.$chunk.mach.mldose.gz\n",
    "        newName=nelson.ea.1000G_p3.chr$chr.$chunk.mach.mldose\n",
    "        processingDir=$dataDir/genotype/imputed/chr$chr\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ea_${chr}_$chunk \\\n",
    "            --script_prefix $processingDir/$newName \\\n",
    "            --mem 15 \\\n",
    "            --nslots 3 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/perl/utilities/prepare_probabel_files.pl \\\n",
    "                    --in_mldose $imputationRoot/ea/chr$chr/convert_to_mach/$baseName \\\n",
    "                    --in_pheno $baseDir/phenotype/nelson.ea.CAT_FTND.AGE.SEX.EVs.v1 \\\n",
    "                    --out_mldose $processingDir/$newName\n",
    "\n",
    "        # copy the legend and mlinfo into this new imputation directory\n",
    "        cp $imputationRoot/ea/chr$chr/convert_to_mach/heroin_icghd.ea.1000G_p3.chr$chr.$chunk.legend.gz \\\n",
    "                $processingDir/nelson.ea.1000G_p3.chr$chr.$chunk.legend.gz\n",
    "        cp $imputationRoot/ea/chr$chr/convert_to_mach/heroin_icghd.ea.1000G_p3.chr$chr.$chunk.mach.mlinfo.gz \\\n",
    "                $processingDir/nelson.ea.1000G_p3.chr$chr.$chunk.mach.mlinfo.gz\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check for completion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check for completion\n",
    "dataDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/data\n",
    "for (( chr=1; chr<23; chr++ )); do\n",
    "    processingDir=$dataDir/genotype/imputed/chr$chr\n",
    "    for chunk in $(grep -P \"^$chr\\s+\" $imputationRoot/final_chunks | perl -lane 'print $F[1];'); do\n",
    "        newName=nelson.ea.1000G_p3.chr$chr.$chunk.mach.mldose\n",
    "        file=$processingDir/$newName.qsub.log\n",
    "        if [ -f $file ]; then\n",
    "            logLineCount=$(wc -l $file | perl -lane 'print $F[0];')\n",
    "            if [ $logLineCount -ne 4 ]; then\n",
    "                echo $file line count: $logLineCount\n",
    "            else\n",
    "                tail -n 1 $file |\n",
    "                perl -ne 'chomp; if (!/^Done/) { print \"'$file'\\n\".$_.\"\\n\"; }'\n",
    "            fi\n",
    "        else\n",
    "            echo $file missing\n",
    "        fi\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Start ProbABEL Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run palinear\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=15\n",
    "    for chunk in $(grep -P \"^$chr\\s+\" /shared/data/studies/nelson/imputed/v1/imputations/final_chunks | perl -lane 'print $F[1];'); do\n",
    "        newName=nelson.ea.1000G_p3.chr$chr.$chunk\n",
    "        imputationDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/data/genotype/imputed/chr$chr\n",
    "        baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002\n",
    "        processingDir=$baseDir/ea/processing/chr$chr\n",
    "\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ea_${chr}_$chunk \\\n",
    "            --script_prefix $processingDir/$newName.CAT_FTND~SNP+AGE+SEX+EVs \\\n",
    "            --mem 10 \\\n",
    "            --nslots 2 \\\n",
    "            --priority -1 \\\n",
    "            --program /shared/bioinformatics/software/third_party/probabel-0.5.0/bin/palinear \\\n",
    "                    --pheno $baseDir/phenotype/nelson.ea.CAT_FTND.AGE.SEX.EVs.v1 \\\n",
    "                    --dose $imputationDir/$newName.mach.mldose \\\n",
    "                    --info $imputationDir/$newName.mach.mlinfo \\\n",
    "                    --map $imputationDir/$newName.legend \\\n",
    "                    --chrom $chr \\\n",
    "                    --out $processingDir/$newName.CAT_FTND~SNP+AGE+SEX+EVs\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check for completion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for completion\n",
    "for (( chr=1; chr<23; chr++ )); do\n",
    "        baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002\n",
    "    for chunk in $(grep -P \"^$chr\\s+\" /shared/data/studies/nelson/imputed/v1/imputations/final_chunks | perl -lane 'print $F[1];'); do\n",
    "        newName=nelson.ea.1000G_p3.chr$chr.$chunk\n",
    "        file=$baseDir/ea/processing/chr$chr/$newName.CAT_FTND~SNP+AGE+SEX+EVs.qsub.log\n",
    "        if [ -f $file ]; then\n",
    "            logLineCount=$(wc -l $file | perl -lane 'print $F[0];')\n",
    "            if [ $logLineCount -eq 0 ]; then\n",
    "                echo $file empty\n",
    "            else\n",
    "                tail -n 1 $file |\n",
    "                    perl -ne 'chomp; if (!/^Analysis:/ || !/done$/) { print \"'$file'\\n\".$_.\"\\n\"; }'\n",
    "            fi\n",
    " else\n",
    "            echo $file missing\n",
    "        fi\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gzip dose files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/data/genotype/imputed\n",
    "for chr in {1..22};do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name zip.chr$chr \\\n",
    "        --script_prefix $dataDir/chr$chr.zip \\\n",
    "        --mem 15 \\\n",
    "        --nslots 2 \\\n",
    "        --priority 0 \\\n",
    "        --program bash gzip.script.sh $chr\n",
    "done\n",
    "\n",
    "cat gzip.script.sh\n",
    "#############################################################################################\n",
    "#!/bin/bash\n",
    "\n",
    "gzip /shared/bioinformatics/jmarks/nicotine/gwas/nelson/data/genotype/imputed/chr$1/*\n",
    "#############################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge ProbABEL results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=21\n",
    "    inFilePrefix=$baseDir/ea/processing/chr$chr/nelson.ea.1000G_p3.chr$chr.\n",
    "    inFileSuffix=.CAT_FTND~SNP+AGE+SEX+EVs_add.out.txt\n",
    "    outFile=$baseDir/ea/processing/chr$chr/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs\n",
    "    echo Chromosome $chr\n",
    "    head -n 1 ${inFilePrefix}1$inFileSuffix > $outFile\n",
    "    for chunk in $(grep -P \"^$chr\\s+\" /shared/data/studies/nelson/imputed/v1/imputations/final_chunks | perl -lane 'print $F[1];'); do\n",
    "        echo Merging $inFilePrefix$chunk$inFileSuffix\n",
    "        tail -n +2 $inFilePrefix$chunk$inFileSuffix >> $outFile\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate chi, P, and OR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/processing\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=21\n",
    "    bash /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ea_$chr \\\n",
    "        --script_prefix $baseDir/chr$chr/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.stats \\\n",
    "        --mem 15 \\\n",
    "        --nslots 3 \\\n",
    "        --priority 0 \\\n",
    "        --program Rscript /shared/bioinformatics/software/R/calculate_stats_for_probabel_results.R \\\n",
    "            --in_file $baseDir/chr$chr/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs \\\n",
    "            --out_file $baseDir/chr$chr/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.stats \\\n",
    "            --complete\n",
    "done\n",
    "\n",
    "baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/processing\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=21\n",
    "Rscript /shared/bioinformatics/software/R/calculate_stats_for_probabel_results.R \\\n",
    "            --in_file $baseDir/chr$chr/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs \\\n",
    "            --out_file $baseDir/chr$chr/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.stats \\\n",
    "            --complete &"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAF > 0.01 in Study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002\n",
    "\n",
    "\n",
    "# Filter MAF <= 0.01 in subjects\n",
    "#for ancestry in ea aa; do\n",
    "ancestry=ea\n",
    "    #for ((chr=1; chr<23; chr++));do\n",
    "chr=21\n",
    "        processingDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/processing/chr$chr\n",
    "        echo \"Processing chr${chr}_${ancestry}\"\n",
    "        head -n 1 $processingDir/nelson.$ancestry.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.stats > \\\n",
    "            $processingDir/nelson.$ancestry.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject.stats\n",
    "\n",
    "        # note column 7 corresponds to the MAF column\n",
    "        awk ' NR>=2 {if ($7 >= 0.01) {print $0}}' \\\n",
    "            $processingDir/nelson.$ancestry.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.stats \\\n",
    "            >> $processingDir/nelson.$ancestry.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject.stats\n",
    "    done &\n",
    "#done\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAF >0.01 in EUR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating a list of SNPs based off of 1000G population \n",
    "# - filter the variants to ones whose MAF <= 1%\n",
    "group=eur\n",
    "for chr in {1..22};do\n",
    "   awk ' { if ($14 >= 0.01) {print $1}}' <(zcat /shared/data/ref_panels/1000G/2013.12/impute2/ALL/ALL.chr$chr.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.nomono.unique_ids.legend.gz) >\\\n",
    "    /shared/data/ref_panels/1000G/2013.12/impute2/ALL/ALL.chr${chr}.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.nomono.unique_ids.maf_gt_0.01_${group}\n",
    "done &\n",
    "\n",
    "\n",
    "\n",
    "ancestry=ea\n",
    "base_dir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002\n",
    "#for ancestry in ea aa; do\n",
    "    if [ $ancestry == \"aa\" ]; then\n",
    "        group=afr\n",
    "    elif [ $ancestry == \"ea\" ]; then\n",
    "        group=eur\n",
    "    fi\n",
    "    for ((chr=1; chr<23; chr++)); do\n",
    "        base_dir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/processing/chr$chr\n",
    "        idList=/shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.legend.maf_lte_0.01_$group \n",
    "        #echo \"Processing chr${chr}_${ancestry}\"      \n",
    "        /shared/bioinformatics/software/perl/utilities/extract_rows.pl \\\n",
    "            --source $base_dir/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject.stats \\\n",
    "            --id_list $idList \\\n",
    "            --out $base_dir/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.stats \\\n",
    "            --header 1 \\\n",
    "            --id_column 0 \n",
    "    done &\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RSQ > 0.30 \n",
    "Filter by imputation quality - remove any variant whose \n",
    "$R^2 < 0.30$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "\n",
    "#for ancestry in ea aa; do\n",
    "ancestry=ea\n",
    "finalDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/final\n",
    "    for ((chr=1; chr<23; chr++));do\n",
    "        baseDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/processing/chr$chr\n",
    "        echo \"Processing chr${chr}_${ancestry}\"\n",
    "        head -n 1 $baseDir/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.stats > $finalDir/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ\n",
    "\n",
    "        awk ' NR>=2 {if ($9 >= 0.30) {print $0}}' \\\n",
    "            $baseDir/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.stats \\\n",
    "            >> $finalDir/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ\n",
    "    done\n",
    "#done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate QQ & Manhattan Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry=ea\n",
    " if [ ea == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  outFile=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/final/nelson.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ.table\n",
    "  #outFile=/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/processing/heroin_icghd.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_$group.table\n",
    "  echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "  for (( chr=1; chr<23; chr++ )); do\n",
    "  inFile=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/final/nelson.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ\n",
    "    echo Processing $inFile\n",
    "    tail -n +2 $inFile |\n",
    "      perl -lne '/^(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)(?:\\s+\\S+){10}\\s+(\\S+)/;\n",
    "                  if (($4 eq \"A\" || $4 eq \"C\" || $4 eq \"G\" || $4 eq \"T\") && ($5 eq \"A\" || $5 eq \"C\" || $5 eq \"G\" || $5 eq \"T\")) {\n",
    "                    print join(\"\\t\",$1,$2,$3,$6,\"snp\");\n",
    "                  } else {\n",
    "                    print join(\"\\t\",$1,$2,$3,$6,\"indel\");\n",
    "                  }' >> $outFile\n",
    "  done\n",
    "\n",
    "\n",
    "\n",
    "outDir=/shared/bioinformatics/jmarks/nicotine/gwas/nelson/association_tests/002/ea/final\n",
    "group=eur\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots \\\n",
    "    --script_prefix $outDir/nelson.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ.plots \\\n",
    "    --mem 7.5 \\\n",
    "    --priority 0 \\\n",
    "    --program Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "        --in  $outDir/nelson.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ.table \\\n",
    "        --in_chromosomes autosomal_nonPAR \\\n",
    "        --in_header \\\n",
    "        --out $outDir/nelson.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+EVs.maf_gt_0.01_subject+eur.RSQ.plots \\\n",
    "        --col_id VARIANT_ID \\\n",
    "        --col_chromosome CHR \\\n",
    "        --col_position POSITION \\\n",
    "        --col_p P \\\n",
    "        --col_variant_type TYPE \\\n",
    "        --generate_snp_indel_manhattan_plot \\\n",
    "        --manhattan_odd_chr_color red \\\n",
    "        --manhattan_even_chr_color blue \\\n",
    "        --manhattan_points_cex 1.5 \\\n",
    "        --generate_snp_indel_qq_plot \\\n",
    "        --qq_lines \\\n",
    "        --qq_points_bg black \\\n",
    "        --qq_lambda\n",
    "\n",
    "  mv /data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/processing/*.png \\\n",
    "   /data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/final\n",
    "\n",
    "### END Generate plots ###\n",
    "\n",
    "\n",
    "### START Filter by p-value ###\n",
    "\n",
    "# MAF > 0.01 in AFR and EUR\n",
    "  if [ ea == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  outFile=/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/processing/heroin_icghd.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_$group.p_lte_0.001\n",
    "  head -n 1 /data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/final/heroin_icghd.ea.1000G_p3.chr1.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_$group > $outFile\n",
    "  for (( chr=1; chr<23; chr++ )); do\n",
    "    inFile=/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/final/heroin_icghd.ea.1000G_p3.chr$chr.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_$group\n",
    "    echo Processing $inFile\n",
    "    tail -n +2 $inFile |\n",
    "      perl -lane 'if ($F[15] <= 0.001) { print; }' >> $outFile\n",
    "  done\n",
    "# Sort\n",
    "R\n",
    "inData=read.table(\"/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/aa/processing/heroin_icghd.aa.1000G_p3.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_afr.p_lte_0.001\", header = TRUE)\n",
    "inData = inData[order(inData$p),]\n",
    "write.csv(inData, file=\"/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/aa/final/heroin_icghd.aa.1000G_p3.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_afr.p_lte_0.001.csv\", row.names = FALSE)\n",
    "inData=read.table(\"/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/processing/heroin_icghd.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_eur.p_lte_0.001\", header = TRUE)\n",
    "inData = inData[order(inData$p),]\n",
    "write.csv(inData, file=\"/data/common/data/studies/heroin_icghd/imputed/v1/association_tests/003/ea/final/heroin_icghd.ea.1000G_p3.CAT_FTND~SNP+AGE+SEX+ALCOHOLDEP4+EVs.maf_gt_0.01_eur.p_lte_0.001.csv\", row.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scripts\n",
    "## calculate_stats_for_probabel_results.R\n",
    "\n",
    "* Note to replace the function `trim` with `trimws`.\n",
    "\n",
    "* replace `model = \"add\" with model = \"addA1\"`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/share/apps/R/bin/Rscript\n",
    "\n",
    "completeCases = FALSE\n",
    "fill = FALSE\n",
    "model = \"additive\"\n",
    "outputPvaluesOnly = FALSE\n",
    "outputChi2Only = FALSE\n",
    "chromosomes = c(1:22)\n",
    "\n",
    "args <- commandArgs(TRUE)\n",
    "\n",
    "loop = TRUE\n",
    "while (loop) {\n",
    "        if (args[1] == \"--in_file\") {\n",
    "                fileIn = args[2]\n",
    "        }\n",
    "\n",
    "        # This should be the file name for the probabel results with \"__CHR__\" inserted where the chr # should be\n",
    "        if (args[1] == \"--in_template\") {\n",
    "                fileInTemplate = args[2]\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--chromosomes\") {\n",
    "                if (args[2] == \"autosomal\") {\n",
    "                        chromosomes = c(1:22)\n",
    "                } else if (args[2] == \"autosomal_nonPAR\") {\n",
    "                        chromosomes = c(1:23)\n",
    "                }\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--out_file\") {\n",
    "                fileOut = args[2]\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--model\") {\n",
    "                model = args[2]\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--complete_cases\") {\n",
    "                completeCases = TRUE\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--fill\") {\n",
    "                fill = TRUE\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--output_pvalues_only\") {\n",
    "                outputPvaluesOnly = TRUE\n",
    "        }\n",
    "\n",
    "        if (args[1] == \"--output_chi2_only\") {\n",
    "                outputChi2Only = TRUE\n",
    "        }\n",
    "\n",
    "        if (length(args)>1) {\n",
    "                args<-args[2:length(args)]\n",
    "        } else {\n",
    "                loop = FALSE\n",
    "        }\n",
    "}\n",
    "\n",
    "\n",
    "if (exists(\"fileIn\")) {\n",
    "        cat(\"Reading \", fileIn, \"...\\n\", sep=\"\")\n",
    "        results = read.table(fileIn, header = TRUE, na.strings = c(\"nan\",\"NA\"), fill = fill)\n",
    "} else if (exists(\"fileInTemplate\")) {\n",
    "        for (chromosome in chromosomes) {\n",
    "                fileIn = gsub(\"__CHR__\",chromosome,fileInTemplate)\n",
    "                cat(\"Reading \", fileIn, \"...\\n\", sep=\"\")\n",
    "                nextResults = read.table(fileIn, header = TRUE, na.strings = c(\"nan\",\"NA\"), fill = fill)\n",
    "                if (exists(\"results\")) {\n",
    "                        results = rbind(results, nextResults)\n",
    "                } else {\n",
    "                        results = nextResults\n",
    "                }\n",
    "        }\n",
    "}\n",
    "\n",
    "cat(\"Calculating stats...\\n\")\n",
    "if (model == \"additive\") {\n",
    "        model = \"addA1\"\n",
    "} else if (model == \"dominant\") {\n",
    "        model = \"domA1\"\n",
    "}\n",
    "beta = paste(sep = \"\", \"beta_SNP_\",model)\n",
    "sebeta = paste(sep = \"\", \"sebeta_SNP_\",model)\n",
    "results$chi = (results[,beta]/results[,sebeta])^2\n",
    "results$p = pchisq(results$chi,df=1,lower=F)\n",
    "if (completeCases) {\n",
    "        results = results[complete.cases(results$p),]\n",
    "}\n",
    "\n",
    "if (outputPvaluesOnly) {\n",
    "        results = results[c(\"name\",\"p\")]\n",
    "} else if (outputChi2Only) {\n",
    "  results = results[c(\"name\",\"chi\")]\n",
    "} else {\n",
    "        if (nrow(results) > 0) {\n",
    "                results$significance[results$p >= 0.05] = \"\"\n",
    "                results$significance[results$p < 0.05] = \"*\"\n",
    "                results$significance[results$p < 0.01] = \"**\"\n",
    "                results$significance[results$p < 0.001] = \"***\"\n",
    "                results$or_95_percent_ci=paste(trim(formatC(exp(results[,beta]), format = \"f\", digits = 2)), \"(\", trim(formatC(exp(results[,beta] - (1.96 * results[,sebeta])), format = \"f\", digits = 2)), \"-\", trim(formatC(exp(results[,beta] + (1.96 * results[,sebeta])), format = \"f\", digits = 2)), \")\", results$significance, sep = \"\")\n",
    "                results$significance = NULL\n",
    "        }\n",
    "        if (\"chrom\" %in% colnames(results) & \"position\" %in% colnames(results)) {\n",
    "                results = cbind(subset(results,select=c(name,chrom,position)),subset(results,select=-c(name,chrom,position)))\n",
    "        } else if (\"chrom\" %in% colnames(results)) {\n",
    "                results = cbind(subset(results,select=c(name,chrom)),subset(results,select=-c(name,chrom)))\n",
    "        } else if (\"position\" %in% colnames(results)) {\n",
    "                results = cbind(subset(results,select=c(name,position)),subset(results,select=-c(name,position)))\n",
    "        }\n",
    "}\n",
    "\n",
    "cat(\"Writing \", fileOut, \"...\\n\", sep=\"\")\n",
    "write.table(results, file = fileOut, row.names = FALSE, quote=FALSE, sep=\"\\t\")\n",
    "\n",
    "cat(\"Done\\n\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nextflow implementation (not used)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/data/0212964/nextflow/nextflow-0.25.1-all\n",
    "\n",
    "/* ######################################################################\n",
    "#                 Association Pipeline v0.1                             #\n",
    "###################################################################### */\n",
    "\n",
    "/*\n",
    " * Defines pipeline parameters\n",
    " */\n",
    "\n",
    "params.final_chunks = \"The final chunking file splited by chromosome\"\n",
    "params.input_pheno = \"The phenotype to be used in this analyses, file generated in ESN Windows environment\"\n",
    "params.imputation_dir = \"The directory contains the post imputation genotype files (mldose files)\"\n",
    "params.example_mldose = \"One of the mldose files in the imputation directory\"\n",
    "params.geno_prefix = \"The prefix of the mldose file\"\n",
    "params.working_dirs = \"The working directory\"\n",
    "params.out = \"The output file name for the stats file\"\n",
    "params.method = \"palinear or palogist\"\n",
    "\n",
    "/*\n",
    " * probabel_header = \"name\\tchrom\\tposition\\tA1\\tA2\\tFreq1\\tMAF\\tQuality\\tRsq\\tn\\tMean_predictor_allele\\tbeta_SNP_add\\tsebeta_SNP_add\\tchi2_SNP\\tchi\\tp\\tor_95_percent_ci\"\n",
    " */\n",
    "\n",
    "// create chunks channels from chunking file\n",
    "chunks = Channel\n",
    "                        .from( file(params.final_chunks) )\n",
    "                        .splitCsv(header:['chr', 'chunk', 'start', 'end'], skip: 1, sep: '\\t')\n",
    "chunks.into { chunks_prep_geno; chunks_merge; }\n",
    "\n",
    "/* *********************************************\n",
    " * Step 1: Start Prepare ProbABEL Phenotype File\n",
    " */\n",
    "\n",
    "process prepare_pheno{\n",
    " input:\n",
    "        // input defined by parameters\n",
    "\n",
    "        output:\n",
    "        file \"probabel_pheno\" into probabel_phenotype_file\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/perl/prepare_probabel_files.pl \\\n",
    "                --in_mldose ${params.example_mldose} \\\n",
    "                --in_pheno ${params.input_pheno} \\\n",
    "                --out_pheno \"probabel_pheno\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* ********************************************\n",
    " * Step 2: Start Prepare ProbABEL Genotype File\n",
    " */\n",
    "\n",
    "process prepare_geno{\n",
    "\n",
    "        input:\n",
    "        file \"probabel_pheno\" from probabel_phenotype_file\n",
    "        set chr, chunk, start, end from chunks_prep_geno\n",
    "\n",
    "        output:\n",
    "        file \"${params.geno_prefix}${chr}.${chunk}.mach_mldose\" into probabel_genotype_files\n",
    "\n",
    "    executor 'sge'\n",
    "    clusterOptions '-S /bin/bash -l mem_free=15G,h_vmem=15G -p 0'\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/perl/prepare_probabel_files.pl \\\n",
    "                --in_mldose \"${params.imputation_dir}/${params.geno_prefix}${chr}.${chunk}.mach.mldose.gz\" \\\n",
    "                --in_pheno \"${probabel_pheno}\" \\\n",
    "                --out_mldose \"${params.geno_prefix}${chr}.${chunk}.mach_mldose\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* *******************************\n",
    " * Step 3: Start ProbABEL Analysis\n",
    " */\n",
    "\n",
    "process do_probabel_analysis{\n",
    "\n",
    "        input:\n",
    "        file probabel_pheno from probabel_phenotype_file\n",
    "        file geno from probabel_genotype_files\n",
    "\n",
    "        output:\n",
    "        file \"${geno.baseName}_add.out.txt\" into probabel_out_files\n",
    "\n",
    "        executor 'sge'\n",
    "    clusterOptions '-S /bin/bash -l mem_free=7G,h_vmem=7G -p 0'\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/${params.method} \\\n",
    "        --pheno $probabel_pheno \\\n",
    "        --dose $geno \\\n",
    "        --info \"${params.imputation_dir}/${geno.baseName}.mach.mlinfo\" \\\n",
    "        --map \"${params.imputation_dir}/${geno.baseName}.legend\" \\\n",
    "        --out \"${geno.baseName}\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* ***************************\n",
    " * Step 4: Start calc chi p OR\n",
    " */\n",
    "process calc_chi_p_or{\n",
    "\n",
    "        input:\n",
    "        file result from probabel_out_files\n",
    "\n",
    "        output:\n",
    "        file \"${result.baseName}.stats\" into stats_files\n",
    "\n",
    "        executor 'sge'\n",
    "    clusterOptions '-S /bin/bash -l mem_free=7G,h_vmem=7G -p 0'\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/R/calculate_stats_for_probabel_results_v2.R \\\n",
    "                --remove_missing_p \\\n",
    "                --in_file \"${result}\" \\\n",
    "                --out_file \"${result.baseName}.stats\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* ***************************\n",
    " * Step 5: merge stats files\n",
    " */\n",
    "\n",
    "stats_files\n",
    "        .collectFile(name: file(params.out), skip: 1)\n",
    "        .println {\"Result saved to file: $it\"}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/data/studies/nelson/imputed/v1/imputations/ea\n",
    "\n",
    "aws s3 sync . s3://rti-nd/nelson/data/imputed/ea/ --quiet &\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
