{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VIDUS  Imputation Preparation\n",
    "__Author__: Jesse Marks\n",
    "**Date**: August 31, 2018\n",
    "\n",
    "This document logs the steps taken to perform phasing and imputation on the dataset [VIDUS](https://www.bccsu.ca/vidus/). The starting point for this analysis is after quality control of observed genotypes. The quality controlled genotypes are oriented on the GRCh37 plus strand. \n",
    "\n",
    "## Software and tools\n",
    "The software and tools used for porcessing these data are\n",
    "* [Michigan Imputation Server](https://imputationserver.sph.umich.edu/index.html) (MIS)\n",
    "* [Amazon Web Services (AWS) - Cloud Computing Services](https://aws.amazon.com/)\n",
    "    * Linux AMI\n",
    "* [PLINK v1.90 beta 4.10](https://www.cog-genomics.org/plink/)\n",
    "* [bgzip](http://www.htslib.org/doc/tabix.html)\n",
    "* [BCF Tools](http://www.htslib.org/doc/bcftools.html)\n",
    "* Windows 10 with [Cygwin](https://cygwin.com/) installed\n",
    "* GNU bash version 4.2.46\n",
    "\n",
    "## Data retrieval and organization\n",
    "\n",
    "The genotype data (QC'd) were retrieved from AWS S3 at `s3://rti-heroin/ngc_vidus_fou/data/genotype/original/ea/`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data processing\n",
    "### GRCh37 strand and allele discordance check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "base_dir=/shared/bioinformatics/jmarks/Heroin/Vidus\n",
    "mkdir -p $base_dir/genotype/original\n",
    "aws s3 sync s3://rti-heroin/ngc_vidus_fou/data/genotype/original/ea/ $base_dir/genotype/original\n",
    "\n",
    "mkdir $base_dir/1000g\n",
    "ancestry=\"ea\"\n",
    "geno_dir=$base_dir/genotype/original\n",
    "\n",
    "# write out the MAF report\n",
    "for ancestry in ea;do\n",
    "    for study in vidus; do\n",
    "        mkdir $base_dir/strand_check\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bfile $geno_dir/${ancestry}_chr_all \\\n",
    "            --freq \\\n",
    "            --out $base_dir/strand_check/${ancestry}_chr_all\n",
    "    done\n",
    "done\n",
    "\n",
    "# Get list of variants from all studies\n",
    "for ancestry in ea; do\n",
    "    cat $geno_dir/${ancestry}_chr_all.bim | \\\n",
    "            perl -lane 'if (($F[0]+0) <= 23) { print $F[1]; }' | \\\n",
    "            sort -u > $base_dir/${ancestry}_chr_all_sorted_variants.txt\n",
    "done\n",
    "\n",
    "# Calculate autosome and chrX MAFs for 1000G EUR\n",
    "pop=\"EUR\"\n",
    "ancestry=\"ea\"\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${pop}_${chr} \\\n",
    "        --script_prefix ${base_dir}/${pop}_chr${chr}.maf \\\n",
    "        --mem 6.8 \\\n",
    "        --nslots 1 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "            --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "            --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "            --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "            --chr ${chr} \\\n",
    "            --out ${base_dir}/1000g/${pop}_chr${chr}.maf \\\n",
    "            --extract ${ancestry}_chr_all_sorted_variants.txt \\\n",
    "            --keep_groups ${pop}\n",
    "done\n",
    "\n",
    "chr=23\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${pop}_${chr} \\\n",
    "    --script_prefix ${base_dir}/${pop}_chr${chr}.maf \\\n",
    "    --mem 6.8 \\\n",
    "    --nslots 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "        --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.hap.gz\\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "        --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "        --chr ${chr} \\\n",
    "        --out ${base_dir}/1000g/${pop}_chr${chr}.maf \\\n",
    "        --extract ${ancestry}_chr_all_sorted_variants.txt \\\n",
    "        --keep_groups ${pop}\n",
    "\n",
    "\n",
    "# Merge per chr MAFs for EUR\n",
    "pop=\"EUR\"\n",
    "head -n 1 1000g/${pop}_chr1.maf > 1000g/${pop}_chr_all.maf\n",
    "tail -q -n +2 1000g/${pop}_chr{1..23}.maf \\\n",
    "    >> 1000g/${pop}_chr_all.maf\n",
    "\n",
    "study=vidus\n",
    "# Run discordance checks for EA groups\n",
    "for ancestry in ea; do\n",
    "    if [ $ancestry = \"ea\" ]; then\n",
    "        pop=EUR\n",
    "    else\n",
    "        pop=AFR\n",
    "    fi\n",
    "\n",
    "#    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "#        --job_name ${ancestry}_${study}_crosscheck \\\n",
    "#        --script_prefix ${base_dir}/strand_check/${ancestry}_allele_discordance_check \\\n",
    "#        --mem 6 \\\n",
    "#        --nslots 4 \\\n",
    "#        --priority 0 \\\n",
    "#        --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "#            --study_bim_file ${geno_dir}/${ancestry}_chr_all.bim\n",
    "#            --study_frq_file ${base_dir}/strand_check/${ancestry}_chr_all.frq\n",
    "#            --ref_maf_file ${base_dir}/1000g/${pop}_chr_all.maf\n",
    "#            --out_prefix ${base_dir}/strand_check/${ancestry}_allele_discordance\"\n",
    "\n",
    "    # just checking chr23 here\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${ancestry}_${study}_crosscheck \\\n",
    "        --script_prefix ${base_dir}/strand_check/${ancestry}_allele_discordance_check \\\n",
    "        --mem 6.8 \\\n",
    "        --nslots 2 \\\n",
    "        --priority 0 \\\n",
    "        --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "            --study_bim_file ${geno_dir}/${ancestry}_chr_all.bim\n",
    "            --study_frq_file ${base_dir}/strand_check/${ancestry}_chr_all.frq\n",
    "            --ref_maf_file ${base_dir}/1000g/${pop}_chr23.maf\n",
    "            --out_prefix ${base_dir}/strand_check/${ancestry}_allele_discordance\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resolving allele discordances\n",
    "The allele discordances will be resolved by\n",
    "* Flipping allele discordances that are fixed by flipping\n",
    "* Removing SNPs with discordant names\n",
    "* Removing SNPs with discordant positions\n",
    "* Removing allele discordances that are not resolved by flipping\n",
    "* Removing alleles with large deviations from the reference population allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "base_dir=/shared/bioinformatics/jmarks/Heroin/Vidus\n",
    "geno_dir=${base_dir}/genotype/original\n",
    "\n",
    "study=vidus\n",
    "\n",
    "# Apply filters\n",
    "for ancestry in ea;do\n",
    "    echo -e \"\\n===============\\nProcessing ${study}\\n\"\n",
    "    # Create remove list\n",
    "    echo \"Making remove list\"\n",
    "    cat <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2) \\\n",
    "        <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.at_cg_snps_freq_diff_gt_0.2 | tail -n +2) \\\n",
    "        <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.discordant_names | tail -n +2) \\\n",
    "        <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.discordant_positions | tail -n +2) \\\n",
    "        <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip | tail -n +2) | \\\n",
    "        sort -u > ${base_dir}/strand_check/${ancestry}_snps.remove\n",
    "\n",
    "    # Create flip list\n",
    "    echo \"Making flip list\"\n",
    "    comm -23 <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.discordant_alleles | tail -n +2 | sort -u) \\\n",
    "        <(cut -f2,2 ${base_dir}/strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2 | sort -u) \\\n",
    "        > ${base_dir}/strand_check/${ancestry}_snps.flip\n",
    "\n",
    "    # Apply filters\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile $geno_dir/${ancestry}_chr_all \\\n",
    "        --exclude $base_dir/strand_check/${ancestry}_snps.remove \\\n",
    "        --flip $base_dir/strand_check/${ancestry}_snps.flip \\\n",
    "        --make-bed \\\n",
    "        --out ${geno_dir}/${ancestry}_filtered\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove monomorphic variants\n",
    "Monomorphic variants prevent MIS from accepting the genotype data. In this case, an arbitrarily small MAF is set that is smaller than the lower bound for these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "base_dir=/shared/bioinformatics/jmarks/Heroin/Vidus\n",
    "geno_dir=$base_dir/genotype/original\n",
    "ancestry=\"ea\"\n",
    "\n",
    "# Apply filters\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile $geno_dir/${ancestry}_filtered \\\n",
    "    --maf 0.000001 \\\n",
    "    --make-bed \\\n",
    "    --out $geno_dir/${ancestry}_filtered_mono"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation preparation for Michigan Imputation Server\n",
    "Visit the [MIS Getting Started Webpage](https://imputationserver.sph.umich.edu/start.html#!pages/help) for more information about the preparing the data for upload to MIS.\n",
    "### VCF File Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "base_dir=/shared/bioinformatics/jmarks/Heroin/Vidus\n",
    "geno_dir=$base_dir/genotype/original\n",
    "ancestry=\"ea\"\n",
    "\n",
    "mkdir $base_dir/phase_prep\n",
    "\n",
    "# Split by chr and remove any individuals with missing data for whole chr\n",
    "#for chr in {1..23}; do\n",
    "for chr in 23; do\n",
    "    # Remove SNPs\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 4000 \\\n",
    "        --bfile ${geno_dir}/${ancestry}_filtered_mono \\\n",
    "        --chr ${chr} \\\n",
    "        --mind 0.99 \\\n",
    "        --make-bed \\\n",
    "        --out ${base_dir}/phase_prep/${ancestry}_chr${chr}_for_phasing \n",
    "done > chr_splitting.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Note__: No subjects were removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "phase_dir=/shared/bioinformatics/jmarks/Heroin/Vidus/phase_prep\n",
    "ancestry=\"ea\"\n",
    "\n",
    "mkdir ea \n",
    "\n",
    "chr=23\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 5000 \\\n",
    "    --bfile ${phase_dir}/${ancestry}_chr${chr}_for_phasing \\\n",
    "    --output-chr M \\\n",
    "    --set-hh-missing \\\n",
    "    --recode vcf bgz \\\n",
    "    --out ${phase_dir}/${ancestry}/${ancestry}_chr${chr}_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to Michigan Imputation Server (MIS)\n",
    "Transfer the \\*.vcf files to local machine (per chromosome) and then upload to MIS.\n",
    "\n",
    "### Uploading parameters EA\n",
    "These are the parameters that were selected on MIS.\n",
    "\n",
    "__Name__: vidus_ea_23\n",
    "\n",
    "__Reference Panel__ 1000G Phase 3 v5\n",
    "\n",
    "__Input Files__ File Upload <br>\n",
    "\n",
    "* Select Files - select VCF files that were downloaded to local machine from cloud. <br>\n",
    "\n",
    "__Phasing__: ShapeIT v2.r790 (unphased) \n",
    "\n",
    "__Population__: EUR\n",
    "\n",
    "__Mode__: Quality Control & Imputation\n",
    "\n",
    "* I will not attempt to re-identify or contact research participants.\n",
    "* I will report any inadvertent data release, security breach or other data management incident of which I become aware."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Imputed Data from MIS\n",
    "First Download the data form the Michigan Imputation Server by clicking on the link provided in the email they send out to alert you that your data has finished. Here you will find commands for downloading the data.\n",
    "## EA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "mkdir -p /shared/impute/kreek/data/genotype/imputed/{aa,ea}\n",
    "cd /shared/impute/kreek/data/genotype/imputed/ea\n",
    "\n",
    "# I put all of the download commands in a bash script so I could submit them as jobs.\n",
    "cat mywget.sh\n",
    "\"\"\"\n",
    "wget https://imputationserver.sph.umich.edu/share/results/2e899483c134f1d2c08e3aabc4414253/chr_1.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/ed48286121726c3bc64ddf024ec72730/chr_10.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/1472b06d90055f4effca30c4924fbeca/chr_11.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/b3e1f5e3fd0bfd0a69fe171e438d3c53/chr_12.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/8bc2b433fed05c8c57e09c240aae9d38/chr_13.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/ed5623de7f6b337dec56abce51908cf5/chr_14.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/55b667fafaa4881bba5c285c859b6d6e/chr_15.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/7744e3469a60729f0cfabef804cb7cf/chr_16.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/3097bb774de0115ed454e48628ad06dc/chr_17.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/761df4b95b0214e19d321cd327e5ac4e/chr_18.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/fd2b2ee5136093e255bdf60f7d575bf8/chr_19.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/dffef070d2e0da537b7b0d2b7c42d5e1/chr_2.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/76d3e3a2e8938a080fda54c80cbbe264/chr_20.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/27e8da7066c8dfe52c890eb1952e6dcd/chr_21.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/f1f5a8bbdbce2b2293081f72b7c5ba6c/chr_22.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/cd7e1fe0b57b45590ec83367bf33ffae/chr_3.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/f405efbacb10d9243aaf3d662c6f9414/chr_4.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/ccba882630a8cce30bf458dca68a2d09/chr_5.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/77aeabce5acbca6e225e3ac02965eb0d/chr_6.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/b7798d33eca337160fb28db9d855e6e0/chr_7.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/f95778ef2f6cb1364e5f272e0b0fcf75/chr_8.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/494429e58125aacc323dce6366f348be/chr_9.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/877e7a71f369a30acf5933880b5edbda/chr_X.no.auto_female.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/1c9f7d9ac07591172ecb01e76b743fd/chr_X.no.auto_male.zip \n",
    "\"\"\"\n",
    "\n",
    "sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ea.download_kreek \\\n",
    "    --script_prefix kreek_imputation_download \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash mywget.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inflate imputation results\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EC2 console ###\n",
    "cd /shared/imputed/kreek/data/genotype/imputed/ea\n",
    "\n",
    "# inflate chr results\n",
    "for f in {1..23};do\n",
    "echo '#!/bin/bash' > chr_$f.sh\n",
    "echo '' >> chr_$f.sh\n",
    "echo 'unzip -P \"\\aKcPO5MYw6qr\" chr_'$f'.zip' >> chr_$f.sh\n",
    "done\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ea.inflate_chr${chr} \\\n",
    "    --script_prefix ea.chr${chr}_inflation \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash chr_${chr}.sh\n",
    "done\n",
    "\n",
    "rm *.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to S3\n",
    "Uploaded to:\n",
    "\n",
    "`s3://rti-heroin/kreek/data/genotype/imputed`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
