{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Procedures for downloading data from dbGaP and moving to S3\n",
    "** Author: Jesse Marks **\n",
    "\n",
    "Note: be sure to remove studies from local machine once they have been uploaded to s3. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Directory structure for S3"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "   rti-common\n",
    "    |\n",
    "    |-- dbgap_studies\n",
    "    |    |\n",
    "    |     ` -- <dbGaP study name>\n",
    "    |    |    |\n",
    "    |    |     ` -- meta\n",
    "    |    |    |\n",
    "    |    |     ` -- genotype\n",
    "    |    |    |    |    \n",
    "    |    |    |     ` -- original\n",
    "    |    |    |    |     |    \n",
    "    |    |    |    |      ` -- unprocessed\n",
    "    |    |    |    |     |    \n",
    "    |    |    |    |      ` -- processing\n",
    "    |    |    |    |     |    \n",
    "    |    |    |    |      ` -- final\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- imputed\n",
    "    |    |    |\n",
    "    |    |     ` -- phenotype\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- unprocessed\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- processing\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- final "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Install Amazon Web Services Command Line Interface (AWS CLI) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Install pip, a package manager for python applications on local machine\n",
    "curl \"https://bootstrap.pypa.io/get-pip.py\" -o \"get-pip.py\"\n",
    "python get-pip.py\n",
    "\n",
    "# Install awscli via pip\n",
    "pip install awscli\n",
    "\n",
    "# Verify installation - should see something similar to the string below\n",
    "aws --version\n",
    "\"aws-cli/1.11.178 Python/2.7.13 CYGWIN_NT-10.0/2.9.0(0.318/5/3) botocore/1.7.36\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configure AWS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Local Machine\n",
    "aws configure\n",
    "\n",
    "AWS Access Key ID [None]: AKIAJONBCJHOJSW2PFJA\n",
    "AWS Secret Access Key [None]:  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\n",
    "Default region name [None]: us-east-1\n",
    "Default output format [None]: text  # could be json, text, or table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Population Controls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000206_pancreatic \n",
    "\n",
    "Genome-wide association study of heroin abuse using public controls\n",
    "CGEMS Pancreatic Cancer (PanScan) (phs000206.v5.p3)\n",
    "\n",
    "* Download 58549"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Local machine\n",
    "\n",
    "# Decrypt (must be executed within project dir) \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58549\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58527/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure (on local)\n",
    "mkdir -p s3/phs000206_pancreatic/{genotype,phenotype,meta}\n",
    "mkdir -p s3/phs000206_pancreatic/genotype/original/{final,processing,unprocessed}\n",
    "mkdir s3/phs000206_pancreatic/phenotype/{final,processing,unprocessed}\n",
    "\n",
    "# Move files/data to appropriate \"unprocessed\" folders\n",
    "cd /58549/PhenoGenotypeFiles/RootStudyConsentSet_phs000206.CGEMS_PanScan.v5.p3.c1.CADM/\n",
    "mv /PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/phenotype/unprocessed/\n",
    "mv /GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/genotype/original/unprocessed/\n",
    "mv /Study\\ MetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/meta/\n",
    "\n",
    "# Start processing the files\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/\n",
    "cp phenotype/unprocessed/* phenotype/processing\n",
    "cp genotype/original/unprocessed/* genotype/original/processing/\n",
    "\n",
    "# unzip and rename\n",
    "cd phenotype/processing/\n",
    "for i in *.gz; do gunzip $i; done\n",
    "for i in *.gz; do mv $i $(echo $i|perl -pi -e s/phs0.+CAN_//g); done\n",
    "for i in *.xml; do mv $i $(echo $i|perl -pi -e s/phs000.+CAN_//g); done\n",
    "\n",
    "$ cd ../../genotype/original/processing/\n",
    "$ for i in *.gz; do tar -xzvf $i; done\n",
    "$ rm *.gz\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "$ aws s3 mv phs000206_pancreatic/ s3://rti-common/dbGaP/phs000206_pancreatic/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000199_chop-ctrl\n",
    "CHOP Control CNV Study (phs000199.v1.p1)\n",
    "* Download 58599"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Local machine\n",
    "\n",
    "# Decrypt (must be executed within project dir) \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58599\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58599/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000199_chop-ctrl/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000199_chop-ctrl/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000199_chop-ctrl/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "\n",
    "# Move files/data to appropriate \"unprocessed\" folders\n",
    "$ /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58599/NIGMS/CHOP_CNV/phs000199v1/p1/\n",
    "$ for i in *.pdf; do mv $i /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/meta/; done\n",
    "$ for i in phs*; do mv $i /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/phenotype/unprocessed/; done\n",
    "$ for i in phg000059v1/*; do mv $i /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/genotype/original/unprocessed/; done\n",
    "\n",
    "# unzip & rename genotype\n",
    "$ cd genotype/original/\n",
    "$ cp unprocessed/ processing/\n",
    "$ for i in processing/*.gz; do tar -xzvf $i; done\n",
    "$ rm *.gz\n",
    "$ mv phg000059.v1.p1.CHOP_CNV.individual-info.MULTI.txt individual-info.MULTI.txt\n",
    "\n",
    "# unzip & rename phenotype\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/phenotype\n",
    "$ cp unprocessed/* processing/\n",
    "$ cd processing/\n",
    "$ for i in *.txt.gz; do gunzip $i; done\n",
    "$ for i in *.gz; do tar -xzvf $i; done$ \n",
    "$ rm *.gz\n",
    "$ for i in *; do mv $i $(echo $i | perl -pi -e s/phs0.+CNV_//g); done\n",
    "\n",
    "# upload to S3\n",
    "$ aws s3 mv phs000199_chop-ctrl s3://rti-common/dbGaP/phs000199_chop-ctrl/ --recursive\n",
    "$ rm -rf phs000199_chop-ctrl/*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000209_MESA-c1\n",
    "MESA Cohort (phs000209.v13.p3.c1)\n",
    "* Download 58601"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Decrypt (must be executed within project dir) \n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58601\n",
    "\n",
    "\n",
    "$ if [ $(find 58601/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "$ mkdir -p s3/phs000209_MESA/{meta,genotype,phenotype}\n",
    "$ mkdir -p s3/phs000209_MESA/genotype/original/{final,processing,unprocessed}\n",
    "$ mkdir -p s3/phs000209_MESA/phenotype/{final,processing,unprocessed}\n",
    "\n",
    "# mv files to appropriate directories\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58601/PhenoGenotypeFiles/\\\n",
    "  RootStudyConsentSet_phs000209.MESA.v13.p3.c1.HMB/\n",
    "$ mv StudyFiles* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/meta/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "\n",
    "$ cd ../ChildStudyConsentSet_phs000420.MESA.v6.p3.c1.HMB/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/genotype/original/unprocessed\n",
    "\n",
    "$ cd ../ChildStudyConsentSet_phs000403.MESA.v3.p3.c1.HMB/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/genotype/original/unprocessed\n",
    "\n",
    "$ cd ../ChildStudyConsentSet_phs000283.MESA.v7.p3.c1.HMB/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/genotype/original/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "$ aws s3 mv phs000209_MESA/ s3://rti-common/dbGaP/phs000209_MESA/ --recursive\n",
    "    \n",
    "    \n",
    "## NOTE: had to rename this study to \"phs000209_MESA-c1\" from \"phs000209_MESA\"\n",
    "## here is the command to perform this action\n",
    "$ aws s3 cp s3://rti-common/dbGaP/phs000209_MESA/ s3://rti-common/dbGaP/phs000209_MESA-c1/ --recursive\n",
    "$ aws s3 rm s3://rti-common/dbGaP/phs000209_MESA/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000209_MESA-c2\n",
    "\n",
    "Multi-Ethnic Study of Atherosclerosis (MESA) Cohort \n",
    "(phs000209.v13.p3.c2)\n",
    "\n",
    "* Download 58619"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Decrypt (must be executed within project dir) \n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58619\n",
    "\n",
    "$ if [ $(find 58619/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "$ mkdir -p s3/phs000209_MESA-c2/{meta,genotype,phenotype}\n",
    "$ mkdir -p s3/phs000209_MESA-c2/genotype/original/{final,processing,unprocessed}\n",
    "$ mkdir -p s3/phs000209_MESA-c2/phenotype/{final,processing,unprocessed}\n",
    "\n",
    "# mv files to appropriate directories\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58619/PhenoGenotypeFiles/\\\n",
    "  RootStudyConsentSet_phs000209.MESA.v13.p3.c2.MB/\n",
    "$ mv StudyFiles* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/meta/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "\n",
    "$ cd ../ChildStudyConsentSet_phs000420.MESA.v6.p3.c2.HMB-NPU/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/genotype/original/unprocessed\n",
    "\n",
    "$ cd ../ChildStudyConsentSet_phs000403.MESA.v3.p3.c2.HMB-NPU/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/genotype/original/unprocessed\n",
    "\n",
    "$ cd ../ChildStudyConsentSet_phs000283.MESA.v7.p3.c2.HMB-NPU/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/genotype/original/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3\n",
    "$ aws s3 cp phs000209_MESA-c2/ s3://rti-common/dbGaP/phs000209_MESA-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000169_visc-adiposity\n",
    "Whole Genome Association Study of Visceral Adiposity in the HABC Study(phs000169.v1.p1)\n",
    "* Download 58623"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58623\n",
    "\n",
    "# check for successful decryption\n",
    "$ if [ $(find 58601/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# resolve problematic decryption\n",
    "for i in $(find ./58623/ -name *ncbi_enc -print)\n",
    "do      \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\"\n",
    "mv ${i} ${j}_tmp.ncbi_enc\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\"\n",
    "mv ${j}_tmp ${i/.ncbi_enc/}\n",
    "let j=$j+1\n",
    "done\n",
    "\n",
    "# double check for successful decryption\n",
    "$ if [ $(find 58601/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000169_visc-adiposity/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000169_visc-adiposity/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000169_visc-adiposity/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# copy files \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58623/JAAMH/CIDR_VisceralAdiposity/phs000169v1/p1\n",
    "$ mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000169_visc-adiposity/meta/\n",
    "$ mv *.gz /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000169_visc-adiposity/phenotype/unprocessed/\n",
    "$ mv phg000052v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000169_visc-adiposity/genotype/original/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "$ aws s3 cp phs000169_visc-adiposity s3://rti-common/dbGaP/phs000169_visc-adiposity/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000170_cataract-lowHDL\n",
    "eMERGE Genome Wide Association Study of Cataract and Low HDL cholesterol(phs000170.v2.p1)\n",
    "* Download 58626\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58626\n",
    "\n",
    "# check for successful decryption\n",
    "$ if [ $(find 58626/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000170_cataract-lowHDL/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000170_cataract-lowHDL/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000170_cataract-lowHDL/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# copy files\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58626/PhenoGenotypeFiles/\\\n",
    "$ RootStudyConsentSet_phs000170.eMERGE_Cataracts.v2.p1.c1.HMB-GSO\n",
    "$ mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000170_cataract-lowHDL/meta/\n",
    "$ mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000170_cataract-lowHDL/genotype/original/unprocessed/\n",
    "$ mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000170_cataract-lowHDL/phenotype/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP/s3/\n",
    "$ aws s3 cp phs000170_cataract-lowHDL/ s3://rti-common/dbGaP/phs000170_cataract-lowHDL/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### phs000287_cardio-health_c1\n",
    "Cardiovascular Health Study (CHS) (phs000287.v6.p1.c2)\n",
    "* Download 58627"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58627\n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find 58627/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000287_cardio-health/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000287_cardio-health/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000287_cardio-health/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "$ cd 58627/PhenoGenotypeFiles/\n",
    "\n",
    "# copy files - meta\n",
    "$ cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/StudyMetaFiles/* ../../s3/phs000287_cardio-health/meta/\n",
    "\n",
    "# copy files - phenotype\n",
    "$ cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "\n",
    "# copy files - genotype\n",
    "$ cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/       \n",
    "$ cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "$ aws s3 cp phs000287_cardio-health/ s3://rti-common/dbGaP/phs000287_cardio-health/ --recursive\n",
    "    \n",
    "# had to rename this folder on s3: appended \"_c1\"\n",
    "$ aws s3 mv s3://rti-common/dbGaP/phs000287_cardio-health/ s3://rti-common/dbGaP/phs000287_cardio-health_c1/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000287_cardio-health_c2\n",
    "Cardiovascular Health Study (CHS) (phs000287.v6.p1.c2)\n",
    "* Download 58634"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58634\n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find 58634/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# resolve problematic decryption\n",
    "$ for i in $(find ./58634/ -name *ncbi_enc -print); do\n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\"; mv ${i} ${j}_tmp.ncbi_enc; \n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc;\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\";\n",
    "mv ${j}_tmp ${i/.ncbi_enc/};\n",
    "let j=$j+1;\n",
    "done\n",
    "\n",
    "# verify the decryption resolution was successful\n",
    "$ if [ $(find 58634/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000287_cardio-health_c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000287_cardio-health_c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000287_cardio-health_c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "\n",
    "# copy meta files \n",
    "$ cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/StudyMetaFiles/* ../../s3/phs000287_cardio-health_c2/meta/\n",
    "\n",
    "# copy phenotype files\n",
    "$ cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "\n",
    "# copy genotype files\n",
    "$ cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "$ cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "$ cd ../../s3/\n",
    "$ aws s3 mv phs000287_cardio-health_c2/ s3://rti-common/dbGaP/phs000287_cardio-health_c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000289_thrombosis\n",
    "GENEVA GWAS of Venous Thrombosis (phs000289.v2.p1)\n",
    "* Download 58637"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58637\n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find 58637/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000289_thrombosis/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000289_thrombosis/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000289_thrombosis/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move data to the directory structure that was created\n",
    "$ cd 58637/NHGRI/VenousThrombosis/phs000289v2/p1/\n",
    "$ mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000289_thrombosis/meta/\n",
    "$ mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000289_thrombosis/phenotype/unprocessed/\n",
    "$ mv genotype/phg000088v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000289_thrombosis/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "$ aws s3 mv phs000289_thrombosis/ s3://rti-common/dbGaP/phs000289_thrombosis/ --recursive\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## phs000285_CARDIA-c1\n",
    "CARDIA_Cohort (phs000285.v3.p2.c1)\n",
    "* Download 58673"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58673\n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find 58673/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000285_CARDIA-c1/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000285_CARDIA-c1/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000285_CARDIA-c1/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv data to the directory structure that was created\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58673/PhenoGenotypeFiles/\n",
    "$ mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c1.HMB-IRB/StudyFiles/* ../../s3/phs000285_CARDIA-c1/meta/\n",
    "$ mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "$ aws s3 mv phs000285_CARDIA-c1/ s3://rti-common/dbGaP/phs000285_CARDIA-c1/ --recursive "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## phs000285_CARDIA-c2\n",
    "CARDIA_Cohort (phs000285.v3.p2.c2)\n",
    "* Download 58657"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58657\n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find 58657/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000285_CARDIA-c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000285_CARDIA-c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000285_CARDIA-c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58657/PhenoGenotypeFiles/\n",
    "mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c2.HMB-IRB-NPU/StudyFiles/* ../../s3/phs000285_CARDIA-c2/meta/\n",
    "mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed\n",
    "\n",
    "cd /cygdrive/c/User/jmarks/ncbi/dbGaP/s3/\n",
    "aws s3 mv phs000285_CARDIA-c2 s3://rti-common/dbGaP/phs000285_CARDIA-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000280_atherosclerosis\n",
    "Atherosclerosis Risk in Communities (ARIC) Cohort (phs000280.v4.p1)\n",
    "* Download 58680"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58680\n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find 58680/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000280_atherosclerosis/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000280_atherosclerosis/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000280_atherosclerosis/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "$ cd 58680/PhenoGenotypeFiles/\n",
    "$ mv RootStudyConsentSet_phs000280.RootStudy.v4.p1.c1.HMB-IRB/StudyMetaFiles/* ../../s3/phs000280_atherosclerosis/meta\n",
    "$ mv RootStudyConsentSet_phs000280.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed\n",
    "$ mv ChildStudyConsentSet_phs000090.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000223.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000398.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000557.RootStudy.v3.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000668.RootStudy.v2.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000860.RootStudy.v2.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000860.RootStudy.v2.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000668.RootStudy.v2.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000557.RootStudy.v3.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000398.RootStudy.v4.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000223.RootStudy.v4.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "$ mv ChildStudyConsentSet_phs000090.RootStudy.v4.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "\n",
    "cd ../../s3\n",
    "aws s3 mv phs000280_atherosclerosis/ s3://rti-common/dbGaP/phs000280_atherosclerosis/ --recursive\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000103_preterm\n",
    "GENEVA Study of Preterm Delivery (DNBC/Iowa) (phs000103.v1.p1)\n",
    "* Download "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v \n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find / -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000103_preterm/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000103_preterm/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000103_preterm/phenotype/{final,unprocessed,processing}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000203_arterial-disease\n",
    "A Genome-Wide Association Study of Peripheral Arterial Disease (phs000203.v1.p1)\n",
    "* Download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "$ cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "$ ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v \n",
    "\n",
    "# Check for successful decryption\n",
    "$ if [ $(find / -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000203_arterial-disease/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000203_arterial-disease/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000203_arterial-disease/phenotype/{final,unprocessed,processing}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  },
  "toc": {
   "nav_menu": {
    "height": "12px",
    "width": "160px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
