{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Opioid plexus data for association analysis\n",
    "\n",
    "We need to several tasks for the HiC project. Ideally we would have this done during January – as soon as different opioid GWAS get run.\n",
    "\n",
    "The attached file contains information on 8 high priority genes for look up in opioid GWAS. In particular the second table in each file gives us a set of regions (positionally defined in hg19) for which we need to:\n",
    "\n",
    "1.\tIdentify variants\n",
    "2.\tLook up those variants for association with opioid addiction phenotypes.\n",
    "3.\tThe first dataset for look up would be MVP (EA, AA, and meta) – probably best to do this with the largest results set (including SAGE and YP)\n",
    "4.\tThe second data set is the PGC cases versus unexposed controls that Ravi is using in the gSEM\n",
    "5.\tThe third results set will be the NGC when it is complete\n",
    "6.\tThe fourth will the gSEM results (EA, then AA, meta).\n",
    "\n",
    "It would be great if one of you could identify the variant sets from these regions and either do the look up or pass it to Ravi or Nathan who have the result files. We want to assign a variable to each of the snps that indicates whether they are from a loss VEL or gain VEL region. This is indicated by the asterisks. Regions with asterisks are loss VELs. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identify variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p /shared/rti-hic/opioid_plexus_data_for_association_analysis/identify_variants/grch37_variants \n",
    "cd /shared/rti-hic/opioid_plexus_data_for_association_analysis/identify_variants/grch37_variants\n",
    "\n",
    "# download chromosomes containing gene regions of interest\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr5.tsv.gz .\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr7.tsv.gz .\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr7.tsv.gz .\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr8.tsv.gz .\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr9.tsv.gz .\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr10.tsv.gz .\n",
    "aws s3 cp s3://rti-common/variants/b153/GRCh37.p13/variants_chr11.tsv.gz .\n",
    "    \n",
    "# upload file containing 8 high priority genes from local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-hic/opioid_plexus_data_for_association_analysis/identify_variants/opioid_plexus_data/\n",
    "\n",
    "unzip plx_opioid.zip\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### python3\n",
    "\n",
    "\"\"\"\n",
    "Identify variants within a set of GRCh37 positional ranges. \n",
    "More specifically, we import a list of positional ranges\n",
    "and we then identify the variants witin these ranges and also indicate if\n",
    "they are a Loss or a Gain VEL, as indicated in the imported list\n",
    "along with the positional ranges.\n",
    "\"\"\"\n",
    "\n",
    "import gzip\n",
    "\n",
    "def main():\n",
    "\n",
    "    gene_list = [\"ADCY2\", \"ANKS6\", \"ASTN2\", \"DUSP4\", \"GABBR2\", \"KCNC1\", \"KCNMA1\", \"WBSCR17\"]\n",
    "\n",
    "    for gene in gene_list:\n",
    "        infile = \"/shared/rti-hic/opioid_plexus_data_for_association_analysis/identify_variants/opioid_plexus_data/plx_{}.txt\".format(gene)\n",
    "        outfile = \"/shared/rti-hic/opioid_plexus_data_for_association_analysis/identify_variants/results/plx_{}_variants_identified.txt\".format(gene)\n",
    "\n",
    "        vel_dict, chrom = create_vel_dict(infile)\n",
    "\n",
    "        varfile = \"/shared/rti-hic/opioid_plexus_data_for_association_analysis/identify_variants/grch37_variants/variants_{}.tsv.gz\".format(chrom)\n",
    "        parse_variant_file(varfile,outfile, vel_dict)\n",
    "\n",
    "\n",
    "\n",
    "# Populate a dictionary that holds the range for the VEL \n",
    "# and if it is a Loss or a Gain {lower_bound: [upperbound, loss]}\n",
    "def create_vel_dict(infile):\n",
    "\n",
    "    with open(infile) as inF:\n",
    "        line = inF.readline()\n",
    "\n",
    "        while line[0:3] != \"chr\":\n",
    "            line = inF.readline()\n",
    "\n",
    "        chrom = line.split()[0]\n",
    "        range_dict = {}\n",
    "\n",
    "        while line[0:3] == \"chr\":\n",
    "            vel_status = \"Gain\"\n",
    "            line = line.split()\n",
    "            lower_bound = line[1]\n",
    "            upper_bound = line[2]\n",
    "            vel = line[3]\n",
    "\n",
    "            if \"*\" in vel:\n",
    "                vel_status = \"Loss\"\n",
    "\n",
    "            range_dict[lower_bound] = [upper_bound, vel_status]\n",
    "            line = inF.readline()\n",
    "\n",
    "    return range_dict, chrom\n",
    "\n",
    "\n",
    "\n",
    "# Helper function that returns the Gain or Loss status of a variant if it is \n",
    "# within the range of interest, else return None\n",
    "def within_range(test_dic, number):\n",
    "    for item in test_dic.items():\n",
    "        smallest = int(item[0])\n",
    "        largest = int(item[1][0])\n",
    "        in_range = smallest <= int(number) <= largest \n",
    "        if in_range:\n",
    "            return item[1][1]\n",
    "            break\n",
    "\n",
    "# Parse each line of the chromosome-specific variant file to determine\n",
    "# which variants are within the positional ranges of interest.\n",
    "# Print out these lines to file along with the VEL Gain or Loss status.\n",
    "def parse_variant_file(varfile, outfile, vel_dict):\n",
    "    with gzip.open(varfile, 'rt') as varF, open(outfile, 'w') as outF:\n",
    "        head = varF.readline().strip()\n",
    "        head += \"\\tVEL\\n\" \n",
    "        outF.write(head)\n",
    "        line = varF.readline()\n",
    "        while line:\n",
    "            sl = line.split()\n",
    "            pos = sl[2]\n",
    "            vel = within_range(vel_dict, pos)\n",
    "            if vel != None:\n",
    "                sl.append(vel)\n",
    "                outline = \"\\t\".join(sl) + \"\\n\"\n",
    "                outF.write(outline)\n",
    "            line = varF.readline()\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
