{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data processing for dbGaP phs000801.v2.p1 (Non-Hodgkin Lymphoma)\n",
    "\n",
    "\n",
    "**Author:** Jesse Marks\n",
    "\n",
    "This document logs several components of data processing for [National Cancer Institute (NCI) Non-Hodgkin Lymphoma Genome-wide Association Study (GWAS)](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000801.v2.p1) including \n",
    "\n",
    "* Data retrieval\n",
    "* Genotype data quality control.\n",
    "\n",
    "The purpose of processing these data are to prepare them for further processing and analysis steps such as haplotype phasing, imputation, and genome-wide association analysis.\n",
    "\n",
    "**Note**: This study consists of five sub-studies - one set of controls and four sets of cases. We will focus only on the controls. See discussion in GitHub [Issue # 68](https://github.com/RTIInternational/bioinformatics/issues/68#issuecomment-365721428).\n",
    "Also note, the `HumanOmniExpressExome-8v1_A` chip was used to genotype the control subjects in this study. The flip file for this chip was not available on EC2 and therefore needed to be downloaded from [here](http://www.well.ox.ac.uk/~wrayner/strand/index.html). The script for generating the flip file is in this notebook [here](#flip-script). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Software and tools\n",
    "The software and tools used for processing these data are\n",
    "* Windows 10 with [Cygwin](https://cygwin.com/) installed \n",
    "* [Aspera Connect](http://downloads.asperasoft.com/downloads)\n",
    "* [KING](http://people.virginia.edu/~wc9c/KING/)\n",
    "* [PLINK v1.9 beta 3.45](https://www.cog-genomics.org/plink/) \n",
    "* [SRA toolkit](https://www.ncbi.nlm.nih.gov/sra/docs/toolkitsoft/)\n",
    "* [STRUCTURE](https://web.stanford.edu/group/pritchardlab/structure.html)\n",
    "* [R v3.2.3](https://www.r-project.org/)\n",
    "* [iGraph (R package)](http://igraph.org/r/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data retieval\n",
    "### Genotypes and phenotypes\n",
    "The genotype and phenotype data were downloaded from dbGaP to a local machine, decrypted (most), and then transfered to [Amazon Simple Storage Service (S3)](https://aws.amazon.com/s3/?sc_channel=PS&sc_campaign=acquisition_US&sc_publisher=google&sc_medium=s3_b&sc_content=s3_e&sc_detail=amazon%20s3&sc_category=s3&sc_segment=192085379926&sc_matchtype=e&sc_country=US&s_kwcid=AL!4422!3!192085379926!e!!g!!amazon%20s3&ef_id=Wd4L7QAAAFUPlk0C:20180115144528:s). \n",
    "Some of the data were not decrypted locally and will thus need to be decrypted on [Amazon Elastic Compute Cloud (EC2)](https://aws.amazon.com/ec2/?sc_channel=PS&sc_campaign=acquisition_US&sc_publisher=google&sc_medium=ec2_b_rlsa_hv&sc_content=ec2_e&sc_detail=amazon%20ec2&sc_category=ec2&sc_segment=213206985258&sc_matchtype=e&sc_country=US&s_kwcid=AL!4422!3!213206985258!e!!g!!amazon%20ec2&ef_id=Wd4L7QAAAFUPlk0C:20180115144653:s).\n",
    "The data were downloaded using the Aspera Connect Browser plug-in. Note that these data require authorized access, so the [authorized access portal](https://dbgap.ncbi.nlm.nih.gov/dbgap/aa/wga.cgi?page=login) must be used (request login information from Eric Johnson). \n",
    "\n",
    "The data files downloaded from dbGaP are encrypted and thus will need to be decrypted using `vdb-decrypt` from the SRA toolkit [(instructions here)](https://www.ncbi.nlm.nih.gov/books/NBK63512/#_Download_Points_often_Ignored_When_Decry_).\n",
    "\n",
    "**Note**: As a way to conserve space, certain genotype data for a given study are excluded from download. General criteria are:\n",
    "* Exclude imputed data\n",
    "* Exclude individual format genotype calls (if the matrix and/or PLINK binary fileset format is available)\n",
    "* Exclude index files that lay out the directory structure for the individual format genotype calls\n",
    "* Exclude raw array data if genotype calls are available\n",
    "\n",
    "To assess which files may be unnecessary for download, the study report available through the [public FTP download site]() (accessible via the dbGaP landing page for a given study accession) should be examined."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General directory structure setup\n",
    "\n",
    "After locally downloading, all of the dbGaP data should be organized within a directory called `ncbi`. For each cohort, its data needs to be placed within a subdirectory of the format dbGaP-x where x signifies the project number of the download (not to be confused with the download request number). The project number can be found on the \"Downloads\" tab of the dbGaP authorized access portal. Also in that tab is the link to download the repository key file. This file should be placed in the top level of the project folder then imported using [vdb-config](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std). Although each download request provides a link to download the repository key, only one key file is needed per project.\n",
    "\n",
    "**Note:** File names longer than the allowed Windows file name character limit will not decrypt and must be renamed before decrypting. It is highly recommended to check the file names after decryption to ensure that they successfully decrypted."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Amazon Web Services\n",
    "#### S3 Directory Structure  \n",
    "For each dbGaP study, we will use the directory structure setup outlined on [GitHub for AWS S3](https://github.com/RTIInternational/bioinformatics/blob/master/documentation/aws_s3_data_organization.md).\n",
    "\n",
    "#### Configure AWS \n",
    "The settings for using the AWS CLI need to be configured before interacting with AWS. These configurations include security credentials and the default region. For more information on this process, see [here](http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html).\n",
    "\n",
    "**Note:** these only need to be configured once. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Local machine (Cygwin)\n",
    "aws configure\n",
    "\n",
    "AWS Access Key ID [None]: AKIAJONBCJHOJSW2PFJA\n",
    "AWS Secret Access Key [None]:  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\n",
    "Default region name [None]: us-east-1\n",
    "Default output format [None]: text  # could be json, text, or table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Modify an EBS Volume from the Command Line\n",
    "The following example demonstrates how an EBS volume can be modified from the command line using hte AWS CLI. An example is detailed [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/cli-modify.html).To make use of the new storage capacity after modifying the ebs volume you need to [extend a linux file system after resizing](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/recognize-expanded-volume-linux.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "\n",
    "# first configure AWS. you will be prompted to enter the following information\n",
    "# the keys and default region can be found in the config file on the cluster launcher\n",
    "# note, these keys are not accurate, they are just examples of what one would see\n",
    "aws configure\n",
    "'''\n",
    "AWS Access Key ID [None]: AKIAI44QH8DHBEXAMPLE\n",
    "AWS Secret Access Key [None]: je7MtGbClwBF/2Zp9Utk/h3yCo8nvbEXAMPLEKEY\n",
    "Default region name [None]: us-east-1\n",
    "Default output format [None]: text\n",
    "'''\n",
    "\n",
    "# this command would modify the volume to 1800 (remove dry-run to actually apply)\n",
    "aws ec2 modify-volume --dry-run --volume-id vol-038921893392154fa --size 1800  \n",
    "\n",
    "# extend file system to the new volume capacity.\n",
    "sudo resize2fs /dev/xvdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='flip-script'></a> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genotype processing\n",
    "## Quality Control Sample Tracking\n",
    "The table below provides statistics on variants and subjects filtered during each step of the QC process.\n",
    "### Initial QC\n",
    "\n",
    "| QC procedure                         | Variants removed | Variants retained | Subjects removed | Subjects retained |\n",
    "|--------------------------------------|------------------|-------------------|------------------|--------------------------|\n",
    "| Initial dbGaP dataset                |0                 |727,545            |0                 |1,683                     |\n",
    "| Genome build 37 and dbGaP 138 update |1,183             |726,362            |0                 |1,683                     |\n",
    "| Duplicate rsID filtering             |0                 |726,362            |0                 |1,683                     |\n",
    "\n",
    "[Genome Build 27](#genome_build)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## European Ancestry\n",
    "### Autosome statistics\n",
    "This table includes filtering statistics prior to merging with chrX.\n",
    "\n",
    "| QC procedure                                    | Variants removed | Variants retained | Subjects removed | Subjects added | Subjects retained |\n",
    "|-------------------------------------------------|------------------|-------------------|------------------|----------------|-------------------|\n",
    "| Pre-partitioning w/initial procedures (all chr) |0                 |726,362            |0                 |0                |1,683               |                    \n",
    "| Partitioning to only autosomes                  |19,845            |706,517            |0                 |0                |1,683               |                    \n",
    "| Remove subjects missing whole autosome data     |0                 |706,517            |0                 |0                |1,683               |                    \n",
    "| STRUCTURE analysis (AFR < 0.25 and EAS < 0.25)  |0                 |706,517            |0                 |0                |1,683               |\n",
    "| Remove variants with missing call rate > 3%     |19,952            |686,565            |0                 |0                |1,683               |                    \n",
    "| Remove variants with HWE p < 0.0001             |1,731             |684,834            |0                 |0                |1,683               |                    \n",
    "\n",
    "[Partitioning Chromosomes](#split_chromosomes) <br>\n",
    "[Call Rate](#call_rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ChrX statistics\n",
    "This table includes filtering statistics prior to merging with autosomes.\n",
    "\n",
    "| QC procedure                                    | Variants removed | Variants retained | Subjects removed | Subjects added | Subjects retained |\n",
    "|-------------------------------------------------|------------------|-------------------|------------------|---------------|--------------|\n",
    "| Pre-partitioning w/initial procedures (all chr) |0                 |726,362            |0                 |0          |1,683         |\n",
    "| Partitioning to only chrX                       |707,901           |18,461             |0                 |0         |1,683         |\n",
    "| Remove subjects missing whole chrX data         |0                 |18,461             |0                 |0               |1,683         |\n",
    "| STRUCTURE analysis (AFR < 0.25 and EAS < 0.25)  |0                 |18,461             |0                 |0                |1,683               |\n",
    "| Remove variants with missing call rate > 3%     |1,147             |17,314             |0                 |0              |1,683         |\n",
    "| Remove variants with HWE p < 0.0001             |8                 |17,306             |0                 |0              |1,683         |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merged autosome and chrX statistics\n",
    "\n",
    "| QC procedure                                            | Variants removed | Variants retained | Subjects removed | Subjects added | Subjects retained |\n",
    "|---------------------------------------------------------|------------------|-------------------|------------------|----------------------|-------------------|\n",
    "| Merge autosomes and chrX                                |0                 |702,140            |0                 |0                      |1,683              |\n",
    "| Remove subjects with IBD > 0.4, IBS > 0.9, KING > 0.177 |0                 |702,140            |85                |0                      |1,598              |\n",
    "| Remove subjects with missing call rate > 3%             |0                 |702,140            |41                |0                      |1,557              |\n",
    "| Sex discordance filter                                  |0                 |702,140            |0                 |0                      |1,557              |\n",
    "| Excessive homozygosity filter                           |0                 |702,140            |0                 |0                      |1,557              |\n",
    "| Duplicate variant ID filter after 1000G renaming        |0                 |702,140            |0                 |0                      |1,557              |\n",
    "| Remove any last ambiguous SNPs                          |2,275             |699,865            |                  | 0                    |1,557              |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EC2 login\n",
    "Login to EC2 master instance. In this case, we use a spot instance called `QCspot`. The compute instance types are `m4.xlarge`. This can be updated with `cfncluster` in the cluster launcher with the command:\n",
    "\n",
    "`cfncluster update QCspot -t default_spot`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# local command line\n",
    "ssh -i ~/.ssh/gwas_rsa ec2-user@35.171.207.199"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieving data from S3 and decompress\n",
    "To avoid writing over the current directory structure, I create a new directory within the directory structure of the other studies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "# Retrieve data from S3 and store in EC2\n",
    "cd /shared/data/studies\n",
    "aws s3 cp s3://rti-common/dbGaP/phs000801_lymphoma_lung_cancer/ phs000801_lymphoma_lung_cancer --recursive\n",
    "\n",
    "# create directory structure\n",
    "cd /shared/data/studies/phs000801_lymphoma_lung_cancer/\n",
    "mkdir -p phenotype/{final,processing} \n",
    "mkdir -p genotype/original/{final,processing}\n",
    "\n",
    "# decompress phenotype data and gunzip\n",
    "cp phenotype/unprocessed/* phenotype/processing\n",
    "for f in phenotype/processing/*.gz; do\n",
    "gunzip \n",
    "done &\n",
    "\n",
    "guzip -r phenotype/processing* &\n",
    "\n",
    "\n",
    "# decompress genotype data and gunzip\n",
    "cp genotype/original/unprocessed/* genotype/original/processing &   \n",
    "cd /genotype/original/processing\n",
    "for f in *.gz; do\n",
    "tar -xzf $f -C .\n",
    "done &\n",
    "\n",
    "\n",
    "rm *.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Renaming the data** can be done, mainly for convenience. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/phenotype/processing\n",
    "for i in phs000818*;do echo \"mv $i $(echo $i | perl -pi -e s/phs.+Set_//g)\"; done\n",
    "\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "for i in phg000579*; do echo \"mv $i $(echo $i | perl -pi -e s/phg.+oma_//g)\";done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exlusion of subjects without phenotype data\n",
    "The `.fam` file may contain more subject IDs than the phenotype file. The subjects without phenotype data are excluded as they provide no benefit for GWA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "mkdir ea\n",
    "\n",
    "# Get EA subject IDs\n",
    "tail -n +12 ../../../phenotype/processing/Subject_Phenotypes.GRU-NPU-MDS.txt |\\\n",
    "    awk '{print $2}' > ../../../phenotype/processing/ea_subject_ids.txt\n",
    "\n",
    "# use this section if IDs are exclusively numerical\n",
    "----------------------------------------------------------------------------------\n",
    "# Append and prepend a space at the beginning of the id so that in the next step\n",
    "# an erroneous ID won't be added because of a over generalized regex expression.\n",
    "# Ex. ID=292  erroneousID=3292\n",
    "#awk '{print \" \" $0 \" \"}' ea_subject_ids.txt > ea_subject_ids.copy\n",
    "#mv ea_subject_ids.copy ea_subject_ids.txt\n",
    "----------------------------------------------------------------------------------\n",
    "\n",
    "# Add family IDs\n",
    "grep -f ../../../phenotype/processing/ea_subject_ids.txt \\\n",
    "    ControlSet.genotype-calls-matrixfmt.c3.GRU-NPU-MDS/ALL_NHL_OmniEx_CONTROL-cg3.fam | \\\n",
    "    cut -d ' ' -f 1,2    \\\n",
    "    > ../../../phenotype/processing/ea_subject_ids.keep\n",
    "\n",
    "# Create filtered PLINK filesets\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile  ControlSet.genotype-calls-matrixfmt.c3.GRU-NPU-MDS/ALL_NHL_OmniEx_CONTROL-cg3 \\\n",
    "        --keep ../../../phenotype/processing/ea_subject_ids.keep \\\n",
    "        --make-bed \\\n",
    "        --out ea/genotypes\n",
    "\n",
    "'727545 variants and 1683 people pass filters and QC.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strand orientation conversion (Attempt 1)\n",
    "<a id='flip-script'></a> \n",
    "\n",
    "For this study, there was ostensibly a discrepancy with the chip that was reported as used. We test a four different chip as well as a test where all of the ambiguous SNPs are removed. The merging results will give us an idea of the actual chip that was used. If after all of this testing, there is still ambiguity, we will use the PLINK flip. The caveat with using the PLINK flip is that it does not handle ambiguous SNPs (i.e. A|T, C|G alleles) and those get removed. This unfortunetly reduced the quality of imputation because of the fewer number of SNPs available. \n",
    "\n",
    "__Note__: Download zip file for chip `HumanOmniExpressExome-8v1_{A,a,B,C}` from <br>\n",
    "http://www.well.ox.ac.uk/~wrayner/strand/ - includes the following (for chip A): <br>\n",
    " `HumanOmniExpressExome-8v1_A.strand` <br>\n",
    " `HumanOmniExpressExome-8v1_A.miss` <br>\n",
    " `HumanOmniExpressExome-8v1_A.multiple`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chip A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "mkdir A_version\n",
    "cd A_version\n",
    "\n",
    "perl -lane 'if ($F[4] eq \"-\") { print $F[0]; }' ~/HumanOmniExpressExome-8v1_A/HumanOmniExpressExome-8v1_A-b37.strand > \\\n",
    "     ~/HumanOmniExpressExome-8v1_A/HumanOmniExpressExome-8v1_A-b37.top_to_plus.flip\n",
    "wc -l ~/HumanOmniExpressExome-8v1_A/HumanOmniExpressExome-8v1_A-b37.top_to_plus.flip\n",
    "'474,237'\n",
    "\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_test\n",
    "'Error: 9729 variants with 3+ alleles present.'\n",
    "\n",
    "# Flip chr1 variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile genotypes \\\n",
    "    --chr 1 \\\n",
    "    --flip ~/HumanOmniExpressExome-8v1_A/HumanOmniExpressExome-8v1_A-b37.top_to_plus.flip \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data using flipped data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_flipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped_test\n",
    "\n",
    "'Error: 23046 variants with 3+ alleles present.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are ambiguous still, so we will attempt the flipping with chip version `a`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chip a Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "mkdir a_version\n",
    "cd a_version\n",
    "\n",
    "# Create flip list for use with PLINK\n",
    "perl -lane 'if ($F[4] eq \"-\") { print $F[0]; }' ~/humanomniexpressexome-8v1-0_a/humanomniexpressexome-8v1-0_a-b37.strand > \\\n",
    "     ~/humanomniexpressexome-8v1-0_a/humanomniexpressexome-8v1-0_a-b37.top_to_plus.flip\n",
    "wc -l ~/humanomniexpressexome-8v1-0_a/humanomniexpressexome-8v1-0_a-b37.top_to_plus.flip\n",
    "'474,238'\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_test\n",
    "'Error: 9729 variants with 3+ alleles present.'\n",
    "\n",
    "\n",
    "# Flip chr1 variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --flip ~/humanomniexpressexome-8v1-0_a/humanomniexpressexome-8v1-0_a-b37.top_to_plus.flip \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data using flipped data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_flipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped_test\n",
    "'Error: 23046 variants with 3+ alleles present.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are ambiguous still, so we will attempt the flipping with chip version `B`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chip B Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea/\n",
    "mkdir B_version\n",
    "cd B_version\n",
    "\n",
    "perl -lane 'if ($F[4] eq \"-\") { print $F[0]; }' ~/HumanOmniExpressExome-8-v1-0-B/HumanOmniExpressExome-8-v1-0-B-b37.strand > \\\n",
    "     ~/HumanOmniExpressExome-8-v1-0-B/HumanOmniExpressExome-8-v1-0-B-b37.top_to_plus.flip\n",
    "wc -l  ~/HumanOmniExpressExome-8-v1-0-B/HumanOmniExpressExome-8-v1-0-B-b37.top_to_plus.flip\n",
    "'474,238'\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_test\n",
    "'Error: 9729 variants with 3+ alleles present.'\n",
    "\n",
    "# Flip chr1 variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --flip ~/HumanOmniExpressExome-8-v1-0-B/HumanOmniExpressExome-8-v1-0-B-b37.top_to_plus.flip \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data using flipped data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_flipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped_test\n",
    "'Error: 23046 variants with 3+ alleles present.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are ambiguous still, so we will attempt the flipping with chip version `C`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chip C Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "mkdir C_version\n",
    "cd C_version\n",
    "\n",
    "\n",
    "perl -lane 'if ($F[4] eq \"-\") { print $F[0]; }' ~/HumanOmniExpressExome-8-v1-1-C/HumanOmniExpressExome-8-v1-1-C-b37.strand > \\\n",
    "     ~/HumanOmniExpressExome-8-v1-1-C/HumanOmniExpressExome-8-v1-1-C-b37.top_to_plus.flip\n",
    "wc -l ~/HumanOmniExpressExome-8-v1-1-C/HumanOmniExpressExome-8-v1-1-C-b37.top_to_plus.flip\n",
    "'477,770'\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_test\n",
    "'Error: 9729 variants with 3+ alleles present.'\n",
    "\n",
    "# Flip chr1 variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --flip ~/HumanOmniExpressExome-8-v1-0-B/HumanOmniExpressExome-8-v1-0-B-b37.top_to_plus.flip \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data using flipped data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_flipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped_test\n",
    "'Error: 23046 variants with 3+ alleles present.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are ambiguous still, so we will remove any ambiguous SNPs (A|T or C|G) and subsequently attempt a merge."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Ambiguous SNPs Removed Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "\n",
    "mkdir remove_ambiguous\n",
    "cd remove_ambiguous\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile ../genotypes \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# create a list of ambiguous SNPs\n",
    "perl -lane 'if (($F[4] eq \"A\" && $F[5] eq \"T\") || ($F[4] eq \"T\" && $F[5] eq \"A\") || \n",
    "($F[4] eq \"C\" && $F[5] eq \"G\") || ($F[4] eq \"G\" && $F[5] eq \"C\") ) {print $F[1];}'\n",
    "chr1_unflipped.bim | sort -u > ambiguous_snps.txt\n",
    "\n",
    "wc -l ambiguous_snps.txt \n",
    "'239'\n",
    "\n",
    "# remove ambiguous snps from chr1\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --exclude ambiguous_snps.txt \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_ambiguous_removed\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 800 \\\n",
    "    --bfile chr1_unflipped_ambiguous_removed \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_ambiguous_removed_test\n",
    "'Error: 9729 variants with 3+ alleles present.'\n",
    "\n",
    "# Clean up\n",
    "rm chr1_*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The merging results are still ambiguous. As a result, we will procede with the QC process and use the PLINK `flip` command to flip our SNPs. This is performed in the section right before the STRUCTURE section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update dbSNP and genome build\n",
    "\n",
    "To ensure that all of the population controls have variant and genomic data in dbSNP 138 and genome build 37 format, I use ID and position mappers to make the updates.\n",
    "\n",
    "<a id=\"genome_build\"><a/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "\n",
    "# Update variant chr \n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile genotypes \\\n",
    "        --update-chr /shared/common/build_conversion/b37/dbsnp_b138/uniquely_mapped_snps.chromosomes \\\n",
    "        --make-bed \\\n",
    "        --out genotypes_chr_b37\n",
    "'727545 variants and 1683 people pass filters and QC.'\n",
    "\n",
    "# Update variant chr coordinate\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile genotypes_chr_b37 \\\n",
    "        --update-map /shared/common/build_conversion/b37/dbsnp_b138/uniquely_mapped_snps.positions \\\n",
    "        --make-bed \\\n",
    "        --out genotypes_chr_position_b37\n",
    "\n",
    "# Filter to only build 37 uniquely mapped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile genotypes_chr_position_b37 \\\n",
    "        --extract /shared/common/build_conversion/b37/dbsnp_b138/uniquely_mapped_snps.ids \\\n",
    "        --make-bed \\\n",
    "        --out genotypes_b37_dbsnp138\n",
    "'726362 variants and 1683 people pass filters and QC.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove duplicate SNPs\n",
    "If multiple rsIDs are present then the one with the better genotype call rate across subjects should be retained. Obtaining the genotype call rates across subjects would need to be calculated using PLINK --missing.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000634_lung_cancer/genotype/original/processing/ea\n",
    "\n",
    "# Find duplicate rsIDs - cut the rsID field then print out duplicated lines (both/all time)\n",
    "cut -f2,2 genotypes_b37_dbsnp138.bim | sort | uniq -D > variant_duplicates.txt\n",
    "\n",
    "wc -l variant_duplicates.txt\n",
    "'0'"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "No duplicate SNPs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Partition into autosome and chrX groups\n",
    "I apply QC to autosomes and chrX separately, so separate subdirectories are created for the processing of each set.\n",
    "\n",
    "<a id=\"split_chromosomes\"><a/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "\n",
    "mkdir autosomes chrX\n",
    "\n",
    "# Autosomes\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile genotypes_b37_dbsnp138 \\\n",
    "    --autosome \\\n",
    "    --make-bed \\\n",
    "    --out autosomes/genotypes_b37_dbsnp138\n",
    "'706517 variants and 1683 people pass filters and QC.'\n",
    "\n",
    "# ChrX (include split PARs)\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile genotypes_b37_dbsnp138 \\\n",
    "    --chr 23,25 \\\n",
    "    --make-bed \\\n",
    "    --out chrX/genotypes_b37_dbsnp138_unmerged\n",
    "'18461 variants and 1683 people pass filters and QC.'\n",
    "\n",
    "# Combine split chrX and PARs\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile chrX/genotypes_b37_dbsnp138_unmerged \\\n",
    "    --merge-x \\\n",
    "    --make-bed \\\n",
    "    --out chrX/genotypes_b37_dbsnp138\n",
    "'18461 variants and 1683 people pass filters and QC.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Missing autosome data subject filtering\n",
    "We calculate the proportion of missing genotype calls per chromosome using PLINK to assess whether any subjects have data missing for whole autosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ea_${chr} \\\n",
    "        --script_prefix autosomes/chr${chr}_missing_call_rate \\\n",
    "        --mem 3.8 \\\n",
    "        --priority 0 \\\n",
    "        --program  /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "            --missing \\\n",
    "            --chr $chr \\\n",
    "            --out autosomes/chr${chr}_missing_call_rate\n",
    "done\n",
    "\n",
    "# if the missing call rate is 1 (i.e. all data was missing for that chromosome)\n",
    "for chr in {1..22}; do\n",
    "    tail -n +2 autosomes/chr${chr}_missing_call_rate.imiss | \\\n",
    "        awk '{ OFS=\"\\t\" } { if($6==1){ print $1,$2 } }' >> autosomes/missing_whole_autosome.remove\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this case none of the EA subjects had missing autosome data. If subjects ever show up as having missing autosome data then further discussions need to be had on whether these subjects should be removed completely or whether they should only be excluded for the missing chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Clean up \n",
    "rm autosomes/chr*missing_call_rate*\n",
    "rm autosomes/missing_whole_autosome.remove"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strand orientation conversion (PLINK)\n",
    "To determine if strand flipping is an issue, I perform a merge between chr1 of the study data and 1000 Genomes Phase 3 data. Doing this will produce a log of problematic variants that may be attributable to strand flipping. These results can be compared to chr1 data with flipped variants to see if strand orientation is truly the issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_test\n",
    "'Error: 9728 variants with 3+ alleles present.'\n",
    "\n",
    "# Extract chr1 flipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "    --chr 1 \\\n",
    "    --flip chr1_unflipped_test-merge.missnp \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped\n",
    "\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile chr1_flipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped_test\n",
    "'Error: 274 variants with 3+ alleles present.'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Between the flipped and the unflipped merge, the flipped merge produced drastically less errors. As a consequence, I will apply a merge for each chromosome to produce a flip list then flip the variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea\n",
    "\n",
    "# Get flip lists\n",
    "for chr in {1..22}; do\n",
    "        # Extract unflipped variants\n",
    "       /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 10000 \\\n",
    "            --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "            --chr ${chr} \\\n",
    "            --make-bed \\\n",
    "            --out chr${chr}_unflipped &>> out.file\n",
    "\n",
    "        # Attempt merge with 1000G data\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 10000 \\\n",
    "            --bfile chr${chr}_unflipped \\\n",
    "            --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "            --make-bed \\\n",
    "            --out chr${chr}_unflipped_test &>>  out.file\n",
    "\n",
    "        # Extract flipped variants\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 10000 \\\n",
    "            --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "            --chr ${chr} \\\n",
    "            --flip chr${chr}_unflipped_test-merge.missnp \\\n",
    "            --make-bed \\\n",
    "            --out chr${chr}_flipped &>>  out.file\n",
    "\n",
    "\n",
    "        # Attempt merge with 1000G data\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 10000 \\\n",
    "            --bfile chr${chr}_flipped \\\n",
    "            --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "            --make-bed \\\n",
    "            --out chr${chr}_flipped_test &>> out.file\n",
    "done\n",
    "\n",
    "\n",
    "# Combine flip lists\n",
    "cat chr*_unflipped_test-merge.missnp | sort -u > chr_all_unflipped_test-merge.missnp\n",
    "cat chr*_flipped_test-merge.missnp | sort -u > chr_all_flipped_test-merge.missnp\n",
    "comm -23 chr_all_unflipped_test-merge.missnp chr_all_flipped_test-merge.missnp \\\n",
    "    > autosomes/chr_all.flip\n",
    "\n",
    "# Perform final flip\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "    --flip autosomes/chr_all.flip \\\n",
    "    --make-bed \\\n",
    "    --out autosomes/genotypes_b37_dbsnp138_flipped\n",
    "\n",
    "# Clean up\n",
    "rm chr*flipped*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Detecting ancestral outliers with STRUCTURE\n",
    "[STRUCTURE](https://web.stanford.edu/group/pritchardlab/structure.html) is a software tool that can be used to identify admixed individuals, among other uses. By comparing the study subjects with the 1000 Genomes Phase 3 reference panel, I can estimate the composition of an individual's ancestry to determine any discrepancies between self-reporting and genetic information. For the study data, I will be comparing the individuals to 3 different superpopulations from 1000 Genomes Phase 3 reference panel\n",
    "\n",
    "* AFR (African)\n",
    "* EAS (East Asian)\n",
    "* EUR (European)\n",
    "\n",
    "### SNP subset selection\n",
    "For compuational efficiency 10,000 SNPs are randomly chosen from the intersection of SNPs in the study data with the three 1000 Genomes superpopulations of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/\n",
    "\n",
    "mkdir structure\n",
    "\n",
    "# Get lists of non-A/T and non-C/G SNPs\n",
    "ancestry=\"ea\"\n",
    "perl -lane 'if (($F[4] eq \"A\" && $F[5] ne \"T\") || ($F[4] eq \"T\" && $F[5] ne \"A\") ||\n",
    "($F[4] eq \"C\" && $F[5] ne \"G\") || ($F[4] eq \"G\" && $F[5] ne \"C\")) { print $F[1]; }' \\\n",
    "${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped.bim | \\\n",
    "    sort -u | \\\n",
    "    grep \"rs\" \\\n",
    "    > structure/no_at_cg_snps.txt\n",
    "\n",
    "\n",
    "# Get list of variants from 1000G\n",
    "mkdir structure/1000g_data\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name merge_1000g_snps \\\n",
    "    --script_prefix structure/1000g_data/merge_1000g_snps \\\n",
    "    --mem 3 \\\n",
    "    --priority 0 \\\n",
    "    --program \"cat /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr{1..22}.bim | \\\n",
    "        cut -f2,2 | \\\n",
    "        sort -u | \\\n",
    "        grep \\\"rs\\\" > structure/1000g_data/1000g_phase3_snps.txt\"\n",
    "\n",
    "\n",
    "\n",
    "study=\"lymphoma\"\n",
    "\n",
    "# Get SNP overlap between study data and 1000G\n",
    "comm -12 structure/no_at_cg_snps.txt structure/1000g_data/1000g_phase3_snps.txt > structure/${study}_1000g_shared_snps.txt\n",
    "\n",
    "# Select 10,000 random SNPs from study and 1000G overlap\n",
    "perl -ne 'print rand().\"\\t\".$_' structure/${study}_1000g_shared_snps.txt | \\\n",
    "    sort -k1,1 | \\\n",
    "    head -10000 | \\\n",
    "    cut -f2,2 \\\n",
    "    > structure/10k_snp_random_sample.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Extract SNP subset PLINK binary filesets for study data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Create ped and map files for study genotype data for SNP subset\n",
    "#for ancestry in {ea,aa}; do\n",
    "\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name merge_1000g_snps \\\n",
    "        --script_prefix structure/1000g_data/merge_1000g_snps \\\n",
    "        --mem 4 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped \\\n",
    "            --extract structure/10k_snp_random_sample.txt \\\n",
    "            --snps-only just-acgt \\\n",
    "            --recode \\\n",
    "            --out structure/${ancestry}_10k_snp_random_sample\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Extract SNP subset PLINK binary files sets for 1000G data\n",
    "The 05/2013 release of the 1000 Genomes data have been previously processed and converted to PLINK binary fileset format, but the files include all the 1000G individuals. We are interested in only three superpopulations, so we create filesets specifically for each of these groups. It was brought to my attention that 1000G Phase 3 rsIDs may be duplicated across chromosomes potentially causing chromosome merging issues. For that reason, it is typically recommended to process the chromosomes separately initially, then combine post SNP subsetting. Subject IDs with superpopulation classifications are available at\n",
    "\n",
    "`/shared/data/ref_panels/1000G/igsr_samples.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Ancestry specific directories\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    mkdir structure/1000g_data/${pop}\n",
    "done\n",
    "\n",
    "# Get subject IDs by ancestry\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"African\"){print $1,$1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > structure/1000g_data/AFR/AFR_subject_ids.txt\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"East Asian\"){print $1,$1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > structure/1000g_data/EAS/EAS_subject_ids.txt\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"European\"){print $1,$1} }' /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > structure/1000g_data/EUR/EUR_subject_ids.txt\n",
    "\n",
    "# Make new binary filesets for each 1000G group\n",
    "# only keeping the data from individuals which are from 1 of the three super populations\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${pop}_${chr}_filter \\\n",
    "            --script_prefix structure/1000g_data/${pop}/ancestry_partition_chr${chr} \\\n",
    "            --mem 8 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink  \\\n",
    "                --noweb \\\n",
    "                --memory 10000 \\\n",
    "                --bfile /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "                --keep structure/1000g_data/${pop}/${pop}_subject_ids.txt \\\n",
    "                --make-bed \\\n",
    "                --out structure/1000g_data/${pop}/${pop}.chr${chr}\n",
    "    done\n",
    "done\n",
    "\n",
    "# Apply SNP subset extraction by chr\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${pop}_${chr}_subsample \\\n",
    "            --script_prefix structure/1000g_data/${pop}/ancestry_partition_chr${chr} \\\n",
    "            --mem 7.6 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "                --noweb \\\n",
    "                --memory 7500 \\\n",
    "                --bfile structure/1000g_data/${pop}/${pop}.chr${chr} \\\n",
    "                --extract structure/10k_snp_random_sample.txt \\\n",
    "                --make-bed \\\n",
    "                --out structure/1000g_data/${pop}/${pop}_chr${chr}_10k_snp_random_sample\n",
    "    done\n",
    "done\n",
    "\n",
    "# Create merge lists and merge autosomes for each 1000G population\n",
    "data_dir=structure/1000g_data\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    echo \"${data_dir}/${pop}/${pop}_chr1_10k_snp_random_sample\" > ${data_dir}/${pop}/${pop}_autosome_merge_list.txt\n",
    "    for chr in {2..22}; do\n",
    "        echo \"${data_dir}/${pop}/${pop}_chr${chr}_10k_snp_random_sample\" \\\n",
    "        >> ${data_dir}/${pop}/${pop}_autosome_merge_list.txt\n",
    "    done\n",
    "done\n",
    "\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${pop}_merge_plink_filesets \\\n",
    "        --script_prefix structure/1000g_data/${pop}/merge_plink_filesets \\\n",
    "        --mem 4 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 4000 \\\n",
    "            --merge-list structure/1000g_data/${pop}/${pop}_autosome_merge_list.txt \\\n",
    "            --snps-only just-acgt \\\n",
    "            --make-bed \\\n",
    "            --out structure/1000g_data/${pop}/${pop}_all_autosomes_10k_snp_random_sample\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Discrepancy assessment between 1000G and study data\n",
    "\n",
    "As a quality check that the SNP data subsampled from the 1000 Genomes and study data are the same, I will attempt to merge an arbitrarily selected group from each data set using PLINK. If any errors are found, PLINK will generate an error file. Likely causes of errors would be:\n",
    "\n",
    "* SNP genomic coordinates not matching\n",
    "* SNP duplicates found\n",
    "* SNP strand orientation flipped\n",
    "\n",
    "If errors are found the two options for moving forward are\n",
    "\n",
    "1. Re-run the 10,000 SNP subsampling and hope the SNPs chosen do not raise issues\n",
    "2. Remove the problematic SNPs\n",
    "\n",
    "Option #2 is the prefered approach and the one I will be taking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "ancestry=\"ea\"\n",
    "study=\"lymphoma\"\n",
    "pop=\"EUR\"\n",
    "# Merge EA and EUR genotype files\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/${ancestry}_10k_snp_random_sample \\\n",
    "    --bmerge structure/1000g_data/${pop}/${pop}_all_autosomes_10k_snp_random_sample \\\n",
    "    --recode \\\n",
    "    --out structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample\n",
    "\n",
    "'''\n",
    "Error: 1 variant with 3+ alleles present.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "The attempted merge generated 2 for the EAs. In the case of an unsuccessful merge, the following are common:\n",
    "\n",
    "* Multiple positions found for a variant\n",
    "* Non-biallelic variants found\n",
    "\n",
    "When these occur, a common approach is to exclude variants with multiple positions and see if flipping the non-biallelic variants resolves the second issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "study=\"lymphoma\"\n",
    "ancestry=\"ea\"\n",
    "pop=\"EUR\"\n",
    "# Get list of multiple position variants\n",
    "grep \"Multiple positions seen for variant\" structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.log | \\\n",
    "    cut -d\"'\" -f2,2 > structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.bad_snps.remove\n",
    "\n",
    "# Flip study EA non-biallelic SNPs and remove multi-position variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/${ancestry}_10k_snp_random_sample \\\n",
    "    --exclude structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.bad_snps.remove \\\n",
    "    --flip structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.missnp \\\n",
    "    --recode \\\n",
    "    --out structure/${ancestry}_10k_snp_random_sample_retry\n",
    "\n",
    "'''\n",
    "--flip: 1 SNP flipped.\n",
    "--exclude: 10000 variants remaining.\n",
    "'''\n",
    "\n",
    "# Remove multi-position variants from 1000G ${pop} (there are none so this step is actually uneccessary)\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile structure/1000g_data/${pop}/${pop}_all_autosomes_10k_snp_random_sample \\\n",
    "    --exclude structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.bad_snps.remove \\\n",
    "    --make-bed \\\n",
    "    --out structure/${pop}_all_autosomes_10k_snp_random_sample_retry\n",
    "\n",
    "\n",
    "# Retry merge\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/${ancestry}_10k_snp_random_sample_retry \\\n",
    "    --bmerge structure/${pop}_all_autosomes_10k_snp_random_sample_retry \\\n",
    "    --make-bed \\\n",
    "    --out structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample_retry\n",
    "\n",
    "'''\n",
    "Error: 1 variant with 3+ alleles present.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "If the merge retry fails then the subset of failed flipped SNPs will get combined with the multi-position SNPs into a blacklist to use for creating the final PED files. Otherwise a separate blacklist and flip list will be used with PLINK --exclude and --flip."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Create final exclusion list\n",
    "study=\"lymphoma\"\n",
    "pop=\"EUR\"\n",
    "ancestry=\"ea\"\n",
    "\n",
    "sort structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.missnp | uniq > \\\n",
    "     structure/10k_snp_random_sample_blacklist.txt\n",
    "\n",
    "cat structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample_retry-merge.missnp\\\n",
    "    structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.bad_snps.remove > structure/10k_snp_random_sample_blacklist.txt\n",
    "\n",
    "# Create final flip list if retry merge was successful\n",
    "cat structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample_retry-merge.missnp structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.missnp | \\\n",
    "sort | uniq -u > structure/10k_snp_random_sample_flip_list.txt\n",
    "\n",
    "# File cleanup\n",
    "rm structure/1000g_${pop}_${study}_${ancestry}_10k_snp_random_sample.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### STUCTURE input file construction\n",
    "Because our initial merge and flip test was unsuccessful, I proceed with appying a blacklist filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "pop=\"EUR\"\n",
    "ancestry=\"ea\"\n",
    "\n",
    "# Create final ped and map files for study genotype data for SNP subset\n",
    "#for ancestry in {ea,aa}; do\n",
    "\n",
    "echo \"=======================${ancestry}========================\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 3000 \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped \\\n",
    "        --extract structure/10k_snp_random_sample.txt \\\n",
    "        --exclude structure/10k_snp_random_sample_blacklist.txt \\\n",
    "        --snps-only just-acgt \\\n",
    "        --recode \\\n",
    "        --out structure/${ancestry}_10k_snp_random_sample.final\n",
    "#done\n",
    "'9999 variants and 1683 people pass filters and QC.'\n",
    "\n",
    "# Create ped and map files for each 1000G population\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 1024 \\\n",
    "        --bfile structure/1000g_data/${pop}/${pop}_all_autosomes_10k_snp_random_sample \\\n",
    "        --exclude structure/10k_snp_random_sample_blacklist.txt \\\n",
    "        --recode \\\n",
    "        --out structure/1000g_data/${pop}_10k_snp_random_sample.final\n",
    "done\n",
    "\n",
    "\n",
    "# Final check for SNP discrepancies EA\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/ea_10k_snp_random_sample.final \\\n",
    "    --merge structure/1000g_data/${pop}_10k_snp_random_sample.final \\\n",
    "    --recode \\\n",
    "    --out structure_input_test\n",
    "\n",
    "# File cleanup\n",
    "rm structure_input_test*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No merging issues were identified in the final check, so I will use the script ped2structure.pl to convert the PED file into a STRUCTURE input file format. This script takes two inputs. The first is an integer that serves as an ID to distinguish between a reference panel population or a study data set group. The second input is an integer that is unique to each group/population regardless of whether it's from the study or 1000G data.\n",
    "\n",
    "The goal of the conversion script is to generate a single STRUCTURE input file containing genotype information for the ~10,000 (post-filtered) subsampled SNPs and the individuals from the study and 1000G data sets. Documentation on the format can be found [here](https://web.stanford.edu/group/pritchardlab/structure_software/release_versions/v2.3.4/structure_doc.pdf). The first three columns contatin the following information respectively\n",
    "\n",
    "1. Subject identifier\n",
    "2. Group/population identifier. Distinct for each ancestry group or superpopulation\n",
    "3. Boolean indicator (1=True, 0=False) specifying reference panel populations. This is used by STRUCTURE to define the ancestry groups\n",
    "\n",
    "We will be running structure assuming that the study subjects descended from three populations. The traditional approach would be to use AFR, EAS, and EUR. I will run STRUCTURE using these 1000G superpopulations.\n",
    "\n",
    "**Note:** for $K$ reference panel populations used for ancestry comparisons, the reference panel populations must be given group IDs between 1 and $K$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "mkdir structure/input_files\n",
    "\n",
    "#### Create STRUCTURE file with AFR, EAS, and EUR ####\n",
    "\n",
    "groupID=1 #distinguish between all groups\n",
    "\n",
    "# Append 1000G populations to STRUCTURE file\n",
    "truncate -s 0 structure/input_files/input_afr_eas_eur\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    cat structure/1000g_data/${pop}_10k_snp_random_sample.final.ped | \\\n",
    "    /shared/bioinformatics/software/perl/file_conversion/ped2structure.pl 1 ${groupID} \\\n",
    "    >> structure/input_files/input_afr_eas_eur\n",
    "    groupID=`echo ${groupID} + 1 | bc`\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Append remaining acestry group (ea) to STRUCTURE file\n",
    "ancestry=\"ea\"\n",
    "cat structure/${ancestry}_10k_snp_random_sample.final.ped | \\\n",
    "/shared/bioinformatics/software/perl/file_conversion/ped2structure.pl 0 ${groupID} \\\n",
    "    >> structure/input_files/input_afr_eas_eur\n",
    "    groupID=`echo ${groupID} + 1 | bc` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run STRUCTURE\n",
    "For running STRUCTURE, the following command line paramters are explicitly specified\n",
    "\n",
    "* m - Main parameter file\n",
    "* e - Additional parameter file\n",
    "* i - Input data file\n",
    "* o - Output file\n",
    "* L - Number of loci in data file\n",
    "* N - Number of subjects\n",
    "* K - Maximum number of populations\n",
    "\n",
    "The remaining parameters are specified in the `mainparams` and `extraparams` files. Any values specified on the command-line overrided the values in the configuration files. Documentation on the parameter options can be found [here](https://web.stanford.edu/group/pritchardlab/structure_software/release_versions/v2.3.4/structure_doc.pdf)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "mkdir structure/output_files\n",
    "\n",
    "ancestry='ea'\n",
    "# Run Structure using AFR, EAS, and EUR\n",
    "L=$(wc -l structure/${ancestry}_10k_snp_random_sample.final.map | perl -lane 'print $F[0];')\n",
    "N=$(wc -l structure/input_files/input_afr_eas_eur | perl -lane 'print $F[0]/2;')\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name structure_afr_eas_eur \\\n",
    "    --script_prefix structure/output_files/structure_afr_eas_eur \\\n",
    "    --mem 7.5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program /shared/bioinformatics/software/third_party/structure_v2.3.4/console/structure \\\n",
    "         -m /shared/bioinformatics/software/third_party/structure_v2.3.4/console/mainparams \\\n",
    "         -e /shared/bioinformatics/software/third_party/structure_v2.3.4/console/extraparams \\\n",
    "         -i structure/input_files/input_afr_eas_eur \\\n",
    "         -o structure/output_files/output_afr_eas_eur \\\n",
    "         -L $L \\\n",
    "         -N $N \\\n",
    "         -K 3\n",
    "'278 0.55500 structure_ ec2-user     r     02/14/2018 17:39:29 all.q@ip-172-31-29-1.ec2.inter     1'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Producing triangle plots\n",
    "\n",
    "[Triangle or ternary plots](https://en.wikipedia.org/wiki/Ternary_plot) provide a means to visualize membership strength. For the purposes of the study data, the triangle plots are used to determine how similar self-reports of individuals from a Retirement ancestry group are to three disparate reference populations. The visualizations allow for the identification of potential outliers that would be removed from downstream processing and analysis.\n",
    "\n",
    "**Note:** The command line Perl script below assumes a specific order of the STRUCTURE output data in regards to the group IDs. Modifications should be made as necessary to ensure that the labels match with the ID. Additionaly, `triagle_plot.R` does not check that the cluster specifications match the given 1000G superpopulations labels. Be sure to assign the right label to the right group ID from previous processing steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "mkdir structure/triangle_plots\n",
    "\n",
    "#### AFR, EAS, and EUR ####\n",
    "\n",
    "# Prepare results for triangle plot\n",
    "perl -ne 'if (/%Miss/) {\n",
    "              $in=1;\n",
    "              print \"num\\tID\\tpop\\tcluster1\\tcluster2\\tcluster3\\n\";\n",
    "          }\n",
    "          if ($in==1 && !/Label/ && !/^\\s+$/) {\n",
    "              #@datasets=(\"AFR\",\"EAS\",\"EUR\",\"Study_AA\", Study_EA);\n",
    "              @datasets=(\"AFR\",\"EAS\",\"EUR\", Study_EA);\n",
    "              s/^\\s+//g;\n",
    "              @F=split /\\s+/;\n",
    "              # Grab only study data set groups by ID\n",
    "              if ($F[3] > 3) {\n",
    "                  print $F[0].\"\\t\".$F[1].\"\\t\".$datasets[$F[3]-1].\"\\t\".$F[5].\"\\t\".$F[6].\"\\t\".$F[7].\"\\n\";\n",
    "              }\n",
    "          } \n",
    "          s/\\s+//g;\n",
    "          if ($_ eq \"\") { $in=0; }' structure/output_files/output_afr_eas_eur_f > \\\n",
    "    structure/triangle_plots/afr_eas_eur.triangle_input\n",
    "\n",
    "\n",
    "# Prepare id xref file, EA\n",
    "perl -ne 'if ($count==0) { print \"num\\tID\\n\"; }\n",
    "          if (!defined $even) { print STDERR \"LALALA\\n\";} \n",
    "          if ($even==0) {/^(\\S+)/; print $count+1; print \"\\t\".$1.\"\\n\"; $count++; $even=1}\n",
    "          else {$even=0}' structure/input_files/input_afr_eas_eur > \\\n",
    "    structure/triangle_plots/afr_eas_eur_id_xref\n",
    "\n",
    "#-------------------------------------------------------------------------------------------\n",
    "# Generate triangle plot\n",
    "Rscript /shared/bioinformatics/software/R/triangle_plot.R \\\n",
    "    --data structure/triangle_plots/afr_eas_eur.triangle_input \\\n",
    "    --prefix structure/triangle_plots/afr_eas_eur \\\n",
    "    --cluster1 African \\\n",
    "    --cluster2 Asian \\\n",
    "    --cluster3 European \\\n",
    "    --xref structure/triangle_plots/afr_eas_eur_id_xref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## local machine ##\n",
    "\n",
    "# copy output triangle plot to local desktop\n",
    "scp -i ~/.ssh/gwas_rsa ec2-user@35.171.207.199:/shared/data/studies/phs000801_lymphoma/genotype/original/processing/structure/triangle_plots/afr_eas_eur_Study_EA.jpg ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEB\nAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/2wBDAQEBAQEBAQEBAQEBAQEB\nAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/wAAR\nCAMgBLADASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAA\nAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkK\nFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWG\nh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl\n5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREA\nAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYk\nNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOE\nhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk\n5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwD+/iiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKK/mm8M/tb/tO/tp/tuftwfs+6P/AMFMvh//AME29X/Z\ne+Pl78B/gz+zDbfBf9n3x38bPjJ4a0fwx4e1qz/aJ8US/tG2utap4t8EfEjU9duR4R0H4R6T\n4Z/s/QLC0/tTXTdalp2sav1nxp/4Kmfti/8ABPL4D/seab+3r8H/AA1qXx++L/8AwUO0P9k7\nxR4v+F3w/wDiV468EfE39l6PxdFpeoftSeBvDfwRg+Jr+Cvil4o8Ka14a1HwX+zl4k1Q/E3x\nv4hg8WR+Bfhpqtvo+paH4dAP6LaK8J8B/tL/AAS+Jvxx+Pv7N/gjxr/bfxo/ZftfhLe/HTwZ\n/wAI54t03/hB7b45+G9Z8XfCyT/hItX0Gw8KeJv+Ep8PeH9X1DZ4P13xBJon2T7J4iTSL64t\nbab5L1b/AIK//wDBObRfgJ8G/wBpa+/aRsW+FH7Q3iPx14T+B02nfDX4y6z8RPipr/wy8X65\n4D+IFl4J+CGj/Du++N2uR+EvFfhzVNL1q+tPh4+nWijT9Qa7Ol61ot7qAB+lVFfzxfsmf8Fi\nfhx8Tf21/wDgpvf+NP2ltN8SfsPfCzxL/wAEvvht+zDqdl8NrqPTPAfxK/ax8NfETwd4y8O+\nLdS8OfDu3+JGgXniP446RofhXxRJ8cri3034U+JLObQtWuPA1uNRtW/bi+/aB+EWnfH3w9+y\n9c+Lc/HbxR8K/Efxt0nwFa6B4n1CZfhX4V8T6H4L1jxjq/iCw0W58KeHLAeKvEek6Hpdt4i1\n3StS8QX0l5H4fstTXStWayAPZaKK/nGtvHf7df7Uf/BUb/gqb+z74O/4KbeMf2P/AIMfsfar\n+xdp/wAKfAfhT9n79jvx6t5H8d/2YdF+I/jeW+8SfGn4ReJvGF60XjCz1DU7cXGvagtsNfms\nYPsum2em2VuAf0c0V8P/AAw1DxV+yH+zH8Z/iZ+1R+2X4i/a40r4XWfxC+MXib4y+Kvh58EP\nhpf+Efhv4N8BabrWq+DLfQPgT4U8EeDb620SPw7r3iW31PU9Ol8SXl14om0+71GTTLLRre1/\nLr/gir+3H+238ZPid8Vvgb/wUO8RaVqPxR+K/wAAfgf/AMFEv2W9KsfCHhLwdP4Z/Za/aBl1\njRNQ+FNxb+FPDvhttWuPgh4t0/w3oV9rXiiLV/Gl7deLyviLV5HhtLeAA/olooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nA/ms/aK8Q2Wt/F79pj4O/wDBWP8A4JKePP22/A7fFbxDrP7GP7Qn7M37Eg/aZsNQ/Zt8Q2ll\nP4R+HHirxB4Nk1n4lfCX4w/D/Um1rR9f8RanN4M0nxH9qOqaZdWGmRW2o618J3P7Kv7ZXwl/\n4J3fsaav4p+Af7TuteBvgp/wX6+B/wC2f8KP2YobDxR+0Z+0p+zB/wAE4vBvjvVr7wX8PfEf\nh7wne+O/GHiXX/A2nm91y48K6RqPizW/Dtp4vsdF1KeOfSdTTT/7P6KAP52dE8V/GT9if/gr\n7/wU++KfiL9kD9q740+Bv25Ph1+wpefs4eMf2f8A4San8SfAuteLP2evhR47+HXi3wN8SPHm\nnXEXhT4Jaje+JvFVobHX/itqfhbwlbaJZ3mu6prllYvYNe/kp/wTD+Bv7Vn7Bui/8EtP21Pj\nr+xT+1F8RPBfg39lT9vD9lX4qfCT4efBHxj4v/aC/Zb+JvxJ/wCCgvxe+Nng/wCMa/AG80/T\n/ihqPh34m/D+/XwlJr3hXw9eX0fgzUrfV7yKTSdY0OLWP7kaKAP5LfgP+yV8Tv25/jd/wcgL\n47/ZU+NX7JXgT9vH4WfsOJ+znf8Axm+HupfDy913xt4C+C3xlg8M/FK2u4kl0A+NfDnxOsPA\nXxN8a+G9J1i88U/DvXNa0nRPiJb6J45OqWafaX/BDXxV8T/2ytN+NP8AwVL+PvhK/wDCPxK/\naA0P4Q/sy+AvDeswxG58KfDX9k3wvL4b+Kj6PLGdtppvj/8AbA8Q/tDa/LZwxwiTTdF8LNdR\nrdWpig/az42fDTVPjD8MPFXw30b4tfFP4G3/AIot7G1T4ofBW98GaZ8TfDVvbarY398vhfUf\nH/gj4ieF7CbWrG0uNAv7278Jahf2mlapfXGg3Wi6/Hpmuad5n4Q/Y++CXgH9ljwZ+xt4JtPi\nF4G+CPgLwZ4S8DeHIfhn8Y/i18HviPZaR4Ov9M1Wyu7f4y/B7xp4B+LGna3rWq6Yt94w1jRf\nGOl3njL+0desPEr6lpHiHWtOvgD6gr+Rzxv4C/Z58Af8FgP+Ct/xI/bg/wCCUfx5/bK8CfGH\nVv2GLj9m/wCJelf8EwvGn7aHgf7D4D/ZI8L+GfivF4P8bf8ACrPGOhaOieKH0XRdfh0HUgt5\nrXh240/Ux9u8PPDb/vJ/w7T/AGdf+ijft/8A/i2L/gqb/wDRkV9v+E/DOneC/Cvhnwdo9z4g\nvNI8J+H9G8M6XeeLPFnirx74qutO0LTrbS7G58TeOvHWs+I/G3jTxBPbWsUus+LPGPiHXfFX\niPUWudY8Q6zqmr3l5fTgH81/7THgfwF+1T+wv4K/Yd/YB/YR+Pn7Enwl/bY/ba8F/Bv9pqDT\nv2FPEP7G118Mvgb4e0Xwd8UP2hPjzq3hbUfhlonhbT4/GngLwh4T+CnhHx54z0aXR/H3iG4u\nvANrPq0/hW50u27v9oX9iT49/shftff8E7v25vhv8bv2zP22rjwB8X4f2Rvjb4G8W+A/gP4h\nvPBf7JP7RWial4f1rxppmg/sqfsw/BfXZvCPwf8AiPpXw98fa6muxeJdM0LTtNk1Sxh0i1t9\nYOof0a0UAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABX\nzB+2F8fPFP7NfwRb4leBvh94f+KXjbVvjB+zP8FPB/gnxZ8QNR+FnhbUvFP7TX7S3wk/Zt8P\n6h4l8f6P8PPixq3hvw/4a1b4sWnijWbrS/hz4t1G507RrmwstKe5u4pofp+viD/gon+zj4p/\nav8A2Vtf+C/g/wAIfCD4iatefGD9lT4j3nw2+PurajoXwb+JvhX4D/tXfBP48eOvhj8QtV0v\n4a/GK5tPD/xH8EfDXxD4JmaX4X+ONNln123g1jQLzSJb4KAeg/DL4n/HvStO1/WP2xvAP7MP\n7POkHxB8OPCfw/1T4ZftW+M/jTp3irxV8RfFS+BdI8M6/ffFL9ln9la28L+INf8AG2u+AfB3\nw40bRJfHWo/ELxV4vXw9bW2javBotn4m7Pxz+0p+zn8MLPxBqPxL+P3wU+Hmn+E/F83w+8VX\n3jn4qeBfCVn4Z8e2/wALF+Oc/gjxBc6/r2nwaN4vg+Cbp8YZvDWovbazF8LGX4gvZL4SYauf\ny7sf2CPH8HwE+Kfgv4T/ALB//BND9hf4jT/Gv9h/47eD7X9lv4ia1feCfjbrf7IP7U/gf9pW\n18IfHbxB4a/YN/Zu13wZpZ/4V+/hDwh4l0vwp8a7vR/+FkeLNbj8Oad/ZB0vxh0XhH9hL9oH\nxb8a9F+NPx+tP2dzJP8A8FUIv29Ne8CeEPE3jbx5oeifD7Rv+CRh/YN8F+GdF1Xxf8KvB7eJ\nfiX4T+Nmn+G/GUeq32g+E9DvfCWkt470+48PeLDZ/DizAP0j8F/tKfs6fEj4i+L/AIQfDv4+\n/BXx78Wvh99q/wCE9+F3gv4qeBfFPxF8EfYbyPTr3/hL/BOh67feJfDX2PUJorG6/tnTLL7P\neSx202yd1Q4fgX9rr9k/4o6L4z8SfDP9p79nn4i+Hfhz4d0nxf8AELX/AAL8avht4u0XwH4S\n1/RW8S6F4o8Z6r4f8S6hY+F/DuteHVbX9J1vXJ7HTNR0VW1WzuZrEGevz1/Zl/YV/aL+GWsf\nsKeAPiJB+zppfwl/4J4Dx4PAXxa+G3ibxzrHxv8A2kv+Ek+C/j/4F2R+KXgXXPhX4Q8NfCX/\nAISiw+IM3xg+Mv8AYvxY+NB+Ifxp8J+F9fhOhwW7vH+df7MH/BMf9pT9pH9gD9kDTvihpvwf\n/Z+bwB/wTS/Zw+E/wy8PfD7xb8b/AAb8QPiPrr/Ff9kH9r2eH9pm88OeEvhH4/8AgPq+ka9+\nzdafDHXYfhT4z8c+ONH1j4v/ABw+JOheK/DniLUdP8MRAH9FXhT9qn9mDx38L/Efxu8D/tH/\nAAF8ZfBfwe18ni34veFPjB8PfEXwv8LPpkcE2pJ4j8f6R4ivPCmiNp8V1bS3y6nq1qbSO4ge\ncRrLGWx7X9sv9kC9+G+k/GOy/as/Ztu/hDr3iDWfCWh/FS1+OfwwuPhvrXirw54Z8QeNPEPh\nnSfHEXih/DGo+INB8HeE/FPizWdGs9Um1LS/DPhrxBr19bQaXo2o3Vt+YXhT/gm18cNE+Cn7\nX7aXonwd8H/HT9o7xF+zFc6bpHj79qj9rb9vDS5/C37Nvj3TvGEdh4y/aG/bk8OfE/UtO8Wa\n/Zah430z4ceLvBf7LNmnwI1y98I+N4tD+K3iDwbobWFz4G/8E6/2idA+LvhXx78X2+DOoaBp\nv/BTe3/4KAanZH4z/Fb43+KYdOm/4JZeNf2Hf+EUHiL4j/BnwdP4l8aeEvjHB8OviT4d8TXL\n+GtGuvC97r8uj6D8P73wL4U8I+JAD9l/Avj3wL8UfCGgfEH4Z+NPCfxF8BeK7BNV8L+N/Avi\nPR/F3hDxJpkjyRR6joHiXw/eaho2sWDyRSxpeade3NuzxyIshZGA6yvlv9jz4G+I/wBnf4Re\nJfh34ouvDV1e6p+0z+2x8YtK/wCESmv59Hs/Bv7Rv7Znx7/aG+H2kN/aGkaJNb61ofgP4o+H\nNG8T2FtYyaVpniaw1fTtE1TXNGtbDW9Q+pKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK\nKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK\nKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKK8D/ai/aF8Gfspf\ns/8AxT/aC8elpfD3wz8L3Osrpkck8Nx4j1+6ng0jwj4Rs7mCy1E2d94v8V6jovhmz1CezlsN\nMuNVj1HU2g021u7iLnxWKw+BwuJxuLrQw+FwdCtisTXqO1Ojh6FOVWtVm9bQp04SnJ20SZ62\nQZFnHFGe5Lwzw9l+JzfP+Is2y7I8kyrBw9pi8zzfNsZRwGW5fhabaU8RjMZiKOHowbSlUqRT\naTufhF/wW7/4KxfFz9lb4h/Dz9nX9k/x/pXhH4lWWlHx78aPFcXhzwB46utCsNaie28B/DuO\n08U/8JXp+hazqFguo+OPFmna54JsdbXQNQ+Fer+G9f8A7I8Qa3Z3cv8AwRA/4KufGH9q34gf\nEn9nj9qzx1p3jH4jy6MvxD+D/i1vDPgLwRcavpWj/Z9O8deALm18KL4V07WNV0+C40jxj4Ss\ntH8FXusyaLa/E7UvEfiD+y9C0Gytv5Dvjd8ZPH37Qvxb+IPxs+J+sTa546+JXifUfE+v3klx\nf3FtbSXkgWx0PR01O91G7sPDPhrS4rHw54U0U3k9voHhrStJ0SxKWOn28aX/ANn344+Ov2av\njV8Nfjv8NrxbPxn8MfFNh4l0pJpb2Gw1WCEvbaz4a1n+zrqxvp/DvizQrnU/DHiSztry1lv9\nA1fUrJbiL7QXX+Ho+M+evxCXEcsxzOPDTzJU3kP1uu8Esmt9Uv8A2eqrwrxyw3+2Oooczx2q\nqKB/1NVf2Z/hTH6HU/BinwbwNPxsjwS8XDxY/sDKo8T1PExVP7f5Fxc8DHPo8Kzzv/jHFhZY\nn2UOFWqcsK66bf8AqbUV5P8AAr4yeDP2hfg78N/jZ8Pr+31Hwj8S/CWk+KdKeC7tr17Fr+3X\n+0tC1CW0kkhi1vw3qqX3h/X7LcJtO1vTNQ0+5jiubWWJPWK/uTD16OKoUcTh6kK2HxNKnXoV\nqb5oVaNaEalKpCS0lCcJRlF9U0z/AJX82yrMcizTMskzjBYjLs3ybMMZlWaZfi6bpYrAZjl+\nJqYTHYLE0pa08RhcTRq0K1N6wqU5ReqCiiitTzwooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACv4pf8Ag4Y/botP\njH8YdD/ZA+HesWuo/D74A6xJrnxLv9OudM1Cx1744T6ddaWdHiu7N71k/wCFU6BqmqeG9RhS\n8sLq28aeI/G/h3xBpAvPCenXA/pi/wCCk/7Z/h/9hz9lXx58VZ9Qjh+ImuWd74F+Cek/Y01G\nXVvirruk6i3h67uLGaN7N9D8Kpa3Xi7xG+ovb2U+k6JNo8M0us6vo+n3/wDnHaUlx478aLce\nLfEWpF9e1a+1zxl4uv3Gv62toWudb8W+J501XVtNm8S65HYx6nq4sJ9Yt9R8R6mq2EN2dQv4\nmb+afpBcaywuCwvBGW1o/W829nic4cJxUqOBjVi8Jg5SuvZyxteCq1VKUGsNQjGalRxbP9uv\n2QH0Y6OfcTZ79KLjbL6i4e8P3jcl8OqeJwladHMOKquBqR4g4ko03Tl9co8MZViHl+AlQp4m\nM87zSvVoSo5lw/BHIV+h83/BMv8AaW1X9kT4Pfti/DLw5efGjwH8VL3xtbal4c+F3h3xT4l8\nYfDqDwf4i1LwsL/xVosOjpf3Fjfav4e8TQ3Wp6JZ6homippdpLfau0es2TD9Dvg38Z/+CDvx\ns8MaF4A+P37Pvjv9mWf4WaFoun+Hvija6z428T678XobOOz07VJviBP8GfDcjyeLdYa2TXNR\nnv8Awe9jb+ZqcXh/xJo5ki0u+/r6+BWi/s/a9+zh4H8Ofs8nw3N+ztq/gF9C8Bf8INdTf2I/\nhG+t7vTrhbG8kdtRXUfOlv11efUnOupr39oPrbHWRemvzzgDwhyjih5o6/FeR5lReVKWBjke\nNrvNcFmDxOH/ANoxuXYzC4arSwlNU8RhJOvRlDErEU62HnBqE1/Y/wBLr9ov4h+BUOBo5T4A\n+KfBOZ0+PpUuK63ilwxlS4A4n4Qhkma2yjhjjPhvPc5wGYcQYyeNyfiCksqzGjiskqZPjMsz\nnDYiFTE4aX83H/BuZ+2B8NfCvhTxr+yZ8QPihDpfjjxp8SrnxX8F/Aur2Oqrb3ySeFbH/hLL\nXSPEb2TaFb3epzaTDd2HhibU7a7uruy1e80+ylu9SlN1/WRX+Zx+1d+z78VP+CeX7XviT4dx\natrWieI/hd4y0/xt8H/iHa4stQ1rwrFqp1n4bfEDS7iFBapqLRWduNVhszcWukeK9M1vQmlm\nk0ufP+iX+yt8YH+P/wCzX8B/jVczaFLqnxN+E/gTxf4ii8NXsWoaJpvi3V/DthceL9DsbmGe\n5Xb4f8UHV9Dnt5J5LizudPms7oi5glUfrvgdxPja+CzHgTOcL9VzXg9uhTT5+epg1iatKrSr\n+0qVP32DxX7pTp8lGeGq4eNOmvZSlP8Azr/aneBfC+V8T8GfSu8Ns+WfcBfSOpwzbFSgsKsN\nhOIXkuAx+Bx+WPC4PCWy7iXI7Y2pQxn1jMqGc4HOK2NxU3jaVHDe+UUUV++H+SQUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFZet63o/hrRdX8R+IdU0/Q/D/h/S7/AFvXNa1a7gsNL0fR9KtJr/U9U1K/unitrLT9\nPsoJ7u8u7iSOC2t4ZJpnSNGYAH8+X/BfH9kz4lftcR/sk+D/AIEWreNvjXoeq/G6/wBM+Elt\neaNpF1rHw5uPD/gW+8b/ABIutf8AEWr6NoWj6R4G8QaB8PvBzf2ldwjU9f8Aix4d06zmF3NH\nBc+Bf8E2/wDggTpei+F/E/xB/b00C8Xxz4i0jxV4P8J/CLQvF+lXmneEPD3ijw1ceHr/AMYe\nJNc8MSahaXvj0R6pqf8AwitlpOuXuieF1itdcvf7U8Qz2Vv4W/d79lzQ9Y8c3njL9rHx1pOo\naT4r+PFvo9t8OfDet2c9jq/w2/Zs8NzajdfCTwpf6ZdCK40bxV41Gs6x8ZPiTYXdnaa7pHin\nx9D8ONdm1Kz+Fvhua3+wK/O8f4XcJZpxbU4xzLC1sdmFWlRhUwWJqU6uVTrYfD0sLQxc8JOi\n5yr08PRpU4xlXeGvBVXh3W/eH9l8J/Tx+kNwH9HrB/Ru4Lz/AC3hbhDAY/M8TguJ8jweNy/x\nAwuXZxnGNz/M8gw/EOHzOGGo5XjM4zHHYutXpZVDOnSxNXL1m6yx/U1/JT+0Z/wbPXkOn6Bc\nfslfHS3v9UOq6knijSv2ibx9MsV0SaOJtHudA174beANUnfVLCZJrfVLPUfDqWuqQ3Ud7Z3u\nkyac2m6r+fX7af8AwTm/4KAfs1eK/gJ8Fvh5Y/HX43fD7wd4P064+Gvir4P6L441PwjoPxH8\nTa1rniD4gxWNj4ZjnTwXrkfijxFqlvZa1rS6ZrGueEIdJvZrw29ndWWk/wBJv/BXD/gqF4g/\n4J16B8JtL8AfDjRfHvxD+Ls3i260668X3WpQeD/Deh+C/wDhH4tRnv7TR7nT9V1nU9Su/Etj\nDp9laapp0NrBa313fXW42Vrd/ih/wSy/4Kl/t/fHv9vr4efD34gfEPUfiv8AD74p6h4oXxx4\nKk8LeHbPRPB+hWnh/V9Yj8S+HZ9F0ayuvC1t4XurOyQK97JYapZyvpeoRXur3+n30H4lxhkH\nhLguJqfB+Wwz7KeIM2x2UZfiamR+1r4PLliJYWtQo1qONxEI1KeLdXCYmssL9ZnRqUade0Zw\nnQrf6f8A0cfF39oXxN4HYr6RvGWJ8J/EHwg4A4W8Q+LcmwfipLA5XxJxm8mo57l+bZnlmY8N\n5PiK+CxmQU8DxDkmWz4glkuGzPB4/F5ZJ4jC4vDZrl325+23/wAEev2o/wBuP4UfshfELVPi\nx4N0v9pz4c/s9eC/hb8ZtL+KOq+Ix4d13UrBpddn1qHxF4V0PxVIni601bXddtvFN0nh65tv\nE84s7mPUrdNPhW6/Sb/gll8Mp/2XvghrX7Evi/UILz4r/s5eJNS1TxJe2k876L418FfGPXfE\nPjr4e/EvwXb3kNtf2/hDVlbxJ4BuYb63huLX4i/DP4h2K/a7K1sdU1L9Jdeg1a60LWrbQLyD\nT9duNJ1GDRb+6j862sdWls5o9OvLiHZJ5sFteNDNLH5b+ZGjLsbO0/wp/slfs3f8FMP2c/27\nY/2hvH3wW/aFWD4S+LNa8b/tI/ESfSddutN8ffCjT5JdS+LMXhvxHdhdL+NGpa/4Ztr+58L+\nFPCV14h1PXvEUWhR6fZ2l/b291YfpOZ4LK+AOLsjzXKeFM6z3NeLHRyHM83pYvE1aeBw9F5b\nRni69L2GIozxeKVKlia7qPCQqRweIqqvT5q7f8UcDcS8d/S5+jv4pcBeIXj94aeFPAP0fIZj\n4rcE+HmO4cybL8TxTm+ZR40zHDZBlWO/tTKcww/D+RVcbj8jyqnhKfEOKwVXiTKMBPLMYqWV\nQh/eFRWZomt6P4l0XSPEfh7VNP1zQNf0uw1vQ9a0m7gv9L1jR9VtIr7TNU02+tXltr3T9Qsp\n4Luzu7eSSC5t5o5oneN1Y6dftB/mYFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABX86f/AAVy/wCCrXwy+BnxY8K/sfar\n8OvEvxF8NWuteBvGn7Ttppms6f4fi8QeAWtZPFujfBnT21Cw1GPU7PxrfR+D7/4oxahbpoWv\nfC24134aMLiXxzqmqeEv3e+M3xW8O/BD4Y+Lvih4og1G/wBO8L2EDWehaJAl54j8XeJdXv7T\nQfBvgTwnp7yQjVfGXj7xdqmh+DPB2jiWOTWPE+u6TpkTrJdqR+QXxL/4Ih/BP9q7UvC3xt/a\nj8X/ABKsP2kfE/meJPjpqHwt8WaTH4S8R6xq16dQs/AmnJ4o8J65LD4U+E2gHTPhH4A1zSIf\nD+r6r4C8KaNqHieC/wDEs82pRfIcb0+MauScvA+IwWGzr65hnKpjo0ZQeCTn9YhS+s0q2HVV\nydJt1ab/AHCrKm1WdJn9GfRcxv0b8D4oe2+lNlHE2deGL4bzunSwnC9bM6OJhxPJYV5PXxzy\nXMMszaeAhRjmEIQwWLh/wqVMtnjYVMshjYv9gvg78VfCXxz+FPw7+MfgSe7uPB3xN8H6B418\nOvf2/wBj1CPS/EGnQajb22o2m+UWuo2YnNpf26yzJDeQTRxzTIqyN6TXIfD7wF4T+Fvgbwh8\nNvAejW3h7wX4E8OaP4T8LaJaF2g0vQtBsINN02zSSVpJpmitbeMS3FxJLcXMu+4uJZZ5JJG6\n+vqMKsSsLhljJUp4xUKKxUqClGhLEqnFV5UYz99UnV5nTUveUGlLW5+EZ9LJp55nM+HKWPoc\nPTzXMZZDRzWdGrmdHJpYys8rpZjUw6VCpj6eBdCGMnQSozxCqSprkcT5e/an/Y1/Z2/bP8Ha\nT4J/aF8AQ+M9M8O6nJrHhq/t9U1bw/4h8OahcRxQ3suja9oV5YalawalbwQwanYPNLp2opb2\nr3dpNNZWctv0vwA/Zd+AH7LvhHT/AAV8CPhZ4U+HukWFj/Z8l5penrP4l1mI3DXckviXxdqB\nu/E/iW6lunMzXGuatfyIRFFCYreC3hi98r+eH/gsD/wV9+M/7Cnxl8AfA74HeA/AOpazqvgD\nS/ib4o8VfErTNf1rT59N1nxF4k0DS/D3h/StD8ReF3jKSeFNQn1jV7i+u/MS7gsdPhtLm1ub\no/KcTY/hPg+liOMs5wOFo4lLD4GpmdDLqVfNKyqSUKGGjXhTVeUUo3tOrGEYU0m7Qgj9+8Du\nEfpA/SNx+UfRs8N+Kc+zDJJyzjinC8EZtxnj8r4DyyWDoyxOZ5zVyuvi5ZXRrynV5efDYCti\nquLxkpQpuVfEVD7K/wCCsv8AwUN8R/8ABPT4I+EPF/gTwPo3jb4h/EjxdceFPCqeLV1N/BWh\nrpmlvq+q6r4ht9F1LR9W1J/J+z2unaRY6tpUl1JPcXUmowxWDw3Pk/7An/BQn44f8FFv2Lv2\nlvEfhzwT4Y+H/wC098O9A8Z+DvBk/hue+i8Aa34+1v4d6hqnwz12wi8TXeqXegiLxSUs9X0f\nU9Z1+Czt7Oz1R9Wli1V7DT9z9mr9oj9jX/gtd+zavgT4x+BPDuv+NPCkeja/8U/gvqlzr2m6\np4F8VIup6NZeNfAXibT7rSdb/sTUo5dRj03XfDutf2jplhrD+G/E0kM97Jb3/wCjf7PP7NHw\nN/ZS+HyfC/8AZ/8Ah5pPw58F/wBp3Gt3WnadcanqN5q2uXdvaWd1rWu65rt9qmu67qs1nYWN\nmb7VtSvJ47Kys7GF4rO0toIvEwEOIOJOI6XEuVcV4TEeHuZZJOjRy/B3jjo4qpQlSlUjL6q1\nQxNDFOVWdaeJjicPUpvBVMKuSTP1HizFeEHgt4M4/wAE+P8AwA4hyj6YfBXijh80zPjDiJU6\nvDFTIcDmtDMKODr0VnkZZrkmbZDCnl+Gy7DZLVyPN8JjKfFGCz6r9Zowf8I3/BPD9qP4mf8A\nBPD9uPwtYftF+JPiz8Kfh3oL6x4K+O/w213TPE17Pb6Nd+EdSTwqupeAbkG6xoevXfhfxJpO\no6XZjUIdDjln0M6lpupy6Zq/+gv4f17RvFWg6J4o8Oalaaz4e8SaRpuvaDrFhKs9jqujaxZw\n6hpepWU6/LNaX1jcQXVtKvyyQyo44Nfz4/8ABSb/AIIXSfte/G/xz+0v8J/jDa+DPG/jDw3b\n3HiH4e+KfD81/ofibxj4T8I2nh/w1/ZXiy01WGbwjZeIbfQtD0nW5rnw94iTSJmvPEdtbaiS\ndDl/Z39knUfh1f8A7M3wNtfhPb65p3gHwv8ADbwx8P8AQtA8Vsh8Z+EH+GunxfD7V/AfjuON\n3Fp4/wDAOueGdS8GePNPZjLpvi/QtasJsTWzgeT4U8N8XcI1uJsiziCrcO0szq4rh/Mq+Mp4\njGYyNebhOTp069V0KM8PSw9erSqUcM6eNq4hwVb2k5U/0D6fvjT9Hn6Q+WeBviv4c4qWX+Me\nP4GweQ+L/BeV8PYvKOG+HK2V0IYrD0o43F5VgVmmY4bOMxzfK8Bj8HmOdU8XwzgcojiJZa8F\nhqWL+iKKKK/Yz/NsKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA\nKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA\nKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAoor5v/aW+MmqfDPwbZ+Hfhy2i6t8e/irrun/DH4HeFdUb7VbXHj7x\nPaancw+K/Eel21xb6hP4B+GXhnRvFHxX+IZtZYLybwJ4C8S2mjST69PplncptK12ld2V3a77\nLu/IqMZS5uWMpcsXKXKm+WKaTlKy0im0m3om1rqecMR+0h+02YsG6+Cv7H2vRyS8B9M+IH7V\n2r+HhLDEckRanoX7OvgDxLHdMCmo6Rd/GH4i2kkcmk+OfgDIsX2vXmHwY+FHh/4IfDHwl8MP\nDdzqOp2XhqxuDqHiHW5IbjxH4y8U61qN54g8a+PvFl5bw20Oo+MfiB4x1XXfGvjDVI7eBdT8\nTa9qt/5MX2nYvp9MkKKKKACvgL/goF/wT1+Dv7ffwov/AAl4xs9O8M/FDSdOkj+F/wAZINHT\nUfEfgO+N/Z6jLatEl3p7634X1hrM2OueHru8SF7W8ub3TJdO1uOz1O3+/aK4M0yvL86wGKyv\nNMLSxuAxlKVHEYatG8Jwkt001OnUg7TpVqcoVaNSMatKcKkIyX1vAnHfF/hlxbkXHfAef5hw\nzxZw1j6OZZNnOW1VTxGGxNGV3CpTnGeHxmCxNPnw2Py7G0cRgMxwVWvgcfhsThK9ajP+E39g\nr9p34C/8Ej/20fjV4H+IreLPjE66pf8AwD8cfFbwZCPD/hfwIvhnxlap4tv9J8CarY3/AIk8\nf2tp4h0VYrrVY9W8MXFtaaHf/wDCO6D4uj1bT7yv7sVZXVXRgyuoZWUgqysMqykcEEEEEcEc\nivxm+Kf/AAQu/Ys+Ln7TmqftLeJW+I0E3ifxg/j/AMbfCnTdb0KL4beL/Fd1f/2prN1qUFx4\nbufEttpvijVHn1HxNpdh4ito7+6vLz7FPplvcPb1+zQAAAAAAGABwAB0AHYCvzzww4a4l4Uw\n2d5Pm6y2nklHM6lThmjgZOrVhhK1bFVcTLEVpp1qkKnPhnRWLnVxUJrEQnP2KoRX9ifTp8bf\nBP6QGd+F3iP4eVONcZ4n5jwRhcJ44ZnxTQhgsBis/wAty7IsBktDJ8uw81lmEr4NYfPIZjLI\nMJgchxGHqZRXw2HlmE80qzK+KEx+zf8AtN+Vza/BX9sHXpJIeAmmfD/9q3SPDxlniADGLTdC\n/aK8AeGpLtQsenaRZ/GH4dXsksureOfj9Ekn2vXmHxn+FHh/43/DHxb8MPElzqOmWXiWxtzp\n3iHRJIIPEfg3xVouo2fiHwT4/wDCd5cQXMOn+Mfh/wCMdJ0Lxr4P1SS3nXTPE2g6Vf8Aky/Z\nvLb9SP4OPT6K+eP2aPiv4g+J3gG+0z4iW2naX8bfhN4jvfhP8ddC0qKW20u0+JHhux029fxH\n4ftLia4urbwR8T/C2r+Gfi38Oo7y4m1CPwB478NwawYdcg1SztfoegAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACv4Rv2eP2Tv+Ci3\nxB/4KP8Ahv4+eEtC+Jer2Oo/tD3njO7/AGqZrXWbX4V+MfhrZ/EHVdN1fxZoPjbVYLbSvFXg\nfxP4P0S+j8MaDpbXSa74Qv8Aw9BpOmDQ9X0WSb+vj9qbXta8a3Hg39k/wFq2oaP4x+P9vrbe\nPfE2h3k1hrPwz/Zw8NSaZbfGHxtp2pWskNzo3ivxNDrui/CD4Z6lZ3Ees6N44+Idn8QNMtNS\n0z4ceJY7b6r0DQdE8K6Fovhfw1pOnaB4c8N6Tpug+H9C0izg0/SdF0TR7OHT9K0nS7C1SK2s\ndO06wt4LOys7eOOC2toYoYUSNFUfE8Y8D4HjOpkVTG5hmWBeQ5j/AGjQWX1oUlXnek3Go5wm\n4zi6MVRr07VaMZ1lD+K2v6g+jf8ASm4q+jVgvFfBcNcI8FcUw8WODJcG5pPi/LcRj5ZZhrY6\nEauCVDFYeNfC1oZjXeZZTi1UwWZ1cPltTELlwMadXWooor7Y/l8KKKKACiiigAooooAKKKKA\nPin475+APxb8K/tY2G638A6zY6D8Gf2qIEJW0s/h/PrN4/wp+ON2p3Rwn4IeN/Eep6R401Em\nwsbP4N/E7x3408U6hc2fwo8NWdv9rVk6/oGieKtC1rwv4m0nTtf8OeJNJ1LQPEGhavZwahpO\ntaJrFnNp2q6TqlhdJLbX2nalYXFxZ3tncRyQXNtNLDMjxuyn5U/ZZ17WvBVx4x/ZQ8e6tqGs\neMvgBb6I3gPxNrl5Nf6z8TP2cPEsmp23wd8b6jqV1JNdaz4r8Mw6FrXwg+JmpXtxJrGs+Ofh\n5e/EDUrTTdL+I/hmO5APsGiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA\nKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA\nKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigArnPGPi/wz8P8Awl4o8eeNdbsPDXg7wV4e1rxZ4r8RarMLfTNB8N+H\ndOudX1zWdRuCCIbHTNMs7m9upcHZBC7YJGD0dfFHxe/4yG+Onhj9m+zH2r4X/Cd/CXxs/aWn\nUF7HXNUt9UOr/AP4EzyKVjn/AOEh8TaFL8YPiRpqTvLa+CfAngrwp4p0e98I/HWGRwDo/wBl\nnwh4n1G28Z/tIfFDQ7/w/wDFf9ombRNZPhXWoTDrPwo+C/htNTX4J/Bi+t23Np+seGtF17W/\nHHxG03zr2G0+NPxK+KEGmajd+HYtBjtfrSiigAooooAKKKKACiiigAooooAKKKKACvkv9qbw\nj4m0228G/tI/C/Q7/wAQfFf9nabW9YHhXRYjNrPxW+C3iRNMHxt+DFjbrhtQ1nxLo2g6J45+\nHOm+dZQ3nxp+Gvwvt9T1K08Oza9HdfWlFAHOeD/F/hn4geEvC/jzwVren+JfB3jXw9ovizwp\n4j0qYXOl694b8Radbavoes6dcAATWOp6ZeW17ay4HmQTI2BnA6Ovij4Q/wDGPPx08Tfs3Xg+\ny/C/4rv4t+Nn7NE7ZSx0PVLjVBq/x8+BMEjFo4P+Ed8Ta7F8YPhvprTxyXXgjx5408J+FdGs\nfCPwKmkT7XoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooA8i+Ovxd0z4G/C7xL8Rb7Sr3xLqFh/Zei+D/BWkzQQa58Q/iL4t1ex8K/Dn4c+H5brFrD\nrvjzxtrGheFdMur1otN0+51VdR1a4tNKs727gwv2cPhFqfwf+HC2Hi/VrLxR8WPHOu6p8TPj\nf41sIZobPxf8WvFy2sviW90pLsHUIPCPhy0s9J8AfDXStRlub3w18LfB/gnwrLdXC6HHK3kX\nh/8A4yO/aXvfGr/6V8F/2Sdc1/wd4B/j07xz+1BfaVe+Gvif43gdcxXumfAXwlq+q/BrQLqC\neSA/Evxp8d9G1vTYNa+G/hrUI/tegAooooAKKKKACiiigAooooAKKKKACiiigAooooA8F/aP\n+EWqfF/4cNY+ENWsvDHxY8Da7pfxM+CHjTUIZprPwh8WvCK3Uvhq81VLQDUJ/CPiO0vNW8Af\nErS9Oltr3xJ8LfGPjbwrFdWy65JMm98Cvi7pnxx+F/hr4i2Gk3vhq/1D+1NF8YeCtWmgn134\nefETwlq994V+Ivw58QS2ubSbXfAfjbR9d8K6pdWTS6bqFzpT6jpNxd6Vd2V3P65XxRr/APxj\nh+0xZeNEza/Bf9rfXdA8HePRymneBf2oLHSrPw38MfG87nENlpnx68JaRpXwb8QXVxcR24+J\nfgv4E6Poumz638SfEmoSAH2vRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFfNP7T3xN8VeDPCWgeAvhXc2sXx2+OniNfhf8AB+S6s49UtPDGrX2m32q+K/ix\nrmlSskF94W+DPgfTdf8AiLqun309hZeKdS0XQvh3b6la+IPHGgJcfSjukaNJIyoiKzu7sFRE\nUFmZmYgKqgEsxIAAJJwK+L/2b0f44+OPFP7Y2rI0vh7xZo9z8N/2XLWcPs079ne21az1PUvi\nZaRSKRDe/tM+LtG0z4hR3lvL5OqfB3wr8AEu7HTPEGmeIIZgD6R+FPwy8KfBn4ceDfhb4Jt7\nuDwx4I0Kz0PTZNSvJNS1rUmgUyX+veI9YnH2vXvFPiPU5bzX/FPiK/aTUvEPiLUtT1vU5p7+\n/uZn9BoooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACvPvit8MvCnxm+HHjL4W+Nre7\nn8M+NtDu9E1GTTbuTTda0x5gsun6/wCHNXgBu9C8U+G9Tis9f8K+IrBo9S8PeItN0zW9Nmgv\n7C2mT0GigD5p/Zh+Jvirxn4R1/wH8Vbi0k+O/wADPETfC/4wvaWcemWfibV7HTbHVfCnxX0L\nS4i0Fl4W+MvgfUtA+IulWFlNfWfhfUdb1z4eXGo3PiDwPr8dv9LV8V/tIo/wN8deFv2x9IR4\nvDvhXR7b4a/tS2sCnZqP7PFzq93qelfE68RAvm3n7Mfi7WdU+IEt5PKYdL+Dni39oAWlhqni\nHUfD0MH2kjpIiyRsro6q6OjBkdGAZWVlJDKwIKsCQQQQcGgB1FFFABRRRQAUUUUAFFFFABRR\nRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRR\nRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRR\nRQAUUUUAFFFFABRRRQAUUUUAFFFcZ8RviD4S+FHgLxj8TPHmrR6H4M8BeG9Y8WeJtVeKe5Nn\no2h2M2oX0kFnaxzXl/eNDA0Vjp1jBcX+o3kkFjY29xd3EMLgHzL+0/eXnxb8QeEv2OvDNzdQ\nn4uaXeeJ/wBoDWdNuJ7e58G/sxaRepp3izSxfWjxT6X4j+PetPD8GPCfl3el6ynhfUfi58QP\nCd82r/CeeGvsWzs7TT7S1sLC1trGwsbaCzsrKzgitrSztLaJYba1tbaFUht7a3hRIoIIkSKK\nJFjjVUUAfMf7Lnw+8WaVofiv4zfFjSm0n44/tCavZeOfHmi3E1veXHw38L2Vm9h8KfgXbXlv\nJc2xs/hJ4NmisPEP9l3cuga78W9b+KvxC0eG1Xx3cwj6koAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooArXtlZ6lZ3enajaW1/p9/bT2V9Y3sEV1Z3tndRPBdWl3azpJBc\n21zBI8M8EyPFNE7xyIyMQfj39mG8vPhLr3iz9jzxNdXUx+EemWfiX4AaxqU8s9z4y/Zi1e8f\nTvCWmfbrqSW41XxH8BdZjm+DHi1pbnU9Zk8M6d8JfiD4tvxq/wAWYIq+y6+XP2ofh/4s1TQ/\nCnxm+E+lPq3xw/Z81e98c+BdDtp7eyufiT4WvbNLH4rfAu5vbmS3thZ/FvwdC9l4eGq3UXh/\nQ/i3ofwq+IOsRXKeBLaEgH1HRXGfDn4g+Eviv4C8HfE3wHq0eueDPHvhvR/FnhnVo4prc3uj\na5Yw6hYyT2l1HDd2F2sM6xXunXsMF/p14k9jfW8F3bzQp2dABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFA\nBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFA\nBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFfFHj3/jIz9orRfg7b/6V8IP2aNY8H/Fb43yKC9h4s+OCpZeMPgH8\nHrhwHinj+HsR0b9pDxzYtJZ6hp2rp+zfLGdS0HxT4ksV9q/aE+L7fBT4Y6r4s0zRB4u8capf\n6R4J+FHgJbs2M/xC+LPjS/i0D4f+DlvFhuZNN07UdfvLa58U+IvstzaeDfBlj4k8baxGmh+H\nNTuIX/s/fCAfBP4Y6T4Qv9a/4S3xnqF7q/jP4pePZLT7DdfET4r+NNQn8Q/EDxrNZtPdPptn\nq/iG9u08OeHlu7my8HeErXw/4L0Z4tB8O6XbQAHtVFFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFFFFABRRRQB8UeAsfs5/tFa18Hbg/ZfhD+0trHjH4rfBCRgUsPCfxvZL\n3xh8ffg9buwSKGP4gxjWf2kPA1gsl5qGpatJ+0hJIdN0Dwr4bsW+168V/aB+EA+Nnwx1bwjY\na1/wiXjTTr3SPGfws8ex2Zvrr4d/FfwZqEPiD4f+NIbNZ7R9Ss9I8Q2VpH4j8PG8trLxj4Ru\nvEPgrWXl0HxFqltOz9nv4vt8bPhjpXizU9EHhHxxpd/q/gn4r+AmuzfT/D34s+Cr+bw/8QPB\nzXrQ2z6lp2m6/Z3Nx4X8RC1trPxl4NvfDnjXR0k0PxHplxMAe20UUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFfLP7Ufj7xbp+ieFPgp8JtWl0f43ftC6pqHgnwVr1nFBdXfwy8H2FpD\nd/Fn47S2txFc23k/Cbwhd/afDB1O1m0HWvi/4h+FHgPWZLaDxtHOgBx3w+/4yL/aJ1340XH+\nlfCH9nDVPGHwl+BcTgSWHir4zqb3wl8f/jNbBg8c0fgkLqf7OPgPUEFlqGm39p+0arf2n4d8\na+Hr0fa9cd8PfAPhP4V+BPB/w18CaTHoXgzwH4a0bwl4X0iKa4uRYaHoNhBpunW8l3dyz3l7\ncLbW8Zur++uLi+v7lpby9uJ7qeaZ+xoAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAr4o+IH/GOn7ROh/Gi3P2T4Q/tIap4Q+Evx0iUBLDwr8aG+xeEvgB8\nZbkKEjgj8bhtL/Zx8eag4vdQ1LULr9nFV/szw74L8RXrfa9cd8QvAPhP4qeBPGHw18d6THrv\ngzx54a1nwl4o0iSa4thf6Hr1hPpuowR3dnLBe2Nw1tcSNa39jcW99YXKxXllcW91BDMgB2NF\nfLP7Lnj7xbqGh+Kvgp8WdWl1j43fs9app/gjxrr15FBa3fxN8H31pNd/Cb47Q2tvFbW3kfFn\nwhafafEx0y1i0HRPi94f+K3gLR5LmHwRLO31NQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBV\nvr6y0yyvNS1K8tdP07T7W4vr+/vriK0srGytInnury8up3jgtrW2gjkmuLiaRIoYkeSR1RWY\nfHn7L1je/FfXfFv7Yviizuref4w6ZYeHPgNo2p28ttd+Cv2YNFvJ9S8FXjWF0qT6T4l+OerX\nN18aPGSS2ul61FoOr/Cv4eeLbB9V+ElrMK/7SDN8dPHvhb9jjSi0/hnxNo1r8TP2p7mFj5Vh\n+z5b6veaZonwsvWTeI7r9pzxfoureBbmyuIvs+q/Bnwb8f44L7S9ftvDk8/2oqqiqiKqIqhV\nVQFVVUYVVUYAUAAAAAADA4oAWiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKA\nCiiigAooooAKKKKACiiigD4y/ahsbz4T674S/bG8L2d3cz/B7S7/AMOfHnRtLt5bi78a/swa\n1eQal40vVsbVXn1bxL8C9Wtrb40eDkjttU1qXQNJ+Knw88I2C6t8W7qavsOxvrLU7Kz1LTby\n11DTtQtbe+sL+xuIruyvrK7iSe1vLO6geSC5tbmCSOa3uIZHimidJI3ZGVjYZVdWR1V0dSro\nwDKysCGVlIIZWBIIIIIJBGK+LP2b2b4F+PfFP7HGqs0Hhnwxo118TP2WLmZj5d/+z3cavaaZ\nrXwtsmfYJLv9mTxfrOleBLeyt4jb6V8GvGPwBSe+1TX7rxFPAAfatFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAV578V/id4V+DPw48ZfFLxrPeReG/BWiXWs38GmWj6lreqzR7YNM8PeG9IhIutd8V+\nJtVnsfD3hTw9Yh9R8Q+I9T0vRNNhmv7+3hf0KvinXB/w0f8AtL2fhBCLr4Lfsj67onivxwMh\n9P8AHX7Ud9pFl4i+G/g2eNsx3ulfALwfrWmfFzXLae3ltJPif43+CWr6LqVv4g+FniOwiAPQ\nf2Yfhj4q8E+D9d8cfFSCzX47/HHxE/xR+MgsrtNTsfDuuahpthpfhn4V6DqkYWG+8KfBrwRp\nvh/4baLqFnBY2vie50HV/iDd6dbeIfGviCW5+laKKACiiigAooooAKKKKACiiigAooooAKKK\nKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAr5q/ae+GXirxr4P0Lxz8KoLNvjv8DvES\nfFH4Ni9u00yy8R63p+nX2l+J/hXr2qSBobHwp8ZfBGpeIPhvrWoXkF9a+GLvXtH+INrp1z4h\n8E+H5bb6VooA89+FHxO8K/Gf4b+Dfil4KnvJfDXjbQ7XWrCDU7OTTdb0qWTdDqXh7xJpE5N1\noPirwzqsN74e8V+Hr4R6l4e8R6ZqmialDDf2FxCnoVfFGh/8Y4ftMXng98WvwW/a413XPFng\nbpHp3gX9qKw0i88RfEnwZAi4istL+Pvg/RtT+L2hW0EEVsnxO8EfG/WNa1K41/4oeHLCX7Xo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKAPBf2jfi3qnwj+HYuvB2l2PiX4s+Pde0v4Y/BDwdqD3CWXir4r+\nLY7xfD8Gr/YidRi8HeFtPsdZ+InxM1LTY7i+8PfC3wZ428TQW1ydFML7/wADPhHpXwO+F/hr\n4c6dql74jvNNGp6x4t8Z6tFDFrvxC+IXizV77xV8RfiP4ijtsWsfiDx/431nXvFusQWSxadZ\n3ury2Ol29ppltZ2kHgvwfI/aF+N/ib9pO8zdfDL4Xv4u+Cf7NEEgD2OtXVtqyaT8ePjtbIwC\nzHxV4q8Pp8JvhzqTwu0HgHwF4n8XeFNYvPCXx1uY5PtWgAooooAKKKKACiiigAooooAKKKKA\nCiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDyT46fCPS/jl8LvE3w5\n1HVb7w3e6kNM1jwl400iKCbXfh78Q/Cer2Hir4dfEfw7HdA2kmv+AfG+jaD4t0i3vUl069vd\nIisdVtrvTLm8tJ+f/Zw+LmqfF74c/a/GOlWPhn4teA9e1T4Y/HDwZp8s8ll4T+LPhKOzXxFb\naSbwjUJfB/iiwvtG+Ifw01PUore+8RfC3xn4J8Tz2tt/bSwp73XxT8XwP2evjj4Z/aUsz9l+\nGXxSbwj8E/2l7dMJY6LeXGqvpPwH+O1wgYpAfC3inX5PhL8R9Rjt0kufAfj3wr4u8VazZeEv\ngTbRIAfa1FFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRR\nRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRR\nRQAUUUUAFFFFABRRRQAUUUUAFfJX7U3i3xNq0Pgv9mv4Y63qHh/4pftETa1pc/irRJzb618K\nPgh4aXTG+NvxisbpAz6brei6Prui/D/4baoIb1LD4z/Ez4aX+oabeeHbHxA9r9O+IfEGh+Et\nA1zxV4n1fT9A8NeGdH1PxB4h17V7uGw0rRND0ayn1LVtX1O+uHjt7LT9NsLa4vL27ndIbe2h\nlmldURiPln9lnw9rvi+bxp+1V8QNJ1LRfG37QUeht4M8La7az2OsfDD9nXww2py/BrwFqGmX\nKQ3Gj+Kdct9d1z4u/EvTby2TWNF+IPxJ1jwHf3up6R8PvDD2wB9O+EvCfhvwH4V8M+BvBui6\nf4b8IeDfD+jeFfCvh3SYFtdL0Hw54e0620nRNG022T5Lew0zTbS2srSBflit4I0HC10NFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABXPeLfCfhvx54V8TeBvGWi6f4k8IeMvD+s+FfFXh3VoFutL17w54h0650nW9G1K2f\n5Liw1PTbu5sruBvllt55EPDV0NFAHyZ+y34t8SaVD4y/Zs+Jut6hr/xT/Z4l0XSofFOtztca\n18Vvgl4kXUz8E/jHf3Tqjalreu6RoWteAfiRqfk2cd78Z/hr8S73T9Os/D154fe6+s6+QP2p\nvD2u+EJ/Bf7VXw/0jUta8bfs/R62fGnhbQrWe+1j4n/s6+J20yX4y+AtP0y1Sa41jxToVvoW\nh/F74Z6bZ20msa18QPhtpHgLT73TNJ+IXieS5+p/D3iDQ/FugaH4q8Mavp+v+GvE2j6Z4g8P\na7pF3Df6Vreh6zZQajpOr6ZfW7yW95p+pWFzb3lldwO8NxbTRTROyOpIBsUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFeW/Gr4\nsaD8D/hh4t+J3iGz1HVrbw5Z2sWk+GtESGbxH428Xa9qVl4c8C/D3wna3EsEF94w+IXjPV9B\n8E+EdPlngjv/ABJr2l2bzQpM0qAHz18dQP2gfi74X/ZRscXXw+8O2nhz4z/tUTLh7W+8FQa1\neP8ACL4G3eSEm/4XN438NX/iDxzprrf2N58Hfhr4v8E+KtOg0/4veHbyf7Xr56/Zq+E+vfC7\nwDe33xBu9O1f41fFXxJffFf46+INKkmuNM1H4meJbHTLK40TQLm5htrufwV8NvC+jeGPhN8O\nGvbeHUR8PPAfhY6v5+tNqV5dfQtABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFfFHwKA/Z9+Lvij9lG+xa/D7x\nFaeI/jP+yvKxCWtj4Jn1qzf4u/A20wSkP/CmPG/iaw1/wNpqJYWNn8HPiV4Q8E+FdPnsPhF4\njvIPtevnr9pb4T698UfANlf/AA+u9O0j41fCrxJY/Ff4FeINVkmt9L074l+G7HU7GDRNfura\nG5u4PBPxK8L6z4n+E3xHayt5tR/4V5488UNo4g1pNNvLYA+haK8t+CvxY0H44fDDwl8TvD1p\nqOk23iOzuotV8Na2kMPiPwT4u0LUr3w545+Hviy1t5Z4bHxh8PfGek694J8XadFPPHp/iTQd\nUs0mmWASP6lQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAV8UQf8AGR/7Tj3Zxd/BP9j/AFya0ss5fTvH/wC1fq2gPBqN4pBWHUtC/Z18\nAeI30q3cHUtHvPi/8R9bhmi0nxx8AraSP0z9pn4q+JPhx4F03QvhrFp198cfjD4ktPhP8DdL\n1S3e90tfH2v6fqWoXHjLxDYxS28t54K+E3hDR/E/xa8eWkNza3mpeEfBGraLo87eIdV0a0uu\n/wDg58KvDfwS+GXhD4XeFZdQvNK8KadJDPrWtTx3niLxVr+p3t1rfi3xx4s1GOGAat4y8d+K\n9S1rxl4y1poY5db8Ua5q+rTqJrySgD0yiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo\noAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigD4om/wCMcP2n\nEuhi0+Cf7YGuQ2t5jKad4A/av0nQFh0+7bJaHTdC/aK8AeHF0udydM0ey+L/AMONFhhj1bxx\n8fbiST7XrzP4x/Crw58bfhl4v+F3iqXULPSvFenRwwa1os8dn4i8Ka/pl7a634S8ceE9Rlhu\nF0nxl4E8V6bovjLwbrSwyy6L4o0PSdWgQzWcdcB+zN8VfEnxI8C6loXxKi06x+OPwf8AEl38\nJ/jnpelQPZaW3j7QNP03ULfxj4esJZbiaz8FfFjwhrHhj4teA7Sa5urzTvCHjfSNF1idfEOl\nazaWoB9F0UUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRXx1+1LrOreP77wd+yR4I1TUNL8S/Hi01m8+J/iTRLuex1b4cfs0+G5tOtPix4kstTtJ\nILnRfFfj5tZ0j4M/Di/s7y017SfEPjrUPiToEWpW3wq8RQQAGb+z/wD8X9+KXin9rq/P2nwP\nDpms/B/9lSB/mtZfhRBrVlc/Ef412WP3crftB+N/Dmkv4W1CKW/07Uvgp8NfhP4s0G4sH8fe\nKbKf7XrN0bRtI8OaPpXh7w/plhomg6Fptjo2iaNpVpBYaXpGkaXaxWOm6ZpthaxxW1lYWFnB\nDa2dpbxRwW1vFHDDGkaKo0qACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKA\nCiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAr4o/aA/wCLBfFLwt+1\n1YE23gebTNG+D/7VcCZW1i+FM+tXtz8OPjXe9Y4m/Z98b+I9WfxVqEslhp2m/BX4lfFfxZr9\nzfP4A8L2UH2vWbrWjaR4j0fVvD3iDTLDW9B17Tb/AEbW9G1W0g1DS9X0jVLWWx1LTNSsLqOW\n2vbC/s55rS8tLmKSC5t5ZIZo3jdlIBpUV8dfsta1q/gG+8Y/skeONU1DVfE3wGtNFu/hj4k1\nu7mvtW+JH7NPiSbUbP4T+JL7U7qSe51rxX4CbRdY+DXxIv7y7ute1bxF4FsPiTr8WnW3xV8O\nwT/YtABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQBzHjb\nxp4W+HHg3xX8QfHGt2XhrwX4G8Oa34u8WeIdSdo9P0Pw34c0251fW9WvXRXdbbT9OtLm6m8t\nHkMcTCNHcqp+dv2WvBnim4sfGP7Q3xU0S90D4v8A7RN1o3iHUfC+roq6v8KfhL4fi1GP4K/A\n+6RWlFnqHgfQNb1bxP4/sYbq/sF+NvxC+LV/o17JoOoaTBbcz8VR/wAND/Hnw3+z5abrn4V/\nBW78IfGj9oy4Qb7DxD4vt79Nf+AHwLnmUiObztc0qP45/EjTY50vNP8ADvhH4U+H/EGmaj4O\n+Nkyv9rUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAfJ/7U3gzxTb2Pg79of4V6He\n+IPi/wDs7XWs+INO8L6Qitq/xW+EviCLTo/jV8D7VC0QvNR8b6Bomk+J/AFhNdWNg3xt+Hvw\nmvtZvYtBsNWgufonwT4z8LfEfwb4U+IPgfW7LxL4M8c+HNE8XeE/EOmu0mn654b8R6bbavom\nrWTuqO1rqGm3dtdw+YiSCOVRIiOGUdPXxR8Kv+Md/j14k/Z6u/8ARfhX8arrxf8AGn9nG4f5\nLHw/4unv3179oH4FwSPmOD7PrmrP8cvhvpr3D3d/4c8YfFXw/wCH9L03wd8EoVQA+16KKKAC\niiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo\noAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAC\niiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooo\noAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK8g+O/xdsfgh8MPEHj6fSL\nnxRrMEuk+HfAvgfT7mG01b4ifErxjq9l4V+HPw90e6uFe2s9R8aeM9X0bQI9UvVXS9Dhvp9d\n1qe00XTNQvLf1+vinwsR+0d+0nqPxEkxd/Br9lDXvFHgL4YAhmsPGv7Slxp1/wCEfjD8RrZh\nmG7sfgn4d1LXPgJ4YvIJh5fjzxN+0TpGs6ebnwv4W1JAD2H9nX4RX3wc+G9vpHifWLTxV8T/\nABdrOq/Eb40+ObS3ktofG/xb8YPDe+LtZsYbkvfWvhfS2isvB/w60S+uLufwj8MfDHgrwZFd\nTWXhy1I91oooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigA\nooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK8J/aL+EN98ZPhtPpHh\nfWLTwp8UfCGs6V8R/gr45vLea5g8EfFzwe8154R1jUIbVor688LaqZr7wd8RtDsbi0m8XfDH\nxR408GS3UNn4iuifdqKAPH/gP8XbH44fC/w94/h0e68La1PJq3h7x14H1G5gu9X+HfxK8H6t\ne+FviN8PNZurYLa3mpeC/Gekaz4fl1Sx36VrkVjBruiT3mi6np15cewV8VeJwP2cv2k9P+IU\nP+ifBv8Aav13wz4E+JqgFdP8FftK22nWPhT4Q/Ea5c7YLOx+NnhzTNF+A3ie8nmJm8f+Gf2d\n9H0bTvtfirxRqT/atABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUVHLLFB\nFJPPJHDDDG8s00rrHFFFGpeSSSRyEjjjQFndiFVQWYgAmgD5o/ag+JPizwr4W8O/DX4T3sFp\n8d/j14gb4afCi8mtItSg8GPNp91qnjr4x6xps4Ntd+H/AIM+BrPWvHDWOpPaaX4r8V2fg74a\nHUrLWPH+iGX1/wCF/wANvCfwe+Hng34X+BrKaw8KeBtA0/w9o0V5dz6jqVxBYwhJtT1vVrtn\nvtb8Q6zdG41fxFr+pSz6pr+uX2oazqlzc6hfXM8nzR+zRFL8bPGPiv8AbJ1qOR9G8e6Q3gD9\nmOxuoyp0P9myx1O31MePbaKVWe21L9pjxVp1n8U7u5ge3Oo/CjRvgBpGtaZYeJPCGrCX7ToA\nKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDg/ih8NvCfxh+HnjH4X+Ob\nKa/8KeONA1Dw9rMVpdTadqVvBfQlYdU0TVbVkvdE8Q6Ndi31fw7r+nSwapoGuWOn6zpdzbah\nY208fkP7MHxJ8V+KvC3iP4bfFe9gu/jt8BtfT4a/Fi8hs4dLg8ZSw6da6r4H+MOj6Zb/AOi2\nnh74y+B73RfG8dlpr3el+FfFN54v+Go1C71nwDrYh+mq+LP2l4pfgn4y8Kftk6LHImjeA9IX\n4f8A7Ttjaxljrn7Nl9qlxqf/AAn1zHEqvc6l+zP4r1G8+KdrczyXB074T61+0BpOjaXqHiTx\ndpCxAH2nRUcUsU8Uc8Ekc0M0aSwzROskUsUih45I5EJSSORCGR1JVlIZSQQakoAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK\nKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK\nKKKACiiigAooooAKKKKACiiigAooooAK+L/2nbi5+Mfibwp+xv4euJ0tfifpMvjH9o3VLCWW\nKbwt+zHpmoNput+GHu7dhLp+u/tF+IIW+DugxmbTtRl8A/8AC7vGXhbU4df+GkKN9M/Er4ie\nE/hH8PvGfxP8d6kdJ8H+AvDereKfEV9HbXF9cx6Zo9nLeTxafp1nHNfarqt2IhZ6To+nQXGp\naxqc9ppmm21zfXdvBJ4t+y98O/Fmg+HfE/xa+LGmjTfjp+0BrVt8QfiRpUlxbX7/AA+0mKwT\nTvhr8ELC/tZbi0m0r4O+ClsfD2pzaVcDQvEvxJu/iR8StPtLK5+IGoxEA+mra2t7O3gtLSCG\n1tLWGK2tra2iSC3treBFihgghiVY4YYY1WOKKNVSNFVEUKABNRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABUNzbW95bz2l3BDdWl1DLbXVrcxJPb3NvOjRTQTwy\nq0c0M0bNHLFIrJIjMjqVJBmooA+L/wBmK4ufg54m8V/sb+Ibi4e1+F+lReMf2ctUv5ZZZvFP\n7Mep6gNN0Pwyl3cs0uoa7+zr4gmX4O6/H52palH4CHwR8Z+KdSm1/wCJkqL9oV8v/tRfDvxZ\nrvh7wv8AFv4T6aNS+On7P2tXXxB+HGlR3FtYv8QdJlsH034l/BC/v7qW2tYdL+MXgpr3w/pk\n2q3H9heGviVZ/Db4k6haXtz8P9OiHtPw1+InhP4ufD/wZ8T/AAJqR1bwf498N6T4p8O38ltc\nWNzLpms2cV5bxX+nXkcN9pWqWglNpq2j6jBb6lpGpQXemajbW19a3EEYB29FFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUV4d+0P8X5/gt8M9Q8SaJosXi34ha/qekeAvhB4Ce6eybx/wDF\nvxpdjRvAvheW7hiuLjTtDfVJRrPjbxFDaXcXgv4f6P4s8cajAdI8NajJGAeM+NP+Mjf2kNJ+\nFMI+1/Bz9l3WfCPxN+MUinfp/jD9oJ4LHxh8DPhPc4EkdxB8KdOn0X9ozxnYyC0u7HxRf/s1\nX9hdXun3HinTV+168Z+APwhh+CPww0TwTNrUvizxTNdav4s+JPjy6tFsdQ+IvxS8Z6pdeJfi\nJ47vrNZrhdP/AOEk8U6lqN5pmhQXEuneFdB/sjwloa2+gaFpdpb+zUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABXxR4L/4xy/aP1f4UzD7J8HP2odZ\n8XfE74OyMdmn+D/2gVgvvGHx0+E9tkRx28HxV06HWv2jPBljH9su77xPYftK39/c2OnWvhbT\nW+168Z+P3whh+N3ww1vwTDrUnhPxTDdaR4s+G3j21tFvtQ+HXxS8Gana+Jfh547sbNprZdQ/\n4RzxTp2nXeqaFPcRad4r0A6v4S1wXGga9qlpcAHs1FeH/s8fF+b40/DPT/Emt6LF4S+IWgan\nq/gL4v8AgJLp71vAHxb8F3baN468LxXc0VvcajoiarCdY8E+IpbS0i8Z+ANY8KeN9Ng/sfxL\np0knuFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUU\nUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAB\nRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAV8UfDb/AIyK/aD8QfHS5xdfCT9n3UfGHwe+\nAELYksfE3xSimuvC37QfxsgRt0cq6Bd2t7+zv8PdQMVpqOmR6J8fLyzuNU8J/FDR7k9b+1J4\n68VwaV4Q+BPwo1m50P40/tEajqnhHwv4j05YptR+Fvw+0i2tLn4xfHYQyx3MMb/DHwtqVrae\nDZ7+zutEv/jT4w+EfhLXFi07xVNcRe++AfAvhT4X+B/CHw38CaNbeHfBXgPw1onhDwpoVo0r\n2+k+HvD2nW+laTYRy3Ek1zP9msbWGJri6mmurh1ae5mmnkkkYA62iiigAooooAKKKKACiiig\nAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK\nKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAPij4k/wDGOv7Qnh/4522L\nX4SftBaj4P8Ag98f4VxHY+GfinLPa+Fv2fPjZOi7Y4l8QXd1Zfs7/ELUPKu9R1NNb+AV3d3G\nl+E/hhrNyPteuS8feBfCnxQ8D+L/AIb+O9GtvEPgrx54a1rwh4r0K7aZLfVvD/iHTrjStWsJ\nJbeSG5g+02V1NGlzazQXVs7LPbTQzxxyL4H+y3478V3GleLvgT8V9ZuNc+NX7O+oaX4R8U+I\n9RSKHUPil8P9Xtru5+D3x2MMUdtDK3xP8LabdWvjKfT7O10PT/jT4P8Ai54R0JZtO8KRXEoB\n9U0UUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAVS1LUtO0bTtQ1jV7+z0rSdKsrrUtU1TUbqGy0/TdOsYJLq\n9v7+8uXit7Szs7aKW4urq4kjht4I5JZXSNGYXa+K/wBonPx4+Inhf9j3Ts3HhDVdI0/4qftV\nXEeTBD8DYdYvdP8ACXwhvW2vH537SnjPQdX8NavYXEM9lq3wT8A/HTSbibStX1TwvdzAFv8A\nZc03Ufidq3i/9sDxdYXllqfxr0/S9G+C2havazWmoeAv2YtBubu/+HcE+nXaRXOj+KfjBfah\nqHxp8eQXFnpmvWUPibwJ8NfFdvdXXwh0q4T7JpAAAAAAAAAAMAAcAADgADoKWgAooooAKKKK\nACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAo\noooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAr41/aj03U\nfhhq3hH9sDwjYXl7qXwU0/U9G+NWhaTazXeoePP2YtdubXUPiHDBp1oktzrHin4P32n2Hxp8\nB29vZ6nr17D4a8d/DXwpb2138XtVuH+yqQgEEEAggggjIIPBBB4II6igCnpupadrOnWGsaRf\n2eqaTqtla6lpep6dcw3un6jp19BHdWV/Y3ls8lvd2d5bSxXFrc28kkM8EiSxO8bqxu18U/s6\nk/Af4i+Kf2PNSJt/COk6PqHxU/ZVuJDtgn+Bc2sWWneLPhDYk7V879mnxnr2j+GdKsbeCCy0\nn4KePvgVpVvNqmr6Z4nuoPtagAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK\nKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAPOfi58UfDHwW+G3i/wCKPjFr99B8\nH6S+oS2Gj2h1LxBr+ozzQ6foPhPwtpCOk2ueL/GGv3mmeFvCHh+1b7b4g8TaxpWi2KveX0CN\n5t+zL8L/ABR4C8F6z4s+J62Mvxz+NXiSf4q/GqfT7v8AtHTtK8V6vpunaXo3w68P6iFjS98I\nfB7wTpHhn4WeF9QhtrH/AISKw8Jt401OxTxL4p1+5uvN9T/4yP8A2mbfw+pF38FP2Q9d03Xf\nEwBEmnePf2qtS0S21fwb4dnjJMOoaL+z34I1+z+IGp288F1p0/xb8ffC/VNKvrDxb8FdZtIv\ntegAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKAPm/wDab+F/ijx74L0XxZ8LxZRfHP4KeJYPit8FZr+7Gnadq/ivSNN1LS9a\n+HPiHUSsiWXhD4w+CtX8S/CzxRqE1tff8I7Y+K08aaZZSeJPCugXFr6X8JPih4Y+NHw38IfF\nDwe18ug+MNIj1GGw1e1Ona/oGowzTWGveFPFOks8k2h+MPB+vWmp+FvF/h66YX3h/wAT6Rq2\ni36R3lhPGvotfFemf8Y4/tMXHh5sWnwW/a613U9e8MDiPT/Af7VGm6Jdaz408OwxooisNG/a\nC8EaBe/EPTYYorXT4Pix4B+KOqarfX3iz40aNaSgH2pRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAV8//ALSfxb1n4TfD\nyH/hBtO0/wAQ/GL4keIdN+FnwM8J6m0/9n+Ifip4qgvpdKuNaS0Zb/8A4QvwNoemeIfij8Tb\nrTRLqWlfC/wL4z1jT7e6vNPgtZvoCvij4LZ/aD+NPin9p++xdfDbwEnij4KfsvQPl7TVNOtt\nXgsvjf8AHa2UlYpj8RfGXh22+Hnw81Ew3SD4YfDh/GvhDWZPDnxz1e1lAPoD4IfCTRvgd8L/\nAAv8NdG1HUNebRYtR1DxF4t1oQf8JD498deKNXv/ABT8QviN4ne1jhtZPFHxC8ca14g8Z+I3\ns4bex/tnXL1bC1tLJbe2i9XoooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoo\nooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKA\nCiiigAooooAKKKKACiiigAooooAKKKKACiiigAryn43/AAl0b44/DDxR8NdZ1DUNCbWYtO1H\nw74s0YQf8JB4D8deF9XsPFXw9+IvhlrlJbWPxR8PfHGi+H/Gfh1ruGey/tnRLJb+2urJri2l\n9WooA8A/Zt+Les/Fn4ezHxxpun+HvjF8OPEWp/Cz45+EtMac6f4d+KfhWCxl1WfREvHbUP8A\nhC/HOh6n4e+KHwyutTEWpar8L/HXgvWdQt7W81Ce1h9/r4o+NP8Axj58afC37T1ji1+G3j1P\nC/wU/aggTKWmlafdavNY/A/47XKAtFCfh34y8RXPw7+IWpLDaofhh8Ro/Gvi7WovDfwM0m1i\n+16ACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAK\nKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiii\ngAooooAKKKw/E/ibw94K8NeIfGXi3WdO8OeFPCWh6t4m8TeIdYuorHSdB8PaDYXGq61rOqXs\n7JDZ6dpem2lzfXt1MyxW9tBLLIyohIAPmL9qbxR4i8QjwX+zD8Nta1HQfiP+0M2t2WteKtDu\nXtNa+FfwD8Mf2YPjX8VNPvYiJtL8Qx6br2h/DH4aapALi40n4t/E3wN4jl07UPD/AIc8SfZv\npzwt4Y8O+CPDPh3wZ4Q0XTvDnhPwjoWkeGPDHh7R7WKy0nQfD2gafb6VoujaXZQqsNpp2l6b\naW1jZWsSrHBbQRRIoVAK+Yv2WfDPiDxIfGn7UXxG0fUtD+IX7Q39hXXh7wrrtrNZaz8LP2f/\nAAwNTb4K/DDUdOuFjn0nxLJY69r3xT+JmmXMb32kfFT4m+MPCbahqWg+EvDLW311QAUUUUAF\nFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRR\nQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAF\nFFFABRRRQAUUUUAYXinwv4d8b+GfEfgvxfouneJPCfi7QtX8MeKPD2sWsd7pOveHtf0+40rW\ntG1SymVobvTtU027ubG9tpVaOe2nlicFXIr5j/ZZ8UeIvD6+M/2YfiTrWo698SP2eW0Sy0bx\nVrly93rXxV+Anib+0x8FPirqF7KTNqniGXTdC1z4Y/EzVJhbz6r8W/hl468RQ6dp/h/xF4b+\n0/XNfIn7U3hnxB4aPgv9qL4daPqWufEH9nka7deIvCuhWs97rPxT/Z/8TjTH+NPwx07TrZZJ\n9W8SxWOg6D8U/hnpltEl/q/xT+GXhDwkuoaboPi7xO1yAfXdFYfhjxN4e8a+GvD3jLwlrOne\nI/Cni3Q9J8TeGfEOj3UV9pOveHtesLfVdF1nS72BnhvNO1TTbu2vrK6hZori2niljZkcE7lA\nBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFA\nBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAV8\nUfGz/jIL4zeF/wBluy/0n4ceCoPDHxr/AGop0w9nqmh2+s3Nz8FfgTdNlUn/AOFoeM/DV744\n+IOmhr23f4VfDe88EeL9I/sL43aJdzfQPxu+LWi/A74X+KviZren6jrg0K3sbPQPCmi+Q3iL\nx5438SarY+GPh/8ADrwtFdSQ20/iv4h+NtY0DwX4YguZoLWXXdcsEu7i2tjNPHyn7Nnwl1r4\nU/D64k8d3+na78ZPiZ4i1H4qfHXxRpZnfTdd+KXii10631Kz0OW7jivm8F+AtA0nw58LPhpD\nqCf2lY/DPwJ4P0/U5bnUbW7u7gA+gaKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigA\nooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKK\nKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA+KPgl/xj78Z\nvFH7Ld7m2+HPjWDxP8a/2XZ3wlnpehXGtW1z8afgTatlo4B8LvGfiWx8b/D7TSbG3j+FPxIs\nvBHhDSToXwR1u7h+16+fv2k/hLrXxW+H1vJ4Ev8ATtC+Mvwz8Rad8VfgV4o1QzppuhfFLwva\n6jb6bZa5LaRTXq+C/H3h/VvEfws+JcOnx/2lffDPx34wsNMlttSurS7t+r+CPxa0T44/C/wp\n8TNEsNQ0Ma7b31nr/hTWvIXxF4D8b+G9UvvDHj/4deKYraSa2g8WfDzxto+v+CvE9vbTT2sO\nu6Ffpa3FzbCGeQA9WooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK\nKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig\nAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK\nKKACiiigAoor5w/ab+KXif4feCdI8L/DD+z7j46fGjxLb/Cn4I2ep2v2/TLDxjrWnalqeqfE\nDxDpwkg+3eDfhB4L0jxN8VvGFgbqyl13RfB83hPSrseJPEehW10Aeb2Gf2j/ANpubWj/AKV8\nFP2Qtcv9H0A/f07x/wDtV6poc+meKdbjYYh1HRP2ePAviC58GafNFLfaXdfFz4hfECw1Cz03\nxl8DdLuYvtevN/hD8LfDHwU+Gng/4XeD/t8uh+ENKFiupaxdDUPEPiPVbq4n1PxH4w8VaqY4\npNb8Y+NPEd9qvizxj4guEF5r/ijWdW1m9L3d9M7ekUAFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFA\nBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFfFF/n9nD9puHWh/ovwU/a91yw0fXj9zTvAH7Vel6HBpvhfW5Ccw6don7Q/gXQLbwbqE0\nkthpdr8XPh54AsNPtNS8ZfHLVLmX7Xrzf4v/AAs8MfGv4aeMPhb4v+3xaH4v0o2Tano10NP8\nQ+HNWtbiDU/DnjDwpqojlk0Txj4L8R2OleLPB3iC3Q3mgeKNF0nWbIpd2MLqAekUV84fsyfF\nLxP8QfBOr+GPid/Z9v8AHT4MeJLj4U/G6z0y2Nhpl94y0bTtN1PS/H/h7TjJObHwb8X/AAXq\n/hn4reD7AXV7LoWi+MIfCmrXZ8SeHNetbX6PoAKKKKACiiigAooooAKKKKACiiigAooooAKK\nKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiig\nAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKK\nKKACiiigAooooAKKKKACiiigAJABJOAOST0A9TXxR+zwP+F9/EnxR+2DqObnwdeaTqnwm/ZU\ngky1uvwUh1myvvG3xjseqNJ+0j418P6PrWhX8E93p+q/BP4d/BLX9OGmah4j8U2c+j+1Fqmp\n/EnU/CH7Ing7Ub3T9Z+N1jquq/GDX9IupbTUvh/+zJ4fuLKx+JWoW2oWskFzo/in4q3ep6Z8\nFvAF3a3mn69p0/i7xZ8S/DDXz/CPWrZPrfS9L0zQ9M07RdF06x0jRtHsLPS9J0nTLWCw03TN\nM0+3jtLDTtPsbWOK2s7GytYYra0tbeKOC3gijhhjSNFUAF+iiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooA+KP2hh/woX4k+F/2wdOzb+DrTSdL+E37VcEeVt2+Cs2tXt94J+Md90R\nZP2bvGviHWda12/nntNP0r4J/ET42a/qP9p6h4b8LWcP2uCCMg5B5BHQj1qhqulaZrumalom\ntadY6vo2sWF5pWraTqdrBfabqmmahbyWl/p2oWN1HLbXlje2s0ttd2txFJBcQSyQyxvG7Kfk\nf9lvVNT+G2p+L/2Q/GWo3uoaz8D7HStU+D3iDWLua91L4gfsya/cXtj8NNRutRu5J7rWPFXw\nrutL1P4L/EC6u7zUNe1G48I+FPiX4neyf4uaLbyAH2PRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQA\nUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFF\nFABRRRQAUUUUAFFFFABRRRQAVynjzxx4U+GXgnxd8RvHWtWvhzwX4E8N614u8V69feZ9k0fw\n94e0641XV9RnWFJZ5EtLG1nm8m3iluJioit4pZnSNurr4p+Jmf2if2gfDnwJtN1z8J/gFqPg\n/wCMn7QVxGWax8S/EmCeDxN+z78EJ5U/dzLpWo2dn+0J8RNPS4t7/TLPw58DdM1Sx1Twh8Wt\nTgIB1f7LXgjxWml+Lvj18VtFudB+M37RN/pXivxF4a1MRNqfws+G+jW95b/Br4FSvG9wkFx8\nNvC+qXupeN7W0vb7Rrn42eNfi94k0GaPSPElpbQfVdFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRR\nRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUA\nFFFFABRRRQAUUUUAFfKv7UfgfxU2meEPj38KtFutd+M37O99q3irw94b0zyk1T4p/DbWreyg\n+MvwKieQwrPP8R/DOlWOqeCbO6vLLR7f42eCPhD4j12d9I8OXdvP9VUUAcr4F8b+FfiX4K8J\nfETwNrNr4i8F+OvDei+LvCmvWXmi11jw94h0631XR9SgWeOKeNLywuoJxDcRQ3EJcxTxRTI8\na9VXxV8M8/s7/tAeIvgRdlrb4T/HvUfGHxj/AGfJ3G2w8NfEee4uPE37QXwQgmc7IhqmpXt5\n+0H8O9PknuL/AFKx8R/HDStKstL8H/CTTLdftWgAooooAKKKKACiiigAooooAKKKKACiiigA\nooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKK\nKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigA\nooooAKKKKACiiigDxr4+/F+3+B/wv13xymiz+LPErXGj+Fvh34DsrqOz1P4i/FDxnq1n4X+H\nXgHTruVJYdPl8UeLdU0vTb3W7qM6X4Y0iTUvFOuy2ug6Jqd5b0v2d/hBcfBf4a2fh/XdZg8W\n/EXxJq2r/EH4xePIrR7L/hPfi34zuf7W8beJILWeW4urDQI794/D/gTw/c3d4PBnw60Lwh4H\nsbl9K8NWCp414O/4yN/aR1b4oTj7V8HP2Wdb8V/Df4Qgnfp/jL9oeWzv/CHxv+KdtgPFc2vw\nj0m81v8AZ28IXsUtvcWni/Vf2lNM1SwuYIPCmpr9r0AFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFA\nBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUU\nUUAFFFFABRRRQAUUUUAFFFFAHhv7Q/whuPjP8Nbzw/oOtQ+EviN4b1XSfiB8HfHstrJef8IF\n8W/BtwdU8FeI7i2glt7u/wDD816svh3x74ftruzHjT4c694w8D31ymleJb9Xu/AL4vW/xw+F\n+heOX0Wfwn4kW41jwt8RPAl7dR3mp/Dr4oeC9WvPC/xF8A6jdxJFDqEvhbxbpeq6bZa3aRjS\n/E+kR6b4p0KW70HWtLvLj2Wvijxj/wAY5ftI6T8UIB9l+Dn7U2t+Ffhv8XgDs0/wb+0PFZWH\nhD4IfFO5yEitrX4uaRZaJ+zt4vvZZrie78YaV+zXpmlWFtBceK9TcA+16KKKACiiigAooooA\nKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACii\nigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooA\nKKKKACiiigAooooAKKKKACiiigAr5g/ai+Ivi3w94a8NfCn4Taimn/HX4/a5P8O/hnqYtYNR\nHgKwFjJqXxG+NmpafcLJaz6P8HPBEWpeKLK21RItF8UfECT4ffDS7vrK++IGlyN9NTzw20M1\nzczRW9vbxSTzzzyJFDBDEhklmmlkKpHFGis8kjsqIilmIAJr80/2bfj78Cfiz4y8Z/te+Kvj\nX8JLWP4gaefAP7O2iar8R/BtteeEv2bdH1Rb+y8TGxu9ZW40vX/2ivFFonxa17Men38vw9tf\ngZ4T8UaVaeJPh1ebgD75+Gnw68J/CL4feDfhh4F05tK8IeBPDul+GdAs5rm4v7wWGlWsdslz\nqep3kkt9q+s37o9/rWt6lPc6prWrXN7qup3Vzf3lzcSdxXjP/DR37PX/AEXj4M/+HQ8Ef/Ly\nj/ho79nr/ovHwZ/8Oh4I/wDl5QB7NRXjP/DR37PX/RePgz/4dDwR/wDLyj/ho79nr/ovHwZ/\n8Oh4I/8Al5QB7NRXjP8Aw0d+z1/0Xj4M/wDh0PBH/wAvKP8Aho79nr/ovHwZ/wDDoeCP/l5Q\nB7NRXjP/AA0d+z1/0Xj4M/8Ah0PBH/y8o/4aO/Z6/wCi8fBn/wAOh4I/+XlAHs1FeM/8NHfs\n9f8ARePgz/4dDwR/8vKP+Gjv2ev+i8fBn/w6Hgj/AOXlAHs1FeM/8NHfs9f9F4+DP/h0PBH/\nAMvKP+Gjv2ev+i8fBn/w6Hgj/wCXlAHs1FeM/wDDR37PX/RePgz/AOHQ8Ef/AC8o/wCGjv2e\nv+i8fBn/AMOh4I/+XlAHs1FeM/8ADR37PX/RePgz/wCHQ8Ef/Lyj/ho79nr/AKLx8Gf/AA6H\ngj/5eUAezUV4z/w0d+z1/wBF4+DP/h0PBH/y8o/4aO/Z6/6Lx8Gf/DoeCP8A5eUAezUV4z/w\n0d+z1/0Xj4M/+HQ8Ef8Ay8o/4aO/Z6/6Lx8Gf/DoeCP/AJeUAezUV4z/AMNHfs9f9F4+DP8A\n4dDwR/8ALyj/AIaO/Z6/6Lx8Gf8Aw6Hgj/5eUAezUV4z/wANHfs9f9F4+DP/AIdDwR/8vKP+\nGjv2ev8AovHwZ/8ADoeCP/l5QB7NRXjP/DR37PX/AEXj4M/+HQ8Ef/Lyj/ho79nr/ovHwZ/8\nOh4I/wDl5QB7NRXjP/DR37PX/RePgz/4dDwR/wDLyj/ho79nr/ovHwZ/8Oh4I/8Al5QB7NRX\njP8Aw0d+z1/0Xj4M/wDh0PBH/wAvKP8Aho79nr/ovHwZ/wDDoeCP/l5QB7NRXjP/AA0d+z1/\n0Xj4M/8Ah0PBH/y8o/4aO/Z6/wCi8fBn/wAOh4I/+XlAHs1FeM/8NHfs9f8ARePgz/4dDwR/\n8vKP+Gjv2ev+i8fBn/w6Hgj/AOXlAHs1FeM/8NHfs9f9F4+DP/h0PBH/AMvKP+Gjv2ev+i8f\nBn/w6Hgj/wCXlAHs1FeM/wDDR37PX/RePgz/AOHQ8Ef/AC8o/wCGjv2ev+i8fBn/AMOh4I/+\nXlAHs1FeM/8ADR37PX/RePgz/wCHQ8Ef/Lyj/ho79nr/AKLx8Gf/AA6Hgj/5eUAezUV4z/w0\nd+z1/wBF4+DP/h0PBH/y8o/4aO/Z6/6Lx8Gf/DoeCP8A5eUAezUV4z/w0d+z1/0Xj4M/+HQ8\nEf8Ay8o/4aO/Z6/6Lx8Gf/DoeCP/AJeUAezUV4z/AMNHfs9f9F4+DP8A4dDwR/8ALyj/AIaO\n/Z6/6Lx8Gf8Aw6Hgj/5eUAezUV4z/wANHfs9f9F4+DP/AIdDwR/8vKP+Gjv2ev8AovHwZ/8A\nDoeCP/l5QB7NRXjP/DR37PX/AEXj4M/+HQ8Ef/Lyj/ho79nr/ovHwZ/8Oh4I/wDl5QB7NRXj\nP/DR37PX/RePgz/4dDwR/wDLyj/ho79nr/ovHwZ/8Oh4I/8Al5QB7NRXjP8Aw0d+z1/0Xj4M\n/wDh0PBH/wAvKP8Aho79nr/ovHwZ/wDDoeCP/l5QB7NRXjP/AA0d+z1/0Xj4M/8Ah0PBH/y8\no/4aO/Z6/wCi8fBn/wAOh4I/+XlAHs1FeM/8NHfs9f8ARePgz/4dDwR/8vKP+Gjv2ev+i8fB\nn/w6Hgj/AOXlAHs1FeM/8NHfs9f9F4+DP/h0PBH/AMvKP+Gjv2ev+i8fBn/w6Hgj/wCXlAHs\n1FeM/wDDR37PX/RePgz/AOHQ8Ef/AC8o/wCGjv2ev+i8fBn/AMOh4I/+XlAHs1FeM/8ADR37\nPX/RePgz/wCHQ8Ef/Lyj/ho79nr/AKLx8Gf/AA6Hgj/5eUAezUV4z/w0d+z1/wBF4+DP/h0P\nBH/y8o/4aO/Z6/6Lx8Gf/DoeCP8A5eUAezUV4z/w0d+z1/0Xj4M/+HQ8Ef8Ay8o/4aO/Z6/6\nLx8Gf/DoeCP/AJeUAezUV4z/AMNHfs9f9F4+DP8A4dDwR/8ALyj/AIaO/Z6/6Lx8Gf8Aw6Hg\nj/5eUAezUV4z/wANHfs9f9F4+DP/AIdDwR/8vKP+Gjv2ev8AovHwZ/8ADoeCP/l5QB7NRXjP\n/DR37PX/AEXj4M/+HQ8Ef/Lyj/ho79nr/ovHwZ/8Oh4I/wDl5QB7NRXjP/DR37PX/RePgz/4\ndDwR/wDLyj/ho79nr/ovHwZ/8Oh4I/8Al5QB7NXD/Ez4deE/i78PvGXww8dac2qeEPHfh7U/\nDOv2cNzcWF4bDVLZ7d7rTNTs5Ib/AEfWdPdo7/Rdb02e21TRdWtrLVdMura/s7e4j5H/AIaO\n/Z6/6Lx8Gf8Aw6Hgj/5eUf8ADR37PX/RePgz/wCHQ8Ef/LygDhv2XfiL4t8Q+GvEvwq+LOoJ\nf/HX4A65D8O/ibqX2WDTh48sTYRal8OvjXpmnW6R2sGj/GPwRLpvim8ttLSbRfC/xAT4g/DS\n1vr2/wDh/qsifT9fmf8AtJfH34E/Cbxl4M/a98K/Gr4SXcfw/wBOHgH9onRNK+I/g25vPFv7\nNusao1/e+JlsrXWWuNU1/wDZ18UXb/FrQsJqF9F8PLr46eE/DGk3niX4i2W39K4J4bmGG5tp\nori3uIo54J4JElhnhlQSRTQyxlkkikRleORGZHRgykgg0AS0UUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAU\nUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFF\nABRRRQAUV8Wf8FDfjT4x+BH7H/xj8WfC2yv9W+NvinStH+Df7Puh6O+kLrOt/tA/HbxHpHwe\n+C9npaa9rHh7RpJbb4h+NdA1a+bVtf0HSLPR9N1LUdZ13Q9Isr/V7L+erTfCfjP4cfsTfGT/\nAIJraR8Mdf8A2cfDXwp/b6/4Jc6p8Nvg7+2D4d+Gfx11bxD+x1+2j+3f8A/DNzonjbR/2fv2\nnPFXw/8AFXgrXP2ndL/aG0vxp4cX4sazb+Lvgtaaj8LNb0nwTqXjmz8R+EwD+uWWKOeOSGaN\nJoZkeKWKVFkjljkUpJHJG4KujqSrowKspIIIJFeW/wDCifgh/wBEb+FX/hvPCP8A8p6/Dj4u\nft4/Hbwr4J8H/tU/GH4UfAH4meD/AIXf8FA/+CnngD4FeBfBXhP4k2HxW0XRP2C/2bf+CzGj\nS+NLfxfqPxJ1fw+vxZ+LPhv9mzw58LBa23gO/wDDmkaB4s+LGsw2N7qHxH8H6P8ABn6m8Tft\naftP/CT4NfGvWviJ8Y/2CPHvxL0nwV+yj4u+E178L/C3x6TS9KvP2p/ivd/CTwxoXjj4N+HP\nGnxn8f8AxB0rxD4ijsdL/Z88beDvG/ge3/aD8YT6p4X1fw58D9F0C98dSgH6Q/8ACifgh/0R\nv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPX47/AA8/b9/bT8f3HhX4QW+m/ADQ/jXd\n/wDBT3xN+wf4k8b+MPg7460DwnD4E03/AIJjeLP2+4viGPgj4c/aj+IOseHPHWiaza6L4RvP\nBN9+0Lr0Ou6Xouq6dqt38NPE/iaT/hXf6WfsUfGn4g/Hr4CL40+K1r4Ot/iN4Z+NX7VPwJ8Y\n3Pw90zW9D8E69rn7L/7U3xm/Zqu/F/hnw/4k1/xZrvhzSvGx+Ey+L7fw1qnirxPd+HBrh0ST\nxHrn2AapdAHrn/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/Cr/w3nhH/wCU9eq0UAeVf8KJ\n+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq0UAeVf8ACifgh/0Rv4Vf+G88\nI/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0Rv4Vf+G88I//ACno/wCFE/BD\n/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/Cr/w3nhH/wCU\n9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq0UAeVf8ACifg\nh/0Rv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0Rv4Vf+G88I//\nACno/wCFE/BD/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/\nCr/w3nhH/wCU9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq\n0UAeVf8ACifgh/0Rv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0\nRv4Vf+G88I//ACno/wCFE/BD/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yn\no/4UT8EP+iN/Cr/w3nhH/wCU9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/\nAMN54R/+U9eq0UAeVf8ACifgh/0Rv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtF\nAHlX/Cifgh/0Rv4Vf+G88I//ACno/wCFE/BD/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4\nVf8AhvPCP/yno/4UT8EP+iN/Cr/w3nhH/wCU9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4\nUT8EP+iN/Cr/AMN54R/+U9eq0UAeVf8ACifgh/0Rv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8A\nDeeEf/lPXqtFAHlX/Cifgh/0Rv4Vf+G88I//ACno/wCFE/BD/ojfwq/8N54R/wDlPXqtFAHl\nX/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/Cr/w3nhH/wCU9eq0UAeVf8KJ+CH/AERv4Vf+\nG88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq0UAeVf8ACifgh/0Rv4Vf+G88I/8Ayno/4UT8\nEP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0Rv4Vf+G88I//ACno/wCFE/BD/ojfwq/8N54R\n/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/Cr/w3nhH/wCU9eq0UAeVf8KJ\n+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq0UAeVf8ACifgh/0Rv4Vf+G88\nI/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0Rv4Vf+G88I//ACno/wCFE/BD\n/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/Cr/w3nhH/wCU\n9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq0UAeVf8ACifg\nh/0Rv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0Rv4Vf+G88I//\nACno/wCFE/BD/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yno/4UT8EP+iN/\nCr/w3nhH/wCU9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/AMN54R/+U9eq\n0UAeVf8ACifgh/0Rv4Vf+G88I/8Ayno/4UT8EP8Aojfwq/8ADeeEf/lPXqtFAHlX/Cifgh/0\nRv4Vf+G88I//ACno/wCFE/BD/ojfwq/8N54R/wDlPXqtFAHlX/Cifgh/0Rv4Vf8AhvPCP/yn\no/4UT8EP+iN/Cr/w3nhH/wCU9eq0UAeVf8KJ+CH/AERv4Vf+G88I/wDyno/4UT8EP+iN/Cr/\nAMN54R/+U9eq1/MR/wAFA/ix8Qbr9uHXf2wfCHwJ+OPxC+Gv/BJXXPhFos3xk8CeIfgbafCT\n4bW3jPT7H4jf8FKz4/0Xxl8dPBnxb8Tt/wAMXfEn4XWmgr8M/g98TbXSPHPg+/tNRu7a/ttR\ng0IA/oi/4UT8EP8Aojfwq/8ADeeEf/lPXqUUUcEccMMaQwwokUUUSLHHFHGoSOOONAFREUBU\nRQFVQAAAAK/La2/bL+M3/DxCL9nXVpPgzpH7P2veNo/hl8Or608N+J/GPjT4h+KU/ZIi/ae1\nNB8X/AvxO8R+D/hj8SdLhi1i7j/Z7+MnwU8A3vib4H6Vb/HHwZ8ZPEP/AAkXh7wFf/mnq3/B\nSr4//AL9gD9mj4k/s/eCf2atB8M/CT/gjt+zH+2z8RPgPpHwd+M3i628PaD4i+FerXPh34d6\nD4s1L45/C/wN+z78ILuz+GXiTwX8KvEmv+N/2mfi5q2oaBrFgPg/rz+GLC88dgH9PFFfjv8A\nET9s79rHwx8Zf2hr/RYP2eP+GfP2ev8Agoz+wv8AsTXfhbU/AnxIvfjF478OftieGP8Agn/a\naz4stfiDafFrTfBfg/XPhl40/bRufEGnCb4YeLtP8c+FtCtvCNxYeDdW0ybxr4l/YigAoooo\nAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACi\niigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoooo\nAKKKKACiiigAooooAKKKKAOb8R+DPB/jFvD7+LvCnhrxS/hLxLp3jPwq3iPQtL1xvDPjDR4r\nuDSfFfh9tTtbo6L4l0uG/vodO13TTbapYxXt3HbXUSXMwfnfEnwd+EfjLxDH4u8X/Cz4c+Kv\nFcMHgy2i8T+JPBHhnXPEMVt8OPH2nfFf4eW8etanpl1qSQeAvilpGk/EnwZEtyI/C/j7S9O8\nY6Gtj4isrbUYvRqKAPPLf4RfCe0j0CG1+GHw8tofCfjzxx8VPC0Vv4K8Nwx+Gvif8TV+ICfE\nn4j6BHHpiro3jz4hL8WPimvjjxhpwtvEPixfiX8QBr2o348ZeIhqXnXhv9kP9k7wb4B8d/Cn\nwh+y/wDs7+Ffhd8UriW8+Jvw28N/BT4baH4B+It3Osaz3Xjvwdpnhq18O+LriZYolln8Qadq\nEsixxh2IRQPoiigDwfwN+yx+zF8ML3StS+Gv7OXwH+Hmo6FrWleJNE1DwN8IPh94SvdH8RaF\n4F8U/C/RNf0q60Dw9p8+n61o3w08c+Nfh3pWqWkkN9p/gXxh4p8JWk8OgeINW0+79Y8M+E/C\nvgvTrnR/B3hnw/4T0i88QeLPFl5pfhnRtO0HTrrxV498Vaz468deJrmx0u2tbafxB408beI/\nEPjHxZrMsTaj4j8Va7rPiHWLm81fVL68n6CigAooooAKKKKACiiigAooooAKKKKACiiigAoo\nooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKA\nCiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAoo\nooAKKKKACiiigAooooAK4SH4W/DK30Dxv4Ug+HXgSDwt8TL7xTqnxH8NQ+EfD8WgfEDU/HMD\n2vjXUfG+jpp66d4rvvGFtJJb+KbvXra/uPEEDvDq0l3GzKe7ooA8Z0j9nH9nnw/8R9O+MWg/\nAf4M6J8XNH8H2Pw90j4p6R8L/BGm/EfS/AOmWUWm6b4H07xvZaHD4msfB+n6dBDYWPhq21SL\nRbSyhitYLKOCNIxy3iP9jj9kTxjZ+AtO8Xfsrfs4eKdP+FXgZPhh8L7DxH8Dvhjrln8N/hom\nkpoKfDzwFa6n4XuoPB/gZNCij0VPCXh5NO0BdJjTThp4s0WEfR9FAHlen/Av4I6Tot54a0v4\nOfCvTfDuoeLPh149v9A0/wCHvhKz0W+8c/B+0+Hdh8JfGd5pVvpEdjc+LPhdY/CH4TWXw68R\nTQPrHgm0+F/w7tvDV5pkPgnw0mmeqUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABR\nRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUU\nAFFFFABRRRQAUUUUAFFFFAH/2Q=="
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(\"IRdisplay\")\n",
    "display_png(file=\"C:/Users/jmarks/Desktop/GitHub/jaamarks_notebooks/Heroin_Project/figures/801_lymphoma/afr_eas_eur_Study_EA.jpg\")  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Removing outliers\n",
    "\n",
    "The triangle plots show an ancestral continuum derived from the genetic data. Since PCA principal components will be utilized as covariates in the GWA model, we can afford to be less conservative with our thresholds here for calling outliers. Based on the figure, I will apply the following inclusion criteria to avoid the loss of usable samples:\n",
    "***\n",
    "\n",
    "| Action Description        | Thresholding Criteria            |\n",
    "|---------------------------|----------------------------------|\n",
    "| For EA retainment         | (AFR < 0.25) $\\wedge$ (EAS < 0.25) |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Create triangle plot input with potential outliers filtered\n",
    "head -1 structure/triangle_plots/afr_eas_eur.triangle_input >\\\n",
    "    structure/triangle_plots/afr_eas_eur_filtered.triangle_input\n",
    "# EA retained\n",
    "tail -n +2 structure/triangle_plots/afr_eas_eur.triangle_input | \\\n",
    "    perl -lane 'if ($F[2] eq \"Study_EA\" && ($F[3] < 0.25 && $F[4] < 0.25)) { print $_; }' \\\n",
    "    >> structure/triangle_plots/afr_eas_eur_filtered.triangle_input\n",
    "\n",
    "wc -l structure/triangle_plots/afr_eas_eur.triangle_input\n",
    "'1684'\n",
    "wc -l structure/triangle_plots/afr_eas_eur_filtered.triangle_input\n",
    "'1684'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Note__: no subjects were removed after applying filter, therefore all of the reassignment and removal procedures do not need to be performed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/ea/autosomes\n",
    "\n",
    "# keep naming convention consistent\n",
    "for ext in {.bed,.bim,.fam,.log}; do\n",
    "    cp genotypes_b37_dbsnp138_flipped${ext} genotypes_b37_dbsnp138_flipped_structure${ext}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotype call rate variant filter\n",
    "I calculate the genotype missing call rate and remove any variants with a rate $> 3\\%$.\n",
    "<a id='call_rate'></a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Get missing call rate\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry='ea'\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure \\\n",
    "        --missing \\\n",
    "        --out ${ancestry}/autosomes/autosome_missing_call_rate\n",
    "\n",
    "    # Get list of variants with missing call rate > 3% \n",
    "    tail -n +2 ${ancestry}/autosomes/autosome_missing_call_rate.lmiss | \\\n",
    "        perl -lane 'if ($F[4] > 0.03) { print $F[1]; }' | \\\n",
    "        sort -u > ${ancestry}/autosomes/variant_missing_rate_gt_3_pct.remove\n",
    "\n",
    "    # Remove variants\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure \\\n",
    "        --exclude ${ancestry}/autosomes/variant_missing_rate_gt_3_pct.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_filter\n",
    "#done\n",
    "\n",
    "wc -l ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_filter.bim\n",
    "'686565'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hardy-Weinberg equilibrium filtering\n",
    "Extensive deviation from Hardy-Weinberg equilibrium (HWE) can be indicative of a genotyping or genotype calling error. Using PLINK  --hardy, I calculate goodness of fit test p-values for allele frequencies compared to HWE. Any variant with p < 0.0001 is removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Calculate HW p-values for each ancestry group\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_filter \\\n",
    "        --hardy \\\n",
    "        --out ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_filter\n",
    "\n",
    "   # Get list of variants with HWE p-value < 0.0001 for each ancestry group\n",
    "    tail -n +2 ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_filter.hwe | \\\n",
    "        perl -lane 'if ($F[2] eq \"ALL(NP)\" && $F[8] < 0.0001) { print $F[1]; }' | \\\n",
    "        sort -u > ${ancestry}/autosomes/hwe_p_lt_0.0001.remove\n",
    "\n",
    "\n",
    "    # Remove variants with HWE p-value < 0.0001\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_filter \\\n",
    "        --exclude ${ancestry}/autosomes/hwe_p_lt_0.0001.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_filter\n",
    "#done\n",
    "wc -l ea/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_filter.bim\n",
    "'684834'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Genotype call rate subject filter\n",
    "**Note**: The genotype call rate subject filtereing and the subsequent LD pruning are specifically for relatedness filtering calculations. Autosomes and chrX merging will be done using the PLINK file sets generated in the preceding step (HWE filtering).\n",
    "\n",
    "I calculate the genotype missing call rate and remove any subjects with a rate > $3\\%.$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Remove variants\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_filter \\\n",
    "        --mind 0.03 \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_lte_0.03_filter\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Linkage disequilibrium pruning\n",
    "\n",
    "Linkage disequilibrium (LD) pruning eliminates a large degree of redundancy in the data and reduces the influence of chromosomal artifacts. The objective of LD pruning is to select a subset of variants based off of LD such that the variants in the subset are indepdendent. This filtering will not carry forward to the final processed results, but this step improves the quality of identity-by-state (IBS) and identity-by-descent (IBD) calculations. Consequently, the LD pruned data will be used as input into IBS and IBD calculations.\n",
    "\n",
    "LD pruning is implemented using [PLINK --indep-pairwise](https://www.cog-genomics.org/plink/1.9/ld#indep)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Run per chromosome LD pruning \n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${ancestry}_${chr}_ld_prune \\\n",
    "            --script_prefix ${ancestry}/autosomes/${ancestry}_${chr}_ld_prune \\\n",
    "            --mem 7.5 \\\n",
    "            --priority 0 \\\n",
    "            --nslots 1 \\\n",
    "            --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "                --noweb \\\n",
    "                --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_lte_0.03_filter \\\n",
    "                --indep-pairwise 50 5 0.5 \\\n",
    "                --chr ${chr} \\\n",
    "                --out ${ancestry}/autosomes/genotypes_for_relatedness_filtering.ld_prune_chr${chr}\n",
    "    done\n",
    "#done\n",
    "\n",
    "\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    # Merge *prune.in files\n",
    "    cat ${ancestry}/autosomes/genotypes_for_relatedness_filtering.ld_prune_chr*.prune.in \\\n",
    "        > ${ancestry}/autosomes/genotypes_for_relatedness_filtering.ld_prune_autosomes.prune.in\n",
    "\n",
    "    # Create new PLINK filesets with only lD pruned variants\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_lte_0.03_filter \\\n",
    "        --extract ${ancestry}/autosomes/genotypes_for_relatedness_filtering.ld_prune_autosomes.prune.in \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_filter\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Identical-by-state and identical-by-descent filtering\n",
    "\n",
    "Identical-by-state or identity-by-state (IBS) filtering provides a means to identify duplicate subjects or monozygotic twins. Identical-by-descent or identity-by-descent allows for detection of high relatedness. For computational efficiency and memory usage limitations, ibd_pipeline.sh is used to perform IBS/IBD calculations on data chunks that are used as input for IBS and IBD filtering. If the study contains AA subjects, then for AA subjects only, [KING](http://people.virginia.edu/~wc9c/KING/manual.html) is used as an additional relationship inference tool for filtering.\n",
    "\n",
    "For these filtering steps, a list of individuals is identified separately for each filtering type then combined into a final list for exclusion. Following IBS filtering, if any individuals are identified for removal, a graph analysis needs to be done to account for an individual being related/identical to multiple individuals. Removal criteria is based on genotype calling rate. These individuals would need to be removed before running IBD filtering and KING. Graph analysis would need to be run after IBD calculations/KING if individuals were selected for removal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Calculate allele frequencies in data chunks\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/scripts/ibd_pipeline.sh \\\n",
    "        --in_prefix ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_filter \\\n",
    "        --out_dir ${ancestry}/autosomes/\n",
    "    sleep 5s\n",
    "#done\n",
    "\n",
    "\n",
    "# Aggregate data\n",
    "#for ancestry in {ea,aa}; do\n",
    "    outfile=${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_filter.genome\n",
    "    head -1 `ls ${ancestry}/autosomes/data.sub.genome* | head -1` > ${outfile}\n",
    "    cat ${ancestry}/autosomes/data.sub.genome* | \\\n",
    "        grep -v FID1 >> ${outfile}\n",
    "\n",
    "    # Remove temporary files\n",
    "    rm ${ancestry}/autosomes/data.sub.*\n",
    "    rm ${ancestry}/autosomes/tmp_sh*.sh\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IBS filtering\n",
    "For individual pairs with greater than 0.9 IBS score, only the individual with the lower genotype missing call rate is retained. I create a list of individuals to remove that will be combined with the IBD list for final filtering but removed from the LD pruned data before running IBD filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Get list of individuals with IBS > 0.9\n",
    "#for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    Rscript /shared/bioinformatics/software/R/merge_genome_missing.R \\\n",
    "        --genome ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_filter.genome \\\n",
    "        --imiss ${ancestry}/autosomes/autosome_missing_call_rate.imiss \\\n",
    "        --type IBS \\\n",
    "        --out ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Graph analysis for IBS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# R console #\n",
    "# Graph analysis to determine minimum # of subjects to remove\n",
    "library(igraph)\n",
    "setwd(\"/shared/data/studies/phs000801_lymphoma/genotype/original/processing\")\n",
    "#for(ancestry in c(\"ea\",\"aa\")){\n",
    "    ancestry=\"ea\"\n",
    "    stats = read.table(paste0(ancestry, \"/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter.stats\"), header = TRUE, colClasses=\"character\")\n",
    "    ibdPairs = as.data.frame(cbind(paste(stats$FID2, stats$IID2, sep=\"__\"), paste(stats$FID1, stats$IID1, sep=\"__\")), stringsAsFactors = FALSE)\n",
    "    colnames(ibdPairs) = c(\"fid_iid_1\",\"fid_iid_2\")\n",
    "    ibdGraph = graph.data.frame(ibdPairs, directed = FALSE)\n",
    "    vertexDegrees = sort(degree(ibdGraph), decreasing = TRUE)\n",
    "    removeList = vector()\n",
    "    while (length(vertexDegrees) > 0 && vertexDegrees[1] > 1) {\n",
    "        if (exists(\"removeList\")) {\n",
    "            removeList = c(removeList, names(vertexDegrees[1]))\n",
    "        } else {\n",
    "            removeList = names(vertexDegrees[1])\n",
    "        }\n",
    "        ibdGraph = ibdGraph - names(vertexDegrees[1])\n",
    "        vertexDegrees = sort(degree(ibdGraph), decreasing = TRUE)\n",
    "    }\n",
    "    ibdPairs = ibdPairs[!(ibdPairs[,1] %in% removeList) & !(ibdPairs[,2] %in% removeList),]\n",
    "    imiss = read.table(paste0(ancestry, \"/autosomes/autosome_missing_call_rate.imiss\"), header = TRUE, colClasses = \"character\")\n",
    "    imiss = as.data.frame(cbind(paste(imiss$FID, imiss$IID, sep=\"__\"), as.numeric(imiss$F_MISS)), stringsAsFactors = FALSE)\n",
    "    colnames(imiss) = c(\"fid_iid_1\",\"f_miss_1\")\n",
    "    ibdPairs = merge(ibdPairs, imiss, by=\"fid_iid_1\", sort = FALSE)\n",
    "    colnames(imiss) = c(\"fid_iid_2\",\"f_miss_2\")\n",
    "    ibdPairs = merge(ibdPairs, imiss, by=\"fid_iid_2\", sort = FALSE)\n",
    "    removeList = c(removeList, ibdPairs[ibdPairs[\"f_miss_1\"] < ibdPairs[\"f_miss_2\"],\"fid_iid_2\"])\n",
    "    removeList = c(removeList, ibdPairs[ibdPairs[\"f_miss_1\"] >= ibdPairs[\"f_miss_2\"],\"fid_iid_1\"])\n",
    "    removeList = cbind(lapply(strsplit(removeList,\"__\"), function(x) x[1]),lapply(strsplit(removeList,\"__\"), function(x) x[2]))\n",
    "    write.table(removeList, file=paste0(ancestry, \"/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_igraph.remove\"), row.names = FALSE, col.names = FALSE, quote = FALSE)\n",
    "    rm(removeList)\n",
    "#}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 85 EA individuals identified for removal."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove IBS problematic subjets\n",
    "Before proceeding with IBD filtering, the problematic subjects need to be removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "'''\n",
    "sort ea/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_igraph.remove | uniq | wc -l\n",
    "85\n",
    "'''\n",
    "\n",
    "# Make new PLINK file set\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_filter \\\n",
    "        --remove ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_igraph.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter\n",
    "done\n",
    "'''\n",
    " wc -l ea/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter.{bim,fam}\n",
    "294562, 1557\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IBD filtering\n",
    "For individual pairs with greateer than 0.4 IBD score, only the individual with the lower genotype missing call rate is retained. I create a list of individuals to remove that will be combined with the IBS list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Calculate allele frequencies in data chunks\n",
    "#for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/scripts/ibd_pipeline.sh \\\n",
    "        --in_prefix ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter \\\n",
    "        --out_dir ${ancestry}/autosomes/\n",
    "    sleep 5s\n",
    "#done\n",
    "\n",
    "# Aggregate data\n",
    "for ancestry in {ea,aa}; do\n",
    "    outfile=${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter.genome\n",
    "    head -1 `ls ${ancestry}/autosomes/data.sub.genome* | head -1` > ${outfile}\n",
    "    cat ${ancestry}/autosomes/data.sub.genome* | \\\n",
    "        grep -v FID1 >> ${outfile}\n",
    "\n",
    "    # Remove temporary files\n",
    "    rm ${ancestry}/autosomes/data.sub.*\n",
    "    rm ${ancestry}/autosomes/tmp_sh*.sh\n",
    "done\n",
    "\n",
    "# Get list of individuals with IBD > 0.4\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    Rscript /shared/bioinformatics/software/R/merge_genome_missing.R \\\n",
    "        --genome ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter.genome \\\n",
    "        --imiss ${ancestry}/autosomes/autosome_missing_call_rate.imiss \\\n",
    "        --type IBD \\\n",
    "        --pi_hat_threshold 0.4 \\\n",
    "        --out ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibd_gt_0.4_filter\n",
    "done\n",
    "\n",
    " wc -l ea/autosomes/genotypes_for_relatedness_ld_prune_ibd_gt_0.4_filter\n",
    "'''0'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Note__: No individuals need to be removed. I can proceed without IBD graph analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Constructing relatedness filtering lists\n",
    "The output of the IBS filtering and IBD filtering are combined into a single remove list. \n",
    "\n",
    "**Note**: there were no subjects identified for the EAs during the IBD filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "\n",
    "ancestry=\"ea\"\n",
    "# Get list of individuals from IBD and IBS filter lists\n",
    "cat <(tail -n +2 ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibs_gt_0.9_filter) <(tail -n +2 ${ancestry}/autosomes/genotypes_for_relatedness_ld_prune_ibd_gt_0.4_filter) | \\\n",
    "    sort -u > ${ancestry}/autosomes/ibs_gt_0.9_ibd_gt_0.4.remove\n",
    "wc -l ea/autosomes/ibs_gt_0.9_ibd_gt_0.4.remove\n",
    "\n",
    "'85'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of subjects marked for removal will be retained as a filtering list for the merged autosome and chrX PLINK file sets."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ChrX Processing\n",
    "\n",
    "### Removing ancestral outliers\n",
    "The subjects within each ancestry group were previously reassigned based on results from STRUCTURE. This change was reflected in the autosomes but need to be updated for chrX.\n",
    "\n",
    "__Note:__ There were no subjects identified for removal after running STRUCTURE. I will keep the naming convention for consistency and convenience, however."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "#for ancestry in {ea,aa}; do\n",
    "#    # Remove outliers identified by STRUCTURE\n",
    "#    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "#    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "#        --noweb \\\n",
    "#        --memory 2048 \\\n",
    "#        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138 \\\n",
    "#        --remove ${ancestry}/autosomes/ancestral_outliers.remove \\\n",
    "#        --chr 23 \\\n",
    "#        --make-bed \\\n",
    "#        --out ${ancestry}/chrX/genotypes_b37_dbsnp138_structure\n",
    "#done\n",
    "\n",
    "# simply renaming the files here\n",
    "for ancestry in {ea,aa}; do\n",
    "    # Remove outliers identified by STRUCTURE\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138 \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX/genotypes_b37_dbsnp138_structure\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Missing data subject filtering\n",
    "We calculate the proportion of missing genotype calls on chrX using PLINK to assess whether any subjects have data missing for the whole chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Get missing call rate per chr\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 1000 \\\n",
    "        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138_structure \\\n",
    "        --missing \\\n",
    "        --chr 23 \\\n",
    "        --out ${ancestry}/chrX/chrX_missing_call_rate\n",
    "\n",
    "# Find subjects that have data missing for whole chrX\n",
    "tail -n +2 ${ancestry}/chrX/chrX_missing_call_rate.imiss | \\\n",
    "    awk '{ OFS=\"\\t\" } { if($6==1){ print $1,$2 } }' >> ${ancestry}/chrX/missing_whole_chrX.remove\n",
    "wc -l ${ancestry}/chrX/missing_whole_chrX.remove\n",
    "'0'\n",
    "\n",
    "# rename files\n",
    "ancestry=\"ea\"\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 1000 \\\n",
    "    --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138_structure \\\n",
    "    --make-bed \\\n",
    "    --out ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx\n",
    "\n",
    "\n",
    "\n",
    "rm {ea,aa}/chrX/chr*missing_call_rate*\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this case none of the subjects had missing chrX data. If subjects ever show up as having missing data then further discussions need to be had on whether these subjects should be removed completely or whether they should only be excluded for the missing chromosome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove duplicate SNPs\n",
    "If multiple rsIDs are present then the one with the better genotype call rate across subjects should be retained. Obtaining the genotype call rates across subjects would need to be calculated using PLINK --missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Find duplicate rsIDs\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    cut -f2,2 ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx.bim | \\\n",
    "    sort | \\\n",
    "    uniq -D \\\n",
    "    > ${ancestry}/chrX/snp_duplicates.txt\n",
    "    echo -e \"\\n\\nNumber of ${ancestry} variant duplicates: $(wc -l ${ancestry}/chrX/snp_duplicates.txt)\\n\\n\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this case there are no duplicated rsIDs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strand flipping\n",
    "If strand flipping was an issue for the autosomes, it would also be an issue for chrX. Unfortunateley the May 2013 release for the 1000 Genomes Phase 3 data lacks IDs for chrX, so strand flipping for chrX will need to be done after converting IDs to 1000 Genomes Phase 3 IMPUTE2 format in a subsequent step. If complete data were available, strand flipping would be executed at this point.\n",
    "\n",
    "### Genotype call rate variant filter\n",
    "I calculate the genotype missing call rate and remove any variants with a rate $> 3\\%$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Get missing call rate\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx \\\n",
    "        --missing \\\n",
    "        --out ${ancestry}/chrX/chrX_missing_call_rate\n",
    "\n",
    "    tail -n +2 ${ancestry}/chrX/chrX_missing_call_rate.lmiss | \\\n",
    "        perl -lane 'if ($F[4] > 0.03) { print $F[1]; }' | \\\n",
    "        sort -u > ${ancestry}/chrX/variant_missing_rate_gt_3_pct.remove\n",
    "\n",
    "    # Remove variants\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx \\\n",
    "        --exclude ${ancestry}/chrX/variant_missing_rate_gt_3_pct.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_filter\n",
    "    echo -e \"\\nNumber of ${ancestry} variants removed: $(wc -l ${ancestry}/chrX/variant_missing_rate_gt_3_pct.remove)\\n\\n\"\n",
    "done\n",
    "\n",
    "'''\n",
    "Number of ea variants removed: 1147 ea/chrX/variant_missing_rate_gt_3_pct.remove\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hardy-Weinberg equilibrium filtering\n",
    "Extensive deviation from Hardy-Weinberg equilibrium (HWE) can be indicative of a genotyping or genotype calling error. Using PLINK  --hardy, I calculate goodness of fit test p-values for allele frequencies compared to HWE. Any variant with p < 0.0001 is removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Calculate HW p-values for each ancestry group\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --memory 3800 \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_filter \\\n",
    "        --hardy \\\n",
    "        --out ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_filter\n",
    "\n",
    "    # Get list of variants with HWE p-value < 0.0001\n",
    "    tail -n +2 ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_filter.hwe | \\\n",
    "        perl -lane 'if ($F[2] eq \"ALL(NP)\" && $F[8] < 0.0001) { print $F[1]; }' | \\\n",
    "        sort -u > ${ancestry}/chrX/hwe_p_lt_0.0001.remove\n",
    "\n",
    "    # Remove variants with HWE p-value < 0.0001\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_filter \\\n",
    "        --exclude ${ancestry}/chrX/hwe_p_lt_0.0001.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_hwe_p_gte_0.0001_filter\n",
    "    echo -e \"\\nNumber of ${ancestry} variants removed: $(wc -l ${ancestry}/chrX/hwe_p_lt_0.0001.remove)\\n\\n\"\n",
    "done\n",
    "\n",
    "'''\n",
    "Number of ea variants removed: 8 ea/chrX/hwe_p_lt_0.0001.remove\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Merge autosomes and chrX\n",
    "The data processed through to HWE filtering for both the autosomes and chrX are merged before applying the final subject filtering steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=ea\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_filter \\\n",
    "        --bmerge ${ancestry}/chrX/genotypes_b37_dbsnp138_structure_subject_missing_chrx_variant_missing_lte_0.03_hwe_p_gte_0.0001_filter \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_filter.all_chr\n",
    "#done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Relatedness filter\n",
    "From relatedness calculations on the autosomal data, I remove subjects that are closely related or seemingly identical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Remove variants\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_filter.all_chr \\\n",
    "        --remove ${ancestry}/autosomes/ibs_gt_0.9_ibd_gt_0.4.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_filter.all_chr\n",
    "    echo -e \"\\nNumber of ${ancestry} subjects removed: $(wc -l  ${ancestry}/autosomes/ibs_gt_0.9_ibd_gt_0.4.remove)\\n\\n\"\n",
    "done\n",
    "\n",
    "'''\n",
    "Number of ea subjects removed: 85 ea/autosomes/ibs_gt_0.9_ibd_gt_0.4.remove\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotype call rate subject filter\n",
    "I calculate the genotype missing call rate and remove any subjects with a rate $>3\\%$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Remove \n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_filter.all_chr \\\n",
    "        --mind 0.03 \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linkage disequilibrium pruning\n",
    "Linkage disequilibrium (LD) pruning eliminates a large degree of redundancy in the data and reduces the influence of chromosomal artifacts. The objective of LD pruning is to select a subset of variants based off of LD such that the variants in the subset are indepdendent. This filtering will not carry forward to the final processed results, but this step improves the quality of sex check and excessive homozygosity calculations. Consequently, the LD pruned data will be used as input for those calculations.\n",
    "\n",
    "LD pruning is implemented using [PLINK --indep-pairwise](https://www.cog-genomics.org/plink/1.9/ld#indep)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Run per chromosome LD pruning for each chr\n",
    "for ancestry in {ea,aa}; do\n",
    "    for chr in {1..23}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${ancestry}_${chr}_ld_prune \\\n",
    "            --script_prefix ${ancestry}/${ancestry}_${chr}_ld_prune \\\n",
    "            --mem 7.5 \\\n",
    "            --nslots 1 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "                --noweb \\\n",
    "                --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr \\\n",
    "                --indep-pairwise 50 5 0.5 \\\n",
    "                --chr ${chr} \\\n",
    "                --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr${chr}_filter\n",
    "    done\n",
    "done\n",
    "\n",
    "# Merge *prune.in files\n",
    "for ancestry in {ea,aa}; do\n",
    "    cat ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr*_filter.prune.in \\\n",
    "        > ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter.prune.in\n",
    "\n",
    "    # Create new PLINK filesets with only lD pruned variants\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr \\\n",
    "        --extract ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter.prune.in \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sex check\n",
    "To verify that self-reported gender status matches with the genetic information, I use PLINK `--split-x` followed by ``--check-sex`` to confirm the data agree. Our discrepancy criteria are\n",
    "\n",
    "* Reported as male (coded as 1) but F < 0.2\n",
    "* Reported as female (coded as 2) but F > 0.8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Break out the pseudo-autosomal region on chrX (will be stored as chr 25)\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter \\\n",
    "        --split-x b37 \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr\n",
    "\n",
    "    # Run sex check\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr \\\n",
    "        --check-sex \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr\n",
    "\n",
    "    # Get subjects that do not meet inbreeding coefficient thresholds\n",
    "    perl -lane 'if ($F[2]==1 && $F[5]<0.2) { print $_; } elsif ($F[2]==2 && $F[5]>0.8) { print $F[0].\"\\t\".$F[1]; }' ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck \\\n",
    "        > ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.threshold_fail\n",
    "\n",
    "    # Get subjects that are problematic due to no inbreeding coefficient value\n",
    "    perl -lane 'if ($F[4] eq \"PROBLEM\" && $F[5] eq \"nan\") { print $F[0].\"\\t\".$F[1]; }' ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck \\\n",
    "        > ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.nan_fail\n",
    "\n",
    "    # Combine discordant subjects into one remove list\n",
    "    cat ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.*_fail | \\\n",
    "        sort -u \\\n",
    "        > ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.remove\n",
    "#done\n",
    "\n",
    "# Print number of discrepancies\n",
    "wc -l {ea,aa}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.remove"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For these data the sex check yielded no discrepancies for the EA subjects under our criteria. If any individuals failed the check then they would be removed before subsequent steps using the file set generated after gentoype call rate subject filtering (before removing the pseudo-autosomal chrX region)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# remove the individuals who failed sex check (note only EA subjecs failed but we will\n",
    "# process the AA subjects as if there were AA subjects who failed too. \n",
    "# this keeps the naming convention easier.)\n",
    "\n",
    "#for ancestry in {ea,aa}; do\n",
    "#    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "#    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "#        --noweb \\\n",
    "#        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr \\\n",
    "#        --remove ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.remove \\\n",
    "#        --make-bed \\\n",
    "#        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_failed_sex_check_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr\n",
    "#done\n",
    "\n",
    "ancestry=\"ea\"\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr \\\n",
    "        --remove ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_split_chrX_filter.all_chr.sexcheck.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_failed_sex_check_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Excessive homozygosity filter\n",
    "Excessive homozygosity checks for inbreeding and population substructure. This check is implemented using PLINK. According to their documentation, the statistics used for this filtering are not LD-sensitive, i.e., an LD pruned dataset is recommended as input. Additionally, the documentation states that only autosomal data are used, so concerns of chrX biasing results are irrelevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "\n",
    "# Get homozygosity stats\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter \\\n",
    "        --het \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter\n",
    "\n",
    "   # Identify problematic subjects\n",
    "    perl -lane 'if ($F[5] < -0.2 || $F[5] > 0.5) { print $F[0].\" \".$F[1]; }' ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_chr_all_filter.het \\\n",
    "    > ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_homozygosity_check_chr_all_filter.remove\n",
    "\n",
    "    # Remove problematic subjects\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_failed_sex_check_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_filter.all_chr \\\n",
    "        --remove ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_homozygosity_check_chr_all_filter.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_failed_sex_check_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_homozygosity_check_filter.all_chr\n",
    "\n",
    "    echo -e \"\\nNumber of ${ancestry} subjects removed: $(wc -l  ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_homozygosity_check_chr_all_filter.remove)\\n\\n\"\n",
    "done\n",
    "\n",
    "'''\n",
    "Number of ea subjects removed: 0 ea/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_ld_prune_homozygosity_check_chr_all_filter.remove\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recoding variants for 1000G phase 3\n",
    "\n",
    "RefSeq IDs (rsIDs) for variants can vary depending on the dbSNP build used and variant IDs do not all use RefSeq nomenclature. To provide a common nomenclature that will make comparisons across data sets feasible, I use a script that recodes all variant names to match 1000G phase 3 variants by position and alleles. The 1000G Phase 3 data I used for STRUCTURE are from /share/nas03/bioinformatics_group/data/ref_panels/1000G/2013.05/plink on MIDAS, but from correspondence with Nathan Gaddis I learned that /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/ also contains 1000G Phase 3 data derived from the May 2013 release. The difference is that it was downloaded from the IMPUTE2 website and reformatted to be directly compatible with IMPUTE2.\n",
    "\n",
    "The data in /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/ is used for variant name recoding, but the 1000G genotype information is acquired from /share/nas03/bioinformatics_group/data/ref_panels/1000G/2013.05/plink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "mkdir 1000g_name_recoding\n",
    "\n",
    "for ancestry in {ea,aa}; do\n",
    "    for ext in {bed,bim,fam}; do\n",
    "        cp  ${ancestry}/genotypes_b37_dbsnp138_structure_variant_missing_lte_0.03_hwe_p_gte_0.0001_subject_missing_chrx_failed_sex_check_ibs_gt_0.9_ibd_gt_0.4_subject_missing_lte_0.03_homozygosity_check_filter.all_chr.${ext} 1000g_name_recoding/${ancestry}_chr_all.${ext}\n",
    "    done\n",
    "done\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant ID updating\n",
    "\n",
    "Because the 1000G data and the study data have RefSeq IDs (rsIDs) from different dbSNP builds, I standardize them using convert_to_1000g_ids.pl. In the study data set, certain indels may be represented as two variants, a monomorphic variant and an indel with the - symbol for one of the alleles. For example:\n",
    "\n",
    "`1   rs201826967  0.809   57873968   0   G`<br>`1   rs11284630   0.809   57873969   -   A`\n",
    "\n",
    "These two variants represent a G:GA indel and is coded as such in the 1000 Genomes data. The script to update the names to 1000 Genomes IMPUTE2 format will assign the same ID to these two variants. The duplicated IDs will cause problems for PLINK filtering, so I will remove the variant from a set of duplicate IDs that has the lower genotype call rate. Duplicates may arise for other reasons, and they will be filtered based on the same criterion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "mkdir 1000g_data\n",
    "\n",
    "# Break out data by chr\n",
    "for ancestry in {ea,aa}; do\n",
    "    for chr in {1..23}; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bfile ${ancestry}_chr_all \\\n",
    "            --chr ${chr} \\\n",
    "            --make-bed \\\n",
    "            --out ${ancestry}_chr${chr}\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "# Rename study autosome variant IDs\n",
    "for ancestry in {ea,aa}; do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name recode_to_1000g_${chr} \\\n",
    "            --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "            --mem 6 \\\n",
    "            --nslots 1 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "            --file_in ${ancestry}_chr${chr}.bim \\\n",
    "            --file_out ${ancestry}_chr${chr}_renamed.bim \\\n",
    "            --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.legend.gz \\\n",
    "            --file_in_header 0 \\\n",
    "            --file_in_id_col 1 \\\n",
    "            --file_in_chr_col 0 \\\n",
    "            --file_in_pos_col 3 \\\n",
    "            --file_in_a1_col 4 \\\n",
    "            --file_in_a2_col 5 \\\n",
    "            --chr ${chr}\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "# Rename study chrX variant IDs\n",
    "chr=23\n",
    "#for ancestry in {ea,aa}; do\n",
    "ancestry=\"ea\"\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "        --mem 6 \\\n",
    "        --nslots 1 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in ${ancestry}_chr${chr}.bim \\\n",
    "        --file_out ${ancestry}_chr${chr}_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "#done\n",
    "\n",
    "\n",
    "# Rename 1000G autosome variant IDs\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix 1000g_chr${chr}_id_rename \\\n",
    "        --mem 8 \\\n",
    "        --nslots 1 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr}.bim \\\n",
    "        --file_out 1000g_data/chr${chr}_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "done\n",
    "\n",
    "\n",
    "# Rename 1000G chrX variant IDs\n",
    "chr=23\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name recode_to_1000g_${chr} \\\n",
    "    --script_prefix 1000g_chr${chr}_id_rename \\\n",
    "    --mem 8 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "    --file_in /shared/data/ref_panels/1000G/2013.05/plink/ALL.chrX.bim \\\n",
    "    --file_out 1000g_data/chr${chr}_renamed.bim \\\n",
    "    --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "    --file_in_header 0 \\\n",
    "    --file_in_id_col 1 \\\n",
    "    --file_in_chr_col 0 \\\n",
    "    --file_in_pos_col 3 \\\n",
    "    --file_in_a1_col 4 \\\n",
    "    --file_in_a2_col 5 \\\n",
    "    --chr ${chr}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove duplicate variant ID\n",
    "\n",
    "Many of the functions in PLINK will throw errors when duplicate variant IDs are found.To resolve this, duplicate variant IDs are identified, and only the duplicate variant with the best genotype call rate is retained."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "for ancestry in {ea,aa}; do\n",
    "    for chr in {1..23}; do\n",
    "        # Append _X (where X is a number) to the end of the variant IDs for all but 1st occurrence of duplicates\n",
    "        perl -lane 'BEGIN { %idCounts = (); }\n",
    "                    if (exists($idCounts{$F[1]})) {\n",
    "                        $idCounts{$F[1]}++;\n",
    "                        print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "                    } else {\n",
    "                        $idCounts{$F[1]} = 1;\n",
    "                        print;\n",
    "                    }' ${ancestry}_chr${chr}_renamed.bim > \\\n",
    "            ${ancestry}_chr${chr}_renamed_dups.bim\n",
    "\n",
    "        # Generate list of duplicate SNPs\n",
    "        grep -P \"_[1-9]\" ${ancestry}_chr${chr}_renamed_dups.bim |\n",
    "            perl -lane 'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' > \\\n",
    "            ${ancestry}_chr${chr}_renamed_dups.dupvar_list\n",
    "    done\n",
    "done\n",
    "\n",
    "#for ancestry in {ea,aa}; do\n",
    "    for chr in $(wc -l *.dupvar_list | perl -lane 'if ($F[0] != 0) { $F[1] =~ /chr(\\d+)/; print $1; }'); do\n",
    "# Get call rates for duplicate SNPs\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed ${ancestry}_chr${chr}.bed \\\n",
    "            --bim ${ancestry}_chr${chr}_renamed_dups.bim \\\n",
    "            --fam ${ancestry}_chr${chr}.fam \\\n",
    "            --extract ${ancestry}_chr${chr}_renamed_dups.dupvar_list \\\n",
    "            --missing \\\n",
    "            --out ${ancestry}_chr${chr}_renamed_dups\n",
    "\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed ${ancestry}_chr${chr}.bed \\\n",
    "            --bim ${ancestry}_chr${chr}_renamed_dups.bim \\\n",
    "            --fam ${ancestry}_chr${chr}.fam \\\n",
    "            --extract ${ancestry}_chr${chr}_renamed_dups.dupvar_list \\\n",
    "            --missing \\\n",
    "            --out ${ancestry}_chr${chr}_renamed_dups\n",
    "\n",
    "        # Create remove list that contains the duplicate with the higher missing rate\n",
    "        tail -n +2 ${ancestry}_chr${chr}_renamed_dups.lmiss | \\\n",
    "            perl -lane 'BEGIN { %missingness = (); }\n",
    "                        if ($F[1] =~ /^(\\S+)\\_/) {\n",
    "                            $duplicateName = $1\n",
    "                        } else {\n",
    "                            $duplicateName = $F[1].\"_2\";\n",
    "                        }\n",
    "                        if (exists($missingness{$duplicateName})) {\n",
    "                            if ($missingness{$duplicateName} > $F[4]) {\n",
    "                                print $duplicateName;\n",
    "                            } else {\n",
    "                                print $F[1];\n",
    "                            }\n",
    "                        } else {\n",
    "                            $missingness{$F[1]} = $F[4];\n",
    "                        }' \\\n",
    "                > ${ancestry}_chr${chr}_renamed_dups.remove\n",
    "\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed ${ancestry}_chr${chr}.bed \\\n",
    "            --bim ${ancestry}_chr${chr}_renamed_dups.bim \\\n",
    "            --fam ${ancestry}_chr${chr}.fam \\\n",
    "            --exclude ${ancestry}_chr${chr}_renamed_dups.remove \\\n",
    "            --make-bed \\\n",
    "            --out ${ancestry}_chr${chr}_renamed_dups_removed\n",
    "\n",
    "        # Remove \"_2\" from variant IDs\n",
    "        perl -i.bak -lne 's/\\_2//; print;' ${ancestry}_chr${chr}_renamed_dups_removed.bim\n",
    "    done\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**: there were no duplicates in the above steps and therefore we got errors. Thus, we don't use:\n",
    "* \\${ancestry}_chr\\${chr}_renamed_dups_removed.bim\n",
    "in the next section. \n",
    "\n",
    "Instead, we use:\n",
    "\n",
    "* \\${ancestry}_chr\\${chr}_renamed.bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge chromosome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "# Add chr where duplicates were removed to merge list\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo -e \"\\n\\n================ ${ancestry} ================\\n\\n\"\n",
    "    for chr in $(wc -l *.dupvar_list | perl -lane 'if ($F[0] != 0) { $F[1] =~ /chr(\\d+)/; if ($1 ne \"1\") { print $1; } }'); do\n",
    "        baseName=${ancestry}_chr${chr}_renamed\n",
    "        echo ${baseName}.bed ${baseName}.bim ${baseName}.fam\n",
    "    done > ${ancestry}_renamed.chr_merge\n",
    "\n",
    "    # Add all other chr to merge list\n",
    "    for chr in $(wc -l *.dupvar_list | perl -lane 'if ($F[0] == 0) { $F[1] =~ /chr(\\d+)/; print $1; }'); do\n",
    "        baseName=${ancestry}_chr${chr}\n",
    "        echo ${baseName}.bed ${baseName}_renamed.bim ${baseName}.fam\n",
    "    done >> ${ancestry}_renamed.chr_merge\n",
    "\n",
    "    # Merge chromosomes\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --merge-list ${ancestry}_renamed.chr_merge \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr_all_renamed\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ChrX Strand flipping\n",
    "Since the original 1000 Genomes Phase 3 data from the May 2013 release lacks ID information, it is not until this stage that strand flipping can be executed. At this point both the 1000 Genomes and the study data have updated IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing\n",
    "\n",
    "# Copy test data\n",
    "for ancestry in {ea,aa}; do\n",
    "    mkdir ${ancestry}/chrX_flip_test\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile 1000g_name_recoding/${ancestry}_chr_all_renamed \\\n",
    "        --chr 23 \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX_flip_test/${ancestry}_chr23_renamed\n",
    "    cp 1000g_name_recoding/1000g_data/chr23_renamed.bim ${ancestry}/chrX_flip_test/\n",
    "    cp /shared/data/ref_panels/1000G/2013.05/plink/ALL.chrX.bed ${ancestry}/chrX_flip_test/chr23_renamed.bed\n",
    "    cp /shared/data/ref_panels/1000G/2013.05/plink/ALL.chrX.fam ${ancestry}/chrX_flip_test/chr23_renamed.fam\n",
    "\n",
    "    # Attempt merge with 1000G chrX data\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/chrX_flip_test/${ancestry}_chr23_renamed \\\n",
    "        --bmerge ${ancestry}/chrX_flip_test/chr23_renamed \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX_flip_test/chrX_unflipped_test\n",
    "\n",
    "'Error: 2082 variants with 3+ alleles present.'\n",
    "    # Flip: variants\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/chrX_flip_test/${ancestry}_chr23_renamed \\\n",
    "        --flip ${ancestry}/chrX_flip_test/chrX_unflipped_test-merge.missnp \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX_flip_test/chrX_flipped\n",
    "\n",
    "    # Attempt merge with 1000G chr1 data\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/chrX_flip_test/chrX_flipped \\\n",
    "        --bmerge ${ancestry}/chrX_flip_test/chr23_renamed \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chrX_flip_test/chrX_flipped_test\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The flip test suggests that flipping is necessary, so I proceed with flipping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/\n",
    "\n",
    "for ancestry in {ea,aa}; do\n",
    "    # Make final fip list\n",
    "    if [ -a ${ancestry}/chrX_flip_test/chrX_flipped_test-merge.missnp ]; then\n",
    "        # Get flip list subset that does not produce errors\n",
    "        echo \"Subsetting flip list\"\n",
    "        comm -23 <(sort -u ${ancestry}/chrX_flip_test/chrX_unflipped_test-merge.missnp) <(sort -u ${ancestry}/chrX_flip_test/chrX_flipped_test-merge.missnp) \\\n",
    "            > ${ancestry}/chrX_flip_test/chrX.flip\n",
    "    else\n",
    "        # No flipped merge errors so keep all in original flip list\n",
    "        sort -u ${ancestry}/chrX_flip_test/chrX_unflipped_test-merge.missnp > ${ancestry}/chrX_flip_test/chrX.flip \n",
    "    fi\n",
    "\n",
    "# Perform final flip\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}/chrX_flip_test/${ancestry}_chr23_renamed \\\n",
    "        --flip ${ancestry}/chrX_flip_test/chrX.flip \\\n",
    "        --make-bed \\\n",
    "        --out 1000g_name_recoding/${ancestry}_chr23_renamed_flipped\n",
    "done\n",
    "\n",
    "# Clean up\n",
    "rm -rf {ea,aa}/chrX_flip_test/chrX_flip_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge flipped chrX with autosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "# Extract flipped autosomes\n",
    "for ancestry in {ea,aa}; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}_chr_all_renamed \\\n",
    "        --autosome \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_autosomes_renamed\n",
    "\n",
    "    # Merge autosomes with flipped chrX\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}_autosomes_renamed \\\n",
    "        --bmerge ${ancestry}_chr23_renamed_flipped \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr_all_renamed_flipped\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Allele discordance\n",
    "As a way of standardizing the data, I use 1000G phase 3 data as a reference. The study data may not match the 1000G phase 3 data for several reasons including\n",
    "\n",
    "* Discordant names\n",
    "* Discordant positions\n",
    "* Discordant alleles (swapped major and minor alleles)\n",
    "* Polymorphic alleles not fixed by strand flipping\n",
    "* Discordant allele frequencies\n",
    "\n",
    "I perform checks for these discordances using the 1000G Phase 3 data derived from\n",
    "`data/ref_panels/1000G/2014.10/` derived from the May 2013 release. These data were downloaded from the [IMPUTE2 website](https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.html) and reformatted to be directly compatible with IMPUTE2. These data will be used for ID recoding and MAF calculations.\n",
    "\n",
    "To determine if any study data variants have flipped alleles, I compare the variants to the appropriate 1000 Genomes Phase 3 reference superpopulation. For example, African ancestry is compared to AFR, European ancestry is compared to EUR, and Hispanic ancestry is compared to AMR.\n",
    "\n",
    "Discordance checks are used mainly to\n",
    "\n",
    "1. Determine allele frequency discordances to remove prior to imputation\n",
    "2. Determine if alleles are properly coded in reference to the forward/positive genome strand\n",
    "\n",
    "A lare number of discordant variants indicates potential allele flipping/mismatching issues. If the number of discordant heterozygous variants is less than approximately 10,000, then it is unlikely that there were major issues with flipping or mismatching that indicate coding alleles to the wrong strand.\n",
    "\n",
    "### Preparing study data\n",
    "The allele discordance scrip requires as input a file of MAFs, so I calculate this using PLINK."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "# Calculate allele frequencies\n",
    "for ancestry in {ea,aa}; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}_chr_all_renamed_flipped \\\n",
    "        --freq \\\n",
    "        --out ${ancestry}_chr_all_renamed_flipped\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparing 1000G phase 3 data\n",
    "\n",
    "MAF data are needed for 1000 Genomes data in addition to the study data, so I calculate these for the superpopulations(s) that matches the study group subject ancestries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding/1000g_data\n",
    "\n",
    "for ancestry in {ea,aa}; do\n",
    "    # Get list of study variant IDs\n",
    "    perl -lane 'if ($F[0] <= 23) { print $F[1]; }' ../${ancestry}_chr_all_renamed_flipped.bim | \\\n",
    "        sort > ../${ancestry}_chr_all_sorted_variants.txt\n",
    "done\n",
    "\n",
    "# Calculate autosome and chrX MAFs for 1000G AFR\n",
    "pop=\"EUR\"\n",
    "ancestry=\"ea\"\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${pop}_${chr} \\\n",
    "        --script_prefix ${pop}_chr${chr}.maf \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "            --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "            --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "            --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "            --chr ${chr} \\\n",
    "            --out ${ancestry}_${pop}_overlap_chr${chr}.maf \\\n",
    "            --extract ../${ancestry}_chr_all_sorted_variants.txt \\\n",
    "            --keep_groups ${pop}\n",
    "done\n",
    "\n",
    "chr=23\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${pop}_${chr} \\\n",
    "    --script_prefix ${pop}_chr${chr}.maf \\\n",
    "    --mem 6.8 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "        --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.hap.gz\\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "        --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "        --chr ${chr} \\\n",
    "        --extract ../${ancestry}_chr_all_sorted_variants.txt \\\n",
    "        --out ${ancestry}_${pop}_overlap_chr${chr}.maf \\\n",
    "        --keep_groups ${pop}\n",
    "\n",
    "# Merge per chr MAFs for EA subject overlap with EUR\n",
    "ancestry=\"ea\"\n",
    "pop=\"EUR\"\n",
    "head -n 1 ${ancestry}_${pop}_overlap_chr1.maf > ${ancestry}_${pop}_overlap_chr_all.maf\n",
    "tail -q -n +2 ${ancestry}_${pop}_overlap_chr{1..23}.maf \\\n",
    "    >> ${ancestry}_${pop}_overlap_chr_all.maf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "#### Discordance check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "# Run discordance checks for each ancestry group\n",
    "    pop=\"EUR\"\n",
    "    ancestry=\"ea\"\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${ancestry}_crosscheck \\\n",
    "        --script_prefix ${ancestry}_chr_all_renamed_1000g_discordance_check \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R \\\n",
    "            --study_bim_file ${ancestry}_chr_all_renamed_flipped.bim\n",
    "            --study_frq_file ${ancestry}_chr_all_renamed_flipped.frq\n",
    "            --ref_maf_file 1000g_data/${ancestry}_${pop}_overlap_chr_all.maf\n",
    "            --out_prefix ${ancestry}_chr_all_renamed_1000g_discordance_check\"\n",
    "\n",
    "# Print summary of discordances\n",
    "for ancestry in {ea,aa}; do\n",
    "    echo \"${ancestry}_chr_all_renamed_1000g_discordance_check.summary\"\n",
    "    cat ${ancestry}_chr_all_renamed_1000g_discordance_check.summary\n",
    "done\n",
    "\n",
    "\n",
    "'''\n",
    "=========================ea=============================\n",
    "discordant_positions    0\n",
    "discordant_names        0\n",
    "discordant_alleles      17949\n",
    "discordant_alleles_not_fixed_by_strand_flip     15529\n",
    "discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip        0\n",
    "at_cg_snps_freq_diff_gt_0.2     547\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "# Flip discordant alleles that are fixed by flipping\n",
    "for ancestry in {ea,aa}; do\n",
    "    comm -23 <(tail -n +2 ${ancestry}_chr_all_renamed_1000g_discordance_check.discordant_alleles | cut -f2,2 | sort -u) \\\n",
    "        <(tail -n +2 ${ancestry}_chr_all_renamed_1000g_discordance_check.discordant_alleles_not_fixed_by_strand_flip | cut -f2,2 | sort -u) \\\n",
    "        > ${ancestry}_chr_all_renamed_1000g_discordance_check.flip\n",
    "\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}_chr_all_renamed_flipped \\\n",
    "        --flip ${ancestry}_chr_all_renamed_1000g_discordance_check.flip \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr_all_renamed_final_flip\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Final check for ambiguous SNPs\n",
    "We perform a final check to determine if there are any more ambiguous SNPs (AT or CG)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "ancestry=\"ea\"\n",
    "perl -lane 'if (($F[4] eq \"A\" && $F[5] eq \"T\") || ($F[4] eq \"T\" && $F[5] eq \"A\") ||\n",
    "($F[4] eq \"C\" && $F[5] eq \"G\") || ($F[4] eq \"G\" && $F[5] eq \"C\")) { print $F[1]; }' \\\n",
    "${ancestry}_chr_all_renamed_final_flip.bim > ${ancestry}_at_cg_snps.txt\n",
    "\n",
    "wc -l ${ancestry}_at_cg_snps.txt\n",
    "'2275'\n",
    "\n",
    "\n",
    "for ancestry in {ea,aa}; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile ${ancestry}_chr_all_renamed_final_flip \\\n",
    "    --exclude ${ancestry}_at_cg_snps.txt \\\n",
    "    --make-bed \\\n",
    "    --out ${ancestry}_chr_all_renamed_final_flip_clean\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Observed genotype finalization\n",
    "As a starting point for haplotype phasing and imputation, I store the final QC processed PLINK binary fileset in a new directory `final`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 command line\n",
    "cd /shared/data/studies/phs000801_lymphoma/genotype/original/\n",
    "\n",
    "mkdir  final\n",
    "\n",
    "# Copy file sets to final directory\n",
    "for ancestry in {ea,aa}; do\n",
    "    for ext in {bed,bim,fam}; do\n",
    "        cp processing/1000g_name_recoding/${ancestry}_chr_all_renamed_final_flip_clean.${ext} final/${ancestry}_chr_all.${ext}\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Final upload to S3\n",
    "Once the QC process has been verified, I can delete the processing files on S3 and upload the compressed final {.bim,.fam,.bed} files to S3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/final\n",
    "\n",
    "# gzip final files \n",
    "for f in *; do\n",
    "gzip $f\n",
    "done\n",
    "\n",
    "aws s3 cp final s3://rti-common/dbGaP/phs000801_lymphoma/genotype/original/final  --recursive --quiet &"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  },
  "toc": {
   "nav_menu": {
    "height": "91px",
    "width": "170px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {
    "height": "812px",
    "left": "0px",
    "right": "1468px",
    "top": "107px",
    "width": "212px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
