{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Applying Nyholt Correction to chrX SNPs\n",
    "__Author:__ Jesse Marks <br>\n",
    "**Date**: September 13, 3018\n",
    "\n",
    "This documentation is relevant to [GitHub Issue #98](https://github.com/RTIInternational/bioinformatics/issues/98#issuecomment-421154584). In order to make corrections for multiple testing in the data, one can apply the [Nyholt Correction Method](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1181954/). The Nyholt Correction Method is a simple correction for multiple testing for SNPs in linkage disequilibrium with each other.\n",
    "\n",
    "We want to run chrX association analyses for HIV acquisition in WIHS1, WIHS2, UHS1+2+3, and VIDUS. There were a list a chrX SNPs from the UKBioBank GWAS results that were significant (listed below). The evidence continues to improve for these chrX SNPs when we combine cohorts during meta-analyses. They're even associated in WIHS2 AAs by themselves at `P<0.05`. The updated meta-analysis across WIHS1 and WIHS2 (all available ancestries, total `N=4025`, best `P=0.007`). Interestingly, much of the signal is driven by AAs, which does have the larger sample size.\n",
    "\n",
    "I will compute the Nyholt correction in the ALL panel for this list (all on chr. X).\n",
    "```\n",
    "rs614503:112046351:A:T\n",
    "rs644416:112048479:C:A\n",
    "rs5903400:112053941:GA:G\n",
    "rs4460547:112055287:C:A\n",
    "rs647000:112055663:C:G\n",
    "rs5974283:112056081:A:G\n",
    "rs604591:112064874:A:C\n",
    "rs17004059:112065240:G:A\n",
    "rs5973962:112067381:T:C\n",
    "rs620730:112073165:C:T\n",
    "rs591816:112073370:A:C\n",
    "rs685830:112074176:T:C\n",
    "rs57610141:112075057:G:GA\n",
    "rs650005:112075127:T:C\n",
    "```\n",
    "\n",
    "## Nyholt Correction\n",
    "Find effective number of independent SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line ##\n",
    "base_dir='/home/ec2-user/jmarks/hiv/wihs1-2_nyholt'\n",
    "processing_dir='/home/ec2-user/jmarks/hiv/wihs1-2_nyholt/matSpD'\n",
    "in_data=${base_dir}/'wihs1_snp_list.txt'\n",
    "out_data=${base_dir}/'wihs2_snp_list.txt'\n",
    "# matSpD should be in ~/bin\n",
    "\n",
    "# Copy variants with phase 3 IDs to final file\n",
    "perl -lane 'if ($F[0] =~ /\\:/) { print $F[0]; }' $in_data >> $out_data\n",
    "# Convert non-phase 3 variant IDs to phase 3 IDs based on name\n",
    "for variant in $(perl -lane 'if ($F[0] !~ /\\:/) { print $F[0]; }' $in_data); do\n",
    "    chr=$(grep -P \"$variant\\s\" $in_data | perl -lane 'print $F[1];')\n",
    "    gunzip -c /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz|\n",
    "        grep \"$variant:\" |\n",
    "        perl -lane 'print $F[0];'\n",
    "done >> $out_data\n",
    "# Convert non-phase 3 variant IDs to phase 3 IDs based on position\n",
    "for position in $(perl -lane 'if ($F[0] !~ /\\:/) { print $F[2]; }' $in_data); do\n",
    "    chr=$(grep -P \"$position\" $in_data | grep -v \":\" | perl -lane 'print $F[1];')\n",
    "    gunzip -c /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz|\n",
    "        grep \":$position:\"\n",
    "done > /home/ec2-user/nyholt/044/044_variant_list.nonphase3.legend\n",
    "perl -lane 'BEGIN { $lastPosition = 0 } if ($F[1] != $lastPosition) { print $F[0]; $lastPosition = $F[1]; }' \\\n",
    "    /home/ec2-user/nyholt/044/044_variant_list.nonphase3.legend >> $out_data\n",
    "\n",
    "\n",
    "### START Extract replication SNPs from 1000G ###\n",
    "# Extract variants from 1000G panel\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=23\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "      --job_name ALL_1000G \\\n",
    "      --script_prefix ${base_dir}/wihs1-2_nyholt \\\n",
    "      --mem 10 \\\n",
    "      --priority 0 \\\n",
    "      --program /shared/bioinformatics/software/perl/file_conversion/convert_reference_panels.pl \\\n",
    "          --impute2_hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.hap.gz \\\n",
    "          --impute2_legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "          --impute2_sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "          --extract $out_data \\\n",
    "          --out ${base_dir}/1000G_ALL.chr$chr.extracted_variants \\\n",
    "          --generate_plink_ped_file \\\n",
    "          --generate_plink_map_file \\\n",
    "          --chr $chr\n",
    "#done\n",
    "\n",
    "### END Extract replication SNPs from 1000G ###\n",
    "\n",
    "\n",
    "### START Generate correlation matrix ###\n",
    "\n",
    "#for (( chr=1; chr<23; chr++ )); do\n",
    "chr=23\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --file ${base_dir}/1000G_ALL.chr$chr.extracted_variants.plink \\\n",
    "        --r \\\n",
    "        --matrix \\\n",
    "        --out ${base_dir}/1000G_ALL.chr$chr.extracted_variants.r\n",
    "#done\n",
    "\n",
    " ### END Generate correlation matrix ###\n",
    "\n",
    "\n",
    "### START Run matSpD analysis ###\n",
    "\n",
    "#for chr in {1..22}; do\n",
    "chr=23\n",
    "    echo $chr\n",
    "    cp ${base_dir}/1000G_ALL.chr$chr.extracted_variants.r.ld ${processing_dir}/correlation.matrix\n",
    "    cd ${processing_dir}\n",
    "    R CMD BATCH ${processing_dir}/matSpDlite.R\n",
    "    mv ${processing_dir}/matSpDlite.out ${base_dir}/1000G_ALL.chr$chr.extracted_variants.matspdlite\n",
    "done\n",
    "\n",
    "\n",
    "veffLi=0\n",
    "chr=23\n",
    "#for chr in {1..22}; do\n",
    "    chrVeffLi=$(grep -A 2 Equation ${base_dir}/1000G_ALL.chr$chr.extracted_variants.matspdlite |\\\n",
    "                tail -n +3 | perl -pe 's/^\\s+//; s/\\s+$//;')\n",
    "    echo $chr\n",
    "    echo $chrVeffLi\n",
    "    veffLi=`echo $veffLi + $chrVeffLi | bc`\n",
    "done\n",
    "echo $veffLi\n",
    "\n",
    "'''4'''\n",
    "\n",
    "### END Run matSpD analysis ###\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
