{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UHS1-4 chrX imputation\n",
    "__Author__: Jesse Marks\n",
    "\n",
    "**Date:** October 04, 2018\n",
    "\n",
    "**GitHub Issue:** [Issue #112](https://github.com/RTIInternational/bioinformatics/issues/112)\n",
    "\n",
    "This document logs the steps taken to perform pre-imputation procedures on the merged datasets of UHS1, UHS2, UHS3, and UHS4—both EA and AA. The starting point for this analysis is after quality control of observed genotypes. The quality controlled genotypes are oriented on the GRCh37 plus strand. Based on findings from [Johnson et al.](https://link.springer.com/article/10.1007/s00439-013-1266-7), the intersection set of variants for will be used for imputation. \n",
    "\n",
    "## Software and tools\n",
    "The software and tools used for porcessing these data are\n",
    "* [Michigan Imputation Server](https://imputationserver.sph.umich.edu/index.html) (MIS)\n",
    "* [Amazon Web Services (AWS) - Cloud Computing Services](https://aws.amazon.com/)\n",
    "    * Linux AMI\n",
    "* [PLINK v1.90 beta 4.10](https://www.cog-genomics.org/plink/)\n",
    "* [bgzip](http://www.htslib.org/doc/tabix.html)\n",
    "* [BCF Tools](http://www.htslib.org/doc/bcftools.html)\n",
    "* Windows 10 with [Cygwin](https://cygwin.com/) installed\n",
    "* GNU bash version 4.2.46\n",
    "\n",
    "## Data retrieval and organization\n",
    "PLINK binary filesets will be obtained from AWS S3 storage.\n",
    "\n",
    "Nathan Gaddis performed the QC and stored the observed genotype data at:\n",
    "\n",
    "```\n",
    "UHS1\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs1.aa.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs1.ea.chr23\n",
    "\n",
    "UHS2\n",
    "\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs2.aa.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs2.ea.chr23\n",
    "\n",
    "UHS3\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs3.aa.V1-2.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-2.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs3.aa.V1-3.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-3.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs3.aa.merged.chr23\n",
    "s3://rti-midas-data/studies/hiv/observed/final/uhs3.ea.merged.chr23\n",
    "\n",
    "UHS4\n",
    "s3://rti-midas-data/studies/uhs4/observed/genotypes/final/uhs4/merged2.ea\n",
    "s3://rti-midas-data/studies/uhs4/observed/genotypes/final/uhs4/merged2.aa\n",
    "```\n",
    "\n",
    "\n",
    "**Note**: For UHS3 I will use v1-2 and v1-3 datasets. I will treat them like two different cohorts. Therefore, when I merge UHS1-4 I will be merging 4 data sets per ancestry. \n",
    "\n",
    "* uhs1.aa, uhs2.aa, uhs3.aa.V1-2,uhs3.aa.V1-3, & uhs4.merged2.aa\n",
    "* uhs1.ea, uhs2.ea, uhs3.ea.V1-2,uhs3.ea.V1-3, & uhs4.merged2.ea\n",
    "\n",
    "**Note2**: The STRUCTURE analysis was performed on the combined UHS1-4 data. This resulted in a new ancestry partitioning. In particular, one UHS1 sample and one UHS4 sample was reassigned ancestry groups.\n",
    "\n",
    "**Note3:** The UHS1 chrX sample counts are in discordance with the autosome sample counts. For example:\n",
    "```\n",
    "  2015 uhs1.aa.chr23.fam\n",
    "  2016 uhs1.aa.fam\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### chrX Statistics Breakdown \n",
    "This table includes the initial number of variants in each study as well as the final number of variants in the intersection set. The `Variants Post-Filtering` is in referral to the filtering steps (1) remove discordant alleles & (2) removal of monomorphic variants.\n",
    "\n",
    "#### EA\n",
    "| Data Set      | Initial Variants | Variants Post-Filtering  |\n",
    "|---------------|------------------|--------------------------|\n",
    "| UHS1          | 17,408           |                          |\n",
    "| UHS2          | 32,857           |                          |\n",
    "| UHS3_V1-2     | 24,202           |                          |\n",
    "| UHS3_V1-3     | 35,440           |                          |\n",
    "| UHS4          | 35,469           |                          |\n",
    "| Intersection  | NA               |                          |\n",
    "\n",
    "\n",
    "#### AA\n",
    "| Data Set      | Initial Variants | Variants Post-Filtering  |\n",
    "|---------------|------------------|--------------------------|\n",
    "| UHS1          | 17,396           |                          |\n",
    "| UHS2          | 23,484           |                          |\n",
    "| UHS3_V1-2     | 41,884           |                          |\n",
    "| UHS3_V1-3     | 44,706           |                          |\n",
    "| UHS4          | 29,488           |                          |\n",
    "| Merged        | NA               |                          |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Directory Structure & Download Data\n",
    "The following section needs to be modified each time to reflect where the data is stored!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EC2 command line (bash) ###\n",
    "\n",
    "# Modify variables below\n",
    "######################################################################\n",
    "base_dir=/shared/jmarks/hiv/uhs1-4/chrX_processing\n",
    "base_name=\"chr23\" # chr23 or chr_all\n",
    "#chr_list={1..23} # or {1..22} \n",
    "ancestry_list=\"aa ea\" # space delimited Ex. \"ea aa ha\"\n",
    "study_list=\"uhs1 uhs2 uhs3_v1-2 uhs3_v1-3 uhs4\" # space delimited \n",
    "######################################################################\n",
    "\n",
    "mkdir -p ${base_dir}/processing/{intersect,1000g,impute_prep}\n",
    "for study in ${study_list};do\n",
    "    for ancestry in ${ancestry_list};do\n",
    "        mkdir -p ${base_dir}/processing/${study}\n",
    "        mkdir -p ${base_dir}/data/${study}/genotype/observed/${ancestry}\n",
    "    done\n",
    "done\n",
    "\n",
    "## download genotype (with AWS CLI tools) to respective directories ##\n",
    "\n",
    "#aws s3 cp s3://rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-3.chr23.fam .\n",
    "#aws s3 cp s3://rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-3.chr23.bed .\n",
    "#aws s3 cp s3://rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-3.chr23.bim .\n",
    "#..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Processing\n",
    "## Data Wrangling\n",
    "The UHS4 chrX genotype data have not been separated from the autosomes like the other sub-groups have. Therefore, we will do this now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to uhs4.merged2.ea.chr23.log.\n",
      "Options in effect:\n",
      "  --bfile uhs4.merged2.ea\n",
      "  --chr 23\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out uhs4.merged2.ea.chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "35469 out of 2073618 variants loaded from .bim file.\n",
      "989 people (659 males, 330 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 989 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.974134.\n",
      "35469 variants and 989 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to uhs4.merged2.ea.chr23.bed + uhs4.merged2.ea.chr23.bim +\n",
      "uhs4.merged2.ea.chr23.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to uhs4.merged2.aa.chr23.log.\n",
      "Options in effect:\n",
      "  --bfile uhs4.merged2.aa\n",
      "  --chr 23\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out uhs4.merged2.aa.chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "29488 out of 1878335 variants loaded from .bim file.\n",
      "1072 people (766 males, 306 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1072 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.971589.\n",
      "29488 variants and 1072 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to uhs4.merged2.aa.chr23.bed + uhs4.merged2.aa.chr23.bim +\n",
      "uhs4.merged2.aa.chr23.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa.chr23\n",
      "  --bmerge /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.ea.chr23\n",
      "  --make-bed\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure\n",
      "\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "2015 people loaded from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa.chr23.fam.\n",
      "1140 people to be merged from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.ea.chr23.fam.\n",
      "Of these, 1140 are new, while 0 are present in the base dataset.\n",
      "17396 markers loaded from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa.chr23.bim.\n",
      "17408 markers to be merged from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.ea.chr23.bim.\n",
      "Of these, 17 are new, while 17391 are present in the base dataset.\n",
      "Warning: Variants 'X:67923735:C:T' and 'X:67923735:A:G' have the same position.\n",
      "Warning: Variants 'X:139580982:C:T' and 'X:139580982:A:G' have the same\n",
      "position.\n",
      "Warning: Variants 'X:150267369:C:T' and 'X:150267369:A:G' have the same\n",
      "position.\n",
      "Performing single-pass merge (3155 people, 17413 variants).\n",
      "Merged fileset written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure-merge.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure-merge.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure-merge.fam\n",
      ".\n",
      "17413 variants loaded from .bim file.\n",
      "3155 people (2374 males, 781 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 3155 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.990882.\n",
      "17413 variants and 3155 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.merged2.aa.chr23\n",
      "  --bmerge /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.merged2.ea.chr23\n",
      "  --make-bed\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure\n",
      "\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "1072 people loaded from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.merged2.aa.chr23.fam.\n",
      "989 people to be merged from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.merged2.ea.chr23.fam.\n",
      "Of these, 989 are new, while 0 are present in the base dataset.\n",
      "29488 markers loaded from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.merged2.aa.chr23.bim.\n",
      "35469 markers to be merged from\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.merged2.ea.chr23.bim.\n",
      "Of these, 6515 are new, while 28954 are present in the base dataset.\n",
      "Performing single-pass merge (2061 people, 36003 variants).\n",
      "Merged fileset written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure-merge.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure-merge.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure-merge.fam\n",
      ".\n",
      "36003 variants loaded from .bim file.\n",
      "2061 people (1425 males, 636 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2061 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.874429.\n",
      "36003 variants and 2061 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "cd /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4\n",
    "## break out chrX data for UHS4\n",
    "for ancestry in ea aa; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile uhs4.merged2.$ancestry \\\n",
    "        --chr 23 \\\n",
    "        --make-bed \\\n",
    "        --out uhs4.merged2.$ancestry.chr23\n",
    "done\n",
    "\n",
    "## uhs1: merge ancestries\n",
    "study=uhs1            \n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile ${base_dir}/$study/uhs1.aa.chr23 \\\n",
    "    --bmerge ${base_dir}/$study/uhs1.ea.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out $base_dir/$study/$study.aa+ea.chr23.1KG.structure\n",
    "\n",
    "## uhs4: merge ancestries\n",
    "study=uhs4\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile ${base_dir}/$study/uhs4.merged2.aa.chr23 \\\n",
    "    --bmerge ${base_dir}/$study/uhs4.merged2.ea.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out $base_dir/$study/$study.aa+ea.chr23.1KG.structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure\n",
      "  --keep /shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding/ea.subject_ids.keep.txt\n",
      "  --make-bed\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure\n",
      "\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "17413 variants loaded from .bim file.\n",
      "3155 people (2374 males, 781 females) loaded from .fam.\n",
      "--keep: 1128 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1128 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.992074.\n",
      "17413 variants and 1128 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs1/uhs1.aa+ea.chr23.1KG.structure\n",
      "  --keep /shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding/aa.subject_ids.keep.txt\n",
      "  --make-bed\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure\n",
      "\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "17413 variants loaded from .bim file.\n",
      "3155 people (2374 males, 781 females) loaded from .fam.\n",
      "--keep: 2008 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2008 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.990263.\n",
      "17413 variants and 2008 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure\n",
      "  --keep /shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding/ea.subject_ids.keep.txt\n",
      "  --make-bed\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure\n",
      "\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "36003 variants loaded from .bim file.\n",
      "2061 people (1425 males, 636 females) loaded from .fam.\n",
      "--keep: 988 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 988 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.959739.\n",
      "36003 variants and 988 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/uhs4/uhs4.aa+ea.chr23.1KG.structure\n",
      "  --keep /shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding/aa.subject_ids.keep.txt\n",
      "  --make-bed\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure\n",
      "\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "36003 variants loaded from .bim file.\n",
      "2061 people (1425 males, 636 females) loaded from .fam.\n",
      "--keep: 1072 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1072 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.795883.\n",
      "36003 variants and 1072 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "## partition UHS1 and UHS4 into respective ancestry groups ##\n",
    "plots=/shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding\n",
    "anlist=\"ea aa\"\n",
    "for study in uhs1 uhs4; do\n",
    "    for an in $anlist; do\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --bfile $base_dir/$study/$study.aa+ea.chr23.1KG.structure \\\n",
    "            --keep $plots/$an.subject_ids.keep.txt \\\n",
    "            --make-bed \\\n",
    "            --out $base_dir/data/$study/genotype/observed/$an/$study.$an.chr23.1KG.structure\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3023 /shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding/ea.subject_ids.keep.txt\n",
      "4026 /shared/jmarks/hiv/uhs1-4/autosome_processing/structure/triangle_plots/thresholding/aa.subject_ids.keep.txt\n"
     ]
    }
   ],
   "source": [
    "ww $plots/ea.subject_ids.keep.txt\n",
    "ww $plots/aa.subject_ids.keep.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  2009 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs1/genotype/observed/aa/uhs1.aa.1KG.structure.fam\n",
      "   767 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs2/genotype/observed/aa/uhs2.aa.1KG.structure.fam\n",
      "    84 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3_v1-2.aa.1KG.structure.fam\n",
      "    94 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3_v1-3.aa.1KG.structure.fam\n",
      "  1072 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs4/genotype/observed/aa/uhs4.aa.1KG.structure.fam\n",
      "  4026 total\n",
      "\n",
      "  1130 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs1/genotype/observed/ea/uhs1.ea.1KG.structure.fam\n",
      "   828 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs2/genotype/observed/ea/uhs2.ea.1KG.structure.fam\n",
      "    33 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3_v1-2.ea.1KG.structure.fam\n",
      "    44 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3_v1-3.ea.1KG.structure.fam\n",
      "   988 /shared/jmarks/hiv/uhs1-4/autosome_processing/data/uhs4/genotype/observed/ea/uhs4.ea.1KG.structure.fam\n",
      "  3023 total\n"
     ]
    }
   ],
   "source": [
    "ww /shared/jmarks/hiv/uhs1-4/autosome_processing/data/*/*/*/aa/*fam\n",
    "echo \"\"\n",
    "ww /shared/jmarks/hiv/uhs1-4/autosome_processing/data/*/*/*/ea/*fam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  2008 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.fam\n",
      "   767 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.fam\n",
      "    84 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.fam\n",
      "    94 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.fam\n",
      "  1072 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.fam\n",
      "  4025 total\n",
      "\n",
      "  1128 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.fam\n",
      "   828 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.fam\n",
      "    33 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.fam\n",
      "    44 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.fam\n",
      "   988 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.fam\n",
      "  3021 total\n"
     ]
    }
   ],
   "source": [
    "echo \"\"\n",
    "ww /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs*/genotype/observed/aa/*fam\n",
    "echo \"\"\n",
    "ww /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs*/genotype/observed/ea/*fam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  17413 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bim\n",
      "  32857 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.bim\n",
      "  24202 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bim\n",
      "  35440 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bim\n",
      "  36003 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bim\n",
      " 145915 total\n",
      "\n",
      "  17413 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bim\n",
      "  23484 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.bim\n",
      "  41884 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bim\n",
      "  44706 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bim\n",
      "  36003 /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bim\n",
      " 163490 total\n"
     ]
    }
   ],
   "source": [
    "ww /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs*/genotype/observed/ea/*bim \n",
    "echo \"\"\n",
    "ww /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs*/genotype/observed/aa/*bim "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GRCh37 strand and allele discordance check\n",
    "### MAF for study data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/aa_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/aa_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "17413 variants loaded from .bim file.\n",
      "2008 people (1423 males, 585 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2008 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.990263.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/aa_chr23.frq\n",
      ".\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/ea_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/ea_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "17413 variants loaded from .bim file.\n",
      "1128 people (942 males, 186 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1128 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.992074.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/ea_chr23.frq\n",
      ".\n",
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/aa_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/aa_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "23484 variants loaded from .bim file.\n",
      "767 people (517 males, 250 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 767 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.974578.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/aa_chr23.frq\n",
      ".\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/ea_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/ea_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "32857 variants loaded from .bim file.\n",
      "828 people (495 males, 333 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 828 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.976798.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/ea_chr23.frq\n",
      ".\n",
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/aa_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/aa_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "41884 variants loaded from .bim file.\n",
      "84 people (59 males, 25 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 84 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.914796.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/aa_chr23.frq\n",
      ".\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/ea_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/ea_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "24202 variants loaded from .bim file.\n",
      "33 people (18 males, 15 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 33 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.884568.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/ea_chr23.frq\n",
      ".\n",
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/aa_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/aa_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "44706 variants loaded from .bim file.\n",
      "94 people (63 males, 31 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 94 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.905208.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/aa_chr23.frq\n",
      ".\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/ea_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/ea_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "35440 variants loaded from .bim file.\n",
      "44 people (24 males, 20 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 44 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.927423.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/ea_chr23.frq\n",
      ".\n",
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/aa_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/aa_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "36003 variants loaded from .bim file.\n",
      "1072 people (766 males, 306 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1072 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.795883.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/aa_chr23.frq\n",
      ".\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/ea_chr23.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bim\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.fam\n",
      "  --freq\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/ea_chr23\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "36003 variants loaded from .bim file.\n",
      "988 people (658 males, 330 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 988 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.959739.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/ea_chr23.frq\n",
      ".\n",
      "\n",
      "Calculating intersection between aa uhs1 uhs2 uhs3_v1-2 uhs3_v1-3 uhs4...\n",
      "\n",
      "5141 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "\n",
      "Calculating intersection between ea uhs1 uhs2 uhs3_v1-2 uhs3_v1-3 uhs4...\n",
      "\n",
      "3412 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n"
     ]
    }
   ],
   "source": [
    "### EC1 command line (Bash) ###\n",
    "\n",
    "# write out the MAF report\n",
    "for study in ${study_list}; do\n",
    "    study_dir=${base_dir}/processing/${study}/strand_check\n",
    "    mkdir ${study_dir}\n",
    "    for ancestry in ${ancestry_list};do\n",
    "        data_dir=${base_dir}/data/${study}/genotype/observed/${ancestry}\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed ${data_dir}/*bed\\\n",
    "            --bim ${data_dir}/*bim\\\n",
    "            --fam ${data_dir}/*fam\\\n",
    "            --freq \\\n",
    "            --out ${study_dir}/${ancestry}_${base_name}\n",
    "    done\n",
    "done\n",
    "\n",
    "# Get list of variants from all studies\n",
    "studies=($study_list)  #studies=(uhs1 uhs2 uhs3_v1-2 uhs3_v1-3 uhs4) # array of study names\n",
    "num=${#studies[@]}\n",
    "\n",
    "## Get intersection set\n",
    "for ancestry in ${ancestry_list};do\n",
    "    bim_files=()\n",
    "    for (( i=0; i<${num}; i++ ));do\n",
    "        bim_files+=(${base_dir}/data/${studies[$i]}/genotype/observed/$ancestry/*bim)\n",
    "    done\n",
    "    \n",
    "    echo -e \"\\nCalculating intersection between $ancestry ${study_list}...\\n\"\n",
    "    cat ${bim_files[@]}| cut -f2 | sort |  uniq -c | awk -v num=$num '$1 == num {print $2}' \\\n",
    "        > ${base_dir}/processing/intersect/${ancestry}_variant_intersection.txt\n",
    "    ww ${base_dir}/processing/intersect/${ancestry}_variant_intersection.txt\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MAF for 1000G\n",
    "This pipeline is currently set up to handle EUR and AFR populations.\n",
    "#### Autosomes\n",
    "Get 1000G MAF for chromosomes 1–22 (autosomes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### EC2 command line (Bash)\n",
    "#\n",
    "## Calculate autosome MAFs for 1000G populations\n",
    "#for ancestry in ${ancestry_list};do\n",
    "#\n",
    "#    if [ $ancestry == \"ea\" ]\n",
    "#    then\n",
    "#        pop=\"EUR\"\n",
    "#    else\n",
    "#        pop=\"AFR\"\n",
    "#    fi\n",
    "#    \n",
    "#    for chr in {1..22}; do\n",
    "#        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "#            --job_name ${pop}_${chr}_MAF \\\n",
    "#            --script_prefix ${base_dir}/processing/1000g/${pop}_chr${chr}.maf \\\n",
    "#            --mem 6.8 \\\n",
    "#            --nslots 1 \\\n",
    "#            --priority 0 \\\n",
    "#            --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "#                --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "#                --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "#                --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "#                --chr ${chr} \\\n",
    "#                --out ${base_dir}/1000g/${pop}_chr${chr}.maf \\\n",
    "#                --extract ${base_dir}/processing/intersect/${ancestry}_variant_intersection.txt \\\n",
    "#                --keep_groups ${pop}\n",
    "#    done\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### chrX\n",
    "Get 1000G MAF for chromosome 23 (chrX)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your job 8231 (\"AFR_23_MAF\") has been submitted\n",
      "Your job 8232 (\"EUR_23_MAF\") has been submitted\n"
     ]
    }
   ],
   "source": [
    "### Bash ###\n",
    "\n",
    "chr=23\n",
    "for ancestry in ${ancestry_list};do\n",
    "\n",
    "    if [ $ancestry == \"ea\" ]\n",
    "    then\n",
    "        pop=\"EUR\"\n",
    "    else\n",
    "        pop=\"AFR\"\n",
    "    fi\n",
    "\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${pop}_23_MAF \\\n",
    "        --script_prefix ${base_dir}/processing/1000g/${pop}_chr${chr}.maf \\\n",
    "        --mem 6.8 \\\n",
    "        --nslots 1 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "            --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.hap.gz\\\n",
    "            --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "            --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "            --chr $chr \\\n",
    "            --out ${base_dir}/processing/1000g/${pop}_chr${chr}.maf \\\n",
    "            --extract ${base_dir}/processing/intersect/${ancestry}_variant_intersection.txt \\\n",
    "            --keep_groups ${pop}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge 1000G chromosomes\n",
    "Only need to perform this if there were multiple chromosomes for which the MAF was calculated—e.g. more than just chrX."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Bash ###\n",
    "##chr_list={1..23}\n",
    "#\n",
    "## Merge per chr MAFs for each 1000G population\n",
    "#for ancestry in ${ancestry_list};do\n",
    "#    if [ $ancestry == \"ea\" ]\n",
    "#    then\n",
    "#        pop=\"EUR\"\n",
    "#    else\n",
    "#        pop=\"AFR\"\n",
    "#    fi\n",
    "#    \n",
    "#        head -n 1 ${base_dir}/1000g/${pop}_chr1.maf > 1000g/${pop}_chr_all.maf\n",
    "#        tail -q -n +2 1000g/${pop}_${chr_list}.maf \\ # chr_list defined in beginning\n",
    "#            >> 1000g/${pop}_chr_all.maf\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Allele Discordances Check\n",
    "The allele discordances will be resolved by\n",
    "* Flipping allele discordances that are fixed by flipping\n",
    "* Removing SNPs with discordant names\n",
    "* Removing SNPs with discordant positions\n",
    "* Removing allele discordances that are not resolved by flipping\n",
    "* Removing alleles with large deviations from the reference population allele frequencies\n",
    "\n",
    "Given that the allele discordance check was done using a union set of SNPs across all studies within an ancestry group, some of the SNPs logged as discordant for a given study may not actually be in the study. Fortunately, if they are not in a given study they will not interfere with the filtering procedures. Note that the intersection set is used for the final studies merger.\n",
    "\n",
    "#### Autosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Bash ###\n",
    "#\n",
    "## Run discordance checks for each ancestry group\n",
    "#for study in ${study_list}; do\n",
    "#    for ancestry in ${ancestry_list};do\n",
    "#        if [ $ancestry = \"ea\" ]; then\n",
    "#            pop=EUR\n",
    "#        else\n",
    "#            pop=AFR\n",
    "#        fi\n",
    "#\n",
    "#       /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "#           --job_name ${ancestry}_${study}_crosscheck \\\n",
    "#           --script_prefix ${base_dir}/strand_check/${ancestry}_allele_discordance_check \\\n",
    "#           --mem 6 \\\n",
    "#           --nslots 4 \\\n",
    "#           --priority 0 \\\n",
    "#           --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "#               --study_bim_file ${base_dir}/${study}/genotype/observed/${ancestry}/*bim\n",
    "#               --study_frq_file ${base_dir}/${study}/strand_check/${ancestry}_chr_all.frq\n",
    "#               --ref_maf_file ${base_dir}/1000g/${pop}_chr_all.maf\n",
    "#               --out_prefix ${base_dir}/${study}/strand_check/${ancestry}_allele_discordance\"\n",
    "#    done\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### chrX "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your job 8233 (\"aa_uhs1_crosscheck\") has been submitted\n",
      "Your job 8234 (\"ea_uhs1_crosscheck\") has been submitted\n",
      "Your job 8235 (\"aa_uhs2_crosscheck\") has been submitted\n",
      "Your job 8236 (\"ea_uhs2_crosscheck\") has been submitted\n",
      "Your job 8237 (\"aa_uhs3_v1-2_crosscheck\") has been submitted\n",
      "Your job 8238 (\"ea_uhs3_v1-2_crosscheck\") has been submitted\n",
      "Your job 8239 (\"aa_uhs3_v1-3_crosscheck\") has been submitted\n",
      "Your job 8240 (\"ea_uhs3_v1-3_crosscheck\") has been submitted\n",
      "Your job 8241 (\"aa_uhs4_crosscheck\") has been submitted\n",
      "Your job 8242 (\"ea_uhs4_crosscheck\") has been submitted\n"
     ]
    }
   ],
   "source": [
    "for study in ${study_list}; do\n",
    "    for ancestry in ${ancestry_list};do\n",
    "        if [ $ancestry = \"ea\" ]; then\n",
    "            pop=EUR\n",
    "        else\n",
    "            pop=AFR\n",
    "        fi\n",
    "\n",
    "        # chr23 discordance check\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${ancestry}_${study}_crosscheck \\\n",
    "            --script_prefix ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance_check \\\n",
    "            --mem 6.8 \\\n",
    "            --nslots 1 \\\n",
    "            --priority 0 \\\n",
    "            --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "                --study_bim_file ${base_dir}/data/${study}/genotype/observed/${ancestry}/*bim\n",
    "                --study_frq_file ${base_dir}/processing/${study}/strand_check/${ancestry}_chr23.frq\n",
    "                --ref_maf_file ${base_dir}/processing/1000g/${pop}_chr23.maf\n",
    "                --out_prefix ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance\"\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resolving Allele Discordances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "===============\n",
      "Processing uhs1_aa\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/aa_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/aa/uhs1.aa.chr23.1KG.structure.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/aa_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "17413 variants loaded from .bim file.\n",
      "2008 people (1423 males, 585 females) loaded from .fam.\n",
      "--flip: 2561 SNPs flipped.\n",
      "--exclude: 17399 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2008 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.990259.\n",
      "17399 variants and 2008 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered.bed +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered.bim +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs1_ea\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/ea_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs1/genotype/observed/ea/uhs1.ea.chr23.1KG.structure.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/strand_check/ea_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "17413 variants loaded from .bim file.\n",
      "1128 people (942 males, 186 females) loaded from .fam.\n",
      "--flip: 1700 SNPs flipped.\n",
      "--exclude: 17410 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1128 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.992074.\n",
      "17410 variants and 1128 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered.bed +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered.bim +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs2_aa\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/aa_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/aa/uhs2.aa.chr23.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/aa_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "23484 variants loaded from .bim file.\n",
      "767 people (517 males, 250 females) loaded from .fam.\n",
      "--flip: 0 SNPs flipped.\n",
      "--exclude: 23484 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 767 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.974578.\n",
      "23484 variants and 767 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered.bed +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered.bim +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs2_ea\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/ea_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs2/genotype/observed/ea/uhs2.ea.chr23.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/strand_check/ea_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "32857 variants loaded from .bim file.\n",
      "828 people (495 males, 333 females) loaded from .fam.\n",
      "--flip: 0 SNPs flipped.\n",
      "--exclude: 32857 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 828 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.976798.\n",
      "32857 variants and 828 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered.bed +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered.bim +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs3_v1-2_aa\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/aa_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/aa_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "41884 variants loaded from .bim file.\n",
      "84 people (59 males, 25 females) loaded from .fam.\n",
      "--flip: 2 SNPs flipped.\n",
      "--exclude: 41847 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 84 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.914726.\n",
      "41847 variants and 84 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs3_v1-2_ea\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/ea_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/strand_check/ea_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "24202 variants loaded from .bim file.\n",
      "33 people (18 males, 15 females) loaded from .fam.\n",
      "--flip: 0 SNPs flipped.\n",
      "--exclude: 24197 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 33 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.884547.\n",
      "24197 variants and 33 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs3_v1-3_aa\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/aa_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/aa_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "44706 variants loaded from .bim file.\n",
      "94 people (63 males, 31 females) loaded from .fam.\n",
      "--flip: 2 SNPs flipped.\n",
      "--exclude: 44668 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 94 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.905134.\n",
      "44668 variants and 94 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs3_v1-3_ea\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/ea_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/strand_check/ea_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "35440 variants loaded from .bim file.\n",
      "44 people (24 males, 20 females) loaded from .fam.\n",
      "--flip: 0 SNPs flipped.\n",
      "--exclude: 35436 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 44 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.927415.\n",
      "35436 variants and 44 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs4_aa\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/aa_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/aa/uhs4.aa.chr23.1KG.structure.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/aa_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1152 MB successfully, after larger attempt(s) failed.\n",
      "36003 variants loaded from .bim file.\n",
      "1072 people (766 males, 306 females) loaded from .fam.\n",
      "--flip: 0 SNPs flipped.\n",
      "--exclude: 36003 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1072 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.795883.\n",
      "36003 variants and 1072 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered.bed +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered.bim +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "===============\n",
      "Processing uhs4_ea\n",
      "\n",
      "Making remove list\n",
      "Making flip list\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered.log.\n",
      "Options in effect:\n",
      "  --bed /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bed\n",
      "  --bim /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.bim\n",
      "  --exclude /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/ea_snps.remove\n",
      "  --fam /shared/jmarks/hiv/uhs1-4/chrX_processing/data/uhs4/genotype/observed/ea/uhs4.ea.chr23.1KG.structure.fam\n",
      "  --flip /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/strand_check/ea_snps.flip\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "36003 variants loaded from .bim file.\n",
      "988 people (658 males, 330 females) loaded from .fam.\n",
      "--flip: 0 SNPs flipped.\n",
      "--exclude: 36003 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 988 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.959739.\n",
      "36003 variants and 988 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered.bed +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered.bim +\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "### Bash ###\n",
    "\n",
    "# Apply filters\n",
    "for study in ${study_list}; do\n",
    "    for ancestry in ${ancestry_list};do\n",
    "        echo -e \"\\n===============\\nProcessing ${study}_${ancestry}\\n\"\n",
    "        echo \"Making remove list\"\n",
    "        cat <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.at_cg_snps_freq_diff_gt_0.2 | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.discordant_names | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.discordant_positions | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip | tail -n +2) | \\\n",
    "              sort -u > ${base_dir}/processing/${study}/strand_check/${ancestry}_snps.remove\n",
    "\n",
    "        # Create flip list\n",
    "        echo \"Making flip list\"\n",
    "        comm -23 <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.discordant_alleles | tail -n +2 | sort -u) \\\n",
    "                 <(cut -f2,2 ${base_dir}/processing/${study}/strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2 | sort -u) \\\n",
    "                 > ${base_dir}/processing/${study}/strand_check/${ancestry}_snps.flip\n",
    "\n",
    "        # Apply filters\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed     ${base_dir}/data/${study}/genotype/observed/${ancestry}/*bed \\\n",
    "            --bim     ${base_dir}/data/${study}/genotype/observed/${ancestry}/*bim \\\n",
    "            --fam     ${base_dir}/data/${study}/genotype/observed/${ancestry}/*fam \\\n",
    "            --exclude ${base_dir}/processing/${study}/strand_check/${ancestry}_snps.remove \\\n",
    "            --flip    ${base_dir}/processing/${study}/strand_check/${ancestry}_snps.flip \\\n",
    "            --make-bed \\\n",
    "            --out     ${base_dir}/processing/${study}/${ancestry}_filtered\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   17399 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered.bim\n",
      "   17410 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered.bim\n",
      "   23484 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered.bim\n",
      "   32857 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered.bim\n",
      "   41847 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered.bim\n",
      "   24197 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered.bim\n",
      "   44668 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered.bim\n",
      "   35436 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered.bim\n",
      "   36003 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered.bim\n",
      "   36003 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered.bim\n",
      "  309304 total\n"
     ]
    }
   ],
   "source": [
    "wc -l $base_dir/processing/*/*filtered.bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove monomorphic variants\n",
    "Monomorphic variants prevent MIS from accepting the genotype data. In this case, an arbitrarily small MAF is set that is smaller than the lower bound for these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "17399 variants loaded from .bim file.\n",
      "2008 people (1423 males, 585 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2008 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.990259.\n",
      "884 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "16515 variants and 2008 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "17410 variants loaded from .bim file.\n",
      "1128 people (942 males, 186 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1128 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.992074.\n",
      "1708 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "15702 variants and 1128 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "23484 variants loaded from .bim file.\n",
      "767 people (517 males, 250 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 767 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.974578.\n",
      "2261 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "21223 variants and 767 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "32857 variants loaded from .bim file.\n",
      "828 people (495 males, 333 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 828 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.976798.\n",
      "8371 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "24486 variants and 828 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "41847 variants loaded from .bim file.\n",
      "84 people (59 males, 25 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 84 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.914726.\n",
      "4773 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "37074 variants and 84 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "24197 variants loaded from .bim file.\n",
      "33 people (18 males, 15 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 33 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.884547.\n",
      "4478 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "19719 variants and 33 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "44668 variants loaded from .bim file.\n",
      "94 people (63 males, 31 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 94 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.905134.\n",
      "4895 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "39773 variants and 94 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "35436 variants loaded from .bim file.\n",
      "44 people (24 males, 20 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 44 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.927415.\n",
      "9045 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "26391 variants and 44 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "36003 variants loaded from .bim file.\n",
      "1072 people (766 males, 306 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1072 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.795883.\n",
      "8037 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "27966 variants and 1072 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered\n",
      "  --maf 0.000001\n",
      "  --make-bed\n",
      "  --memory 2048\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 2048 MB for main workspace.\n",
      "Allocated 1536 MB successfully, after larger attempt(s) failed.\n",
      "36003 variants loaded from .bim file.\n",
      "988 people (658 males, 330 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 988 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.959739.\n",
      "9358 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "26645 variants and 988 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "### Bash ###\n",
    "\n",
    "# Apply filters\n",
    "for study in ${study_list}; do\n",
    "    for ancestry in ${ancestry_list};do\n",
    "        geno_dir=${base_dir}/processing/${study}\n",
    "\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bfile ${geno_dir}/${ancestry}_filtered \\\n",
    "            --maf 0.000001 \\\n",
    "            --make-bed \\\n",
    "            --out ${geno_dir}/${ancestry}_filtered_mono\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   16515 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono.bim\n",
      "   15702 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono.bim\n",
      "   21223 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono.bim\n",
      "   24486 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono.bim\n",
      "   37074 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono.bim\n",
      "   19719 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono.bim\n",
      "   39773 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono.bim\n",
      "   26391 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono.bim\n",
      "   27966 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono.bim\n",
      "   26645 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono.bim\n",
      "  255494 total\n"
     ]
    }
   ],
   "source": [
    "wc -l $base_dir/processing/*/*mono.bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Calculating intersection between aa uhs1 uhs2 uhs3_v1-2 uhs3_v1-3 uhs4...\n",
      "\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_aa_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/aa_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_aa_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "16515 variants loaded from .bim file.\n",
      "2008 people (1423 males, 585 females) loaded from .fam.\n",
      "--extract: 4581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 2008 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99358.\n",
      "4581 variants and 2008 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_aa_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_aa_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_aa_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_aa_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/aa_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_aa_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "21223 variants loaded from .bim file.\n",
      "767 people (517 males, 250 females) loaded from .fam.\n",
      "--extract: 4581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 767 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.97656.\n",
      "4581 variants and 767 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_aa_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_aa_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_aa_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_aa_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/aa_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_aa_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "37074 variants loaded from .bim file.\n",
      "84 people (59 males, 25 females) loaded from .fam.\n",
      "--extract: 4581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 84 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.973729.\n",
      "4581 variants and 84 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_aa_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_aa_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_aa_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_aa_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/aa_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_aa_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "39773 variants loaded from .bim file.\n",
      "94 people (63 males, 31 females) loaded from .fam.\n",
      "--extract: 4581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 94 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.965617.\n",
      "4581 variants and 94 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_aa_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_aa_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_aa_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_aa_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/aa_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_aa_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "27966 variants loaded from .bim file.\n",
      "1072 people (766 males, 306 females) loaded from .fam.\n",
      "--extract: 4581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1072 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.959536.\n",
      "4581 variants and 1072 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_aa_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_aa_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_aa_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "\n",
      "Calculating intersection between ea uhs1 uhs2 uhs3_v1-2 uhs3_v1-3 uhs4...\n",
      "\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_ea_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs1/ea_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_ea_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "15702 variants loaded from .bim file.\n",
      "1128 people (942 males, 186 females) loaded from .fam.\n",
      "--extract: 3251 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1128 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.99276.\n",
      "3251 variants and 1128 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_ea_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_ea_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs1_ea_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_ea_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs2/ea_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_ea_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "24486 variants loaded from .bim file.\n",
      "828 people (495 males, 333 females) loaded from .fam.\n",
      "--extract: 3251 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 828 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.97431.\n",
      "3251 variants and 828 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_ea_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_ea_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs2_ea_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_ea_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-2/ea_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_ea_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "19719 variants loaded from .bim file.\n",
      "33 people (18 males, 15 females) loaded from .fam.\n",
      "--extract: 3251 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 33 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.942423.\n",
      "3251 variants and 33 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_ea_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_ea_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-2_ea_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_ea_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs3_v1-3/ea_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_ea_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "26391 variants loaded from .bim file.\n",
      "44 people (24 males, 20 females) loaded from .fam.\n",
      "--extract: 3251 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 44 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.95809.\n",
      "3251 variants and 44 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_ea_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_ea_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs3_v1-3_ea_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_ea_filtered_snp_intersection.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/uhs4/ea_filtered_mono\n",
      "  --extract /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n",
      "  --make-bed\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_ea_filtered_snp_intersection\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 978 MB for main workspace.\n",
      "26645 variants loaded from .bim file.\n",
      "988 people (658 males, 330 females) loaded from .fam.\n",
      "--extract: 3251 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 988 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.959802.\n",
      "3251 variants and 988 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_ea_filtered_snp_intersection.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_ea_filtered_snp_intersection.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/uhs4_ea_filtered_snp_intersection.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "studies=($study_list)  #studies=(UHS1 UHS2 UHS3_v1-2 UHS3_v1-3) # array of study names\n",
    "num=${#studies[@]}\n",
    "\n",
    "# Get intersection set\n",
    "for ancestry in ${ancestry_list};do\n",
    "    bim_files=()\n",
    "    for (( i=0; i<${num}; i++ ));do\n",
    "        bim_files+=(${base_dir}/processing/${studies[$i]}/${ancestry}_filtered_mono.bim)\n",
    "    done\n",
    "    \n",
    "    echo -e \"\\nCalculating intersection between $ancestry ${study_list}...\\n\"\n",
    "    cat ${bim_files[@]} | cut -f2 | sort | uniq -c | awk -v num=$num '$1 == num {print $2}' \\\n",
    "        > ${base_dir}/processing/intersect/${ancestry}_variant_intersection.txt\n",
    "\n",
    "    # Make new PLINK binary file sets\n",
    "    for study in ${studies[@]}; do\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile ${base_dir}/processing/${study}/${ancestry}_filtered_mono \\\n",
    "            --extract ${base_dir}/processing/intersect/${ancestry}_variant_intersection.txt \\\n",
    "            --make-bed \\\n",
    "            --out ${base_dir}/processing/intersect/${study}_${ancestry}_filtered_snp_intersection\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  4581 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_variant_intersection.txt\n",
      "  3251 /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_variant_intersection.txt\n",
      "  7832 total\n"
     ]
    }
   ],
   "source": [
    "ww $base_dir/processing/intersect/*section.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge test\n",
    "As a final check to confirm that our data sets are all compatible, a PLINK file set merge is conducted. If any issues persist then an error will be raised."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating aa merge-list\n",
      "\n",
      "\n",
      "======== aa ========\n",
      "\n",
      "\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --memory 4000\n",
      "  --merge-list /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_merge_list.txt\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 4000 MB for main workspace.\n",
      "Allocated 1265 MB successfully, after larger attempt(s) failed.\n",
      "Performing single-pass merge (4025 people, 4581 variants).\n",
      "Merged fileset written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged-merge.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged-merge.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged-merge.fam\n",
      ".\n",
      "4581 variants loaded from .bim file.\n",
      "4025 people (2828 males, 1197 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 4025 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.980202.\n",
      "4581 variants and 4025 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/aa_studies_merged.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "Creating ea merge-list\n",
      "\n",
      "\n",
      "======== ea ========\n",
      "\n",
      "\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --memory 4000\n",
      "  --merge-list /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_merge_list.txt\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 4000 MB for main workspace.\n",
      "Allocated 1265 MB successfully, after larger attempt(s) failed.\n",
      "Performing single-pass merge (3021 people, 3251 variants).\n",
      "Merged fileset written to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged-merge.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged-merge.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged-merge.fam\n",
      ".\n",
      "3251 variants loaded from .bim file.\n",
      "3021 people (2137 males, 884 females) loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 3021 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.975869.\n",
      "3251 variants and 3021 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "--make-bed to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged.bed\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged.bim\n",
      "+\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/intersect/ea_studies_merged.fam\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "chr=23 # comment out if not running chrX exclusively\n",
    "for ancestry in $ancestry_list;do\n",
    "\n",
    "    echo \"Creating $ancestry merge-list\"\n",
    "    touch ${base_dir}/processing/intersect/${ancestry}_merge_list.txt\n",
    "    for study in $study_list;do\n",
    "        echo ${base_dir}/processing/intersect/${study}_${ancestry}_filtered_snp_intersection >>\\\n",
    "             ${base_dir}/processing/intersect/${ancestry}_merge_list.txt\n",
    "    done\n",
    "    \n",
    "    if [ $chr==23 ]; then\n",
    "        out_file=${ancestry}_studies_merged_chrx\n",
    "    else\n",
    "        out_file=${ancestry}_studies_merged\n",
    "    fi\n",
    "\n",
    "# Merge file sets\n",
    "    echo -e \"\\n\\n======== ${ancestry} ========\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 4000 \\\n",
    "        --merge-list ${base_dir}/processing/intersect/${ancestry}_merge_list.txt \\\n",
    "        --make-bed \\\n",
    "        --out ${base_dir}/processing/intersect/$out_file\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wc -l $base_dir/processing/intersect/*merged*bim\n",
    "  4581 /shared/jmarks/hiv/uhs1234/chrX_processing//processing/intersect/aa_studies_merged_chrx.bim\n",
    "  3251 /shared/jmarks/hiv/uhs1234/chrX_processing//processing/intersect/ea_studies_merged_chrx.bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation preparation for Michigan Imputation Server\n",
    "Visit the [MIS Getting Started Webpage](https://imputationserver.sph.umich.edu/start.html#!pages/help) for more information about the preparing the data for upload to MIS.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VCF File Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No aa subjects removed\n",
      "No ea subjects removed\n"
     ]
    }
   ],
   "source": [
    "### Split by chr and remove any individuals with missing data for whole chr\n",
    "\n",
    "chr=23\n",
    "for ancestry in $ancestry_list;do\n",
    "\n",
    "\n",
    "    if [ $chr==23 ]; then\n",
    "        myfile=${ancestry}_studies_merged_chrx\n",
    "    else\n",
    "        myfile=${ancestry}_studies_merged\n",
    "    fi\n",
    "    \n",
    "    # Remove SNPs\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 4000 \\\n",
    "        --bfile ${base_dir}/processing/intersect/$myfile \\\n",
    "        --chr ${chr} \\\n",
    "        --mind 0.99 \\\n",
    "        --make-bed \\\n",
    "        --out ${base_dir}/processing/impute_prep/${ancestry}_chr${chr}_for_phasing \n",
    "done > ${base_dir}/processing/impute_prep/chr_splitting.log \n",
    "\n",
    "\n",
    "### Split by chr and remove any individuals with missing data for whole chr\n",
    "\n",
    "# if merged data sets together\n",
    "#for ancestry in $ancestry_list;do\n",
    "#    for chr in {1..23};do\n",
    "#        # Remove SNPs\n",
    "#        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "#            --noweb \\\n",
    "#            --memory 4000 \\\n",
    "#            --bfile ${base_dir}/processing/intersect/${ancestry}_studies_merged \\\n",
    "#            --chr ${chr} \\\n",
    "#            --mind 0.99 \\\n",
    "#            --make-bed \\\n",
    "#            --out ${base_dir}/processing/impute_prep/${ancestry}_chr${chr}_for_phasing \n",
    "#    done > ${base_dir}/processing/impute_prep/chr_splitting.log \n",
    "#done\n",
    "\n",
    "for ancestry in $ancestry_list; do\n",
    "    grep removed $base_dir/processing/impute_prep/$ancestry*log |\n",
    "        perl -lne '/(\\d+)(\\speople)/;\n",
    "             $mycount += $1; \n",
    "             print $mycount if eof'  > $base_dir/processing/impute_prep/$ancestry.removed\n",
    "    any_removed=$(cat $base_dir/processing/impute_prep/$ancestry.removed)\n",
    "    if [ \"$any_removed\" == 0 ]; then\n",
    "        echo \"No $ancestry subjects removed\"\n",
    "    else\n",
    "        echo \"Some $ancestry subjects removed\"\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/aa_chr23_final.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/../aa_chr23_for_phasing\n",
      "  --memory 5000\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/aa_chr23_final\n",
      "  --output-chr M\n",
      "  --recode vcf bgz\n",
      "  --set-hh-missing\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 5000 MB for main workspace.\n",
      "Allocated 1185 MB successfully, after larger attempt(s) failed.\n",
      "4581 variants loaded from .bim file.\n",
      "4025 people (2828 males, 1197 females) loaded from .fam.\n",
      "Using up to 2 threads (change this with --threads).\n",
      "Before main variant filters, 4025 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.980202.\n",
      "4581 variants and 4025 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Warning: Underscore(s) present in sample IDs.\n",
      "--recode vcf bgz to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/aa_chr23_final.vcf.gz\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/ea’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/ea/ea_chr23_final.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/ea/../ea_chr23_for_phasing\n",
      "  --memory 5000\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/ea/ea_chr23_final\n",
      "  --output-chr M\n",
      "  --recode vcf bgz\n",
      "  --set-hh-missing\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 5000 MB for main workspace.\n",
      "Allocated 1185 MB successfully, after larger attempt(s) failed.\n",
      "3251 variants loaded from .bim file.\n",
      "3021 people (2137 males, 884 females) loaded from .fam.\n",
      "Using up to 2 threads (change this with --threads).\n",
      "Before main variant filters, 3021 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.975869.\n",
      "3251 variants and 3021 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Warning: Underscore(s) present in sample IDs.\n",
      "--recode vcf bgz to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/ea/ea_chr23_final.vcf.gz\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "# EC2 command line #\n",
    "#for chr in ${chr_list};do\n",
    "chr=23\n",
    "    for ancestry in ${ancestry_list};do\n",
    "        final_dir=${base_dir}/processing/impute_prep/${ancestry}\n",
    "        mkdir $final_dir\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 5000 \\\n",
    "            --bfile ${final_dir}/../${ancestry}_chr${chr}_for_phasing \\\n",
    "            --output-chr M \\\n",
    "            --set-hh-missing \\\n",
    "            --recode vcf bgz \\\n",
    "            --out ${final_dir}/${ancestry}_chr${chr}_final\n",
    "    done\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transfer the *.vcf.gz files to local machine (per chromosome) and then upload to MIS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to Michigan Imputation Server (MIS)\n",
    "\n",
    "### Uploading parameters\n",
    "These are the parameters that were selected on MIS.\n",
    "\n",
    "__Name__: UHS1-3_merged_chrX_EA_v02\n",
    "\n",
    "__Reference Panel__ 1000G Phase 3 v5\n",
    "\n",
    "__Input Files__ File Upload <br>\n",
    "\n",
    "* Select Files - select VCF (.gz) files that were downloaded to local machine from cloud. <br>\n",
    "\n",
    "__Phasing__: ShapeIT v2.r790 (unphased) \n",
    "\n",
    "__Population__: EUR\n",
    "\n",
    "__Mode__: Quality Control & Imputation\n",
    "\n",
    "* I will not attempt to re-identify or contact research participants.\n",
    "* I will report any inadvertent data release, security breach or other data management incident of which I become aware."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Troubleshooting\n",
    "\n",
    "```\n",
    "Chromosome X check failed! \n",
    "java.io.IOException: Found haplotype 0/1 at pos 2788707 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 2825403 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 2862721 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 2985672 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3002687 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3028289 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3296294 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3542110 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3627414 for male proband 8002221030_HHG6599\n",
    "Error during manifest file creation.\n",
    "```\n",
    "\n",
    "**We are going to remove subject listed above in the MIS error message from the study.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/remove_subject’: File exists\n",
      "PLINK v1.90b4.9 64-bit (13 Oct 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/remove_subject/aa_chr23_final_v02.log.\n",
      "Options in effect:\n",
      "  --bfile /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/remove_subject/../../aa_chr23_for_phasing\n",
      "  --memory 5000\n",
      "  --noweb\n",
      "  --out /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/remove_subject/aa_chr23_final_v02\n",
      "  --output-chr M\n",
      "  --recode vcf bgz\n",
      "  --remove-fam /shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/remove_subject/remove_list.tsv\n",
      "  --set-hh-missing\n",
      "\n",
      "Note: --noweb has no effect since no web check is implemented yet.\n",
      "1957 MB RAM detected; reserving 5000 MB for main workspace.\n",
      "Allocated 1185 MB successfully, after larger attempt(s) failed.\n",
      "4581 variants loaded from .bim file.\n",
      "4025 people (2828 males, 1197 females) loaded from .fam.\n",
      "--remove-fam: 4024 people remaining.\n",
      "Using up to 2 threads (change this with --threads).\n",
      "Before main variant filters, 4024 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.980223.\n",
      "4581 variants and 4024 people pass filters and QC.\n",
      "Note: No phenotypes present.\n",
      "Warning: Underscore(s) present in sample IDs.\n",
      "--recode vcf bgz to\n",
      "/shared/jmarks/hiv/uhs1-4/chrX_processing/processing/impute_prep/aa/remove_subject/aa_chr23_final_v02.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "remove_dir=$base_dir/processing/impute_prep/aa/remove_subject\n",
    "mkdir $remove_dir\n",
    "ancestry=aa\n",
    "chr=23\n",
    "echo 8002221030_HHG6599_AS93-4798 8002221030_HHG6599_AS93-4798 > ${remove_dir}/remove_list.tsv\n",
    "\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 5000 \\\n",
    "        --bfile ${remove_dir}/../../${ancestry}_chr${chr}_for_phasing \\\n",
    "        --output-chr M \\\n",
    "        --set-hh-missing \\\n",
    "        --remove-fam $remove_dir/remove_list.tsv \\\n",
    "        --recode vcf bgz \\\n",
    "        --out ${remove_dir}/${ancestry}_chr${chr}_final_v02"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Imputed Results from MIS\n",
    "First, download the data form the Michigan Imputation Server (MIS) by clicking on the link provided in the email they send out to alert you that your data has finished. Here you will find commands for downloading the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 (share drive)\n",
    "# Note: jmarks is a symlink; you will have to save results\n",
    "# to the share drive because they are very large)\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p /home/ec2-user/jmarks/hiv/uhs123/data/uhs123_merged/genotype/imputed/{ea,aa}/chr23"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ChrX\n",
    "### EA\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email.\n",
    "* EA password: z@j1CRuP:Diw3Y\n",
    "\n",
    "Then, inflate imputation results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry=ea\n",
    "study=uhs1-4\n",
    "passW=\"'z@j1CRuP:Diw3Y'\"\n",
    "cd /shared/jmarks/hiv/uhs1234/gwas/genotype/imputed/$ancestry\n",
    "\n",
    "\n",
    "# download.file\n",
    "####################################################################################################\n",
    "####################################################################################################\n",
    "# QC-results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/3e003fabde51f68742c3df72b19720dc/qcreport.html\n",
    "\n",
    "# Imputation Results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/ce8b8d3a6e056b00d429efb863350952/chr_X.no.auto_female.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/20d74f774326075473cf154d4a619622/chr_X.no.auto_male.zip\n",
    "\n",
    "# Logs\n",
    "wget https://imputationserver.sph.umich.edu/share/results/d5d5d852884d1b437eec9665d880641/chr_X.no.auto_female.log\n",
    "wget https://imputationserver.sph.umich.edu/share/results/1afd40c6ed931bf4530a50e2d41eb764/chr_X.no.auto_male.log\n",
    "\n",
    "\n",
    "# SNP Statistics\n",
    "wget https://imputationserver.sph.umich.edu/share/results/d7270cc7271a914540a9a11aed68baf3/statistics.txt\n",
    "\n",
    "####################################################################################################\n",
    "####################################################################################################\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name MIS.dowload.$study.$ancestry.chrx \\\n",
    "    --script_prefix imputed.data.download.$ancestry.chrx \\\n",
    "    --mem 3 \\\n",
    "    --nslots 1 \\\n",
    "    --priority 0 \\\n",
    "    --program bash download.file\n",
    "\n",
    "# inflate chr results\n",
    "for file in *zip; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name unzip.$study.$ancestry.chrx \\\n",
    "        --script_prefix unzip.imputed.$study.$ancestry.chrx.data \\\n",
    "        --mem 3 \\\n",
    "        --nslots 2 \\\n",
    "        --priority 0 \\\n",
    "        --program unzip -P $passW $file \n",
    "done\n",
    "\n",
    "# we can remove the original imputed data from MIS after we inflate the zip files\n",
    "rm -rf *zip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AA\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email.\n",
    "* AA password: LFkee+tMo31WV:\n",
    "\n",
    "Then, inflate imputation results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry=aa\n",
    "study=uhs1-4\n",
    "passW=\"'LFkee+tMo31WV:'\"\n",
    "cd /shared/jmarks/hiv/uhs1234/gwas/genotype/imputed/$ancestry\n",
    "\n",
    "\n",
    "# download.file\n",
    "####################################################################################################\n",
    "####################################################################################################\n",
    "# QC-results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/dae3adc9d1820e3e7b6108ed82c8851c/qcreport.html\n",
    "\n",
    "# Imputation Results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/fef581304eaa1ce12cfaaf6d570a5dfb/chr_X.no.auto_female.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/851f851b7dbe42aac869cfc6c4b465c9/chr_X.no.auto_male.zip\n",
    "\n",
    "# Logs\n",
    "wget https://imputationserver.sph.umich.edu/share/results/f65c749682821647a9cd40597588807a/chr_X.no.auto_female.log\n",
    "wget https://imputationserver.sph.umich.edu/share/results/7a8fca2dc09fe487c09fe80220be8ea6/chr_X.no.auto_male.log\n",
    "\n",
    "# SNP Statistics\n",
    "wget https://imputationserver.sph.umich.edu/share/results/15699900e18d5b91e416476d33b1f452/statistics.txt\n",
    "\n",
    "####################################################################################################\n",
    "####################################################################################################\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name MIS.dowload.$study.$ancestry.chrx \\\n",
    "    --script_prefix imputed.data.download.$ancestry.chrx \\\n",
    "    --mem 3 \\\n",
    "    --nslots 1 \\\n",
    "    --priority 0 \\\n",
    "    --program bash download.file\n",
    "\n",
    "# inflate chr results\n",
    "for file in *zip; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name unzip.$study.$ancestry.chrx \\\n",
    "        --script_prefix unzip.imputed.$study.$ancestry.chrx.data \\\n",
    "        --mem 3 \\\n",
    "        --nslots 2 \\\n",
    "        --priority 0 \\\n",
    "        --program unzip -P $passW $file \n",
    "done\n",
    "\n",
    "# we can remove the original imputed data from MIS after we inflate the zip files\n",
    "rm -rf *zip\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry=aa\n",
    "# mv inflated results to respective directories and upload to s3\n",
    "files=\"chrX.no.auto_male.dose.vcf.gz chrX.no.auto_male.dose.vcf.gz.tbi chrX.no.auto_male.info.gz chr_X.no.auto_male.log \"\n",
    "for file in $files; do\n",
    "    aws s3 cp $file s3://rti-hiv/hiv_uhs1234/data/genotype/imputed/$ancestry/$file --quiet &\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
