{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\" style=\"margin-top: 1em;\"><ul class=\"toc-item\"><li><span><a href=\"#UHS1–4-EA+AA-2df-Meta-Analysis-for-HIV-Acquisition\" data-toc-modified-id=\"UHS1–4-EA+AA-2df-Meta-Analysis-for-HIV-Acquisition-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>UHS1–4 EA+AA 2df Meta-Analysis for HIV Acquisition</a></span></li><li><span><a href=\"#Download-Data-&amp;-Create-Directory-Structure\" data-toc-modified-id=\"Download-Data-&amp;-Create-Directory-Structure-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Download Data &amp; Create Directory Structure</a></span></li><li><span><a href=\"#Results\" data-toc-modified-id=\"Results-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Results</a></span><ul class=\"toc-item\"><li><span><a href=\"#cross-ancestry\" data-toc-modified-id=\"cross-ancestry-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>cross-ancestry</a></span></li></ul></li><li><span><a href=\"#Code\" data-toc-modified-id=\"Code-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Code</a></span><ul class=\"toc-item\"><li><span><a href=\"#methods\" data-toc-modified-id=\"methods-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>methods</a></span></li><li><span><a href=\"#Secondary-scripts\" data-toc-modified-id=\"Secondary-scripts-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Secondary scripts</a></span><ul class=\"toc-item\"><li><span><a href=\"#merge_metal\" data-toc-modified-id=\"merge_metal-4.2.1\"><span class=\"toc-item-num\">4.2.1&nbsp;&nbsp;</span>merge_metal</a></span></li><li><span><a href=\"#qsub_metal.sh\" data-toc-modified-id=\"qsub_metal.sh-4.2.2\"><span class=\"toc-item-num\">4.2.2&nbsp;&nbsp;</span>qsub_metal.sh</a></span></li></ul></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UHS1–4 EA+AA 2df Meta-Analysis for HIV Acquisition\n",
    "**Author:** Jesse Marks <br>\n",
    "**Results**: The results of this meta-analysis will be stored on AWS S3 at: `s3:///rti-hiv/meta_new/020`<br>\n",
    "**GitHub Issue:** [#97](https://github.com/RTIInternational/bioinformatics/issues/97)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Data & Create Directory Structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseD=/shared/jmarks/hiv/meta\n",
    "\n",
    "## Download data from S3\n",
    "\n",
    "## Create directory structure then mv data over to respective directories\n",
    "mkdir -p $baseD/data/uhs1234/{1df,2df}/{ea,aa}\n",
    "mkdir -p $baseD/results/uhs1234/{1df,2df}/001/{processing,final}\n",
    "\n",
    "## unzip data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cross-ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"IRdisplay\")\n",
    "display_png(file=\"C://Users/Jess/OneDrive - Research Triangle Institute/Projects/HIV/uhs1-4/meta/002/uhs1234_hiv_acquisition_1df_meta_analysis_afr+eur.1df.snps+indels.manhattan.png\", width=800, height=800)\n",
    "display_png(file=\"C://Users/Jess/OneDrive - Research Triangle Institute/Projects/HIV/uhs1-4/meta/002/uhs1234_hiv_acquisition_1df_meta_analysis_afr+eur.1df.snps+indels.qq.png\", width=800, height=800)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code\n",
    "Below are the methods to perform the meta-analysis in the AWS environment. The main parts that need modified are the paths to the cohorts' data which has been modified. For example, the R section at the end on lines 315 and 316.\n",
    "\n",
    "Also note that we are converting everything to the 1000 Genome Phase 3 format and therefore we will include the script for performing this below.\n",
    "\n",
    "## methods\n",
    "Perform meta-analysis with METAL and produce the Manhattan and QQ plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=\"2df\" # degrees of freedom (1df or 2df)\n",
    "study=\"uhs1-4_aa+uhs1-4_ea\" # prepended to results files and for directory structure\n",
    "study_num=020\n",
    "filePrefix=\"hiv_acquisition_${df}_meta_analysis_${study}\"\n",
    "outDir=\"/shared/jmarks/hiv/meta/results/${study_num}\"\n",
    "dataDir=\"/shared/jmarks/hiv/meta/data\"\n",
    "minchr=1 # first chr in sequence\n",
    "maxchr=2  # 22 or 23 \n",
    "\n",
    "interactive_beta=\"beta_SNP_gender\"\n",
    "interactive_std=\"sebeta_SNP_gender\"\n",
    "interactive_cov=\"cov_SNP_int_SNP_gender\"\n",
    "mkdir -p $outDir/{processing,final}\n",
    "\n",
    "\n",
    "# functions to create cohort paths. arg1=\"base/directory/path\" arg2=\"chromosome#\"\n",
    "uhs1234_aa() { echo $1/uhs1234/$df/aa/uhs1234.aa.$df.1000G_p3.chr$2.HIV_ACQ~SNP+AGE+SEX+SNPbySEX+PC10+PC9+PC2+PC6.maf_gt_0.01_subject+afr.rsq_gt_0.30.stats ;}\n",
    "uhs1234_ea() { echo $1/uhs1234/$df/ea/uhs1234.ea.$df.1000G_p3.chr$2.HIV_ACQ~SNP+AGE+SEX+SNPbySEX+PC1+PC9+PC10.maf_gt_0.01_subject+eur.rsq_gt_0.30.stats ;}\n",
    "#vidus_ea() { echo $1/vidus/$df/ea/ ;}\n",
    "\n",
    "\n",
    "#NOTE, the following loop only works running in a new vim shell without any\n",
    "# predefined function. You could do this manually if you otherwise wanted to.\n",
    "# create capitalized cohort labels based off of the above functions\n",
    "cohortLabels=\"\"\n",
    "for function in $(compgen -A function);do\n",
    "    cohortLabels+=$(echo ${function^^})\n",
    "    cohortLabels+=\" \"\n",
    "done\n",
    "\n",
    "# remove \"MODULE\" function\n",
    "echo $cohortLabels\n",
    "cohortLabels=${cohortLabels//\"MODULE\"}\n",
    "echo $cohortLabels\n",
    "\n",
    "# the cohort labels below were retrieved from running the loop above\n",
    "cohortLabels=\"UHS1234_AA UHS1234_EA\"\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n",
    "##name    chrom   position        A1      A2      Freq1   MAF     Quality Rsq     n       Mean_predictor_allele   beta_SNP_addA1  sebeta_SNP_addA1      chi2_SNP_add     chi     p       or_95_percent_ci ##\n",
    "\n",
    "if [ $df == \"1df\" ];then\n",
    "    pv=\"p\"\n",
    "    scheme=\"STDERR\"\n",
    "elif [ $df == \"2df\" ]; then\n",
    "    pv=\"p_2df\"\n",
    "    scheme=\"INTERACTION\"\n",
    "fi\n",
    "\n",
    "### START METAL analysis ###\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++)); do\n",
    "    my_string=\"\"\n",
    "    for cohort in ${cohortLabels};do\n",
    "        string_fun() { echo \"--PROCESS\" $(${cohort,,} $1 $2); }\n",
    "        my_string+=$(string_fun $dataDir $chr)\n",
    "        my_string+=\" \"\n",
    "    done\n",
    "    bash ~/bin/qsub_metal.sh \\\n",
    "        --SCHEME $scheme \\\n",
    "        --PVALUE $pv \\\n",
    "        --MARKER name \\\n",
    "        --ALLELE A1 A2 \\\n",
    "        --EFFECT beta_SNP_addA1 \\\n",
    "        --STDERR sebeta_SNP_addA1 \\\n",
    "        --INTEFFECTLABEL $interactive_beta \\\n",
    "        --INTSTDERRLABEL $interactive_std \\\n",
    "        --INTCOVLABEL $interactive_cov \\\n",
    "        --GENOMICCONTROL OFF \\\n",
    "        ${my_string} \\\n",
    "        --OUTFILE $outDir/processing/${filePrefix}.chr$chr.$df .metal \\\n",
    "        --ANALYZE HETEROGENEITY \\\n",
    "        --script_prefix $outDir/processing/${filePrefix}.chr$chr.$df \\\n",
    "        --mem 3.8 \\\n",
    "        --priority 0\n",
    "done\n",
    "\n",
    "## START Merge ProbABEL and METAL results ###\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "    my_string=\"\"\n",
    "    for cohort in ${cohortLabels};do\n",
    "        string_fun() { echo $(${cohort,,} $1 $2); }\n",
    "        my_string+=$(string_fun $dataDir $chr)\n",
    "        my_string+=\" \"\n",
    "    done\n",
    "\n",
    "num_cohorts=\"$(echo $cohortLabels  | wc -w)\"\n",
    "chrom=$(printf '%0.s chrom' $(seq 1 ${num_cohorts}))\n",
    "position=$(printf '%0.s position' $(seq 1 ${num_cohorts}))\n",
    "name=$(printf '%0.s name' $(seq 1 ${num_cohorts}))\n",
    "#name=\"$(for ((i=1; i<=${num_cohorts}; i++ )); do printf '%0s name'; done)\"\n",
    "\n",
    "  /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name merge_meta_${chr} \\\n",
    "    --script_prefix $outDir/processing/${filePrefix}.chr$chr.$df \\\n",
    "    --mem 10 \\\n",
    "    --nslots 2 \\\n",
    "    --priority 0 \\\n",
    "    --program Rscript /shared/bioinformatics/software/R/merge_probabel_metal.R \\\n",
    "        --metal $outDir/processing/${filePrefix}.chr$chr.$df.metal \\\n",
    "        --chr_columns  $chrom \\\n",
    "        --pos_columns  $position \\\n",
    "        --probabel_merge_bys $name \\\n",
    "        --probabel \\\n",
    "        ${my_string} \\\n",
    "        --probabel_labels ${cohortLabels} \\\n",
    "        --out $outDir/final/${filePrefix}.chr$chr.$df \n",
    "        ${my_string} \\\n",
    "        --probabel_labels ${cohortLabels} \\\n",
    "        --out $outDir/final/${filePrefix}.chr$chr.$df \n",
    "done\n",
    "### END Merge ProbABEL and METAL results ###\n",
    "\n",
    "# Check for completion (should return/print out nothing)\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "  file=$outDir/processing/${filePrefix}.chr$chr.$df.qsub.log\n",
    "  if [ -f $file ]\n",
    "  then\n",
    "    logLineCount=$(wc -l $file | perl -lane 'print $F[0];')\n",
    "    if [ $logLineCount -eq 0 ]; then\n",
    "      echo $file empty\n",
    "    else\n",
    "      tail -n 1 $file |\n",
    "        perl -ne 'chomp; if (!/Done$/) { print \"'$file'\\n\".$_.\"\\n\"; }'\n",
    "    fi\n",
    "  else\n",
    "    echo $file missing\n",
    "  fi\n",
    "done\n",
    "\n",
    "\n",
    "### START Generate plots ###\n",
    "# create SNP table for input to plotting script\n",
    "if [ $df == \"1df\" ]; then\n",
    "    skip_col=2\n",
    "else\n",
    "\tskip_col=6\n",
    "fi\n",
    "\n",
    "\n",
    "outFile=$outDir/processing/${filePrefix}.$df.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "    inFile=$outDir/final/${filePrefix}.chr$chr.$df  \n",
    "    echo Processing $inFile\n",
    "    tail -n +2 $inFile |\n",
    "    perl -slne '/^(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)(?:\\s+\\S+){$skip_col}\\s+(\\S+)/;\n",
    "                if (($4 eq \"a\" || $4 eq \"c\" || $4 eq \"g\" || $4 eq \"t\") && ($5 eq \"a\" || $5 eq \"c\" || $5 eq \"g\" || $5 eq \"t\")) {\n",
    "                  print join(\"\\t\",$1,$2,$3,$6,\"snp\");\n",
    "                } else {\n",
    "                  print join(\"\\t\",$1,$2,$3,$6,\"indel\");\n",
    "                }' -- -skip_col=${skip_col} >> $outFile\n",
    "done\n",
    "\n",
    "\n",
    "# generate plots\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots \\\n",
    "    --script_prefix $outDir/processing/${filePrefix}.$df.plots \\\n",
    "    --mem 10 \\\n",
    "    --nslots 3 \\\n",
    "    --priority 0 \\\n",
    "    --program Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "    --in $outDir/processing/${filePrefix}.$df.table \\\n",
    "    --in_chromosomes autosomal_nonPAR \\\n",
    "    --in_header \\\n",
    "    --out $outDir/processing/${filePrefix}.$df\\\n",
    "    --col_id VARIANT_ID \\\n",
    "    --col_chromosome CHR \\\n",
    "    --col_position POSITION \\\n",
    "    --col_p P \\\n",
    "    --col_variant_type TYPE \\\n",
    "    --generate_snp_indel_manhattan_plot \\\n",
    "    --manhattan_odd_chr_color red \\\n",
    "    --manhattan_even_chr_color blue \\\n",
    "    --manhattan_points_cex 1.5 \\\n",
    "    --manhattan_cex_lab 1.5 \\\n",
    "    --manhattan_cex_axis 1.5 \\\n",
    "    --generate_snp_indel_qq_plot \\\n",
    "    --qq_lines \\\n",
    "    --qq_points_bg black \\\n",
    "    --qq_lambda\n",
    "\n",
    "mv $outDir/processing/*.png \\\n",
    " $outDir/final\n",
    "### END Generate plots ###\n",
    "\n",
    "                  \n",
    "## remove singleton SNP/indels (snps founnd in only one cohort) ###\n",
    "# python script\n",
    "python \n",
    "\n",
    "#####################################################################\n",
    "minchr = 1 # starting chr\n",
    "maxchr = 24 # ending chr plus 1\n",
    "study = \"uhs1-4_aa+uhs1-4_ea\" \n",
    "df = \"2df\"\n",
    "filePrefix = \"hiv_acquisition_{}_meta_analysis_{}.chr\".format(df, study)\n",
    "outDir = \"/shared/jmarks/hiv/meta/results/020/final/\" \n",
    "\n",
    "#####################################################################\n",
    "\n",
    "if df == \"1df\":\n",
    "    direction_col = 8\n",
    "elif df == \"2df\":\n",
    "    direction_col = 12\n",
    "\n",
    "for chrom in range(minchr,maxchr):\n",
    "    print(\"Processing chromosome \" + str(chrom) + \"...\")\n",
    "    inF = open(outDir + filePrefix + str(chrom) + \".\" + df, \"r\")\n",
    "    outF = open(outDir + filePrefix + str(chrom) + \".exclude_singletons.\" + df, \"w\")\n",
    "    firstLine = inF.readline()\n",
    "    outF.write(firstLine)\n",
    "    line = inF.readline()\n",
    "    while(line):\n",
    "        tmp = line.split()\n",
    "        if(sum([d != \"?\" for d in tmp[direction_col]]) > 1):\n",
    "            outF.write(line)\n",
    "        line = inF.readline()\n",
    "    inF.close()\n",
    "    outF.close()\n",
    "## end python script\n",
    "\n",
    "\n",
    "### START removed SNPs with p_2df==0\n",
    "python\n",
    "#####################################################################\n",
    "minchr = 1 # starting chr\n",
    "maxchr = 24 # ending chr plus 1\n",
    "study = \"uhs1-4_aa+uhs1-4_ea\"\n",
    "df = \"2df\"\n",
    "filePrefix = \"hiv_acquisition_{}_meta_analysis_{}.chr\".format(df, study)\n",
    "outDir = \"/shared/jmarks/hiv/meta/results/020/final/\"\n",
    "\n",
    "#####################################################################\n",
    "\n",
    "\n",
    "\n",
    "for chrom in range(minchr,maxchr):\n",
    "    print(\"Processing chromosome \" + str(chrom) + \"...\")\n",
    "    inF = open(outDir + filePrefix + str(chrom) + \".exclude_singletons.\" + df, \"r\")\n",
    "    outF = open(outDir + filePrefix + str(chrom) + \".exclude_singletons+p2df_eq_0.\" + df, \"w\")\n",
    "    firstLine = inF.readline()\n",
    "    print(firstLine)\n",
    "    sl = firstLine.split()\n",
    "    p2df_list = []\n",
    "    for item in sl:\n",
    "        if \"p_2df.\" in item:\n",
    "            p2df_list.append(sl.index(item))\n",
    "\n",
    "    outF.write(firstLine)\n",
    "    line = inF.readline()\n",
    "    while(line):\n",
    "        zero = False\n",
    "        tmp = line.split()\n",
    "        for index in p2df_list:\n",
    "            if tmp[index] == \"0\":\n",
    "                print(line)\n",
    "                zero = True\n",
    "                break\n",
    "        if zero == False:\n",
    "            outF.write(line)\n",
    "\n",
    "        line = inF.readline()\n",
    "    inF.close()\n",
    "    outF.close()\n",
    "### end python script\n",
    "\n",
    "\n",
    "### START Generate plots with singletons removed and SNPs with p_2df==0###\n",
    "# create SNP table for input to plotting script\n",
    "outFile=$outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outFile\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "    inFile=$outDir/final/${filePrefix}.chr$chr.exclude_singletons+p2df_eq_0.$df  \n",
    "    echo Processing $inFile\n",
    "    tail -n +2 $inFile |\n",
    "    perl -slne '/^(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)\\s+(\\S+)(?:\\s+\\S+){$skip_col}\\s+(\\S+)/;\n",
    "                if (($4 eq \"a\" || $4 eq \"c\" || $4 eq \"g\" || $4 eq \"t\") && ($5 eq \"a\" || $5 eq \"c\" || $5 eq \"g\" || $5 eq \"t\")) {\n",
    "                  print join(\"\\t\",$1,$2,$3,$6,\"snp\");\n",
    "                } else {\n",
    "                  print join(\"\\t\",$1,$2,$3,$6,\"indel\");\n",
    "                }' -- -skip_col=${skip_col} >> $outFile\n",
    "done\n",
    "\n",
    "\n",
    "# Generate plots with singletons removed\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "  --job_name gwas_plots_no_singletons \\\n",
    "  --script_prefix $outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df.plots \\\n",
    "  --mem 10 \\\n",
    "  --nslots 3 \\\n",
    "  --priority 0 \\\n",
    "  --program Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "  --in $outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df.table \\\n",
    "  --in_chromosomes autosomal_nonPAR \\\n",
    "  --in_header \\\n",
    "  --out $outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df \\\n",
    "  --col_id VARIANT_ID \\\n",
    "  --col_chromosome CHR \\\n",
    "  --col_position POSITION \\\n",
    "  --col_p P \\\n",
    "  --col_variant_type TYPE \\\n",
    "  --generate_snp_indel_manhattan_plot \\\n",
    "  --manhattan_odd_chr_color red \\\n",
    "  --manhattan_even_chr_color blue \\\n",
    "  --manhattan_points_cex 1.5 \\\n",
    "  --manhattan_cex_lab 1.5 \\\n",
    "  --manhattan_cex_axis 1.5 \\\n",
    "  --generate_snp_indel_qq_plot \\\n",
    "  --qq_lines \\\n",
    "  --qq_points_bg black \\\n",
    "  --qq_lambda\n",
    "\n",
    "\n",
    "# plot with ylim\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "  --job_name gwas_plots_no_singletons_ylim \\\n",
    "  --script_prefix $outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df.ylim_9.plots \\\n",
    "  --mem 10 \\\n",
    "  --nslots 3 \\\n",
    "  --priority 0 \\\n",
    "  --program Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "  --in $outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df.table \\\n",
    "  --in_chromosomes autosomal_nonPAR \\\n",
    "  --in_header \\\n",
    "  --out $outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.ylim_9.$df \\\n",
    "  --col_id VARIANT_ID \\\n",
    "  --col_chromosome CHR \\\n",
    "  --col_position POSITION \\\n",
    "  --col_p P \\\n",
    "  --col_variant_type TYPE \\\n",
    "  --generate_snp_indel_manhattan_plot \\\n",
    "  --manhattan_odd_chr_color red \\\n",
    "  --manhattan_even_chr_color blue \\\n",
    "  --manhattan_points_cex 1.5 \\\n",
    "  --manhattan_cex_lab 1.5 \\\n",
    "  --manhattan_cex_axis 1.5 \\\n",
    "  --generate_snp_indel_qq_plot \\\n",
    "  --qq_lines \\\n",
    "  --qq_points_bg black \\\n",
    "  --qq_lambda \\\n",
    "  --manhattan_ylim 9\n",
    "\n",
    "mv $outDir/processing/*.png \\\n",
    " $outDir/final\n",
    "\n",
    "### END Generate plots ###\n",
    "\n",
    "\n",
    "### START Filter by p-value  < 0.001 ###\n",
    "if [ $df == \"1df\" ]; then\n",
    "    pcol=7\n",
    "else\n",
    "    pcol=11\n",
    "fi\n",
    "\n",
    "outFile=$outDir/processing/${filePrefix}.$df.p_lte_0.001\n",
    "head -n 1 $outDir/final/${filePrefix}.chr1.$df > $outFile\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "  inFile=$outDir/final/${filePrefix}.chr$chr.$df\n",
    "  echo Processing $inFile\n",
    "  tail -n +2 $inFile |\n",
    "    perl -slane 'if ($F[$pcol] <= 0.001) { print; }' -- -pcol=$pcol >> $outFile\n",
    "done\n",
    "\n",
    "\n",
    "# Filter by p-value < 0.001 (exclude_singletons)\n",
    "outFile=$outDir/processing/${filePrefix}.exclude_singletons+p2df_eq_0.$df.p_lte_0.001\n",
    "head -n 1 $outDir/final/${filePrefix}.chr1.exclude_singletons+p2df_eq_0.$df > $outFile\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "  inFile=$outDir/final/${filePrefix}.chr$chr.exclude_singletons+p2df_eq_0.$df\n",
    "  echo Processing $inFile\n",
    "  tail -n +2 $inFile |\n",
    "    perl -slane 'if ($F[$pcol] <= 0.001) { print; }' -- -pcol=$pcol >> $outFile\n",
    "done\n",
    "\n",
    "# end p-value filtering\n",
    "\n",
    "\n",
    "# Sort p-value filtered file (Rscript)\n",
    "R\n",
    "df <- \"2df\"\n",
    "outDir <- \"/shared/jmarks/hiv/meta/results/020\"\n",
    "study <- \"uhs1-4_aa+uhs1-4_ea\"\n",
    "filePrefix <- paste(\"hiv_acquisition_\", df, \"_meta_analysis_\", study, sep=\"\")\n",
    "\n",
    "# excluding singletons\n",
    "inData2 <- read.table(paste(sep=\"\", outDir, \"/processing/\", filePrefix, \".exclude_singletons+p2df_eq_0.\", df, \".p_lte_0.001\"), header = TRUE)\n",
    "outData2 <- inData2[order(inData2$P.value),]\n",
    "write.csv(outData2, file=paste(sep=\"\", outDir, \"/final/\", filePrefix, \".exclude_singletons+p2df_eq_0.\", df, \".p_lte_0.001.csv\"), row.names = FALSE, quote=FALSE)\n",
    "### END Filter by p-value ###\n",
    "\n",
    "\n",
    "\n",
    "### Create FUMA file ###\n",
    "# Chr \n",
    "# Position\n",
    "# MarkerName\n",
    "# P-value\n",
    "# Allele1\n",
    "# Allele2\n",
    "# Effect\n",
    "# StdErr\n",
    "\n",
    "#MarkerName chr position Allele1 Allele2 Effect StdErr P-value Direction HetISq HetChiSq HetDf HetPVal\n",
    "base_dir=$outDir/final\n",
    "\n",
    "in_file() {\n",
    "echo $1/hiv_acquisition_2df_meta_analysis_uhs1-4_aa+uhs1-4_ea.chr$2.exclude_singletons+p2df_eq_0.2df;\n",
    " }\n",
    "outfile=$base_dir/hiv_acquisition_2df_meta_analysis_uhs1-4_aa+uhs1-4_ea.exclude_singletons+p2df_eq_0.ALL_CHR.FUMA\n",
    "echo $(in_file $base_dir 1)\n",
    "\n",
    "# print header to file\n",
    "awk '{print $1,$2,$3,$4,$5,$6,$7,$8;exit}'\\\n",
    "    $(in_file $base_dir 1) > $outfile\n",
    "for (( chr=${minchr}; chr<=${maxchr}; chr++ )); do\n",
    "    echo processing chr$chr\n",
    "    awk 'NR>=2{print $1,$2,$3,$4,$5,$6,$7,$8}'\\\n",
    "     $(in_file $base_dir $chr) >> $outfile\n",
    "done\n",
    "\n",
    "### End FUMA creation ###"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Secondary scripts\n",
    "These were scripts called in the methods file above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### merge_metal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/share/apps/R/bin/Rscript\n",
    "\n",
    "args <- commandArgs(TRUE)\n",
    "\n",
    "loop = TRUE\n",
    "hasMetalFile = FALSE\n",
    "hasProbabelFiles = FALSE\n",
    "hasProbabelLabels = FALSE\n",
    "hasChrColumns = FALSE\n",
    "hasPosColumns = FALSE\n",
    "hasProbabelMergeBys = FALSE\n",
    "hasOutFile = FALSE\n",
    "metalCsv = FALSE\n",
    "probabelCsv = FALSE\n",
    "outCsv = FALSE\n",
    "\n",
    "while (loop) {\n",
    "\n",
    "    if (args[1] == \"--metal\") {\n",
    "        metalFile = args[2]\n",
    "        hasMetalFile = TRUE\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--metal_csv\") {\n",
    "        metalCsv = TRUE\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--probabel\") {\n",
    "        moreProbabelFiles = TRUE\n",
    "        while (moreProbabelFiles) {\n",
    "            if (exists(\"probabelFiles\")) {\n",
    "                probabelFiles = c(probabelFiles, args[2])\n",
    "            } else {\n",
    "                probabelFiles = args[2]\n",
    "                hasProbabelFiles = TRUE\n",
    "            }\n",
    "            if ((length(args) > 2) && (substr(args[3],1,2) != \"--\")) {\n",
    "                args = args[2:length(args)]\n",
    "            } else {\n",
    "                moreProbabelFiles = FALSE\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--probabel_labels\") {\n",
    "        moreProbabelLabels = TRUE\n",
    "        while (moreProbabelLabels) {\n",
    "            if (exists(\"probabelLabels\")) {\n",
    "                probabelLabels = c(probabelLabels, args[2])\n",
    "            } else {\n",
    "                probabelLabels = args[2]\n",
    "                hasProbabelLabels = TRUE\n",
    "            }\n",
    "            if ((length(args) > 2) && (substr(args[3],1,2) != \"--\")) {\n",
    "                args = args[2:length(args)]\n",
    "            } else {\n",
    "                moreProbabelLabels = FALSE\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--probabel_csv\") {\n",
    "        probabelCsv = TRUE\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--chr_columns\") {\n",
    "        moreChrColumns = TRUE\n",
    "        while (moreChrColumns) {\n",
    "            if (exists(\"chrColumns\")) {\n",
    "                chrColumns = c(chrColumns, args[2])\n",
    "            } else {\n",
    "                chrColumns = args[2]\n",
    "                hasChrColumns = TRUE\n",
    "            }\n",
    "            if ((length(args) > 2) && (substr(args[3],1,2) != \"--\")) {\n",
    "                args = args[2:length(args)]\n",
    "            } else {\n",
    "                moreChrColumns = FALSE\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--pos_columns\") {\n",
    "        morePosColumns = TRUE\n",
    "        while (morePosColumns) {\n",
    "            if (exists(\"posColumns\")) {\n",
    "                posColumns = c(posColumns, args[2])\n",
    "            } else {\n",
    "                posColumns = args[2]\n",
    "                hasPosColumns = TRUE\n",
    "            }\n",
    "            if ((length(args) > 2) && (substr(args[3],1,2) != \"--\")) {\n",
    "                args = args[2:length(args)]\n",
    "            } else {\n",
    "                morePosColumns = FALSE\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--probabel_merge_bys\") {\n",
    "        moreProbabelMergeBys = TRUE\n",
    "        while (moreProbabelMergeBys) {\n",
    "            if (exists(\"probabelMergeBys\")) {\n",
    "                probabelMergeBys = c(probabelMergeBys, args[2])\n",
    "            } else {\n",
    "                probabelMergeBys = args[2]\n",
    "                hasProbabelMergeBys = TRUE\n",
    "            }\n",
    "            if ((length(args) > 2) && (substr(args[3],1,2) != \"--\")) {\n",
    "                args = args[2:length(args)]\n",
    "            } else {\n",
    "                moreProbabelMergeBys = FALSE\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--out\") {\n",
    "        outFile = args[2]\n",
    "        hasOutFile = TRUE\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--out_csv\") {\n",
    "        outCsv = TRUE\n",
    "    }\n",
    "    \n",
    "    if (length(args) > 1) {\n",
    "        args = args[2:length(args)]\n",
    "    } else {\n",
    "        loop=FALSE\n",
    "    }\n",
    "}\n",
    "\n",
    "if (!hasMetalFile) {\n",
    "    cat(\"No METAL file specified\\n\")\n",
    "    q(save = \"no\", status = 1, runLast = FALSE)  # terminate R session if no METAL files included in argument\n",
    "}\n",
    "if (!hasProbabelFiles) {\n",
    "    cat(\"No Probabel files specified\\n\")\n",
    "    q(\"no\", 1, FALSE) \n",
    "}\n",
    "if (!hasProbabelLabels) {\n",
    "    cat(\"No Probabel labels specified\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (!hasChrColumns) {\n",
    "    cat(\"No chromosome columns specified\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (!hasPosColumns) {\n",
    "    cat(\"No position columns specified\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (!hasProbabelMergeBys) {\n",
    "    cat(\"No merge by columns specified\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (!hasOutFile) {\n",
    "    cat(\"No output file specified\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "\n",
    "probabelFileCount = length(probabelFiles)\n",
    "if (length(probabelLabels) != probabelFileCount) {\n",
    "    cat(\"# of Probabel labels must match # of Probabel files\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (length(chrColumns) != probabelFileCount) {\n",
    "    cat(\"# of chromosome columns must match # of Probabel files\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (length(posColumns) != probabelFileCount) {\n",
    "    cat(\"# of position columns must match # of Probabel files\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "if (length(probabelMergeBys) != probabelFileCount) {\n",
    "    cat(\"# of merge by columns must match # of Probabel files\\n\")\n",
    "    q(\"no\", 1, FALSE)\n",
    "}\n",
    "\n",
    "cat(\"Reading\", metalFile, \"\\n\")\n",
    "if (metalCsv) {\n",
    "    merged = read.csv(metalFile, header=TRUE)\n",
    "} else {\n",
    "    merged = read.table(metalFile, header=TRUE)\n",
    "}\n",
    "for (i in 1:probabelFileCount) {\n",
    "    cat(\"Merging\", probabelFiles[i], \"\\n\")\n",
    "    if (probabelCsv) {\n",
    "        probabel = read.csv(probabelFiles[i], header=TRUE)\n",
    "    } else {\n",
    "        probabel = read.table(probabelFiles[i], header=TRUE)\n",
    "    }\n",
    "    colnames(probabel) = paste(colnames(probabel), probabelLabels[i], sep=\".\")\n",
    "    merged = merge(merged, probabel, by.x = \"MarkerName\", by.y = paste(probabelMergeBys[i], probabelLabels[i], sep=\".\"), all.x = TRUE)\n",
    "    chrColumn = paste(chrColumns[i], probabelLabels[i], sep=\".\")\n",
    "    if (chrColumn %in% colnames(merged)) {\n",
    "        if (\"chr\" %in% colnames(merged)) {\n",
    "            merged[is.na(merged$chr),\"chr\"] = merged[is.na(merged$chr),chrColumn]\n",
    "        } else {\n",
    "            colnames(merged)[colnames(merged)==chrColumn] = \"chr\"\n",
    "        }\n",
    "        merged = merged[,!(colnames(merged) %in% chrColumn)]\n",
    "    }\n",
    "    posColumn = paste(posColumns[i], probabelLabels[i], sep=\".\")\n",
    "    if (posColumn %in% colnames(merged)) {\n",
    "        if (\"position\" %in% colnames(merged)) {\n",
    "            merged[is.na(merged$position),\"position\"] = merged[is.na(merged$position), posColumn]\n",
    "        } else {\n",
    "            colnames(merged)[colnames(merged)==posColumn] = \"position\"\n",
    "        }\n",
    "        merged = merged[,!(colnames(merged) %in% posColumn)]\n",
    "    }\n",
    "}\n",
    "leadingColumns = c(\"MarkerName\",\"chr\",\"position\")\n",
    "merged = cbind(merged[,colnames(merged) %in% leadingColumns], merged[,!(colnames(merged) %in% leadingColumns)])\n",
    "merged = merged[order(merged$P.value),]\n",
    "cat(\"Writing\", outFile, \"\\n\")\n",
    "if (outCsv) {\n",
    "    write.csv(merged, file = outFile, row.names = FALSE)\n",
    "} else {\n",
    "    write.table(merged, file = outFile, row.names = FALSE, quote = FALSE)\n",
    "}\n",
    "cat(\"Done\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### qsub_metal.sh\n",
    "\n",
    "Note the change at the top for metalScript. We are using a different METAL version. We had to add a patch in order to perform 2df (interaction) calculations. See email below:\n",
    "\n",
    "```css\n",
    "Dana asked me to send you the methods for running the joint 2df meta-analysis. I have attached 3 files to this email:\n",
    "1) METAL-i.docx - instructions for installing the needed software and running the meta-analysis.\n",
    "2) metal-2010-02-08-patch2.tar.gz - a patch for METAL (the meta-analysis software) that adds the joint 2df functionality.\n",
    "2) generic-metal-2010-02-08.tar.gz - the version of METAL that you will need to apply the patch to (the patch will not work on newer versions of METAL).\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/sh\n",
    "\n",
    "#metalScript=\"#!/shared/bioinformatics/software/third_party/metal_v3.25.2011/bin/metal\\n\\n\"\n",
    "metalScript=\"#!/shared/bioinformatics/software/third_party/generic-metal/bin/metal\\n\\n\"\n",
    "outFile=\"\"\n",
    "finalOutFile=\"\"\n",
    "scriptPrefix=\"\"\n",
    "mem=3.8\n",
    "priority=0\n",
    "\n",
    "while [ \"$1\" != \"\" ]; \n",
    "do\n",
    "    if [ \"$1\" == \"--MARKER\" ] || [ \"$1\" == \"--EFFECT\" ] || [ \"$1\" == \"--PVALUE\" ] || [ \"$1\" == \"--WEIGHT\" ] || [ \"$1\" == \"--GENOMICCONTROL\" ] || [ \"$1\" == \"--PROCESS\" ] || [ \"$1\" == \"--SCHEME\" ] || [ \"$1\" == \"--STDERR\" ] || [ \"$1\" == \"--INTEFFECTLABEL\" ] || [ \"$1\" == \"--INTSTDERRLABEL\" ] || [ \"$1\" == \"--INTCOVLABEL\" ]; then\n",
    "        metalScript=$metalScript$(echo $1 | sed 's/--//')\" \"$2\"\\n\"\n",
    "        shift\n",
    "    elif [ \"$1\" == \"--ALLELE\" ] || [ \"$1\" == \"--OUTFILE\" ]; then\n",
    "        metalScript=$metalScript$(echo $1 | sed 's/--//')\" \"$2\" \"$3\"\\n\"\n",
    "        if [ \"$1\" == \"--OUTFILE\" ]; then\n",
    "            outFile=$2\"1\"$3\n",
    "            finalOutFile=$2$3\n",
    "        fi\n",
    "        shift\n",
    "        shift\n",
    "    elif [ \"$1\" == \"--ANALYZE\" ]; then\n",
    "        if [ \"$2\" == \"HETEROGENEITY\" ]; then\n",
    "            metalScript=$metalScript\"ANALYZE HETEROGENEITY\\n\"\n",
    "            shift\n",
    "        else\n",
    "            metalScript=$metalScript\"ANALYZE\\n\"\n",
    "        fi\n",
    "    else\n",
    "        case $1 in\n",
    "            --script_prefix )            shift\n",
    "                                        scriptPrefix=$1\n",
    "                                        ;;\n",
    "            --mem )                        shift\n",
    "                                        mem=$1\n",
    "                                        ;;\n",
    "            --priority )                shift\n",
    "                                        priority=$1\n",
    "                                        ;;\n",
    "        esac\n",
    "    fi\n",
    "    shift\n",
    "done\n",
    "\n",
    "metalScript=$metalScript\"QUIT\\n\"\n",
    "fileMetalScript=$scriptPrefix\".metal.sh\"\n",
    "echo -e $metalScript > $fileMetalScript\n",
    "chmod 775 $fileMetalScript\n",
    "\n",
    "fileQsubScript=$scriptPrefix\".metal.qsub.sh\"\n",
    "fileQsubScriptLog=$scriptPrefix\".metal.qsub.log\"\n",
    "fileQsubScriptError=$scriptPrefix\".metal.qsub.error\"\n",
    "echo \\#$ -S /bin/bash > $fileQsubScript\n",
    "echo \\#$ -v LD_LIBRARY_PATH >> $fileQsubScript\n",
    "echo \\#  >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\#$ -l h_rt=600:00:00  >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# Job name  >> $fileQsubScript\n",
    "echo \\#$ -N METAL >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# Use current working directory >> $fileQsubScript\n",
    "echo \\#$ -cwd >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# Join stdout and stderr >> $fileQsubScript\n",
    "echo \\#$ -j y >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# pe request for MPICH2. Set your number of processors here.  >> $fileQsubScript\n",
    "echo \\# SMP stands for \"Shared Memory Jobs\" >> $fileQsubScript\n",
    "echo \\#$ -pe smp 1 >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# Run job through bash shell >> $fileQsubScript\n",
    "echo \\#$ -S /bin/bash >> $fileQsubScript\n",
    "echo \\# >> $fileQsubScript\n",
    "echo \\# The following is for reporting only. It is not really needed >> $fileQsubScript\n",
    "echo \\# to run the job. It will show up in your output file. >> $fileQsubScript\n",
    "echo echo \\\"Got \\$NSLOTS processors.\\\" >> $fileQsubScript\n",
    "echo $fileMetalScript >> $fileQsubScript\n",
    "echo mv $outFile $finalOutFile >> $fileQsubScript\n",
    "echo mv $outFile.info $finalOutFile.info >> $fileQsubScript\n",
    "\n",
    "chmod 775 $fileQsubScript\n",
    "\n",
    "vmem=`echo \"$mem + 2\" | bc`\n",
    "\n",
    "qsub -q all.q -l mem_free=${mem}G,h_vmem=${vmem}G,h_rt=24:00:00 -p $priority -o $fileQsubScriptLog -e $fileQsubScriptError $fileQsubScript\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": true,
   "toc_position": {
    "height": "586px",
    "left": "0px",
    "right": "1351.46px",
    "top": "110px",
    "width": "167px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
