{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UHS1–4 Genotype Array QC\n",
    "__Author__: Jesse Marks <br>\n",
    "\n",
    "**GitHub Issue:** [Issue #125](https://github.com/RTIInternational/bioinformatics/issues/125)\n",
    "\n",
    "This document logs the steps taken to perform pre-imputation procedures on the UHS datasets: UHS1, UHS2, UHS3 (chip version 1–2), UHS3 (chip version 1–3), and UHS4. The starting point for this analysis is after quality control of observed genotypes for each batch/wave. We will merge these data and run through our automated [genotype array QC](https://github.com/RTIInternational/biocloud_gwas_workflows/tree/master/genotype_array_qc) workflow. This will remove any overlapping samples there may be. The quality controlled genotypes are oriented on the GRCh37 plus strand. \n",
    "\n",
    "## Data retrieval and organization\n",
    "PLINK binary filesets were obtained from AWS S3 storage. Nathan Gaddis detail the whereabouts of the UHS1–3 genotype data in [this post from GitHub Issue #117](https://github.com/RTIInternational/bioinformatics/issues/117#issuecomment-469845859) and for UHS4 by email correspondence.\n",
    "\n",
    "* UHS1: `s3://rti-midas-data/studies/hiv/observed/final/uhs1.ea*`\n",
    "* UHS2: `s3://rti-midas-data/studies/hiv/observed/final/uhs2{ea,aa}*`\n",
    "* UHS3_V–2: `s3://rti-midas-data/studies/hiv/observed/final/uhs3.{ea,aa}.V1-2*`\n",
    "* UHS3_V–3: `s3://rti-midas-data/studies/hiv/observed/final/uhs3.{ea,aa}.V1-3*`\n",
    "* UHS4: `s3://rti-shared/shared_data/post_qc/uhs4/genotype/array/observed/0007/eur/uhs4.{bed,bim,fam}.gz`\n",
    "\n",
    "**Note:** that we are not using any HA data for UHS."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare analysis environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study=\"uhs1234\"\n",
    "\n",
    "## create directory structure\n",
    "baseDir=/shared/rti-shared/shared_data/pre_qc/$study/genotype/array/observed/0001\n",
    "mkdir -p $baseDir/{eur,afr}/{uhs1,uhs2,uhs3_v1-2,uhs3_v1-3,uhs4}\n",
    "mkdir -p /shared/rti-common/ref_panels/1000G/2014.10/legend_with_chr/dbsnp_b153_ids/\n",
    "\n",
    "## download reference files\n",
    "cd /shared/rti-common/ref_panels/1000G/2014.10/legend_with_chr/dbsnp_b153_ids/\n",
    "aws s3 sync s3://rti-common/ref_panels/1000G/2014.10/legend_with_chr/dbsnp_b153_ids/ .\n",
    "\n",
    "## download study genotype data\n",
    "cd $baseDir\n",
    "for ext in {bed,bim,fam}; do\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs1.aa.$ext.gz afr/uhs1/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs1.aa.chr23.$ext.gz afr/uhs1/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs1.ea.$ext.gz eur/uhs1/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs1.ea.chr23.$ext.gz eur/uhs1/\n",
    "\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs2.aa.$ext.gz afr/uhs2/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs2.aa.chr23.$ext.gz afr/uhs2/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs2.ea.$ext.gz eur/uhs2/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs2.ea.chr23.$ext.gz eur/uhs2/\n",
    "\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.aa.V1-2.$ext.gz afr/uhs3_v1-2/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.aa.V1-2.chr23.$ext.gz afr/uhs3_v1-2/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-2.$ext.gz eur/uhs3_v1-2/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-2.chr23.$ext.gz eur/uhs3_v1-2/\n",
    "\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.aa.V1-3.$ext.gz afr/uhs3_v1-3/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.aa.V1-3.chr23.$ext.gz afr/uhs3_v1-3/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-3.$ext.gz eur/uhs3_v1-3/\n",
    "    aws s3 cp s3://rti-hiv/rti-midas-data/studies/hiv/observed/final/uhs3.ea.V1-3.chr23.$ext.gz eur/uhs3_v1-3/\n",
    "\n",
    "    aws s3 cp s3://rti-shared/shared_data/post_qc/uhs4/genotype/array/observed/0007/afr/uhs4.$ext.gz afr/uhs4/\n",
    "    aws s3 cp s3://rti-shared/shared_data/post_qc/uhs4/genotype/array/observed/0007/eur/uhs4.$ext.gz eur/uhs4/\n",
    "done\n",
    "\n",
    "gunzip -r * \n",
    "\n",
    "wc -l */*/*bim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "   805863 afr/uhs1/uhs1.aa.bim\n",
    "    17396 afr/uhs1/uhs1.aa.chr23.bim\n",
    "  1395852 afr/uhs2/uhs2.aa.bim\n",
    "    23484 afr/uhs2/uhs2.aa.chr23.bim\n",
    "  1845925 afr/uhs3_v1-2/uhs3.aa.V1-2.bim\n",
    "    41884 afr/uhs3_v1-2/uhs3.aa.V1-2.chr23.bim\n",
    "  1806719 afr/uhs3_v1-3/uhs3.aa.V1-3.bim\n",
    "    44706 afr/uhs3_v1-3/uhs3.aa.V1-3.chr23.bim\n",
    "  1715530 afr/uhs4/uhs4.bim\n",
    "  \n",
    "   808822 eur/uhs1/uhs1.ea.bim\n",
    "    17408 eur/uhs1/uhs1.ea.chr23.bim\n",
    "  1820881 eur/uhs2/uhs2.ea.bim\n",
    "    32857 eur/uhs2/uhs2.ea.chr23.bim\n",
    "  1924758 eur/uhs3_v1-2/uhs3.ea.V1-2.bim\n",
    "    24202 eur/uhs3_v1-2/uhs3.ea.V1-2.chr23.bim\n",
    "  1868610 eur/uhs3_v1-3/uhs3.ea.V1-3.bim\n",
    "    35440 eur/uhs3_v1-3/uhs3.ea.V1-3.chr23.bim\n",
    "  2009074 eur/uhs4/uhs4.bim\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge autosomes with chrX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UHS1\n",
    "cd $baseDir/afr/uhs1/\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs1.aa \\\n",
    "    --bmerge uhs1.aa.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs1_afr\n",
    "#wc -l *fam\n",
    "#  2015 uhs1.aa.chr23.fam\n",
    "#  2016 uhs1.aa.fam\n",
    "#  2016 uhs1_afr.fam\n",
    "\n",
    "#wc -l *bim\n",
    "#  805863 uhs1.aa.bim\n",
    "#   17396 uhs1.aa.chr23.bim\n",
    "#  805863 uhs1_afr.bim\n",
    "\n",
    "cd $baseDir/eur/uhs1/\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs1.ea \\\n",
    "    --bmerge uhs1.ea.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs1_eur\n",
    "# wc -l *fam\n",
    "#  1140 uhs1.ea.chr23.fam\n",
    "#  1142 uhs1.ea.fam\n",
    "#  1142 uhs1_eur.fam\n",
    "\n",
    "#wc -l *bim\n",
    "#  808822 uhs1.ea.bim\n",
    "#   17408 uhs1.ea.chr23.bim\n",
    "#  808822 uhs1_eur.bim\n",
    "\n",
    "\n",
    "# UHS2\n",
    "cd $baseDir/afr/uhs2/\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs2.aa \\\n",
    "    --bmerge uhs2.aa.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs2_afr\n",
    "# wc -l *fam\n",
    "#   767 uhs2.aa.chr23.fam\n",
    "#   767 uhs2.aa.fam\n",
    "#   767 uhs2_afr.fam\n",
    "\n",
    "#wc -l *bim\n",
    "#  1395852 uhs2.aa.bim\n",
    "#    23484 uhs2.aa.chr23.bim\n",
    "#  1416262 uhs2_afr.bim\n",
    "\n",
    "cd $baseDir/eur/uhs2/\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs2.ea \\\n",
    "    --bmerge uhs2.ea.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs2_eur\n",
    "#wc -l *fam\n",
    "#   828 uhs2.ea.chr23.fam\n",
    "#   828 uhs2.ea.fam\n",
    "#   828 uhs2_eur.fam\n",
    "\n",
    "#wc -l *bim\n",
    "#  1820881 uhs2.ea.bim\n",
    "#    32857 uhs2.ea.chr23.bim\n",
    "#  1841954 uhs2_eur.bim\n",
    "\n",
    "\n",
    "# UHS3_v1-2\n",
    "cd $baseDir/afr/uhs3_v1-2\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs3.aa.V1-2 \\\n",
    "    --bmerge uhs3.aa.V1-2.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs3_v1-2_afr\n",
    "#wc -l *bim *fam\n",
    "#  1845925 uhs3.aa.V1-2.bim\n",
    "#    41884 uhs3.aa.V1-2.chr23.bim\n",
    "#  1845927 uhs3_v1-2_afr.bim\n",
    "#       84 uhs3.aa.V1-2.chr23.fam\n",
    "#       84 uhs3.aa.V1-2.fam\n",
    "#       84 uhs3_v1-2_afr.fam\n",
    "\n",
    "cd $baseDir/eur/uhs3_v1-2\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs3.ea.V1-2 \\\n",
    "    --bmerge uhs3.ea.V1-2.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs3_v1-2_eur\n",
    "#wc -l *bim *fam\n",
    "#  1924758 uhs3.ea.V1-2.bim\n",
    "#    24202 uhs3.ea.V1-2.chr23.bim\n",
    "#  1924759 uhs3_v1-2_eur.bim\n",
    "#       33 uhs3.ea.V1-2.chr23.fam\n",
    "#       33 uhs3.ea.V1-2.fam\n",
    "#       33 uhs3_v1-2_eur.fam\n",
    "\n",
    "\n",
    "# UHS3_v1-3\n",
    "cd $baseDir/afr/uhs3_v1-3\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs3.aa.V1-3 \\\n",
    "    --bmerge uhs3.aa.V1-3.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs3_v1-3_afr\n",
    "#wc -l *bim *fam\n",
    "#  1806719 uhs3.aa.V1-3.bim\n",
    "#    44706 uhs3.aa.V1-3.chr23.bim\n",
    "#  1806722 uhs3_v1-3_afr.bim\n",
    "#       94 uhs3.aa.V1-3.chr23.fam\n",
    "#       94 uhs3.aa.V1-3.fam\n",
    "#       94 uhs3_v1-3_afr.fam\n",
    "\n",
    "cd $baseDir/eur/uhs3_v1-3\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile uhs3.ea.V1-3 \\\n",
    "    --bmerge uhs3.ea.V1-3.chr23 \\\n",
    "    --make-bed \\\n",
    "    --out uhs3_v1-3_eur\n",
    "#wc -l *bim *fam\n",
    "#  1868610 uhs3.ea.V1-3.bim\n",
    "#    35440 uhs3.ea.V1-3.chr23.bim\n",
    "#  1868612 uhs3_v1-3_eur.bim\n",
    "#       44 uhs3.ea.V1-3.chr23.fam\n",
    "#       44 uhs3.ea.V1-3.fam\n",
    "#       44 uhs3_v1-3_eur.fam\n",
    "\n",
    "\n",
    "# UHS4\n",
    "cd $baseDir/afr/uhs4/\n",
    "# rename files for consistency \n",
    "for ext in {bim,bed,fam}; do\n",
    "    cp uhs4.$ext uhs4_afr.$ext\n",
    "done\n",
    "#wc -l *afr.bim *afr.fam\n",
    "# 1715530 uhs4_afr.bim\n",
    "#    1131 uhs4_afr.fam\n",
    "\n",
    "cd $baseDir/eur/uhs4/\n",
    "# rename files for consistency \n",
    "for ext in {bim,bed,fam}; do\n",
    "    cp uhs4.$ext uhs4_eur.$ext\n",
    "done\n",
    "#wc -l *eur.bim *eur.fam\n",
    "# 2009074 uhs4_eur.bim\n",
    "#    1225 uhs4_eur.fam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge ancestry specific batches/waves"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AFR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $baseDir\n",
    "\n",
    "# create merge list\n",
    "wc -l afr/uhs*/*afr.fam | awk '{print $2}' | head -n -1 | awk -F\".\" '{print $1}' > uhs_afr_merge_list.txt\n",
    "\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --merge-list uhs_afr_merge_list.txt \\\n",
    "    --make-bed \\\n",
    "    --out afr/uhs1234_afr\n",
    "# Error: 6723 variants with 3+ alleles present.\n",
    "\n",
    "# remove those variants and try the merge again.\n",
    "while read line; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --bfile $line \\\n",
    "        --exclude afr/uhs1234_afr-merge.missnp \\\n",
    "        --make-bed \\\n",
    "        --out ${line}_remove_missnps\n",
    "done < uhs_afr_merge_list.txt\n",
    "\n",
    "# create new merge list\n",
    "wc -l afr/uhs*/*remove_missnps.fam | awk '{print $2}' | head -n -1 | awk -F\".\" '{print $1}' > uhs_afr_new_merge_list.txt\n",
    "\n",
    "# try merge again\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --merge-list uhs_afr_new_merge_list.txt \\\n",
    "    --make-bed \\\n",
    "    --out afr/uhs1234_afr\n",
    "\n",
    "gzip afr/uhs1234_afr*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EUR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $baseDir\n",
    "\n",
    "# create merge list\n",
    "wc -l eur/uhs*/*eur.fam | awk '{print $2}' | head -n -1 | awk -F\".\" '{print $1}' > uhs_eur_merge_list.txt\n",
    "\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --merge-list uhs_eur_merge_list.txt \\\n",
    "    --make-bed \\\n",
    "    --out eur/uhs1234_eur\n",
    "#Error: 6717 variants with 3+ alleles present.\n",
    "\n",
    "# remove those variants and try the merge again.\n",
    "while read line; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --bfile $line \\\n",
    "        --exclude eur/uhs1234_eur-merge.missnp \\\n",
    "        --make-bed \\\n",
    "        --out ${line}_remove_missnps\n",
    "done < uhs_eur_merge_list.txt\n",
    "\n",
    "# create new merge list\n",
    "wc -l eur/uhs*/*remove_missnps.fam | awk '{print $2}' | head -n -1 | awk -F\".\" '{print $1}' > uhs_eur_new_merge_list.txt\n",
    "\n",
    "# try merge again\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --merge-list uhs_eur_new_merge_list.txt \\\n",
    "    --make-bed \\\n",
    "    --out eur/uhs1234_eur\n",
    "\n",
    "gzip eur/uhs1234_eur*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $baseDir/afr/\n",
    "\n",
    "# include a README within the version-level directory\n",
    "for file in uhs1234_afr*; do\n",
    "    aws s3 mv $file  s3://rti-shared/shared_data/pre_qc/uhs1234/genotype/array/observed/0001/afr/\n",
    "done\n",
    "\n",
    "cd $baseDir/eur/\n",
    "# include a README within the version-level directory\n",
    "for file in uhs1234_eur*; do\n",
    "    aws s3 mv $file  s3://rti-shared/shared_data/pre_qc/uhs1234/genotype/array/observed/0001/eur/\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p /shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/{wf_input,wf_output}\n",
    "mkdir -p /shared/bioinformatics/methods/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Set up config file for QC pipeline\n",
    "\n",
    "Copy JSON file from previous run and modify settings. Then edit this config file to include the appropriate cohort information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/bioinformatics/methods/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/\n",
    "cp /shared/biocloud_gwas_workflows/genotype_array_qc/test/ukb_qc_wf.json uhs1234_afr_genotype_qc_wf.json\n",
    "cp /shared/biocloud_gwas_workflows/genotype_array_qc/test/ukb_qc_wf.json uhs1234_eur_genotype_qc_wf.json\n",
    "\n",
    "# Edit wf config files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zip biocloud_gwas_workflows repo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/biocloud_gwas_workflows\n",
    "git pull\n",
    "git submodule update --init --recursive\n",
    "git rev-parse HEAD > /shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_input/git_hash.txt\n",
    "cd /shared\n",
    "zip \\\n",
    "    --exclude=*/var/* \\\n",
    "    --exclude=*.git/* \\\n",
    "    --exclude=*/test/* \\\n",
    "    --exclude=*/.idea/* \\\n",
    "    -r /shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_input/biocloud_gwas_workflows.zip \\\n",
    "    biocloud_gwas_workflows/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submit job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open session in terminal 1\n",
    "ssh -i ~/.ssh/gwas_rsa -L localhost:8000:localhost:8000 ec2-user@54.174.185.7\n",
    "\n",
    "# Submit job in terminal 2\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/shared/biocloud_gwas_workflows/genotype_array_qc/genotype_array_qc_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@/shared/bioinformatics/methods/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/uhs1234_afr_genotype_qc_wf.json\" \\\n",
    "    -F \"workflowDependencies\"=@/shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_input/biocloud_gwas_workflows.zip\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/shared/biocloud_gwas_workflows/genotype_array_qc/genotype_array_qc_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@/shared/bioinformatics/methods/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/uhs1234_eur_genotype_qc_wf.json\" \\\n",
    "    -F \"workflowDependencies\"=@/shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_input/biocloud_gwas_workflows.zip\n",
    "\n",
    "job=a06ddc1c-278a-446a-908c-24d8878e1de7 # afr\n",
    "job=4dfd555f-4897-41b4-ba47-209aa250839f # eur\n",
    "\n",
    "# Monitor job in terminal 1\n",
    "#tail -f /tmp/cromwell-server.log\n",
    "\n",
    "# check job status in terminal 2\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get final outputs\n",
    "\n",
    "In terminal with credentials for accessing rti-code S3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AFR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_output/\n",
    "echo $job # to verify you have the correct job ID saved in variable \"job\"\n",
    "\n",
    "# Download output from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${afr_job}/outputs\" -H \"accept: application/json\" \\\n",
    "  > ${job}_final_outputs.json\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${eur_job}/outputs\" -H \"accept: application/json\" \\\n",
    "  > ${job}_final_outputs.json\n",
    "\n",
    "# Copy final outputs json to S3\n",
    "aws s3 cp ${job}_final_outputs.json \\\n",
    "  s3://rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_output/\n",
    "\n",
    "# Copy final qc files from cromwell-output to final location\n",
    "python - <<EOF\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "\n",
    "job = \"$job\"\n",
    "dir = \"rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/afr\"\n",
    "\n",
    "def traverse(o, tree_types=(list, tuple)):\n",
    "    if isinstance(o, tree_types):\n",
    "        for value in o:\n",
    "            for subvalue in traverse(value, tree_types):\n",
    "                yield subvalue\n",
    "    else:\n",
    "        yield o\n",
    "\n",
    "with open(job + '_final_outputs.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for key in outputs:\n",
    "        if (type(outputs[key]) == list):\n",
    "            for value in traverse(outputs[key]):\n",
    "                if (str(value)[0:2] == \"s3\"):\n",
    "                    message = \"aws s3 cp {} s3://{}/wf_output/\".format(value, dir)\n",
    "                    os.system(message)\n",
    "        else:\n",
    "            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "                message = \"aws s3 cp {} s3://{}/wf_output/\".format(outputs[key], dir)\n",
    "                os.system(message)\n",
    "\n",
    "# Move and rename final genotype plink file set\n",
    "with open('/shared/' + dir + '/wf_output/' + job + '_final_outputs.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for ext in [\"bed\", \"bim\", \"fam\"]:\n",
    "        final = \"genotype_array_qc_wf.final_qc_{}\".format(ext)\n",
    "        for original in outputs[final]:\n",
    "            fileName = re.sub(r'.+/', '', original)\n",
    "            new = re.sub(r'uhs1234.YRI.+\\.', 'afr/uhs1234.', fileName)\n",
    "            new = re.sub(r'uhs1234.CEU.+\\.', 'eur/uhs1234.', new)\n",
    "            new = re.sub(r'uhs1234.CHB.+\\.', 'amr/uhs1234.', new)\n",
    "            final = \"s3://{}/wf_output/{}\".format(dir, new)\n",
    "            os.system(\"aws s3 cp {} {}\".format(original, final))\n",
    "\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### EUR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_output/\n",
    "echo $job # to verify you have the correct job ID saved in variable \"job\"\n",
    "\n",
    "# Download output from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/outputs\" -H \"accept: application/json\" \\\n",
    "  > ${job}_final_outputs.json\n",
    "\n",
    "# Copy final outputs json to S3\n",
    "aws s3 cp ${job}_final_outputs.json \\\n",
    "  s3://rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/wf_output/\n",
    "\n",
    "# Copy final qc files from cromwell-output to final location\n",
    "python - <<EOF\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "\n",
    "job = \"$job\"\n",
    "dir = \"rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0002/eur\"\n",
    "\n",
    "def traverse(o, tree_types=(list, tuple)):\n",
    "    if isinstance(o, tree_types):\n",
    "        for value in o:\n",
    "            for subvalue in traverse(value, tree_types):\n",
    "                yield subvalue\n",
    "    else:\n",
    "        yield o\n",
    "\n",
    "with open(job + '_final_outputs.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for key in outputs:\n",
    "        if (type(outputs[key]) == list):\n",
    "            for value in traverse(outputs[key]):\n",
    "                if (str(value)[0:2] == \"s3\"):\n",
    "                    message = \"aws s3 cp {} s3://{}/wf_output/\".format(value, dir)\n",
    "                    os.system(message)\n",
    "        else:\n",
    "            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "                message = \"aws s3 cp {} s3://{}/wf_output/\".format(outputs[key], dir)\n",
    "                os.system(message)\n",
    "\n",
    "# Move and rename final genotype plink file set\n",
    "with open('/shared/' + dir + '/wf_output/' + job + '_final_outputs.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for ext in [\"bed\", \"bim\", \"fam\"]:\n",
    "        final = \"genotype_array_qc_wf.final_qc_{}\".format(ext)\n",
    "        for original in outputs[final]:\n",
    "            fileName = re.sub(r'.+/', '', original)\n",
    "            new = re.sub(r'uhs1234.YRI.+\\.', 'afr/uhs1234.', fileName)\n",
    "            new = re.sub(r'uhs1234.CEU.+\\.', 'eur/uhs1234.', new)\n",
    "            new = re.sub(r'uhs1234.CHB.+\\.', 'amr/uhs1234.', new)\n",
    "            final = \"s3://{}/wf_output/{}\".format(dir, new)\n",
    "            os.system(\"aws s3 cp {} {}\".format(original, final))\n",
    "\n",
    "EOF"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
