{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stratified LDSC using DEGs by HIV status: UPDATE\n",
    "**author**: Jesse Marks <br>\n",
    "**date**: Dec 7, 2022 <br>\n",
    "**GitHub issue**: [166](https://github.com/RTIInternational/bioinformatics/issues/166)\n",
    "\n",
    "## Description\n",
    "See [this comment](https://github.com/RTIInternational/bioinformatics/issues/166#issuecomment-1317677579).\n",
    "**Updating previous ananlyis** \n",
    "```\n",
    "@jaamarks This paper recently came out that includes a bigger EUR-only HIV acquisition GWAS than what we currently have. Would you be able to run sLDSC as done before using the summary stats from this new study?\n",
    "Meta-analysis sumstats link: https://doi.org/10.18742/18166406.v1\n",
    "```\n",
    "\n",
    "So I downloaded the new HIV acquisition GWAS results and stored them on S3 at `s3://rti-shared/gwas_publicly_available_sumstats/hiv_acquisition_duarte2022_iscience/raw/HIV_acquisition_meta.zip`\n",
    "\n",
    "\n",
    "In addition to the HIV acquisition GWAS, we will apply the paritioned h2 analysis using the DEGs by HIV status we will apply analyze the traits:\n",
    "\n",
    "| Trait                             | Location                                                                                                                                             |\n",
    "|-----------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "| Alzheimer's disease               | s3://rti-shared/ldsc/data/alzheimers_disease_lambert2013_nat_genet/munged/alzheimers_disease_lambert2013_nat_genet.sumstats.gz                       |\n",
    "| Amyotrophic lateral sclerosis     | s3://rti-shared/ldsc/data/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet/munged/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz |\n",
    "| Asthma                            | s3://rti-shared/ldsc/data/asthma_han2020_nat_commun/munged/asthma_han2020_nat_commun.sumstats.gz                                                     |\n",
    "| Atopic dermatitis (Eczema)        | s3://rti-shared/ldsc/data/atopic_dermatitis_paternoster2015_nat_genet/munged/munged_sumstats_readme.txt                                              |\n",
    "| Crohn's disease                   | s3://rti-shared/ldsc/data/crohns_disease_liu2015_nat_genet/munged/crohns_disease_liu2015_nat_genet.sumstats.gz                                       |\n",
    "| Inflammatory Bowel Disease (Euro) | s3://rti-shared/ldsc/data/inflammatory_bowel_disease_liu2015_nat_genet/munged/inflammatory_bowel_disease_liu2015_nat_genet.sumstats.gz               |\n",
    "| Neuroticism                       | s3://rti-shared/ldsc/data/neuroticism_okbay2016_nat_genet/munged/neuroticism_okbay2016_nat_genet.sumstats.gz                                         |\n",
    "| Parkinson's disease               | s3://rti-shared/ldsc/data/parkinsons_disease_sanchez2009_nat_genet/munged/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz                       |\n",
    "| Platelet count                    | s3://rti-shared/ldsc/data/platelet_count_vuckovic2020_cell/munged/platelet_count_vuckovic2020_cell.sumstats.gz                                       |\n",
    "| Primary biliary cirrhosis         | s3://rti-shared/ldsc/data/primary_biliary_cirrhosis_cordell2015_nat_commun/munged/PASS_Primary_biliary_cirrhosis.sumstats.gz                         |\n",
    "| Primary sclerosing cholangitis    | s3://rti-shared/ldsc/data/primary_sclerosing_cholangitis_ji2017_nat_genet/munged/primary_sclerosing_cholangitis_ji2017_nat_genet.sumstats.gz         |\n",
    "| Red blood cell count              | s3://rti-shared/ldsc/data/red_blood_cell_count_vuckovic2020_cell/munged/red_blood_cell_count_vuckovic2020_cell.sumstats.gz                           |\n",
    "| Rheumatoid Arthritis              | s3://rti-shared/ldsc/data/rheumatoid_arthritis_okada2014_nature/munged/rheumatoid_arthritis_okada2014_nature.sumstats.gz                             |\n",
    "| Systemic lupus erythematosus      | s3://rti-shared/ldsc/data/systemic_lupus_erythematosus_bentham2015_nat_genet/munged/PASS_Lupus.sumstats.gz                                           |\n",
    "| Type 2 Diabetes                   | s3://rti-shared/ldsc/data/type2_diabetes_xue2018_nat_commun/munged/type2_diabetes_xue2018_nat_commun.sumstats.gz                                     |\n",
    "| Ulcerative colitis                | s3://rti-shared/ldsc/data/ulcerative_colitis_liu2015_nat_genet/munged/ulcerative_colitis_liu2015_nat_genet.sumstats.gz                               |\n",
    "| White blood cell count            | s3://rti-shared/ldsc/data/white_blood_cell_count_vuckovic2020_cell/munged/white_blood_cell_count_vuckovic2020_cell.sumstats.gz                       |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ~/rti-hiv/ldsc/hiv_acquisition/partitioned_h2/0003/{1000g,deg_bedfiles,results,sumstats}/\n",
    "baseD=~/rti-hiv/ldsc/hiv_acquisition/partitioned_h2/0003/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download ldsc files\n",
    "cd $baseD/1000g/\n",
    "\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_baseline_ldscores.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_plinkfiles.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_frq.tgz\n",
    "#wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/weights_hm3_no_hla.tgz\n",
    "\n",
    "# decompress\n",
    "for file in *tgz; do\n",
    "  tar xvzf $file\n",
    "  rm $file \n",
    "done  \n",
    "\n",
    "\n",
    "# Download HIV acquisition sumstats results\n",
    "cd $baseD/sumstats/\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/hiv_acquisition_duarte2022_iscience/raw/HIV_acquisition_meta.zip . # HIV acquisition\n",
    "unzip HIV_acquisition_meta.zip\n",
    "\n",
    "head HIV_acquisition_meta_sample_size1_filtered_annotated.txt\n",
    "#CHR_BP_hg19 CHR BP A1 A2 Neff Z P Direction HetISq HetChiSq HetDf HetPVal RSID MAF_dbsnp151 BETA SE\n",
    "#12_126890980 12 126890980 A G 17014.55 0.768 0.4425 ++++ 0 0.256 3 0.9681 rs1000000 0.165735 0.011196166098801 0.0145783412744804\n",
    "#4_57561647 4 57561647 A G 17014.55 -0.147 0.8833 +--- 0 1.443 3 0.6954 rs10000003 0.220248 -0.00192290395219447 0.013080979266629\n",
    "\n",
    "\n",
    "# munge:  docker interactive mode\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_0bb574e bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats HIV_acquisition_meta_sample_size1_filtered_annotated.txt \\\n",
    "    --snp RSID \\\n",
    "    --N-col Neff \\\n",
    "    --a1 A1 \\\n",
    "    --a2 A2 \\\n",
    "    --p P \\\n",
    "    --signed-sumstats BETA,0 \\\n",
    "    --out HIV_acquisition_meta_sample_size1_filtered_annotated_munged\n",
    "    \n",
    "# upload to s3\n",
    "aws s3 cp HIV_acquisition_meta_sample_size1_filtered_annotated_munged.sumstats.gz s3://rti-shared/ldsc/data/hiv_acquisition_duarte2022_iscience/munged/\n",
    "aws s3 cp HIV_acquisition_meta_sample_size1_filtered_annotated_munged.log s3://rti-shared/ldsc/data/hiv_acquisition_duarte2022_iscience/munged/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download sumstats formatted (munged) results\n",
    "aws s3 cp s3://rti-shared/ldsc/data/alzheimers_disease_lambert2013_nat_genet/munged/alzheimers_disease_lambert2013_nat_genet.sumstats.gz . # Alzheimer's disease\n",
    "aws s3 cp s3://rti-shared/ldsc/data/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet/munged/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz . # Amyotrophic lateral sclerosis\n",
    "aws s3 cp s3://rti-shared/ldsc/data/asthma_han2020_nat_commun/munged/asthma_han2020_nat_commun.sumstats.gz . # Asthma\n",
    "aws s3 cp s3://rti-shared/ldsc/data/atopic_dermatitis_paternoster2015_nat_genet/munged/eczema_paternoster2015_nat_genet.sumstats.gz . # Atopic dermatitis (Eczema)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/crohns_disease_liu2015_nat_genet/munged/crohns_disease_liu2015_nat_genet.sumstats.gz . # Crohn's disease\n",
    "aws s3 cp s3://rti-shared/ldsc/data/inflammatory_bowel_disease_liu2015_nat_genet/munged/inflammatory_bowel_disease_liu2015_nat_genet.sumstats.gz . # Inflammatory Bowel Disease (Euro)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/neuroticism_okbay2016_nat_genet/munged/neuroticism_okbay2016_nat_genet.sumstats.gz . # Neuroticism\n",
    "aws s3 cp s3://rti-shared/ldsc/data/parkinsons_disease_sanchez2009_nat_genet/munged/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz . # Parkinson's disease\n",
    "aws s3 cp s3://rti-shared/ldsc/data/platelet_count_vuckovic2020_cell/munged/platelet_count_vuckovic2020_cell.sumstats.gz . # Platelet count\n",
    "aws s3 cp s3://rti-shared/ldsc/data/primary_biliary_cirrhosis_cordell2015_nat_commun/munged/PASS_Primary_biliary_cirrhosis.sumstats.gz . # Primary biliary cirrhosis\n",
    "aws s3 cp s3://rti-shared/ldsc/data/primary_sclerosing_cholangitis_ji2017_nat_genet/munged/primary_sclerosing_cholangitis_ji2017_nat_genet.sumstats.gz . # Primary sclerosing cholangitis\n",
    "aws s3 cp s3://rti-shared/ldsc/data/red_blood_cell_count_vuckovic2020_cell/munged/red_blood_cell_count_vuckovic2020_cell.sumstats.gz . # Red blood cell count\n",
    "aws s3 cp s3://rti-shared/ldsc/data/rheumatoid_arthritis_okada2014_nature/munged/rheumatoid_arthritis_okada2014_nature.sumstats.gz . # Rheumatoid Arthritis\n",
    "aws s3 cp s3://rti-shared/ldsc/data/systemic_lupus_erythematosus_bentham2015_nat_genet/munged/PASS_Lupus.sumstats.gz . # Systemic lupus erythematosus\n",
    "aws s3 cp s3://rti-shared/ldsc/data/type2_diabetes_xue2018_nat_commun/munged/type2_diabetes_xue2018_nat_commun.sumstats.gz . # Type 2 Diabetes\n",
    "aws s3 cp s3://rti-shared/ldsc/data/ulcerative_colitis_liu2015_nat_genet/munged/ulcerative_colitis_liu2015_nat_genet.sumstats.gz . # Ulcerative colitis\n",
    "aws s3 cp s3://rti-shared/ldsc/data/white_blood_cell_count_vuckovic2020_cell/munged/white_blood_cell_count_vuckovic2020_cell.sumstats.gz . # White blood cell count\n",
    "    \n",
    "# Download DEGs by HIV status\n",
    "cd $baseD/deg_bedfiles/\n",
    "aws s3 sync s3://rti-hiv/scratch/bquach/hiv/stratified_ldsc/deg_bedfiles/ ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Partitioned h2 analysis\n",
    "See the LDSC wiki page  [LD-Score-Estimation-Tutorial](https://github.com/bulik/ldsc/wiki/LD-Score-Estimation-Tutorial#partitioned-ld-scores).\n",
    "\n",
    "1. Create annotation file based off of a BED formatted file (file containing chr, chr-start, chr-end)\n",
    "2. Compute the annotation-specific (partitioned) LD scores.\n",
    "3. Compute the partitioned heritability estimate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# interactive session\n",
    "docker run -it -v $PWD:/data/ \\\n",
    "    rtibiocloud/ldsc:v1.0.1_0bb574e bash\n",
    "\n",
    "# loop through each BED file\n",
    "for window in {cis10k,cis100k,cis400k}; do\n",
    "    case $window in\n",
    "        cis10k) deg_file=/data/deg_bedfiles/hiv_status_vl_suppressed_degs_cis10k.bed.gz ;;\n",
    "        cis100k) deg_file=/data/deg_bedfiles/hiv_status_vl_suppressed_degs_cis100k.bed.gz ;;\n",
    "        cis400k) deg_file=/data/deg_bedfiles/hiv_status_vl_suppressed_degs_cis400k.bed.gz ;;\n",
    "    esac \n",
    "    \n",
    "    mkdir -p /data/{annotations_ldscores,results}\n",
    "    \n",
    "    # loop through each chromosome\n",
    "        for j in {1..22}; do\n",
    "        # create annotation files\n",
    "        python /opt/ldsc/make_annot.py \\\n",
    "            --bed-file $deg_file \\\n",
    "            --bimfile \"/data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim\" \\\n",
    "            --annot-file \"/data/annotations_ldscores/hiv_status_vl_suppressed_degs_${window}_chr$j.annot.gz\"\n",
    "\n",
    "        # compute LD scores\n",
    "        python /opt/ldsc/ldsc.py \\\n",
    "            --l2 \\\n",
    "            --thin-annot \\\n",
    "            --ld-wind-cm 1 \\\n",
    "            --bfile \"/data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j\" \\\n",
    "            --print-snps \"/data/1000g/1000G_EUR_Phase3_baseline/print_snps.txt\" \\\n",
    "            --annot \"/data/annotations_ldscores/hiv_status_vl_suppressed_degs_${window}_chr$j.annot.gz\" \\\n",
    "            --out \"/data/annotations_ldscores/hiv_status_vl_suppressed_degs_${window}_chr$j\"\n",
    "    done # end chr loop\n",
    "    \n",
    "    \n",
    "    # loop through all traits\n",
    "    for trait in {\"alzheimers_disease\",\"amyotrophic_lateral_sclerosis\",\"asthma\",\"atopic_dermatitis\",\"crohns_disease\",\"hiv_acquisition\",\"inflammatory_bowel_disease\",\"neuroticism\",\"parkinsons_disease\",\"platelet_count\",\"primary_biliary_cirrhosis\",\"primary_sclerosing_cholangitis\",\"red_blood_cell_count\",\"rheumatoid_arthritis\",\"systemic_lupus_erythematosus\",\"type2_diabetes\",\"ulcerative_colitis\",\"white_blood_cell_count\"}; do\n",
    "\n",
    "        # store processing files for each meta in separate dir\n",
    "        mkdir -p /data/annotations_ldscores/${trait}/\n",
    "\n",
    "        # use sumstats files that corresponds to the trait name for the h2 estimate\n",
    "        case $trait in \n",
    "            \"alzheimers_disease\")  stats=/data/sumstats/alzheimers_disease_lambert2013_nat_genet.sumstats.gz ;;\n",
    "            \"amyotrophic_lateral_sclerosis\") stats=/data/sumstats/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz ;;\n",
    "            \"asthma\") stats=/data/sumstats/asthma_han2020_nat_commun.sumstats.gz ;;\n",
    "            \"atopic_dermatitis\") stats=/data/sumstats/eczema_paternoster2015_nat_genet.sumstats.gz ;;\n",
    "            \"crohns_disease\") stats=/data/sumstats/crohns_disease_liu2015_nat_genet.sumstats.gz ;;\n",
    "            \"hiv_acquisition\") stats=/data/sumstats/HIV_acquisition_meta_sample_size1_filtered_annotated_munged.sumstats.gz ;;\n",
    "            \"inflammatory_bowel_disease\") stats=/data/sumstats/inflammatory_bowel_disease_liu2015_nat_genet.sumstats.gz ;;\n",
    "            \"neuroticism\") stats=/data/sumstats/neuroticism_okbay2016_nat_genet.sumstats.gz ;;\n",
    "            \"parkinsons_disease\") stats=/data/sumstats/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz ;;\n",
    "            \"platelet_count\") stats=/data/sumstats/platelet_count_vuckovic2020_cell.sumstats.gz ;;\n",
    "            \"primary_biliary_cirrhosis\") stats=/data/sumstats/PASS_Primary_biliary_cirrhosis.sumstats.gz ;;\n",
    "            \"primary_sclerosing_cholangitis\") stats=/data/sumstats/primary_sclerosing_cholangitis_ji2017_nat_genet.sumstats.gz ;;\n",
    "            \"red_blood_cell_count\") stats=/data/sumstats/red_blood_cell_count_vuckovic2020_cell.sumstats.gz ;;\n",
    "            \"rheumatoid_arthritis\") stats=/data/sumstats/rheumatoid_arthritis_okada2014_nature.sumstats.gz ;;\n",
    "            \"systemic_lupus_erythematosus\") stats=/data/sumstats/PASS_Lupus.sumstats.gz ;;\n",
    "            \"type2_diabetes\") stats=/data/sumstats/type2_diabetes_xue2018_nat_commun.sumstats.gz ;;\n",
    "            \"ulcerative_colitis\") stats=/data/sumstats/ulcerative_colitis_liu2015_nat_genet.sumstats.gz ;;\n",
    "            \"white_blood_cell_count\") stats=/data/sumstats/white_blood_cell_count_vuckovic2020_cell.sumstats.gz ;;\n",
    "        esac\n",
    "    \n",
    "        \n",
    "        # computed partitioned heritability estimate\n",
    "        python /opt/ldsc/ldsc.py \\\n",
    "            --h2 $stats \\\n",
    "            --overlap-annot \\\n",
    "            --print-coefficients \\\n",
    "            --w-ld-chr \"/data/weights_hm3_no_hla/weights.\" \\\n",
    "            --ref-ld-chr \"/data/annotations_ldscores/hiv_status_vl_suppressed_degs_${window}_chr,/data/1000g/1000G_EUR_Phase3_baseline/baseline.\" \\\n",
    "            --frqfile-chr \"/data/1000g/1000G_Phase3_frq/1000G.EUR.QC.\" \\\n",
    "            --out \"/data/results/${trait}_with_hiv_status_vl_suppressed_degs_${window}_results\"\n",
    "    \n",
    "    done # end trait file loop\n",
    "done # end BED file loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop through all traits\n",
    "for trait in \"hiv_acquisition\"; do\n",
    "\n",
    "    # store processing files for each meta in separate dir\n",
    "    mkdir -p /data/annotations_ldscores/${trait}/\n",
    "    \n",
    "    # use sumstats files that corresponds to the trait name for the h2 estimate\n",
    "    stats=/data/sumstats/hiv_acquisition_gwas_meta_eur.sumstats.gz\n",
    "    \n",
    "    # loop through each BED file\n",
    "    for window in {cis10k,cis100k,cis400k}; do\n",
    "        case $window in\n",
    "            cis10k) deg_file=/data/deg_bedfiles/hiv_status_vl_suppressed_degs_cis10k.bed.gz ;;\n",
    "            cis100k) deg_file=/data/deg_bedfiles/hiv_status_vl_suppressed_degs_cis100k.bed.gz ;;\n",
    "            cis400k) deg_file=/data/deg_bedfiles/hiv_status_vl_suppressed_degs_cis400k.bed.gz ;;\n",
    "        esac \n",
    "        \n",
    "        # loop through each chromosome\n",
    "        for j in {1..22}; do\n",
    "        \n",
    "            # create annotation files\n",
    "            python /opt/ldsc/make_annot.py \\\n",
    "                --bed-file $deg_file \\\n",
    "                --bimfile \"/data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim\" \\\n",
    "                --annot-file \"/data/annotations_ldscores/$trait/${trait}_degs_${window}.$j.annot.gz\"\n",
    "\n",
    "            # compute LD scores\n",
    "            python /opt/ldsc/ldsc.py \\\n",
    "                --l2 \\\n",
    "                --bfile \"/data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j\" \\\n",
    "                --ld-wind-cm 1 \\\n",
    "                --annot \"/data/annotations_ldscores/$trait/${trait}_degs_${window}.$j.annot.gz\" \\\n",
    "                --thin-annot \\\n",
    "                --out \"/data/annotations_ldscores/$trait/${trait}_degs_${window}.$j\" \\\n",
    "                --print-snps \"/data/1000g/1000G_EUR_Phase3_baseline/print_snps.txt\"\n",
    "        done # end chr loop\n",
    "        \n",
    "        # computed partitioned heritability estimate\n",
    "        python /opt/ldsc/ldsc.py \\\n",
    "            --h2 $stats \\\n",
    "            --w-ld-chr \"/data/1000g/weights_hm3_no_hla/weights.\" \\\n",
    "            --ref-ld-chr \"/data/annotations_ldscores/$trait/${trait}_degs_${window}.,/data/1000g/1000G_EUR_Phase3_baseline/baseline.\" \\\n",
    "            --overlap-annot \\\n",
    "            --out \"/data/results/${trait}_hiv_status_vl_suppressed_degs_${window}_results\" \\\n",
    "            --print-coefficients \\\n",
    "            --frqfile-chr \"/data/1000g/1000G_Phase3_frq/1000G.EUR.QC.\"\n",
    "    \n",
    "    done # end BED file loop\n",
    "done # end trait file loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine results files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for window in {\"10k\",\"100k\",\"400k\"}; do\n",
    "    outfile=hiv_acquisition_vl_suppressed_degs_cis${window}_combined_traits_results.tsv\n",
    "    head -1 hiv_acquisition_with_hiv_status_vl_suppressed_degs_cis${window}_results.results > $outfile\n",
    "        \n",
    "    for file in *degs_cis${window}_results.results; do\n",
    "        trait=$(echo $file |  sed 's/_with_hiv_status_vl_suppressed_degs_cis.*//')\n",
    "        awk -v trait=$trait \\\n",
    "        '$1 = trait {print $0}' OFS=\"\\t\" <(tail -n +2 $file | head -1) >> $outfile\n",
    "    done\n",
    "done\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
