{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d63da397-b4fa-48a4-8719-7c01009d3e26",
   "metadata": {},
   "source": [
    "# LiftOver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baff4c73-22c2-4b99-ba9e-cff9c7e9afb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# zip dependencies\n",
    "cd ../../\n",
    "zip \\\n",
    "    --exclude=*/var/* \\\n",
    "    --exclude=*.git/* \\\n",
    "    --exclude=*/test/* \\\n",
    "    --exclude=*/.idea/* \\\n",
    "    -r biocloud_gwas_workflows/helper_workflows/biocloud_gwas_workflows.zip \\\n",
    "    biocloud_gwas_workflows/\n",
    "\n",
    "cd -\n",
    "\n",
    "/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/helper_workflows/convert_variant_ids.wdl\n",
    "/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/helper_workflows/convert_variant_ids_wf.wdl\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/helper_workflows/convert_variant_ids_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/helper_workflows/inputs.json\" \\\n",
    "    -F \"workflowDependencies=@/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/helper_workflows/biocloud_gwas_workflows.zip\" \\\n",
    "    -F \"workflowOptions=@/home/ec2-user/bin/cromwell/hiv_gnetii_charge_code.json\" \\\n",
    "    >> job_id.txt\n",
    "\n",
    "job=55a5cf74-74df-4619-9adb-a75fd68c6b7e\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdbe48f9-c49f-4372-91c7-91907ca4dfc5",
   "metadata": {},
   "source": [
    "## Convert variant IDs input.json\n",
    "```\n",
    "CHR SNP BP A1 A2 FRQ_A_6334 FRQ_U_7247 INFO OR SE P ngt Direction HetISqt HetChiSq HetDf HetPVa BETA Z SE_BETA MAF\n",
    "1 rs12564456 2936386 C G 0.824 0.822 0.958 1.003 0.0344 0.939 0 -++-+- -4.9 6.674 5 0.246 0.00299550897979837 0.0765267914162924 0.039143271583194 0.177067226271998\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f896c1e-ed74-42ca-8180-e9bb4ec19c67",
   "metadata": {},
   "outputs": [],
   "source": [
    "{\n",
    "    \"convert_variant_ids_wf.input_files\": [\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr1_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr2_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr3_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr4_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr5_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr6_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr7_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr8_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr9_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr10_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr11_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr12_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr13_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr14_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr15_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr16_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr17_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr18_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr19_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr20_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr21_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\",\n",
    "        \"s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/dan_chr22_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz\"\n",
    "    ],\n",
    "    \"convert_variant_ids_wf.ref_files\": [\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr1.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr2.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr3.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr4.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr5.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr6.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr7.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr8.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr9.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr10.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr11.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr12.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr13.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr14.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr15.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr16.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr17.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr18.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr19.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr20.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr21.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh37.p13/id_conversion/variant_lookup_chr22.tsv.gz\"\n",
    "        \n",
    "    ],\n",
    "    \"convert_variant_ids_wf.chrs\": [\n",
    "        \"1\",\n",
    "        \"2\",\n",
    "        \"3\",\n",
    "        \"4\",\n",
    "        \"5\",\n",
    "        \"6\",\n",
    "        \"7\",\n",
    "        \"8\",\n",
    "        \"9\",\n",
    "        \"10\",\n",
    "        \"11\",\n",
    "        \"12\",\n",
    "        \"13\",\n",
    "        \"14\",\n",
    "        \"15\",\n",
    "        \"16\",\n",
    "        \"17\",\n",
    "        \"18\",\n",
    "        \"19\",\n",
    "        \"20\",\n",
    "        \"21\",\n",
    "        \"22\"\n",
    "    ],\n",
    "    \"convert_variant_ids_wf.output_files\": [\n",
    "        \"dan_chr1_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr2_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr3_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr4_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr5_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr6_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr7_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr8_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr9_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr10_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr11_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr12_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr13_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr14_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr15_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr16_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr17_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr18_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr19_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr20_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr21_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "        \"dan_chr22_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\"\n",
    "    ],\n",
    "\n",
    "    \"convert_variant_ids_wf.in_header\": 1,\n",
    "    \"convert_variant_ids_wf.in_sep\": \"space\",\n",
    "    \"convert_variant_ids_wf.in_id_col\": 1,\n",
    "    \"convert_variant_ids_wf.in_chr_col\": 0,\n",
    "    \"convert_variant_ids_wf.in_pos_col\": 2,\n",
    "    \"convert_variant_ids_wf.in_a1_col\": 3,\n",
    "    \"convert_variant_ids_wf.in_a2_col\": 4,\n",
    "    \"convert_variant_ids_wf.in_missing_allele\": \"0\",\n",
    "    \"convert_variant_ids_wf.in_deletion_allele\": \"-\",\n",
    "    \"convert_variant_ids_wf.ref_deletion_allele\": \".\",\n",
    "    \"convert_variant_ids_wf.output_compression\": \"gzip\",\n",
    "    \"convert_variant_ids_wf.cpu\": 1,\n",
    "    \"convert_variant_ids_wf.mem_gb\": 1,\n",
    "    \"convert_variant_ids_wf.in_chunk_size\": 50000,\n",
    "    \"convert_variant_ids_wf.ref_chunk_size\": 500000\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fab8cc9-741e-4a31-bb08-6b0632178f6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for shard in {0..21}; do\n",
    "    aws s3 cp s3://rti-cromwell-output/cromwell-execution/convert_variant_ids_wf/55a5cf74-74df-4619-9adb-a75fd68c6b7e/call-convert_variant_ids/shard-${shard}/ \\\n",
    "    s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/ \\\n",
    "        --recursive  \\\n",
    "        --exclude \"*\"  \\\n",
    "        --include \"*assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\"\n",
    "done\n",
    "\n",
    "aws s3 sync  s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/split_by_chromosome/ .\n",
    "\n",
    "# combine\n",
    "zcat dan_chr1_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz |\\\n",
    "    head -1 >  dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    zcat dan_chr${chr}_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz |\\\n",
    "    tail -n +2 >> dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt\n",
    "done\n",
    "\n",
    "gzip dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt\n",
    "\n",
    "# upload to S3 (note the name change)\n",
    "aws s3 cp dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz \\\n",
    "    s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg19/hg19_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\n",
    "\n",
    "# apply MAF, RSQ, and variants-in-all-cohorts filters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9993c84-11a7-4137-8bee-3ff25ae8be28",
   "metadata": {},
   "source": [
    "## hg38 1000g_p3 ID conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "238a19b4-7b4e-4d9a-984e-0f14a35bd099",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg38/hg38_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz .\n",
    "\n",
    "# ORS= output record separator \n",
    "awk 'NR==1{h = $0} NR>1{ print (!a[$1]++ ? h ORS $0: $0) > \"hg38_dan_chr\"$1\"_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt\"}' <(zcat hg38_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz)\n",
    "gzip hg38_dan_chr{1..22}_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt\n",
    "\n",
    "for file in hg38_dan_chr{1..22}_assoc_dosage_meta_ngt_metadaner_beta_sebeta.txt.gz; do\n",
    "    aws s3 cp $file s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg38/split_by_chromosome/ \n",
    "done"
   ]
  },
  {
   "cell_type": "raw",
   "id": "dbacdfc1-a847-467f-9f25-53129cf958ad",
   "metadata": {},
   "source": [
    "Submit the code below once the GRCh38 variants have been restored.\n",
    "Note that I changed \"D\" for deletion in the input file. I haven't done this before, so keep an eye out for any mysterious errors if it happens.\n",
    "    -Feb 11, 2022"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97e9f576-26bd-4c08-b469-513b747f6bd5",
   "metadata": {},
   "source": [
    "### inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c4d74cb-27e2-4412-8695-8a6a8128f565",
   "metadata": {},
   "outputs": [],
   "source": [
    "{\n",
    "    \"convert_variant_ids_wf.input_files\": [\n",
    "    ],\n",
    "    \"convert_variant_ids_wf.ref_files\": [\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr1.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr2.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr3.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr4.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr5.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr6.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr7.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr8.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr9.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr10.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr11.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr12.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr13.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr14.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr15.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr16.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr17.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr18.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr19.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr20.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr21.tsv.gz\",\n",
    "        \"s3://rti-common/variants/b153/GRCh38/id_conversion/variant_lookup_chr22.tsv.gz\"\n",
    "\n",
    "    ],\n",
    "    \"convert_variant_ids_wf.chrs\": [\n",
    "        \"1\",\n",
    "        \"2\",\n",
    "        \"3\",\n",
    "        \"4\",\n",
    "        \"5\",\n",
    "        \"6\",\n",
    "        \"7\",\n",
    "        \"8\",\n",
    "        \"9\",\n",
    "        \"10\",\n",
    "        \"11\",\n",
    "        \"12\",\n",
    "        \"13\",\n",
    "        \"14\",\n",
    "        \"15\",\n",
    "        \"16\",\n",
    "        \"17\",\n",
    "        \"18\",\n",
    "        \"19\",\n",
    "        \"20\",\n",
    "        \"21\",\n",
    "        \"22\"\n",
    "    ],\n",
    "    \"convert_variant_ids_wf.output_files\": [\n",
    "    \"hg38_dan_chr1_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr2_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr3_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr4_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr5_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr6_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr7_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr8_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr9_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr10_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr11_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr12_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr13_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr14_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr15_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr16_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr17_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr18_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr19_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr20_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr21_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\",\n",
    "    \"hg38_dan_chr22_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\"\n",
    "    ],\n",
    "    \"convert_variant_ids_wf.in_header\": 1,\n",
    "    \"convert_variant_ids_wf.in_sep\": \"space\",\n",
    "    \"convert_variant_ids_wf.in_id_col\": 1,\n",
    "    \"convert_variant_ids_wf.in_chr_col\": 0,\n",
    "    \"convert_variant_ids_wf.in_pos_col\": 2,\n",
    "    \"convert_variant_ids_wf.in_a1_col\": 3,\n",
    "    \"convert_variant_ids_wf.in_a2_col\": 4,\n",
    "    \"convert_variant_ids_wf.in_missing_allele\": \"0\",\n",
    "    \"convert_variant_ids_wf.in_deletion_allele\": \"D\",\n",
    "    \"convert_variant_ids_wf.ref_deletion_allele\": \".\",\n",
    "    \"convert_variant_ids_wf.output_compression\": \"gzip\",\n",
    "    \"convert_variant_ids_wf.cpu\": 1,\n",
    "    \"convert_variant_ids_wf.mem_gb\": 1,\n",
    "    \"convert_variant_ids_wf.in_chunk_size\": 50000,\n",
    "    \"convert_variant_ids_wf.ref_chunk_size\": 500000\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cadc4551-e6b3-460d-b79a-44f700a895de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#git clone --recursive https://github.com/RTIInternational/biocloud_gwas_workflows\n",
    "\n",
    "cd biocloud_gwas_workflows/biocloud_wdl_tools/convert_variant_ids\n",
    "\n",
    "# edit inputs.json\n",
    "\n",
    "# zip dependencies\n",
    "cd ../../../\n",
    "zip \\\n",
    "    --exclude=*/var/* \\\n",
    "    --exclude=*.git/* \\\n",
    "    --exclude=*/test/* \\\n",
    "    --exclude=*/.idea/* \\\n",
    "    -r biocloud_gwas_workflows/biocloud_wdl_tools/convert_variant_ids/biocloud_gwas_workflows.zip \\\n",
    "    biocloud_gwas_workflows/\n",
    "\n",
    "cd -\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/helper_workflows/convert_variant_ids_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/biocloud_wdl_tools/convert_variant_ids/inputs.json\" \\\n",
    "    -F \"workflowDependencies=@/home/ec2-user/rti-hiv/gwas/mclaren/biocloud_gwas_workflows/biocloud_wdl_tools/convert_variant_ids/biocloud_gwas_workflows.zip\" \\\n",
    "    -F \"workflowOptions=@/home/ec2-user/bin/cromwell/hiv_gnetii_charge_code.json\" \\\n",
    "    >> job_id.txt\n",
    "\n",
    "job=5edc42ce-5c61-4e39-854d-9d837251fe49  #28b4ee9c-d841-4fe6-b10a-f53e5d894408 #da05daad-a789-4766-9e48-3713393e7ab4  #7eebcddc-e7a7-438f-bb1c-7ac8763c4d9f 08757d4d-321f-4084-b1a7-f2d11e1eabd2  fe8f7f02-ba03-49b7-b27b-b9006062acb0\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"\n",
    "\n",
    "#Next combine the SNP-name-converted files\n",
    "for shard in {0..21}; do\n",
    "    aws s3 cp s3://rti-cromwell-output/cromwell-execution/convert_variant_ids_wf/$job/call-convert_variant_ids/shard-${shard}/ \\\n",
    "    s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg38/split_by_chromosome/ \\\n",
    "        --recursive  \\\n",
    "        --exclude \"*\"  \\\n",
    "        --include \"*assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz\"\n",
    "done\n",
    "\n",
    "aws s3 sync  s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg38/split_by_chromosome/ .\n",
    "\n",
    "# combine chromosomes\n",
    "zcat hg38_dan_chr1_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz  \\\n",
    "    >  hg38_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat hg38_dan_chr${chr}_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz |\\\n",
    "    tail -n +2 >> hg38_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt\n",
    "done\n",
    "\n",
    "gzip hg38_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt\n",
    "\n",
    "# upload to S3\n",
    "aws s3 cp hg38_dan_chr_all_assoc_dosage_meta_ngt_metadaner_beta_sebeta_1000g_p3.txt.gz \\\n",
    "    s3://rti-hiv/gwas/mclaren/processed/hiv_acquisition/final_stats/hg38/\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf31f18a-23e6-454b-a497-4c8264670bac",
   "metadata": {},
   "source": [
    "## EUR specific meta 0036"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93ac2a88-1d43-402f-9088-7004a4997891",
   "metadata": {},
   "outputs": [],
   "source": [
    "git clone --recursive https://github.com/RTIInternational/biocloud_gwas_workflows/\n",
    "cd 0036/biocloud_gwas_workflows/meta_analysis/metal/gwas/\n",
    "\n",
    "# edit inputs.json\n",
    "\n",
    "# zip dependencies\n",
    "cd ../../../../\n",
    "zip \\\n",
    "    --exclude=*/var/* \\\n",
    "    --exclude=*.git/* \\\n",
    "    --exclude=*/test/* \\\n",
    "    --exclude=*/.idea/* \\\n",
    "    -r biocloud_gwas_workflows/meta_analysis/metal/gwas/biocloud_gwas_workflows.zip \\\n",
    "    biocloud_gwas_workflows/\n",
    "\n",
    "cd -\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0036/biocloud_gwas_workflows/meta_analysis/metal/gwas/main.wdl\" \\\n",
    "    -F \"workflowInputs=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0036/biocloud_gwas_workflows/meta_analysis/metal/gwas/inputs.json\" \\\n",
    "    -F \"workflowDependencies=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0036/biocloud_gwas_workflows/meta_analysis/metal/gwas/biocloud_gwas_workflows.zip\" \\\n",
    "    -F \"workflowOptions=@/home/ec2-user/bin/cromwell/hiv_gnetii_charge_code.json\" \\\n",
    "    >> job_id.txt\n",
    "\n",
    "job=0965a6f3-8731-44f4-a609-53710af98734  #16194238-2343-4265-b58f-93c2f325172e  60d5add2-878f-474e-b3b8-eaa3f84b46fa   #56864af4-f4c1-4417-89a9-9f59f5d24738 #7416ce5d-843c-4e22-ac7b-4efa9b2b5dc9\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"\n",
    "\n",
    "\n",
    "# copy results\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/metal_gwas_meta_analysis_wf/$job/call-postprocessing/POSTPROCESS.postprocessing/ \\\n",
    "    s3://rti-hiv/gwas_meta/hiv_acquisition/results/0036/cromwell_results/ \\\n",
    "        --recursive \\\n",
    "        --exclude \"*\" \\\n",
    "        --include \"*output_file.txt\" \\\n",
    "        --include \"*final_results_p_lte_0.001.tsv\" \\\n",
    "        --include \"*png\" \\\n",
    "        --include \"*combined_sorted_results.txt.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82bccef3-a953-4639-8c09-5a9361e56813",
   "metadata": {},
   "source": [
    "# AFR-specific meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35fa32a9-19a7-4d05-b98c-faf5b0578dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "git clone --recursive https://github.com/RTIInternational/biocloud_gwas_workflows/\n",
    "cd biocloud_gwas_workflows/meta_analysis/metal/gwas/\n",
    "\n",
    "# edit input file\n",
    "\n",
    "cd ../../../../\n",
    "zip \\\n",
    "    --exclude=*/var/* \\\n",
    "    --exclude=*.git/* \\\n",
    "    --exclude=*/test/* \\\n",
    "    --exclude=*/.idea/* \\\n",
    "    -r biocloud_gwas_workflows/meta_analysis/metal/gwas/biocloud_gwas_workflows.zip \\\n",
    "    biocloud_gwas_workflows/\n",
    "\n",
    "cd -\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0037/biocloud_gwas_workflows/meta_analysis/metal/gwas/main.wdl\" \\\n",
    "    -F \"workflowInputs=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0037/biocloud_gwas_workflows/meta_analysis/metal/gwas/inputs.json\" \\\n",
    "    -F \"workflowDependencies=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0037/biocloud_gwas_workflows/meta_analysis/metal/gwas/biocloud_gwas_workflows.zip\" \\\n",
    "    -F \"workflowOptions=@/home/ec2-user/bin/cromwell/hiv_gnetii_charge_code.json\" \\\n",
    "    >> job_id.txt\n",
    "\n",
    "# paste job id here and check status\n",
    "job=1cdfaac0-2341-4f43-82d2-f8b8a188056d   #f84aeba6-150c-406f-9af0-f7acf6b770c8\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"\n",
    "\n",
    "\n",
    "\n",
    "# copy results\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/metal_gwas_meta_analysis_wf/$job/call-postprocessing/POSTPROCESS.postprocessing/ \\\n",
    "    s3://rti-hiv/gwas_meta/hiv_acquisition/results/0037/cromwell_results/ \\\n",
    "        --recursive \\\n",
    "        --exclude \"*\" \\\n",
    "        --include \"*output_file.txt\" \\\n",
    "        --include \"*final_results_p_lte_0.001.tsv\" \\\n",
    "        --include \"*png\" \\\n",
    "        --include \"*combined_sorted_results.txt.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51ce0333-2714-4afb-bfb9-954a75bc8c3d",
   "metadata": {},
   "source": [
    "# Cross-ancestry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6e49ff6-a9e3-41c9-a986-6115e0f0b7ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "git clone --recursive https://github.com/RTIInternational/biocloud_gwas_workflows/\n",
    "cd biocloud_gwas_workflows/meta_analysis/metal/gwas/\n",
    "\n",
    "# edit input file\n",
    "\n",
    "cd ../../../../\n",
    "zip \\\n",
    "    --exclude=*/var/* \\\n",
    "    --exclude=*.git/* \\\n",
    "    --exclude=*/test/* \\\n",
    "    --exclude=*/.idea/* \\\n",
    "    -r biocloud_gwas_workflows/meta_analysis/metal/gwas/biocloud_gwas_workflows.zip \\\n",
    "    biocloud_gwas_workflows/\n",
    "\n",
    "cd -\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0038/biocloud_gwas_workflows/meta_analysis/metal/gwas/main.wdl\" \\\n",
    "    -F \"workflowInputs=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0038/biocloud_gwas_workflows/meta_analysis/metal/gwas/inputs.json\" \\\n",
    "    -F \"workflowDependencies=@/home/ec2-user/rti-hiv/gwas_meta/hiv_acquisition/0038/biocloud_gwas_workflows/meta_analysis/metal/gwas/biocloud_gwas_workflows.zip\" \\\n",
    "    -F \"workflowOptions=@/home/ec2-user/bin/cromwell/hiv_gnetii_charge_code.json\" \\\n",
    "    >> job_id.txt\n",
    "\n",
    "# paste job id here and check status\n",
    "job=5065cbce-9d15-4899-8ad4-dbd6feb32ea4  #f1a0440b-5454-4383-97c2-e6e477a8524b  5f7d4da3-3aca-432e-a477-50353070921d\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"\n",
    "\n",
    "\n",
    "# copy results\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/metal_gwas_meta_analysis_wf/$job/call-postprocessing/POSTPROCESS.postprocessing/ \\\n",
    "    s3://rti-hiv/gwas_meta/hiv_acquisition/results/0038/cromwell_results/ \\\n",
    "        --recursive \\\n",
    "        --exclude \"*\" \\\n",
    "        --include \"*output_file.txt\" \\\n",
    "        --include \"*final_results_p_lte_0.001.tsv\" \\\n",
    "        --include \"*png\" \\\n",
    "        --include \"*call-merge/*\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e98ec8-9c6a-4e12-ad2b-c0b46c9b3a8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ecae143-c667-4273-b592-1321bf3e9f2e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2f57efd-1361-4563-9edd-e4bbaa5f21e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc6a5bcd-51d3-467a-9351-1ce3025a1ed2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
