{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hi-C Opioids Genotype Data Processing\n",
    "**Author:** Jesse Marks <br>\n",
    "**GitHub Issue**: [#159](https://github.com/RTIInternational/bioinformatics/issues/159)<br>\n",
    "**Initial sample size:** 155\n",
    "\n",
    "The samples were genotyped on [Illumina Infinium Multi-Ethnic Global-8 v1.0](https://support.illumina.com/downloads/infinium-multi-ethnic-global-8-v1-product-files.html) genotyping arrays. The genotype data are provided in [Illumina TOP/BOT format](https://www.illumina.com/documents/products/technotes/technote_topbot.pdf) with A/B allele nomenclature.  \n",
    "\n",
    "The genotype data are stored on S3 at `s3://rti-hic/shared_data/raw/cohort_0001/genotype/array/CaseWestern_Genotyping_Report_20201016.txt.gz`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[How to interpret DNA strand and allele information for Infinium genotyping array data](https://support.illumina.com/bulletins/2017/06/how-to-interpret-dna-strand-and-allele-information-for-infinium-.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-QC reformatting\n",
    "\n",
    "Note that the gentoype data are provided in Illumina TOP/BOT format with A/B nomenclature. We will need to extract the genotypes, first. Then we will identify whether the variant is TOP or BOT, then convert the A/B to A,C,G, or T. We will then convert the genotype data Plink format and orient the genotypes to GRCh37 forward strand. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## make directory structure\n",
    "mkdir -p /shared/rti-hic/shared_data/raw/cohort_0001/genotype/array/info/\n",
    "mkdir -p /shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/\n",
    "\n",
    "## download manifest files for genotype array\n",
    "cd /shared/rti-hic/shared_data/raw/cohort_0001/genotype/array/info\n",
    "wget ftp://webdata2:webdata2@ussd-ftp.illumina.com/downloads/productfiles/multiethnic-global-8/v1-0/infinium-multi-ethnic-global-8-d1-csv.zip\n",
    "\n",
    "unzip infinium-multi-ethnic-global-8-d1-csv.zip\n",
    "\n",
    "## download genotype data\n",
    "cd /shared/rti-hic/shared_data/raw/cohort_0001/genotype/array/\n",
    "aws s3 cp s3://rti-hic/shared_data/raw/cohort_0001/genotype/array/CaseWestern_Genotyping_Report_20201016.txt.gz .\n",
    "\n",
    "## upload genotype QC report and sample sheet from local\n",
    "#CaseWestern_Multi-Ethnic_QC_Report_20201016.xlsx\n",
    "#CaseWestern_Sample-Sheet_20201016.csv.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract genotypes (A/B) from raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Python3\n",
    "\"\"\"\n",
    "Process the raw genotype data for cohort 0001.\n",
    "Pull out the pertinent data:\n",
    "Name    Chr     Position        *.GType\n",
    "\"\"\"\n",
    "\n",
    "import gzip\n",
    "\n",
    "infile = \"/shared/rti-hic/shared_data/raw/cohort_0001/genotype/array/CaseWestern_Genotyping_Report_20201016.txt.gz\"\n",
    "outfile = \"/shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/cohort_0001_casewestern_genotypes_a_b.txt\"\n",
    "\n",
    "with gzip.open(infile, 'rt') as inF, open(outfile, 'w') as outF:\n",
    "    head = inF.readline()\n",
    "    split_head = head.split(\"\\t\")\n",
    "\n",
    "    # indices to keep\n",
    "    snp = 1\n",
    "    chrom = 3\n",
    "    pos = 4\n",
    "\n",
    "    # keep only headers of interest\n",
    "    keep_head = []\n",
    "    a_list = [ split_head[snp], split_head[chrom], split_head[pos] ]\n",
    "    # genotypes are listed every fourth column starting at the 10th index\n",
    "    b_list = [ split_head[i+6] for i in range(1, len(split_head)-6) if (i % 4) == 0 ]\n",
    "    keep_head.extend(a_list)\n",
    "    keep_head.extend(b_list)\n",
    "    keep_head = \"\\t\".join(keep_head) + \"\\n\"\n",
    "    outF.write(keep_head)\n",
    "\n",
    "    line = inF.readline()\n",
    "    # keep only columns of interest\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        keep_list = []\n",
    "\n",
    "        a_list = [ sl[snp], sl[chrom], sl[pos] ]\n",
    "        # genotypes are listed every fourth column starting at the 10th index\n",
    "        b_list = [ sl[i+6] for i in range(1, len(sl)-6) if (i % 4) == 0 ]\n",
    "        keep_list.extend(a_list)\n",
    "        keep_list.extend(b_list)\n",
    "\n",
    "        keep_line = \"\\t\".join(keep_list) + \"\\n\"\n",
    "        outF.write(keep_line)\n",
    "        line = inF.readline()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note*: Lost 20,382 variants from our genotyped data upon converting from A/B nomenclature of Illumina TOP/BOT format using the [manifest file](https://support.illumina.com/downloads/infinium-multi-ethnic-global-8-v1-product-files.html). These SNPs were either not in the manifest file, or were indels whose coding was ambiguous. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert to ACTG & GRCh37 + strand\n",
    "\n",
    "Using the [Manifest file (build 37)](https://support.illumina.com/downloads/infinium-multi-ethnic-global-8-v1-product-files.html) we will determine which Illumina strand the alleles are oriented to (TOP/BOT) and then convert the A/B nomenclature to ACGT nomenclature. Also orient the alleles to the GRCh37 + strand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Python3\n",
    "\"\"\"\n",
    "Convert the Illuma TOP/BOT formatted A/B nomenclature to\n",
    "ACGT coding using the manifest for that specific \n",
    "Illumina Infinium chip.\n",
    "\"\"\"\n",
    "\n",
    "import re\n",
    "\n",
    "manifest = '/shared/rti-hic/shared_data/raw/cohort_0001/genotype/array/info/Multi-EthnicGlobal_D1.csv'\n",
    "infile = '/shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/cohort_0001_casewestern_genotypes_a_b.txt'\n",
    "outfile = '/shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/cohort_0001_casewestern_genotypes_acgt_plus_strand.txt'\n",
    "\n",
    "\n",
    "# flip alleles if GRCh37 strand is \"-\"\n",
    "def flip_alleles(allele):\n",
    "    flipper = {\n",
    "            \"A\": \"T\",\n",
    "            \"T\": \"A\",\n",
    "            \"C\": \"G\",\n",
    "            \"G\": \"C\",\n",
    "            }\n",
    "\n",
    "    new_alleles = [flipper[i] for i in allele] \n",
    "    new_alleles = \"\".join(new_alleles)\n",
    "    return new_alleles\n",
    "\n",
    "\n",
    "# determine which is allele A and B using Illumina TOP/BOT format\n",
    "# input line and return both snp_name and a_b\n",
    "#def orientation(ill_strand, grch_strand, alleles, markername):\n",
    "def orientation(line):\n",
    "    line = line.split(\",\")\n",
    "    ill_strand = line[2]\n",
    "    grch_strand = line[20]\n",
    "    alleles = line[3]\n",
    "    markername = line[1]\n",
    "\n",
    "    a_b = re.sub(r'[\\[\\]]', '', alleles) # remove square brackets\n",
    "    a_b = a_b.split('/')\n",
    "    orient_dict = {}\n",
    "\n",
    "    if ill_strand == \"TOP\" or ill_strand == \"BOT\":\n",
    "        if grch_strand == \"-\":\n",
    "            orient_dict[\"A\"] = flip_alleles(a_b[0])\n",
    "            orient_dict[\"B\"] = flip_alleles(a_b[1])\n",
    "            return orient_dict\n",
    "        else:\n",
    "            orient_dict[\"A\"] = a_b[0]\n",
    "            orient_dict[\"B\"] = a_b[1]\n",
    "            return orient_dict\n",
    "\n",
    "    else: # else if the ill_strand is PLUS/MINUS (an indel)\n",
    "        m_names = markername.split(\"-\")\n",
    "        m_names = m_names[1:]\n",
    "        insertion = max(m_names) # largest string will be an insertion\n",
    "        deletion = min(m_names)  # smallest string will be a deletion\n",
    "\n",
    "        if a_b[0] == \"I\":\n",
    "            a_b[0] = insertion\n",
    "            a_b[1] = deletion\n",
    "        else:\n",
    "            a_b[1] = insertion\n",
    "            a_b[0] = deletion\n",
    "\n",
    "        if grch_strand == \"-\":\n",
    "            orient_dict[\"A\"] = flip_alleles(a_b[0])\n",
    "            orient_dict[\"B\"] = flip_alleles(a_b[1])\n",
    "            return orient_dict\n",
    "        else:\n",
    "            orient_dict[\"A\"] = a_b[0]\n",
    "            orient_dict[\"B\"] = a_b[1]\n",
    "            return orient_dict\n",
    "\n",
    "\n",
    "def convert_genotype(line, tmp_dict):\n",
    "    line = line.split()\n",
    "    snp = line[0]\n",
    "    chrom = line[1]\n",
    "    pos = line[2]\n",
    "    genotypes = line[3:]\n",
    "\n",
    "    for i in range(len(genotypes)):\n",
    "        if genotypes[i] == \"NC\":\n",
    "            genotypes[i] = '0:0' # (Per PLINK: '0' = no call)\n",
    "        else:\n",
    "            chrom_pos = \"{}:{}\".format(chrom, pos)\n",
    "            if chrom_pos == \"0:0\":\n",
    "                chrom_pos = snp\n",
    "            genoA = tmp_dict[chrom_pos][genotypes[i][0]]\n",
    "            genoB = tmp_dict[chrom_pos][genotypes[i][1]]\n",
    "\n",
    "            sample = \"{}:{}\".format(genoA, genoB)\n",
    "            genotypes[i] = sample\n",
    "    new_list = line[0:3] + genotypes\n",
    "    new_line = \"\\t\".join(new_list) + \"\\n\"\n",
    "    return new_line\n",
    "\n",
    "####################################################################################################\n",
    "####################################################################################################\n",
    "\n",
    "with open(manifest) as manF, open(infile) as inF, open(outfile, 'w') as outF:\n",
    "\n",
    "    # skip comment lines in the header\n",
    "    line = manF.readline()\n",
    "    while (line[0:6] != 'IlmnID'):\n",
    "        line = manF.readline()\n",
    "    print(line)\n",
    "        \n",
    "    line = line.strip().split(',')\n",
    "    markername = 1\n",
    "    chrom = 9\n",
    "    pos = 10\n",
    "\n",
    "    # dictionary will contain allele A and B for the GRCh + strand\n",
    "    man_dict = {}\n",
    "\n",
    "    # run through whole manifest file\n",
    "    line = manF.readline()\n",
    "    count = 0\n",
    "    while line:\n",
    "        sl = line.strip().split(',')\n",
    "        try:\n",
    "            a_b = orientation(line) # GRCh37 + strand alleles A & B\n",
    "            name = \"{}:{}\".format(sl[chrom], sl[pos])\n",
    "            \n",
    "            if name == \"0:0\": # some of the variants had 0:0 for chrom:pos\n",
    "                name = sl[markername] # just use the given variant name\n",
    "\n",
    "            man_dict[name] = a_b\n",
    "            line = manF.readline()\n",
    "        except:\n",
    "            line = manF.readline()\n",
    "    \n",
    "\n",
    "    # now run through our actual gentypes\n",
    "    head = inF.readline()\n",
    "    outF.write(head)\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        try:\n",
    "            out_line = convert_genotype(line, man_dict) # convert the A's and B's to the appropriate ACGT format\n",
    "            outF.write(out_line)\n",
    "            line = inF.readline()\n",
    "        except:\n",
    "            line = inF.readline()  # some SNPs were not found in the manifest (e.g.  1:155204987)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## tped & tfam\n",
    "The data must be convert to tped and tfam before we can convert to Plink's Bed, Bim, Fam format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### python3\n",
    "\"\"\"\n",
    "Convert file to a tped and a tfam file.\n",
    "\"\"\"\n",
    "import re\n",
    "\n",
    "infile = '/shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/cohort_0001_casewestern_genotypes_acgt_plus_strand.txt'\n",
    "outfile1 = '/shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/cohort_0001_casewestern_genotypes_acgt_plus_strand.tfam'\n",
    "outfile2 = '/shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/cohort_0001_casewestern_genotypes_acgt_plus_strand.tped'\n",
    "\n",
    "with open(infile) as inF, open(outfile1, 'w') as tFAM, open(outfile2, 'w') as tPED:\n",
    "    head = inF.readline()\n",
    "    head_ids = head.split()\n",
    "    head_ids = head_ids[3:]\n",
    "\n",
    "    # create tfam file\n",
    "    for item in head_ids:\n",
    "        item = re.sub(r'\\.GType', '', item)\n",
    "        tfam_line = '{}\\t{}\\t0\\t0\\t0\\t0\\n'.format(item, item)\n",
    "        tFAM.write(tfam_line)\n",
    "\n",
    "\n",
    "    # create tped file\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        variant_id = sl[0]\n",
    "        chrom = sl[1]\n",
    "        pos = sl[2]\n",
    "        genotypes = sl[3:]\n",
    "        genotypes = [ re.sub(r':', '\\t', item) for item in genotypes ]\n",
    "        genotypes = '\\t'.join(genotypes)\n",
    "        outline = '{}\\t{}\\t0\\t{}\\t{}\\n'.format(chrom, variant_id, pos, genotypes) \n",
    "        tPED.write(outline)\n",
    "        line = inF.readline()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert to BED/BIM/FAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001\n",
    "\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --tfile cohort_0001_casewestern_genotypes_acgt_plus_strand \\\n",
    "    --make-bed \\\n",
    "    --out casewestern_cohort_0001_genotypes\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add sex information\n",
    "Add the sex information to the FAM file. This can be found in the file \n",
    "`CaseWestern_Multi-Ethnic_QC_Report_20201016.xlsx`\n",
    "We will create a map file from this Excel file and then add the sex info to the FAM file.\n",
    "\n",
    "```\n",
    "head CaseWestern_Multi-Ethnic_QC_Report_20201016_sexmap.txt\n",
    "\n",
    "Sample_ID       Gender\n",
    "8042579044_JN0000001_HCCCP      Male\n",
    "8042579056_JN0000013_HCTVE      Male\n",
    "8042579068_JN0000025_HCTQQ      Male\n",
    "8042579080_JN0000037_NIH1889    Female\n",
    "8042579092_JN0000049_HCTMB      Male\n",
    "8042579104_JN0000061_HCTGD      Male\n",
    "8042579116_JN0000073_HCTBA      Female\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Python3\n",
    "\"\"\"\n",
    "Use a map file to enter sex information into the FAM file\n",
    "\"\"\"\n",
    "\n",
    "famfile = 'casewestern_cohort_0001_genotypes.fam'\n",
    "infile = 'CaseWestern_Multi-Ethnic_QC_Report_20201016_sexmap.txt'\n",
    "outfile = 'casewestern_cohort_0001_genotypes_sex_mapped.fam'\n",
    "\n",
    "with open(famfile) as famF, open(infile) as inF, open(outfile, 'w') as outF:\n",
    "    head = inF.readline()\n",
    "\n",
    "    # contains map to sex\n",
    "    sex_dic = {}\n",
    "\n",
    "    # populate the sex dictionary by looping through map file\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        sex = sl[1]\n",
    "        if sex == 'Male':\n",
    "            sex = '1'\n",
    "        elif sex == 'Female':\n",
    "            sex = '2'\n",
    "        sex_dic[sl[0]] = sex\n",
    "        line = inF.readline()\n",
    "\n",
    "    # now loop through the fam file (also remove positive controls)\n",
    "    line = famF.readline()\n",
    "    count = 0\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        \n",
    "        # we will remove these with PLINK\n",
    "        # need to rename NA subjects because they throw an error in PLINK when trying to remove them \n",
    "        if sl[0] == 'Positive_Control_NA_NA' or sl[0] == 'NA':\n",
    "            count += 1\n",
    "            sl[0] += str(count)\n",
    "            sl[1] += str(count)\n",
    "            outline = ' '.join(sl) + '\\n'\n",
    "            outF.write(outline)\n",
    "            line = famF.readline()\n",
    "            continue\n",
    "            \n",
    "        sl[4] = sex_dic[sl[0]]\n",
    "        new_line = ' '.join(sl) + '\\n'\n",
    "        outF.write(new_line)\n",
    "        line = famF.readline()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove positive controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001\n",
    "\n",
    "# the bottom 5 subjects were either NA or positive controls\n",
    "cut -d' ' -f1 casewestern_cohort_0001_genotypes_sex_mapped.fam | head -155 > keep_list.txt\n",
    "    \n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bed casewestern_cohort_0001_genotypes.bed \\\n",
    "    --bim casewestern_cohort_0001_genotypes.bim \\\n",
    "    --fam casewestern_cohort_0001_genotypes_sex_mapped.fam \\\n",
    "    --keep-fam keep_list.txt \\\n",
    "    --make-bed \\\n",
    "    --out casewestern_cohort_0001\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001\n",
    "\n",
    "for ext in {bed,bim,fam}; do\n",
    "    gzip casewestern_cohort_0001.$ext \n",
    "    aws s3 cp  casewestern_cohort_0001.$ext.gz s3://rti-hic/shared_data/pre_qc/cohort_0001/genotype/array/observed/0001/\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QC\n",
    "\n",
    "View processing [here](https://github.com/RTIInternational/bioinformatics/blob/master/methods/rti-hic/shared_data/post_qc/cohort_0001/genotype/array/observed/0001/post_qc_cohort_0001_genotype_array_0001.md)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
