{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "00b29198-0b36-4cce-ae08-cc11f9c16bc4",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "# Link Phenotype to DNAm & Genotype Data\n",
    "\n",
    "email from Fang\n",
    "\n",
    "```css\n",
    "I have done the phenotype summarization for the ALSPAC data. Also, I realized that it is pretty complicated to deliver all the information in the task assignment on Teams, so along with that, I am writing down more details in this email for you to check.\n",
    " \n",
    "The DNA methylation data includes three waves for Mother and four waves for Child:\n",
    "```\n",
    "\n",
    "**Methylation**:\n",
    "Mother & Child\n",
    "\n",
    "| Mother               | Mother           | Mother         | Child   | Child     | Child       | Child       |\n",
    "|----------------------|------------------|----------------|---------|-----------|-------------|-------------|\n",
    "| Antenatal (~28,7yrs) | TF1-3 (~42.9yrs) | FOM (~47.8yrs) |   Cord  | 7yrs (F7) | 15yrs (TF3) | 17yrs (F17) |\n",
    "|          n/a         |        n/a       |       n/a      | n = 914 |  n = 980  |   n = 254   |   n = 727   |\n",
    "|        n = 987       |      n = 182     |     n = 810    |   n/a   |    n/a    |     n/a     |     n/a     |\n",
    "\n",
    "```css\n",
    "I have created phenotype files separately for each wave with cannabis use information (mothers_antenatal, mothers_TF1_3, mothers_FOM, kids_F15, kids_F17) at: s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/.\n",
    " \n",
    "To link the phenotype files, DNA methylation and genetic data, you will need the linking id file: s3://rti-cannabis/shared_data/raw_data/alspac/B3176_datasetids.csv\n",
    " \n",
    "The raw DNA methylation data is at s3://rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas/data.Robj\n",
    " \n",
    "The genetic data are at s3://rti-cannabis/shared_data/raw_data/alspac/gwa_550_g1/ and s3://rti-cannabis/shared_data/raw_data/alspac/gwa_660_g0m/.\n",
    " \n",
    "So I think the first thing is to extract DNA methylation data and genetic data matching the phenotype data, giving a demographic summary for each wave. \n",
    "```\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5899adae-272d-46cb-a0a9-7ee8548dbf23",
   "metadata": {},
   "source": [
    "## Download data\n",
    "Use the GOBOT sandbox (https://github.rti.org/RTI/Bio-research-machine) because it has a lot of storage and memory.\n",
    "\n",
    "Note that the GOBOT sandbox server has ARM achitecture and therefore PLINK will not work as of 10/26/2021. To resolve this issue, we will have to use a different server when we use PLINK."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34135f4f-00ab-479f-a7cd-377d2fa300c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## on a large memory node (r5.2xlarge with 64GB)\n",
    "\n",
    "sudo yum update -y\n",
    "aws configure # enter credentials and configurations\n",
    "\n",
    "# install R\n",
    "#sudo amazon-linux-extras install R4 -y\n",
    "\n",
    "# install Docker\n",
    "sudo amazon-linux-extras install docker\n",
    "sudo service docker start\n",
    "sudo usermod -a -G docker ec2-user\n",
    "# then logout and back in\n",
    "\n",
    "\n",
    "mkdir -p rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas/ \n",
    "mkdir -p rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas/ \n",
    "mkdir -p rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\n",
    "mkdir -p rti-cannabis/shared_scripts/DNAm_xreactive/\n",
    "\n",
    "aws s3 cp s3://rti-cannabis/shared_scripts/DNAm_xreactive/cross.csv \\ # cross-reactive probes\n",
    "    ~/rti-cannabis/shared_scripts/DNAm_xreactive/ \n",
    "\n",
    "mkdir -p rti-cannabis/shared_data/post_qc/alspac/phenotypes\n",
    "mkdir -p rti-cannabis/shared_data/raw_data/alspac/gwa_550_g1/\n",
    "mkdir -p rti-cannabis/shared_data/raw_data/alspac/gwa_660_g0m/\n",
    "\n",
    "# DNAm file\n",
    "cd rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas/\n",
    "aws s3 cp s3://rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas/data.Robj .\n",
    "\n",
    "# phenotype files\n",
    "cd ~/rti-cannabis/shared_data/post_qc/alspac/phenotypes/\n",
    "aws s3 sync s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/ .\n",
    "\n",
    "# genetic data\n",
    "cd ~/rti-cannabis/shared_data/raw_data/alspac/gwa_550_g1/\n",
    "aws s3 sync s3://rti-cannabis/shared_data/raw_data/alspac/gwa_550_g1/version1/ .\n",
    "\n",
    "cd ~/rti-cannabis/shared_data/raw_data/alspac/gwa_660_g0m/\n",
    "aws s3 sync s3://rti-cannabis/shared_data/raw_data/alspac/gwa_660_g0m/version1/ .\n",
    "\n",
    "# linking ID file\n",
    "cd ~/rti-cannabis/shared_data/raw_data/alspac\n",
    "aws s3 cp s3://rti-cannabis/shared_data/raw_data/alspac/B3176_datasetids.csv ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c71a699-e5be-4e9a-ace0-7c665192f607",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# DNAm Data\n",
    "Use the EWAS Docker image that Fang Fang created.\n",
    "\n",
    "*Note*: an updated ewas image is available at `rtibiocloud/ewas:v0.0.2_99db04b`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bac58461-f7ac-4fa9-af03-6c69aa87448c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas/\n",
    "\n",
    "## start interactive session\n",
    "docker run \\\n",
    "    --rm \\\n",
    "    -v $PWD:$PWD \\\n",
    "    -it ffang8/ewas:v041221 \\\n",
    "    /bin/bash\n",
    "\n",
    "cd /home/ec2-user/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/betas "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ee6c24b-e8ce-4071-bfc7-8c367631e312",
   "metadata": {},
   "source": [
    "## Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf034e7-dbee-4ba9-b0c1-ac501173f5fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Start R\n",
    "R\n",
    "\n",
    "##############################\n",
    "##LOAD LIBRARIES\n",
    "##############################\n",
    "\n",
    "library(minfi)\n",
    "library(ggplot2)\n",
    "    theme_set(theme_bw(base_size=18) + \n",
    "        theme(panel.grid.major = element_blank(),\n",
    "         panel.grid.minor = element_blank(),\n",
    "         plot.title = element_text(hjust = 0.5),\n",
    "         legend.position=\"none\"))\n",
    "library(ggrepel)\n",
    "library(jaffelab)\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(dplyr)\n",
    "library(RColorBrewer)\n",
    "library(FlowSorted.Blood.EPIC)\n",
    "library(data.table)# to process results\n",
    "library(MASS) # rlm function for robust linear regression\n",
    "library(sandwich) #Huberís estimation of the standard error\n",
    "library(lmtest) # to use coeftest\n",
    "library(parallel) # to use multicore approach - part of base R – can be omitted if lapply() used instead of mclapply()\n",
    "# library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19) #This is the newest version, LIBD processed data using older version#\n",
    "library(IlluminaHumanMethylation450kanno.ilmn12.hg19)\n",
    "library(ENmix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cca1db8d-13e6-41a4-affd-b20108a9911c",
   "metadata": {},
   "source": [
    "\n",
    "```\n",
    "We strongly recommend to drop the probes that contain either a SNP at the CpG interrogation or at the single nucleotide extension.The function dropLociWithSnps allows to drop the corresponding probes. \n",
    "```\n",
    "-- https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "This step is to filter out probes with SNPs, since SNPs will affect the probe hybridization and then the estimation of the methylation results will be inaccurate.\n",
    "```\n",
    "-- Fang Fang\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "Hybridization probe\n",
    "In molecular biology, a hybridization probe is a fragment of DNA or RNA of variable length which can be radioactively or fluorescently labeled. It can then be used in DNA or RNA samples to detect the presence of nucleotide substances that are complementary to the sequence in the probe.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc089a36-9cb1-438d-86b8-80d85e13d8e6",
   "metadata": {},
   "source": [
    "## Apply probe filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f63c97-9653-470c-92cb-b0593d9934cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### R \n",
    "load(\"data.Robj\", verbose=TRUE)\n",
    "# betas\n",
    "\n",
    "##############################\n",
    "##PROBE FILTERING\n",
    "##############################\n",
    "#In order to use minfi for additional annotation, we need to create a GenomicRatioSet\n",
    "#beta_grs<-makeGenomicRatioSetFromMatrix(beta,array = \"IlluminaHumanMethylationEPIC\", annotation = \"ilm10b4.hg19\") \n",
    "beta_grs <- makeGenomicRatioSetFromMatrix(betas,\n",
    "                                          array = \"IlluminaHumanMethylation450k\",\n",
    "                                          annotation = \"ilmn12.hg19\")\n",
    "dim(beta_grs) #482855 x 4854\n",
    "\n",
    "#DROP PROBES WITH SNP IN SBE OR CPG\n",
    "beta_grs <- addSnpInfo(beta_grs)\n",
    "beta_grs_filter <- dropLociWithSnps(beta_grs, snps=c(\"SBE\",\"CpG\"), maf=0.01) \n",
    "\n",
    "dim(beta_grs_filter) #465873 x 4854\n",
    "\n",
    "save(beta_grs_filter,file=\"~/rti-cannabis/shared_data/post_qc/alspac/dna_methylation/beta_grs_filter.Rdata\")\n",
    "\n",
    "## close R session to free up memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e4893bb-3cbc-492a-b9c4-3f774f888848",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\n",
    "\n",
    "## Start R session with new Rdata\n",
    "library(IlluminaHumanMethylation450kanno.ilmn12.hg19)\n",
    "\n",
    "load(\"beta_grs_filter.Rdata\", verbose=TRUE)\n",
    "\n",
    ">head(rownames(beta_grs_filter))\n",
    "#[1] \"cg13869341\" \"cg14008030\" \"cg12045430\" \"cg20826792\" \"cg00381604\"\n",
    "#[6] \"cg20253340\"\n",
    "\n",
    ">head(colnames(betas))\n",
    "#[1] \"SLIDE210_R04C02\" \"SLIDE380_R02C01\" \"SLIDE306_R02C02\" \"SLIDE16_R06C01\"\n",
    "#[5] \"SLIDE189_R04C02\" \"SLIDE132_R03C01\"\n",
    "\n",
    "#DROP PROBES WITH CROSSREACTIVE BEHAVIOR\n",
    "xReactiveProbes<-read.csv(file=\"~/rti-cannabis/shared_scripts/DNAm_xreactive/cross.csv\", stringsAsFactors=FALSE)\n",
    "keep <- !(featureNames(beta_grs_filter) %in% xReactiveProbes[,1])\n",
    "table(keep) # 26,769 to remove\n",
    "beta_grs_filter <- beta_grs_filter[keep,] \n",
    "\n",
    "dim(beta_grs_filter) # 439104   4854\n",
    "\n",
    "beta_grs_filter <- beta_grs_filter[keep,] \n",
    "\n",
    "#DROP PROBES THAT MAP TO THE CHRY, but keep CHRX\n",
    "annEPIC = getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)\n",
    "table(annEPIC$chr)\n",
    "#  chr1 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19  chr2 chr20\n",
    "# 46857 24388 28794 24539 12285 15078 15259 21969 27879  5922 25521 34810 10379\n",
    "# chr21 chr22  chr3  chr4  chr5  chr6  chr7  chr8  chr9  chrX  chrY\n",
    "# 4243  8552 25159 20464 24327 36611 30017 20950  9861 11232   416\n",
    "\n",
    "\n",
    "for(chr in names(table(annEPIC$chr))){\n",
    "    keep_chr <- featureNames(beta_grs_filter) %in% annEPIC$Name[annEPIC$chr==chr]\n",
    "    beta_chr <- beta_grs_filter[keep_chr,]\n",
    "    bVals_chr <- getBeta(beta_chr)\n",
    "    save(bVals_chr, file = paste0('alspac_dnameth_betas_',chr,'.rda'))\n",
    "}\n",
    "\n",
    "keep <- !(featureNames(beta_grs_filter) %in% annEPIC$Name[annEPIC$chr %in% c(\"chrY\")])\n",
    "table(keep) #303 to remove\n",
    "beta_grs_filter <- beta_grs_filter[keep,]\n",
    "dim(beta_grs_filter) # 438801   4854\n",
    "   \n",
    "setwd(\"~/rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\")\n",
    "save(beta_grs_filter, file = 'alspac_dnameth_withchrX_beta_grs_filter.rda')\n",
    "\n",
    "\n",
    "######################################\n",
    "##CONVERT BACK TO BETA MATRIX\n",
    "######################################\n",
    "\n",
    "bVals <- getBeta(beta_grs_filter)\n",
    "dim(bVals) # 438801   4854\n",
    "\n",
    "##########################################\n",
    "##SAVE METH data\n",
    "###########################################\n",
    "save(bVals, file = 'alspac_dnameth_withchrX_bVals.rda')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b311980c-eb74-4ea2-a354-4dfe22ea161d",
   "metadata": {},
   "source": [
    "## Upload DNAm to S3\n",
    "Upload to S3 after removing cross-reactive and SNP-associated probes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04cf7c2a-03e4-4971-a79f-608a6460da33",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\n",
    "\n",
    "aws s3 cp alspac_dnameth_withchrX_beta_grs_filter.rda s3://rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\n",
    "aws s3 cp alspac_dnameth_withchrX_bVals.rda s3://rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\n",
    "\n",
    "for chr in {1..22} X Y; do\n",
    "    aws s3 cp alspac_dnameth_betas_chr$chr.rda s3://rti-cannabis/shared_data/post_qc/alspac/dna_methylation/\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f134b56b-83d1-4ef1-8d0e-1aedd8944b78",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Phenotype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f79b11a-4e7e-41b7-ac82-deb2e45becb6",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## SVA and celltype proportions\n",
    "Surrogate variables from non-negative control probes.\n",
    "\n",
    "We do not have the rgset, but Fang calculated the SVs. Add these to the phenotype file.\n",
    "\n",
    "```css\n",
    "[10/29/21 12:35 PM] Fang, Fang\n",
    "Hey, Jesse, I finally made the sva files based on the summary control probes they provided in the raw data, using customized code mimicking the core code in ENmix. The file contains 6 SVs that should be included in our EWAS. \n",
    "\n",
    "s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/alspac_control_probe_sva.txt\n",
    "\n",
    "The codes are recorded in s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/ALSPAC surrogate variables based on control probes.ipynb \n",
    "\n",
    "I would like to try another way using the R package meffil later on.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b42456a-651c-4ba9-8098-36e92d3bacdf",
   "metadata": {},
   "source": [
    "### Issue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b7d275e-9546-478d-b5ad-ca0eb2d43b79",
   "metadata": {},
   "source": [
    "- ALSPAC does not provide raw data (IDAT files) or rgSet, only beta values and qc.objects using meffil.\n",
    "\n",
    "`Min, J. L., Hemani, G., Davey Smith, G., Relton, C., & Suderman, M. (2018). Meffil: efficient normalization and analysis of very large DNA methylation datasets. Bioinformatics, 34(23), 3983-3989.` \n",
    "\n",
    "- We cannot generate surrogate variables as we usually do with ENmix.\n",
    "\n",
    "`sv <- ctrlsva(rgSet,percvar=0.95,npc=6,flag=1)`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aed2e25-2f2b-43ac-97c2-ffe2634bde7f",
   "metadata": {},
   "source": [
    "### Solution 1 - mimic the method in ENmix (ctrlsva)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28dabf93-f6ec-4b2d-ac26-e81cd2e104ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the core codes for ctrlsva\n",
    "percvar=0.95\n",
    "npc=1\n",
    "flag=1\n",
    "pca <- prcomp(t(ctrl_nneg))\n",
    "eigenvalue = pca$sdev^2\n",
    "perc = eigenvalue/sum(eigenvalue)\n",
    "if (flag == 1) {\n",
    "    npc=1\n",
    "    while(sum(perc[1:npc] < percvar)){\n",
    "        npc <- npc + 1\n",
    "    }\n",
    "    npc\n",
    "    ctrlsva = pca$x[,1:npc]\n",
    "}\n",
    "else {\n",
    "    ctrlsva = pca$x[,1:npc]\n",
    "}\n",
    "cat(npc,\" surrogate variables explain \",sum(perc[1:npc])*100,\"% of \\n data variation\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3403e3d5-de93-4e56-aed1-c27ed0a8d028",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/control_matrix/\n",
    "aws s3 cp s3://rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/control_matrix/data.txt .\n",
    "\n",
    "### R\n",
    "# start from the data.txt with the control probe intensities, renamed as \"control.data.txt\"\n",
    "control <- read.table(\"control.data.txt\",header=T,stringsAsFactors=F)\n",
    "\n",
    "percvar=0.95\n",
    "npc=1\n",
    "flag=1\n",
    "pca <- prcomp(control[,2:43])\n",
    "eigenvalue = pca$sdev^2\n",
    "perc = eigenvalue/sum(eigenvalue)\n",
    "if (flag == 1) {\n",
    "    npc=1\n",
    "    while(sum(perc[1:npc]) < percvar){\n",
    "        npc <- npc + 1\n",
    "    }\n",
    "    npc\n",
    "    ctrlsva = pca$x[,1:npc]\n",
    "}else {\n",
    "    ctrlsva = pca$x[,1:npc]\n",
    "}\n",
    "\n",
    "cat(npc,\" surrogate variables explain \",sum(perc[1:npc])*100,\"% of \\n data variation\\n\")\n",
    "# 6  surrogate variables explain  95.68676 % of\n",
    "#  data variation\n",
    "\n",
    "\n",
    "\n",
    "sva <- data.frame(Sample_Name=control$Sample_Name,\n",
    "                  sv1=ctrlsva[,1],sv2=ctrlsva[,2],sv3=ctrlsva[,3],\n",
    "                  sv4=ctrlsva[,4],sv5=ctrlsva[,5],sv6=ctrlsva[,6])\n",
    "\n",
    "write.table(sva,file=\"alspac_control_probe_sva.txt\",quote=F,row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f87120e1-29e9-435f-8898-42ae23d6eaeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp alspac_control_probe_sva.txt s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09bdded9-3080-4dfd-acca-6a632d3f7003",
   "metadata": {},
   "source": [
    "### Cell type estimation\n",
    "\n",
    "ALSPAC precomputed the cell-type proportions for us at `s3://rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/derived/cellcounts/houseman/data.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daf5058b-4c00-43e3-91c9-455ff4dbb4d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ~/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/derived/cellcounts/houseman/\n",
    "cd ~/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/derived/cellcounts/houseman/\n",
    "aws s3 cp s3://rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/derived/cellcounts/houseman/data.txt ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a394af75-df63-4255-a27e-c8ee0369878d",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## Mothers phenotype files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b3d1529-793c-47a8-b03a-e0c6689ce17a",
   "metadata": {},
   "source": [
    "### combine TF1–3 and the FOM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afd3655e-80f2-4664-b034-01b1db0563b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes\n",
    "\n",
    "### R \n",
    "tf13 <- read.csv(\"pheno_mothers_TF1_3_ewas.csv\")\n",
    "head(tf13)\n",
    "#  cidB3176 qlet age_at_DNAm cannabisUse smoking drinking\n",
    "#1        7    A    51.63860           2   Never  Current\n",
    "#2       92    A    34.71321           2 Current     Ever\n",
    "#3       97    A    43.97810           2 Current  Current\n",
    "\n",
    "fom <- read.csv(\"pheno_mothers_FOM_ewas.csv\")\n",
    "dim(fom) # 795 x 7\n",
    "head(fom)\n",
    "#   cidB3176 qlet age_at_DNAm cannabisUse smoking drinking      BMI\n",
    "#1       21    A          46           2   Never  Current 21.03090\n",
    "#2       84    A          57           2    Ever  Current 37.19008\n",
    "#3      113    A          52           2    Ever  Current 24.00127\n",
    "\n",
    "# combine datasets\n",
    "tf13[,\"BMI\"] <- NA\n",
    "dim(tf13) # 182 x 7\n",
    "combined_data <- rbind(tf13, fom)\n",
    "dim(combined_data)# 977 x 7\n",
    "head(combined_data)\n",
    "#  cidB3176 qlet age_at_DNAm cannabisUse smoking drinking BMI\n",
    "#1        7    A    51.63860           2   Never  Current  NA\n",
    "\n",
    "write.table(combined_data,file=\"pheno_mothers_combined_FOM_TF1_3_n977_ewas.txt\",quote=F,row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4a2ccd-29d5-4a39-9586-af57d28c650a",
   "metadata": {},
   "source": [
    "### Remove outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e33fe4d9-bd3c-41df-944e-2133cce024a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### R\n",
    "# update the phenotype file sample IDs from cidB3176 to Sample_Name\n",
    "mothers_pheno <- read.csv(\"pheno_mothers_combined_FOM_TF1_3_n977_ewas.txt\", sep = \" \")\n",
    "linking_file <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/raw_data/alspac/B3176_datasetids.csv\")\n",
    "load(\"~/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/samplesheet/data.Robj\")\n",
    "\n",
    "dim(mothers_pheno) # 977 x 7\n",
    "dim(linking_file) # 21506 x 6\n",
    "dim(samplesheet) # 4854 x 16\n",
    "\n",
    "# keep only mothers in TF1-3 and FOM\n",
    "samplesheet_mothers <-  samplesheet[which(samplesheet$time_code == \"TF1-3\" | \n",
    "                                          samplesheet$time_code == \"FOM\"),]\n",
    "dim(samplesheet_mothers) # 992 x 16\n",
    "\n",
    "# Remove duplicate and population stratification samples \n",
    "sample.rm <- which(samplesheet_mothers$duplicate.rm==\"Remove\" | \n",
    "                   samplesheet_mothers$genotypeQCkids == \"ETHNICITY\" | \n",
    "                   samplesheet_mothers$genotypeQCkids==\"HZT;ETHNICITY\"|\n",
    "                   samplesheet_mothers$genotypeQCmums==\"/strat\" )\n",
    "\n",
    "samplesheet_mothers <- samplesheet_mothers[-sample.rm, ]                                     \n",
    "\n",
    "dim(samplesheet_mothers) # 947 x 16\n",
    "\n",
    "# remove duplicated cidB3176 IDs from phenotype file\n",
    "remove_dups <- which(duplicated(mothers_pheno$cidB3176))\n",
    "mothers_pheno <- mothers_pheno[-remove_dups, ]\n",
    "\n",
    "dim(mothers_pheno) # 975 x 7\n",
    "\n",
    "# add ALN IDs to the phenotype file from the linking file (ALN == dnam_450_g0m_g1)\n",
    "mothers_pheno$ALN <- linking_file$dnam_450_g0m_g1[match(mothers_pheno$cidB3176, linking_file$cidB3176)]\n",
    "\n",
    "# NOTE: there are no duplicated ALNs this time. If there were, we would have to address \n",
    "# them using the \"qlet\" to get the unique Sample_Name\n",
    "which(duplicated(mothers_pheno$ALN)) # integer(0)\n",
    "\n",
    "# remove any mothers from the phenotype file that were not in the curated samplesheet\n",
    "keep <- mothers_pheno$ALN %in% samplesheet_mothers$ALN\n",
    "mothers_pheno <- mothers_pheno[keep,] # 946 x 8\n",
    "\n",
    "# add Sample_Name to the phenotype file\n",
    "mothers_pheno$Sample_Name <- samplesheet_mothers$Sample_Name[match(mothers_pheno$ALN, samplesheet_mothers$ALN)]\n",
    "\n",
    "dim(mothers_pheno) # 946 x 9\n",
    "                            \n",
    "outname = \"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/pheno_mothers_combined_FOM_TF1_3_n946_with_sample_name_ids.txt\"\n",
    "write.table(mothers_pheno, outname, quote = F, row.names = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e64a1c1c-1e6e-4f90-98ad-40d2814e5b20",
   "metadata": {},
   "source": [
    "### Combine SVs and cell-type proportions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e52f6a50-e265-467e-a7f3-6cb1809c3979",
   "metadata": {},
   "outputs": [],
   "source": [
    "### R \n",
    "pheno_mothers <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/pheno_mothers_combined_FOM_TF1_3_n946_with_sample_name_ids.txt\",\n",
    "                          sep = \" \")\n",
    "cellcounts <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/derived/cellcounts/houseman/data.txt\",\n",
    "                      sep = \"\\t\")\n",
    "sv_data <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/alspac_control_probe_sva.txt\",\n",
    "                    sep = \" \")\n",
    "\n",
    "dim(cellcounts) # 4854 x 7\n",
    "dim(pheno_mothers) # 946 x 9\n",
    "dim(sv_data) # 4854 x 7\n",
    "\n",
    "# add cellcounts\n",
    "add_cellcounts <- which(cellcounts$Sample_Name %in% pheno_mothers$Sample_Name)\n",
    "cellcounts[add_cellcounts, 1]\n",
    "\n",
    "# attach cell type estimations and SVs to pheno\n",
    "pheno_mothers_cellcounts <- merge(pheno_mothers,cellcounts,by.x=\"Sample_Name\",by.y=\"Sample_Name\")\n",
    "pheno_mothers_cellcounts_svs <- merge(pheno_mothers_cellcounts,sv_data,by.x=\"Sample_Name\",by.y=\"Sample_Name\")\n",
    "\n",
    "# reorder\n",
    "pheno_mothers_cellcounts_svs <- pheno_mothers_cellcounts_svs[, c(1,2,9,3,4,5,6,7,8,10,11,12,13,14,15,16,17,18,19,20,21)]\n",
    "dim(pheno_mothers_cellcounts_svs) # 946 x 21\n",
    "\n",
    "outfile <- \"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/pheno_mothers_combined_FOM_TF1_3_n946_ewas_final.txt\"\n",
    "write.table(pheno_mothers_cellcounts_svs, file=outfile, quote=F, row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1989c33-f06c-428f-93c7-395a22438af4",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## Childrens phenotype files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0754cdbe-4dbc-4155-a2b0-ab8a5787441d",
   "metadata": {},
   "source": [
    "### combine 15yrs and 17yrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc03da7a-aedc-4552-b451-e17a81deb1b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes\n",
    "\n",
    "### R \n",
    "setwd(\"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes\")\n",
    "f15 <- read.csv(\"pheno_kids_F15_ewas.csv\")\n",
    "f17 <- read.csv(\"pheno_kids_F17_ewas.csv\")\n",
    "linking_file <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/raw_data/alspac/B3176_datasetids.csv\")\n",
    "load(\"/home/ec2-user/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/samplesheet/data.Robj\")\n",
    "\n",
    "head(f15); head(f17)\n",
    "dim(f15) # 253 x 6\n",
    "dim(f17) # 716 x 6\n",
    "\n",
    "# combine phenotype datasets\n",
    "combined_pheno <- rbind(f15, f17)\n",
    "dim(combined_pheno) # 969 x 6\n",
    "head(combined_pheno)\n",
    "\n",
    "# see which samples have duplicate cidB3176 IDs. This indicates twins. These samples will have\n",
    "# separate Sample_Name IDs in the samplesheet. Take into account the qlet to determine the Sample_Name\n",
    "dup_ids <- which(duplicated(combined_pheno$cidB3176)) # 295 441\n",
    "combined_pheno[c(dup_ids-1,dup_ids), ]\n",
    "#    cidB3176 qlet age_at_DNAm cannabisUse smoking drinking\n",
    "#294     1525    A    18.00000           2    Ever  Current\n",
    "#440     5646    A    17.58333           2   Never     Ever\n",
    "#295     1525    B    18.00000           2    Ever     Ever\n",
    "#441     5646    B    17.58333           2   Never     Ever\n",
    "\n",
    "\n",
    "# keep only kids in F17 (17yr old) and TF3 (15yr old) in the samplesheet\n",
    "samplesheet_kids <-  samplesheet[which(samplesheet$time_code == \"F17\" | \n",
    "                                          samplesheet$time_code == \"TF3\"),]\n",
    "dim(samplesheet_kids) # 981 x 16\n",
    "\n",
    "# Remove duplicate and population stratification samples \n",
    "sample.rm <- which(samplesheet_kids$duplicate.rm==\"Remove\" | \n",
    "                   samplesheet_kids$genotypeQCkids == \"ETHNICITY\" | \n",
    "                   samplesheet_kids$genotypeQCkids==\"HZT;ETHNICITY\"|\n",
    "                   samplesheet_kids$genotypeQCmums==\"/strat\" )\n",
    "\n",
    "samplesheet_kids <- samplesheet_kids[-sample.rm, ]                                     \n",
    "\n",
    "dim(samplesheet_kids) # 925 x 16\n",
    "\n",
    "samplesheet_kids[which(duplicated(samplesheet_kids$ALN)), c(1,2,3,7,9)]\n",
    "#        Sample_Name   ALN QLET time_code Sex\n",
    "#4357 SLIDE84_R06C01 37785    B       F17   M\n",
    "#2347 SLIDE78_R03C02 36044    A       F17   F\n",
    "\n",
    "\n",
    "# add ALN IDs to the phenotype file from the linking file (ALN == dnam_450_g0m_g1)\n",
    "combined_pheno$ALN <- linking_file$dnam_450_g0m_g1[match(combined_pheno$cidB3176, linking_file$cidB3176)]\n",
    "\n",
    "# remove kids samples from phenotype file that are not in the curated samplesheet\n",
    "keep <- combined_pheno$ALN %in% samplesheet_kids$ALN\n",
    "combined_pheno <- combined_pheno[keep,]\n",
    "dim(combined_pheno) # 924 x 7\n",
    "\n",
    "# add Sample_Name to the phenotype file\n",
    "combined_pheno$Sample_Name <- samplesheet_kids$Sample_Name[match(combined_pheno$ALN, samplesheet_kids$ALN)]\n",
    "\n",
    "# add Sex to the phenotype file\n",
    "combined_pheno$sex <- samplesheet_kids$Sex[match(combined_pheno$ALN, samplesheet_kids$ALN)]\n",
    "\n",
    "dim(combined_pheno) # 924 x 9\n",
    "\n",
    "# NOTE: These below duplicates indicate twins. We will separate them out using\n",
    "# the qlet. This will enable us to get the unique Sample_Name\n",
    "dup_ids <- which(duplicated(combined_pheno$Sample_Name)) # 278, 420\n",
    "combined_pheno[c(dup_ids-1,dup_ids), ] # 277:278 and 419:420\n",
    "#    cidB3176 qlet age_at_DNAm cannabisUse smoking drinking   ALN Sample_Name    sex\n",
    "#294     1525    A    18.00000           2    Ever  Current 36044 SLIDE159_R04C01    F\n",
    "#295     1525    B    18.00000           2    Ever     Ever 36044 SLIDE159_R04C01    M\n",
    "#440     5646    A    17.58333           2   Never     Ever 37785 SLIDE345_R06C02    F\n",
    "#441     5646    B    17.58333           2   Never     Ever 37785 SLIDE345_R06C02    M\n",
    "\n",
    "samplesheet_kids[which(samplesheet_kids$ALN == 36044 | samplesheet_kids$ALN == 37785), 1:4]\n",
    "#         Sample_Name   ALN QLET    Slide\n",
    "#2350 SLIDE159_R04C01 36044    B SLIDE159\n",
    "#2347  SLIDE78_R03C02 36044    A  SLIDE78\n",
    "#4355 SLIDE345_R06C02 37785    A SLIDE345\n",
    "#4357  SLIDE84_R06C01 37785    B  SLIDE84\n",
    "\n",
    "# manually assign the Sample_Name to the twins\n",
    "combined_pheno[277, \"Sample_Name\"] <- \"SLIDE78_R03C02\"\n",
    "combined_pheno[420, \"Sample_Name\"] <- \"SLIDE84_R06C01\"\n",
    "\n",
    "which(duplicated(combined_pheno$Sample_Name)) # integer(0)\n",
    "\n",
    "dim(combined_pheno) # 924 x 9\n",
    "\n",
    "# reorder columns\n",
    "combined_pheno <- combined_pheno[, c(8,1,7,2,3,4,5,6,9)]\n",
    "\n",
    "# convert sex coding to PLINK format (M=1, F=2)\n",
    "combined_pheno[which(combined_pheno$sex == \"F\"), \"sex\"] <- 2\n",
    "combined_pheno[which(combined_pheno$sex == \"M\"), \"sex\"] <- 1\n",
    "\n",
    "outfile <- \"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/pheno_kids_combined_f15_f17_n924_ewas.txt\"\n",
    "write.table(combined_pheno,file=outfile,quote=F,row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "969e631c-adba-4352-b5f1-e69a666aa577",
   "metadata": {},
   "source": [
    "### add SVs and Cell-type proportions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e8c2a89-a3e1-4715-9e0b-1b673786e4e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellcounts <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/raw_data/alspac/dnam_450_g0m_g1/version1/derived/cellcounts/houseman/data.txt\",\n",
    "                      sep = \"\\t\")\n",
    "sv_data <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/alspac_control_probe_sva.txt\", \n",
    "                    sep = \" \")\n",
    "kids_pheno <- read.csv(\"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/pheno_kids_combined_f15_f17_n924_ewas.txt\",\n",
    "                      sep = \" \")\n",
    "\n",
    "dim(cellcounts) # 4854 x 7\n",
    "dim(sv_data) # 4854 x 7\n",
    "dim(kids_pheno) # 924 x 9\n",
    "\n",
    "# attach cell type estimations and SVs to pheno\n",
    "pheno_kids_cellcounts <- merge(kids_pheno,cellcounts,by.x=\"Sample_Name\",by.y=\"Sample_Name\")\n",
    "pheno_kids_cellcounts_svs <- merge(pheno_kids_cellcounts,sv_data,by.x=\"Sample_Name\",by.y=\"Sample_Name\")\n",
    "dim(pheno_kids_cellcounts_svs) # 924 x 21\n",
    "\n",
    "outfile <- \"/home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/pheno_kids_combined_f15_f17_n924_ewas_final.txt\"\n",
    "write.table(pheno_kids_cellcounts_svs, file=outfile, quote=F, row.names=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7520428-0a4a-43ef-9c37-88da3530c63b",
   "metadata": {},
   "source": [
    "### S3 upload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7796b6e-ead2-46f3-aa26-807538a0049c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# upload to S3\n",
    "cd /home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes\n",
    "\n",
    "aws s3 cp pheno_kids_combined_f15_f17_n924_ewas_final.txt \\\n",
    "    s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/\n",
    "\n",
    "aws s3 cp pheno_mothers_combined_FOM_TF1_3_n946_ewas_final.txt \\\n",
    "    s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab353b5-ddfc-4020-a71b-f3fe27a64677",
   "metadata": {},
   "source": [
    "## Childrens Cannabis Use in Never Smokers\n",
    "| Never Smokers | cannabisUse 1 | cannabisUse 2 |\n",
    "|---------------|---------------|---------------|\n",
    "| 218           | 20            | 198           |\n",
    "\n",
    "According to Fang Fang 1=ever and 2=never."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0faaf8fe-7b7a-4cbc-9f2d-ac1e8a0229a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/ec2-user/rti-cannabis/shared_data/post_qc/alspac/phenotypes/children_cannabis_in_never_smokers/\n",
    "\n",
    "aws s3 cp s3://rti-cannabis/ewas/alspac/data/pheno_kids_combined_f15_f17_n924_ewas_final.txt .\n",
    "\n",
    "wc -l pheno_kids_combined_f15_f17_n924_ewas_final.txt\n",
    "#925 pheno_kids_combined_f15_f17_n924_ewas_final.txt\n",
    "#Sample_Name cidB3176 ALN qlet age_at_DNAm cannabisUse smoking drinking sex Bcell CD4T CD8T Gran Mono NK sv1 sv2 sv3 sv4 sv5 sv6\n",
    "\n",
    "cat pheno_kids_combined_f15_f17_n924_ewas_final.txt  | awk '$7==\"Never\" {print $7}'|wc -l\n",
    "#218\n",
    "cat pheno_kids_combined_f15_f17_n924_ewas_final.txt  | awk '$7!=\"Never\" {print $7}'| wc -l\n",
    "#707\n",
    "\n",
    "\n",
    "head -1 pheno_kids_combined_f15_f17_n924_ewas_final.txt >\\\n",
    "    pheno_kids_combined_f15_f17_n218_never_smokers_ewas_final.txt\n",
    "\n",
    "cat pheno_kids_combined_f15_f17_n924_ewas_final.txt  |\\\n",
    "    awk '$7==\"Never\" {print $0}' >>\\\n",
    "    pheno_kids_combined_f15_f17_n218_never_smokers_ewas_final.txt\n",
    "\n",
    "awk '$6==2' pheno_kids_combined_f15_f17_n218_never_smokers_ewas_final.txt  | wc -l\n",
    "#198\n",
    "ec2-user@ip-172-31-15-238:children_cannabis_in_never_smokers$ awk '$6==1' pheno_kids_combined_f15_f17_n218_never_smokers_ewas_final.txt  | wc -l\n",
    "#20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e19bd92-2949-4b19-acb5-d079e5fdd6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## upload to S3\n",
    "aws s3 cp pheno_kids_combined_f15_f17_n218_never_smokers_ewas_final.txt \\\n",
    "    s3://rti-cannabis/ewas/alspac/data/\n",
    "\n",
    "aws s3 cp pheno_kids_combined_f15_f17_n218_never_smokers_ewas_final.txt \\\n",
    "    s3://rti-cannabis/shared_data/post_qc/alspac/phenotypes/\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  },
  "toc-autonumbering": true,
  "toc-showcode": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
