{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VCF file from MIS\n",
    "##fileformat=VCFv4.1\n",
    "##filedate=2018.3.17\n",
    "##source=Minimac3\n",
    "##contig=<ID=15>\n",
    "##FILTER=<ID=GENOTYPED,Description=\"Marker was genotyped AND imputed\">\n",
    "##FILTER=<ID=GENOTYPED_ONLY,Description=\"Marker was genotyped but NOT imputed\">\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##FORMAT=<ID=DS,Number=1,Type=Float,Description=\"Estimated Alternate Allele Dosage : [P(0/1)+2*P(1/1)]\">\n",
    "##FORMAT=<ID=GP,Number=3,Type=Float,Description=\"Estimated Posterior Probabilities for Genotypes 0/0, 0/1 and 1/1 \">\n",
    "##INFO=<ID=AF,Number=1,Type=Float,Description=\"Estimated Alternate Allele Frequency\">\n",
    "##INFO=<ID=MAF,Number=1,Type=Float,Description=\"Estimated Minor Allele Frequency\">\n",
    "##INFO=<ID=R2,Number=1,Type=Float,Description=\"Estimated Imputation Accuracy\">\n",
    "##INFO=<ID=ER2,Number=1,Type=Float,Description=\"Empirical (Leave-One-Out) R-square (available only for genotyped variants)\n",
    "\">\n",
    "#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  EMERGE-1_16224699       EMERGE-1_16224832       EM\n",
    "ERGE-1_16225552 EMERGE-1_16225624       EMERGE-1_16226154       EMERGE-1_16233448       EMERGE-2_16229819       EMERGE-3_1\n",
    "6218877 EMERGE-3_16219796       EMERGE-3_16220543       EMERGE-3_16220698       EMERGE-3_16220967       EMERGE-3_16221838EMERGE-3_16222896        EMERGE-3_16223176       EMERGE-3_16223472       EMERGE-3_16223514       EMERGE-3_16223639       EM\n",
    "ERGE-3_16225243 EMERGE-3_16225590       EMERGE-3_16225928       EMERGE-3_16226294       EMERGE-3_16228484       EMERGE-3_1\n",
    "6228513 EMERGE-3_16229771       EMERGE-3_16229817       EMERGE-3_16229912       EMERGE-3_16230463       EMERGE-3_16230480EMERGE-3_16230920        EMERGE-3_16232087       EMERGE-3_16232393       EMERGE-3_16232421       EMERGE-3_16232533       EM\n",
    "ERGE-3_16233580 EMERGE-3_16234445       EMERGE-3_16234448       EMERGE-4_16226213       EMERGE-4_16226273       EMERGE-5_1\n",
    "6223236 EMERGE-6_16221489       EMERGE-7_16220844       EMERGE-8_16220003       EMERGE-9_16217786       EMERGE-9_16231695EMERGE-10_16230033       EMERGE-10_16230638      EMERGE-11_16214874      EMERGE-11_16215381      EMERGE-11_16215457  \n",
    "\n",
    "---\n",
    "\n",
    "\n",
    "\n",
    "**rs16969968 gwas results**\n",
    "rs16969968 15 78882925 G A 0.34688 0.34688 0.997 0.98853 730 0.352004 0.0525009 0.0338811 2.42038 2.4011455861625 0.121246432020861 1.05(0.99-1.13)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## info file\n",
    "SNP     REF(0)  ALT(1)  ALT_Frq MAF     AvgCall Rsq     Genotyped       LooRsq  EmpR    EmpRsq  Dose0   Dose1\n",
    "15:20000041     T       A       0.00594 0.00594 0.99406 0.00013 Imputed -       -       -       -       -\n",
    "15:20000075     C       T       0.01359 0.01359 0.98641 0.00019 Imputed -       -       -       -       -\n",
    "15:20000114     G       T       0.00998 0.00998 0.99002 0.00040 Imputed -       -       -       -       -\n",
    "15:20000123     T       C       0.00113 0.00113 0.99887 0.00001 Imputed -       -       -       -       -\n",
    "15:20000127     A       T       0.00363 0.00363 0.99637 0.00006 Imputed -       -       -       -       -\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter by P-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 console ##\n",
    "cd /shared/s3/emerge/data/assoc_tests/002/final\n",
    "\n",
    "ls \n",
    "\n",
    "\"\"\"emerge_ea.1000G.chr10.CAT_FTND~SNP+SEX+EVs.subject_1000g_maf_gt_0.01_rsq_gt_0.30_ea.gz\n",
    "...\n",
    "\"\"\"\n",
    "\n",
    "outFile=emerge.ea.1000G_p3.CAT_FTND~SNP+SEX+EVs.maf_gt_0.01.rsq_gt_0.30.p_lte_0.001\n",
    "zcat emerge_ea.1000G.chr10.CAT_FTND~SNP+SEX+EVs.subject_1000g_maf_gt_0.01_rsq_gt_0.30_ea.gz |\\\n",
    "    head > $ouFile\n",
    "for file in *; do\n",
    "    echo Processing $file\n",
    "    zcat $file | tail -n +2 |\n",
    "    perl -lane 'if ($F[15] <= 0.001) { print; }' >> \\\n",
    "    $outFile\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### START Filter by p-value ###\n",
    "\n",
    "# MAF > 0.01 in AFR and EUR\n",
    "#for ancestry in aa; do\n",
    "  if [ $ancestry == \"aa\" ]; then\n",
    "    group=afr\n",
    "  else\n",
    "    group=eur\n",
    "  fi\n",
    "  for ext in $group ${group}+$COHORT.rsq_gt_0.3; do\n",
    "    outFile=$ASSOCIATION_ROOT/$ancestry/processing/$COHORT.$ancestry.1000G_p3.$MODEL.maf_gt_0.01_$ext.p_lte_0.001\n",
    "    head -n 1 $ASSOCIATION_ROOT/$ancestry/final/$COHORT.$ancestry.1000G_p3.chr1.$MODEL.maf_gt_0.01_$ext > \\\n",
    "      $outFile\n",
    "    for (( chr=1; chr<24; chr++ )); do\n",
    "      echo Processing ${ASSOCIATION_ROOT}/$ancestry/final/$COHORT.$ancestry.1000G_p3.chr$chr.$MODEL.maf_gt_0.01_$ext\n",
    "      tail -n +2 $ASSOCIATION_ROOT/$ancestry/final/$COHORT.$ancestry.1000G_p3.chr$chr.$MODEL.maf_gt_0.01_$ext |\n",
    "        perl -lane 'if ($F[15] <= 0.001) { print; }' >> \\\n",
    "        $outFile\n",
    "    done\n",
    "done\n",
    "#done\n",
    "\n",
    "# Sort\n",
    "#for (ancestry in c(\"ea\",\"aa\")){\n",
    "for (ancestry in c(\"aa\")){\n",
    "        cohort=\"copdgene2\"\n",
    "        if (ancestry == \"aa\") { group = \"afr\" } else if (ancestry == \"ea\") { group = \"eur\" }\n",
    "        dat=read.table(paste0('/share/nas04/jmarks/gwas/',cohort, '/association_tests/001/',\n",
    "                              ancestry,'/processing/', cohort, '.',ancestry,\n",
    "                              '.1000G_p3.CAT_FTND~SNP+AGE+SEX+GOLD+EVs.maf_gt_0.01_',\n",
    "                              group,'.p_lte_0.001'), header = TRUE)\n",
    "        dat = dat[order(dat$p),]\n",
    "        write.csv(dat,\n",
    "            file = paste0('/share/nas04/jmarks/gwas/',cohort,'/association_tests/001/',\n",
    "                          ancestry,'/final/', cohort, '.',ancestry,'.1000G_p3.CAT_FTND~SNP+AGE+SEX+GOLD+EVs.maf_gt_0.01_',\n",
    "                          group,'.p_lte_0.001.csv'), row.names = FALSE)\n",
    "\n",
    "\n",
    "        dat=read.table(paste0('/share/nas04/jmarks/gwas/', cohort,'/association_tests/001/',\n",
    "                              ancestry,'/processing/', cohort, '.',ancestry,\n",
    "                              '.1000G_p3.CAT_FTND~SNP+AGE+SEX+GOLD+EVs.maf_gt_0.01_',group,\n",
    "                              '+', cohort, '.rsq_gt_0.3.p_lte_0.001'), header=TRUE)\n",
    "        dat = dat[order(dat$p),]\n",
    "        write.csv(dat,\n",
    "                  file = paste0('/share/nas04/jmarks/gwas/', cohort, '/association_tests/001/',\n",
    "                         ancestry,'/final/', cohort, '.', ancestry, '.1000G_p3.CAT_FTND~SNP+AGE+SEX+GOLD+EVs.maf_gt_0.01_',\n",
    "                         group,'+',cohort,'.rsq_gt_0.3.p_lte_0.001.csv'), row.names = FALSE)\n",
    "}\n",
    "\n",
    "## END Filter by p-value ###\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
