{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data processing for dbGaP phs000428.v2.p2.c1 (Health and Retirement Study)\n",
    "Genetics Resource with the Health and Retirement Study (phs000428.v2.p2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Author:** Jesse Marks\n",
    "\n",
    "This document logs several components of data processing for [dbGaP study phs000428.v2.p2.c1](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000428.v2.p2) including \n",
    "\n",
    "* Data retrieval\n",
    "* Genotype data quality control.\n",
    "\n",
    "The purpose of processing these data are to prepare them for further processing and analysis steps such as haplotype phasing, imputation, and genome-wide association analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Software and tools\n",
    "The software and tools used for processing these data are\n",
    "* Windows 10 with [Cygwin](https://cygwin.com/) installed \n",
    "* [Aspera Connect](http://downloads.asperasoft.com/downloads)\n",
    "* [KING](http://people.virginia.edu/~wc9c/KING/)\n",
    "* [PLINK v1.9 beta 3.45](https://www.cog-genomics.org/plink/) \n",
    "* [SRA toolkit](https://www.ncbi.nlm.nih.gov/sra/docs/toolkitsoft/)\n",
    "* [STRUCTURE](https://web.stanford.edu/group/pritchardlab/structure.html)\n",
    "* [R v3.2.3](https://www.r-project.org/)\n",
    "* [iGraph (R package)](http://igraph.org/r/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data retieval\n",
    "### Genotypes and phenotypes\n",
    "The genotype and phenotype data were downloaded from dbGaP to a local machine, decrypted, and then transfered to [Amazon Simple Storage Service (S3)](https://aws.amazon.com/s3/?sc_channel=PS&sc_campaign=acquisition_US&sc_publisher=google&sc_medium=s3_b&sc_content=s3_e&sc_detail=amazon%20s3&sc_category=s3&sc_segment=192085379926&sc_matchtype=e&sc_country=US&s_kwcid=AL!4422!3!192085379926!e!!g!!amazon%20s3&ef_id=Wd4L7QAAAFUPlk0C:20180115144528:s). \n",
    "Some of the data were not decrypted locally and will thus need to be decrypted on [Amazon Elastic Compute Cloud (EC2)](https://aws.amazon.com/ec2/?sc_channel=PS&sc_campaign=acquisition_US&sc_publisher=google&sc_medium=ec2_b_rlsa_hv&sc_content=ec2_e&sc_detail=amazon%20ec2&sc_category=ec2&sc_segment=213206985258&sc_matchtype=e&sc_country=US&s_kwcid=AL!4422!3!213206985258!e!!g!!amazon%20ec2&ef_id=Wd4L7QAAAFUPlk0C:20180115144653:s).\n",
    "The data were downloaded using the Aspera Connect Browser plug-in. Note that these data require authorized access, so the [authorized access portal](https://dbgap.ncbi.nlm.nih.gov/dbgap/aa/wga.cgi?page=login) must be used (request login information from Eric Johnson). The data files downloaded from dbGaP are encrypted and thus will need to be decrypted using `vdb-decrypt` from the SRA toolkit [(instructions here)](https://www.ncbi.nlm.nih.gov/books/NBK63512/#_Download_Points_often_Ignored_When_Decry_).\n",
    "\n",
    "Note: As a way to conserve space, certain genotype data for a given study are excluded from download. General criteria are:\n",
    "* Exclude imputed data\n",
    "* Exclude individual format genotype calls if the matrix and/or PLINK binary fileset format is available\n",
    "* Exclude index files that lay out the directory structure for the individual format genotype calls\n",
    "* Exclude raw array data if genotype calls are available\n",
    "\n",
    "To assess which files may be unnecessary for download, the study report available through the [public FTP download site]() (accessible via the dbGaP landing page for a given study accession) should be examined."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General directory structure setup\n",
    "\n",
    "After locally downloading, all of the dbGaP data should be organized within a directory called `ncbi`. For each cohort, its data needs to be placed within a subdirectory of the format dbGaP-x where x signifies the project number of the download (not to be confused with the download request number). The project number can be found on the \"Downloads\" tab of the dbGaP authorized access portal. Also in that tab is the link to download the repository key file. This file should be placed in the top level of the project folder then imported using [vdb-config](https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std). Although each download request provides a link to download the repository key, only one key file is needed per project.\n",
    "\n",
    "**Note:** File names longer than the allowed Windows file name character limit will not decrypt and must be renamed before decrypting. It is highly recommended to check the file names after decryption to ensure that they successfully decrypted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Local machine cygwin terminal ##\n",
    "\n",
    "# Create NCBI and project specific directory\n",
    "mkdir -p /cygdrive/c/Users/bquach/ncbi/dbGaP-2556\n",
    "\n",
    "# Download data and repo key to the appropriate project directory \n",
    "cd /cygdrive/c/Users/bquach/ncbi/dbGaP-2556"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Amazon Web Services S3 directory structure setup\n",
    "For each dbGaP study, we will use the directory structure setup outlined on [GitHub for AWS S3](https://github.com/RTIInternational/bioinformatics/blob/master/documentation/aws_s3_data_organization.md)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Connecting to AWS - using Cygwin\n",
    "To connect to AWS using the SSH client Cygwin, you will need to utilize PuTTYgen on Windows to generate SSH key pairs. The two links [here](https://stackoverflow.com/questions/2224066/how-to-convert-ssh-keypairs-generated-using-puttygenwindows-into-key-pairs-use) and [here](https://stackoverflow.com/questions/2419566/best-way-to-use-multiple-ssh-private-keys-on-one-client) explain how to convert the Putty key into key-pairs and then how to conveniently login to AWS. The private key from PuTTygen output is saved to:\n",
    "\n",
    "```\n",
    "~/.ssh/gwas_rsa\n",
    "```\n",
    "This private key is then referenced in the ssh config file ```~/.ssh/config```. In the config file, I use the shortname ```AWS``` so that I can login with \n",
    "\n",
    "```ssh AWS\n",
    "```.\n",
    "\n",
    "**Note:**  `User` for the config file is ```jmarks``` - in my case - and ```HostName``` is the IP address ```50.19.195.254```.\n",
    "\n",
    "* Changed maintain_initial_size parameter to ```true```.\n",
    "* ebs_snapshot_id = snap-06d1702c532ba16c7\n",
    "* volume_type = gp2\n",
    "\n",
    "**Note:** the Perl scripts will have a slightly different path on EC2 than the ones that were on MIDAS.\n",
    "For other software, such as R, the paths will probably be the same but the names slightly altered."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  Install Amazon Web Services Command Line Interface (AWS CLI) \n",
    "The Amazon Web Services Command Line Interface (AWS CLI) needs to be installed in order to upload the data from a local machine to S3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:6: unexpected string constant\n1: # Install pip, a package manager for python applications on local machine\n2: curl \"https://bootstrap.pypa.io/get-pip.py\"\n        ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:6: unexpected string constant\n1: # Install pip, a package manager for python applications on local machine\n2: curl \"https://bootstrap.pypa.io/get-pip.py\"\n        ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "# Install pip, a package manager for python applications on local machine\n",
    "curl \"https://bootstrap.pypa.io/get-pip.py\" -o \"get-pip.py\"\n",
    "python get-pip.py\n",
    "\n",
    "# Install awscli via pip\n",
    "pip install awscli\n",
    "\n",
    "# Verify installation - should see something similar to the string below\n",
    "aws --version\n",
    "\"aws-cli/1.11.178 Python/2.7.13 CYGWIN_NT-10.0/2.9.0(0.318/5/3) botocore/1.7.36\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Configure AWS \n",
    "The settings for using the AWS CLI need to be configured before interacting with AWS. These configurations include security credentials and the default region. For more information on this process, see [here](http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html).\n",
    "\n",
    "**Note:** these only need to be configured once. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Local machine (Cygwin)\n",
    "aws configure\n",
    "\n",
    "AWS Access Key ID [None]: AKIAJONBCJHOJSW2PFJA\n",
    "AWS Secret Access Key [None]:  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\n",
    "Default region name [None]: us-east-1\n",
    "Default output format [None]: text  # could be json, text, or table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Modify an EBS Volume from the Command Line\n",
    "The following example demonstrates how an EBS volume can be modified from the command line using hte AWS CLI. An example is detailed [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/cli-modify.html).To make use of the new storage capacity after modifying the ebs volume you need to [extend a linux file system after resizing](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/recognize-expanded-volume-linux.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "\n",
    "# first configure AWS. you will be prompted to enter the following information\n",
    "# the keys and default region can be found in the config file on the cluster launcher\n",
    "aws configure\n",
    "'''\n",
    "AWS Access Key ID [None]: AKIAI44QH8DHBEXAMPLE\n",
    "AWS Secret Access Key [None]: je7MtGbClwBF/2Zp9Utk/h3yCo8nvbEXAMPLEKEY\n",
    "Default region name [None]: us-east-1\n",
    "Default output format [None]: text\n",
    "'''\n",
    "\n",
    "\n",
    "aws ec2 modify-volume --dry-run --volume-id vol-038921893392154fa --size 950  \n",
    "\n",
    "# extend file system to the new volume capacity.\n",
    "sudo resize2fs /dev/xvdb\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotype processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genetics Resource with the Health and Retirement Study (phs000428.v2.p2)\n",
    "Genetics Resource with the Health and Retirement Study \n",
    "[dbGaP study phs000428.v2.p2.c1](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000428.v2.p2)\n",
    "* Download 58712"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Local machine (Cygwin) ##\n",
    "\n",
    "# Decrypt (must be executed within project directory)\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58712\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58712/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# Create directory structure\n",
    "mkdir -p s3/phs000428_retirement/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000428_retirement/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000428_retirement/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# Move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58712/PhenoGenotypeFiles/RootStudyConsentSet_phs000428.CIDR_Aging_Omni1.v2.p2.c1.NPR\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000428_retirement/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000428_retirement/phenotype/unprocessed/\n",
    "mv StudyMetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000428_retirement/meta/\n",
    "\n",
    "# Upload to S3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000428_retirement/ s3://rti-common/dbGaP/phs000428_retirement/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating an instance on EC2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# local machine\n",
    "ssh AWS\n",
    "\n",
    "# cluster config server\n",
    "cnfcluster create 428retirement\n",
    "\n",
    "'''   \n",
    "Output:\"MasterPublicIP\"=\"35.169.161.38\"\n",
    "Output:\"MasterPrivateIP\"=\"172.31.29.63\"\n",
    "Output:\"GangliaPublicURL\"=\"http://35.169.161.38/ganglia/\"\n",
    "Output:\"GangliaPrivateURL\"=\"http://172.31.29.63/ganglia/\"\n",
    "'''\n",
    "\n",
    "# local machine, note that username is ec2-user\n",
    "    ssh -i ~/.ssh/gwas_rsa ec2-user@35.169.161.38"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieving data from S3\n",
    "To avoid writing over the current directory structure, I create a new directory within the directory structure of the other studies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "# Retrieve data from S3 and store in EC2\n",
    "cd /shared/data/studies\n",
    "aws s3 cp s3://rti-common/dbGaP/phs000428_retirement/ phs000428_retirement --recursive\n",
    "\n",
    "# create directory structure\n",
    "cd /shared/data/studies/phs000428_retirement/\n",
    "mkdir phenotype/{final,processing} \n",
    "mkdir genotype/original/{final,processing}\n",
    "\n",
    "# gunzip phenotype data\n",
    "cp phenotype/unprocessed/*.gz phenotype/processing\n",
    "gunzip -r phenotype/processing/*    \n",
    "\n",
    "# gunzip genotype data\n",
    "cp genotype/original/unprocessed/* genotype/original/processing &\n",
    "gunzip -r genotype/original/processing/*.gz &"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quality Control Sample Tracking\n",
    "#### Pre-chromosome type partitioning\n",
    "The table below provides statistics on variants and subjects filtered during each step of the QC process."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| QC procedure                         | Variants removed | Variants retained | Subjects removed | Subjects retained |\n",
    "|--------------------------------------|------------------|-------------------|------------------|-------------------|\n",
    "| Initial dbGaP dataset                | 0                | 2315518           | 0                | 2559              |\n",
    "| Genome build 37 and dbGaP 138 update | 200149           | 2115369           | 0                | 2559              |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Autosome statistics\n",
    "This table includes filtering statistics prior to merging with chrX.\n",
    "\n",
    "| QC procedure                                    | Variants removed | Variants retained | Subjects removed | Subjects added | Subjects retained |\n",
    "|-------------------------------------------------|------------------|-------------------|------------------|----------------|-------------------|\n",
    "| Pre-partitioning w/initial procedures (all chr) |    200149        |    2115369        |      0           |      0         | 2559             |\n",
    "| Partitioning to only autosomes                  |     39206        |    2076163        |      0           |      0         |   2559             |\n",
    "| Remove subjects missing whole autosome data     |       0          |    2076163        |      0           |      0         |   2559             |\n",
    "| Duplicate rsID filtering                        |       0          |    2076163        |      0           |      0         |   2559             |\n",
    "| Remove ancestral outliers                       |                 |           |                 |               |              |\n",
    "| Remove sujects with re-assigned ancestry        |                 |           |                |               |              |\n",
    "| Add subjects re-assigned by STRUCTURE           |                 |            |                 |              |              |\n",
    "| Remove variants with missing call rate > 3%     |              |          |                 |               |              |\n",
    "| Remove variants with HWE p < 0.0001             |             |            |                 |               |              |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ChrX statistics¶\n",
    "This table includes filtering statistics prior to merging with autosomes.\n",
    "\n",
    "| QC procedure                                    | Variants removed | Variants retained | Subjects removed | Subjects added | Subjects retained |\n",
    "|-------------------------------------------------|------------------|-------------------|------------------|----------------|-------------------|\n",
    "| Pre-partitioning w/initial procedures (all chr) |    200149        |       2115369     |         0        |        0       |     2559           |\n",
    "| Partitioning to only chrX                       |    2077494       |       37875       |         0        |        0       |     2559           |\n",
    "| Remove subjects missing whole autosome data     |         0        |       37875       |         0        |        0       |     2559           |\n",
    "| Duplicate rsID filtering                        |         0        |       37875       |         0        |        0       |     2559           |\n",
    "| Remove ancestral outliers                       |                 |            |                 |               |              |\n",
    "| Remove sujects with re-assigned ancestry        |                 |            |                |               |              |\n",
    "| Add subjects re-assigned by STRUCTURE           |                 |            |                 |              |              |\n",
    "| Remove variants with missing call rate > 3%     |            |            |                 |               |              |\n",
    "| Remove variants with HWE p < 0.0001             |             |            |                 |               |              |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merged autosome and chrX statistics\n",
    "\n",
    "| QC procedure                                            | Variants removed | Variants retained | Subjects removed | Subjects added | Subjects retained |\n",
    "|---------------------------------------------------------|------------------|-------------------|------------------|----------------|-------------------|\n",
    "| Merge autosomes and chrX                                |                 |            |                 |             |              |\n",
    "| Remove subjects with IBD > 0.4, IBS > 0.9, KING > 0.177 |                 |            |                 |             |              |\n",
    "| Remove subjects with missing call rate > 3%             |                 |            |                 |             |              |\n",
    "| Sex discordance filter                                  |                 |            |                 |             |              |\n",
    "| Excessive homozygosity filter                           |                 |            |                 |             |              |\n",
    "| Duplicate variant ID filter after 1000G renaming        |                 |            |                 |             |              |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ancestry partitioning\n",
    "The data contains AA subjects as well as non-AA subjects. We will only focus on the AA subjects for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/phenotype/processing\n",
    "\n",
    "# grab the subject ids of the AA subjects\n",
    "tail -n +12 phs000428.v2.pht002614.v2.p2.c1.HRS_Subject_Phenotypes.NPR.txt | awk '{ if($5==1){print $2 } }\n",
    "' > aa_subjects_ids.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exclusion of subjects without phenotype data\n",
    "The .fam file may contain more subject IDs than the phenotype file. The subjects without phenotype data are excluded as they provide no benefit for GWA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/\n",
    "\n",
    "mkdir aa\n",
    "\n",
    "# Add family IDs\n",
    "grep -f ../../../phenotype/processing/aa_subject_ids.txt \\\n",
    "    phase123.genotype-calls-matrixfmt.c1/subject_level_PLINK_sets/HRS_phase123_TOP.fam | \\\n",
    "    cut -d ' ' -f 1,2    \\\n",
    "    > ../../../phenotype/processing/aa_subject_ids.keep\n",
    "\n",
    "# Create filtered PLINK filesets\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile  phase123.genotype-calls-matrixfmt.c1/subject_level_PLINK_sets/HRS_phase123_TOP \\\n",
    "        --keep ../../../phenotype/processing/aa_subject_ids.keep \\\n",
    "        --make-bed \\\n",
    "        --out aa/genotypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map kgp IDs to rs IDs\n",
    "The genotype information is given in kgp format. We need to map this to rs IDs before we update the dbSNP and genonome build. Duplicates may arise and will need to be filtered based on missing call rate. The variant with the highest missing call rate will be removed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "/shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# change conversion file to tab-separated\n",
    "sed 's/,/\\t/g' phase1.marker-info.MULTI/SNP_kgpID2rsID.csv >\\\n",
    "phase1.marker-info.MULTI/SNP_kgpID2rsID.tsv\n",
    "\n",
    "# remove duplicate IDs \n",
    "sort -u -k 1,1 phase1.marker-info.MULTI/SNP_kgpID2rsID.tsv >\\\n",
    "phase1.marker-info.MULTI/SNP_kgpID2rsID.sorted\n",
    "\n",
    "\n",
    "# convert kgp ID to rs ID\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile aa/genotypes \\\n",
    "        --update-name phase1.marker-info.MULTI/SNP_kgpID2rsID.sorted \\\n",
    "        --make-bed \\\n",
    "        --out aa/genotypes_kgpID2rsID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  Remove duplicate rsIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# Append _X (where X is a number) to the end of the rs IDs for all but 1st occurrence of duplicates\n",
    "perl -lane 'BEGIN { %idCounts = (); }\n",
    "                    if (exists($idCounts{$F[1]})) {\n",
    "                        $idCounts{$F[1]}++;\n",
    "                        print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "                    } else {\n",
    "                        $idCounts{$F[1]} = 1;\n",
    "                        print;\n",
    "                    } ' aa/genotypes_kgpID2rsID.bim > aa/genotypes_kgpID2rsID_renamed.bim\n",
    "\n",
    "# create a list of the duplicate snps\n",
    "grep _ aa/genotypes_kgpID2rsID_renamed.bim | perl -lane \\\n",
    "'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' >\\\n",
    "aa/genotypes_kgpID2rsID_renamed.duplicate_snps\n",
    "\n",
    "\n",
    "# Get call rates for duplicate SNPs\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --bed aa/genotypes_kgpID2rsID.bed \\\n",
    "            --bim aa/genotypes_kgpID2rsID_renamed.bim \\\n",
    "            --fam aa/genotypes_kgpID2rsID.fam \\\n",
    "            --extract aa/genotypes_kgpID2rsID_renamed.duplicate_snps \\\n",
    "            --missing \\\n",
    "            --out aa/genotypes_kgpID2rsID.duplicate_snps.missing\n",
    "\n",
    "# Create remove list for duplicates containing duplicate with higher missing rate\n",
    "tail -n +2 aa/genotypes_kgpID2rsID.duplicate_snps.missing.lmiss |\\\n",
    " perl -lane 'BEGIN { %missingness = (); }\n",
    "                        if ($F[1] =~ /^(\\S+)_/) {\n",
    "                            $duplicateName = $1\n",
    "                        } else {\n",
    "                            $duplicateName = $F[1].\"_2\";\n",
    "                        }\n",
    "                        if (exists($missingness{$duplicateName})) {\n",
    "                            if ($missingness{$duplicateName} > $F[4]) {\n",
    "                                print $duplicateName;\n",
    "                            } else {\n",
    "                                print $F[1];\n",
    "                            }\n",
    "                        } else {\n",
    "                            $missingness{$F[1]} = $F[4];\n",
    "                        }' > aa/genotypes_kgpID2rsID.duplicate_snps.remove\n",
    "\n",
    "# Remove duplicates with higher missing rate\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "   --bed /shared/data/studies/phs000428_retirement/genotype/original/processing/aa/genotypes_kgpID2rsID.bed \\\n",
    "   --bim /shared/data/studies/phs000428_retirement/genotype/original/processing/aa/genotypes_kgpID2rsID_renamed.bim \\\n",
    "   --fam /shared/data/studies/phs000428_retirement/genotype/original/processing/aa/genotypes_kgpID2rsID.fam \\\n",
    "   --exclude /shared/data/studies/phs000428_retirement/genotype/original/processing/aa/genotypes_kgpID2rsID.duplicate_snps.remove \\\n",
    "   --make-bed \\\n",
    "   --out /shared/data/studies/phs000428_retirement/genotype/original/processing/aa/genotypes_kgpID2rsID_duplicates_removed\n",
    "\n",
    "# Remove \"_2\" from rs IDs\n",
    "perl -i.bak -lne 's/_2//; print;'  aa/genotypes_kgpID2rsID_duplicates_removed.bim\n",
    " \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Update dbSNP and genome build\n",
    "\n",
    "To ensure that all of the population controls have variant and genomic data in dbSNP 138 and genome build 37 format, I use ID and position mappers to make the updates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC1 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# update name to dbSNP138\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile aa/genotypes_kgpID2rsID_duplicates_removed \\\n",
    "        --update-name /shared/bioinformatics/methods/nas03/bioinformatics_group/common/snp_id_conversion/b138/old_to_current.xref \\\n",
    "        --make-bed \\\n",
    "        --out aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update\n",
    "\n",
    "\n",
    "# There are still some duplicates that need to be taken care of again.\n",
    "# Append _X (where X is a number) to the end of the rs IDs for all but 1st occurrence of duplicates\n",
    "perl -lane 'BEGIN { %idCounts = (); }\n",
    "                    if (exists($idCounts{$F[1]})) {\n",
    "                        $idCounts{$F[1]}++;\n",
    "                        print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "                    } else {\n",
    "                        $idCounts{$F[1]} = 1;\n",
    "                        print;\n",
    "                    } ' aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update.bim > aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_X.bim\n",
    "\n",
    "# create a list of the duplicate snps\n",
    "grep _ aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_X.bim | perl -lane \\\n",
    "'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' >\\\n",
    "aa/genotypes_kgpID2rsID_renamed.duplicate_snps_X\n",
    "\n",
    "# remove the snps from the list that do not have _X in the id\n",
    "grep '_' aa/genotypes_kgpID2rsID_renamed.duplicate_snps_X > aa/genotypes_kgpID2rsID_renamed.duplicate_snps_X.remove\n",
    "aa/genotypes_kgpID2rsID_renamed.duplicate_snps_X.remove\n",
    "\n",
    "# Remove duplicates with _ in name\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "   --bed aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update.bed \\\n",
    "   --bim aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_X.bim \\\n",
    "   --fam aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update.fam \\\n",
    "   --exclude aa/genotypes_kgpID2rsID_renamed.duplicate_snps_X.remove \\\n",
    "   --make-bed \\\n",
    "   --out aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_clean\n",
    "\n",
    "\n",
    "\n",
    "# Update variant chr \n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_clean \\\n",
    "        --update-chr /shared/bioinformatics/methods/nas03/bioinformatics_group/common/\\\n",
    "build_conversion/b37/dbsnp_b138/uniquely_mapped_snps.chromosomes \\\n",
    "        --make-bed \\\n",
    "        --out aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_clean_chr_b37\n",
    "# Note:  688235 values updated, 60400964 variant IDs not present.\n",
    "\n",
    "\n",
    "# Update variant chr coordinate\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_clean_chr_b37 \\\n",
    "        --update-map /shared/bioinformatics/methods/nas03/bioinformatics_group/common/\\\n",
    "build_conversion/b37/dbsnp_b138/uniquely_mapped_snps.positions \\\n",
    "        --make-bed \\\n",
    "        --out aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_clean_chr_position_b37\n",
    "\n",
    "# Filter to only build 37 uniquely mapped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile aa/genotypes_kgpID2rsID_duplicates_removed_rsid_update_clean_chr_position_b37 \\\n",
    "        --extract /shared/bioinformatics/methods/nas03/bioinformatics_group/common/\\\n",
    "build_conversion/b37/dbsnp_b138/uniquely_mapped_snps.ids \\\n",
    "        --make-bed \\\n",
    "        --out aa/genotypes_b37_dbsnp138"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### intermittent upload to S3\n",
    "As a precaution, we should upload our results to S3 periodically for safe keeping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/\n",
    "aws s3 cp aa/ s3://rti-common/dbGaP/phs000428_retirement/genotype/original/processing/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Partition into autosome and chrX groups\n",
    "I apply QC to autosomes and chrX separately, so separate subdirectories are created for the processing of each set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/aa\n",
    "\n",
    "mkdir autosomes chrX\n",
    "\n",
    "# Autosomes\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile genotypes_b37_dbsnp138 \\\n",
    "    --autosome \\\n",
    "    --make-bed \\\n",
    "    --out autosomes/genotypes_b37_dbsnp138\n",
    "\n",
    "# ChrX (include split PARs)\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile genotypes_b37_dbsnp138 \\\n",
    "    --chr 23,25 \\\n",
    "    --make-bed \\\n",
    "    --out chrX/genotypes_b37_dbsnp138_unmerged\n",
    "\n",
    "# Combine split chrX and PARs\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --bfile chrX/genotypes_b37_dbsnp138_unmerged \\\n",
    "    --merge-x \\\n",
    "    --make-bed \\\n",
    "    --out chrX/genotypes_b37_dbsnp138"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Missing autosome data subject filtering\n",
    "We calculate the proportion of missing genotype calls per chromosome using PLINK to assess whether any subjects have data missing for whole autosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/aa\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name aa_${chr} \\\n",
    "        --script_prefix autosomes/chr${chr}_missing_call_rate \\\n",
    "        --mem 3.8 \\\n",
    "        --priority 0 \\\n",
    "        --program  /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "            --missing \\\n",
    "            --chr $chr \\\n",
    "            --out autosomes/chr${chr}_missing_call_rate\n",
    "done\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    tail -n +2 autosomes/chr${chr}_missing_call_rate.imiss | \\\n",
    "        awk '{ OFS=\"\\t\" } { if($6==1){ print $1,$2 } }' >> autosomes/missing_whole_autosome.remove\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this case none of the subjects had missing autosome data. If subjects ever show up as having missing autosome data then further discussions need to be had on whether these subjects should be removed completely or whether they should only be excluded for the missing chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Clean up \n",
    "rm autosomes/chr*missing_call_rate*\n",
    "rm autosomes/missing_whole_autosome.remove"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove duplicate SNPs\n",
    "If multiple rsIDs are present then the one with the better genotype call rate across subjects should be retained. Obtaining the genotype call rates across subjects would need to be calculated using PLINK --missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/aa\n",
    "\n",
    "# Find duplicate rsIDs\n",
    "cut -f2,2 autosomes/genotypes_b37_dbsnp138.bim | sort | uniq -D > autosomes/variant_duplicates.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this case there are no duplicated rsIDs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Strand orientation conversion\n",
    "To determine if the strands are oriented correctly from the HumanOmni2.5-4v1 chip, I will flip chromosome 1 in 4 different ways. For each different orientation of chromosome 1, I will perform the requisite actions for updating the variant ID until I get to the Allele discordance section. I will consult Bryan Quach on the actions to take based on the results of this test."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chromosome 1 flip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding\n",
    "mkdir Chromosome1\n",
    "cp aa_chr1.* Chromosome1\n",
    "\n",
    "cd Chromosome1/\n",
    "mkdir noflip top_to_plus source_to_plus ilmn_to_plus\n",
    "\n",
    "cp aa* noflip\n",
    "cp aa* top_to_plus\n",
    "cp aa* source_to_plus\n",
    "cp aa* ilmn_to_plus\n",
    "\n",
    "# remove duplicate lines from flip files \n",
    "sort /shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.ilmn_to_plus.flip | uniq >\\\n",
    "/shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.ilmn_to_plus.flip.clean\n",
    "\n",
    "sort /shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.source_to_plus.flip | uniq >\\\n",
    "/shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.source_to_plus.flip.clean\n",
    "\n",
    "sort /shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.top_to_plus.flip | uniq >\\\n",
    "/shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.top_to_plus.flip.clean\n",
    "\n",
    "\n",
    "# flip illumina to plus strand from HumanOmni2.5-4v1 chip\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/ilmn_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile aa_chr1 \\\n",
    "    --flip /shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.ilmn_to_plus.flip.clean \\\n",
    "    --make-bed \\\n",
    "    --out aa_chr1_flipped\n",
    "\n",
    "\n",
    "# flip source to plus strand from HumanOmni2.5-4v1 chip\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/source_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile aa_chr1 \\\n",
    "    --flip /shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.source_to_plus.flip.clean \\\n",
    "    --make-bed \\\n",
    "    --out aa_chr1_flipped\n",
    "\n",
    "\n",
    "# flip top to plus strand from HumanOmni2.5-4v1 chip\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/top_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile aa_chr1 \\\n",
    "    --flip /shared/common/chip_info/HumanOmni2.5-4v1_D/HumanOmni2.5-4v1.dbsnp_b141.b141_id.top_to_plus.flip.clean \\\n",
    "    --make-bed \\\n",
    "    --out aa_chr1_flipped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recoding variants for 1000G phase 3\n",
    "RefSeq IDs (rsIDs) for variants can vary depending on the dbSNP build used and variant IDs do not all use RefSeq nomenclature. To provide a common nomenclature that will make comparisons across data sets feasible, I use a script that recodes all variant names to match 1000G phase 3 variants by position and alleles. The 1000G Phase 3 data I used for STRUCTURE are from /share/nas03/bioinformatics_group/data/ref_panels/1000G/2013.05/plink on MIDAS, but from correspondence with Nathan Gaddis I learned that /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/ also contains 1000G Phase 3 data derived from the May 2013 release. The difference is that it was downloaded from the IMPUTE2 website and reformatted to be directly compatible with IMPUTE2.\n",
    "\n",
    "The data in /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/ is used for variant name recoding, but the 1000G genotype information is acquired from /share/nas03/bioinformatics_group/data/ref_panels/1000G/2013.05/plink."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filtered study data file renaming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "mkdir 1000g_name_recoding\n",
    "\n",
    "ancestry=\"aa\"\n",
    "for ext in {bed,bim,fam}; do\n",
    "    cp  ${ancestry}/genotypes_b37_dbsnp138.${ext} 1000g_name_recoding/${ancestry}_chr_all.${ext}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Variant ID updating\n",
    "Because the 1000G data and the study data have RefSeq IDs (rsIDs) from different dbSNP builds, I standardize them using convert_to_1000g_ids.v4.pl. In the study data set, certain indels may be represented as two variants, a monomorphic variant and an indel with the - symbol for one of the alleles. For example:\n",
    "```\n",
    "1   rs201826967  0.809   57873968   0   G\n",
    "1   rs11284630   0.809   57873969   -   A\n",
    "```\n",
    "\n",
    "These two variants represent a G:GA indel and is coded as such in the 1000 Genomes data. The script to update the names to 1000 Genomes IMPUTE2 format will assign the same ID to these two variants. The duplicated IDs will cause problems for PLINK filtering, so I will remove the variant from a set of duplicate IDs that has the lower genotype call rate. Duplicates may arise for other reasons, and they will be filtered based on the same criterion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/\n",
    "mkdir 1000g_data\n",
    "\n",
    "# Break out data by chr\n",
    "ancestry=\"aa\"\n",
    "for chr in {1..23}; do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}_chr_all \\\n",
    "        --chr ${chr} \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr${chr}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Variant ID updating continued"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "# rename the rs id to the 1000G form\n",
    "# ex. rs 3094315 -->  rs3094315:752566:G:A\n",
    "\n",
    "\n",
    "ancestry=\"aa\"\n",
    "chr=1\n",
    "\n",
    "# illumina to plus\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/ilmn_to_plus\n",
    "\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program perl /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in ${ancestry}_chr${chr}_flipped.bim \\\n",
    "        --file_out ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "\n",
    "\n",
    "# source to plus\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/source_to_plus\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program perl /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in ${ancestry}_chr${chr}_flipped.bim \\\n",
    "        --file_out ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "\n",
    "\n",
    "# top to plus\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program perl /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in ${ancestry}_chr${chr}_flipped.bim \\\n",
    "        --file_out ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "\n",
    "\n",
    "# no flip (note the name distinction for the 'file in' section)\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/noFlip\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program perl /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in ${ancestry}_chr${chr}.bim \\\n",
    "        --file_out ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "\n",
    "# Rename 1000G autosome variant IDs\n",
    "#-------------------------------------------------------------------------------------------------------------------\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/\n",
    "\n",
    "ancestry=\"aa\"\n",
    "chr=1\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name recode_to_1000g_${chr} \\\n",
    "    --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "    --mem 8 \\\n",
    "    --priority 0 \\\n",
    "    --program  perl /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl\\\n",
    "    --file_in /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr}.bim \\\n",
    "    --file_out 1000g_data/chr${chr}_renamed.bim \\\n",
    "    --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "    --file_in_header 0 \\\n",
    "    --file_in_id_col 1 \\\n",
    "    --file_in_chr_col 0 \\\n",
    "    --file_in_pos_col 3 \\\n",
    "    --file_in_a1_col 4 \\\n",
    "    --file_in_a2_col 5 \\\n",
    "    --chr ${chr}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove duplicate variant IDs\n",
    "Many of the functions in PLINK will throw errors when duplicate variant IDs are found.To resolve this, duplicate variant IDs are identified, and only the duplicate variant with the best genotype call rate is retained."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/noFlip\n",
    "\n",
    "ancestry=\"aa\"\n",
    "chr=1\n",
    "# Append _X (where X is a number) to the end of the variant IDs for all but 1st occurrence of duplicates\n",
    "perl -lane 'BEGIN { %idCounts = (); }\n",
    "            if (exists($idCounts{$F[1]})) {\n",
    "                $idCounts{$F[1]}++;\n",
    "                print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "            } else {\n",
    "                $idCounts{$F[1]} = 1;\n",
    "                print;\n",
    "            }' ${ancestry}_chr${chr}_flipped_renamed.bim > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.bim\n",
    "\n",
    "\n",
    "# Generate list of duplicate SNPs\n",
    "grep -P \"_[1-9]\" ${ancestry}_chr${chr}_flipped_renamed_dups.bim |\n",
    "    perl -lane 'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.dupvar_list\n",
    "\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/top_to_plus\n",
    "\n",
    "# Append _X (where X is a number) to the end of the variant IDs for all but 1st occurrence of duplicates\n",
    "perl -lane 'BEGIN { %idCounts = (); }\n",
    "            if (exists($idCounts{$F[1]})) {\n",
    "                $idCounts{$F[1]}++;\n",
    "                print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "            } else {\n",
    "                $idCounts{$F[1]} = 1;\n",
    "                print;\n",
    "            }' ${ancestry}_chr${chr}_flipped_renamed.bim > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.bim\n",
    "\n",
    "# Generate list of duplicate SNPs\n",
    "grep -P \"_[1-9]\" ${ancestry}_chr${chr}_flipped_renamed_dups.bim |\n",
    "    perl -lane 'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.dupvar_list\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/ilmn_to_plus\n",
    "\n",
    "# Append _X (where X is a number) to the end of the variant IDs for all but 1st occurrence of duplicates\n",
    "perl -lane 'BEGIN { %idCounts = (); }\n",
    "            if (exists($idCounts{$F[1]})) {\n",
    "                $idCounts{$F[1]}++;\n",
    "                print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "            } else {\n",
    "                $idCounts{$F[1]} = 1;\n",
    "                print;\n",
    "            }' ${ancestry}_chr${chr}_flipped_renamed.bim > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.bim\n",
    "\n",
    "\n",
    "# Generate list of duplicate SNPs\n",
    "grep -P \"_[1-9]\" ${ancestry}_chr${chr}_flipped_renamed_dups.bim |\n",
    "    perl -lane 'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.dupvar_list\n",
    "\n",
    "\n",
    "\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/source_to_plus\n",
    "\n",
    "# Append _X (where X is a number) to the end of the variant IDs for all but 1st occurrence of duplicates\n",
    "perl -lane 'BEGIN { %idCounts = (); }\n",
    "            if (exists($idCounts{$F[1]})) {\n",
    "                $idCounts{$F[1]}++;\n",
    "                print join(\"\\t\",$F[0],$F[1].\"_\".$idCounts{$F[1]},$F[2],$F[3],$F[4],$F[5]);\n",
    "            } else {\n",
    "                $idCounts{$F[1]} = 1;\n",
    "                print;\n",
    "            }' ${ancestry}_chr${chr}_flipped_renamed.bim > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.bim\n",
    "\n",
    "\n",
    "# Generate list of duplicate SNPs\n",
    "grep -P \"_[1-9]\" ${ancestry}_chr${chr}_flipped_renamed_dups.bim |\n",
    "    perl -lane 'print substr($F[1], 0, index($F[1],\"_\")).\"\\n\".$F[1];' > \\\n",
    "    ${ancestry}_chr${chr}_flipped_renamed_dups.dupvar_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were no duplicate variant IDs found in any of the data. If there had been, the approach to remove the duplicate with the highest missing call rate would have been accomplished with code similar to:\n",
    "\n",
    "```\n",
    "for chr in $(wc -l *.dupvar_list | perl -lane 'if ($F[0] != 0) { $F[1] =~ /chr(\\d+)/; print $1; }'); do\n",
    "    # Get call rates for duplicate SNPs\n",
    "    /share/nas03/bioinformatics_group/software/plink_1.9_beta3.45/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bed ${ancestry}_chr${chr}.bed \\\n",
    "        --bim ${ancestry}_chr${chr}_renamed_dups.bim \\\n",
    "        --fam ${ancestry}_chr${chr}.fam \\\n",
    "        --extract ${ancestry}_chr${chr}_renamed_dups.dupvar_list \\\n",
    "        --missing \\\n",
    "        --out ${ancestry}_chr${chr}_renamed_dups\n",
    "\n",
    "    # Create remove list that contains the duplicate with the higher missing rate\n",
    "    tail -n +2 ${ancestry}_chr${chr}_renamed_dups.lmiss | \\\n",
    "        perl -lane 'BEGIN { %missingness = (); }\n",
    "                    if ($F[1] =~ /^(\\S+)\\_/) {\n",
    "                        $duplicateName = $1\n",
    "                    } else {\n",
    "                        $duplicateName = $F[1].\"_2\";\n",
    "                    }\n",
    "                    if (exists($missingness{$duplicateName})) {\n",
    "                        if ($missingness{$duplicateName} > $F[4]) {\n",
    "                            print $duplicateName;\n",
    "                        } else {\n",
    "                            print $F[1];\n",
    "                        }\n",
    "                    } else {\n",
    "                        $missingness{$F[1]} = $F[4];\n",
    "                    }' \\\n",
    "            > ${ancestry}_chr${chr}_renamed_dups.remove\n",
    "\n",
    "    # Remove duplicates with higher missing rate\n",
    "    /share/nas03/bioinformatics_group/software/plink_1.9_beta3.45/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bed ${ancestry}_chr${chr}.bed \\\n",
    "        --bim ${ancestry}_chr${chr}_renamed_dups.bim \\\n",
    "        --fam ${ancestry}_chr${chr}.fam \\\n",
    "        --exclude ${ancestry}_chr${chr}_renamed_dups.remove \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr${chr}_renamed_dups_removed\n",
    "\n",
    "    # Remove \"_2\" from variant IDs\n",
    "    perl -i.bak -lne 's/\\_2//; print;' ${ancestry}_chr${chr}_renamed_dups_removed.bim\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Allele discordance\n",
    "As a way of standardizing the data, I use 1000G phase 3 data as a reference. The study data may not match the 1000G phase 3 data for several reasons including\n",
    "\n",
    "* Discordant names\n",
    "* Discordant positions\n",
    "* Discordant alleles (swapped major and minor alleles)\n",
    "* Polymorphic alleles not fixed by strand flipping\n",
    "* Discordant allele frequencies\n",
    "\n",
    "I perform checks for these discordances using the 1000G Phase 3 data derived from\n",
    "`data/ref_panels/1000G/2014.10/` derived from the May 2013 release. This data was downloaded from the [IMPUTE2 website](https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.html) and reformatted to be directly compatible with IMPUTE2. These data will be used for ID recoding and MAF calculations.\n",
    "\n",
    "To determine if any study data variants have flipped alleles, I compare the variants to the appropriate 1000 Genomes Phase 3 reference superpopulation. For example, African ancestry is compared to AFR, European ancestry is compared to EUR, and Hispanic ancestry is compared to AMR.\n",
    "\n",
    "Discordance checks are used mainly to\n",
    "1. Determine allele frequency discordances to remove prior to imputation\n",
    "2. Determine if alleles are properly coded in reference to the forward/positive genome strand\n",
    "\n",
    "A lare number of discordant variants indicates potential allele flipping/mismatching issues. If the number of discordant heterozygous variants is less than approximately 10,000, then it is unlikely that there were major issues with flipping or mismatching that indicate coding alleles to the wrong strand.\n",
    "\n",
    "#### Preparing study data\n",
    "The allele discordance scrip requires as input a file of MAFs, so I calculate this using PLINK. In order to use PLINK, I need all three files of the form: .fam, .bed, & .bim. After flipping and renaming chromosome 1, I have a .bim file but I need to generate the other two files as well. I will use PLINK to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/Chromosome1/\n",
    "ancestry=\"aa\"\n",
    "chr=1\n",
    "\n",
    "cd ilmn_to_plus/\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bed ${ancestry}_chr${chr}.bed \\\n",
    "        --bim ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --fam ${ancestry}_chr${chr}.fam \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr${chr}_flipped_renamed_clean\n",
    "\n",
    "\n",
    "cd ../top_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bed ${ancestry}_chr${chr}.bed \\\n",
    "        --bim ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --fam ${ancestry}_chr${chr}.fam \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr${chr}_flipped_renamed_clean\n",
    "\n",
    "\n",
    "cd ../source_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bed ${ancestry}_chr${chr}.bed \\\n",
    "        --bim ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --fam ${ancestry}_chr${chr}.fam \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr${chr}_flipped_renamed_clean\n",
    "\n",
    "\n",
    "cd ../noFlip\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bed ${ancestry}_chr${chr}.bed \\\n",
    "        --bim ${ancestry}_chr${chr}_flipped_renamed.bim \\\n",
    "        --fam ${ancestry}_chr${chr}.fam \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_chr${chr}_flipped_renamed_clean\n",
    "\n",
    "\n",
    "# note that I use chr1_flipped_renamed because (1) I am only using chr1 for this\n",
    "# discordance check and (2) because there were no dups found and therefore this file\n",
    "# will suffice for calculating the freq with.\n",
    "\n",
    "# illumina to plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile ilmn_to_plus/${ancestry}_chr1_flipped_renamed_clean \\\n",
    "    --freq \\\n",
    "    --out ilmn_to_plus/${ancestry}_chr1_flipped_renamed_clean\n",
    "\n",
    "# top_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile top_to_plus/${ancestry}_chr1_flipped_renamed_clean \\\n",
    "    --freq \\\n",
    "    --out top_to_plus/${ancestry}_chr1_flipped_renamed_clean\n",
    "\n",
    "\n",
    "# source_to_plus\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile source_to_plus/${ancestry}_chr1_flipped_renamed_clean \\\n",
    "    --freq \\\n",
    "    --out source_to_plus/${ancestry}_chr1_flipped_renamed_clean\n",
    "\n",
    "\n",
    "# noFlip\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile noFlip/${ancestry}_chr1_flipped_renamed_clean \\\n",
    "    --freq \\\n",
    "    --out noFlip/${ancestry}_chr1_flipped_renamed_clean"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preparing 1000G phase 3 data\n",
    "MAF data are needed for 1000 Genomes data in addition to the study data, so I calculate these for the superpopulation(s) that matches the study group subject ancestries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "/shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding/1000g_data\n",
    "ancestry=\"aa\"\n",
    "\n",
    "# Get list of study variant IDs (chr1, illumina to plus)\n",
    "perl -lane 'if ($F[0] <= 23) { print $F[1]; }' ../Chromosome1/ilmn_to_plus/${ancestry}_chr1_flipped_renamed_clean.bim | \\\n",
    "    sort > ../Chromosome1/ilmn_to_plus/${ancestry}_chr1_sorted_variants.txt\n",
    "\n",
    "# Get list of study variant IDs (chr1, top to plus)\n",
    "perl -lane 'if ($F[0] <= 23) { print $F[1]; }' ../Chromosome1/top_to_plus/${ancestry}_chr1_flipped_renamed_clean.bim | \\\n",
    "    sort > ../Chromosome1/top_to_plus/${ancestry}_chr1_sorted_variants.txt\n",
    "\n",
    "# Get list of study variant IDs (chr1, source to plus)\n",
    "perl -lane 'if ($F[0] <= 23) { print $F[1]; }' ../Chromosome1/source_to_plus/${ancestry}_chr1_flipped_renamed_clean.bim | \\\n",
    "    sort > ../Chromosome1/source_to_plus/${ancestry}_chr1_sorted_variants.txt\n",
    "\n",
    "# Get list of study variant IDs (chr1, no flip)\n",
    "perl -lane 'if ($F[0] <= 23) { print $F[1]; }' ../Chromosome1/noFlip/${ancestry}_chr1_flipped_renamed_clean.bim | \\\n",
    "    sort > ../Chromosome1/noFlip/${ancestry}_chr1_sorted_variants.txt\n",
    "\n",
    "\n",
    "# make directories for each of the 4 flip orientations\n",
    "mkdir ilmn_to_plus top_to_plus source_to_plus noFlip\n",
    "\n",
    "pop=\"AFR\"\n",
    "ancestry=\"aa\"\n",
    "chr=1\n",
    "\n",
    "# Calculate chr1 MAF for 1000G AFR (illumina to plus)\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${pop}_${chr} \\\n",
    "    --script_prefix ilmn_to_plus/${pop}_chr${chr}.maf \\\n",
    "    --mem 6.8 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "        --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "        --chr ${chr} \\\n",
    "        --out ilmn_to_plus/${ancestry}_${pop}_overlap_chr${chr}.maf \\\n",
    "        --extract ../Chromosome1/ilmn_to_plus/${ancestry}_chr1_sorted_variants.txt \\\n",
    "        --keep_groups ${pop}\n",
    "\n",
    "\n",
    "pop=\"AFR\"\n",
    "ancestry=\"aa\"\n",
    "chr=1\n",
    "\n",
    "# Calculate chr1 MAF for 1000G AFR (top to plus)\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${pop}_${chr} \\\n",
    "    --script_prefix top_to_plus/${pop}_chr${chr}.maf \\\n",
    "    --mem 6.8 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "        --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "        --chr ${chr} \\\n",
    "        --out top_to_plus/${ancestry}_${pop}_overlap_chr${chr}.maf \\\n",
    "        --extract ../Chromosome1/top_to_plus/${ancestry}_chr1_sorted_variants.txt \\\n",
    "        --keep_groups ${pop}\n",
    "\n",
    "\n",
    "# Calculate chr1 MAF for 1000G AFR (source to plus)\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${pop}_${chr} \\\n",
    "    --script_prefix source_to_plus/${pop}_chr${chr}.maf \\\n",
    "    --mem 6.8 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "        --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "        --chr ${chr} \\\n",
    "        --out source_to_plus/${ancestry}_${pop}_overlap_chr${chr}.maf \\\n",
    "        --extract ../Chromosome1/source_to_plus/${ancestry}_chr1_sorted_variants.txt \\\n",
    "        --keep_groups ${pop}\n",
    "\n",
    "\n",
    "# Calculate chr1 MAF for 1000G AFR (no flip)\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${pop}_${chr} \\\n",
    "    --script_prefix noFlip/${pop}_chr${chr}.maf \\\n",
    "    --mem 6.8 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "        --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "        --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "        --chr ${chr} \\\n",
    "        --out noFlip/${ancestry}_${pop}_overlap_chr${chr}.maf \\\n",
    "        --extract ../Chromosome1/noFlip/${ancestry}_chr1_sorted_variants.txt \\\n",
    "        --keep_groups ${pop}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Discordance check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/1000g_name_recoding\n",
    "\n",
    "# illumina to plus\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "# Run discordance checks for each ancestry group\n",
    "pop=\"AFR\"\n",
    "ancestry=\"aa\"\n",
    "\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${ancestry}_crosscheck \\\n",
    "    --script_prefix 1000g_data/ilmn_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check \\\n",
    "    --mem 6 \\\n",
    "    --priority 0 \\\n",
    "    --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "        --study_bim_file Chromosome1/ilmn_to_plus/${ancestry}_chr1_flipped_renamed_clean.bim\n",
    "        --study_frq_file Chromosome1/ilmn_to_plus/${ancestry}_chr1_flipped_renamed_clean.frq\n",
    "        --ref_maf_file 1000g_data/ilmn_to_plus/${ancestry}_${pop}_overlap_chr1.maf\n",
    "        --out_prefix 1000g_data/ilmn_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check\"\n",
    "\n",
    "\n",
    "# Print summary of discordances\n",
    "ancestry=\"aa\"\n",
    "    echo \"1000g_data/ilmn_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check.summary\"\n",
    "    cat 1000g_data/ilmn_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check.summary\n",
    "\n",
    "# results\n",
    "'''\n",
    "discordant_positions    0\n",
    "discordant_names        0\n",
    "discordant_alleles      80149\n",
    "discordant_alleles_not_fixed_by_strand_flip     95\n",
    "discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip        0\n",
    "at_cg_snps_freq_diff_gt_0.2     1313\n",
    "'''\n",
    "\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "\n",
    "# top to plus\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "# Run discordance checks for each ancestry group\n",
    "pop=\"AFR\"\n",
    "ancestry=\"aa\"\n",
    "\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${ancestry}_crosscheck \\\n",
    "    --script_prefix 1000g_data/top_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check \\\n",
    "    --mem 6 \\\n",
    "    --priority 0 \\\n",
    "    --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "        --study_bim_file Chromosome1/top_to_plus/${ancestry}_chr1_flipped_renamed_clean.bim\n",
    "        --study_frq_file Chromosome1/top_to_plus/${ancestry}_chr1_flipped_renamed_clean.frq\n",
    "        --ref_maf_file 1000g_data/top_to_plus/${ancestry}_${pop}_overlap_chr1.maf\n",
    "        --out_prefix 1000g_data/top_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check\"\n",
    "\n",
    "\n",
    "# Print summary of discordances\n",
    "ancestry=\"aa\"\n",
    "    echo \"1000g_data/top_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check.summary\"\n",
    "    cat 1000g_data/top_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check.summary\n",
    "\n",
    "'''\n",
    "discordant_positions    0\n",
    "discordant_names        0\n",
    "discordant_alleles      82401\n",
    "discordant_alleles_not_fixed_by_strand_flip     96\n",
    "discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip        0\n",
    "at_cg_snps_freq_diff_gt_0.2     1379\n",
    "'''\n",
    "\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# source to plus\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "# Run discordance checks for each ancestry group\n",
    "pop=\"AFR\"\n",
    "ancestry=\"aa\"\n",
    "\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${ancestry}_crosscheck \\\n",
    "    --script_prefix 1000g_data/source_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check \\\n",
    "    --mem 6 \\\n",
    "    --priority 0 \\\n",
    "    --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "        --study_bim_file Chromosome1/source_to_plus/${ancestry}_chr1_flipped_renamed_clean.bim\n",
    "        --study_frq_file Chromosome1/source_to_plus/${ancestry}_chr1_flipped_renamed_clean.frq\n",
    "        --ref_maf_file 1000g_data/source_to_plus/${ancestry}_${pop}_overlap_chr1.maf\n",
    "        --out_prefix 1000g_data/source_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check\"\n",
    "\n",
    "\n",
    "# Print summary of discordances\n",
    "ancestry=\"aa\"\n",
    "    echo \"1000g_data/source_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check.summary\"\n",
    "    cat 1000g_data/source_to_plus/${ancestry}_chr1_renamed_1000g_discordance_check.summary\n",
    "\n",
    "'''\n",
    "discordant_positions    0\n",
    "discordant_names        0\n",
    "discordant_alleles      79252\n",
    "discordant_alleles_not_fixed_by_strand_flip     97\n",
    "discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip        0\n",
    "at_cg_snps_freq_diff_gt_0.2     1327\n",
    "'''\n",
    "\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "# no flip\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "# Run discordance checks for each ancestry group\n",
    "pop=\"AFR\"\n",
    "ancestry=\"aa\"\n",
    "\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${ancestry}_crosscheck \\\n",
    "    --script_prefix 1000g_data/noFlip/${ancestry}_chr1_renamed_1000g_discordance_check \\\n",
    "    --mem 6 \\\n",
    "    --priority 0 \\\n",
    "    --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "        --study_bim_file Chromosome1/noFlip/${ancestry}_chr1_flipped_renamed_clean.bim\n",
    "        --study_frq_file Chromosome1/noFlip/${ancestry}_chr1_flipped_renamed_clean.frq\n",
    "        --ref_maf_file 1000g_data/noFlip/${ancestry}_${pop}_overlap_chr1.maf\n",
    "        --out_prefix 1000g_data/noFlip/${ancestry}_chr1_renamed_1000g_discordance_check\"\n",
    "\n",
    "\n",
    "# Print summary of discordances\n",
    "ancestry=\"aa\"\n",
    "    echo \"1000g_data/noFlip/${ancestry}_chr1_renamed_1000g_discordance_check.summary\"\n",
    "    cat 1000g_data/noFlip/${ancestry}_chr1_renamed_1000g_discordance_check.summary\n",
    "\n",
    "'''\n",
    "discordant_positions    0\n",
    "discordant_names        0\n",
    "discordant_alleles      80247\n",
    "discordant_alleles_not_fixed_by_strand_flip     96\n",
    "discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip        0\n",
    "at_cg_snps_freq_diff_gt_0.2     1304\n",
    "'''\n",
    "\n",
    "#---------------------------------------------------------------------------------------------------\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ** Because the results of the discordant allele check were ambiguous - the strand flipping did not seem to fix the high number of discordant_alleles for any of the flips - we will move forward with the strand flipping using PLINK.**\n",
    " \n",
    " The issue could be because the genotype data used in this study was merged from 3 phases which used 2 different versions of the OmniHuman2.5 chip - namely version *C* and *D*."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNA strand flipping\n",
    "\n",
    "To determine if strand flipping is an issue, I perform a merge between chr1 of the study data and 1000 Genomes Phase 3 data. Doing this will produce a log of problematic variants that may be attributable to strand flipping. These results can be compared to chr1 data with flipped variants to see if strand orientation is truly the issue.\n",
    "\n",
    "As of 01/10/2018 the binary filesets per chromosome in PLINK format for 1000G Phase 3 can be found in on AWS at:\n",
    "`/shared/data/ref_panels/1000G/2013.05/plink/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/aa\n",
    "\n",
    "# Extract chr1 unflipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "    --chr 1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile chr1_unflipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_unflipped_test\n",
    "\n",
    "# results\n",
    "'''\n",
    "5226909 markers to be merged from\n",
    "/shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1.bim.\n",
    "Of these, 5097895 are new, while 129014 are present in the base dataset.\n",
    "15 more multiple-position warnings: see log file.\n",
    "Error: 63419 variants with 3+ alleles present.\n",
    "'''\n",
    "\n",
    "# Extract chr1 flipped variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile autosomes/genotypes_b37_dbsnp138 \\\n",
    "    --chr 1 \\\n",
    "    --flip chr1_unflipped_test-merge.missnp \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped\n",
    "\n",
    "'''\n",
    "Total genotyping rate is 0.984139.\n",
    "164729 variants and 2559 people pass filters and QC.\n",
    "Note: No phenotypes present.\n",
    "'''\n",
    "\n",
    "# Attempt merge with 1000G chr1 data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile chr1_flipped \\\n",
    "    --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1 \\\n",
    "    --make-bed \\\n",
    "    --out chr1_flipped_test\n",
    "\n",
    "'''\n",
    "5226909 markers to be merged from\n",
    "/shared/data/ref_panels/1000G/2013.05/plink/ALL.chr1.bim.\n",
    "Of these, 5097895 are new, while 129014 are present in the base dataset.\n",
    "15 more multiple-position warnings: see log file.\n",
    "Error: 344 variants with 3+ alleles present.\n",
    "'''\n",
    "rm chr1_*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Between the flipped and unflipped merge, the flipped merge produced drastically fewer errors. As a consequence, I will apply a merge for each chromosome to produced a flip list then flip the variants. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/\n",
    "\n",
    "# Get flip lists\n",
    "ancestry=aa\n",
    "for chr in {1..22}; do\n",
    "    # Extract unflipped variants\n",
    "   /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138 \\\n",
    "        --chr ${chr} \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}/chr${chr}_unflipped\n",
    "\n",
    "    # Attemp merge with 1000G data\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile ${ancestry}/chr${chr}_unflipped \\\n",
    "            --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "            --make-bed \\\n",
    "            --out ${ancestry}/chr${chr}_unflipped_test\n",
    "\n",
    "    # Extract flipped variants\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138 \\\n",
    "            --chr ${chr} \\\n",
    "            --flip ${ancestry}/chr${chr}_unflipped_test-merge.missnp \\\n",
    "            --make-bed \\\n",
    "            --out ${ancestry}/chr${chr}_flipped\n",
    "\n",
    "    # Attempt merge with 1000G data\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile ${ancestry}/chr${chr}_flipped \\\n",
    "            --bmerge /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "            --make-bed \\\n",
    "            --out ${ancestry}/chr${chr}_flipped_test\n",
    "done\n",
    "\n",
    "\n",
    "# Combine flip lists\n",
    "ancestry=aa\n",
    "cat ${ancestry}/chr*_unflipped_test-merge.missnp | sort -u > ${ancestry}/chr_all_unflipped_test-merge.missnp\n",
    "cat ${ancestry}/chr*_flipped_test-merge.missnp | sort -u > ${ancestry}/chr_all_flipped_test-merge.missnp\n",
    "comm -23 ${ancestry}/chr_all_unflipped_test-merge.missnp ${ancestry}/chr_all_flipped_test-merge.missnp \\\n",
    "    > ${ancestry}/autosomes/chr_all.flip\n",
    "\n",
    "# Perform final flip\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138 \\\n",
    "    --flip ${ancestry}/autosomes/chr_all.flip \\\n",
    "    --make-bed \\\n",
    "    --out ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped\n",
    "\n",
    "# Clean up\n",
    "rm ${ancestry}/chr*flipped*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Detecting ancestral outliers with STRUCTURE\n",
    "[STRUCTURE](https://web.stanford.edu/group/pritchardlab/structure.html) is a software tool that can be used to identify admixed individuals, among other uses. By comparing the study subjects with the 1000 Genomes Phase 3 reference panel, I can estimate the composition of an individual's ancestry to determine any discrepancies between self-reporting and genetic information. For the study data, I will be comparing the individuals to 3 different superpopulations from 1000 Genomes Phase 3 reference panel\n",
    "\n",
    "* AFR (African)\n",
    "* EAS (East Asian)\n",
    "* EUR (European)\n",
    "\n",
    "**SNP subset selection**\n",
    "For compuational efficiency 10,000 SNPs are randomly chosen from the intersection of SNPs in the study data with the three 1000 Genomes superpopulations of interest. As of `01/15/2018` the bineary filesets per chromosome in PLINK format for 1000G Phase 3 can be found in `/shared/data/ref_panels/1000G/2013.05/plink/`. Although I am unsure of the dbSNP build for the rsIDs, the discrepancies should not be an issue since in the end we are only using a subset of the SNPs for input into STRUCTURE.\n",
    "\n",
    "**Note:** `/share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/` also contains 1000G Phase 3 data derived from the May 2013 release. The difference is that it was downloaded from the IMPUTE2 website and reformatted to be directly compatible with IMPUTE2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line #\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing/\n",
    "mkdir structure\n",
    "\n",
    "# Get lists of non-A/T and non-C/G SNPs\n",
    "ancestry=\"aa\"\n",
    "perl -lane 'if (($F[4] eq \"A\" && $F[5] ne \"T\") || ($F[4] eq \"T\" && $F[5] ne \"A\") || ($F[4] eq \"C\" && $F[5] ne \"G\") || ($F[4] eq \"G\" && $F[5] ne \"C\")) { print $F[1]; }' \\\n",
    "${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped.bim | \\\n",
    "    sort -u | \\\n",
    "    grep \"rs\" \\\n",
    "    > structure/no_at_cg_snps.txt\n",
    "\n",
    "# Get list of variants from 1000G\n",
    "mkdir structure/1000g_data\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name merge_1000g_snps \\\n",
    "    --script_prefix structure/1000g_data/merge_1000g_snps \\\n",
    "    --mem 3 \\\n",
    "    --priority 0 \\\n",
    "    --program \"cat /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr{1..22}.bim | \\\n",
    "        cut -f2,2 | \\\n",
    "        sort -u | \\\n",
    "        grep \\\"rs\\\" > structure/1000g_data/1000g_phase3_snps.txt\"\n",
    "\n",
    "# Get SNP overlap between study data and 1000G\n",
    "comm -12 structure/no_at_cg_snps.txt structure/1000g_data/1000g_phase3_snps.txt > structure/retirement_1000g_shared_snps.txt\n",
    "\n",
    "# Select 10,000 random SNPs from study and 1000G overlap\n",
    "perl -ne 'print rand().\"\\t\".$_' structure/retirement_1000g_shared_snps.txt | \\\n",
    "    sort -k1,1 | \\\n",
    "    head -10000 | \\\n",
    "    cut -f2,2 \\\n",
    "    > structure/10k_snp_random_sample.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Extract SNP subset PLINK binary filesets for study data**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line #\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "ancestry=aa\n",
    "# Create ped and map files for study genotype data for SNP subset\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile ${ancestry}/autosomes/genotypes_b37_dbsnp138_flipped \\\n",
    "    --extract structure/10k_snp_random_sample.txt \\\n",
    "    --snps-only just-acgt \\\n",
    "    --recode \\\n",
    "    --out structure/${ancestry}_10k_snp_random_sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Extract SNP subset PLINK binary fileets for 1000G data**\n",
    "The 05/2013 release of the 1000 Genomes data have been previously processed and converted to PLINK binary fileset format, but the files include all the 1000G individuals. We are interested in only three superpopulations, so we create filesets specifically for each of these groups. It was brought to my attention that 1000G Phase 3 rsIDs may be duplicated across chromosomes potentially causing chromosome merging issues. For that reason, it is typically recommended to process the chromosomes separately initially, then combine post SNP subsetting. Subject IDs with superpopulation classifications are available at\n",
    "\n",
    "`/shared/data/ref_panels/1000G/igsr_samples.tsv`\n",
    "\n",
    "**Note:** This file was not included in the snapshot for EC2 as of `01/15/2018`. I had to scp the file from MIDAS, which was located at `/share/nas03/bioinformatics_group/data/ref_panels/1000G/igsr_samples.tsv`. The following code includes this procedure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## local machine ##\n",
    "cd /cygdrive/c/Users/jmarks/Desktop/MIDAS\n",
    "\n",
    "# download file from MIDAS\n",
    "scp jmarks@rtplhpc01.rti.ns:/share/nas03/bioinformatics_group/ref_panels/1000G/igsr_samples.tsv .\n",
    "\n",
    "# upload to EC2\n",
    "scp -i ~/.ssh/gwas_rsa igsr_samples.tsv ec2-user@35.169.161.38:/shared/data/ref_panels/1000G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# Ancestry specific directories\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    mkdir structure/1000g_data/${pop}\n",
    "done\n",
    "\n",
    "# Get subject IDs by ancestry\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"African\"){print $1,$1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > structure/1000g_data/AFR/AFR_subject_ids.txt\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"East Asian\"){print $1,$1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > structure/1000g_data/EAS/EAS_subject_ids.txt\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"European\"){print $1,$1} }' /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > structure/1000g_data/EUR/EUR_subject_ids.txt\n",
    "\n",
    "\n",
    "\n",
    "# Make new binary filesets for each 1000G group\n",
    "# only keeping the data from individuals which are from 1 of the three super populations\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${pop}_${chr}_filter \\\n",
    "            --script_prefix structure/1000g_data/${pop}/ancestry_partition_chr${chr} \\\n",
    "            --mem 15.5 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink  \\\n",
    "                --noweb \\\n",
    "                --memory 10000 \\\n",
    "                --bfile /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "                --keep structure/1000g_data/${pop}/${pop}_subject_ids.txt \\\n",
    "                --make-bed \\\n",
    "                --out structure/1000g_data/${pop}/${pop}.chr${chr}\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# Apply SNP subset extraction by chr\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${pop}_${chr}_subsample \\\n",
    "            --script_prefix structure/1000g_data/${pop}/ancestry_partition_chr${chr} \\\n",
    "            --mem 7.6 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "                --noweb \\\n",
    "                --memory 7500 \\\n",
    "                --bfile structure/1000g_data/${pop}/${pop}.chr${chr} \\\n",
    "                --extract structure/10k_snp_random_sample.txt \\\n",
    "                --make-bed \\\n",
    "                --out structure/1000g_data/${pop}/${pop}_chr${chr}_10k_snp_random_sample\n",
    "    done\n",
    "done\n",
    "\n",
    "# Create merge lists and merge autosomes for each 1000G population\n",
    "data_dir=structure/1000g_data\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    echo \"${data_dir}/${pop}/${pop}_chr1_10k_snp_random_sample\" > ${data_dir}/${pop}/${pop}_autosome_merge_list.txt\n",
    "    for chr in {2..22}; do\n",
    "        echo \"${data_dir}/${pop}/${pop}_chr${chr}_10k_snp_random_sample\" \\\n",
    "        >> ${data_dir}/${pop}/${pop}_autosome_merge_list.txt\n",
    "    done\n",
    "done\n",
    "\n",
    "for pop in {AFR,EAS,EUR}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${pop}_merge_plink_filesets \\\n",
    "        --script_prefix structure/1000g_data/${pop}/merge_plink_filesets \\\n",
    "        --mem 4 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 4000 \\\n",
    "            --merge-list structure/1000g_data/${pop}/${pop}_autosome_merge_list.txt \\\n",
    "            --snps-only just-acgt \\\n",
    "            --make-bed \\\n",
    "            --out structure/1000g_data/${pop}/${pop}_all_autosomes_10k_snp_random_sample\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "**Discrepancy assessment between 1000G and study data**\n",
    "\n",
    "As a quality check that the SNP data subsampled from the 1000 Genomes and study data are the same, I will attempt to merge an arbitrarily selected group from each data set using PLINK. If any errors are found, PLINK will generate an error file. Likely causes of errors would be:\n",
    "\n",
    "* SNP genomic coordinates not matching\n",
    "* SNP duplicates found\n",
    "* SNP strand orientation flipped\n",
    "\n",
    "If errors are found the two options for moving forward are\n",
    "\n",
    "1. Re-run the 10,000 SNP subsampling and hope the SNPs chosen do not raise issues\n",
    "2. Remove the problematic SNPs\n",
    "\n",
    "Option #2 is the prefered approach and the one I will be taking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line #\n",
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# Merge AA and AFR genotype files\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/aa_10k_snp_random_sample \\\n",
    "    --bmerge structure/1000g_data/AFR/AFR_all_autosomes_10k_snp_random_sample \\\n",
    "    --recode \\\n",
    "    --out structure/1000g_AFR_retirement_aa_10k_snp_random_sample\n",
    "\n",
    "'''\n",
    "Error: 2916 variants with 3+ alleles present.\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "The attempted merge generated 1 error. In the case of an unsuccessful merge, the following are common:\n",
    "\n",
    "* Multiple positions found for a variant\n",
    "* Non-biallelic variants found\n",
    "\n",
    "When these occur, I exclude variants with multiple positions and see if flipping the non-biallelic variants resolves the second issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/data/studies/phs000428_retirement/genotype/original/processing\n",
    "\n",
    "# Get list of multiple position variants\n",
    "grep \"Multiple positions seen for variant\" structure/1000g_AFR_retirement_aa_10k_snp_random_sample.log | \\\n",
    "    cut -d\"'\" -f2,2 > structure/1000g_AFR_retirement_aa_10k_snp_random_sample.bad_snps.remove\n",
    "\n",
    "# Flip study AA non-biallelic SNPs and remove multi-position variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/aa_10k_snp_random_sample \\\n",
    "    --exclude structure/1000g_AFR_retirement_aa_10k_snp_random_sample.bad_snps.remove \\\n",
    "    --flip structure/1000g_AFR_retirement_aa_10k_snp_random_sample.missnp \\\n",
    "    --recode \\\n",
    "    --out structure/aa_10k_snp_random_sample_retry\n",
    "\n",
    "'''\n",
    "--flip: 2916 SNPs flipped.\n",
    "--exclude: 10000 variants remaining.\n",
    "Using 1 thread.\n",
    "Before main variant filters, 2556 founders and 3 nonfounders present.\n",
    "Calculating allele frequencies... done.\n",
    "Total genotyping rate is 0.988509.\n",
    "10000 variants and 2559 people pass filters and QC.\n",
    "Note: No phenotypes present.\n",
    "'''\n",
    "\n",
    "\n",
    "# Remove multi-position variants from 1000G AFR (there are none so this step is actually uneccessary)\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile structure/1000g_data/AFR/AFR_all_autosomes_10k_snp_random_sample \\\n",
    "    --exclude structure/1000g_AFR_retirement_aa_10k_snp_random_sample.bad_snps.remove \\\n",
    "    --make-bed \\\n",
    "    --out structure/AFR_all_autosomes_10k_snp_random_sample_retry\n",
    "\n",
    "\n",
    "# Retry merge\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --file structure/aa_10k_snp_random_sample_retry \\\n",
    "    --bmerge structure/AFR_all_autosomes_10k_snp_random_sample_retry \\\n",
    "    --make-bed \\\n",
    "    --out structure/1000g_AFR_retirement_aa_10k_snp_random_sample_retry\n",
    "\n",
    "'''Error: 4 variants with 3+ alleles present.'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {
    "height": "812px",
    "left": "0px",
    "right": "1468px",
    "top": "107px",
    "width": "212px"
   },
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
