{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kreek Phasing + Imputation\n",
    "__Author__: Jesse Marks\n",
    "\n",
    "\n",
    "This document logs the steps taken to perform phasing and imputation on the dataset [Kreek]() & [Lung Cancer in Never Smokers](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000634.v1.p1). The starting point for this analysis is after quality control of observed genotypes and after checking for compatibility of population control dbGaP data sets. The quality controlled genotypes are oriented on the GRCh37 plus strand. Based on findings from [Johnson et al.](https://link.springer.com/article/10.1007/s00439-013-1266-7), the intersection set of variants for will be used for imputation. \n",
    "\n",
    "## Software and tools\n",
    "The software and tools used for porcessing these data are\n",
    "* [Michigan Imputation Server](https://imputationserver.sph.umich.edu/index.html) (MIS)\n",
    "* [Amazon Web Services (AWS) - Cloud Computing Services](https://aws.amazon.com/)\n",
    "    * Linux AMI\n",
    "* [PLINK v1.90 beta 4.10](https://www.cog-genomics.org/plink/)\n",
    "* [bgzip](http://www.htslib.org/doc/tabix.html)\n",
    "* [BCF Tools](http://www.htslib.org/doc/bcftools.html)\n",
    "* Windows 10 with [Cygwin](https://cygwin.com/) installed\n",
    "* GNU bash version 4.2.46\n",
    "\n",
    "## Data retrieval and organization\n",
    "The temporary working directory for this analysis will be: `/shared/sandbox/ngc_vidus-lung_cancer_oaall_case_control` <br>\n",
    "on `EC2`. PLINK binary filesets will be obtained from AWS S3 storage.\n",
    "\n",
    "\n",
    "\n",
    "* `/share/nas04/bioinformatics_group/data/studies/kreek`\n",
    "\n",
    "\n",
    "As a constituent of an opioid GWAS meta-analysis in progress referred to as the \"UHS-Nelson\" study, we need to show replication in an independent cohort. We will take advantage of a data set released by Mary-Jeanne Kreek's research group, which we refer to as the \"Kreek\" data set, to assess whether the UHS-Nelson study findings are replicable with the Kreek data. The three types of opioid GWAS we will run are:\n",
    "\n",
    "* Opioid cases vs. clean controls (no abuse or dependence of any drug)\n",
    "* Opioid cases vs. all controls with no covariates adjusting for other drugs\n",
    "* Opioid cases vs. all controls with covariates adjusting for cocaine and alcohol\n",
    "\n",
    "These three GWAS will be run separately for 2 different ancestry groups:\n",
    "\n",
    "* European (EA)\n",
    "* African (AA)\n",
    "\n",
    "**Note**: It was decided that we would not proceed with the the Hispanic ancestry group."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data processing\n",
    "### GRCh37 strand and allele discordance check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /shared/impute/kreek/fou\n",
    "\n",
    "mkdir 1000g\n",
    "ancestry=\"ea\"\n",
    "genoLoc=/shared/impute/kreek/data/genotype/original\n",
    "workingDir=/shared/impute/kreek/fou\n",
    "#for study in {lung_cancer,VIDUS}; do\n",
    "for ancestry in {ea,aa};do\n",
    "    for study in kreek; do\n",
    "        mkdir strand_check\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bfile $genoLoc/${ancestry}_chr_all \\\n",
    "            --freq \\\n",
    "            --out $workingDir/strand_check/${ancestry}_chr_all\n",
    "    done\n",
    "done\n",
    "# Get list of variants from all studies\n",
    "for ancestry in {ea,aa}; do\n",
    "    cat $genoLoc/${ancestry}_chr_all.bim | \\\n",
    "            perl -lane 'if (($F[0]+0) <= 23) { print $F[1]; }' | \\\n",
    "            sort -u > $workingDir/${ancestry}_chr_all_sorted_variants.txt\n",
    "done\n",
    "\n",
    "# Calculate autosome MAFs for 1000G EUR and AFR \n",
    "for ancestry in {ea,aa}; do\n",
    "    if [ $ancestry = \"ea\" ]; then\n",
    "        pop=EUR\n",
    "    else\n",
    "        pop=AFR\n",
    "    fi\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${pop}_${chr} \\\n",
    "            --script_prefix ${pop}_chr${chr}.maf \\\n",
    "            --mem 8 \\\n",
    "            --nslots 1 \\\n",
    "            --priority 0 \\\n",
    "            --program /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "                --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.hap.gz\\\n",
    "                --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz \\\n",
    "                --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "                --chr ${chr} \\\n",
    "                --out 1000g/${pop}_chr${chr}.maf \\\n",
    "                --extract ${ancestry}_chr_all_sorted_variants.txt \\\n",
    "                --keep_groups ${pop}\n",
    "    done\n",
    "done\n",
    "\n",
    "# Merge per chr MAFs for EUR\n",
    "\n",
    "for pop in {EUR,AFR}; do\n",
    "    head -n 1 1000g/${pop}_chr1.maf > 1000g/${pop}_chr_all.maf\n",
    "    tail -q -n +2 1000g/${pop}_chr{1..22}.maf \\\n",
    "        >> 1000g/${pop}_chr_all.maf\n",
    "done\n",
    "\n",
    "# Run discordance checks for EA group\n",
    "\n",
    "for ancestry in {ea,aa}; do\n",
    "    study=kreek\n",
    "\n",
    "    if [ $ancestry = \"ea\" ]; then\n",
    "        pop=EUR\n",
    "    else\n",
    "        pop=AFR\n",
    "    fi\n",
    "\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${ancestry}_${study}_crosscheck \\\n",
    "        --script_prefix ${workingDir}/strand_check/test/${ancestry}_allele_discordance_check \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "            --study_bim_file ${genoLoc}/${ancestry}_chr_all.bim\n",
    "            --study_frq_file ${workingDir}/strand_check/${ancestry}_chr_all.frq\n",
    "            --ref_maf_file 1000g/${pop}_chr_all.maf\n",
    "            --out_prefix ${workingDir}/strand_check/${ancestry}_allele_discordance\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resolving allele discordances\n",
    "The allele discordances will be resolved by\n",
    "* Flipping allele discordances that are fixed by flipping\n",
    "* Removing SNPs with discordant names\n",
    "* Removing SNPs with discordant positions\n",
    "* Removing allele discordances that are not resolved by flipping\n",
    "* Removing alleles with large deviations from the reference population allele frequencies\n",
    "\n",
    "Given that the allele discordance check was done using a union set of SNPs across all studies within an ancestry group, some of the SNPs logged as discordant for a given study may not actually be in the study. Fortunately, if they are not in a given study they will not interfere with the filtering procedures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /shared/impute/kreek/fou\n",
    "mkdir final\n",
    "genoLoc=/shared/impute/kreek/data/genotype/original\n",
    "\n",
    "study=kreek\n",
    "for ancestry in {ea,aa};do\n",
    "    echo -e \"\\n===============\\nProcessing ${study}\\n\"\n",
    "    # Create remove list\n",
    "    echo \"Making remove list\"\n",
    "    cat <(cut -f2,2 strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2) \\\n",
    "        <(cut -f2,2 strand_check/${ancestry}_allele_discordance.at_cg_snps_freq_diff_gt_0.2 | tail -n +2) \\\n",
    "        <(cut -f2,2 strand_check/${ancestry}_allele_discordance.discordant_names | tail -n +2) \\\n",
    "        <(cut -f2,2 strand_check/${ancestry}_allele_discordance.discordant_positions | tail -n +2) \\\n",
    "        <(cut -f2,2 strand_check/${ancestry}_allele_discordance.discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip | tail -n +2) | \\\n",
    "        sort -u > ${ancestry}_snps.remove\n",
    "\n",
    "    # Create flip list\n",
    "    echo \"Making flip list\"\n",
    "    comm -23 <(cut -f2,2 strand_check/${ancestry}_allele_discordance.discordant_alleles | tail -n +2 | sort -u) \\\n",
    "        <(cut -f2,2 strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2 | sort -u) \\\n",
    "        > ${ancestry}_snps.flip\n",
    "\n",
    "    # Apply filters\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile $genoLoc/${ancestry}_chr_all \\\n",
    "        --exclude ${ancestry}_snps.remove \\\n",
    "        --flip ${ancestry}_snps.flip \\\n",
    "        --make-bed \\\n",
    "        --out final/${ancestry}_filtered\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove monomorphic variants\n",
    "Monomorphic variants prevent MIS from accepting the genotype data. In this case, an arbitrarily small MAF is set that is smaller than the lower bound for these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /shared/impute/kreek/fou/final\n",
    "\n",
    "ancestry=\"ea\"\n",
    "study=kreek\n",
    "for ancestry in {ea,aa};do\n",
    "    # Apply filters\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 2048 \\\n",
    "        --bfile ${ancestry}_filtered \\\n",
    "        --maf 0.000001 \\\n",
    "        --make-bed \\\n",
    "        --out ${ancestry}_filtered_mono\n",
    "done\n",
    "\n",
    "\"\"\"\n",
    "482792 aa_filtered.bim\n",
    "457378 aa_filtered_mono.bim\n",
    "\n",
    "490863 ea_filtered.bim\n",
    "434239 ea_filtered_mono.bim\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation preparation for Michigan Imputation Server\n",
    "Visit the [MIS Getting Started Webpage](https://imputationserver.sph.umich.edu/start.html#!pages/help) for more information about the preparing the data for upload to MIS.\n",
    "### VCF File Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /shared/impute/kreek/fou/final\n",
    "\n",
    "mkdir phase_prep\n",
    "\n",
    "# Split by chr and remove any individuals with missing data for whole chr\n",
    "for ancestry in {ea,aa};do\n",
    "    for chr in {1..23}; do\n",
    "        # Remove SNPs\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 4000 \\\n",
    "            --bfile ${ancestry}_filtered_mono \\\n",
    "            --chr ${chr} \\\n",
    "            --mind 0.99 \\\n",
    "            --make-bed \\\n",
    "            --out phase_prep/${ancestry}_chr${chr}_for_phasing \n",
    "    done > chr_splitting.log\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Note__: No subjects were removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /shared/impute/kreek/fou/final/phase_prep\n",
    "mkdir ea aa\n",
    "\n",
    "for ancestry in {ea,aa};do\n",
    "    for chr in {1..22}; do\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 5000 \\\n",
    "            --bfile ${ancestry}_chr${chr}_for_phasing \\\n",
    "            --recode vcf bgz \\\n",
    "            --out ${ancestry}/${ancestry}_chr${chr}_final\n",
    "    done\n",
    "done\n",
    "\n",
    "chr=23\n",
    "for ancestry in {ea,aa};do\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 5000 \\\n",
    "        --bfile ${ancestry}_chr${chr}_for_phasing \\\n",
    "        --output-chr M \\\n",
    "        --set-hh-missing \\\n",
    "        --recode vcf bgz \\\n",
    "        --out ${ancestry}/${ancestry}_chr${chr}_final\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to Michigan Imputation Server (MIS)\n",
    "Transfer the \\*.vcf files to local machine (per chromosome) and then upload to MIS.\n",
    "\n",
    "### Uploading parameters EA\n",
    "These are the parameters that were selected on MIS.\n",
    "\n",
    "__Name__: kreek_ea\n",
    "\n",
    "__Reference Panel__ 1000G Phase 3 v5\n",
    "\n",
    "__Input Files__ File Upload <br>\n",
    "\n",
    "* Select Files - select VCF files that were downloaded to local machine from cloud. <br>\n",
    "\n",
    "__Phasing__: ShapeIT v2.r790 (unphased) \n",
    "\n",
    "__Population__: EUR\n",
    "\n",
    "__Mode__: Quality Control & Imputation\n",
    "\n",
    "* I will not attempt to re-identify or contact research participants.\n",
    "* I will report any inadvertent data release, security breach or other data management incident of which I become aware.\n",
    "\n",
    "\n",
    "### Uploading parameters AA\n",
    "These are the parameters that were selected on MIS.\n",
    "\n",
    "__Name__: chr1-23\n",
    "\n",
    "__Reference Panel__ 1000G Phase 3 v5\n",
    "\n",
    "__Input Files__ File Upload <br>\n",
    "\n",
    "* Select Files - select VCF files that were downloaded to local machine from cloud. <br>\n",
    "\n",
    "__Phasing__: ShapeIT v2.r790 (unphased) \n",
    "\n",
    "__Population__: AFR\n",
    "\n",
    "__Mode__: Quality Control & Imputation\n",
    "\n",
    "* I will not attempt to re-identify or contact research participants.\n",
    "* I will report any inadvertent data release, security breach or other data management incident of which I become aware."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Imputed Data from MIS\n",
    "First Download the data form the Michigan Imputation Server by clicking on the link provided in the email they send out to alert you that your data has finished. Here you will find commands for downloading the data.\n",
    "## EA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "mkdir -p /shared/impute/kreek/data/genotype/imputed/{aa,ea}\n",
    "cd /shared/impute/kreek/data/genotype/imputed/ea\n",
    "\n",
    "# I put all of the download commands in a bash script so I could submit them as jobs.\n",
    "cat mywget.sh\n",
    "\"\"\"\n",
    "wget https://imputationserver.sph.umich.edu/share/results/2e899483c134f1d2c08e3aabc4414253/chr_1.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/ed48286121726c3bc64ddf024ec72730/chr_10.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/1472b06d90055f4effca30c4924fbeca/chr_11.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/b3e1f5e3fd0bfd0a69fe171e438d3c53/chr_12.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/8bc2b433fed05c8c57e09c240aae9d38/chr_13.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/ed5623de7f6b337dec56abce51908cf5/chr_14.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/55b667fafaa4881bba5c285c859b6d6e/chr_15.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/7744e3469a60729f0cfabef804cb7cf/chr_16.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/3097bb774de0115ed454e48628ad06dc/chr_17.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/761df4b95b0214e19d321cd327e5ac4e/chr_18.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/fd2b2ee5136093e255bdf60f7d575bf8/chr_19.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/dffef070d2e0da537b7b0d2b7c42d5e1/chr_2.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/76d3e3a2e8938a080fda54c80cbbe264/chr_20.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/27e8da7066c8dfe52c890eb1952e6dcd/chr_21.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/f1f5a8bbdbce2b2293081f72b7c5ba6c/chr_22.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/cd7e1fe0b57b45590ec83367bf33ffae/chr_3.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/f405efbacb10d9243aaf3d662c6f9414/chr_4.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/ccba882630a8cce30bf458dca68a2d09/chr_5.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/77aeabce5acbca6e225e3ac02965eb0d/chr_6.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/b7798d33eca337160fb28db9d855e6e0/chr_7.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/f95778ef2f6cb1364e5f272e0b0fcf75/chr_8.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/494429e58125aacc323dce6366f348be/chr_9.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/877e7a71f369a30acf5933880b5edbda/chr_X.no.auto_female.zip \n",
    "wget https://imputationserver.sph.umich.edu/share/results/1c9f7d9ac07591172ecb01e76b743fd/chr_X.no.auto_male.zip \n",
    "\"\"\"\n",
    "\n",
    "sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name download_kreek \\\n",
    "    --script_prefix kreek_imputation_download \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash mywget.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inflate imputation results\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EC2 console ###\n",
    "cd /shared/imputed/kreek/data/genotype/imputed/ea\n",
    "\n",
    "# inflate chr results\n",
    "for f in {1..23};do\n",
    "echo '#!/bin/bash' > chr_$f.sh\n",
    "echo '' >> chr_$f.sh\n",
    "echo 'unzip -P \"YHv(tBF1rCx?3b\" chr_'$f'.zip' >> chr_$f.sh\n",
    "done\n",
    "\n",
    "for chr in {1..23}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ea.inflate_chr${chr} \\\n",
    "    --script_prefix ea.chr${chr}_inflation \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash chr_${chr}.sh\n",
    "done\n",
    "\n",
    "rm *.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AA \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "cd /shared/impute/kreek/data/genotype/imputed/aa\n",
    "\n",
    "# I put all of the download commands in a bash script so I could submit them as jobs.\n",
    "cat mywget.sh\n",
    "\"\"\"\n",
    "wget https://imputationserver.sph.umich.edu/share/results/8fce96789a782e1eae8564174708f790/chr_1.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/2818409a432b4f031522994106fb5b00/chr_10.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/eb99a005ea72e5a9dcf1f32e7ad06f89/chr_11.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/775f95824d1ad4ec0a9d9d4f677b2837/chr_12.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/6848af95103b8cf57c84d758afa0240c/chr_13.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/90d53e022cf5f290f9b35cd2536f5a4/chr_14.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/c95be565f4e16408dcf0c2be15fd0166/chr_15.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/33fc07c12dc27a27e99e0108ff6f3efb/chr_16.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/efa12bc6a1e571d43fd0a56bc02f49c5/chr_17.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/9aafd32ba54edba436ba648e9ade8a5e/chr_18.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/bf0cc57426f359bd178c17b8261ae09d/chr_19.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/665ba49c9066e3d68fcc5a3d81cc68e6/chr_2.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/7faaf92de24d36f56fad80c74abe8417/chr_20.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/89db9a7ee57ca03779087290d9328530/chr_21.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/f37ce24b6a80ffa12ff07a09458eab9f/chr_22.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/3be4ba7c31591f3cbcc4a1661e039e9f/chr_3.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/84c9a1580141c54eb726a3dff15204c2/chr_4.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/10e69b2563ce96499b3dcee78e402411/chr_5.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/9252a2a74b2a78b6a51d59810821ce68/chr_6.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/d6a22c97c0b306e3e6019eacf512d583/chr_7.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/534afd90e14af150a2437add4834b8b4/chr_8.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/8c905522864028a306ca36669f244c50/chr_9.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/14d26b9017c84fe2b5e970639dd1a0ab/chr_X.no.auto_female.zip\n",
    "wget https://imputationserver.sph.umich.edu/share/results/57d7fb05f5e2b20df4e5b4972e2618bc/chr_X.no.auto_male.zip\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name aa_kreek.download \\\n",
    "    --script_prefix aa_kreek_imputation_download \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash mywget.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inflate imputation results\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EC2 console ###\n",
    "cd /shared/imputed/kreek/data/genotype/imputed/aa\n",
    "\n",
    "# inflate chr results\n",
    "for f in {1..23};do\n",
    "echo '#!/bin/bash' > chr_$f.sh\n",
    "echo '' >> chr_$f.sh\n",
    "echo 'unzip -P \"T7RPRoezqRvn*4\" chr_'$f'.zip' >> chr_$f.sh\n",
    "done\n",
    "\n",
    "for chr in {1..23}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name aa.inflate_chr${chr} \\\n",
    "    --script_prefix aa.chr${chr}_inflation \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash chr_${chr}.sh\n",
    "done\n",
    "\n",
    "rm *.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
