{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NGC LDSC Regression\n",
    "**Author**: Jesse Marks <br>\n",
    "**GitHub Issue:** [#140](https://github.com/RTIInternational/bioinformatics/issues/140) <br>\n",
    "**Results Location:** `s3://rti-heroin/ldsc/results/20200109_fou_vs_oaall`\n",
    "**Local:** `~/Projects/heroin/ldsc/fou_vs_oaall/`\n",
    "\n",
    "**TLDR**: performing ldsc for fou vs oaall with only UHS to see if correlation is still inverse of what we expect. <br>\n",
    "**Description**: The FOU phenotype for NGC continues to be flipped in the opposite direction from the other NGC phenotypes (OAall and OAexp). See figures in [this GitHub comment](https://github.com/RTIInternational/bioinformatics/issues/140#issuecomment-566154881). We are going to investigate this issue by performing an LDSC regression of OAall vs FOU. This analysis will only have UHS contributing to each phenotype. Furthermore, there will be two separate FOU analyses going into the LDSC: one with old UHS4 and one with new UHS4. Remember that there were issues with the UHS4 data coming from Rutgers where they had to resend some of the data because of corrupt/erroneous samples that had poor GCR quality. So, we have the following analyses that we will be comparing:\n",
    "\n",
    "* FOU_old_uhs4 -- UHS1(897) + UHS2-3(772) + UHS4(861) = 2,530\n",
    "* FOU_new_uhs4 -- UHS1(897) + UHS2-3(772) + UHS4(1,067) = 2,736\n",
    "* OAall -- UHS1(9,245)\n",
    "\n",
    "We are going to utilize the [LD score regression pipeline](https://github.com/RTIInternational/ld-regression-pipeline) that Alex Waldrop developed to perform LD score regression. \n",
    "\n",
    "## Data\n",
    "NGC summary stats results location:\n",
    "* FOU with new UHS4: s3://rti-midas-data/studies/ngc/meta/095/processing/fou/uhs1-4.ea.fou.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "* FOU with old UHS4: s3://rti-midas-data/studies/ngc/meta/096/processing/fou/uhs1-4.ea.fou.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "\n",
    "OAall:\n",
    "* UHS1: s3://rti-midas-data/studies/ngc/uhs1/association_tests/002/ea/oaall/uhs1.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz\n",
    "* CATS-MOLE: s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* CATS-PERHUNT: s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* COGA: s3://rti-midas-data/studies/ngc/coga/association_tests/001/ea/oaall/coga.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_coga.se.rsq_gt_0.3.gz\n",
    "* Kreek: s3://rti-midas-data/studies/ngc/kreek/association_tests/003/ea/oaall/kreek.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_kreek.rsq_gt_0.3.gz\n",
    "* Bulgaria: s3://rti-heroin/gwas/op_dep_bulgaria/results/oaall/0001/mis_minimac4_eagle2.4/1000g_p3/eur/chr{1..22}.1000g_ids.maf_gt_0.01_eur_obd.rsq_gt_0.3.gz\n",
    "* VIDUS: s3://rti-midas-data/studies/ngc/vidus/association_tests/002/ea/oaall/vidus.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz\n",
    "* Yale-Penn-CIDR: s3://rti-midas-data/studies/ngc/yale-penn/association_tests/004/ea/oaall/yale-penn.cidr.ea.chr{1..22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "* Yale-Penn-GO: s3://rti-midas-data/studies/ngc/yale-penn/association_tests/005/ea/oaall/yale-penn.go.ea.chr{1..22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "* deCODE OAall (N=275468â€¬): s3://rti-midas-data/studies/ngc/decode/association_tests/001/ea/oaall/decode.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz\n",
    "* deCODE OAexp (N=2581): s3://rti-midas-data/studies/ngc/decode/association_tests/002/ea/oaexp/decode.ea.oaexp.chr{1..22}.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz\n",
    "\n",
    "FOU: \n",
    "* alive: s3://rti-midas-data/studies/ngc/alive/association_tests/001/ea/fou/alivess.ea.fou.chr{1..22}.1000g_ids.maf_gt_0.01_eur_alive.se.rsq_gt_0.3.gz\n",
    "* cats: s3://rti-midas-data/studies/ngc/cats/association_tests/012/ea/fou/cats.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* cogend: s3://rti-midas-data/studies/ngc/cogend/association_tests/001/ea/fou/cogend.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_cogend.rsq_gt_0.3.gz\n",
    "* start: s3://rti-midas-data/studies/ngc/start/association_tests/001/ea/fou/start.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_start.se.rsq_gt_0.3.gz\n",
    "* uhs1: s3://rti-midas-data/studies/ngc/uhs1/association_tests/001/ea/fou/uhs1.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz\n",
    "* uhs2-3: s3://rti-midas-data/studies/ngc/uhs2-3/association_tests/001/ea/fou/uhs2-3.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.03_eur_uhs2-3.se.rsq_gt_0.3.gz\n",
    "* uhs4: s3://rti-shared/gwas/uhs4/results/box_cox_totopioid_tot_30d/0001/mis_minimac4_eagle2.4/1000g_p3/eur/chr{1.22}.1000g_ids.maf_gt_0.01_eur_uhs4.rsq_gt_0.3.gz\n",
    "* vidus: s3://rti-shared/gwas/vidus/results/box_cox_useropioid6mfq/0001/mis_minimac3_shapeit2/1000g_p3/eur/chr{1.22}.1000g_ids.maf_gt_0.01_eur_vidus.rsq_gt_0.3.gz\n",
    "* yale-penn: s3://rti-midas-data/studies/ngc/yale-penn/association_tests/001/ea/fou/yale-penn.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "\n",
    "**Note**: The effect allele for the FOU analyses is the same as for the other NGC metas (A2). The effect allele for the OAall analysis is ALT.\n",
    "\n",
    "<br>\n",
    "\n",
    "**sample sizes**\n",
    "* OAall_UHS1: 9,245\n",
    "* FOU_old_UHS4: 2,530\n",
    "* FOU_new_UHS4: 2,736\n",
    "\n",
    "OAall:\n",
    "* Bulgaria (N=2,765)\n",
    "* CATS-MOLE (N=3,162)\n",
    "* CATS-PERHUNT (N=1,920)\n",
    "* COGA (N=7,631)\n",
    "* deCODE OAall (N=275,468)\n",
    "* deCODE OAexp (N=2,581)\n",
    "* Kreek (N=556)\n",
    "* UHS1 (N=9,245)\n",
    "* VIDUS (N=2,177)\n",
    "* Yale-Penn-CIDR (N=666)\n",
    "* Yale-Penn-GO (N=917)\n",
    "\n",
    "FOU:\n",
    "* alive (N=152)\n",
    "* cats (N=1226)\n",
    "* cogend (N=99)\n",
    "* start (N=231)\n",
    "* uhs1 (N=897)\n",
    "* uhs2-3 (N=772)\n",
    "* uhs4 (N=1067)\n",
    "* vidus (N=300)\n",
    "* yale-penn (N=850)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data wrangling\n",
    "Format the summary stats for input into cromwell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "## FOU meta 087\n",
    "for chr in {1..22}; do\n",
    "   aws s3 cp s3://rti-midas-data/studies/ngc/meta/087/processing/fou/alive+cats+cogend+start+uhs1-4+vidus+yale-penn.ea.fou.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=fou_meta087_n5833.txt\n",
    "zcat alive+cats+cogend+start+uhs1-4+vidus+yale-penn.ea.fou.chr22.maf_gt_0.01.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$6,$8; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=alive+cats+cogend+start+uhs1-4+vidus+yale-penn.ea.fou.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "## upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## OAall meta  089\n",
    "for chr in {1..22}; do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/089/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=oaall_meta089_n304507.txt\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr22.maf_gt_0.01.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$6,$8; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "## upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_oaall/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## FOU_old_UHS4\n",
    "cd /shared/jmarks/heroin/ldsc/fou_oaall/001/processing\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/096/processing/fou/uhs1-4.ea.fou.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz . --quiet &\n",
    "done\n",
    "\n",
    "outf=fou_meta096_n2530.txt\n",
    "for chr in {1..22};do\n",
    "    inf=uhs1-4.ea.fou.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "## upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "    \n",
    "    \n",
    "########################################################################################################################################################################## \n",
    "## FOU_new_UHS4\n",
    "cd /shared/jmarks/heroin/ldsc/fou_oaall/001/processing\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/095/processing/fou/uhs1-4.ea.fou.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz . --quiet &\n",
    "done\n",
    "\n",
    "outf=fou_meta095_n2736.txt\n",
    "for chr in {1..22};do\n",
    "    inf=uhs1-4.ea.fou.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "## upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/opioid_fou/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## OAall_decode_subset\n",
    "cd /shared/jmarks/heroin/ldsc/fou_oaall/002/processing\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/094/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz . --quiet &\n",
    "done\n",
    "\n",
    "outf=oaall_meta094_n31620.txt\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr22.maf_gt_0.01.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$6,$8; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "## upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_oaall/$outf.gz\n",
    "\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## CATS-MOLE (3162)\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "\n",
    "outf=catsmole_oaall010_n3162.txt\n",
    "zcat catsmole.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=catsmole.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## CATS-PERHUNT \n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=catsperthunt_oaall011_n1920.txt\n",
    "zcat catsperthunt.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=catsperthunt.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## COGA\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/coga/association_tests/001/ea/oaall/coga.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_coga.se.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=coga_oaall001_n7631.txt\n",
    "zcat coga.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_coga.se.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=coga.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_coga.se.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done  &\n",
    "\n",
    "gzip $outf &\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## Kreek\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/kreek/association_tests/003/ea/oaall/kreek.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_kreek.rsq_gt_0.3.gz . --quiet \n",
    "done &\n",
    "\n",
    "outf=kreek_oaall003_n556.txt\n",
    "zcat kreek.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_kreek.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=kreek.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_kreek.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## Bulgaria\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-heroin/gwas/op_dep_bulgaria/results/oaall/0001/mis_minimac4_eagle2.4/1000g_p3/eur/chr$chr.1000g_ids.maf_gt_0.01_eur_obd.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=bulgaria_oaall0001_n2765.txt\n",
    "zcat chr22.1000g_ids.maf_gt_0.01_eur_obd.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=chr$chr.1000g_ids.maf_gt_0.01_eur_obd.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done  &\n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## VIDUS\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/vidus/association_tests/002/ea/oaall/vidus.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=vidus_oaall002_n2177.txt\n",
    "zcat vidus.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=vidus.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value , N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf &\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## Yale-Penn-CIDR\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/yale-penn/association_tests/004/ea/oaall/yale-penn.cidr.ea.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=yalepenn_cidr_oaall004_n666.txt\n",
    "zcat yale-penn.cidr.ea.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$5,$4,$16,$20, $10; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=yale-penn.cidr.ea.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, n\n",
    "    awk '{print $1,$2,$3,$5,$4,$16,$20, $10}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## Yale-Penn-GO\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/yale-penn/association_tests/005/ea/oaall/yale-penn.go.ea.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=yalepenn_go_oaall005_n917.txt\n",
    "zcat yale-penn.go.ea.chr22.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$5,$4,$16,$20, $10; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=yale-penn.go.ea.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, n\n",
    "    awk '{print $1,$2,$3,$5,$4,$16,$20, $10}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## UHS1\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/uhs1/association_tests/002/ea/oaall/uhs1.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "\n",
    "outf=uhs1_oaall002_n9245.txt\n",
    "zcat uhs1.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=uhs1.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## deCODE OAall (N=275,468â€¬) \n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/decode/association_tests/001/ea/oaall/decode.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=decode_oaall001_n275468.txt\n",
    "zcat decode.ea.oaall.chr22.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$5,$4,$8,$9; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=decode.ea.oaall.chr$chr.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$5,$4,$8,$9}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## deCODE OAexp (N=2,581)\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/decode/association_tests/002/ea/oaexp/decode.ea.oaexp.chr$chr.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=decode_oaexp002_n2581.txt\n",
    "zcat decode.ea.oaexp.chr22.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$5,$4,$8,$9; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=decode.ea.oaexp.chr$chr.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$5,$4,$8,$9}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/$outf.gz\n",
    "\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## OAall (N=302,330) no VIDUS\n",
    "\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/098/processing/oaall/cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=oaall_meta098_n302330.txt\n",
    "zcat cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr22.maf_gt_0.01.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$6,$8; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2 ) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_oaall/$outf.gz\n",
    "\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## OAall_099 (N=29,443) no VIDUS and deCODE OAexp\n",
    "\n",
    "\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/099/processing/oaall/cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz .\n",
    "done\n",
    "\n",
    "outf=oaall_meta099_n29443.txt\n",
    "zcat cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr22.maf_gt_0.01.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$6,$8; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6,$8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_oaall/$outf.gz\n",
    "\n",
    "\n",
    "########################################################################################################################################################################## \n",
    "## CATS FOU (N=1226)\n",
    "\n",
    "for chr in {1..22};\n",
    "    do aws s3 cp s3://rti-midas-data/studies/ngc/cats/association_tests/012/ea/fou/cats.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz .\n",
    "done \n",
    "\n",
    "\n",
    "outf=cats_fou012_n1226.txt\n",
    "zcat cats.ea.fou.chr22.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18, $6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18, $6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done \n",
    "\n",
    "gzip $outf\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_oaall/$outf.gz\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "## OAall meta105\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/meta/105/processing/oaall/cats+kreek+odb+uhs+vidus.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "\n",
    "outf=oaall_meta105_n18825.txt\n",
    "zcat cats+kreek+odb+uhs+vidus.ea.chr2.maf_gt_0.01.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$6, $8; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats+kreek+odb+uhs+vidus.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value\n",
    "    awk '{print $1,$2,$3,$4,$5,$6, $8}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf &\n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_oaall/$outf.gz\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "## FOU alive\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/alive/association_tests/001/ea/fou/alivess.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_alive.se.rsq_gt_0.3.gz . --quiet \n",
    "done \n",
    "\n",
    "outf=alive_fou001_n152.txt\n",
    "zcat alivess.ea.fou.chr7.1000g_ids.maf_gt_0.01_eur_alive.se.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=alivess.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_alive.se.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "## FOU cats\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/cats/association_tests/012/ea/fou/cats.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "\n",
    "outf=cats_fou001_n152.txt\n",
    "zcat cats.ea.fou.chr2.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cats.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "####################################################################################################\n",
    "## FOU cogend\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/cogend/association_tests/001/ea/fou/cogend.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_cogend.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "\n",
    "outf=cogend_fou001_n99.txt\n",
    "zcat cogend.ea.fou.chr1.1000g_ids.maf_gt_0.01_eur_cogend.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=cogend.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_cogend.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "####################################################################################################\n",
    "## FOU start\n",
    "\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/start/association_tests/001/ea/fou/start.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_start.se.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "\n",
    "outf=start_fou001_n231.txt\n",
    "zcat start.ea.fou.chr1.1000g_ids.maf_gt_0.01_eur_start.se.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=start.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_start.se.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "####################################################################################################\n",
    "## FOU uhs1\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/uhs1/association_tests/001/ea/fou/uhs1.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "\n",
    "outf=uhs1_fou001_n897.txt\n",
    "zcat uhs1.ea.fou.chr1.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=uhs1.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "####################################################################################################\n",
    "## FOU uhs2-3\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/uhs2-3/association_tests/001/ea/fou/uhs2-3.ea.fou.chr$chr.1000g_ids.maf_gt_0.03_eur_uhs2-3.se.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=uhs2_3_fou001_n772.txt\n",
    "zcat uhs2-3.ea.fou.chr3.1000g_ids.maf_gt_0.03_eur_uhs2-3.se.rsq_gt_0.3.gz   |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=uhs2-3.ea.fou.chr$chr.1000g_ids.maf_gt_0.03_eur_uhs2-3.se.rsq_gt_0.3.gz                                                                \n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "####################################################################################################\n",
    "## FOU uhs4\n",
    "        \n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-shared/gwas/uhs4/results/box_cox_totopioid_tot_30d/0001/mis_minimac4_eagle2.4/1000g_p3/eur/chr$chr.1000g_ids.maf_gt_0.01_eur_uhs4.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=uhs4_fou0001_n1067.txt\n",
    "zcat chr7.1000g_ids.maf_gt_0.01_eur_uhs4.rsq_gt_0.3.gz  |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=chr$chr.1000g_ids.maf_gt_0.01_eur_uhs4.rsq_gt_0.3.gz                                                                \n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "####################################################################################################\n",
    "## FOU vidus\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-shared/gwas/vidus/results/box_cox_useropioid6mfq/0001/mis_minimac3_shapeit2/1000g_p3/eur/chr$chr.1000g_ids.maf_gt_0.01_eur_vidus.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=vidus_fou0001_n300.txt\n",
    "zcat  chr1.1000g_ids.maf_gt_0.01_eur_vidus.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$17,$18,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=chr$chr.1000g_ids.maf_gt_0.01_eur_vidus.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$17,$18,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "####################################################################################################\n",
    "## FOU yale-penn\n",
    "\n",
    "for chr in {1..22};do\n",
    "    aws s3 cp s3://rti-midas-data/studies/ngc/yale-penn/association_tests/001/ea/fou/yale-penn.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz . --quiet\n",
    "done &\n",
    "\n",
    "outf=yalepenn_fou001_n850.txt\n",
    "zcat  yale-penn.ea.fou.chr1.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=yale-penn.ea.fou.chr$chr.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n",
    "\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "## FOU UHS2-4\n",
    "for chr in {1..22}; do\n",
    "    aws s3 cp s3://rti-heroin/gwas/uhs234/results/eur/stats/final/uhs234.eur.1000G_p3.fou.chr$chr.rsq.maf_study_eur.txt.gz .\n",
    "done \n",
    "\n",
    "\n",
    "outf=uhs234_fou_n1562.txt\n",
    "zcat  uhs234.eur.1000G_p3.fou.chr1.rsq.maf_study_eur.txt.gz |\\\n",
    "    awk 'NR==1{print $1,$2,$3,$4,$5,$16,$17,$6; exit}' > $outf\n",
    "for chr in {1..22};do\n",
    "    inf=uhs234.eur.1000G_p3.fou.chr$chr.rsq.maf_study_eur.txt.gz\n",
    "    # MarkerName, CHR, POS, Allele1, Allele2, Effect, P-value, N\n",
    "    awk '{print $1,$2,$3,$4,$5,$16,$17,$6}' <(zcat $inf | tail -n +2) >> $outf\n",
    "done &\n",
    "\n",
    "\n",
    "gzip $outf \n",
    "# upload to S3\n",
    "aws s3 cp $outf.gz s3://rti-heroin/ldsc/data/opioid_fou/$outf.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create WorkFlow inputs\n",
    "Here is an example entry in the Excel Phenotype File:\n",
    "\n",
    "**trait\tplot_label\tsumstats_path\tpmid\tcategory\tsample_size\tid_col\tchr_col\tpos_col\teffect_allele_col\tref_allele_col\teffect_col\tpvalue_col\tsample_size_col\teffect_type\tw_ld_chr**\n",
    "```\n",
    "COPDGWAS Hobbs et al.\tCOPD\ts3://rti-nd/LDSC/COPDGWAS_HobbsEtAl/modGcNoOtherMinMissSorted.withchrpos.txt.gz\t28166215\tRespiratory\t51772\t3\t1\t2\t4\t5\t10\t12\t\tbeta\ts3://clustername--files/eur_w_ld_chr.tar.bz2\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "#screen\n",
    "#qrsh\n",
    "\n",
    "#phenD=20200109_heroin_ldsc_phenotypes_local.xlsx\n",
    "#procD=/shared/jmarks/heroin/ldsc/fou_oaall/001\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "#mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "## upload phenotype file (Excel) to */workflow_inputs/\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "# edit this file\n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phenD >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Analysis Workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above ld-regression-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## copy cromwell config file from S3 to EC2 instance\n",
    "cd /shared/jmarks/bin/cromwell\n",
    "#aws s3 cp s3://rti-cromwell-output/cromwell-config/cromwell_default_genomics_queue.conf .\n",
    "\n",
    "## Run workflowâ€”Navigate to cromwell directory\n",
    "java -Dconfig.file=/shared/jmarks/bin/cromwell/cromwell_default_genomics_queue.conf \\\n",
    "    -jar cromwell-44.jar \\\n",
    "    run $procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl \\\n",
    "    -i $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json \\\n",
    "    -p $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Record the workflow log-ID. Then get the results on s3 at `s3:///rti-cromwell-output/cromwell-execution/full_ld_regression_wf/<log-ID>/` <br>\n",
    "You can find the log-ID in the directory `/shared/jmarks/bin/cromwell/cromwell-workflow-logs/` (for example).\n",
    "<br>\n",
    "<br>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"png\")\n",
    "#setwd(\"C:/Users/jmarks/OneDrive - Research Triangle Institute/Projects/heroin/ldsc/ngc_all/fou/001/processing/output/cromwell\")\n",
    "\n",
    "#orig <- readPNG(\"ngc_fou.ld_regression_results-1.png\")\n",
    "#zoom <- readPNG(\"20191213_heroin_ngc_ldsc_fou_rg_results-1.png\")\n",
    "\n",
    "grid::grid.raster(orig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run Analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 001\n",
    "FOU_old_uhs4 -- UHS1(897) + UHS2-3(772) + UHS4(861) = 2,530 <br>\n",
    "FOU_new_uhs4 -- UHS1(897) + UHS2-3(772) + UHS4(1,067) = 2,736 <br>\n",
    "\\*OAall -- UHS1(9,245) <br>\n",
    "\\* reference trait\n",
    "\n",
    "`df0c2159-798c-4b5b-a0d5-40b1defaee9a`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200109_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/001"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 002\n",
    "This analysis will be FOU_new_UHS4 (N=2,736) compared to:\n",
    "    * OAall_094 (deCODE subset of OAexp, N=31,620)\n",
    "    * OAall_089 (all, N=304,507)\n",
    "\n",
    "\n",
    "`44a93a8c-2047-4896-88ce-16342857670a`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200110_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/002"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 003\n",
    "Complete FOU_087(N=5,388) compared to:\n",
    "  * OAall_UHS1 GWAS results (N=9,245)\n",
    "\n",
    "\n",
    "`0b3a6bd5-7d63-47fc-a70e-ae93a9894bf6`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200110_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/003"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 004\n",
    "Complete FOU_087(N=5,388) compared to OAall GWAS results:\n",
    "  * CATS-MOLE (N=3,162)\n",
    "  * CATS-PERHUNT (N=1,920)\n",
    "  * COGA (N=7,631)\n",
    "  * Kreek (N=556)\n",
    "  * Bulgaria (N=2,765)\n",
    "  * VIDUS (N=2,177)\n",
    "  * Yale-Penn-CIDR (N=666)\n",
    "  * Yale-Penn-GO (N=917)\n",
    "  * deCODE OAall (N=275,468)\n",
    "  * deCODE OAexp (N=2,581)\n",
    "\n",
    "`dec6127d-106d-45c1-93e3-dc6df5794bb1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200121_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/004"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 005\n",
    "Complete FOU_087(N=5,388) compared to OAall GWAS results:\n",
    "  * CATS-MOLE (N=3,162)\n",
    "  * CATS-PERHUNT (N=1,920)\n",
    "  * COGA (N=7,631)\n",
    "  * Kreek (N=556)\n",
    "  * Bulgaria (N=2,765)\n",
    "  * VIDUS (N=2,177)\n",
    "  * Yale-Penn-CIDR (N=666)\n",
    "  * Yale-Penn-GO (N=917)\n",
    "  * deCODE OAall (N=275,468)\n",
    "  * deCODE OAexp (N=2,581)\n",
    "  * UHS1 (N=9,245)\n",
    "\n",
    "\n",
    "`d068dcb8-aa95-4c34-b284-b5110c1cd10e`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200122_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/005"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 006 \n",
    "(no VIDUS)\n",
    "EUR metas without VIDUS. For OAall, there were 2177 VIDUS individuals removed. For FOU, there were 300 VIDUS individuals removed. You can find the results at the following locations in S3:\n",
    "\n",
    "FOU: s3://rti-midas-data/studies/ngc/meta/097/processing/fou/alive+cats+cogend+start+uhs1-4+yale-penn.ea.fou.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "OAall (deCODE OAall): s3://rti-midas-data/studies/ngc/meta/098/processing/oaall/cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "OAall (deCODE OAexp): s3://rti-midas-data/studies/ngc/meta/099/processing/oaall/cats+coga+decode+kreek+odb+uhs+yale-penn.ea.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.gz\n",
    "\n",
    "`59304359-aa17-42ca-b341-e11aec16a98e`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200127_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/006"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 007\n",
    "VIDUS_FOU vs VIDUS_OAall\n",
    "\n",
    "`99c0b4c7-cf16-46e9-8e9e-f73552dc82e1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200130_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/007"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 008\n",
    "FOU UHS1â€“4 (with old UHS4) compared to UHS1_OAall\n",
    "\n",
    "`c5aafd20-ec79-4be5-be1c-501d0136ca03`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200131_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/008"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 009\n",
    "FOU_092 (N=4,527) compared to\n",
    "* FOU_087 (N=5388)\n",
    "* OAall_UHS1 GWAS results (N=9,245)\n",
    "* COGA (N=7,631)\n",
    "\n",
    "`3e0994ff-3628-4cb9-9d94-946a83ce4300`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200131_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/009"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 010\n",
    "FOU_092 (N=4,527) compared to\n",
    "* FOU_087 (N=5388)\n",
    "* OAall_UHS1 GWAS results (N=9,245)\n",
    "* COGA (N=7,631)\n",
    "\n",
    "`fd1dacd6-5318-4b8f-82cd-7a7fe5d700c6`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200131_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/010"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 011\n",
    "`CATS_FOU (N=1226) vs CATSPERTHUNT_OAall (N=1920) & CATSMOLE_OAall (N=3162)`\n",
    "\n",
    "FOU location:\n",
    "```\n",
    "s3://rti-midas-data/studies/ngc/cats/association_tests/012/ea/fou/cats.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "```\n",
    "OAall location:\n",
    "```\n",
    "CATS-MOLE: s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "\n",
    "CATS-PERHUNT: s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "```\n",
    "\n",
    "\n",
    "`6cdb291a-c900-4963-b2ae-24803e9090ea`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200211_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/011"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 012\n",
    "Same as 011 but changing the reference trait vs compared to traits. Testing to see if order matters in the LDSC regression pipeline. <br>\n",
    "`CATSMOLE_OAall (N=3162) vs CATS_FOU (N=1226) & CATSPERTHUNT_OAall (N=1920) `\n",
    "\n",
    "FOU location:\n",
    "```\n",
    "s3://rti-midas-data/studies/ngc/cats/association_tests/012/ea/fou/cats.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "```\n",
    "OAall location:\n",
    "```\n",
    "CATS-MOLE: s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "\n",
    "CATS-PERHUNT: s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "```\n",
    "\n",
    "`fd9bdc16-346c-4bba-bdf4-3e277d69380c`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200211_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/012\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 013\n",
    "Same as 011 & 012 but changing the reference trait vs compared to traits. Testing to see if order matters in the LDSC regression pipeline. <br>\n",
    "CATSPERTHUNT_OAall (N=1920) vs CATS_FOU (N=1226) & CATSMOLE_OAall (N=3162)\n",
    "\n",
    "FOU location:\n",
    "```\n",
    "s3://rti-midas-data/studies/ngc/cats/association_tests/012/ea/fou/cats.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "```\n",
    "OAall location:\n",
    "```\n",
    "CATS-MOLE: s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "\n",
    "CATS-PERHUNT: s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "```\n",
    "\n",
    "\n",
    "`d684a632-d265-4403-9855-aefefe279b7c`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200211_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/013\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 014\n",
    "Yale-Penn FOU (N=850) vs Yale-Penn-CIDR (N=666) & Yale-Penn-GO (N=917)\n",
    "\n",
    "FOU location\n",
    "```\n",
    "s3://rti-midas-data/studies/ngc/yale-penn/association_tests/001/ea/fou/yale-penn.ea.fou.chr{1.22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "```\n",
    "\n",
    "OAall location:\n",
    "```\n",
    "Yale-Penn-CIDR: s3://rti-midas-data/studies/ngc/yale-penn/association_tests/004/ea/oaall/yale-penn.cidr.ea.chr{1..22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "\n",
    "Yale-Penn-GO: s3://rti-midas-data/studies/ngc/yale-penn/association_tests/005/ea/oaall/yale-penn.go.ea.chr{1..22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "```\n",
    "\n",
    "\n",
    "`c1da79b8-567f-4e6a-ad92-77bbc766150f`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200211_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/014\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 015\n",
    "Complete FOU_087(N=5,388) compared to OAall GWAS results:\n",
    "\n",
    "* Bulgaria (N=2,765)\n",
    "* CATS-MOLE (N=3,162)\n",
    "* CATS-PERHUNT (N=1,920)\n",
    "* COGA (N=7,631)\n",
    "* deCODE OAall (N=275,468)\n",
    "* deCODE OAexp (N=2,581)\n",
    "* Kreek (N=556)\n",
    "* UHS1 (N=9,245)\n",
    "* VIDUS (N=2,177)\n",
    "* Yale-Penn-CIDR (N=666)\n",
    "* Yale-Penn-GO (N=917)\n",
    "\n",
    "`5e5449fb-1242-410c-b26d-1a145eff51e2`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200211_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/015\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 016\n",
    "`FOU meta087 vs Yale-Penn_GO`\n",
    "try to get Yale-Penn_GO (N=917) to run\n",
    "\n",
    "`638d5ae7-7c30-4661-91fc-ac64f746cf1a`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200211_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/016"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 017\n",
    "`FOU meta106 (N=4592) vs OAall meta105 (N=19,825)`  <br>\n",
    "Excluding cohorts that ran LMM. \n",
    "\n",
    "`165b5b52-f932-4046-8da5-97bacf508753`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200217_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/017"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 018\n",
    "FOU meta106 (N=4592) vs OAall GWAS cohorts:\n",
    "* CATS-MOLE: s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* CATS-PERHUNT: s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* Kreek: s3://rti-midas-data/studies/ngc/kreek/association_tests/003/ea/oaall/kreek.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_kreek.rsq_gt_0.3.gz\n",
    "* Bulgaria: s3://rti-heroin/gwas/op_dep_bulgaria/results/oaall/0001/mis_minimac4_eagle2.4/1000g_p3/eur/chr{1..22}.1000g_ids.maf_gt_0.01_eur_obd.rsq_gt_0.3.gz\n",
    "* UHS1: s3://rti-midas-data/studies/ngc/uhs1/association_tests/002/ea/oaall/uhs1.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz\n",
    "* VIDUS: s3://rti-midas-data/studies/ngc/vidus/association_tests/002/ea/oaall/vidus.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz\n",
    "\n",
    "\n",
    "`5bec8de6-0987-4462-97f7-b1a29939a984`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200217_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/018"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 019\n",
    "VIDUS_FOU (N=300) vs OAall_meta105 individual cohorts\n",
    "\n",
    "**reference**:\n",
    "VIDUS_FOU\n",
    "* s3://rti-midas-data/studies/ngc/vidus/association_tests/001/ea/fou/vidus.ea.fou.chr{1..22}.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz\n",
    "\n",
    "\n",
    "**compared to**:\n",
    "* Bulgaria (N=2,765)  s3://rti-heroin/gwas/op_dep_bulgaria/results/oaall/0001/mis_minimac4_eagle2.4/1000g_p3/eur/chr{1..22}.1000g_ids.maf_gt_0.01_eur_obd.rsq_gt_0.3.gz\n",
    "* CATS-MOLE (N=3,162)  s3://rti-midas-data/studies/ngc/cats/association_tests/010/ea/oaall/catsmole.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* CATS-PERHUNT (N=1,920)   s3://rti-midas-data/studies/ngc/cats/association_tests/011/ea/oaall/catsperthunt.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_cats.rsq_gt_0.3.gz\n",
    "* COGA (N=7,631)   s3://rti-midas-data/studies/ngc/coga/association_tests/001/ea/oaall/coga.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_coga.se.rsq_gt_0.3.gz\n",
    "* deCODE OAall (N=275,468) s3://rti-midas-data/studies/ngc/decode/association_tests/001/ea/oaall/decode.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz\n",
    "* deCODE OAexp (N=2,581)  s3://rti-midas-data/studies/ngc/decode/association_tests/002/ea/oaexp/decode.ea.oaexp.chr{1..22}.1000g_ids.maf_gt_0.01_eur_decode.beta_se.rsq_gt_0.3.gz\n",
    "* Kreek (N=556)   s3://rti-midas-data/studies/ngc/kreek/association_tests/003/ea/oaall/kreek.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_kreek.rsq_gt_0.3.gz\n",
    "* UHS1 (N=9,245)   s3://rti-midas-data/studies/ngc/uhs1/association_tests/002/ea/oaall/uhs1.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_uhs1.rsq_gt_0.3.gz\n",
    "* VIDUS (N=2,177)   s3://rti-midas-data/studies/ngc/vidus/association_tests/002/ea/oaall/vidus.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz\n",
    "* Yale-Penn-CIDR (N=666)   s3://rti-midas-data/studies/ngc/yale-penn/association_tests/004/ea/oaall/yale-penn.cidr.ea.chr{1..22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "* Yale-Penn-GO (N=917)   s3://rti-midas-data/studies/ngc/yale-penn/association_tests/005/ea/oaall/yale-penn.go.ea.chr{1..22}.1000g_ids.maf_gt_0.01_eur_yale-penn.rsq_gt_0.3.gz\n",
    "\n",
    "\n",
    "**old**\n",
    "a:\n",
    "`acdb0afb-2a04-41e7-bc66-ec9db2bdc7f8`\n",
    "\n",
    "b:\n",
    "`6d6d2e6d-fa00-42f1-b594-71152a90ddd1`\n",
    "\n",
    "\n",
    "___\n",
    "**new**\n",
    "a:\n",
    "`c66d4e68-013a-4b84-ace1-ac9ae26540aa` <br>\n",
    "`f358dd7e-078f-45cd-9bda-6b91a6322dea` <br>\n",
    "`8da94fa6-4e36-4ce2-ae0c-880f6bbbc6c9` <br>\n",
    "\n",
    "b:\n",
    "`02a88a65-6871-4c4d-9905-ad5c2e2fe2fd` <br>\n",
    "`8d98e8bc-6c77-45a1-af37-fc9b9803796a`\n",
    "\n",
    "c:\n",
    "`fe7c6248-b96f-41dc-b1e1-8c007d4d28b3`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200225_heroin_ldsc_phenotypes_local.xlsx\n",
    "#procD=/shared/jmarks/heroin/ldsc/fou_oaall/019/a\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/019/b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in {decode-oaexp,decode-oaall,coga,cats-perthunt,cats-mole,bulgaria,yp-go,yp-cidr,vidus,uhs1,kreek}/*; do\n",
    "    ~/bin/ldsc/ldsc.py \\\n",
    "    --h2 $file \\\n",
    "    --ref-ld-chr eur_w_ld_chr/ \\\n",
    "    --w-ld-chr eur_w_ld_chr/ \\\n",
    "    --out $file\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "bulgaria/bulgaria_oaall0001_n2765.txt.munged.merged.txt.log:Total Observed scale h2: 0.2539 (0.1536)\n",
    "cats-mole/catsmole_oaall010_n3162.txt.munged.merged.txt.log:Total Observed scale h2: 0.3934 (0.1543)\n",
    "cats-perthunt/catsperthunt_oaall011_n1920.txt.munged.merged.txt.log:Total Observed scale h2: 0.7583 (0.2346)\n",
    "coga/coga_oaall001_n7631.txt.munged.merged.txt.log:Total Observed scale h2: -0.2965 (0.0534)\n",
    "decode-oaall/decode_oaall001_n275468.txt.munged.merged.txt.log:Total Observed scale h2: 0.0068 (0.0017)\n",
    "decode-oaexp/decode_oaexp002_n2581.txt.munged.merged.txt.log:Total Observed scale h2: 0.0385 (0.1841)\n",
    "kreek/kreek_oaall003_n556.txt.munged.merged.txt.log:Total Observed scale h2: 2.8865 (0.9397)\n",
    "uhs1/uhs1_oaall002_n9245.txt.munged.merged.txt.log:Total Observed scale h2: 0.1642 (0.0478)\n",
    "vidus/vidus_oaall002_n2177.txt.munged.merged.txt.log:Total Observed scale h2: 0.4593 (0.2154)\n",
    "yp-cidr/yalepenn_cidr_oaall004_n666.txt.munged.merged.txt.log:Total Observed scale h2: -0.4239 (0.6054)\n",
    "yp-go/yalepenn_go_oaall005_n917.txt.munged.merged.txt.log:Total Observed scale h2: 0.1106 (0.5307)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 020\n",
    "  VIDUS_OAall (2177) vs FOU_meta cohorts\n",
    "\n",
    "```\n",
    "    reference trait:\n",
    "        VIDUS_OAall (s3://rti-midas-data/studies/ngc/vidus/association_tests/002/ea/oaall/vidus.ea.oaall.chr{1..22}.1000g_ids.maf_gt_0.01_eur_vidus.se.rsq_gt_0.3.gz)\n",
    "\n",
    "    compared to:\n",
    "        * alive (N=152)   s3://rti-heroin/ldsc/data/opioid_fou/alive_fou001_n152.txt.gz\n",
    "        * cats (N=1226)   s3://rti-heroin/ldsc/data/opioid_fou/cats_fou001_n152.txt.gz\n",
    "        * cogend (N=99)   s3://rti-heroin/ldsc/data/opioid_fou/cogend_fou001_n99.txt.gz\n",
    "        * start (N=231)   s3://rti-heroin/ldsc/data/opioid_fou/start_fou001_n231.txt.gz\n",
    "        * uhs1 (N=897)    s3://rti-heroin/ldsc/data/opioid_fou/uhs1_fou001_n897.txt.gz\n",
    "        * uhs2-3 (N=772)  s3://rti-heroin/ldsc/data/opioid_fou/uhs2_3_fou001_n772.txt.gz\n",
    "        * uhs4 (N=1067)   s3://rti-heroin/ldsc/data/opioid_fou/uhs4_fou0001_n1067.txt.gz\n",
    "        * vidus (N=300)   s3://rti-heroin/ldsc/data/opioid_fou/vidus_fou0001_n300.txt.gz\n",
    "        * yale-penn (N=850)  s3://rti-heroin/ldsc/data/opioid_fou/yalepenn_fou001_n850.txt.gz\n",
    "```\n",
    "\n",
    "`dc83a580-8b5e-4a23-92d0-1c0a0cdab14b`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200220_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/020\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 021\n",
    "VIDUS_OAall (N=2177) vs UHS2-4 FOU (N=1562).\n",
    "\n",
    "`723b1e4c-8454-427d-a4fd-4f47b1ce87c5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenD=20200317_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/jmarks/heroin/ldsc/fou_oaall/021\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in {alive,cats,cogend,start,uhs1,uhs23,uhs4,vidus,yp}/*; do\n",
    "    ~/bin/ldsc/ldsc.py \\\n",
    "    --h2 $file \\\n",
    "    --ref-ld-chr eur_w_ld_chr/ \\\n",
    "    --w-ld-chr eur_w_ld_chr/ \\\n",
    "    --out $file.log\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "alive/alive_fou001_n152.txt.munged.merged.txt.log.log:Total Observed scale h2: -1.2532 (2.6704)\n",
    "cats/cats_fou001_n1226.txt.munged.merged.txt.log.log:Total Observed scale h2: 1.8512 (0.4473)\n",
    "cogend/cogend_fou001_n99.txt.munged.merged.txt.log.log:Total Observed scale h2: -6.3417 (4.0912)\n",
    "start/start_fou001_n231.txt.munged.merged.txt.log.log:Total Observed scale h2: 0.0173 (1.7918)\n",
    "uhs1/uhs1_fou001_n897.txt.munged.merged.txt.log.log:Total Observed scale h2: 0.6334 (0.491)\n",
    "uhs23/uhs2_3_fou001_n772.txt.munged.merged.txt.log.log:Total Observed scale h2: 0.03 (0.6457)\n",
    "uhs4/uhs4_fou0001_n1067.txt.munged.merged.txt.log.log:Total Observed scale h2: 0.2717 (0.386)\n",
    "vidus/vidus_fou0001_n300.txt.munged.merged.txt.log.log:Total Observed scale h2: -1.3093 (1.4063)\n",
    "yp/yalepenn_fou001_n850.txt.munged.merged.txt.log.log:Total Observed scale h2: -0.8698 (0.4161)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LD Hub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Selection\n",
    "* Neurological diseases\n",
    "* Personality traits\n",
    "* Cognitive\n",
    "* Education\n",
    "* Brain Volume (ENIGMA)\n",
    "* Psychiatric diseases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cd /shared/jmarks/heroin/ldsc/ngc_all/oaall/001/ldhub\n",
    "#\n",
    "#outF=ngc_oaall_ldhub_with_pvalues.txt # name of file to create for ldhub\n",
    "#samp_size=304507\n",
    "#\n",
    "#### Download outputs for each ref chr from rftm_sumstats step ###\n",
    "#aws s3 sync s3://rti-heroin/ldsc/ngc_20191211/oaall/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/ec7b872f-5fd0-431f-a6de-69f43a787aab/call-munge_chr_wf .\n",
    "#        \n",
    "#mv  */MUNGE_CHR.munge_sumstats_chr_wf/*/call-rfmt_sumstats/*.standardized.phase3ID.munge_ready.txt .\n",
    "#rm -rf shard*\n",
    "#\n",
    "### Concat into single file ##\n",
    "#cat *.chr1.*.standardized.phase3ID.munge_ready.txt > $outF\n",
    "#for chr in {2..22}; do\n",
    "#    tail -n +2  *.chr$chr.*.standardized.phase3ID.munge_ready.txt >> $outF\n",
    "#done\n",
    "#\n",
    "### Remove unnecessary columns (need snpID, A1, A2 Beta, Pvalue) in that order ##\n",
    "#head -1 $outF | cut -f1,4,5,6,7 > tmp\n",
    "#tail -n +2 $outF | awk 'BEGIN{OFS=\"\\t\"}{print $1, $4, $5, $6, $7}'  >> tmp && mv tmp $outF\n",
    "#\n",
    "### Add sample size column (sample = ) and change header names ##\n",
    "#cat $outF | awk -v var=$samp_size -F \"\\t\"  \\\n",
    "#    'BEGIN{OFS=\"\\t\";} NR==1{print \"snpid\", \"A1\", \"A2\", \"BETA\", \"N\", \"P-value\"} \\\n",
    "#    NR>1{print $1,$2,$3,$4,var, $5}' > tmp && mv tmp $outF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## local ##\n",
    "cd /home/jmarks/Projects/heroin/ldsc/ngc_all/oaall/001/processing/input/ldhub\n",
    "scp -i ~/.ssh/gwas_rsa    ec2-user@34.195.174.206:/shared/jmarks/heroin/ldsc/ngc_all/oaall/001/ldhub/ngc_oaall_ldhub_with_pvalues.txt .\n",
    "    \n",
    "# zip file with 7zip\n",
    "ngc_oaall_ldhub_with_pvalues.txt.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## upload input file\n",
    "Follow the steps above to zip and upload input file. Essentially, \n",
    "* zip the LDHub input file created above (and only this file)\n",
    "* download the file to your local machine.\n",
    "* Login to [LDHub](http://ldsc.broadinstitute.org/ldhub/) by clicking on `Get Started with LDHub` and then sign in with your Google email account. (jambqc for me)\n",
    "* Click `Go Test Center`\n",
    "* Click `Continue`\n",
    "* Upload zipped file by clicking `Choose File`, naming your trait, and clicking `Continue`.\n",
    "* Select traits of interest from LDHub by checking the box next to the trait of interest and then clicking `Submit your request`\n",
    "\n",
    "**Note**: keep browser open during LDSC analysis on LDHub.\n",
    "\n",
    "<br>\n",
    "\n",
    "___\n",
    "\n",
    "```\n",
    "Important notes for your uploaded file:\n",
    "\n",
    "1. To save the uploading time, LD Hub only accepts zipped files as input (e.g. mydata.zip).\n",
    "\n",
    "2. Please check that there is ONLY ONE plain TXT file (e.g. mydata.txt) in your zipped file.\n",
    "\n",
    "3. Please make sure you do NOT zip any folder together with the plain txt file (e.g. /myfolder/mydata.txt), otherwise you will get an error: [Errno 2] No such file or directory\n",
    "\n",
    "4. Please do NOT zip multiple files (e.g. zip mydata.zip file1.txt file2.txt ..) or zip a file with in a folder (e.g. zip mydata.zip /path/to/my/file/mydata.txt).\n",
    "\n",
    "5. Please keep the file name of your plain txt file short (less than 50 characters), otherwise you may get an error: [Errno 2] No such file or directory\n",
    "\n",
    "6. Please zip your plain txt file using following command (ONE file at a time):\n",
    "\n",
    "For Windows system: 1) Locate the file that you want to compress. 2) Right-click the file, point to Send to, and then click Compressed (zipped) folder.\n",
    "\n",
    "For Linux and Mac OS system: zip mydata.zip mydata.txt\n",
    "\n",
    "Reminder: for Mac OS system, please do NOT zip you file by right click mouse and click \"Compress\" to zip your file, this will automatically create a folder called \"__MACOS\". You will get an error: [Errno 2] No such file or directory.\n",
    "\n",
    "Upload the trait of interest\n",
    "To save your upload time, we highly recommend you to use the SNP list we used in LD Hub to reduce the number of SNPs in your uploaded file. Click here to download our SNP list (w_hm3.noMHC.snplist.zip).\n",
    "\n",
    "Please upload the zipped file you just created. Click here to download an input example.\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/jmarks/heroin/ldsc/ngc_all/oaexp/002/ldhub\n",
    "outF=oaexp_with_pvalues.txt # name of file to create for ldhub\n",
    "samp_size=5561\n",
    "\n",
    "### Download outputs for each ref chr from rftm_sumstats step ###\n",
    "aws s3 sync s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/b5167ef6-cc76-4f70-8eef-f7414e441cc8/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/b25fcc19-9146-4fd4-b7bf-2ca291b908c9/call-munge_chr_wf/ . \n",
    "        \n",
    "mv  */MUNGE_CHR.munge_sumstats_chr_wf/*/call-rfmt_sumstats/*.standardized.phase3ID.munge_ready.txt .\n",
    "rm -rf shard*\n",
    "\n",
    "## Concat into single file ##\n",
    "cat *.chr1.*.standardized.phase3ID.munge_ready.txt > $outF\n",
    "for chr in {2..22}; do\n",
    "    tail -n +2  *.chr$chr.*.standardized.phase3ID.munge_ready.txt >> $outF\n",
    "done\n",
    "\n",
    "## Remove unnecessary columns (need snpID, A1, A2 Beta, Pvalue) in that order ##\n",
    "head -1 $outF | cut -f1,4,5,6,7 > tmp\n",
    "tail -n +2 $outF | awk 'BEGIN{OFS=\"\\t\"}{print $1, $4, $5, $6, $7}'  >> tmp && mv tmp $outF\n",
    "\n",
    "## Add sample size column (sample = 18245.00) and change header names ##\n",
    "cat $outF | awk -v var=$samp_size -F \"\\t\"  \\\n",
    "    'BEGIN{OFS=\"\\t\";} NR==1{print \"snpid\", \"A1\", \"A2\", \"BETA\", \"N\", \"P-value\"} \\\n",
    "    NR>1{print $1,$2,$3,$4,var, $5}' > tmp && mv tmp $outF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Final Plot\n",
    "Merge the output tables from cromwell and LDHub. The merged table should have the header:\n",
    "```\n",
    "trait2\tTrait_Label\tTrait_Group\trg\tse\tz\tp\th2_obs\th2_obs_se\th2_int\th2_int_se\tgcov_int\tgcov_int_se\n",
    "```\n",
    "\n",
    "**Note**: upload the plot table to EC2 instance to run docker and create the plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## enter interactive mode ##\n",
    "# note that the image tag corresponds to the latest tag for this image\n",
    "\n",
    "\n",
    "\n",
    "docker run -it -v\"/shared/jmarks/heroin/ldsc/fou_oaall/019/plot/:/data/\" \\\n",
    "    rticode/plot_ld_regression_results:b018e08753390ee773ed7e9eb2ca851c88eee749  /bin/bash\n",
    "\n",
    "\n",
    "Rscript /opt/plot_ld_regression/plot_ld_regression_results.R  \\\n",
    "    --input_file 20200226_vidus_fou_vs_oaall.ld_regression_results.csv \\\n",
    "    --output_file 20200226_vidus_fou_vs_oaall.ld_regression_results.pdf  \\\n",
    "    --comma_delimited \\\n",
    "    --xmax 1 \\\n",
    "    --xmin -1.6 \\\n",
    "    --title \"VIDUS_FOU vs OAall Cohorts\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "java -Dconfig.file=/shared/jmarks/bin/cromwell/cromwell_default_genomics_queue.conf \\\n",
    "    -jar /shared/jmarks/bin/cromwell/cromwell-44.jar \\\n",
    "    run hello.wdl \\\n",
    "    -i inputs.json \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "714.716px",
    "left": "0px",
    "right": "1299.83px",
    "top": "110.284px",
    "width": "211.989px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
