{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FTND GWAS of Nelson Data\n",
    "\n",
    "**Author**: Jesse Marks\n",
    "\n",
    "**GitHub Issue**: [issue 96](https://github.com/RTIInternational/bioinformatics/issues/96#issuecomment-401903152)\n",
    "\n",
    "This document logs the steps taken to process:\n",
    "\n",
    "[*A Genome-Wide Association Study of Heroin Dependence*](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000277.v1.p1&phv=171252&phd=&pha=&pht=2797&phvf=&phdf=&phaf=&phtf=&dssp=1&consent=&temp=1) (Nelson data set) and perform an FTND GWAS to test for association with Nicotine Dependence (ND). This Nelson data set is a collaboration of Australian and American investigators aims to identify genes associated with liability for heroin dependence. There is FTND phenotype data included however, so we will perform a ND GWAS.\n",
    "\n",
    "The [FagerstrÃ¶m Test for Nicotine Dependence (FTND](https://cde.drugabuse.gov/instrument/d7c0b0f5-b865-e4de-e040-bb89ad43202b) is a standard instrument for assessing the intensity of physical addiction to nicotine.\n",
    "\n",
    "The data for this study are located on S3 at:\n",
    "\n",
    "`s3://rti-midas-data/studies/heroin_icghd/`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Nelson Data set has been downloaded to an EC2 instance from S3 to the location:\n",
    "`/shared/data/studies/Nelson`\n",
    "\n",
    "\n",
    "`/shared/data/studies/Nelson/UHS_heroin`\n",
    "\n",
    "The head of the data looks like the follow:\n",
    "\n",
    "`head imputed.aa_pop_pheno_controls_overlap.chr1.heroin_10EVs.palogist.stats`\n",
    "\n",
    "`name position A1 A2 Freq1 MAF Quality Rsq n Mean_predictor_allele beta_SNP_add sebeta_SNP_add loglik chi p or_95_percent_ci       chr\n",
    "rs58108140 10583 G A 0.923 0.077 0 0.471 3662 0.923191 0.00656649 0.133317 -2320.59 0.00242602632930075 0.960716312898352 1.01(0.78-1.31) 1\n",
    "`\n",
    "\n",
    "Where is the phenotype file for these data? What data do I need in order to perform a gwas? What I should do is look at my old notebooks to see how I performed the GWAS and then replicate this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Set autoscaling tags**\n",
    "\n",
    "This can be done in the AWS console."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aws autoscaling create-or-update-tags --tags \"ResourceId=cfncluster-JMspot2-ComputeFleet-IQXOI7AVONI2,ResourceType=auto-scaling-group,Key=project-number,Value=0215457.000.001,PropagateAtLaunch=true\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 ##\n",
    "aws s3 sync s3://rti-midas-data/studies/heroin_icghd/ /shared/data/studies/nelson --quiet &\n",
    "mkdir -p /shared/jmarks/nicotine/gwas/nelson/phenotype\n",
    "cp /shared/data/studies/nelson/phenotypes/probabel/heroin_icghd.ea.CAT_FTND.AGE.SEX.ALCOHOLDEP4.EVs.v1.gz \\\n",
    "    /shared/jmarks/nicotine/gwas/nelson/phenotype\n",
    "\n",
    "cd /shared/jmarks/nicotine/gwas/nelson/phenotype\n",
    "gunzip heroin_icghd.ea.CAT_FTND.AGE.SEX.ALCOHOLDEP4.EVs.v1.gz\n",
    "\n",
    "# only what to run a baseline gwas and therefore we will remove the alcohol covariate column\n",
    "awk '{print $1,$2,$3,$4,$6,$7,$8, $9}' heroin_icghd.ea.CAT_FTND.AGE.SEX.ALCOHOLDEP4.EVs.v1 > nelson.ea.CAT_FTND.AGE.SEX.EVs.v1\n",
    "\n",
    "mkdir -p ../association_test/001\n",
    "cd association_tests/001\n",
    "\n",
    "cp /shared/data/studies/nelson/imputed/v1/association_tests/003/_methods.heroin_icghd.v1.association_tests.003.sh \\\n",
    "    _methods.nelson.v1.association_tests.sh\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/data/0212964/nextflow/nextflow-0.25.1-all\n",
    "\n",
    "/* ######################################################################\n",
    "#                 Association Pipeline v0.1                             #\n",
    "###################################################################### */\n",
    "\n",
    "/*\n",
    " * Defines pipeline parameters\n",
    " */\n",
    "\n",
    "params.final_chunks = \"The final chunking file splited by chromosome\"\n",
    "params.input_pheno = \"The phenotype to be used in this analyses, file generated in ESN Windows environment\"\n",
    "params.imputation_dir = \"The directory contains the post imputation genotype files (mldose files)\"\n",
    "params.example_mldose = \"One of the mldose files in the imputation directory\"\n",
    "params.geno_prefix = \"The prefix of the mldose file\"\n",
    "params.working_dirs = \"The working directory\"\n",
    "params.out = \"The output file name for the stats file\"\n",
    "params.method = \"palinear or palogist\"\n",
    "\n",
    "/*\n",
    " * probabel_header = \"name\\tchrom\\tposition\\tA1\\tA2\\tFreq1\\tMAF\\tQuality\\tRsq\\tn\\tMean_predictor_allele\\tbeta_SNP_add\\tsebeta_SNP_add\\tchi2_SNP\\tchi\\tp\\tor_95_percent_ci\"\n",
    " */\n",
    "\n",
    "// create chunks channels from chunking file\n",
    "chunks = Channel\n",
    "                        .from( file(params.final_chunks) )\n",
    "                        .splitCsv(header:['chr', 'chunk', 'start', 'end'], skip: 1, sep: '\\t')\n",
    "chunks.into { chunks_prep_geno; chunks_merge; }\n",
    "\n",
    "/* *********************************************\n",
    " * Step 1: Start Prepare ProbABEL Phenotype File\n",
    " */\n",
    "\n",
    "process prepare_pheno{\n",
    " input:\n",
    "        // input defined by parameters\n",
    "\n",
    "        output:\n",
    "        file \"probabel_pheno\" into probabel_phenotype_file\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/perl/prepare_probabel_files.pl \\\n",
    "                --in_mldose ${params.example_mldose} \\\n",
    "                --in_pheno ${params.input_pheno} \\\n",
    "                --out_pheno \"probabel_pheno\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* ********************************************\n",
    " * Step 2: Start Prepare ProbABEL Genotype File\n",
    " */\n",
    "\n",
    "process prepare_geno{\n",
    "\n",
    "        input:\n",
    "        file \"probabel_pheno\" from probabel_phenotype_file\n",
    "        set chr, chunk, start, end from chunks_prep_geno\n",
    "\n",
    "        output:\n",
    "        file \"${params.geno_prefix}${chr}.${chunk}.mach_mldose\" into probabel_genotype_files\n",
    "\n",
    "    executor 'sge'\n",
    "    clusterOptions '-S /bin/bash -l mem_free=15G,h_vmem=15G -p 0'\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/perl/prepare_probabel_files.pl \\\n",
    "                --in_mldose \"${params.imputation_dir}/${params.geno_prefix}${chr}.${chunk}.mach.mldose.gz\" \\\n",
    "                --in_pheno \"${probabel_pheno}\" \\\n",
    "                --out_mldose \"${params.geno_prefix}${chr}.${chunk}.mach_mldose\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* *******************************\n",
    " * Step 3: Start ProbABEL Analysis\n",
    " */\n",
    "\n",
    "process do_probabel_analysis{\n",
    "\n",
    "        input:\n",
    "        file probabel_pheno from probabel_phenotype_file\n",
    "        file geno from probabel_genotype_files\n",
    "\n",
    "        output:\n",
    "        file \"${geno.baseName}_add.out.txt\" into probabel_out_files\n",
    "\n",
    "        executor 'sge'\n",
    "    clusterOptions '-S /bin/bash -l mem_free=7G,h_vmem=7G -p 0'\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/${params.method} \\\n",
    "        --pheno $probabel_pheno \\\n",
    "        --dose $geno \\\n",
    "        --info \"${params.imputation_dir}/${geno.baseName}.mach.mlinfo\" \\\n",
    "        --map \"${params.imputation_dir}/${geno.baseName}.legend\" \\\n",
    "        --out \"${geno.baseName}\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* ***************************\n",
    " * Step 4: Start calc chi p OR\n",
    " */\n",
    "process calc_chi_p_or{\n",
    "\n",
    "        input:\n",
    "        file result from probabel_out_files\n",
    "\n",
    "        output:\n",
    "        file \"${result.baseName}.stats\" into stats_files\n",
    "\n",
    "        executor 'sge'\n",
    "    clusterOptions '-S /bin/bash -l mem_free=7G,h_vmem=7G -p 0'\n",
    "\n",
    "        \"\"\"\n",
    "        /share/nas03/bioinformatics_group/software/R/calculate_stats_for_probabel_results_v2.R \\\n",
    "                --remove_missing_p \\\n",
    "                --in_file \"${result}\" \\\n",
    "                --out_file \"${result.baseName}.stats\"\n",
    "        \"\"\"\n",
    "}\n",
    "\n",
    "/* ***************************\n",
    " * Step 5: merge stats files\n",
    " */\n",
    "\n",
    "stats_files\n",
    "        .collectFile(name: file(params.out), skip: 1)\n",
    "        .println {\"Result saved to file: $it\"}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation Conversion\n",
    "\n",
    "There are some apparent issues with the `mldose` files. Specifically, the minor allele frequencies (MAF) are not lining up to what they should be. The issue was found after performing a baseline GWAS and finding that all of the results were NULL. After comparing the MAF of the `mldose` files and the `gen.gz` files, one can see that the MAF are in discordance. As an example, let us compare the MAF of SNP `rs16969968` (Mr. Big). \n",
    "\n",
    "```\n",
    "heroin_icghd.ea.1000G_p3.chr15.21.mach.mlinfo:\n",
    "SNP     Al1     Al2     Freq1   MAF     Quality Rsq\n",
    "rs16969968:78882925:G:A   G       A       0.002   0.002   0       0\n",
    "\n",
    " heroin_icghd.ea.1000G_p3.chr15.21.gen_info\n",
    "snp_id rs_id position exp_freq_a1 info certainty type info_type0 concord_type0 r2_type0\n",
    "--- rs16969968:78882925:G:A 78882925 0.349 0.996 0.998 0 -1 -1 -1\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#perl /shared/bioinformatics/software/perl/file_conversion/convert_post-v2.3.2_impute2_files.pl\\\n",
    "dataDir=/shared/data/studies/nelson/imputed/v1/imputations/ea/chr15/sandbox\n",
    "\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name impute_conversion_test \\\n",
    "    --script_prefix $dataDir/test2/mytest \\\n",
    "    --mem 15 \\\n",
    "    --nslots 4 \\\n",
    "    --priority 0 \\\n",
    "    --program perl /shared/bioinformatics/software/perl/file_conversion/convert_pre-v2.3.2_impute2_files.pl\\\n",
    "        --gen $dataDir/heroin_icghd.ea.1000G_p3.chr15.21.gen.gz \\\n",
    "        --sample $dataDir/heroin_icghd.ea.for_imputation.chr15.phased.sample.gz\\\n",
    "        --gen_info $dataDir/heroin_icghd.ea.1000G_p3.chr15.21.gen_info \\\n",
    "        --remove $dataDir/heroin_icghd.ea.1000G_p3.chr15.gen.remove\\\n",
    "        --out $dataDir/test2/heroin_icghd.ea.1000G_p3.chr15.21\\\n",
    "        --generate_mach_mldose_file\\\n",
    "        --generate_mach_mlinfo_file\\\n",
    "        --generate_legend_file \\\n",
    "        --variant_type all\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
