{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d1c0af67-5044-4cce-b10a-352777677a8d",
   "metadata": {},
   "source": [
    "# LDSC-SEG on TWAS Meta-Analysis Results for OUD\n",
    "\n",
    "**Author**: Jesse Marks<br>\n",
    "**NIH Project**: [Harnessing Knowledge of Gene Function in Brain Tissue for Discovering Biology Underlying Heroin Addiction](https://reporter.nih.gov/search/RC99reuHhEW0n_3WuFPU6g/project-details/10116351) <br>\n",
    "**Charge Code**:0218755.001.002.001 ~0215889.001.001~<br>\n",
    "**GitHub Issue**:  [Opioid Use Disorder TWAS Meta-analysis (Uniform Processing) #183](https://github.com/RTIInternational/bioinformatics/issues/183)<br>\n",
    "\n",
    "\n",
    "**Description**:<br>\n",
    "This notebook details the process used to test for heritability enrichment in the significantly differentially expressed genes from our TWAS meta-analysis for ~41~ 47 phenotypes.\n",
    "\n",
    "\n",
    "We performed Stratified LD Score Regression (S-LDSC) analyses using the LD score regression approach described in [Finucane et al. 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5896795/) on specifically expressed genes (LDSC-SEG).\n",
    "LDSC-SEG applies stratified LD score regression to test whether disease heritability is enriched in the regions surrounding genes with the highest specific expression in a given tissue.\n",
    "This approach helps to interpret GWAS signal by leveraging gene expression data.\n",
    "We used gene sets from a TWAS meta-analysis of 4 OUD case/control datasets:\n",
    "\n",
    "- [Corradin et al. (2022) Molecular Psychiatry](https://doi.org/10.1038/s41380-022-01477-y)\n",
    "- [Mendez et al. (2021) Molecular Psychiatry](https://doi.org/10.1038/s41380-021-01259-y)\n",
    "- [Seney et al. (2021) Biological Psychiatry](https://doi.org/10.1016/j.biopsych.2021.06.007)\n",
    "- [Sosnowski et al. (2022) Drug and Alcohol Dependence Reports](https://doi.org/10.1016/j.dadr.2022.100040)\n",
    "\n",
    "The meta-analysis results are published in [THIS](https://github.com/RTIInternational/bioinformatics/issues/183#issuecomment-1198311459) GitHub comment, and we used the results with singletons removed: genes that only appear in one study.\n",
    "The file name is `meta_analysis_sumstats_no_singletons_20220727.tsv.gz` and is located under the Results section.\n",
    "\n",
    "We filtered the results to obtain two sets of significantly expressed genes:\n",
    "\n",
    "- List of genes with a Benjamini-Hochberg FDR <0.05\n",
    "- List of genes with a Benjamini-Hochberg FDR <0.10\n",
    "\n",
    "The wfisher_adj_pvalue column was used to create these gene lists.\n",
    "This column contains the [weighted Fisher's p-value](https://www.nature.com/articles/s41598-021-86465-y) that we applied the Benjamini-Hochberg FDR thresholds to.\n",
    "\n",
    "We used the same set of Opioid-Use GWAS results used in the paper [Multi-trait genome-wide association study of opioid addiction: OPRM1 and Beyond](https://www.medrxiv.org/content/10.1101/2021.09.13.21263503v1) for the LDSC-SEG analyses.\n",
    "Our aim was to determine whether genes (and their proximal genomic regions) that show differential expression by opioid use disorder are also enriched for genetic signal associated with opioid addiction and related phenotypes.\n",
    "\n",
    "Overall, this notebook will address the biological question of whether significantly differentially expressed genes (and their proximal regions) from the opioid use disorder TWAS meta-analysis results showed enrichment for genetic signal associated with opioid addiction or any related phenotypes from the following EUR-specific studies:\n",
    "\n",
    "___\n",
    "\n",
    "<br><br>\n",
    "\n",
    "<details>\n",
    "    <summary>phenotype list</summary>\n",
    "    \n",
    "   add the following 7 sleep studies to the list:\n",
    "* Insomnia - Jansen 2019\n",
    "* Insomnia - Lane 2019\n",
    "* LongSleepDur - Dashti 2019\n",
    "* ShortSleepDur - Dashti 2019\n",
    "* sleepDuration - Dashti 2019\n",
    "* Sleepdur - Jansen 2019\n",
    "    \n",
    "___    \n",
    "    \n",
    "* Age of Initiation  (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Alcohol Dependence (Walters et al., 2018 Nat Neurosci [30482948](https://pubmed.ncbi.nlm.nih.gov/30482948))\n",
    "* Alcohol Drinks per Week (DPW) (Liu et al., 2019 Nat Genet [30643251]())\n",
    "* Alzheimer's Disease (Lambert et al., 2013 Nat Genet [24162737](https://pubmed.ncbi.nlm.nih.gov/24162737))\n",
    "* Amyotrophic Lateral Sclerosis (Rheenen et al., 2016 Nat Genet [27455348](https://pubmed.ncbi.nlm.nih.gov/27455348))\n",
    "* Anorexia Nervosa (Watson et al., 2019 Nat Genet [31308545](https://pubmed.ncbi.nlm.nih.gov/31308545))\n",
    "* Attention Deficit Hyperactivity Disorder (Demontis et al., 2019 Nat Genet [30478444]())\n",
    "* Autism Spectrum Disorders (Grove et al., 2019 Nat Genet [30804558](https://pubmed.ncbi.nlm.nih.gov/30804558))\n",
    "* Bipolar Disorder (Stahl et al., 2019 Nat Genet [31043756](https://pubmed.ncbi.nlm.nih.gov/31043756))\n",
    "* Cannabis Use Disorder (CUD) (Demontis et al., 2019 Nat Neurosci [31209380](https://pubmed.ncbi.nlm.nih.gov/31209380))\n",
    "* Childhood IQ (Benyamin et al., 2014 Mol Psychiatry [23358156](https://pubmed.ncbi.nlm.nih.gov/23358156))\n",
    "* Cigarettes Per Day (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* College Completion (Rietveld et al., 2013 Science [23722424](https://pubmed.ncbi.nlm.nih.gov/23722424))\n",
    "* Cotinine Levels (Ware et al., 2016 Sci Rep [26833182](https://pubmed.ncbi.nlm.nih.gov/26833182/))\n",
    "* Fagerstrom Test for Nicotine Dependence (FTND) (Quach et al., 2020 Nat Commun [33144568](https://pubmed.ncbi.nlm.nih.gov/33144568/))\n",
    "* Heaviness of Smoking Index (HSI) (Quach et al., 2020 Nat Commun [33144568](https://pubmed.ncbi.nlm.nih.gov/33144568/))\n",
    "* Intelligence (Sniekers et al., 2017 Nat Genet [28530673](https://pubmed.ncbi.nlm.nih.gov/28530673))\n",
    "* Lifetime Cannabis Use (Ever vs. Never) (Pasman et al., 2018 Nat Neurosci [30150663](https://pubmed.ncbi.nlm.nih.gov/30150663))\n",
    "* Major Depressive Disorder (Howard et al., 2018 Nat Commun [29662059](https://pubmed.ncbi.nlm.nih.gov/29662059))\n",
    "* Mean Accumbens Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Caudate Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Hippocampus Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Pallidum Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Putamen Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Thalamus Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry [21173776](https://pubmed.ncbi.nlm.nih.gov/21173776))\n",
    "* Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry [21173776](https://pubmed.ncbi.nlm.nih.gov/21173776))\n",
    "* Neuroticism (Okbay et al., 2016 Nat Genet [27089181]())\n",
    "* Opioid Addiction: GENOA GWAS meta-analysis\n",
    "* Opioid Addiction: gSEM OA GWAS meta-analysis (i.e., GENOA, MVP-SAGE-YP, PGC-SUD, and Partners Health)\n",
    "* Parkinson's Disease (Sanchez et al., 2009 Nat Genet [19915575](https://pubmed.ncbi.nlm.nih.gov/19915575))\n",
    "* Post-traumatic Stress Disorder (Nievergelt et al., 2019 Nat Commun [31594949](https://pubmed.ncbi.nlm.nih.gov/31594949))\n",
    "* Psychiatric Genetics Consortium Cross-disorder GWAS (Schizophrenia, Bipolar Disorder, MDD, ASD and ADHD) (Cross-Disorder Group of the Psychiatric Genomics Consortium, 2013 Lancet [23453885](https://pubmed.ncbi.nlm.nih.gov/23453885))\n",
    "* Schizophrenia (Ripke et al., 2014 Nature [25056061](https://pubmed.ncbi.nlm.nih.gov/25056061))\n",
    "* Smoking Cessation (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Smoking Initiation (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Subjective Well Being (Okbay et al., 2016 Nat Genet [27089181](https://pubmed.ncbi.nlm.nih.gov/27089181))\n",
    "* Total Intracranial Volume (ICV) (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Years of Schooling (Okbay et al., 2016 Nature [27225129](https://pubmed.ncbi.nlm.nih.gov/27225129))\n",
    "</details><br><br>\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e67c8a81-412e-452e-996b-4643635e2e89",
   "metadata": {},
   "source": [
    "## Genomic start-end locations for GENCODE v30 IDS\n",
    "Retrieve the genomic start and end locations for the GENCODE v30 IDs for meta-analysis significant genes at FDR <0.05 when singletons are excluded. Meta-analysis summary statistics without singletons are posted in this [comment](https://github.com/RTIInternational/bioinformatics/issues/183#issuecomment-1198311459) as file `meta_analysis_sumstats_no_singletons_20220727.tsv.gz`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "975c6590-0f2c-4fe9-93d1-a7c9075c08ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gencode_id\tgene_name\tbase_mean_expression_corradin\tbase_mean_log2cpm_corradin\tfold_change_corradin\tlog2_fold_change_corradin\tlog2_fold_change_se_corradin\ttest_statistic_corradin\tb_statistic_corradin\tpvalue_corradin\tadjusted_pvalue_corradin\tbase_mean_expression_mendez\tbase_mean_log2cpm_mendez\tfold_change_mendez\tlog2_fold_change_mendez\tlog2_fold_change_se_mendez\ttest_statistic_mendez\tb_statistic_mendez\tpvalue_mendez\tadjusted_pvalue_mendez\tbase_mean_expression_seney\tbase_mean_log2cpm_seney\tfold_change_seney\tlog2_fold_change_seney\tlog2_fold_change_se_seney\ttest_statistic_seney\tb_statistic_seney\tpvalue_seney\tadjusted_pvalue_seney\tbase_mean_expression_sosnowski\tbase_mean_log2cpm_sosnowski\tfold_change_sosnowski\tlog2_fold_change_sosnowski\tlog2_fold_change_se_sosnowski\ttest_statistic_sosnowski\tb_statistic_sosnowski\tpvalue_sosnowski\tadjusted_pvalue_sosnowski\tnum_datasets\tfc_sign_corradin\tfc_sign_mendez\tfc_sign_seney\tfc_sign_sosnowski\twfisher_fc_sign\twfisher_pvalue\twfisher_adj_pvalue\n",
      "ENSG00000000003.14\tTSPAN6\t280.590909090909\t2.15463993595764\t0.972324915595254\t-0.0404896043186799\t0.115514949361746\t-0.350513977129339\t-5.56185774421871\t0.728294376660881\t0.941113912764389\t171.515675\t2.87339596889899\t1.29794088555831\t0.376224677594399\t0.158993917816571\t2.36628345763794\t-3.38865413716201\t0.02332628916534\t0.192483045460433\t151.4161\t3.00917901371786\t1.05677694215658\t0.0796708938403902\t0.0723683795231596\t1.10090752847234\t-5.62032119746489\t0.278135223021147\t0.592633128020453\t176.66272972973\t2.6468452333018\t1.02623331886532\t0.0373587715905427\t0.0519246798255103\t0.719480056806985\t-5.43229896328812\t0.473151309979134\t0.887644640704241\t4\t-1\t1\t1\t1\t1\t0.0913917604863265\t0.32213328944865\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | head -2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dfb81a3f-32f9-4109-9f43-0329ff7c45fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wfisher_adj_pvalue\n",
      "0.32213328944865\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# verify the last column is the wfisher_adj_pvalue column (the Benjamini-Hochberg pvalue)\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | head -2 | awk \\\n",
    "'{print $NF}'\n",
    "\n",
    "# create FDR filtered file: 0.05\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | \\\n",
    "  head -1 > meta_analysis_sumstats_no_singletons_20220727_fdr0.05.tsv\n",
    "\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | tail -n +2 | awk \\\n",
    "'$NF < 0.05 ' >> meta_analysis_sumstats_no_singletons_20220727_fdr0.05.tsv\n",
    "\n",
    "\n",
    "# create FDR filtered file: 0.10\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | \\\n",
    "  head -1 > meta_analysis_sumstats_no_singletons_20220727_fdr0.10.tsv\n",
    "\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | tail -n +2 | awk \\\n",
    "'$NF < 0.10 ' >> meta_analysis_sumstats_no_singletons_20220727_fdr0.10.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f0d69cd-2805-48a0-b118-7dac9b096862",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# download gencode v30 annotation file\n",
    "wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_30/GRCh37_mapping/gencode.v30lift37.annotation.gtf.gz\n",
    "\n",
    "gunzip --to-stdout gencode.v30lift37.annotation.gtf.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3d10ced6-e212-411a-8eff-2f3f3c72b3dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#description: evidence-based annotation of the human genome, version 30 (Ensembl 96), mapped to GRCh37 with gencode-backmap\n",
      "#provider: GENCODE\n",
      "#contact: gencode-help@ebi.ac.uk\n",
      "#format: gtf\n",
      "#date: 2019-04-02\n",
      "chr1\tHAVANA\tgene\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; level 2; havana_gene \"OTTHUMG00000000961.2_2\"; remap_status \"full_contig\"; remap_num_mappings 1; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_num_mappings 1; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 1; exon_id \"ENSE00002234944.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:11869-12227\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t12613\t12721\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 2; exon_id \"ENSE00003582793.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:12613-12721\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t13221\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 3; exon_id \"ENSE00002312635.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:13221-14409\"; remap_status \"full_contig\";\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f2accd5e-b7d4-42a6-89dc-1c5c977009d4",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "\n",
    "get the start and end positions for each gene\n",
    "https://www.gencodegenes.org/pages/data_format.html\n",
    "\n",
    "| column-number   | content                | values/format           |\n",
    "|-----------------|------------------------|-------------------------|\n",
    "| 4               | genomic start location | integer-value (1-based) |\n",
    "| 5               | genomic end location   | integer-value           |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f0e41733-37d5-48cb-b39f-c1fe021f448e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>base_mean_expression_corradin</th>\n",
       "      <th>base_mean_log2cpm_corradin</th>\n",
       "      <th>fold_change_corradin</th>\n",
       "      <th>log2_fold_change_corradin</th>\n",
       "      <th>log2_fold_change_se_corradin</th>\n",
       "      <th>test_statistic_corradin</th>\n",
       "      <th>b_statistic_corradin</th>\n",
       "      <th>pvalue_corradin</th>\n",
       "      <th>...</th>\n",
       "      <th>pvalue_sosnowski</th>\n",
       "      <th>adjusted_pvalue_sosnowski</th>\n",
       "      <th>num_datasets</th>\n",
       "      <th>fc_sign_corradin</th>\n",
       "      <th>fc_sign_mendez</th>\n",
       "      <th>fc_sign_seney</th>\n",
       "      <th>fc_sign_sosnowski</th>\n",
       "      <th>wfisher_fc_sign</th>\n",
       "      <th>wfisher_pvalue</th>\n",
       "      <th>wfisher_adj_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>280.590909</td>\n",
       "      <td>2.154640</td>\n",
       "      <td>0.972325</td>\n",
       "      <td>-0.040490</td>\n",
       "      <td>0.115515</td>\n",
       "      <td>-0.350514</td>\n",
       "      <td>-5.561858</td>\n",
       "      <td>0.728294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.473151</td>\n",
       "      <td>0.887645</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.091392</td>\n",
       "      <td>0.322133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>393.045477</td>\n",
       "      <td>2.667048</td>\n",
       "      <td>0.888755</td>\n",
       "      <td>-0.170143</td>\n",
       "      <td>0.092914</td>\n",
       "      <td>-1.831176</td>\n",
       "      <td>-4.277967</td>\n",
       "      <td>0.076581</td>\n",
       "      <td>...</td>\n",
       "      <td>0.762672</td>\n",
       "      <td>0.959600</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.363189</td>\n",
       "      <td>0.609168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>543.378159</td>\n",
       "      <td>3.128846</td>\n",
       "      <td>0.934389</td>\n",
       "      <td>-0.097905</td>\n",
       "      <td>0.064771</td>\n",
       "      <td>-1.511541</td>\n",
       "      <td>-4.832444</td>\n",
       "      <td>0.140654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.439635</td>\n",
       "      <td>0.876792</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.403248</td>\n",
       "      <td>0.637638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>155.691750</td>\n",
       "      <td>1.329310</td>\n",
       "      <td>0.975023</td>\n",
       "      <td>-0.036492</td>\n",
       "      <td>0.127130</td>\n",
       "      <td>-0.287042</td>\n",
       "      <td>-5.374356</td>\n",
       "      <td>0.775967</td>\n",
       "      <td>...</td>\n",
       "      <td>0.401802</td>\n",
       "      <td>0.865274</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.832877</td>\n",
       "      <td>0.935349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>FGR</td>\n",
       "      <td>44.590909</td>\n",
       "      <td>-0.829102</td>\n",
       "      <td>0.824932</td>\n",
       "      <td>-0.277653</td>\n",
       "      <td>0.283458</td>\n",
       "      <td>-0.979521</td>\n",
       "      <td>-4.768122</td>\n",
       "      <td>0.334812</td>\n",
       "      <td>...</td>\n",
       "      <td>0.706047</td>\n",
       "      <td>0.949224</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.266319</td>\n",
       "      <td>0.525174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gencode_id gene_name  base_mean_expression_corradin  \\\n",
       "0  ENSG00000000003.14    TSPAN6                     280.590909   \n",
       "1  ENSG00000000419.12      DPM1                     393.045477   \n",
       "2  ENSG00000000457.14     SCYL3                     543.378159   \n",
       "3  ENSG00000000460.17  C1orf112                     155.691750   \n",
       "4  ENSG00000000938.13       FGR                      44.590909   \n",
       "\n",
       "   base_mean_log2cpm_corradin  fold_change_corradin  \\\n",
       "0                    2.154640              0.972325   \n",
       "1                    2.667048              0.888755   \n",
       "2                    3.128846              0.934389   \n",
       "3                    1.329310              0.975023   \n",
       "4                   -0.829102              0.824932   \n",
       "\n",
       "   log2_fold_change_corradin  log2_fold_change_se_corradin  \\\n",
       "0                  -0.040490                      0.115515   \n",
       "1                  -0.170143                      0.092914   \n",
       "2                  -0.097905                      0.064771   \n",
       "3                  -0.036492                      0.127130   \n",
       "4                  -0.277653                      0.283458   \n",
       "\n",
       "   test_statistic_corradin  b_statistic_corradin  pvalue_corradin  ...  \\\n",
       "0                -0.350514             -5.561858         0.728294  ...   \n",
       "1                -1.831176             -4.277967         0.076581  ...   \n",
       "2                -1.511541             -4.832444         0.140654  ...   \n",
       "3                -0.287042             -5.374356         0.775967  ...   \n",
       "4                -0.979521             -4.768122         0.334812  ...   \n",
       "\n",
       "   pvalue_sosnowski  adjusted_pvalue_sosnowski  num_datasets  \\\n",
       "0          0.473151                   0.887645             4   \n",
       "1          0.762672                   0.959600             4   \n",
       "2          0.439635                   0.876792             4   \n",
       "3          0.401802                   0.865274             4   \n",
       "4          0.706047                   0.949224             4   \n",
       "\n",
       "   fc_sign_corradin  fc_sign_mendez  fc_sign_seney  fc_sign_sosnowski  \\\n",
       "0              -1.0             1.0            1.0                1.0   \n",
       "1              -1.0            -1.0            1.0               -1.0   \n",
       "2              -1.0            -1.0            1.0                1.0   \n",
       "3              -1.0             1.0           -1.0               -1.0   \n",
       "4              -1.0             1.0            1.0                1.0   \n",
       "\n",
       "   wfisher_fc_sign  wfisher_pvalue  wfisher_adj_pvalue  \n",
       "0                1        0.091392            0.322133  \n",
       "1               -1        0.363189            0.609168  \n",
       "2               -1        0.403248            0.637638  \n",
       "3               -1        0.832877            0.935349  \n",
       "4                1        0.266319            0.525174  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>annotation_source</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>genomic_start_location</th>\n",
       "      <th>genomic_end_location</th>\n",
       "      <th>score(not_used)</th>\n",
       "      <th>genomic_strand</th>\n",
       "      <th>genomic_phase</th>\n",
       "      <th>additional_information</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; gene_type \"tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome_name annotation_source feature_type  genomic_start_location  \\\n",
       "0            chr1            HAVANA         gene                   11869   \n",
       "1            chr1            HAVANA   transcript                   11869   \n",
       "2            chr1            HAVANA         exon                   11869   \n",
       "3            chr1            HAVANA         exon                   12613   \n",
       "4            chr1            HAVANA         exon                   13221   \n",
       "\n",
       "   genomic_end_location score(not_used) genomic_strand genomic_phase  \\\n",
       "0                 14409               .              +             .   \n",
       "1                 14409               .              +             .   \n",
       "2                 12227               .              +             .   \n",
       "3                 12721               .              +             .   \n",
       "4                 14409               .              +             .   \n",
       "\n",
       "                              additional_information  \n",
       "0  gene_id \"ENSG00000223972.5_2\"; gene_type \"tran...  \n",
       "1  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  \n",
       "2  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  \n",
       "3  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  \n",
       "4  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import pandas as pd\n",
    "\n",
    "#df = pd.read_csv(\"meta_analysis_sumstats_no_singletons_20220727.tsv.gz\", compression=\"gzip\", sep=\"\\t\")\n",
    "#ann = pd.read_csv(\"gencode.v30lift37.annotation.gtf.gz\", sep=\"\\t\", compression=\"gzip\", skiprows=5, header=None)\n",
    "#ann.columns = [\"chromosome_name\", \"annotation_source\", \"feature_type\",\"genomic_start_location\",\"genomic_end_location\",\"score(not_used)\",\"genomic_strand\",\"genomic_phase\", \"additional_information\"]\n",
    "\n",
    "df.head()\n",
    "ann.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2d96fcc9-fa7d-4118-acaa-5a5eeeca8e79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ENSG00000223972': ['chr1', '11869', '14409'], 'ENSG00000227232': ['chr1', '14404', '29570']}\n",
      "ENSG00000270629\n",
      "ENSG00000274020\n",
      "ENSG00000275143\n",
      "ENSG00000277209\n",
      "ENSG00000278774\n",
      "ENSG00000280105\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "# https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/make_annot_sample_files/ENSG_coord.txt\n",
    "\n",
    "fdr = \"0.10\"\n",
    "in1 = \"meta_analysis_sumstats_no_singletons_20220727.tsv.gz\"\n",
    "out1 = f\"meta_analysis_sumstats_no_singletons_20220727_fdr{fdr}_coord.tsv\"\n",
    "annfile = \"gencode.v30lift37.annotation.gtf.gz\"\n",
    "\n",
    "with gzip.open(in1, 'rt') as inF, gzip.open(annfile, 'rt') as annF, open(out1, 'w') as outF:\n",
    "    for _ in range(5):\n",
    "        next(annF)\n",
    "    line = annF.readline()\n",
    "    \n",
    "    gencode_dic  = {}\n",
    "    while line:\n",
    "        sl = line.split(\"\\t\")\n",
    "        if sl[2] == \"gene\":\n",
    "            gencode = sl[8].split(\";\")[0] # remove all additional_info except \"gene_id <ENSG...>\"\n",
    "            gencode = gencode.split(\" \")[1].strip('\"') # remove \"gene_id\" portion, and double quotes\n",
    "            gencode = gencode.split(\".\")[0] # remove suffix  <ENSG...>\n",
    "            gencode_dic[gencode] = [sl[0], sl[3], sl[4]] # chr, start-, and end-genomic position\n",
    "        line = annF.readline()\n",
    "    \n",
    "    print(dict(list(gencode_dic.items())[0:2]))\n",
    "\n",
    "    outF.write(\"GENE\\tCHR\\tSTART\\tEND\\n\")\n",
    "    next(inF)\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        gencode = sl[0].split(\".\")[0]\n",
    "        wfisher_adj_p = sl[-1]\n",
    "        if float(wfisher_adj_p) < float(fdr):\n",
    "            if gencode in gencode_dic:\n",
    "                chrom = gencode_dic[gencode][0]\n",
    "                start = gencode_dic[gencode][1]\n",
    "                end = gencode_dic[gencode][2]\n",
    "                outline = f\"{gencode}\\t{chrom}\\t{start}\\t{end}\\n\"\n",
    "                outF.write(outline)\n",
    "            else:\n",
    "                print(gencode)\n",
    "        line = inF.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f153d332-64b3-44bc-b015-2ba331e5d5b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# extract just the gene\n",
    "tail -n +2 meta_analysis_sumstats_no_singletons_20220727_fdr0.05_coord.tsv | \\\n",
    "  cut -f1 > meta_analysis_sumstats_no_singletons_20220727_fdr0.05_geneset.tsv\n",
    "\n",
    "tail -n +2 meta_analysis_sumstats_no_singletons_20220727_fdr0.10_coord.tsv | \\\n",
    "  cut -f1 > meta_analysis_sumstats_no_singletons_20220727_fdr0.10_geneset.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53879ab8-34a3-4288-b227-b5f27e286ef5",
   "metadata": {},
   "source": [
    "## munge sumstats\n",
    "Some of the sumstats were already munged, so we will just use them when we can.\n",
    "Others we will have to download an munge.\n",
    "\n",
    "\n",
    "* ~Age of Initiation (Liu et al., 2019 Nat Genet 30643251)~\n",
    "* ~Alcohol Dependence (Walters et al., 2018 Nat Neurosci 30482948)~\n",
    "* ~Alcohol Drinks per Week (DPW) (Liu et al., 2019 Nat Genet 30643251)~\n",
    "* ~Alzheimer's Disease (Lambert et al., 2013 Nat Genet 24162737)~\n",
    "* ~Amyotrophic Lateral Sclerosis (Rheenen et al., 2016 Nat Genet 27455348)~\n",
    "* ~Anorexia Nervosa (Watson et al., 2019 Nat Genet 31308545)~\n",
    "* ~Attention Deficit Hyperactivity Disorder (Demontis et al., 2019 Nat Genet 30478444)~\n",
    "* ~Autism Spectrum Disorders (Grove et al., 2019 Nat Genet 30804558)~\n",
    "* ~Bipolar Disorder (Stahl et al., 2019 Nat Genet 31043756)~\n",
    "* ~Brain Volume: Mean Accumbens Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Mean Amygdala Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Mean Caudate Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Mean Hippocampus Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Mean Pallidum Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Mean Putamen Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Mean Thalamus Volume (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Brain Volume: Total Intracranial Volume (ICV) (Hibar et al., 2015 Nature 25607358)~\n",
    "* ~Cannabis Use Disorder (CUD) (Demontis et al., 2019 Nat Neurosci 31209380)~\n",
    "* ~Childhood IQ (Benyamin et al., 2014 Mol Psychiatry 23358156)~\n",
    "* ~Cigarettes Per Day (Liu et al., 2019 Nat Genet 30643251)~\n",
    "* ~College Completion (Rietveld et al., 2013 Science 23722424)~\n",
    "* ~Cotinine Levels (Ware et al., 2016 Sci Rep 26833182)~\n",
    "* ~Depressive Symptoms (Okbay et al., 2016 Nat Genet 27089181)~\n",
    "* ~Fagerstrom Test for Nicotine Dependence (FTND) (Quach et al., 2020 Nat Commun 33144568)~\n",
    "* ~Heaviness of Smoking Index (HSI) (Quach et al., 2020 Nat Commun 33144568)~\n",
    "* ~Intelligence (Sniekers et al., 2017 Nat Genet 28530673)~\n",
    "* ~Lifetime Cannabis Use (Ever vs. Never) (Pasman et al., 2018 Nat Neurosci 30150663)~\n",
    "* ~Major Depressive Disorder (Howard et al., 2018 Nat Commun 29662059)~\n",
    "* ~Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry 21173776)~\n",
    "* ~Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry 21173776)~\n",
    "* ~Neuroticism (Okbay et al., 2016 Nat Genet 27089181)~\n",
    "* ~Opioid Addiction: GENOA GWAS meta-analysis~\n",
    "* ~Opioid Addiction: gSEM OA GWAS meta-analysis (i.e., GENOA, MVP-SAGE-YP, PGC-SUD, and Partners Health)~\n",
    "* ~Parkinson's Disease (Sanchez et al., 2009 Nat Genet 19915575)~\n",
    "* ~Post-traumatic Stress Disorder (Nievergelt et al., 2019 Nat Commun 31594949)~\n",
    "* ~Psychiatric Genetics Consortium Cross-disorder GWAS (Schizophrenia, Bipolar Disorder, MDD, ASD and ADHD) (Cross-Disorder Group of the Psychiatric Genomics Consortium, 2013 Lancet 23453885)~\n",
    "* ~Schizophrenia (Ripke et al., 2014 Nature 25056061)~\n",
    "* ~Smoking Cessation (Liu et al., 2019 Nat Genet 30643251)~\n",
    "* ~Smoking Initiation (Liu et al., 2019 Nat Genet 30643251)~\n",
    "* ~Subjective Well Being (Okbay et al., 2016 Nat Genet 27089181)~\n",
    "* ~Years of Education (Okbay et al., 2022 Nature Genetics  35361970)~\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee7f402b-6d0a-4a6e-89f2-86bfeed367ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd sumstats/\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/AgeOfInitiation.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/alcohol_dependence_walters2018_nat_neurosci/munged/pgc_alcdep.eur_discovery.aug2018_release.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/DrinksPerWeek.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/alzheimers_disease_lambert2013_nat_genet/munged/alzheimers_disease_lambert2013_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet/munged/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/anorexia_watson2019_nat_genet/munged/anorexia_watson2019_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/adhd_demontis2018_nat_genet/munged/daner_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.meta.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/autism_spectrum_disorder_grove2019_nat_genet/munged/iPSYCH-PGC_ASD_Nov2017.munged.merged.txt.gz . \n",
    "aws s3 cp s3://rti-shared/ldsc/data/bipolar_disorder_stahl2019_nat_genet/munged/daner_PGC_BIP32b_mds7a_0416a.munged.merged.txt.gz .\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Accumbens Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Caudate Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Hippocampus Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Pallidum Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Putamen Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Thalamus Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.sumstats.gz .\n",
    "#17\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/cannabis_use_disorder_demontis2019_nat_neurosci/munged/CUD_GWAS_iPSYCH_June2019.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/childhood_intelligence_benyamin2014_mol_psych/munged/CHIC_Summary_Benyamin2014.sumstats.gz . # Childhood IQ s3://rti-shared/gwas_publicly_available_sumstats/childhood_intelligence_benyamin2014_mol_psych/raw/CHIC_Summary_Benyamin2014.txt.gz\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/CigarettesPerDay.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/educational_attainment_rietveld2013_science/munged/SSGAC_College_Rietveld2013_publicrelease.sumstats.gz . # College completion \n",
    "aws s3 cp s3://rti-shared/ldsc/data/cotinine_levels_ware2016_sci_rep/munged/cotinine_ware2016_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/depressive_symptoms_okbay2016/munged/DS_Full.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/nicotine_dependence_quach2020_nat_commun/munged/ftnd_wave3_eur_quach2020_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/ukb_hsi/munged/ukb_gwa_003_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/intelligence_sniekers2017_nat_genet/munged/intelligence_sniekers2017_nat_genet_sumstats_formatted.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/lifetime_cannabis_use_pasman2018_nat_neurosci/munged/cannabis_icc_ukb_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/major_depressive_disorder_howard2018_nat_commun/munged/pgc_ukb_depression_gwas_workflow_ready.txt.munged.merged.txt.gz .\n",
    "#28\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.sumstats.gz . # Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry 21173776)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/GPC-1.NEO-OPENNESS.full.with_header.hg19.sumstats.gz .# Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry 21173776)\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/neuroticism_okbay2016_nat_genet/munged/neuroticism_okbay2016_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/opioid_addiction_gaddis_mathur2022_sci_rep/munged/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted.sumstats.gz .\n",
    "aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/GenomicSEM/results/29/gSEM/final/munged/genomicSEM_GWAS.oaALL.MVP1_MVP2_YP_SAGE.PGC.Song.table.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/parkinsons_disease_sanchez2009_nat_genet/munged/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/ptsd_nievergelt2019_nat_commun/munged/pts_eur_freeze2_overall.results.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/cross_disorder_gwas_pgc2013_lancet/munged/pgc.cross.full.2013-03.hg19.sumstats.gz . # pgc needs a liftover from hg18 (s3://rti-shared/gwas_publicly_available_sumstats/cross_disorder_gwas_pgc2013_lancet/raw/)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/schizophrenia_ripke2014_nature/munged/daner_natgen_pgc_eur.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/SmokingCessation.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/SmokingInitiation.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/subjective_wellbeing_okbay2016_nat_genet/munged/SWB_Full.sumstats.gz . # Subjective Well Being (Okbay et al., 2016 Nat Genet 27089181)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/years_schooling_okbay2022_nat_genet/munged/EA4_additive_excl_23andMe.sumstats.gz . # Years of Education (Okbay et al., 2022 Nature Genetics  35361970)\n",
    "#41"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c4bf15e-4c0c-458f-8a4a-aba847bdebb1",
   "metadata": {},
   "source": [
    "### Opioid Addiction\n",
    "initiated restore 11/2/2022"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42e05aed-f552-455e-a817-a52d91ee19bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use ldsc tool munge_sumstats.py to convert to sumstats format (https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format)\n",
    "\n",
    "s3://rti-heroin/rti-midas-data/studies/ngc/meta/144/processing/\n",
    "for chr in {1..22}; do\n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/144/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.8.tsv.gz .\n",
    "done\n",
    "\n",
    "# combine into 1 file\n",
    "# keep only SNPs with rsIDs, and just keep the rsID portion of MarkerName\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.8.tsv.gz  | head -1 > \\\n",
    "    cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.tsv\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    awk '{split($1,a,\":\")} \n",
    "        {\n",
    "            $1=a[1] \n",
    "            b=substr($1,1,2)\n",
    "            {if (b==\"rs\") \n",
    "                {print $0}\n",
    "            }\n",
    "        }' OFS=\"\\t\" <(zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.8.tsv.gz | tail -n +2) \\\n",
    "            >> cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.tsv\n",
    "done\n",
    "\n",
    "\n",
    "# munge:  docker interactive mode\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.tsv \\\n",
    "    --snp MarkerName \\\n",
    "    --N-cas 7281 \\\n",
    "    --N-con 297550 \\\n",
    "    --a1 Allele1 \\\n",
    "    --a2 Allele2 \\\n",
    "    --p P-value \\\n",
    "    --signed-sumstats Effect,0 \\\n",
    "    --out cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted\n",
    "\n",
    "\n",
    "    \n",
    "# upload to s3\n",
    "aws s3 cp cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted.sumstats.gz s3://rti-shared/ldsc/data/opioid_addiction_gaddis_mathur2022_sci_rep/munged/\n",
    "aws s3 cp cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted.log s3://rti-shared/ldsc/data/opioid_addiction_gaddis_mathur2022_sci_rep/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abd7242a-7c47-4bf7-9f7b-bfa1fb15846c",
   "metadata": {},
   "source": [
    "### Brain Volume\n",
    "use ldsc tool munge_sumstats.py to convert to sumstats format (https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3dacd20-6e3e-481d-ba42-d0b5a52ccb3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Accumbens Volume (Hibar et al., 2015 Nature 25607358)\n",
    "\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.gz  | head\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 0.9454 1.1303 0.4029 13112\n",
    "\n",
    "# interactive mode\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5406b241-7110-4356-98b4-37ab349c7460",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Amygdala Volume (Hibar et al., 2015 Nature 25607358)\n",
    "\n",
    "aws s3 cp  s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.gz  | head -2\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 2.3536 2.4545 0.3376 13160\n",
    "\n",
    "# interactive mode\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb62161b-1d08-433d-b8a8-afa82039cad2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Caudate Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.gz  | head\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 3.1005 5.1190 0.5447 13171\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f54e7e62-67bd-4b92-9603-d7a5872696b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Hippocampus Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.gz | head -2\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 -7.4896 4.9232 0.1282 13163\n",
    "\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "    \n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9a0cab7-a3a9-4b77-adac-f08fd695a9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Pallidum Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.gz  | head -2\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 -3.0672 2.0149 0.1279 13142\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eeb0c70-10a0-48c3-aff2-5005554a478b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Putamen Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.gz  | head\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 -3.2910 6.2791 0.6002 13145\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "# note that I had to manually edit the munge_sumstats.py file, in particular I substituted line 1 for line 2. This was because I was getting an error saying (ValueError: WARNING: median value of SIGNED_SUMSTATS is 0.13 (should be close to 0.0). This column may be mislabeled.). Raymond Walters suggested lessening the tolerance threshold a bit. see https://groups.google.com/g/ldsc_users/c/RLbVw3e_PU0\n",
    "# verifying median value in R > median(df$Effect_Beta) [1] 0.129\n",
    "# 1. check_median(dat.SIGNED_SUMSTAT, signed_sumstat_null, 0.10, sign_cname))\n",
    "# 2. check_median(dat.SIGNED_SUMSTAT, signed_sumstat_null, 0.15, sign_cname))\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61653270-36a2-4897-aae9-41bef6b5ac57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean Thalamus Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.gz  | head -2\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 -3.0636 6.5794 0.6415 13193\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f5f162f-6ec9-4cce-8ed9-90a9a9950990",
   "metadata": {},
   "source": [
    "####  Total Intracranial Volume: to-do\n",
    "\n",
    "\n",
    "There were issues with this that I need to resolve.\n",
    "```\n",
    "I attempted to perform LDSC with this set of GWAS results. However, when I attempted to munge the sumstats I received the following error message:\n",
    "\n",
    "\n",
    "ValueError: WARNING: median value of SIGNED_SUMSTATS is -22.41 (should be close to 0.0). This column may be mislabeled.\n",
    "\n",
    " \n",
    "\n",
    "I calculated this in R to verify (note it was not exactly the same but that is because the munge_sumstats.py script provided by the LDSC repo removed some SNPs during QC):\n",
    "\n",
    "> df <- read.table(\"ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.gz\", header=TRUE)\n",
    "\n",
    "> median(df$Effect_Beta).   # [1] -20.121\n",
    "\n",
    " \n",
    "\n",
    "Do you know why the Effect_Beta column would not be close to 0? Here is a sample of the file â€“ it looks like the beta values are way away from 0 and the SE is huge.\n",
    "\n",
    "zcat ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.gz | head\n",
    "\n",
    "RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "\n",
    "rs667647 5:29439275 T C 0.347 -148.8340 2029.8618 0.9415 11373\n",
    "\n",
    "rs113534962 5:85928892 T C 0.05013 6239.4568 4368.8262 0.1532 11373\n",
    "\n",
    "rs2366866 10:128341232 T C 0.4578 1025.8514 1895.1917 0.5883 11373\n",
    "\n",
    "rs472303 3:62707519 T C 0.06332 5058.2986 4040.6084 0.2106 11373\n",
    "```\n",
    "\n",
    "**response**:\n",
    "```\n",
    "This is because the GWAS was run in the raw scale of the data not on residuals or z scores.\n",
    "The variable you are looking at is the basically volume the brain in mm3, as a consequence all beta and se are scaled accordingly\n",
    "```\n",
    "\n",
    "\n",
    "<br>\n",
    "\n",
    "**What we did**:\n",
    "Edited the script `munge_sumstats.py` so that it would not throw an error when checking the median."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c06aa04f-6f42-4e74-9ea7-49ebb73e60ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Total Intracranial Volume (ICV) (Hibar et al., 2015 Nature 25607358) \n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/brain_volume_hibar2015_nature/ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.gz .\n",
    "\n",
    "zcat ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.gz  | head -2\n",
    "#RSID CHR_BP_hg19b37 Effect_Allele Non_Effect_Allele Freq_European_1000Genomes Effect_Beta StdErr Pvalue N\n",
    "#rs667647 5:29439275 T C 0.347 -148.8340 2029.8618 0.9415 11373\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "\n",
    "# we edited tolerance in munge_sumstats.py so this would pass QC\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.gz \\\n",
    "    --snp RSID \\\n",
    "    --N-col N \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Non_Effect_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats Effect_Beta,0 \\\n",
    "    --out ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl\n",
    "\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.sumstats.gz s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/\n",
    "aws s3 cp ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.log s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b2be437-d795-496f-9ca2-ab1028171758",
   "metadata": {},
   "source": [
    "### Childhood IQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b5c441f-50eb-4406-b0ae-cb18e19b9315",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/childhood_intelligence_benyamin2014_mol_psych/raw/CHIC_Summary_Benyamin2014.txt.gz .\n",
    "\n",
    "zcat CHIC_Summary_Benyamin2014.txt.gz  | head\n",
    "SNP CHR BP A1 A2 FREQ_A1 EFFECT_A1 SE P\n",
    "#rs1000000 chr12 125415860 A G 0.373 -0.0143 0.0156 0.3604\n",
    "#rs10000010 chr4 21294943 T C 0.575 0.0028 0.0127 0.8237\n",
    "\n",
    "\n",
    "# munge:  docker interactive mode\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats CHIC_Summary_Benyamin2014.txt.gz \\\n",
    "    --snp SNP \\\n",
    "    --N 17989 \\\n",
    "    --a1 A1 \\\n",
    "    --a2 A2 \\\n",
    "    --p P \\\n",
    "    --signed-sumstats EFFECT_A1,0 \\\n",
    "    --out CHIC_Summary_Benyamin2014\n",
    "    \n",
    "# upload to s3\n",
    "aws s3 cp CHIC_Summary_Benyamin2014.sumstats.gz s3://rti-shared/ldsc/data/childhood_intelligence_benyamin2014_mol_psych/munged/\n",
    "aws s3 cp CHIC_Summary_Benyamin2014.log s3://rti-shared/ldsc/data/childhood_intelligence_benyamin2014_mol_psych/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3683f7d7-7e71-4a00-9d25-956b297400af",
   "metadata": {},
   "source": [
    "### College Completion\n",
    "Rietvald Science 2013\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "423ca40e-af72-42f6-abc2-8beb732b68ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/educational_attainment_rietveld2013_science/raw/SSGAC_Rietveld2013.zip .\n",
    "\n",
    "unzip SSGAC_Rietveld2013.zip\n",
    "\n",
    "head SSGAC_College_Rietveld2013_publicrelease.txt\n",
    "#MarkerName      Effect_Allele   Other_Allele    EAF     OR      SE      Pvalue\n",
    "#rs3813193       C       G       0.16    0.985   0.014   0.2823\n",
    "#rs4075116       T       C       0.78    1.018   0.012   0.1199\n",
    "\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats SSGAC_College_Rietveld2013_publicrelease.txt \\\n",
    "    --snp MarkerName \\\n",
    "    --N 95427 \\\n",
    "    --a1 Effect_Allele  \\\n",
    "    --a2 Other_Allele \\\n",
    "    --p Pvalue \\\n",
    "    --signed-sumstats OR,1 \\\n",
    "    --out SSGAC_College_Rietveld2013_publicrelease\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp SSGAC_College_Rietveld2013_publicrelease.sumstats.gz s3://rti-shared/ldsc/data/educational_attainment_rietveld2013_science/munged/\n",
    "aws s3 cp SSGAC_College_Rietveld2013_publicrelease.log s3://rti-shared/ldsc/data/educational_attainment_rietveld2013_science/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47270e25-e871-4dea-96c0-3f771bbdf701",
   "metadata": {},
   "source": [
    "### Cross-Disorder (PGC)\n",
    "build 36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13aec680-b516-4364-9401-71dba55e7503",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/cross_disorder_gwas_pgc2013_lancet/raw/pgc.cross.full.2013-03.zip .\n",
    "\n",
    "unzip pgc.cross.full.2013-03.zip\n",
    "\n",
    "head pgc.cross.full.2013-03.txt\n",
    "#snpid hg18chr bp a1 a2 or se pval info ngt CEUaf\n",
    "#rs3131972       1       742584  A       G       1.047   0.0361  0.2032  0.774   0       0.16055\n",
    "#rs3131969       1       744045  A       G       1.067   0.0377  0.08597 0.939   0       0.133028\n",
    "\n",
    "# need to convert to build 37\n",
    "# use liftover workflow: liftover_genomic_annotations\n",
    "git clone --recurse-submodules https://github.com/RTIInternational/biocloud_gwas_workflows\n",
    "\n",
    "# once liftover completed, now we can use munge_sumstats.py\n",
    "\n",
    "head pgc.cross.full.2013-03.hg19.txt\n",
    "#snpid chr GRCh37_POS a1 a2 or se pval info ngt CEUaf\n",
    "#rs3131972 1 752721 A G 1.047 0.0361 0.2032 0.774 0 0.16055\n",
    "\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats pgc.cross.full.2013-03.hg19.txt \\\n",
    "    --snp snpid \\\n",
    "    --N-cas 33332 \\\n",
    "    --N-con 27888 \\\n",
    "    --a1 a1  \\\n",
    "    --a2 a2 \\\n",
    "    --p pval \\\n",
    "    --signed-sumstats or,1 \\\n",
    "    --out pgc.cross.full.2013-03.hg19\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp pgc.cross.full.2013-03.hg19.sumstats.gz s3://rti-shared/ldsc/data/cross_disorder_gwas_pgc2013_lancet/munged/\n",
    "aws s3 cp pgc.cross.full.2013-03.hg19.log s3://rti-shared/ldsc/data/cross_disorder_gwas_pgc2013_lancet/munged/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8139ec9-ba60-4743-8af6-6875bfde6e22",
   "metadata": {},
   "source": [
    "### Intelligence\n",
    "Sniekers et al., 2017 Nat Genet\t28530673\n",
    "\n",
    "```\n",
    "## Association results of the meta-analysis for intelligence based on 78,308 individuals in 13 cohorts. \n",
    "\n",
    "## Version date: 10-07-2017\n",
    "\n",
    "#Columns:\n",
    "Chromosome: chromosome number\n",
    "position: base pair position of the SNP on the chromosome (reported on GRCh37)\n",
    "rsid: SNP rs number\n",
    "ref: effect allele\n",
    "alt: non-effect allele\n",
    "N: sample size\n",
    "MAF: minor allele frequency in UK Biobank\n",
    "Beta: effect size of the effect allele\n",
    "SE: standard error of the effect\n",
    "Zscore: Z-score computed in METAL by a weighted Z-score method\n",
    "p_value: P-value computed in METAL by a weighted Z-score method\n",
    "direction: direction of the effect in each of the cohorts, order: CHIC (consisting of 6 cohorts), UKB-wb, UKB-ts, ERF, GENR, HU, MCTFR, STR\n",
    "\n",
    "Beta/SE were calculated from METAL Z-scores using the formula from Zhu et al (Nature Genetics, 2016):\n",
    "\n",
    "Beta = Zscore / sqrt( 2 * MAF * ( 1 - MAF) * ( N + Zscore^2 ) )\n",
    "SE = 1 / sqrt( 2 * MAF * ( 1 - MAF ) * ( N + Zscore^2 ) )\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60d5e480-e047-4cc7-b2ee-fc2965d10c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use ldsc tool munge_sumstats.py to convert to sumstats format (https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format)\n",
    "\n",
    "# intelligence s3://rti-shared/gwas_publicly_available_sumstats/intelligence_sniekers2017_nat_genet/raw/sumstats.txt.gz\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/intelligence_sniekers2017_nat_genet/raw/sumstats.txt.gz .\n",
    "zcat sumstats.txt.gz  | head\n",
    "#Chromosome      position        rsid    ref     alt     MAF     Beta    SE      Zscore  p_value direction\n",
    "#1       100000012       rs10875231      T       G       0.234588        0.000298163293384453    0.00596326586768906     0.05    0.9599  +-++--+-\n",
    "\n",
    "# interactive mode\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash \\\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats sumstats.txt.gz \\\n",
    "    --snp rsid \\\n",
    "    --N 78308 \\\n",
    "    --a1 ref \\\n",
    "    --a2 alt \\\n",
    "    --p p_value \\\n",
    "    --signed-sumstats Beta,0 \\\n",
    "    --out intelligence_sniekers2017_nat_genet_sumstats_formatted\n",
    "\n",
    "\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp intelligence_sniekers2017_nat_genet_sumstats_formatted.log s3://rti-shared/ldsc/data/intelligence_sniekers2017_nat_genet/munged/\n",
    "aws s3 cp intelligence_sniekers2017_nat_genet_sumstats_formatted.sumstats.gz s3://rti-shared/ldsc/data/intelligence_sniekers2017_nat_genet/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "200ef86e-8cf3-45bb-bb6b-7ffe4b2ad23a",
   "metadata": {},
   "source": [
    "### Personality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bc84848-9ffc-40a3-b54f-a88f3cb18ebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry 21173776)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/personality_demoor_2012_mol_psych/raw/GPC-1.NEO-CONSCIENTIOUSNESS.full.txt .\n",
    "\n",
    "# Need to add headers - they were missing. The readme says \"The files contain the following information (columns): SNPID CHR BP A1 A2 BETA SE PVALUE INFO NCOH MAF\"\n",
    "echo -e \"SNPID\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tSE\\tPVALUE\\tINFO\\tNCOH\\tMAF\" > GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.txt\n",
    "cat GPC-1.NEO-CONSCIENTIOUSNESS.full.txt >> GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.txt\n",
    "\n",
    "# perform liftover to build37 (currently build 36)\n",
    "# use liftover workflow: liftover_genomic_annotations\n",
    "git clone --recurse-submodules https://github.com/RTIInternational/biocloud_gwas_workflows\n",
    "\n",
    "# once liftover completed, now we can use munge_sumstats.py\n",
    "\n",
    "zcat GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.txt.gz | head -2\n",
    "#SNPID CHR HG19_POS A1 A2 BETA SE PVALUE INFO NCOH MAF\n",
    "#rs3121561 1 990380 t c -.0144 .1282 .9105 .531 8 .267\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.txt.gz \\\n",
    "    --snp SNPID \\\n",
    "    --N 17375 \\\n",
    "    --a1 A1  \\\n",
    "    --a2 A2 \\\n",
    "    --p PVALUE \\\n",
    "    --signed-sumstats BETA,0 \\\n",
    "    --out GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.sumstats.gz s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/\n",
    "aws s3 cp GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.log s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09f82517-dc42-4789-a4f9-34399a0de82b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry 21173776)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/personality_demoor_2012_mol_psych/raw/GPC-1.NEO-OPENNESS.full.txt .\n",
    "\n",
    "# Need to add headers - they were missing. The readme says \"The files contain the following information (columns): SNPID CHR BP A1 A2 BETA SE PVALUE INFO NCOH MAF\"\n",
    "echo -e \"SNPID\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tSE\\tPVALUE\\tINFO\\tNCOH\\tMAF\" > GPC-1.NEO-OPENNESS.full.with_header.txt\n",
    "cat GPC-1.NEO-OPENNESS.full.txt >> GPC-1.NEO-OPENNESS.full.with_header.txt\n",
    "\n",
    "# perform liftover to build37 (currently build 36)\n",
    "# use liftover workflow: liftover_genomic_annotations\n",
    "git clone --recurse-submodules https://github.com/RTIInternational/biocloud_gwas_workflows\n",
    "\n",
    "# once liftover completed, now we can use munge_sumstats.py\n",
    "\n",
    "zcat GPC-1.NEO-OPENNESS.full.with_header.hg19.txt.gz | head -2\n",
    "#SNPID CHR HG19_POS A1 A2 BETA SE PVALUE INFO NCOH MAF\n",
    "#rs3121561 1 990380 t c -.0259 .1199 .8293 .536 8 .267\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats GPC-1.NEO-OPENNESS.full.with_header.hg19.txt.gz \\\n",
    "    --snp SNPID \\\n",
    "    --N 17375 \\\n",
    "    --a1 A1  \\\n",
    "    --a2 A2 \\\n",
    "    --p PVALUE \\\n",
    "    --signed-sumstats BETA,0 \\\n",
    "    --out GPC-1.NEO-OPENNESS.full.with_header.hg19\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp GPC-1.NEO-OPENNESS.full.with_header.hg19.sumstats.gz s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/\n",
    "aws s3 cp GPC-1.NEO-OPENNESS.full.with_header.hg19.log s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b754a8b-158e-4d07-b5ac-c985ed019d22",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/personality_demoor_2012_mol_psych/raw/GPC-1.NEO-EXTRAVERSION.full.txt ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d9e601b-bb9f-4733-a056-b03cb554b609",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/personality_demoor_2012_mol_psych/raw/GPC-1.NEO-AGREEABLENESS.full.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6109ff42-92ec-4986-872d-c54beffb8f13",
   "metadata": {},
   "source": [
    "### Subjective Wellbeing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4f78b9d-f843-43b4-88de-160ea1f3dc32",
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/subjective_wellbeing_okbay2016_nat_genet/raw/SWB_Full.txt.gz .\n",
    "\n",
    "zcat SWB_Full.txt.gz  | head -2\n",
    "#MarkerName      CHR     POS     A1      A2      EAF     Beta    SE      Pval\n",
    "#rs2075677       20      47701024        A       G       0.7743  0.021   0.004   1.879e-08\n",
    "\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats SWB_Full.txt.gz \\\n",
    "    --snp MarkerName \\\n",
    "    --N 298420 \\\n",
    "    --a1 A1  \\\n",
    "    --a2 A2\\\n",
    "    --p Pval \\\n",
    "    --signed-sumstats Beta,0 \\\n",
    "    --out SWB_Full\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp SWB_Full.sumstats.gz s3://rti-shared/ldsc/data/subjective_wellbeing_okbay2016_nat_genet/munged/\n",
    "aws s3 cp SWB_Full.log s3://rti-shared/ldsc/data/subjective_wellbeing_okbay2016_nat_genet/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7272397a-c402-451a-8b70-a58ee60d25a3",
   "metadata": {},
   "source": [
    "###  Years of Education \n",
    "(Okbay et al., 2022 Nature Genetics  35361970)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06684032-5638-46a8-b45b-874efc225638",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use ldsc tool munge_sumstats.py to convert to sumstats format (https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format)\n",
    "\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/years_schooling_okbay2022_nat_genet/raw/EA4_additive_excl_23andMe.txt.gz .\n",
    "\n",
    "zcat EA4_additive_excl_23andMe.txt.gz  | head -2\n",
    "#rsID    Chr     BP      Effect_allele   Other_allele    EAF_HRC Beta    SE      SE_unadj        P       P_unadj\n",
    "#rs667647        5       29439275        T       C       0.376548        -0.00032        0.00179 0.00167 0.86    0.8504\n",
    "\n",
    "docker run -it -v $PWD:/data/ rtibiocloud/ldsc:v1.0.1_9501d4d bash\n",
    "python /opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats EA4_additive_excl_23andMe.txt.gz \\\n",
    "    --snp rsID \\\n",
    "    --N 765283 \\\n",
    "    --a1 Effect_allele  \\\n",
    "    --a2 Other_allele \\\n",
    "    --p P \\\n",
    "    --signed-sumstats Beta,0 \\\n",
    "    --out EA4_additive_excl_23andMe\n",
    "\n",
    "# upload to s3\n",
    "aws s3 cp EA4_additive_excl_23andMe.sumstats.gz s3://rti-shared/ldsc/data/years_schooling_okbay2022_nat_genet/munged/\n",
    "aws s3 cp EA4_additive_excl_23andMe.log s3://rti-shared/ldsc/data/years_schooling_okbay2022_nat_genet/munged/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45adb2cf-21c1-4d0d-bc58-f3fd963485fb",
   "metadata": {},
   "source": [
    "# Stratified LDSC\n",
    "https://github.com/bulik/ldsc/wiki/LD-Score-Estimation-Tutorial#partitioned-ld-scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d53343d-291c-45a8-acf7-4c2fa4396526",
   "metadata": {},
   "outputs": [],
   "source": [
    "# download files needed for partitioned heritability analysis\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_phase3_baseline_ldscores.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_plinkfiles.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_frq.tgz\n",
    "\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/weights_hm3_no_hla.tgz\n",
    "#wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/hapmap3_snps.tgz\n",
    "\n",
    "# extract files\n",
    "tar -xvf 1000G_Phase3_baseline_ldscores.tgz\n",
    "tar -xvf 1000G_Phase3_plinkfiles.tgz\n",
    "tar -xvf 1000G_Phase3_frq.tgz\n",
    "#tar -xvf hapmap3_snps.tgz\n",
    "tar -xvf weights_hm3_no_hla.tgz\n",
    "\n",
    "\n",
    "# interactive session\n",
    "docker run -it -v $PWD:/data/ \\\n",
    "    rtibiocloud/ldsc:v1.0.1_0bb574e bash\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf6f7fb-51fd-4950-b7cb-dc7583c23614",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "window=100000\n",
    "for fdr in {\"0.05\",\"0.10\"}; do # loop through each BED file\n",
    "    coord_file=/data/deg_bedfiles/meta_analysis_sumstats_no_singletons_20220727_fdr${fdr}_coord.tsv \n",
    "    geneset_file=/data/deg_bedfiles/meta_analysis_sumstats_no_singletons_20220727_fdr${fdr}_geneset.tsv\n",
    "\n",
    "    # store processing files for each meta in separate dir\n",
    "    mkdir -p /data/{annotations_ldscores,results}/fdr$fdr/\n",
    "\n",
    "    for j in {1..22}; do # loop through each chromosome\n",
    "        python /opt/ldsc/make_annot.py \\ # create annotation files\n",
    "            --gene-set-file $geneset_file \\\n",
    "            --gene-coord-file $coord_file \\\n",
    "            --windowsize $window \\\n",
    "            --bimfile /data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim \\\n",
    "            --annot-file /data/annotations_ldscores/fdr$fdr/oa_twas_meta_fdr${fdr}genes_window${window}_chr$j.annot.gz\n",
    "\n",
    "        python /opt/ldsc/ldsc.py \\ # compute LD scores\n",
    "            --l2 \\\n",
    "            --thin-annot \\\n",
    "            --ld-wind-cm 1 \\\n",
    "            --print-snps /data/1000g/1000G_EUR_Phase3_baseline/print_snps.txt \\\n",
    "            --bfile /data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j \\\n",
    "            --annot /data/annotations_ldscores/fdr$fdr/oa_twas_meta_fdr${fdr}genes_window${window}_chr$j.annot.gz \\\n",
    "            --out /data/annotations_ldscores/fdr$fdr/oa_twas_meta_fdr${fdr}genes_window${window}_chr$j\n",
    "    done # end chr loop\n",
    "\n",
    "    for trait in {\"adhd\",\"age_of_initiation\",\"alcohol_dependence\",\"alzheimers_disease\",\"amyotrophic_lateral_sclerosis\",\"anorexia\",\"autism\",\"bipolar\",\"brain_volume_mean_accumbens\",\"brain_volume_mean_amygdala\",\"brain_volume_mean_caudate\",\"brain_volume_mean_hippocampus\",\"brain_volume_mean_pallidum\",\"brain_volume_mean_putamen\",\"brain_volume_mean_thalamus\",\"brain_volume_total_intracranial\",\"cannabis_use_disorder\",\"childhood_intelligence\",\"cigarettes_per_day\",\"college_completion\",\"cotinine_levels\",\"cross_disorder\",\"depressive_symptoms\",\"drinks_per_week\",\"ftnd\",\"heaviness_smoking_index\",\"intelligence\",\"lifetime_cannabis_use\",\"major_depressive_disorder\",\"neo_conscientiousness\",\"neo_openness\",\"neuroticism\",\"opioid_addiction_144\",\"opioid_addiction_gsem\",\"parkinsons\",\"ptsd\",\"schizophrenia\",\"smoking_cessation\",\"smoking_initiation\",\"subjective_wellbeing\",\"years_of_education\"}; do # loop through all traits\n",
    "        case $trait in  # use sumstats files that corresponds to the trait name for the h2 estimate\n",
    "            \"adhd\") stats=/data/sumstats/daner_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.meta.munged.merged.txt.gz ;;\n",
    "            \"age_of_initiation\") stats=/data/sumstats/AgeOfInitiation.txt.munged.merged.txt.gz ;;\n",
    "            \"alcohol_dependence\") stats=/data/sumstats/pgc_alcdep.eur_discovery.aug2018_release.txt.munged.merged.txt.gz ;;\n",
    "            \"alzheimers_disease\") stats=/data/sumstats/alzheimers_disease_lambert2013_nat_genet.sumstats.gz ;;\n",
    "            \"amyotrophic_lateral_sclerosis\") stats=/data/sumstats/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz ;;\n",
    "            \"anorexia\") stats=/data/sumstats/anorexia_watson2019_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"autism\") stats=/data/sumstats/iPSYCH-PGC_ASD_Nov2017.munged.merged.txt.gz ;;\n",
    "            \"bipolar\") stats=/data/sumstats/daner_PGC_BIP32b_mds7a_0416a.munged.merged.txt.gz ;;\n",
    "            \"brain_volume_mean_accumbens\") stats=/data/sumstats/ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_amygdala\") stats=/data/sumstats/ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_caudate\") stats=/data/sumstats/ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_hippocampus\") stats=/data/sumstats/ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_pallidum\") stats=/data/sumstats/ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_putamen\") stats=/data/sumstats/ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_thalamus\") stats=/data/sumstats/ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_total_intracranial\") stats=/data/sumstats/ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"cannabis_use_disorder\") stats=/data/sumstats/CUD_GWAS_iPSYCH_June2019.munged.merged.txt.gz ;;\n",
    "            \"childhood_intelligence\") stats=/data/sumstats/CHIC_Summary_Benyamin2014.sumstats.gz ;;\n",
    "            \"cigarettes_per_day\") stats=/data/sumstats/CigarettesPerDay.txt.munged.merged.txt.gz ;;\n",
    "            \"college_completion\") stats=/data/sumstats/SSGAC_College_Rietveld2013_publicrelease.sumstats.gz ;;\n",
    "            \"cotinine_levels\") stats=/data/sumstats/cotinine_ware2016_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"cross_disorder\") stats=/data/sumstats/pgc.cross.full.2013-03.hg19.sumstats.gz ;;\n",
    "            \"depressive_symptoms\") stats=/data/sumstats/DS_Full.txt.munged.merged.txt.gz ;;\n",
    "            \"drinks_per_week\") stats=/data/sumstats/DrinksPerWeek.txt.munged.merged.txt.gz ;;\n",
    "            \"ftnd\") stats=/data/sumstats/ftnd_wave3_eur_quach2020_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"heaviness_smoking_index\") stats=/data/sumstats/ukb_gwa_003_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"intelligence\") stats=/data/sumstats/intelligence_sniekers2017_nat_genet_sumstats_formatted.sumstats.gz ;;\n",
    "            \"lifetime_cannabis_use\") stats=/data/sumstats/cannabis_icc_ukb_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"major_depressive_disorder\") stats=/data/sumstats/pgc_ukb_depression_gwas_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"neo_conscientiousness\") stats=/data/sumstats/GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.sumstats.gz ;;\n",
    "            \"neo_openness\") stats=/data/sumstats/GPC-1.NEO-OPENNESS.full.with_header.hg19.sumstats.gz ;;\n",
    "            \"neuroticism\") stats=/data/sumstats/neuroticism_okbay2016_nat_genet.sumstats.gz ;;\n",
    "            \"opioid_addiction_144\") stats=/data/sumstats/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted.sumstats.gz ;;\n",
    "            \"opioid_addiction_gsem\") stats=/data/sumstats/genomicSEM_GWAS.oaALL.MVP1_MVP2_YP_SAGE.PGC.Song.table.sumstats.gz ;;\n",
    "            \"parkinsons\") stats=/data/sumstats/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz ;;\n",
    "            \"ptsd\") stats=/data/sumstats/pts_eur_freeze2_overall.results.munged.merged.txt.gz ;;\n",
    "            \"schizophrenia\") stats=/data/sumstats/daner_natgen_pgc_eur.munged.merged.txt.gz ;;\n",
    "            \"smoking_cessation\") stats=/data/sumstats/SmokingCessation.txt.munged.merged.txt.gz ;;\n",
    "            \"smoking_initiation\") stats=/data/sumstats/SmokingInitiation.txt.munged.merged.txt.gz ;;\n",
    "            \"subjective_wellbeing\") stats=/data/sumstats/SWB_Full.sumstats.gz ;;\n",
    "            \"years_of_education\") stats=/data/sumstats/EA4_additive_excl_23andMe.sumstats.gz ;;\n",
    "        esac\n",
    "\n",
    "        # computed partitioned heritability estimate\n",
    "        python /opt/ldsc/ldsc.py \\\n",
    "            --h2 $stats \\\n",
    "            --overlap-annot \\\n",
    "            --print-coefficients \\\n",
    "            --w-ld-chr \"/data/weights_hm3_no_hla/weights.\" \\\n",
    "            --frqfile-chr \"/data/1000g/1000G_Phase3_frq/1000G.EUR.QC.\" \\\n",
    "            --ref-ld-chr \"/data/annotations_ldscores/fdr$fdr/oa_twas_meta_fdr${fdr}genes_window${window}_chr,/data/1000g/1000G_EUR_Phase3_baseline/baseline.\" \\\n",
    "            --out \"/data/results/fdr$fdr/${trait}_with_oa_twas_meta_analysis_deg_genes_fdr${fdr}_window${window}\"\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "326f5ee5-b18f-4633-a609-347ced4a5495",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd results/\n",
    "for fdr in {\"0.05\",\"0.10\"}; do\n",
    "    outfile=fdr${fdr}/all_phenotypes_oa_twas_meta_analysis_deg_fdr${fdr}_window100000_final_results.tsv\n",
    "    touch $outfile\n",
    "    head -1 fdr${fdr}/smoking_initiation_with_oa_twas_meta_analysis_deg_genes_fdr${fdr}_window100000.results > $outfile\n",
    "        \n",
    "    for file in   fdr${fdr}/*_fdr${fdr}_window100000.results; do\n",
    "        trait=$(echo $file |  sed \"s/_with_oa_twas_meta_analysis_deg_genes_fdr.*//\") # remove suffix\n",
    "        trait=$(echo $trait |  sed \"s/fdr$fdr\\///\") # remove directory prefix\n",
    "        #echo $trait\n",
    "        awk -v trait=$trait \\\n",
    "        '$1 = trait {print $0}' OFS=\"\\t\" <(tail -n +2 $file | head -1) >> $outfile\n",
    "    done\n",
    "done\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R4.2.1",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
