{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d1c0af67-5044-4cce-b10a-352777677a8d",
   "metadata": {},
   "source": [
    "# LDSC-SEG on TWAS Meta-Analysis Results for OUD\n",
    "\n",
    "**Author**: Jesse Marks<br>\n",
    "**NIH Project**: [Harnessing Knowledge of Gene Function in Brain Tissue for Discovering Biology Underlying Heroin Addiction](https://reporter.nih.gov/search/RC99reuHhEW0n_3WuFPU6g/project-details/10116351) <br>\n",
    "**Charge Code**:0218755.001.002.001 ~0215889.001.001~<br>\n",
    "**GitHub Issue**:  [Opioid Use Disorder TWAS Meta-analysis (Uniform Processing) #183](https://github.com/RTIInternational/bioinformatics/issues/183)<br>\n",
    "\n",
    "**Overview**:<br>\n",
    "Differential gene expression (DGE) analysis commonly uses RNA-seq (RNA sequencing) data as a primary source of information. RNA-seq is a powerful and widely used technique for measuring gene expression levels and identifying differentially expressed genes between different biological conditions. It provides a comprehensive view of gene expression and allows researchers to compare expression patterns across different conditions, treatments, or stages of development. It has become a valuable tool for studying gene regulation, identifying disease-associated genes, and understanding biological processes at the transcriptomic level.\n",
    "\n",
    "We received nucleus accumbens (NAc) RNA-seq data from the New York Genome Center (NYGC). The GitHub issue  [NYGC NAc RNA-seq data processing and QC #189](https://github.com/RTIInternational/bioinformatics/issues/189) describes the process of retrieving, processing, and QC'ing these nucleus accumbens (NAc) data.\n",
    "\n",
    "\n",
    "**Description**:<br>\n",
    "This notebook details the process used to test for heritability enrichment in the significantly differentially expressed genes from our DEG meta-analysis for 47 phenotypes.\n",
    "\n",
    "\n",
    "We performed Stratified LD Score Regression (S-LDSC) analyses using the LD score regression approach described in [Finucane et al. 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5896795/) on specifically expressed genes (LDSC-SEG).\n",
    "LDSC-SEG applies stratified LD score regression to test whether disease heritability is enriched in the regions surrounding genes with the highest specific expression in a given tissue.\n",
    "This approach helps to interpret GWAS signal by leveraging gene expression data.\n",
    "\n",
    "\n",
    "\n",
    "The meta-analysis results are published in [THIS](https://github.com/RTIInternational/bioinformatics/issues/189#issuecomment-1545939252) GitHub comment.\n",
    "The file name is `hic_NAc_Seney_published_meta_analysis_results_20230427.txt`.\n",
    "These results are from the meta-analysis with Seney et al. published summary stats (Batch 1, Batch 2, Seney et al.):\n",
    "96 significant DEGs (FDR < 0.05)\n",
    "\n",
    "\n",
    "We will filter the results to obtain two sets of significantly expressed genes:\n",
    "\n",
    "- List of genes with a Benjamini-Hochberg FDR <0.05\n",
    "- List of genes with a Benjamini-Hochberg FDR <0.10\n",
    "\n",
    "\n",
    "Overall, this notebook will address the biological question of whether significantly differentially expressed genes (and their proximal regions) from the opioid use disorder DEG meta-analysis results showed enrichment for genetic signal associated with opioid overdose deaths or any related phenotypes from the following EUR-specific studies:\n",
    "\n",
    "___\n",
    "\n",
    "<details>\n",
    "    <summary>phenotype list</summary>\n",
    "    \n",
    "`sleep studies`:\n",
    "* Insomnia (Jansen et al., 2019 Nat Genet [30804565](https://pubmed.ncbi.nlm.nih.gov/30804565/))\n",
    "* Insomnia (Lane et al., 2019 Nat Genet [30804566](https://pubmed.ncbi.nlm.nih.gov/30804566/))\n",
    "* LongSleepDur (Dashti et al., 2019 Nat Commun [30846698](https://pubmed.ncbi.nlm.nih.gov/30846698/))\n",
    "* ShortSleepDur (Dashti et al., 2019 Nat Commun [30846698](https://pubmed.ncbi.nlm.nih.gov/30846698/))\n",
    "* sleepDuration (Dashti et al., 2019 Nat Commun [30846698](https://pubmed.ncbi.nlm.nih.gov/30846698/))\n",
    "* Sleepdur (Jansen et al., 2019 Nat Genet [30804565](https://pubmed.ncbi.nlm.nih.gov/30804565/))\n",
    "    \n",
    "`41 other studies:`\n",
    "* Age of Initiation  (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Alcohol Dependence (Walters et al., 2018 Nat Neurosci [30482948](https://pubmed.ncbi.nlm.nih.gov/30482948))\n",
    "* Alcohol Drinks per Week (DPW) (Liu et al., 2019 Nat Genet [30643251]())\n",
    "* Alzheimer's Disease (Lambert et al., 2013 Nat Genet [24162737](https://pubmed.ncbi.nlm.nih.gov/24162737))\n",
    "* Amyotrophic Lateral Sclerosis (Rheenen et al., 2016 Nat Genet [27455348](https://pubmed.ncbi.nlm.nih.gov/27455348))\n",
    "* Anorexia Nervosa (Watson et al., 2019 Nat Genet [31308545](https://pubmed.ncbi.nlm.nih.gov/31308545))\n",
    "* Attention Deficit Hyperactivity Disorder (Demontis et al., 2019 Nat Genet [30478444]())\n",
    "* Autism Spectrum Disorders (Grove et al., 2019 Nat Genet [30804558](https://pubmed.ncbi.nlm.nih.gov/30804558))\n",
    "* Bipolar Disorder (Stahl et al., 2019 Nat Genet [31043756](https://pubmed.ncbi.nlm.nih.gov/31043756))\n",
    "* Cannabis Use Disorder (CUD) (Demontis et al., 2019 Nat Neurosci [31209380](https://pubmed.ncbi.nlm.nih.gov/31209380))\n",
    "* Childhood IQ (Benyamin et al., 2014 Mol Psychiatry [23358156](https://pubmed.ncbi.nlm.nih.gov/23358156))\n",
    "* Cigarettes Per Day (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* College Completion (Rietveld et al., 2013 Science [23722424](https://pubmed.ncbi.nlm.nih.gov/23722424))\n",
    "* Cotinine Levels (Ware et al., 2016 Sci Rep [26833182](https://pubmed.ncbi.nlm.nih.gov/26833182/))\n",
    "* Fagerstrom Test for Nicotine Dependence (FTND) (Quach et al., 2020 Nat Commun [33144568](https://pubmed.ncbi.nlm.nih.gov/33144568/))\n",
    "* Heaviness of Smoking Index (HSI) (Quach et al., 2020 Nat Commun [33144568](https://pubmed.ncbi.nlm.nih.gov/33144568/))\n",
    "* Intelligence (Sniekers et al., 2017 Nat Genet [28530673](https://pubmed.ncbi.nlm.nih.gov/28530673))\n",
    "* Lifetime Cannabis Use (Ever vs. Never) (Pasman et al., 2018 Nat Neurosci [30150663](https://pubmed.ncbi.nlm.nih.gov/30150663))\n",
    "* Major Depressive Disorder (Howard et al., 2018 Nat Commun [29662059](https://pubmed.ncbi.nlm.nih.gov/29662059))\n",
    "* Mean Accumbens Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Caudate Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Hippocampus Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Pallidum Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Putamen Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Thalamus Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry [21173776](https://pubmed.ncbi.nlm.nih.gov/21173776))\n",
    "* Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry [21173776](https://pubmed.ncbi.nlm.nih.gov/21173776))\n",
    "* Neuroticism (Okbay et al., 2016 Nat Genet [27089181]())\n",
    "* Opioid Addiction: GENOA GWAS meta-analysis\n",
    "* Opioid Addiction: gSEM OA GWAS meta-analysis (i.e., GENOA, MVP-SAGE-YP, PGC-SUD, and Partners Health)\n",
    "* Parkinson's Disease (Sanchez et al., 2009 Nat Genet [19915575](https://pubmed.ncbi.nlm.nih.gov/19915575))\n",
    "* Post-traumatic Stress Disorder (Nievergelt et al., 2019 Nat Commun [31594949](https://pubmed.ncbi.nlm.nih.gov/31594949))\n",
    "* Psychiatric Genetics Consortium Cross-disorder GWAS (Schizophrenia, Bipolar Disorder, MDD, ASD and ADHD) (Cross-Disorder Group of the Psychiatric Genomics Consortium, 2013 Lancet [23453885](https://pubmed.ncbi.nlm.nih.gov/23453885))\n",
    "* Schizophrenia (Ripke et al., 2014 Nature [25056061](https://pubmed.ncbi.nlm.nih.gov/25056061))\n",
    "* Smoking Cessation (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Smoking Initiation (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Subjective Well Being (Okbay et al., 2016 Nat Genet [27089181](https://pubmed.ncbi.nlm.nih.gov/27089181))\n",
    "* Total Intracranial Volume (ICV) (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Years of Schooling (Okbay et al., 2016 Nature [27225129](https://pubmed.ncbi.nlm.nih.gov/27225129))\n",
    "</details><br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fa2f49b-d950-4d75-9c25-881b6bc8aa5f",
   "metadata": {},
   "source": [
    "Download file from GitHub https://github.com/RTIInternational/bioinformatics/files/11465503/hic_NAc_Seney_published_meta_analysis_results_20230427.txt\n",
    "\n",
    "There was an issue with the file missing a column header. Caryn instructed Jesse to add in the header manually.\n",
    "\n",
    "```css\n",
    "\"It looks like I have the Ensembl id’s as the row names so it didn’t have a column name but you could add Gene before SYMBOL to fix that.\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2aac5335-378c-40fb-8261-650693577788",
   "metadata": {},
   "outputs": [],
   "source": [
    "sed -i.bak '1s/^/\"Gene\"\\t/' hic_NAc_Seney_published_meta_analysis_results_20230427.txt\n",
    "sed -i.bak 's/\"//g' hic_NAc_Seney_published_meta_analysis_results_20230427.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a0975e98-a331-4db3-87ad-dad91bbf733c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tSYMBOL\tpvalue\toverall.eff.direction\tp.adjust.bonferroni\tp.adjust.fdr\n",
      "ENSG00000000003\tTSPAN6\t0.435375129889047\t+\t1\t0.801291554092373\n",
      "ENSG00000000419\tDPM1\t0.48850757113392\t+\t1\t0.845234634361694\n",
      "ENSG00000000457\tSCYL3\t0.681123215633089\t+\t1\t0.979037097735917\n",
      "ENSG00000000460\tC1orf112\t0.0516570130885042\t+\t1\t0.353211250563386\n",
      "ENSG00000000938\tFGR\t0.742674561619552\t-\t1\t1\n",
      "ENSG00000000971\tCFH\t0.139491690296594\t-\t1\t0.500835281920593\n",
      "ENSG00000001036\tFUCA2\t0.619340681287323\t+\t1\t0.941892016492497\n",
      "ENSG00000001084\tGCLC\t0.711857457254194\t-\t1\t0.996942214950086\n",
      "ENSG00000001167\tNFYA\t0.929884555311393\t+\t1\t1\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head ~/Downloads/hic_NAc_Seney_published_meta_analysis_results_20230427.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35f39e5b-c638-4a37-aba6-ec7ac0d3a9c4",
   "metadata": {},
   "source": [
    "## Create FDR filtered files\n",
    "Only need the \"Gene\" column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dfb81a3f-32f9-4109-9f43-0329ff7c45fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   13419 /Users/jmarks/Downloads/hic_NAc_Seney_published_meta_analysis_results_20230427.txt\n",
      "     240 /Users/jmarks/Downloads/hic_NAc_Seney_published_meta_analysis_results_20230427_fdr0.10_genes.txt\n",
      "      96 /Users/jmarks/Downloads/hic_NAc_Seney_published_meta_analysis_results_20230427_fdr0.05_genes.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "meta=~/Downloads/hic_NAc_Seney_published_meta_analysis_results_20230427\n",
    "\n",
    "\n",
    "# create FDR filtered file: 0.05\n",
    "# use last column \"p.adjust.fdr\"\n",
    "> ${meta}_fdr0.05_genes.txt\n",
    "tail -n +2 ${meta}.txt | \\\n",
    "    awk '$NF < 0.05 {print $1}' >> ${meta}_fdr0.05_genes.txt\n",
    "\n",
    "\n",
    "# create FDR filtered file: 0.10\n",
    "# use last column \"p.adjust.fdr\"\n",
    "> ${meta}_fdr0.10_genes.txt\n",
    "tail -n +2 ${meta}.txt | \\\n",
    "    awk '$NF < 0.10 {print $1}' >> ${meta}_fdr0.10_genes.txt\n",
    "\n",
    "\n",
    "wc -l ${meta}.txt\n",
    "wc -l ${meta}_fdr0.10_genes.txt\n",
    "wc -l ${meta}_fdr0.05_genes.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33b8310-ac82-4b20-b3f2-ed5ea6b336e3",
   "metadata": {},
   "source": [
    "## Create Gene Coordinate File\n",
    "One of the requirements of running [partioned heritability analysis with LDSC software](https://github.com/bulik/ldsc/wiki/LD-Score-Estimation-Tutorial#partitioned-ld-scores) is to provide a gene coordinate file.\n",
    "With columns:\n",
    "```css\n",
    "GENE, CHR, START, END\n",
    "```\n",
    "where START and END are base pair coordinates of TSS and TES (Transcription Start Site and Transcription End Site, respectively).\n",
    "This file can contain more genes than are in the gene set. \n",
    "It's used to define the genomic regions or intervals corresponding to individual genes.\n",
    "It allows researchers to assess the contribution of specific gene regions or functional elements to the overall heritability of a trait.\n",
    "\n",
    "By associating genetic variants with these defined gene regions, partitioned heritability analysis can estimate the proportion of trait heritability explained by each gene or functional element. This analysis helps identify which specific genes or genomic regions are significantly associated with the trait of interest, providing insights into the genetic architecture of the trait and potential biological mechanisms involved.\n",
    "\n",
    "In summary, the gene coordinate file is a crucial input in partitioned heritability analysis as it defines the boundaries of genes and enables the investigation of the genetic contribution of different genomic regions to complex traits or diseases."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e67c8a81-412e-452e-996b-4643635e2e89",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "### Use GENCODE v40 IDS annotation file\n",
    "We will download the GENCODE v40 annotation file to capture the TSS and TES for our gene coordinate file.\n",
    "\n",
    "Note that our meta-analysis results are in in gencode v40 hg38.\n",
    "But the LD-scores for LDSC are in build hg19. \n",
    "Thus, we will have to use the [gencode GRCh37_mapping](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/GRCh37_mapping/) to get the TSS and TES in GRCh37 coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f0d69cd-2805-48a0-b118-7dac9b096862",
   "metadata": {},
   "outputs": [],
   "source": [
    "# download gencode v40 annotation file\n",
    "wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/GRCh37_mapping/gencode.v40lift37.annotation.gtf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1a853620-6a67-4816-a628-f0a206e18d95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#description: evidence-based annotation of the human genome, version 30 (Ensembl 96), mapped to GRCh37 with gencode-backmap\n",
      "#provider: GENCODE\n",
      "#contact: gencode-help@ebi.ac.uk\n",
      "#format: gtf\n",
      "#date: 2019-04-02\n",
      "chr1\tHAVANA\tgene\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; level 2; havana_gene \"OTTHUMG00000000961.2_2\"; remap_status \"full_contig\"; remap_num_mappings 1; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_num_mappings 1; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 1; exon_id \"ENSE00002234944.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:11869-12227\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t12613\t12721\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 2; exon_id \"ENSE00003582793.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:12613-12721\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t13221\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 3; exon_id \"ENSE00002312635.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:13221-14409\"; remap_status \"full_contig\";\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "gunzip --to-stdout ~/Downloads/gencode.v40lift37.annotation.gtf.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2accd5e-b7d4-42a6-89dc-1c5c977009d4",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "\n",
    "get the start and end positions for each gene\n",
    "https://www.gencodegenes.org/pages/data_format.html\n",
    "\n",
    "| column-number |              content             |                                      values/format                                     |\n",
    "|:-------------:|:--------------------------------:|:--------------------------------------------------------------------------------------:|\n",
    "| 1             | chromosome name                  | chr{1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,M} or GRC accession a |\n",
    "| 2             | annotation source                | {ENSEMBL,HAVANA}                                                                       |\n",
    "| 3             | feature type                     | {gene,transcript,exon,CDS,UTR,start_codon,stop_codon,Selenocysteine}                   |\n",
    "| 4             | genomic start location           | integer-value (1-based)                                                                |\n",
    "| 5             | genomic end location             | integer-value                                                                          |\n",
    "| 6             | score(not used)                  | .                                                                                      |\n",
    "| 7             | genomic strand                   | {+,-}                                                                                  |\n",
    "| 8             | genomic phase (for CDS features) | {0,1,2,.}                                                                              |\n",
    "\n",
    "as well as gene_id in column 9."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caca306d-7e5b-44c4-9632-f6b9edacc614",
   "metadata": {},
   "source": [
    "We want \n",
    "\n",
    "```css\n",
    "GENE CHR START END\n",
    "\n",
    "```\n",
    "for our coordinate file, so columns: 9, 1, 4, 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d96fcc9-fa7d-4118-acaa-5a5eeeca8e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "annfile = \"/Users/jmarks/Downloads/gencode.v40lift37.annotation.gtf.gz\"\n",
    "outfile = \"/Users/jmarks/Downloads/gencode.v40lift37.coordinate_file.txt\"\n",
    "\n",
    "with gzip.open(annfile, 'rt') as annfile, open(outfile, 'w') as outfile:\n",
    "    for _ in range(5):\n",
    "        next(annfile)\n",
    "    line = annfile.readline()\n",
    "    outfile.write(\"GENE\\tCHR\\tSTART\\tEND\\n\")\n",
    "\n",
    "    while line:\n",
    "        sl = line.split(\"\\t\")\n",
    "        chrom = sl[0]\n",
    "        #if chrom[:3] == \"chr\": # some instances of GL000204.1 genes\n",
    "        if sl[2] == \"gene\" and chrom not in (\"chrX\", \"chrY\", \"chrM\"): # LDSC only contains LDscores for autosomes\n",
    "            #chrom = chrom.split(\"chr\")[1] # remove \"chr\" prefix\n",
    "            start = sl[3]\n",
    "            end = sl[4]\n",
    "            gencode = sl[8].split(\";\")[0] # remove all additional info after ;\n",
    "            gencode = gencode.split(\" \")[1].strip('\"') # remove \"gene_id\" portion, and double quotes\n",
    "            gencode = gencode.split(\".\")[0] # remove suffix <ENSG...>\n",
    "            outline = f\"{gencode}\\t{chrom}\\t{start}\\t{end}\\n\" # gene, chr, start-, and end-genomic position\n",
    "            outfile.write(outline)\n",
    "        line = annfile.readline()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf71d249-b553-41c1-9d81-6b394b1d37c6",
   "metadata": {},
   "source": [
    "Note, another option could have been to use Bryan Quach's \n",
    "GENCODE data munging R functions. See https://github.com/bryancquach/omixjutsu\n",
    "`subset_gencode_gtf`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53879ab8-34a3-4288-b227-b5f27e286ef5",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "## munge sumstats\n",
    "Summary statisics should be in [Summary Statistics File Format](https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format).\n",
    "\n",
    "For our analysis,some of the sumstats were already munged, so we will just use them when we can.\n",
    "Others we will have to download an munge. \n",
    "\n",
    "Note that we did this for another analysis, so nothing to do here.\n",
    "See the following notebook for notes on processing: https://github.com/RTIInternational/jaamarks_notebooks/blob/master/heroin/develop/oud_twas_stratified_ldsc/20230426_oud_stratified_ldsc_on_twas_meta.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45adb2cf-21c1-4d0d-bc58-f3fd963485fb",
   "metadata": {},
   "source": [
    "# Stratified LDSC\n",
    "https://github.com/bulik/ldsc/wiki/LD-Score-Estimation-Tutorial#partitioned-ld-scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "331b378b-d14f-43db-9acd-fa556f38b421",
   "metadata": {},
   "source": [
    "## Download munged sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c566687f-1805-471c-a235-9b053cf1da06",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir sumstats && cd sumstats\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/AgeOfInitiation.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/alcohol_dependence_walters2018_nat_neurosci/munged/pgc_alcdep.eur_discovery.aug2018_release.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/DrinksPerWeek.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/alzheimers_disease_lambert2013_nat_genet/munged/alzheimers_disease_lambert2013_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet/munged/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/anorexia_watson2019_nat_genet/munged/anorexia_watson2019_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/adhd_demontis2018_nat_genet/munged/daner_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.meta.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/autism_spectrum_disorder_grove2019_nat_genet/munged/iPSYCH-PGC_ASD_Nov2017.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/bipolar_disorder_stahl2019_nat_genet/munged/daner_PGC_BIP32b_mds7a_0416a.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Accumbens Volume (Hibar et al., 2015 Nature 25607358)\n",
    "#10\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Caudate Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Hippocampus Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Pallidum Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Putamen Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz . # Mean Thalamus Volume (Hibar et al., 2015 Nature 25607358)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/brain_volume_hibar2015_nature/munged/ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/cannabis_use_disorder_demontis2019_nat_neurosci/munged/CUD_GWAS_iPSYCH_June2019.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/childhood_intelligence_benyamin2014_mol_psych/munged/CHIC_Summary_Benyamin2014.sumstats.gz . # Childhood IQ s3://rti-shared/gwas_publicly_available_sumstats/childhood_intelligence_benyamin2014_mol_psych/raw/CHIC_Summary_Benyamin2014.txt.gz\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/CigarettesPerDay.txt.munged.merged.txt.gz .\n",
    "#20\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/educational_attainment_rietveld2013_science/munged/SSGAC_College_Rietveld2013_publicrelease.sumstats.gz . # College completion\n",
    "aws s3 cp s3://rti-shared/ldsc/data/cotinine_levels_ware2016_sci_rep/munged/cotinine_ware2016_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/depressive_symptoms_okbay2016/munged/DS_Full.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/nicotine_dependence_quach2020_nat_commun/munged/ftnd_wave3_eur_quach2020_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/ukb_hsi/munged/ukb_gwa_003_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/intelligence_sniekers2017_nat_genet/munged/intelligence_sniekers2017_nat_genet_sumstats_formatted.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/lifetime_cannabis_use_pasman2018_nat_neurosci/munged/cannabis_icc_ukb_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/major_depressive_disorder_howard2018_nat_commun/munged/pgc_ukb_depression_gwas_workflow_ready.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.sumstats.gz . # Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry 21173776)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/personality_demoor_2012_mol_psych/munged/GPC-1.NEO-OPENNESS.full.with_header.hg19.sumstats.gz .# Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry 21173776)\n",
    "#30\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/neuroticism_okbay2016_nat_genet/munged/neuroticism_okbay2016_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/opioid_addiction_gaddis_mathur2022_sci_rep/munged/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted.sumstats.gz .\n",
    "aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/GenomicSEM/results/29/gSEM/final/munged/genomicSEM_GWAS.oaALL.MVP1_MVP2_YP_SAGE.PGC.Song.table.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/parkinsons_disease_sanchez2009_nat_genet/munged/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/ptsd_nievergelt2019_nat_commun/munged/pts_eur_freeze2_overall.results.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/cross_disorder_gwas_pgc2013_lancet/munged/pgc.cross.full.2013-03.hg19.sumstats.gz . # pgc needs a liftover from hg18 (s3://rti-shared/gwas_publicly_available_sumstats/cross_disorder_gwas_pgc2013_lancet/raw/)\n",
    "aws s3 cp s3://rti-shared/ldsc/data/schizophrenia_ripke2014_nature/munged/daner_natgen_pgc_eur.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/SmokingCessation.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/gscan_liu2019/munged/SmokingInitiation.txt.munged.merged.txt.gz .\n",
    "aws s3 cp s3://rti-shared/ldsc/data/subjective_wellbeing_okbay2016_nat_genet/munged/SWB_Full.sumstats.gz . # Subjective Well Being (Okbay et al., 2016 Nat Genet 27089181)\n",
    "#40\n",
    "\n",
    "aws s3 cp s3://rti-shared/ldsc/data/years_schooling_okbay2022_nat_genet/munged/EA4_additive_excl_23andMe.sumstats.gz . # Years of Education (Okbay et al., 2022 Nature Genetics  35361970)\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/insomnia_jansen2019_nat_genet/raw/Insomnia-Jansen_2019.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/insomnia_lane2019_nat_genet/raw/Insomnia-Lane_2019.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/long_sleep_duration_dashti2019_nat_commun/raw/LongSleepDur-Dashti_2019.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/short_sleep_duration_dashti2019_nat_commun/raw/ShortSleepDur-Dashti_2019.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/sleep_duration_dashti2019_nat_commun/raw/SleepDuration-Dashti_2019.sumstats.gz .\n",
    "aws s3 cp s3://rti-shared/gwas_publicly_available_sumstats/sleep_duration_jansen2019_nat_genet/raw/Sleepdur-Jansen_2019.sumstats.gz .\n",
    "#47"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8b46295-195e-4967-9237-fedfb1b72b2f",
   "metadata": {},
   "source": [
    "## Download LD-scores\n",
    "And other necessary files for sLDSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0645bcac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make dir for gene coordinate BED file and geneset, logs, and 1000g files\n",
    "# make dir for logs\n",
    "mkdir {gene_files,logs,1000g} \n",
    "\n",
    "cd 1000g\n",
    "# download files needed for partitioned heritability analysis\n",
    "aws s3 cp s3://rti-shared/ldsc/ld_score_reference/1000G/LDSCORE-1000G_Phase3_baseline_ldscores.tar .\n",
    "aws s3 cp s3://rti-shared/ldsc/ld_score_reference/1000G/LDSCORE-1000G_Phase3_frq.tar .\n",
    "aws s3 cp s3://rti-shared/ldsc/ld_score_reference/1000G/LDSCORE-1000G_Phase3_plinkfiles.tar .\n",
    "aws s3 cp s3://rti-shared/ldsc/ld_score_reference/1000G/LDSCORE-weights_hm3_no_hla.tar .\n",
    "#wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_phase3_baseline_ldscores.tgz\n",
    "#wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_plinkfiles.tgz\n",
    "#wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_frq.tgz\n",
    "#wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/weights_hm3_no_hla.tgz\n",
    "\n",
    "\n",
    "# extract files\n",
    "tar -xvf LDSCORE-1000G_Phase3_baseline_ldscores.tar\n",
    "tar -xvf LDSCORE-1000G_Phase3_plinkfiles.tar\n",
    "tar -xvf LDSCORE-1000G_Phase3_frq.tar\n",
    "tar -xvf LDSCORE-weights_hm3_no_hla.tar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7a10975-729f-4f66-9a65-d38f15cd73a0",
   "metadata": {},
   "source": [
    "## Start sLDSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1439d23-2971-46e2-8d74-f08187465689",
   "metadata": {},
   "outputs": [],
   "source": [
    "# interactive session\n",
    "docker run -it -v $PWD:/data/ \\\n",
    "    rtibiocloud/ldsc:v1.0.1_0bb574e bash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b0b7036",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "date=20230718\n",
    "window=100000\n",
    "\n",
    "for fdr in {\"0.05\",\"0.10\"}; do # loop through each BED file\n",
    "    coord_file=/data/gene_files/gencode.v40lift37.coordinate_file.txt \n",
    "    geneset_file=/data/gene_files/hic_NAc_Seney_published_meta_analysis_results_20230427_fdr${fdr}_genes.txt\n",
    "\n",
    "    # store processing files for each meta in separate dir\n",
    "    mkdir -p /data/{annotations_ldscores,results}/fdr$fdr/\n",
    "\n",
    "    for j in {1..22}; do # loop through each chromosome\n",
    "        # create annotation files\n",
    "        python /opt/ldsc/make_annot.py \\\n",
    "            --gene-set-file $geneset_file \\\n",
    "            --gene-coord-file $coord_file \\\n",
    "            --windowsize $window \\\n",
    "            --bimfile /data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim \\\n",
    "            --annot-file /data/annotations_ldscores/fdr$fdr/meta_fdr${fdr}genes_window${window}_chr$j.annot.gz | \\\n",
    "        tee -a /data/logs/sldsc_${date}.log\n",
    "\n",
    "        # compute LD scores\n",
    "        python /opt/ldsc/ldsc.py \\\n",
    "            --l2 \\\n",
    "            --thin-annot \\\n",
    "            --ld-wind-cm 1 \\\n",
    "            --print-snps /data/1000g/1000G_EUR_Phase3_baseline/print_snps.txt \\\n",
    "            --bfile /data/1000g/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j \\\n",
    "            --annot /data/annotations_ldscores/fdr$fdr/meta_fdr${fdr}genes_window${window}_chr$j.annot.gz \\\n",
    "            --out /data/annotations_ldscores/fdr$fdr/meta_fdr${fdr}genes_window${window}_chr$j | \\\n",
    "        tee -a /data/logs/sldsc_${date}.log\n",
    "    done # end chr loop\n",
    "\n",
    "    for trait in {\"insomnia_jansen\",\"insomnia_lane\",\"long_sleep_duration_dashti\",\"short_sleep_duration_dashti\",\"sleep_duration_dashti\",\"sleep_duration_jansen\",\"adhd\",\"age_of_initiation\",\"alcohol_dependence\",\"alzheimers_disease\",\"amyotrophic_lateral_sclerosis\",\"anorexia\",\"autism\",\"bipolar\",\"brain_volume_mean_accumbens\",\"brain_volume_mean_amygdala\",\"brain_volume_mean_caudate\",\"brain_volume_mean_hippocampus\",\"brain_volume_mean_pallidum\",\"brain_volume_mean_putamen\",\"brain_volume_mean_thalamus\",\"brain_volume_total_intracranial\",\"cannabis_use_disorder\",\"childhood_intelligence\",\"cigarettes_per_day\",\"college_completion\",\"cotinine_levels\",\"cross_disorder\",\"depressive_symptoms\",\"drinks_per_week\",\"ftnd\",\"heaviness_smoking_index\",\"intelligence\",\"lifetime_cannabis_use\",\"major_depressive_disorder\",\"neo_conscientiousness\",\"neo_openness\",\"neuroticism\",\"opioid_addiction_144\",\"opioid_addiction_gsem\",\"parkinsons\",\"ptsd\",\"schizophrenia\",\"smoking_cessation\",\"smoking_initiation\",\"subjective_wellbeing\",\"years_of_education\"}; do # loop through all traits\n",
    "        case $trait in  # use sumstats files that corresponds to the trait name for the h2 estimate\n",
    "            \"insomnia_jansen\") stats=/data/sumstats/Insomnia-Jansen_2019.sumstats.gz ;;\n",
    "            \"insomnia_lane\") stats=/data/sumstats/Insomnia-Lane_2019.sumstats.gz ;;\n",
    "            \"long_sleep_duration_dashti\") stats=/data/sumstats/LongSleepDur-Dashti_2019.sumstats.gz ;;\n",
    "            \"short_sleep_duration_dashti\") stats=/data/sumstats/ShortSleepDur-Dashti_2019.sumstats.gz ;;\n",
    "            \"sleep_duration_dashti\") stats=/data/sumstats/SleepDuration-Dashti_2019.sumstats.gz ;;\n",
    "            \"sleep_duration_jansen\") stats=/data/sumstats/Sleepdur-Jansen_2019.sumstats.gz ;;\n",
    "            \"adhd\") stats=/data/sumstats/daner_meta_filtered_NA_iPSYCH23_PGC11_sigPCs_woSEX_2ell6sd_EUR_Neff_70.meta.munged.merged.txt.gz ;;\n",
    "            \"age_of_initiation\") stats=/data/sumstats/AgeOfInitiation.txt.munged.merged.txt.gz ;;\n",
    "            \"alcohol_dependence\") stats=/data/sumstats/pgc_alcdep.eur_discovery.aug2018_release.txt.munged.merged.txt.gz ;;\n",
    "            \"alzheimers_disease\") stats=/data/sumstats/alzheimers_disease_lambert2013_nat_genet.sumstats.gz ;;\n",
    "            \"amyotrophic_lateral_sclerosis\") stats=/data/sumstats/amyotrophic_lateral_sclerosis_rheenen2016_nat_genet.sumstats.gz ;;\n",
    "            \"anorexia\") stats=/data/sumstats/anorexia_watson2019_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"autism\") stats=/data/sumstats/iPSYCH-PGC_ASD_Nov2017.munged.merged.txt.gz ;;\n",
    "            \"bipolar\") stats=/data/sumstats/daner_PGC_BIP32b_mds7a_0416a.munged.merged.txt.gz ;;\n",
    "            \"brain_volume_mean_accumbens\") stats=/data/sumstats/ENIGMA2_MeanAccumbens_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_amygdala\") stats=/data/sumstats/ENIGMA2_MeanAmygdala_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_caudate\") stats=/data/sumstats/ENIGMA2_MeanCaudate_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_hippocampus\") stats=/data/sumstats/ENIGMA2_MeanHippocampus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_pallidum\") stats=/data/sumstats/ENIGMA2_MeanPallidum_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_putamen\") stats=/data/sumstats/ENIGMA2_MeanPutamen_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_mean_thalamus\") stats=/data/sumstats/ENIGMA2_MeanThalamus_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"brain_volume_total_intracranial\") stats=/data/sumstats/ENIGMA2_ICV_Combined_GenomeControlled_Jan23.tbl.sumstats.gz ;;\n",
    "            \"cannabis_use_disorder\") stats=/data/sumstats/CUD_GWAS_iPSYCH_June2019.munged.merged.txt.gz ;;\n",
    "            \"childhood_intelligence\") stats=/data/sumstats/CHIC_Summary_Benyamin2014.sumstats.gz ;;\n",
    "            \"cigarettes_per_day\") stats=/data/sumstats/CigarettesPerDay.txt.munged.merged.txt.gz ;;\n",
    "            \"college_completion\") stats=/data/sumstats/SSGAC_College_Rietveld2013_publicrelease.sumstats.gz ;;\n",
    "            \"cotinine_levels\") stats=/data/sumstats/cotinine_ware2016_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"cross_disorder\") stats=/data/sumstats/pgc.cross.full.2013-03.hg19.sumstats.gz ;;\n",
    "            \"depressive_symptoms\") stats=/data/sumstats/DS_Full.txt.munged.merged.txt.gz ;;\n",
    "            \"drinks_per_week\") stats=/data/sumstats/DrinksPerWeek.txt.munged.merged.txt.gz ;;\n",
    "            \"ftnd\") stats=/data/sumstats/ftnd_wave3_eur_quach2020_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"heaviness_smoking_index\") stats=/data/sumstats/ukb_gwa_003_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"intelligence\") stats=/data/sumstats/intelligence_sniekers2017_nat_genet_sumstats_formatted.sumstats.gz ;;\n",
    "            \"lifetime_cannabis_use\") stats=/data/sumstats/cannabis_icc_ukb_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"major_depressive_disorder\") stats=/data/sumstats/pgc_ukb_depression_gwas_workflow_ready.txt.munged.merged.txt.gz ;;\n",
    "            \"neo_conscientiousness\") stats=/data/sumstats/GPC-1.NEO-CONSCIENTIOUSNESS.full.with_header.h19.sumstats.gz ;;\n",
    "            \"neo_openness\") stats=/data/sumstats/GPC-1.NEO-OPENNESS.full.with_header.hg19.sumstats.gz ;;\n",
    "            \"neuroticism\") stats=/data/sumstats/neuroticism_okbay2016_nat_genet.sumstats.gz ;;\n",
    "            \"opioid_addiction_144\") stats=/data/sumstats/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chrall.maf_gt_0.01.rsq_gt_0.8.sumstats_formatted.sumstats.gz ;;\n",
    "            \"opioid_addiction_gsem\") stats=/data/sumstats/genomicSEM_GWAS.oaALL.MVP1_MVP2_YP_SAGE.PGC.Song.table.sumstats.gz ;;\n",
    "            \"parkinsons\") stats=/data/sumstats/parkinsons_disease_sanchez2009_nat_genet.sumstats.gz ;;\n",
    "            \"ptsd\") stats=/data/sumstats/pts_eur_freeze2_overall.results.munged.merged.txt.gz ;;\n",
    "            \"schizophrenia\") stats=/data/sumstats/daner_natgen_pgc_eur.munged.merged.txt.gz ;;\n",
    "            \"smoking_cessation\") stats=/data/sumstats/SmokingCessation.txt.munged.merged.txt.gz ;;\n",
    "            \"smoking_initiation\") stats=/data/sumstats/SmokingInitiation.txt.munged.merged.txt.gz ;;\n",
    "            \"subjective_wellbeing\") stats=/data/sumstats/SWB_Full.sumstats.gz ;;\n",
    "            \"years_of_education\") stats=/data/sumstats/EA4_additive_excl_23andMe.sumstats.gz ;;\n",
    "        esac\n",
    "\n",
    "        # computed partitioned heritability estimate\n",
    "        python /opt/ldsc/ldsc.py \\\n",
    "            --h2 $stats \\\n",
    "            --overlap-annot \\\n",
    "            --print-coefficients \\\n",
    "            --w-ld-chr \"/data/1000g/weights_hm3_no_hla/weights.\" \\\n",
    "            --frqfile-chr \"/data/1000g/1000G_Phase3_frq/1000G.EUR.QC.\" \\\n",
    "            --ref-ld-chr \"/data/annotations_ldscores/fdr$fdr/meta_fdr${fdr}genes_window${window}_chr,/data/1000g/1000G_EUR_Phase3_baseline/baseline.\" \\\n",
    "            --out \"/data/results/fdr$fdr/${trait}_with_meta_analysis_deg_genes_fdr${fdr}_window${window}\" | \\\n",
    "          tee -a /data/logs/sldsc_${date}.log\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0080b525-44cb-4437-bb16-5416b35a900c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd results/\n",
    "\n",
    "for fdr in {\"0.05\",\"0.10\"}; do\n",
    "    outfile=${date}_all_phenotypes_with_meta_analysis_deg_fdr${fdr}_window${window}_final_results.tsv\n",
    "    touch $outfile\n",
    "    head -1 fdr${fdr}/smoking_initiation_with_meta_analysis_deg_genes_fdr${fdr}_window${window}.results > $outfile\n",
    "        \n",
    "    for file in fdr${fdr}/*_fdr${fdr}_window${window}.results; do\n",
    "        trait=$(echo $file | sed \"s/_with_oa_twas_meta_analysis_deg_genes_fdr.*//\") # remove suffix\n",
    "        trait=$(echo $trait | sed \"s/fdr$fdr\\///\") # remove directory prefix\n",
    "        #echo $trait\n",
    "        awk -v trait=$trait \\\n",
    "        '$1 = trait {print $0}' OFS=\"\\t\" <(tail -n +2 $file | head -1) >> $outfile\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b93d33db-099a-4a38-8c7d-29293d522142",
   "metadata": {},
   "source": [
    "## fdr0.05_window100000_final_results\n",
    "\n",
    "| Category                                                                                  | Prop._SNPs       | Prop._h2         | Prop._h2_std_error | Enrichment     | Enrichment_std_error | Enrichment_p          | Coefficient        | Coefficient_std_error | Coefficient_z-score |\n",
    "|-------------------------------------------------------------------------------------------|------------------|------------------|--------------------|----------------|----------------------|-----------------------|--------------------|-----------------------|---------------------|\n",
    "| adhd_with_meta_analysis_deg_genes_fdr0.05_window100000.results                            | 0.00870166355234 | 0.0255842592481  | 0.00657035062018   | 2.94015725778  | 0.755068336147       | 0.010888320497495454  | 1.58162412662e-07  | 6.24872734529e-08     | 2.53111400006       |\n",
    "| age_of_initiation_with_meta_analysis_deg_genes_fdr0.05_window100000.results               | 0.00870166355234 | 0.00852239296764 | 0.00573337652427   | 0.979398125011 | 0.658882808993       | 0.97504144727272768   | -1.69118995272e-09 | 4.42773376733e-09     | -0.381953848535     |\n",
    "| alcohol_dependence_with_meta_analysis_deg_genes_fdr0.05_window100000.results              | 0.00870166355234 | -0.0113700059354 | 0.0129930101884    | -1.3066473861  | 1.49316393471        | 0.10549005764853481   | -3.40616927178e-08 | 2.32045860688e-08     | -1.46788624528      |\n",
    "| alzheimers_disease_with_meta_analysis_deg_genes_fdr0.05_window100000.results              | 0.00870166355234 | 0.0186995406635  | 0.0152086748395    | 2.14896158085  | 1.74778934488        | 0.51104650620035352   | -1.11801656058e-09 | 1.99593257386e-08     | -0.0560147459502    |\n",
    "| amyotrophic_lateral_sclerosis_with_meta_analysis_deg_genes_fdr0.05_window100000.results   | 0.00870166355234 | 0.0299201123831  | 0.023644626653     | 3.43843590402  | 2.71725360453        | 0.35574426658909342   | 8.74784811041e-09  | 2.11194911505e-08     | 0.414207333315      |\n",
    "| anorexia_with_meta_analysis_deg_genes_fdr0.05_window100000.results                        | 0.00870166355234 | 0.0101780761051  | 0.00450954485894   | 1.16967014915  | 0.518239395469       | 0.74379771697718977   | 4.67766272693e-09  | 1.93243217994e-08     | 0.242060900014      |\n",
    "| autism_with_meta_analysis_deg_genes_fdr0.05_window100000.results                          | 0.00870166355234 | 0.0125350176942  | 0.00686898242557   | 1.44053118336  | 0.789387268797       | 0.57691947688725809   | 1.16727349486e-08  | 2.54015554614e-08     | 0.459528353149      |\n",
    "| bipolar_with_meta_analysis_deg_genes_fdr0.05_window100000.results                         | 0.00870166355234 | 0.0159776877082  | 0.00478182037281   | 1.83616473012  | 0.549529448484       | 0.12769024041357641   | 6.17345393874e-08  | 6.22356140949e-08     | 0.991948746472      |\n",
    "| brain_volume_mean_accumbens_with_meta_analysis_deg_genes_fdr0.05_window100000.results     | 0.00870166355234 | 0.0267181245146  | 0.0316329927592    | 3.07046168286  | 3.63528106654        | 0.55843363391535727   | 3.36531395958e-08  | 6.5659671409e-08      | 0.512538958444      |\n",
    "| brain_volume_mean_amygdala_with_meta_analysis_deg_genes_fdr0.05_window100000.results      | 0.00870166355234 | -12.4289582327   | 480.050450446      | -1428.34277124 | 55167.6639253        | 0.39104708595636217   | 7.02958403975e-08  | 7.40856474903e-08     | 0.948845596668      |\n",
    "| brain_volume_mean_caudate_with_meta_analysis_deg_genes_fdr0.05_window100000.results       | 0.00870166355234 | 0.030681178022   | 0.0128432647291    | 3.52589798921  | 1.47595510351        | 0.067885192929666535  | 5.25309930191e-08  | 4.75170454492e-08     | 1.10551892532       |\n",
    "| brain_volume_mean_hippocampus_with_meta_analysis_deg_genes_fdr0.05_window100000.results   | 0.00870166355234 | 0.0172497389492  | 0.0244419908054    | 1.98234956402  | 2.80888713501        | 0.72405545896062196   | 2.01356344781e-08  | 6.82737039435e-08     | 0.294925180781      |\n",
    "| brain_volume_mean_pallidum_with_meta_analysis_deg_genes_fdr0.05_window100000.results      | 0.00870166355234 | 0.0618915110152  | 0.0286407132045    | 7.1126067611   | 3.29140664107        | 0.04513575183616262   | 1.37412145539e-07  | 7.34355250045e-08     | 1.87119443254       |\n",
    "| brain_volume_mean_putamen_with_meta_analysis_deg_genes_fdr0.05_window100000.results       | 0.00870166355234 | 0.0331245894108  | 0.0112278867955    | 3.80669618074  | 1.29031497575        | 0.022836830909067775  | 1.37986525472e-07  | 7.07127128666e-08     | 1.95136800553       |\n",
    "| brain_volume_mean_thalamus_with_meta_analysis_deg_genes_fdr0.05_window100000.results      | 0.00870166355234 | 0.0353614401703  | 0.0209399508546    | 4.06375631023  | 2.4064307622         | 0.20150793472635628   | 9.01729858465e-08  | 6.84154155757e-08     | 1.31802145887       |\n",
    "| brain_volume_total_intracranial_with_meta_analysis_deg_genes_fdr0.05_window100000.results | 0.00870166355234 | 0.0224669869814  | 0.0197274175583    | 2.58191860053  | 2.2670857635         | 0.46032037386542457   | 1.04647501501e-08  | 6.69703352281e-08     | 0.156259485852      |\n",
    "| cannabis_use_disorder_with_meta_analysis_deg_genes_fdr0.05_window100000.results           | 0.00870166355234 | 0.0358067828576  | 0.0334324804442    | 4.11493533876  | 3.84207918901        | 0.39372587821911198   | 6.39060267738e-08  | 8.77059722888e-08     | 0.728639397136      |\n",
    "| childhood_intelligence_with_meta_analysis_deg_genes_fdr0.05_window100000.results          | 0.00870166355234 | 0.0269144089793  | 0.0173519437005    | 3.09301880238  | 1.99409499071        | 0.29002641992119332   | 5.26512368144e-08  | 5.47292073895e-08     | 0.962031780208      |\n",
    "| cigarettes_per_day_with_meta_analysis_deg_genes_fdr0.05_window100000.results              | 0.00870166355234 | 0.00882226075015 | 0.00267171493721   | 1.01385909684  | 0.307034961894       | 0.96391921548052362   | -4.13177312802e-09 | 3.59131423511e-09     | -1.15049056071      |\n",
    "| college_completion_with_meta_analysis_deg_genes_fdr0.05_window100000.results              | 0.00870166355234 | 0.0157432250945  | 0.00879440272481   | 1.80922015656  | 1.01065763712        | 0.42424939702119346   | 9.12932957909e-09  | 1.51344302921e-08     | 0.603215938946      |\n",
    "| cotinine_levels_with_meta_analysis_deg_genes_fdr0.05_window100000.results                 | 0.00870166355234 | -0.0267032307046 | 0.356074748577     | -3.06875007796 | 40.9203075291        | 0.69927949961895353   | -8.64478505763e-08 | 1.61641777396e-07     | -0.534811309113     |\n",
    "| cross_disorder_with_meta_analysis_deg_genes_fdr0.05_window100000.results                  | 0.00870166355234 | 0.0197167738156  | 0.00482340161274   | 2.26586257869  | 0.554307987631       | 0.02140110266598633   | 2.50744761433e-08  | 1.41426584899e-08     | 1.77296766101       |\n",
    "| depressive_symptoms_with_meta_analysis_deg_genes_fdr0.05_window100000.results             | 0.00870166355234 | 0.00385336846433 | 0.00522366848795   | 0.442831240389 | 0.600306878855       | 0.35780241802341906   | -3.52406640977e-09 | 4.23176860374e-09     | -0.832764439591     |\n",
    "| drinks_per_week_with_meta_analysis_deg_genes_fdr0.05_window100000.results                 | 0.00870166355234 | 0.0173733080875  | 0.00545523036202   | 1.9965501979   | 0.62691809781        | 0.11298475028019651   | 5.21972816647e-09  | 4.58279546495e-09     | 1.13898344502       |\n",
    "| ftnd_with_meta_analysis_deg_genes_fdr0.05_window100000.results                            | 0.00870166355234 | 0.00244859955802 | 0.0100871388038    | 0.281394418814 | 1.15921958406        | 0.54022940840461409   | -1.30495111341e-08 | 1.45968630697e-08     | -0.893994214498     |\n",
    "| heaviness_smoking_index_with_meta_analysis_deg_genes_fdr0.05_window100000.results         | 0.00870166355234 | 0.0313835268977  | 0.0223231049784    | 3.60661231142  | 2.56538360097        | 0.2605824506068018    | 2.29322410408e-08  | 2.72719250825e-08     | 0.840873571317      |\n",
    "| insomnia_jansen_with_meta_analysis_deg_genes_fdr0.05_window100000.results                 | 0.00870166355234 | 0.0126445868593  | 0.00559884651045   | 1.45312293255  | 0.643422545215       | 0.48102074330057976   | 1.50538939024e-09  | 4.45788985373e-09     | 0.337691024146      |\n",
    "| insomnia_lane_with_meta_analysis_deg_genes_fdr0.05_window100000.results                   | 0.00870166355234 | 0.0113274708255  | 0.0036983994818    | 1.30175922769  | 0.425022118995       | 0.47872315194756998   | 9.94958164423e-10  | 3.88225611801e-09     | 0.256283494488      |\n",
    "| intelligence_with_meta_analysis_deg_genes_fdr0.05_window100000.results                    | 0.00870166355234 | 0.0112892274711  | 0.00702356691955   | 1.2973642802   | 0.807152204553       | 0.71302031779648145   | 3.78349530597e-09  | 2.31080260849e-08     | 0.163730787393      |\n",
    "| lifetime_cannabis_use_with_meta_analysis_deg_genes_fdr0.05_window100000.results           | 0.00870166355234 | 0.00410613525034 | 0.00420221000728   | 0.471879339582 | 0.482920303917       | 0.2716715467211599    | -6.4355988417e-09  | 4.91153543718e-09     | -1.31030284196      |\n",
    "| long_sleep_duration_dashti_with_meta_analysis_deg_genes_fdr0.05_window100000.results      | 0.00870166355234 | 0.01609664238    | 0.00455575901374   | 1.84983506696  | 0.523550351761       | 0.096333946836897319  | 1.89714493871e-09  | 1.76466390009e-09     | 1.07507437457       |\n",
    "| major_depressive_disorder_with_meta_analysis_deg_genes_fdr0.05_window100000.results       | 0.00870166355234 | 0.0122596562669  | 0.00350094434651   | 1.40888649546  | 0.402330465371       | 0.31006190385531446   | 4.03657690862e-09  | 4.01296434185e-09     | 1.00588407092       |\n",
    "| neo_conscientiousness_with_meta_analysis_deg_genes_fdr0.05_window100000.results           | 0.00870166355234 | 0.0705666311747  | 0.0378682327354    | 8.10955637968  | 4.35183830168        | 0.014310877788372605  | 1.03267772153e-07  | 4.13380737103e-08     | 2.49812734082       |\n",
    "| neo_openness_with_meta_analysis_deg_genes_fdr0.05_window100000.results                    | 0.00870166355234 | 0.00241864634677 | 0.0223771975123    | 0.27795217917  | 2.57159994497        | 0.77992005390694774   | -8.42759725349e-09 | 4.2579406784e-08      | -0.197926600909     |\n",
    "| neuroticism_with_meta_analysis_deg_genes_fdr0.05_window100000.results                     | 0.00870166355234 | 0.00465984994444 | 0.00286150349493   | 0.535512539231 | 0.328845568174       | 0.14653116793940507   | -6.95472797336e-09 | 4.30821564615e-09     | -1.61429430293      |\n",
    "| opioid_addiction_144_with_meta_analysis_deg_genes_fdr0.05_window100000.results            | 0.00870166355234 | 0.0105155912136  | 0.011137231661     | 1.20845757255  | 1.27989683742        | 0.87124902622824241   | 5.37327683463e-11  | 2.60328988903e-09     | 0.020640332286      |\n",
    "| opioid_addiction_gsem_with_meta_analysis_deg_genes_fdr0.05_window100000.results           | 0.00870166355234 | 0.0193503254118  | 0.0075047450931    | 2.22375012495  | 0.862449467043       | 0.14913595563902779   | 1.28774046578e-08  | 1.12164744292e-08     | 1.14807952705       |\n",
    "| parkinsons_with_meta_analysis_deg_genes_fdr0.05_window100000.results                      | 0.00870166355234 | 0.00709567742566 | 0.0216633006012    | 0.815439183897 | 2.48955851613        | 0.9407982041464189    | -8.18387504095e-08 | 1.91987792714e-07     | -0.426270593836     |\n",
    "| ptsd_with_meta_analysis_deg_genes_fdr0.05_window100000.results                            | 0.00870166355234 | 0.00318593490283 | 0.0135908916632    | 0.366129405448 | 1.56187280529        | 0.67962597620851795   | -4.21241955683e-09 | 1.00373220291e-08     | -0.419675641032     |\n",
    "| schizophrenia_with_meta_analysis_deg_genes_fdr0.05_window100000.results                   | 0.00870166355234 | 0.00927002423178 | 0.00215060532504   | 1.0653163244   | 0.247148756339       | 0.79184147178278197   | -2.62662874262e-08 | 3.67305877862e-08     | -0.715106645694     |\n",
    "| short_sleep_duration_dashti_with_meta_analysis_deg_genes_fdr0.05_window100000.results     | 0.00870166355234 | 0.0147399460566  | 0.00485126584814   | 1.69392277326  | 0.557510161012       | 0.21500012220386022   | 3.19179037301e-09  | 3.88403412989e-09     | 0.821771968594      |\n",
    "| sleep_duration_dashti_with_meta_analysis_deg_genes_fdr0.05_window100000.results           | 0.00870166355234 | 0.0175284667724  | 0.00452470849018   | 2.01438112     | 0.519982008379       | 0.051951234785041059  | 7.77581504945e-09  | 5.42096313407e-09     | 1.43439733072       |\n",
    "| sleep_duration_jansen_with_meta_analysis_deg_genes_fdr0.05_window100000.results           | 0.00870166355234 | 0.0161025017011  | 0.0042959261384    | 1.85050842338  | 0.493690213665       | 0.088935493980230262  | 6.29221011282e-09  | 5.18271670187e-09     | 1.2140756431        |\n",
    "| smoking_cessation_with_meta_analysis_deg_genes_fdr0.05_window100000.results               | 0.00870166355234 | 0.0250847443565  | 0.00549737442477   | 2.88275272948  | 0.631761316868       | 0.0026814229978564964 | 8.22986600201e-09  | 3.30570220558e-09     | 2.4895969117        |\n",
    "| smoking_initiation_with_meta_analysis_deg_genes_fdr0.05_window100000.results              | 0.00870166355234 | 0.0123411035138  | 0.00385444756888   | 1.41824645823  | 0.442955251683       | 0.34520790564350401   | 3.30344574513e-09  | 4.57539921232e-09     | 0.722001642225      |\n",
    "| subjective_wellbeing_with_meta_analysis_deg_genes_fdr0.05_window100000.results            | 0.00870166355234 | 0.00552113099052 | 0.00413708640266   | 0.634491434576 | 0.475436263166       | 0.4458039789443945    | -2.03451727232e-09 | 1.91619673433e-09     | -1.06174759401      |\n",
    "| years_of_education_with_meta_analysis_deg_genes_fdr0.05_window100000.results              | 0.00870166355234 | 0.0134958598266  | 0.00189801692418   | 1.55095169394  | 0.218121157266       | 0.011974623946558864  | 8.43490564125e-09  | 4.59925395533e-09     | 1.83397257972       |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb11f2b6-15e2-49a7-8a88-4d1b0049f3e1",
   "metadata": {},
   "source": [
    "## fdr0.10_window100000_final_results\n",
    "\n",
    "| Category                                                                                  | Prop._SNPs      | Prop._h2           | Prop._h2_std_error | Enrichment       | Enrichment_std_error | Enrichment_p          | Coefficient        | Coefficient_std_error | Coefficient_z-score |\n",
    "|-------------------------------------------------------------------------------------------|-----------------|--------------------|--------------------|------------------|----------------------|-----------------------|--------------------|-----------------------|---------------------|\n",
    "| adhd_with_meta_analysis_deg_genes_fdr0.10_window100000.results                            | 0.0235019733579 | 0.0318551174922    | 0.00766241339458   | 1.35542309606    | 0.326032766607       | 0.27762915094615515   | 2.91765508929e-08  | 2.74711937659e-08     | 1.06207801312       |\n",
    "| age_of_initiation_with_meta_analysis_deg_genes_fdr0.10_window100000.results               | 0.0235019733579 | 0.025052529551     | 0.00928418473288   | 1.06597557446    | 0.39503851832        | 0.86729497427276281   | -6.10123021093e-10 | 2.63015730117e-09     | -0.231972065253     |\n",
    "| alcohol_dependence_with_meta_analysis_deg_genes_fdr0.10_window100000.results              | 0.0235019733579 | -0.000497586197079 | 0.0209129735571    | -0.0211721028486 | 0.889839046221       | 0.24270659086880342   | -1.4499484047e-08  | 1.47077641792e-08     | -0.985838763141     |\n",
    "| alzheimers_disease_with_meta_analysis_deg_genes_fdr0.10_window100000.results              | 0.0235019733579 | 0.0404742623042    | 0.0205822188447    | 1.72216441947    | 0.875765559399       | 0.40951306648821562   | -5.5368507959e-09  | 9.77709076447e-09     | -0.566308621786     |\n",
    "| amyotrophic_lateral_sclerosis_with_meta_analysis_deg_genes_fdr0.10_window100000.results   | 0.0235019733579 | 0.0295629598465    | 0.0356802333351    | 1.25789266273    | 1.51818031583        | 0.86481047528682342   | -6.6861336738e-09  | 1.25922442845e-08     | -0.530972360663     |\n",
    "| anorexia_with_meta_analysis_deg_genes_fdr0.10_window100000.results                        | 0.0235019733579 | 0.0303146775314    | 0.00754723399557   | 1.28987796343    | 0.321131927122       | 0.36952132526941572   | 1.05993303891e-08  | 1.23437842885e-08     | 0.858677545014      |\n",
    "| autism_with_meta_analysis_deg_genes_fdr0.10_window100000.results                          | 0.0235019733579 | 0.0396037489417    | 0.00938373852413   | 1.68512440801    | 0.399274494156       | 0.088413684373080184  | 2.12268975135e-08  | 1.30171681637e-08     | 1.63068474238       |\n",
    "| bipolar_with_meta_analysis_deg_genes_fdr0.10_window100000.results                         | 0.0235019733579 | 0.03155003084      | 0.00619308507269   | 1.3424417754     | 0.263513407083       | 0.19444137571484557   | 9.00450641274e-09  | 3.01355287189e-08     | 0.298800346154      |\n",
    "| brain_volume_mean_accumbens_with_meta_analysis_deg_genes_fdr0.10_window100000.results     | 0.0235019733579 | 0.037305353313     | 0.0490361527884    | 1.58732855088    | 2.08646959307        | 0.77440176167903862   | 5.64124701666e-09  | 3.8995301822e-08      | 0.144664786605      |\n",
    "| brain_volume_mean_amygdala_with_meta_analysis_deg_genes_fdr0.10_window100000.results      | 0.0235019733579 | 0.107754834966     | 52.9731284671      | 4.58492711762    | 2253.98640618        | 0.99487935825594043   | 5.42886640797e-09  | 3.62610810718e-08     | 0.149716066028      |\n",
    "| brain_volume_mean_caudate_with_meta_analysis_deg_genes_fdr0.10_window100000.results       | 0.0235019733579 | 0.0969001339914    | 0.0287363470951    | 4.12306373239    | 1.22272060552        | 0.0071337912452482952 | 7.91187696614e-08  | 3.84124737807e-08     | 2.05971555264       |\n",
    "| brain_volume_mean_hippocampus_with_meta_analysis_deg_genes_fdr0.10_window100000.results   | 0.0235019733579 | 0.0103406959784    | 0.0366169092383    | 0.439992668739   | 1.55803551816        | 0.71578586749287587   | -1.9337864561e-08  | 3.80180144465e-08     | -0.50865003979      |\n",
    "| brain_volume_mean_pallidum_with_meta_analysis_deg_genes_fdr0.10_window100000.results      | 0.0235019733579 | 0.104600197932     | 0.0399634255256    | 4.45069851535    | 1.70042850943        | 0.029592036154114224  | 7.39479770128e-08  | 3.85590468295e-08     | 1.9177853991        |\n",
    "| brain_volume_mean_putamen_with_meta_analysis_deg_genes_fdr0.10_window100000.results       | 0.0235019733579 | 0.0488804678261    | 0.0158251651787    | 2.07984525732    | 0.673354740803       | 0.086752669029321189  | 4.1457730765e-08   | 3.6843406482e-08      | 1.12524152144       |\n",
    "| brain_volume_mean_thalamus_with_meta_analysis_deg_genes_fdr0.10_window100000.results      | 0.0235019733579 | 0.0589704924027    | 0.0292140673601    | 2.50917195355    | 1.24304742054        | 0.22348482441572115   | 3.89606318979e-08  | 3.60782163249e-08     | 1.07989351655       |\n",
    "| brain_volume_total_intracranial_with_meta_analysis_deg_genes_fdr0.10_window100000.results | 0.0235019733579 | 0.0302331307405    | 0.0288105513058    | 1.28640817859    | 1.22587796638        | 0.80735661098468159   | -2.54991866379e-08 | 3.80128976819e-08     | -0.670803548082     |\n",
    "| cannabis_use_disorder_with_meta_analysis_deg_genes_fdr0.10_window100000.results           | 0.0235019733579 | 0.096072976466     | 0.0577597829014    | 4.08786849526    | 2.45765672618        | 0.14075985139800018   | 7.04545220168e-08  | 5.1493505088e-08      | 1.36822152418       |\n",
    "| childhood_intelligence_with_meta_analysis_deg_genes_fdr0.10_window100000.results          | 0.0235019733579 | 0.0599643377625    | 0.0235432483297    | 2.55145969445    | 1.00175623431        | 0.12126885055377402   | 4.14792607518e-08  | 2.85241119766e-08     | 1.45418236984       |\n",
    "| cigarettes_per_day_with_meta_analysis_deg_genes_fdr0.10_window100000.results              | 0.0235019733579 | 0.0289905494269    | 0.00582235917887   | 1.23353681776    | 0.247739161738       | 0.3351657751266427    | -5.52960715934e-10 | 2.90607815312e-09     | -0.190277303912     |\n",
    "| college_completion_with_meta_analysis_deg_genes_fdr0.10_window100000.results              | 0.0235019733579 | 0.0298925716507    | 0.0110595938822    | 1.27191751924    | 0.470581500277       | 0.56495301696106903   | 2.54080724985e-09  | 7.25335751201e-09     | 0.350293949477      |\n",
    "| cotinine_levels_with_meta_analysis_deg_genes_fdr0.10_window100000.results                 | 0.0235019733579 | -0.0593686374002   | 0.833660063733     | -2.5261128713    | 35.4719176572        | 0.4780019487839352    | -6.39592471541e-08 | 7.75374647499e-08     | -0.824881847252     |\n",
    "| cross_disorder_with_meta_analysis_deg_genes_fdr0.10_window100000.results                  | 0.0235019733579 | 0.0354514890529    | 0.00834091287368   | 1.50844733389    | 0.354902660584       | 0.1515223867377706    | 7.48572202e-09     | 9.16680031746e-09     | 0.816612314086      |\n",
    "| depressive_symptoms_with_meta_analysis_deg_genes_fdr0.10_window100000.results             | 0.0235019733579 | 0.0027896383329    | 0.00851415203342   | 0.118698046774   | 0.362273920738       | 0.016739901577230848  | -5.92908018752e-09 | 2.63488861253e-09     | -2.25022043031      |\n",
    "| drinks_per_week_with_meta_analysis_deg_genes_fdr0.10_window100000.results                 | 0.0235019733579 | 0.0398396450619    | 0.00690985656583   | 1.69516169793    | 0.294011760656       | 0.019571434476542655  | 3.60646788007e-09  | 2.14914928328e-09     | 1.67809091165       |\n",
    "| ftnd_with_meta_analysis_deg_genes_fdr0.10_window100000.results                            | 0.0235019733579 | 0.0259541107133    | 0.0169714940567    | 1.10433750894    | 0.722130597219       | 0.88410559523537158   | -1.56565426635e-09 | 9.02702823143e-09     | -0.17344071894      |\n",
    "| heaviness_smoking_index_with_meta_analysis_deg_genes_fdr0.10_window100000.results         | 0.0235019733579 | 0.0404016637869    | 0.0315164776434    | 1.71907538025    | 1.34101409969        | 0.56941903102791502   | 3.78576120838e-09  | 1.52804033628e-08     | 0.247752701189      |\n",
    "| insomnia_jansen_with_meta_analysis_deg_genes_fdr0.10_window100000.results                 | 0.0235019733579 | 0.0384662127229    | 0.00822291112551   | 1.63672267589    | 0.349881731219       | 0.070676849027460339  | 3.28417269495e-09  | 2.4745865651e-09      | 1.32716015728       |\n",
    "| insomnia_lane_with_meta_analysis_deg_genes_fdr0.10_window100000.results                   | 0.0235019733579 | 0.0274587273551    | 0.00476314686948   | 1.16835837302    | 0.20267008208        | 0.40643426368333679   | 3.17315691577e-10  | 1.90321654951e-09     | 0.166726004804      |\n",
    "| intelligence_with_meta_analysis_deg_genes_fdr0.10_window100000.results                    | 0.0235019733579 | 0.026747271713     | 0.00915518274402   | 1.13808620688    | 0.38954953291        | 0.72304113814713367   | 6.30839507957e-10  | 1.1244744926e-08      | 0.0561008286187     |\n",
    "| lifetime_cannabis_use_with_meta_analysis_deg_genes_fdr0.10_window100000.results           | 0.0235019733579 | 0.0151580906649    | 0.00801635420692   | 0.64497097474    | 0.341092813138       | 0.29368230946567037   | -4.10428942975e-09 | 3.51610005746e-09     | -1.16728459449      |\n",
    "| long_sleep_duration_dashti_with_meta_analysis_deg_genes_fdr0.10_window100000.results      | 0.0235019733579 | 0.0410998577354    | 0.0104065097907    | 1.74878326639    | 0.442793021345       | 0.094822666089747862  | 1.94263583651e-09  | 1.56359545423e-09     | 1.24241588913       |\n",
    "| major_depressive_disorder_with_meta_analysis_deg_genes_fdr0.10_window100000.results       | 0.0235019733579 | 0.0266474285535    | 0.00472985315447   | 1.13383791853    | 0.201253447209       | 0.5054426014355724    | 1.56185655051e-09  | 2.01921316591e-09     | 0.773497606332      |\n",
    "| neo_conscientiousness_with_meta_analysis_deg_genes_fdr0.10_window100000.results           | 0.0235019733579 | 0.191199463231     | 0.0923204898197    | 8.13546421485    | 3.92820162009        | 0.0065487473360090753 | 1.05378959074e-07  | 3.79625880908e-08     | 2.7758634059        |\n",
    "| neo_openness_with_meta_analysis_deg_genes_fdr0.10_window100000.results                    | 0.0235019733579 | -0.0402690040514   | 0.0339392626408    | -1.71343075912   | 1.444102677          | 0.029851977188901199  | -4.11971310857e-08 | 2.13237235019e-08     | -1.93198580361      |\n",
    "| neuroticism_with_meta_analysis_deg_genes_fdr0.10_window100000.results                     | 0.0235019733579 | 0.0203983314152    | 0.00546109195136   | 0.867941219429   | 0.232367379037       | 0.56485779269466119   | -1.79401922086e-09 | 3.16895958793e-09     | -0.566122467352     |\n",
    "| opioid_addiction_144_with_meta_analysis_deg_genes_fdr0.10_window100000.results            | 0.0235019733579 | 0.0497879452604    | 0.0196320492697    | 2.11845807593    | 0.835336206486       | 0.18506441122330691   | 2.12478348547e-09  | 1.73209007289e-09     | 1.22671650784       |\n",
    "| opioid_addiction_gsem_with_meta_analysis_deg_genes_fdr0.10_window100000.results           | 0.0235019733579 | 0.0287271577149    | 0.00991415419262   | 1.2223296009     | 0.421843478488       | 0.59366869326507166   | 1.02928599286e-09  | 5.68032184259e-09     | 0.181202055338      |\n",
    "| parkinsons_with_meta_analysis_deg_genes_fdr0.10_window100000.results                      | 0.0235019733579 | 0.0484728452976    | 0.0442486414162    | 2.06250107425    | 1.8827628107         | 0.57481340565007755   | 2.07524565868e-08  | 1.49914822022e-07     | 0.13842831754       |\n",
    "| ptsd_with_meta_analysis_deg_genes_fdr0.10_window100000.results                            | 0.0235019733579 | 0.0222675444771    | 0.02247054064      | 0.947475522076   | 0.956112931362       | 0.95606940455335865   | 3.24883974243e-11  | 6.2567410316e-09      | 0.00519254309236    |\n",
    "| schizophrenia_with_meta_analysis_deg_genes_fdr0.10_window100000.results                   | 0.0235019733579 | 0.0320347438846    | 0.00407749818819   | 1.36306613052    | 0.173495992277       | 0.038411519832770552  | 2.76211947757e-08  | 2.63627832305e-08     | 1.04773439641       |\n",
    "| short_sleep_duration_dashti_with_meta_analysis_deg_genes_fdr0.10_window100000.results     | 0.0235019733579 | 0.0388831967692    | 0.012106646039     | 1.65446518797    | 0.515133170081       | 0.2050333120228881    | 3.46706167464e-09  | 3.636766782e-09       | 0.953336268853      |\n",
    "| sleep_duration_dashti_with_meta_analysis_deg_genes_fdr0.10_window100000.results           | 0.0235019733579 | 0.0452085255736    | 0.0119805059708    | 1.92360551539    | 0.509765958305       | 0.07477291210259239   | 7.87005113402e-09  | 5.49361657973e-09     | 1.43258107292       |\n",
    "| sleep_duration_jansen_with_meta_analysis_deg_genes_fdr0.10_window100000.results           | 0.0235019733579 | 0.040230098511     | 0.0108342441223    | 1.71177534322    | 0.460992953967       | 0.12426233186812584   | 5.81800822085e-09  | 4.89702952422e-09     | 1.18806884706       |\n",
    "| smoking_cessation_with_meta_analysis_deg_genes_fdr0.10_window100000.results               | 0.0235019733579 | 0.0429445834425    | 0.00806419324428   | 1.82727564144    | 0.343128345926       | 0.015429063153835835  | 3.21813129221e-09  | 1.90050106857e-09     | 1.69330675232       |\n",
    "| smoking_initiation_with_meta_analysis_deg_genes_fdr0.10_window100000.results              | 0.0235019733579 | 0.0260994151404    | 0.00481810422193   | 1.11052015688    | 0.20500849646        | 0.58935135712504594   | 5.65573628823e-10  | 2.17780073932e-09     | 0.259699438342      |\n",
    "| subjective_wellbeing_with_meta_analysis_deg_genes_fdr0.10_window100000.results            | 0.0235019733579 | 0.030702222292     | 0.0106354178114    | 1.30636784514    | 0.452532970293       | 0.49620905833319662   | 7.33866582491e-10  | 1.81175741758e-09     | 0.405057860048      |\n",
    "| years_of_education_with_meta_analysis_deg_genes_fdr0.10_window100000.results              | 0.0235019733579 | 0.0294987394144    | 0.00254148101557   | 1.25516010784    | 0.108139047597       | 0.019124108434994555  | 3.19824863693e-09  | 2.23486200656e-09     | 1.43107208747       |"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
