{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coloc review\n",
    "Perform a sanity check on the coloc analysis.\n",
    "|     Dataset1    |     Dataset2               |     CpG           |     Pval vs   Beta/se    |     stdev   methyl    |\n",
    "|-----------------|----------------------------|-------------------|--------------------------|-----------------------|\n",
    "|     GSCAN       |     mQTL   filtered        |     cg12073833    |     P-value              |     Yes               |\n",
    "|     GSCAN       |     mQTL   filtered        |     cg12073833    |     Beta/se              |     Yes               |\n",
    "|     GSCAN       |     mQTL   filtered        |     cg12073833    |     P-value              |     no                |\n",
    "|     GSCAN       |     mQTL   filtered        |     cg12073833    |     Beta/se              |     no                |\n",
    "|     GSCAN       |     mQTL   non-filtered    |     cg12073833    |     P-value              |     Yes               |\n",
    "|     GSCAN       |     mQTL   non-filtered    |     cg12073833    |     Beta/se              |     Yes               |\n",
    "|     GSCAN       |     mQTL   non-filtered    |     cg12073833    |     P-value              |     no                |\n",
    "|     GSCAN       |     mQTL   non-filtered    |     cg12073833    |     Beta/se              |     no                |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ubuntu/rti-nd/coloc_review'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Split up input data\n",
    "Currently the baseline mQTL results are merged with the GSCAN results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Non-filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t1.2353351130762\t0.0405091\t30.4952180873861\t1.83259348812666e-79\trs6944877:1952582:T:G\tT\t0.424946\tG\t0.0311\t0.00711\t0.00306\t262990\t258485\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t1.2353351130762\t0.0405091\t30.4952180873861\t1.83259348812666e-79\trs6944877:1952582:T:G\tT\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t0.424946\tG\t0.0311\t0.00711\t0.00306\t262990\t258485\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# split data up by GSCAN and mQTL results\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "head -2 baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
    "echo \n",
    "\n",
    "cut -f1-10  baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt > coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt\n",
    "\n",
    "head -2 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt\n",
    "echo \n",
    "\n",
    "cut -f1-4,11-17  baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt > coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt\n",
    "\n",
    "head -2 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t1.2353351130762\t0.0405091\t30.4952180873861\t1.83259348812666e-79\trs6944877:1952582:T:G\tT\t0.424946\tG\t0.0311\t0.00711\t0.00306\t262990\t258485\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t1.2353351130762\t0.0405091\t30.4952180873861\t1.83259348812666e-79\trs6944877:1952582:T:G\tT\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t0.424946\tG\t0.0311\t0.00711\t0.00306\t262990\t258485\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# split filterd data up by GSCAN and mQTL results\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "head -2 MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
    "echo \n",
    "\n",
    "cut -f1-10  MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt > coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
    "\n",
    "head -2 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
    "echo \n",
    "\n",
    "cut -f1-4,11-17  MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt > coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n",
    "\n",
    "head -2 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cg12073833 \n",
    "Extract only the results for this CpG. This is one that was way off the diagonal when we plotted the coloc results indicating that using p-value inputs was way different that using beta/se."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n",
      "coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
      "\n",
      "1856 coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t1979425\trs11770899:1979425:C:T\tcg12073833\t0.570940965506915\t0.0881546\t6.47658412528508\t6.43513092936016e-10\trs11770899:1979425:C:T\tC\n",
      "\n",
      "1186 coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t1979425\trs11770899:1979425:C:T\tcg12073833\t0.570940965506915\t0.0881546\t6.47658412528508\t6.43513092936016e-10\trs11770899:1979425:C:T\tC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "ls \n",
    "echo\n",
    "\n",
    "# mQTL non-filtered\n",
    "head -1 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt > coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
    "grep cg12073833 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt >> coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
    "\n",
    "wc -l coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
    "echo\n",
    "\n",
    "# mQTL filtered\n",
    "head -1 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt > coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
    "grep cg12073833 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt >> coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
    "\n",
    "wc -l coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
    "echo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n",
      "\n",
      "1856 coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1979425\trs11770899:1979425:C:T\tcg12073833\t0.383123\tT\t0.000324\t0.0113\t0.00311\t262990\t257466\n",
      "\n",
      "1186 coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1979425\trs11770899:1979425:C:T\tcg12073833\t0.383123\tT\t0.000324\t0.0113\t0.00311\t262990\t257466\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "ls \n",
    "echo\n",
    "\n",
    "# GSCAN non-filtered\n",
    "head -1 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt > coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
    "grep cg12073833 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt >> coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
    "\n",
    "wc -l coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
    "echo\n",
    "\n",
    "# mQTL filtered\n",
    "head -1 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt > coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt\n",
    "grep cg12073833 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt >> coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt\n",
    "\n",
    "wc -l coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0.719655345726588</li><li>0.705241523327758</li><li>0.729659636474802</li><li>0.716479310658444</li><li>0.716685466360408</li><li>0.766853165346677</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.719655345726588\n",
       "\\item 0.705241523327758\n",
       "\\item 0.729659636474802\n",
       "\\item 0.716479310658444\n",
       "\\item 0.716685466360408\n",
       "\\item 0.766853165346677\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.719655345726588\n",
       "2. 0.705241523327758\n",
       "3. 0.729659636474802\n",
       "4. 0.716479310658444\n",
       "5. 0.716685466360408\n",
       "6. 0.766853165346677\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.7196553 0.7052415 0.7296596 0.7164793 0.7166855 0.7668532"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.0269402510301528"
      ],
      "text/latex": [
       "0.0269402510301528"
      ],
      "text/markdown": [
       "0.0269402510301528"
      ],
      "text/plain": [
       "[1] 0.02694025"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(coloc)\n",
    "\n",
    "# read in input data\n",
    "qtls <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg12073833_baseline.txt\", header=TRUE)\n",
    "#head(qtls, 2)\n",
    "\n",
    "gscan <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\", header=TRUE)\n",
    "#head(gscan, 2)\n",
    "\n",
    "meth <- read.table(\"/home/ubuntu/rti-nd/coloc_review/expression_file/mqtl_methylation_data/LIBD_NAc_MEGA_OMIC_full_matrixeQTL_input.txt\", header=TRUE)\n",
    "\n",
    "# flip GSCAN beta if the effect alleles don't match QTL effect alleles\n",
    "qtls$mQTL_Effect_Allele<-as.character(qtls$mQTL_Effect_Allele)\n",
    "gscan$GSCAN_Effect_Allele<-as.character(gscan$GSCAN_Effect_Allele)\n",
    "gscan$GSCAN_Beta<-as.numeric(gscan$GSCAN_Beta)\n",
    "for (i in 1:length(gscan$GSCAN_Beta)) {\n",
    "    if (qtls$mQTL_Effect_Allele[i] != gscan$GSCAN_Effect_Allele[i]) {\n",
    "        gscan$GSCAN_Beta[i] = -1 * gscan$GSCAN_Beta[i]\n",
    "    }\n",
    "}\n",
    "#head(gscan, 2)\n",
    "\n",
    "\n",
    "meth <- meth[which(meth$cg_id==\"cg12073833\"),]\n",
    "meth <- as.numeric(meth[1,2:ncol(meth)])\n",
    "head(meth)\n",
    "meth_sd <- sd(meth)\n",
    "meth_sd # 0.0269402510301528"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 001: non-filtered, beta/se, no stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n",
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 1.48e-05  1.46e-02  4.71e-05  4.57e-02  9.40e-01 \n",
      "[1] \"PP abf for shared variant: 94%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se\n",
    "my.res1 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, \n",
    "                                   varbeta=(qtls$mQTL_SE)**2, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\"),\n",
    "                     dataset2=list(beta=gscan$GSCAN_Beta, \n",
    "                                   varbeta=(gscan$GSCAN_SE)**2, N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"),\n",
    "                    MAF=gscan$MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 002: non-filtered, p-value, no stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 9.63e-05  3.00e-02  1.69e-04  5.16e-02  9.18e-01 \n",
      "[1] \"PP abf for shared variant: 91.8%\"\n"
     ]
    }
   ],
   "source": [
    "# perform coloc with p-values\n",
    "my.res2 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\"),\n",
    "                     dataset2=list(pvalues=gscan$GSCAN_pvalue,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"), \n",
    "                    MAF=gscan$MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 003: non-filtered, beta/se, WITH stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.1890    0.0351    0.6020    0.1120    0.0621 \n",
      "[1] \"PP abf for shared variant: 6.21%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se with meth_sd\n",
    "my.res3 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, \n",
    "                                   varbeta=(qtls$mQTL_SE)**2, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\", \n",
    "                                   sdY=meth_sd),\n",
    "                     dataset2=list(beta=gscan$GSCAN_Beta, \n",
    "                                   varbeta=(gscan$GSCAN_SE)**2,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\", \n",
    "                                   MAF=gscan$MAF))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 004: non-filtered, p-value, with stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 9.63e-05  3.00e-02  1.69e-04  5.16e-02  9.18e-01 \n",
      "[1] \"PP abf for shared variant: 91.8%\"\n"
     ]
    }
   ],
   "source": [
    "# p-value using standard deviation of methylation data as estimate of sdY for mqtls and using maf for gscan\n",
    "my.res4 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\", \n",
    "                                   sdY=meth_sd),\n",
    "                     dataset2=list(pvalues=gscan$GSCAN_pvalue,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"), \n",
    "                    MAF=gscan$MAF)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(vbeta = (qtls$mQTL_SE)^2, maf = gscan$MAF, n = 220):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    }
   ],
   "source": [
    "sdY.est <- function(vbeta, maf, n) {\n",
    "    warning(\"estimating sdY from maf and varbeta, please directly supply sdY if known\")\n",
    "    oneover <- 1/vbeta\n",
    "    nvx <- 2 * n * maf * (1-maf)\n",
    "    m <- lm(nvx ~ oneover - 1)\n",
    "    cf <- coef(m)[['oneover']]\n",
    "    if(cf < 0)\n",
    "        stop(\"estimated sdY is negative - this can happen with small datasets, or those with errors.  A reasonable estimate of sdY is required to continue.\")\n",
    "    return(sqrt(cf))\n",
    "}\n",
    "\n",
    "meth_sd2 <- sdY.est(vbeta=(qtls$mQTL_SE)**2, maf=gscan$MAF, n=220) # 0.961596439610978"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ubuntu/rti-nd/coloc_review'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t1.2353351130762\t0.0405091\t30.4952180873861\t1.83259348812666e-79\trs6944877:1952582:T:G\tT\t0.424946\tG\t0.0311\t0.00711\t0.00306\t262990\t258485\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t1.2353351130762\t0.0405091\t30.4952180873861\t1.83259348812666e-79\trs6944877:1952582:T:G\tT\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t1952582\trs6944877:1952582:T:G\tcg11050988\t0.424946\tG\t0.0311\t0.00711\t0.00306\t262990\t258485\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# split filterd data up by GSCAN and mQTL results\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "head -2 MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
    "echo \n",
    "\n",
    "cut -f1-10  MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt > coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
    "\n",
    "head -2 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
    "echo \n",
    "\n",
    "cut -f1-4,11-17  MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt > coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n",
    "\n",
    "head -2 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cg05226172 \n",
    "Extract only the results for this CpG. This is one that was way off the diagonal when we plotted the coloc results indicating that using p-value inputs was way different that using beta/se."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
      "coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
      "coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n",
      "coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt\n",
      "coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt\n",
      "\n",
      "1867 coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t2034193\trs3857706:2034193:A:G\tcg05226172\t-0.394324096947506\t0.101795\t-3.87372104447984\t0.000142826728710564\trs3857706:2034193:A:G\tA\n",
      "\n",
      "1080 coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t2034193\trs3857706:2034193:A:G\tcg05226172\t-0.394324096947506\t0.101795\t-3.87372104447984\t0.000142826728710564\trs3857706:2034193:A:G\tA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "ls \n",
    "echo\n",
    "\n",
    "# mQTL non-filtered\n",
    "head -1 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt > coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
    "grep cg05226172 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt >> coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
    "\n",
    "wc -l coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
    "echo\n",
    "\n",
    "# mQTL filtered\n",
    "head -1 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt > coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
    "grep cg05226172 coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt >> coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
    "\n",
    "wc -l coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
    "echo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAF_0.05_GENO_0.1_AA_CAUC_filtered_baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "baseline_coloc_input_dataset_GSCAN_age_of_initiation_locus_chr7_1532865_2532865.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline.txt\n",
      "coloc_input_dataset1_locus_chr7_1532865_2532865_baseline_filtered.txt\n",
      "coloc_input_dataset1_locus_cpg05226172_baseline.txt\n",
      "coloc_input_dataset1_locus_cpg05226172_baseline_filtered.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline.txt\n",
      "coloc_input_dataset1_locus_cpg12073833_baseline_filtered.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt\n",
      "coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt\n",
      "coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\n",
      "coloc_input_dataset2_locus_cpg12073833_gscan_ai_filtered.txt\n",
      "\n",
      "1867 coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t2034193\trs3857706:2034193:A:G\tcg05226172\t0.412827\tG\t3.39e-07\t0.0158\t0.00304\t262990\t259154\n",
      "\n",
      "1080 coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt\n",
      "\n",
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tMAF\tGSCAN_Effect_Allele\tGSCAN_pvalue\tGSCAN_Beta\tGSCAN_SE\tGSCAN_N\tGSCAN_Effective_N\n",
      "7\t2034193\trs3857706:2034193:A:G\tcg05226172\t0.412827\tG\t3.39e-07\t0.0158\t0.00304\t262990\t259154\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "cd mqtls/input_datasets/baseline/age_of_initiation/\n",
    "ls \n",
    "echo\n",
    "\n",
    "# GSCAN non-filtered\n",
    "head -1 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt > coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
    "grep cg05226172 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai.txt >> coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
    "\n",
    "wc -l coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\n",
    "echo\n",
    "\n",
    "# mQTL filtered\n",
    "head -1 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt > coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt\n",
    "grep cg05226172 coloc_input_dataset2_locus_chr7_1532865_2532865_gscan_ai_filtered.txt >> coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt\n",
    "\n",
    "wc -l coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt\n",
    "echo\n",
    "head -2 coloc_input_dataset2_locus_cpg05226172_gscan_ai_filtered.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chromosome</th><th scope=col>GRCh37.p13_position</th><th scope=col>SNP_ID</th><th scope=col>Gene</th><th scope=col>mQTL_Beta</th><th scope=col>mQTL_SE</th><th scope=col>mQTL_test_stat</th><th scope=col>mQTL_pvalue</th><th scope=col>mQTL_original_SNP_ID</th><th scope=col>mQTL_Effect_Allele</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>7</td><td>2034193</td><td>rs3857706:2034193:A:G </td><td>cg05226172</td><td>-0.3943241</td><td>0.101795</td><td>-3.873721</td><td>0.0001428267</td><td>rs3857706:2034193:A:G </td><td>A </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>7</td><td>1837416</td><td>rs5881908:1837416:TA:T</td><td>cg05226172</td><td>-0.4212490</td><td>0.111240</td><td>-3.786841</td><td>0.0001986860</td><td>rs5881908:1837416:TA:T</td><td>TA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & Chromosome & GRCh37.p13\\_position & SNP\\_ID & Gene & mQTL\\_Beta & mQTL\\_SE & mQTL\\_test\\_stat & mQTL\\_pvalue & mQTL\\_original\\_SNP\\_ID & mQTL\\_Effect\\_Allele\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 7 & 2034193 & rs3857706:2034193:A:G  & cg05226172 & -0.3943241 & 0.101795 & -3.873721 & 0.0001428267 & rs3857706:2034193:A:G  & A \\\\\n",
       "\t2 & 7 & 1837416 & rs5881908:1837416:TA:T & cg05226172 & -0.4212490 & 0.111240 & -3.786841 & 0.0001986860 & rs5881908:1837416:TA:T & TA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 10\n",
       "\n",
       "| <!--/--> | Chromosome &lt;int&gt; | GRCh37.p13_position &lt;int&gt; | SNP_ID &lt;chr&gt; | Gene &lt;chr&gt; | mQTL_Beta &lt;dbl&gt; | mQTL_SE &lt;dbl&gt; | mQTL_test_stat &lt;dbl&gt; | mQTL_pvalue &lt;dbl&gt; | mQTL_original_SNP_ID &lt;chr&gt; | mQTL_Effect_Allele &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 7 | 2034193 | rs3857706:2034193:A:G  | cg05226172 | -0.3943241 | 0.101795 | -3.873721 | 0.0001428267 | rs3857706:2034193:A:G  | A  |\n",
       "| 2 | 7 | 1837416 | rs5881908:1837416:TA:T | cg05226172 | -0.4212490 | 0.111240 | -3.786841 | 0.0001986860 | rs5881908:1837416:TA:T | TA |\n",
       "\n"
      ],
      "text/plain": [
       "  Chromosome GRCh37.p13_position SNP_ID                 Gene       mQTL_Beta \n",
       "1 7          2034193             rs3857706:2034193:A:G  cg05226172 -0.3943241\n",
       "2 7          1837416             rs5881908:1837416:TA:T cg05226172 -0.4212490\n",
       "  mQTL_SE  mQTL_test_stat mQTL_pvalue  mQTL_original_SNP_ID  \n",
       "1 0.101795 -3.873721      0.0001428267 rs3857706:2034193:A:G \n",
       "2 0.111240 -3.786841      0.0001986860 rs5881908:1837416:TA:T\n",
       "  mQTL_Effect_Allele\n",
       "1 A                 \n",
       "2 TA                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chromosome</th><th scope=col>GRCh37.p13_position</th><th scope=col>SNP_ID</th><th scope=col>Gene</th><th scope=col>MAF</th><th scope=col>GSCAN_Effect_Allele</th><th scope=col>GSCAN_pvalue</th><th scope=col>GSCAN_Beta</th><th scope=col>GSCAN_SE</th><th scope=col>GSCAN_N</th><th scope=col>GSCAN_Effective_N</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>7</td><td>2034193</td><td>rs3857706:2034193:A:G </td><td>cg05226172</td><td>0.412827</td><td>G</td><td>3.39e-07</td><td>0.0158</td><td>0.00304</td><td>262990</td><td>259154</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>7</td><td>1837416</td><td>rs5881908:1837416:TA:T</td><td>cg05226172</td><td>0.312899</td><td>T</td><td>7.02e-02</td><td>0.0115</td><td>0.00625</td><td>122844</td><td>122315</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Chromosome & GRCh37.p13\\_position & SNP\\_ID & Gene & MAF & GSCAN\\_Effect\\_Allele & GSCAN\\_pvalue & GSCAN\\_Beta & GSCAN\\_SE & GSCAN\\_N & GSCAN\\_Effective\\_N\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 7 & 2034193 & rs3857706:2034193:A:G  & cg05226172 & 0.412827 & G & 3.39e-07 & 0.0158 & 0.00304 & 262990 & 259154\\\\\n",
       "\t2 & 7 & 1837416 & rs5881908:1837416:TA:T & cg05226172 & 0.312899 & T & 7.02e-02 & 0.0115 & 0.00625 & 122844 & 122315\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 11\n",
       "\n",
       "| <!--/--> | Chromosome &lt;int&gt; | GRCh37.p13_position &lt;int&gt; | SNP_ID &lt;chr&gt; | Gene &lt;chr&gt; | MAF &lt;dbl&gt; | GSCAN_Effect_Allele &lt;chr&gt; | GSCAN_pvalue &lt;dbl&gt; | GSCAN_Beta &lt;dbl&gt; | GSCAN_SE &lt;dbl&gt; | GSCAN_N &lt;int&gt; | GSCAN_Effective_N &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 7 | 2034193 | rs3857706:2034193:A:G  | cg05226172 | 0.412827 | G | 3.39e-07 | 0.0158 | 0.00304 | 262990 | 259154 |\n",
       "| 2 | 7 | 1837416 | rs5881908:1837416:TA:T | cg05226172 | 0.312899 | T | 7.02e-02 | 0.0115 | 0.00625 | 122844 | 122315 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chromosome GRCh37.p13_position SNP_ID                 Gene       MAF     \n",
       "1 7          2034193             rs3857706:2034193:A:G  cg05226172 0.412827\n",
       "2 7          1837416             rs5881908:1837416:TA:T cg05226172 0.312899\n",
       "  GSCAN_Effect_Allele GSCAN_pvalue GSCAN_Beta GSCAN_SE GSCAN_N\n",
       "1 G                   3.39e-07     0.0158     0.00304  262990 \n",
       "2 T                   7.02e-02     0.0115     0.00625  122844 \n",
       "  GSCAN_Effective_N\n",
       "1 259154           \n",
       "2 122315           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chromosome</th><th scope=col>GRCh37.p13_position</th><th scope=col>SNP_ID</th><th scope=col>Gene</th><th scope=col>MAF</th><th scope=col>GSCAN_Effect_Allele</th><th scope=col>GSCAN_pvalue</th><th scope=col>GSCAN_Beta</th><th scope=col>GSCAN_SE</th><th scope=col>GSCAN_N</th><th scope=col>GSCAN_Effective_N</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>7</td><td>2034193</td><td>rs3857706:2034193:A:G </td><td>cg05226172</td><td>0.412827</td><td>G</td><td>3.39e-07</td><td>-0.0158</td><td>0.00304</td><td>262990</td><td>259154</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>7</td><td>1837416</td><td>rs5881908:1837416:TA:T</td><td>cg05226172</td><td>0.312899</td><td>T</td><td>7.02e-02</td><td>-0.0115</td><td>0.00625</td><td>122844</td><td>122315</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Chromosome & GRCh37.p13\\_position & SNP\\_ID & Gene & MAF & GSCAN\\_Effect\\_Allele & GSCAN\\_pvalue & GSCAN\\_Beta & GSCAN\\_SE & GSCAN\\_N & GSCAN\\_Effective\\_N\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 7 & 2034193 & rs3857706:2034193:A:G  & cg05226172 & 0.412827 & G & 3.39e-07 & -0.0158 & 0.00304 & 262990 & 259154\\\\\n",
       "\t2 & 7 & 1837416 & rs5881908:1837416:TA:T & cg05226172 & 0.312899 & T & 7.02e-02 & -0.0115 & 0.00625 & 122844 & 122315\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 11\n",
       "\n",
       "| <!--/--> | Chromosome &lt;int&gt; | GRCh37.p13_position &lt;int&gt; | SNP_ID &lt;chr&gt; | Gene &lt;chr&gt; | MAF &lt;dbl&gt; | GSCAN_Effect_Allele &lt;chr&gt; | GSCAN_pvalue &lt;dbl&gt; | GSCAN_Beta &lt;dbl&gt; | GSCAN_SE &lt;dbl&gt; | GSCAN_N &lt;int&gt; | GSCAN_Effective_N &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 7 | 2034193 | rs3857706:2034193:A:G  | cg05226172 | 0.412827 | G | 3.39e-07 | -0.0158 | 0.00304 | 262990 | 259154 |\n",
       "| 2 | 7 | 1837416 | rs5881908:1837416:TA:T | cg05226172 | 0.312899 | T | 7.02e-02 | -0.0115 | 0.00625 | 122844 | 122315 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chromosome GRCh37.p13_position SNP_ID                 Gene       MAF     \n",
       "1 7          2034193             rs3857706:2034193:A:G  cg05226172 0.412827\n",
       "2 7          1837416             rs5881908:1837416:TA:T cg05226172 0.312899\n",
       "  GSCAN_Effect_Allele GSCAN_pvalue GSCAN_Beta GSCAN_SE GSCAN_N\n",
       "1 G                   3.39e-07     -0.0158    0.00304  262990 \n",
       "2 T                   7.02e-02     -0.0115    0.00625  122844 \n",
       "  GSCAN_Effective_N\n",
       "1 259154           \n",
       "2 122315           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.0319845914097218"
      ],
      "text/latex": [
       "0.0319845914097218"
      ],
      "text/markdown": [
       "0.0319845914097218"
      ],
      "text/plain": [
       "[1] 0.03198459"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(coloc)\n",
    "\n",
    "# read in input data\n",
    "qtls <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg05226172_baseline.txt\", header=TRUE)\n",
    "head(qtls, 2)\n",
    "\n",
    "gscan <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset2_locus_cpg05226172_gscan_ai.txt\", header=TRUE)\n",
    "head(gscan, 2)\n",
    "\n",
    "meth <- read.table(\"/home/ubuntu/rti-nd/coloc_review/expression_file/mqtl_methylation_data/LIBD_NAc_MEGA_OMIC_full_matrixeQTL_input.txt\", header=TRUE)\n",
    "\n",
    "# flip GSCAN beta if the effect alleles don't match QTL effect alleles\n",
    "qtls$mQTL_Effect_Allele<-as.character(qtls$mQTL_Effect_Allele)\n",
    "gscan$GSCAN_Effect_Allele<-as.character(gscan$GSCAN_Effect_Allele)\n",
    "gscan$GSCAN_Beta<-as.numeric(gscan$GSCAN_Beta)\n",
    "for (i in 1:length(gscan$GSCAN_Beta)) {\n",
    "    if (qtls$mQTL_Effect_Allele[i] != gscan$GSCAN_Effect_Allele[i]) {\n",
    "        gscan$GSCAN_Beta[i] = -1 * gscan$GSCAN_Beta[i]\n",
    "    }\n",
    "}\n",
    "head(gscan, 2)\n",
    "\n",
    "\n",
    "meth <- meth[which(meth$cg_id==\"cg05226172\"),]\n",
    "meth_sd <- sd(meth[1,2:ncol(meth)])\n",
    "meth_sd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 001: non-filtered, beta/se, no stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in check_dataset(d = dataset1, 1):\n",
      "“minimum p value is: 0.00010719\n",
      "If this is what you expected, this is not a problem.\n",
      "If this is not as small as you expected, please check the 02_data vignette.”\n",
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n",
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0482    0.0195    0.1500    0.0600    0.7220 \n",
      "[1] \"PP abf for shared variant: 72.2%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se\n",
    "my.res1 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, varbeta=(qtls$mQTL_SE)**2, N=220, type=\"quant\"),\n",
    "                    dataset2=list(beta=gscan$GSCAN_Beta, varbeta=(gscan$GSCAN_SE)**2, N=gscan$GSCAN_Effective_N, type=\"quant\"),\n",
    "                    MAF=gscan$MAF)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 002: non-filtered, p-value, no stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in check_dataset(d = dataset1, 1):\n",
      "“minimum p value is: 0.00014283\n",
      "If this is what you expected, this is not a problem.\n",
      "If this is not as small as you expected, please check the 02_data vignette.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.0860    0.0326    0.1480    0.0555    0.6780 \n",
      "[1] \"PP abf for shared variant: 67.8%\"\n"
     ]
    }
   ],
   "source": [
    "# perform coloc with p-values\n",
    "my.res2 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, N=220, type=\"quant\"),\n",
    "                    dataset2=list(pvalues=gscan$GSCAN_pvalue, N=gscan$GSCAN_Effective_N, type=\"quant\"),\n",
    "                    MAF=gscan$MAF)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 003: non-filtered, beta/se, WITH stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in check_dataset(d = dataset1, 1):\n",
      "“minimum p value is: 0.00010719\n",
      "If this is what you expected, this is not a problem.\n",
      "If this is not as small as you expected, please check the 02_data vignette.”\n",
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.1930    0.0360    0.5990    0.1120    0.0605 \n",
      "[1] \"PP abf for shared variant: 6.05%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se with meth_sd\n",
    "my.res3 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, varbeta=(qtls$mQTL_SE)**2, N=220, type=\"quant\", sdY=meth_sd),\n",
    "                    dataset2=list(beta=gscan$GSCAN_Beta, varbeta=(gscan$GSCAN_SE)**2, N=gscan$GSCAN_Effective_N, type=\"quant\", MAF=gscan$MAF))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(vbeta = (qtls$mQTL_SE)^2, maf = gscan$MAF, n = 220):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "1.03252151235383"
      ],
      "text/latex": [
       "1.03252151235383"
      ],
      "text/markdown": [
       "1.03252151235383"
      ],
      "text/plain": [
       "[1] 1.032522"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sdY.est <- function(vbeta, maf, n) {\n",
    "    warning(\"estimating sdY from maf and varbeta, please directly supply sdY if known\")\n",
    "    oneover <- 1/vbeta\n",
    "    nvx <- 2 * n * maf * (1-maf)\n",
    "    m <- lm(nvx ~ oneover - 1)\n",
    "    cf <- coef(m)[['oneover']]\n",
    "    if(cf < 0)\n",
    "        stop(\"estimated sdY is negative - this can happen with small datasets, or those with errors.  A reasonable estimate of sdY is required to continue.\")\n",
    "    return(sqrt(cf))\n",
    "}\n",
    "\n",
    "meth_sd2 <- sdY.est(vbeta=(qtls$mQTL_SE)**2, maf=gscan$MAF, n=220) # 1.03252151235383\n",
    "meth_sd2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MAF\n",
    "\n",
    "Megan:\n",
    "```\n",
    "When we use this: meth_sd2 <- sdY.est(vbeta=(qtls$mQTL_SE)**2, maf=gscan$MAF, n=220) # 0.961596439610978\n",
    "\n",
    "We should see what happens when we replace maf=gscan$MAF with maf=mQTLs$MAF Then re-run beta/se, with new sd-meth\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/rti-nd/coloc_review\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "pwd\n",
    "aws s3 cp s3://rti-nd/scratch/mcarnes/nac_mqtl/mQTLs_MEGA_baseline.afreq ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome\tGRCh37.p13_position\tSNP_ID\tGene\tmQTL_Beta\tmQTL_SE\tmQTL_test_stat\tmQTL_pvalue\tmQTL_original_SNP_ID\tmQTL_Effect_Allele\n",
      "7\t1979425\trs11770899:1979425:C:T\tcg12073833\t0.570940965506915\t0.0881546\t6.47658412528508\t6.43513092936016e-10\trs11770899:1979425:C:T\tC\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "head -2 /home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg12073833_baseline.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add MAF to qtl data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Python3\n",
    "freqf = \"mQTLs_MEGA_baseline.afreq\"\n",
    "qtlf = \"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg12073833_baseline.txt\"\n",
    "outf = \"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg12073833_baseline_with_qtl_maf.txt\"\n",
    "\n",
    "with open(freqf) as frqF, open(qtlf) as qtlF, open(outf, \"w\") as outF:\n",
    "    head = frqF.readline() # #CHROM  ID      REF     ALT     ALT_FREQS       OBS_CT\n",
    "    line = frqF.readline() # 1       rs140435168:256586:T:G  T       G       0.0148515       404\n",
    "\n",
    "    freq_set = dict()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        if sl[0] == \"7\": # chromosome 7\n",
    "            rsid = sl[1]\n",
    "            maf = sl[4]\n",
    "            freq_set[rsid] = maf\n",
    "        if sl[0] == \"8\":\n",
    "            break\n",
    "        line = frqF.readline()\n",
    "\n",
    "    #print(dict(list(freq_set.items())[0:2]))\n",
    "\n",
    "\n",
    "    head = qtlF.readline().split() # Chromosome      GRCh37.p13_position     SNP_ID  Gene    mQTL_Beta       mQTL_SE mQTL_test_stat  mQTL_pvalue     mQTL_original_SNP_ID    mQTL_Effect_Allele\n",
    "    head.append(\"MEGA_MAF\")\n",
    "    head = \"\\t\".join(head) + \"\\n\"\n",
    "    outF.write(head)\n",
    "    line = qtlF.readline() # 7       1979425 rs11770899:1979425:C:T  cg12073833      0.570940965506915       0.0881546       6.47658412528508        6.43513092936016e-10    rs11770899:1979425:C:T C\n",
    "\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        rsid = sl[8]\n",
    "        if rsid in freq_set:\n",
    "            maf = freq_set[rsid]\n",
    "            sl.append(maf)\n",
    "            outline = \"\\t\".join(sl) + \"\\n\"\n",
    "            outF.write(outline)\n",
    "        line = qtlF.readline()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## coloc: cg12073833"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chromosome</th><th scope=col>GRCh37.p13_position</th><th scope=col>SNP_ID</th><th scope=col>Gene</th><th scope=col>mQTL_Beta</th><th scope=col>mQTL_SE</th><th scope=col>mQTL_test_stat</th><th scope=col>mQTL_pvalue</th><th scope=col>mQTL_original_SNP_ID</th><th scope=col>mQTL_Effect_Allele</th><th scope=col>MEGA_MAF</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>7</td><td>1979425</td><td>rs11770899:1979425:C:T</td><td>cg12073833</td><td>0.570941</td><td>0.0881546</td><td>6.476584</td><td>6.435131e-10</td><td>rs11770899:1979425:C:T</td><td>C</td><td>0.495098</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>7</td><td>2025040</td><td>rs4721282:2025040:A:G </td><td>cg12073833</td><td>0.537640</td><td>0.0860770</td><td>6.246037</td><td>2.269349e-09</td><td>rs4721282:2025040:A:G </td><td>A</td><td>0.345794</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Chromosome & GRCh37.p13\\_position & SNP\\_ID & Gene & mQTL\\_Beta & mQTL\\_SE & mQTL\\_test\\_stat & mQTL\\_pvalue & mQTL\\_original\\_SNP\\_ID & mQTL\\_Effect\\_Allele & MEGA\\_MAF\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 7 & 1979425 & rs11770899:1979425:C:T & cg12073833 & 0.570941 & 0.0881546 & 6.476584 & 6.435131e-10 & rs11770899:1979425:C:T & C & 0.495098\\\\\n",
       "\t2 & 7 & 2025040 & rs4721282:2025040:A:G  & cg12073833 & 0.537640 & 0.0860770 & 6.246037 & 2.269349e-09 & rs4721282:2025040:A:G  & A & 0.345794\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 11\n",
       "\n",
       "| <!--/--> | Chromosome &lt;int&gt; | GRCh37.p13_position &lt;int&gt; | SNP_ID &lt;chr&gt; | Gene &lt;chr&gt; | mQTL_Beta &lt;dbl&gt; | mQTL_SE &lt;dbl&gt; | mQTL_test_stat &lt;dbl&gt; | mQTL_pvalue &lt;dbl&gt; | mQTL_original_SNP_ID &lt;chr&gt; | mQTL_Effect_Allele &lt;chr&gt; | MEGA_MAF &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 7 | 1979425 | rs11770899:1979425:C:T | cg12073833 | 0.570941 | 0.0881546 | 6.476584 | 6.435131e-10 | rs11770899:1979425:C:T | C | 0.495098 |\n",
       "| 2 | 7 | 2025040 | rs4721282:2025040:A:G  | cg12073833 | 0.537640 | 0.0860770 | 6.246037 | 2.269349e-09 | rs4721282:2025040:A:G  | A | 0.345794 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chromosome GRCh37.p13_position SNP_ID                 Gene       mQTL_Beta\n",
       "1 7          1979425             rs11770899:1979425:C:T cg12073833 0.570941 \n",
       "2 7          2025040             rs4721282:2025040:A:G  cg12073833 0.537640 \n",
       "  mQTL_SE   mQTL_test_stat mQTL_pvalue  mQTL_original_SNP_ID  \n",
       "1 0.0881546 6.476584       6.435131e-10 rs11770899:1979425:C:T\n",
       "2 0.0860770 6.246037       2.269349e-09 rs4721282:2025040:A:G \n",
       "  mQTL_Effect_Allele MEGA_MAF\n",
       "1 C                  0.495098\n",
       "2 A                  0.345794"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chromosome</th><th scope=col>GRCh37.p13_position</th><th scope=col>SNP_ID</th><th scope=col>Gene</th><th scope=col>MAF</th><th scope=col>GSCAN_Effect_Allele</th><th scope=col>GSCAN_pvalue</th><th scope=col>GSCAN_Beta</th><th scope=col>GSCAN_SE</th><th scope=col>GSCAN_N</th><th scope=col>GSCAN_Effective_N</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>7</td><td>1979425</td><td>rs11770899:1979425:C:T</td><td>cg12073833</td><td>0.383123</td><td>T</td><td>0.000324</td><td>0.0113</td><td>0.00311</td><td>262990</td><td>257466</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>7</td><td>2025040</td><td>rs4721282:2025040:A:G </td><td>cg12073833</td><td>0.371466</td><td>G</td><td>0.000533</td><td>0.0106</td><td>0.00307</td><td>262990</td><td>258825</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Chromosome & GRCh37.p13\\_position & SNP\\_ID & Gene & MAF & GSCAN\\_Effect\\_Allele & GSCAN\\_pvalue & GSCAN\\_Beta & GSCAN\\_SE & GSCAN\\_N & GSCAN\\_Effective\\_N\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 7 & 1979425 & rs11770899:1979425:C:T & cg12073833 & 0.383123 & T & 0.000324 & 0.0113 & 0.00311 & 262990 & 257466\\\\\n",
       "\t2 & 7 & 2025040 & rs4721282:2025040:A:G  & cg12073833 & 0.371466 & G & 0.000533 & 0.0106 & 0.00307 & 262990 & 258825\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 11\n",
       "\n",
       "| <!--/--> | Chromosome &lt;int&gt; | GRCh37.p13_position &lt;int&gt; | SNP_ID &lt;chr&gt; | Gene &lt;chr&gt; | MAF &lt;dbl&gt; | GSCAN_Effect_Allele &lt;chr&gt; | GSCAN_pvalue &lt;dbl&gt; | GSCAN_Beta &lt;dbl&gt; | GSCAN_SE &lt;dbl&gt; | GSCAN_N &lt;int&gt; | GSCAN_Effective_N &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 7 | 1979425 | rs11770899:1979425:C:T | cg12073833 | 0.383123 | T | 0.000324 | 0.0113 | 0.00311 | 262990 | 257466 |\n",
       "| 2 | 7 | 2025040 | rs4721282:2025040:A:G  | cg12073833 | 0.371466 | G | 0.000533 | 0.0106 | 0.00307 | 262990 | 258825 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chromosome GRCh37.p13_position SNP_ID                 Gene       MAF     \n",
       "1 7          1979425             rs11770899:1979425:C:T cg12073833 0.383123\n",
       "2 7          2025040             rs4721282:2025040:A:G  cg12073833 0.371466\n",
       "  GSCAN_Effect_Allele GSCAN_pvalue GSCAN_Beta GSCAN_SE GSCAN_N\n",
       "1 T                   0.000324     0.0113     0.00311  262990 \n",
       "2 G                   0.000533     0.0106     0.00307  262990 \n",
       "  GSCAN_Effective_N\n",
       "1 257466           \n",
       "2 258825           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Chromosome</th><th scope=col>GRCh37.p13_position</th><th scope=col>SNP_ID</th><th scope=col>Gene</th><th scope=col>MAF</th><th scope=col>GSCAN_Effect_Allele</th><th scope=col>GSCAN_pvalue</th><th scope=col>GSCAN_Beta</th><th scope=col>GSCAN_SE</th><th scope=col>GSCAN_N</th><th scope=col>GSCAN_Effective_N</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>7</td><td>1979425</td><td>rs11770899:1979425:C:T</td><td>cg12073833</td><td>0.383123</td><td>T</td><td>0.000324</td><td>-0.0113</td><td>0.00311</td><td>262990</td><td>257466</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>7</td><td>2025040</td><td>rs4721282:2025040:A:G </td><td>cg12073833</td><td>0.371466</td><td>G</td><td>0.000533</td><td>-0.0106</td><td>0.00307</td><td>262990</td><td>258825</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & Chromosome & GRCh37.p13\\_position & SNP\\_ID & Gene & MAF & GSCAN\\_Effect\\_Allele & GSCAN\\_pvalue & GSCAN\\_Beta & GSCAN\\_SE & GSCAN\\_N & GSCAN\\_Effective\\_N\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 7 & 1979425 & rs11770899:1979425:C:T & cg12073833 & 0.383123 & T & 0.000324 & -0.0113 & 0.00311 & 262990 & 257466\\\\\n",
       "\t2 & 7 & 2025040 & rs4721282:2025040:A:G  & cg12073833 & 0.371466 & G & 0.000533 & -0.0106 & 0.00307 & 262990 & 258825\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 11\n",
       "\n",
       "| <!--/--> | Chromosome &lt;int&gt; | GRCh37.p13_position &lt;int&gt; | SNP_ID &lt;chr&gt; | Gene &lt;chr&gt; | MAF &lt;dbl&gt; | GSCAN_Effect_Allele &lt;chr&gt; | GSCAN_pvalue &lt;dbl&gt; | GSCAN_Beta &lt;dbl&gt; | GSCAN_SE &lt;dbl&gt; | GSCAN_N &lt;int&gt; | GSCAN_Effective_N &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 7 | 1979425 | rs11770899:1979425:C:T | cg12073833 | 0.383123 | T | 0.000324 | -0.0113 | 0.00311 | 262990 | 257466 |\n",
       "| 2 | 7 | 2025040 | rs4721282:2025040:A:G  | cg12073833 | 0.371466 | G | 0.000533 | -0.0106 | 0.00307 | 262990 | 258825 |\n",
       "\n"
      ],
      "text/plain": [
       "  Chromosome GRCh37.p13_position SNP_ID                 Gene       MAF     \n",
       "1 7          1979425             rs11770899:1979425:C:T cg12073833 0.383123\n",
       "2 7          2025040             rs4721282:2025040:A:G  cg12073833 0.371466\n",
       "  GSCAN_Effect_Allele GSCAN_pvalue GSCAN_Beta GSCAN_SE GSCAN_N\n",
       "1 T                   0.000324     -0.0113    0.00311  262990 \n",
       "2 G                   0.000533     -0.0106    0.00307  262990 \n",
       "  GSCAN_Effective_N\n",
       "1 257466           \n",
       "2 258825           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.0269402510301528"
      ],
      "text/latex": [
       "0.0269402510301528"
      ],
      "text/markdown": [
       "0.0269402510301528"
      ],
      "text/plain": [
       "[1] 0.02694025"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(coloc)\n",
    "\n",
    "# read in input data\n",
    "qtls <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg12073833_baseline_with_qtl_maf.txt\", header=TRUE)\n",
    "head(qtls, 2)\n",
    "\n",
    "gscan <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\", header=TRUE)\n",
    "head(gscan, 2)\n",
    "\n",
    "meth <- read.table(\"/home/ubuntu/rti-nd/coloc_review/expression_file/mqtl_methylation_data/LIBD_NAc_MEGA_OMIC_full_matrixeQTL_input.txt\", header=TRUE)\n",
    "\n",
    "# flip GSCAN beta if the effect alleles don't match QTL effect alleles\n",
    "qtls$mQTL_Effect_Allele<-as.character(qtls$mQTL_Effect_Allele)\n",
    "gscan$GSCAN_Effect_Allele<-as.character(gscan$GSCAN_Effect_Allele)\n",
    "gscan$GSCAN_Beta<-as.numeric(gscan$GSCAN_Beta)\n",
    "for (i in 1:length(gscan$GSCAN_Beta)) {\n",
    "    if (qtls$mQTL_Effect_Allele[i] != gscan$GSCAN_Effect_Allele[i]) {\n",
    "        gscan$GSCAN_Beta[i] = -1 * gscan$GSCAN_Beta[i]\n",
    "    }\n",
    "}\n",
    "head(gscan, 2)\n",
    "\n",
    "meth <- meth[which(meth$cg_id==\"cg12073833\"),]\n",
    "meth_sd <- sd(meth[1,2:ncol(meth)])\n",
    "meth_sd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 001: non-filtered, beta/se, no stdev-methyl, qtl MAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n",
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 1.57e-05  1.46e-02  4.99e-05  4.56e-02  9.40e-01 \n",
      "[1] \"PP abf for shared variant: 94%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se\n",
    "my.res1 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, \n",
    "                                   varbeta=(qtls$mQTL_SE)**2, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\",\n",
    "                                   MAF=qtls$MEGA_MAF),\n",
    "                     dataset2=list(beta=gscan$GSCAN_Beta, \n",
    "                                   varbeta=(gscan$GSCAN_SE)**2, \n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\",\n",
    "                                   MAF=gscan$MAF))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 002: non-filtered, p-value, no stdev-methyl, qtl MAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 0.000091  0.023500  0.000160  0.040200  0.936000 \n",
      "[1] \"PP abf for shared variant: 93.6%\"\n"
     ]
    }
   ],
   "source": [
    "# p-value, mqtl maf, no sdY\n",
    "my.res2 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\", \n",
    "                                   MAF=qtls$MEGA_MAF),\n",
    "                     dataset2=list(pvalues=gscan$GSCAN_pvalue,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"), \n",
    "                                   MAF=gscan$MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 003: beta/se, with stdev-methyl, qtl MAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.1890    0.0351    0.6020    0.1120    0.0621 \n",
      "[1] \"PP abf for shared variant: 6.21%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se\n",
    "my.res3 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, \n",
    "                                   varbeta=(qtls$mQTL_SE)**2, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\",\n",
    "                                   MAF=qtls$MEGA_MAF,\n",
    "                                   sdY=meth_sd),\n",
    "                     dataset2=list(beta=gscan$GSCAN_Beta, \n",
    "                                   varbeta=(gscan$GSCAN_SE)**2, \n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\",\n",
    "                                   MAF=gscan$MAF))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 004: pvalue, with stdev-methyl, qtl MAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 0.000091  0.023500  0.000160  0.040200  0.936000 \n",
      "[1] \"PP abf for shared variant: 93.6%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se\n",
    "my.res4 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\",\n",
    "                                   MAF=qtls$MEGA_MAF,\n",
    "                                   sdY=meth_sd),\n",
    "                     dataset2=list(pvalues=gscan$GSCAN_pvalue, \n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\",\n",
    "                                   MAF=gscan$MAF))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# coloc 003\n",
    "Applying RINT function to methlylation data for sdY calculation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC91BMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqdnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///+aDNxdAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3df3xcZZ3o8SeTJoFtS1vakJYWWtsC1+VCk3ZJg6z8EETAUhZJtC21KP2BgrQLrNgWt6z11xqu7NWruEoLVVbWlbBBFi+/BK6CC4ReDWJXIuVKCyhQwLbhR5ucP+6cmTnJmWSekydnvpN5cp7P+/WyczLz5HvOZM6HJJMxUR6AoqlyHwCQBIQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQjJQqZR6rNwHEXLgnxZOqFDqo0UNkb5Ttn2QRhghDbAgfT6cnd1ckd6c7m8UPkfu2rRp010jemyBj6iMGCGFjll74i/ITq8+fP6ah0NX/6zCv/YF7WjtvPJ9nEYSIQ1gHtKa9LVrRvTYcp70z+hDjp55xfDfNXTMQ4WU8Yne4Nq971ExQyrbx2lEEdIAhUK6au3atc8PWlm2E+Q76R0ftTfWuw43JPW98HtGhlT4g+QRkqMKhVRY2U6QG9I7/kC8dzUN6eQnnnh4owrt516laqJDMtlnghHSAFFf2r3znbOOrDlkxl+t/M5r3uK+/2hP9pc8teYvD6uq++D/6s6+6/6/P7Zm6tJn9vq3dwYTXvnMrDGneV77NR84ZvKY8cd8rD271l/zWNeyI8Y2/djz9n1+Ts2MT/1pwEGFp1/dt+Mf9d0eMaLziv8+oXra+T/yv0bLO+bMIb157ZyaqcvzP5H0fQTOSm/MyV735lFKfW1wSOG7FXyQMpevXTM7N3fAxymxCGmAiJDeauo7Ke7LP0EOXBG8cfST/ju8emLmjUP+NRzSv6ZPRnWq583re88L3vUX+1vfGJ+54oZXT8hcHvNm+JDyp+tCKjii57MVucVnvFYgpB9mv/GZmpdHXkgLstd9UqlLfls4pOBuhUP63rS+uYTkKP80mnV1xvEDQvp6Jo7JYwaH9Km+N9ThXel3OC/3Rk04pFr/jfdnQzpsbOb26/3p/kZVdn11kOqG8CHlT9eFVHBEZnFqnP/vXx8YHNK43JsfH/gRSH9p98jn/Qavy1x1d/qbsjc0IQV3KxxSVf9cQnJU3rfa+SGdn764o9c7+JtvnfaA98edS9NvLt25c+cfvMf9lSfds/2rfmOLPe9h/+2Gf/959jNJEJKqvODvll/geZfcsGO/5/3xKv/c8r/gyiz68O2fy3zumPI/b5mTvpgbOqIB0/fs3JC+fF96x/v7luhGbE9fUdH6lvefR6ev+G7eMWcPacGWb/ufPMYe1HwEmjLPabx+pKq419OEFNytcEihuXn7TDBCGiAipLPT/3EPfTvR/030qvTW4f4p96X0RsUfM7dM9L+0Wh4OqfKRvB0dODR93TNetoIT02fy+/yNe9LfQ/mX3f0LB073WtXAJxt0I/wDWe4v+H56o9Eb9GTDjH2e96C/9tnCH4FpT2WuuVipT3m6kIK7FQ4pby5PNjgpIiT/c0jNKSv/x4NvZVb2nyDvTW9d5m+87L/LHd789L8r/bcfCYe0PLeHntubjx2Xyo6/38tWcKOX+T5EzUh/itrpX/FS/xENnK4JqdCIv0xfnHBeml9Y5buDQvpS+vJNf+0Tmo9A6rb0FXcqNWefNqTloe3HCs0lJCdFPNnw+wm502v8NW974RNkUnrrHzNbf5He+pY3I/3vl/03d4dD2paduu/U0Il6p5et4O705WfSl2ekL18acMIOnK4JqdCISXn/UXhxUEj+84YHlcp/Hjz3EXjnVx9Ibxz6J8+brVL/x9OGtC20/VihuYTkpKinv5+5sCZ3TvonRv8JMjG99bXMlv/l2re96UFIu8IhPZSd6n+Ho05cvmbNIenLNi/33HX6cq3KvurnhQEn7MDpmpAKjZiYF9LOwj9H0oTkeX/yb/mm503OmxJuInS38p/+Ds8lJCdFv0Ro/y9uXt+Qfqvmragv7fwZq/23Hw6HlDtZj01vfiZ9eaDaMCTDL+0KjfDf9cbwwmGF5PlPqK8bKqTHBmwTEnwRIR3IXn1wavrN5zzPf04u+7zxwKcDLgveXlYgpPG5fu5WhiEZPtlQaMTq9EVj7hm5l37phY/ZIKTf+LdcXXRIoX0mGCENEBHS1Rd8///1er0/8Z+GftXz/iF9Me3B53a+ln2CuvGn/zd4+jvzHMNJdz6c/QHQgJD8n4Ge80bvQ0eahjRw+jBC6vAvW571vNfv/vihm7zwMUeHdPITT/zie/6T6OpWz9t6U8b1/ptfuSn83KNRSKF9JhghDRARkn+aVk/OfJt0cvrae3L/iV6b9yPTSeEfyI4tENKV/nWVY1VVlWFIg6abh+T9bea9Ds286GFT/jFHhtRnev+LYzVPNgwZUmifCUZIAwwRUtbUp9PXHmzoO0EOXB7ccnTmieRXsi/TGXuH/+9/eXln3KvHZG6r3DrZNKSB04cRUs/ngpcIKXVD/jEbhXTkU/3Xxw0ptM8EI6QBIkL6w82rTjr60Koj3v+V1zO3v/aZOdXBCdKx5r3jqo4485u5Vxvs//zc6qnLfveUfwLt8fLPuD3rZlbVLn7UMw5p4PRhhOR5O66aP6ly7NzzW3874JiHDOnQoz70zfD/WSNuSHkfp8QipFL6XPr8OabcB4GRQEil8B8r7k9/8nj96/7TA18s98FgJBBSKbQplZoyJfPtyYn7h16O0Y+QSqGt77v1M18aejUSgJBK4dVvfeS4w8dMPOHS+8t9JBghhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQUE1LvjvZtt7bv6BU7GGC0ih9S9+bpKmPG5m7BAwJGo9gh7VuoUg0tq1a31KdU037JQwJGn9ghbVDLdme3di1RG6UOBxidYoc0e0FPsNkzf67MwQCjVeyQqtf1b6+tkTgUYPSKHVLt4v7tRXUShwKMXrFDWpK6JdjcUrFU5mCA0Sp2SF0TVMP6rW1tW9fXq4ldkocEjD7xf47U2ahyGjsFDwgYjYp5ZUNH68rm5pWtHWIHA4xWpXmtXc/P7utz7/dLsosR8OJ9wl4s9z1CyZQmpOdqJ/UZr94tyT5K79Kqw0RVXVrue4SSKTakJ2668c59kSt+od4pch/lcsniTlGLLyn3PULJxA7pwY2ved7Lp/pPNky5K2ohIRGSA2KHdF5tj9fbpKZfsvYMVR31fAMhEZIDYoc07UzPu099yH/d950VF0YsJCRCckDskKpaPO+L6unM9rlTIhYSEiE5IP5r7U71vI1qb2b78uqIhYRESA6IHdL5Nbu929TDme2mWRELCYmQHBA7pPvVyS93z33vDs979zp1ZcRCQiIkB8T/OdK1auyyz1SOOeGUKWrWKxHrCImQHFDED2Rvnpp9zWrFhbujlhESITmgmFc2vHPPFy7/9IZbdkWvIiRCckDpf0EkIRGSAwhJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISU86pL8+4VpZj5b7I4Q+hKQnHdLRtU2iavkMZw9C0hMPiS8Vk4uQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEmPkGCsmJB6d7Rvu7V9R2/0KkIKEFKCxQ+pe/N0lTFjc3fUOkIKEFKCxQ5p30KVamhZtbqlPqWa9kcsJKQAISVY7JA2qGW7s1u7lqiNEQsJKUBICRY7pNkLeoLNnvlzIxYSUoCQEix2SNXr+rfX1kQsJKQAISVY7JBqF/dvL6qLWEhIAUJKsNghLUndEmxuqVgasZCQAoSUYLFD6pqgGtZvbWvbur5eTeyKWEhIAUJKsPg/R+psVDmNnVHrCClASAlWzCsbOlpXNjevbO2IXkVIAUJKsNK81u7lc8/sc5J6uyT7KD1CgrHShLRv07V9LuYzUg4hJRiv/tYjJBgrMqRHzpk8bt4NB6KWEFKAkBIsdkh1V6b/ua0y87Td4qj/JwUhBQgpwWKHpFZ43qvjUtc9t+eOaer7EQsJKUBICVZUSN9RmRfc/VKdFbGQkAKElGBFhfRp9ZvMdn1txEJCChBSghUV0idU9v8be2FVxEJCChBSghUV0pfUy5nt06dELCSkACElWPyQUjU1Ver+zPashoiFhBQgpASLHdJxGV/2NzvUZRELCSlASAkm8cqGx1ufiriVkAKElGC8REiPkGCMkPQICcYISY+QYIyQ9AgJxghJj5BgjJD0CAnGCEnP9pCajl0t6yfl/oiPYoSkZ3tIR8+8SNRMPsPFR0h61ofEl4r2ICQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgjJD1CgjFC0iMkGCMkPUKCMULSIyQYIyQ9QoIxQtIjJBgrJqTeHe3bbm3f0Ru9ipAChJRg8UPq3jxdZczY3B21jpAChJRgsUPat1ClGlpWrW6pT6mm/RELCSlASAkWO6QNatnu7NauJWpjxEJCChBSgsUOafaCnmCzZ/7ciIWEFCCkBIsdUvW6/u21NRELCSlASAkWO6Taxf3bi+oiFhJSgJASLHZIS1K3BJtbKpZGLCSkACElWOyQuiaohvVb29q2rq9XE7siFhJSgJASLP7PkTobVU5jZ9Q6QgoQUoIV88qGjtaVzc0rWzuiVxFSgJASrESvtXv6yT43E1IOISVYaULqqlAhb5dkH6VHSDBWfEiXbi1w5Z/39Pkpn5FyCCnBig9JXRp9O98jBQgpwWKHtDGg6tP/RCwkpAAhJVjskFSeiIWEFCCkBIsf0rjP35ihmtL/RCwkpAAhJVjskNqPmPaT7AS+RzJESAkW/8mGP12gPvGmR0jmCCnBinnW7ubxR91LSOYIKcGKevp75/vVZXsJyRQhJVhxP0fq+VrNbEIyRUgJVuwPZH99IiGZIqQEK/qVDb0HeqIXEFKAkBKM37SqR0gwRkh6hARjhKRHSDBGSHqEBGOEpEdIMEZIeoQEY4SkR0gwRkh6hARjhKRHSDBGSHqEBGOEpEdIMEZIeoQEY4SkR0gwRkh6hARjhKRHSDBGSHqEBGOEpEdIMEZIeoQEY4SkR0gwRkh6hARjhKRHSDBGSHqEBGOEpEdIMEZIeoQEY4SkR0gwRkh6hARjhKRHSDBGSHqEBGOEpEdIMEZIeoQEY4SkR0gwFg5pT0n2QEgBQkqwcEiHrni0BHsgpAAhJVg4pGOUOvFbb0rvgZAChJRg4ZB6H2ipVmMvfVx2D4QUIKQEG/Bkwx+/Okep+f+8V3APhBQgpAQb9Kxd770fqVLjL+sU2wMhBQgpwQY//f38dUcopSo+9obQHggpQEgJNiCkg/9+bkod9YVd/3GaWia0B0IKEFKC5YX0h03TVcVZbQfTm72LJgntgZAChJRg4ZA+XKkOv+rZ3BtfknrNAyEFCCnBwrmok7a81fdGxzeF9kBIAUJKsHBIT5ZkD4QUIKQE40WreoQEY+GQbj/thczlC6f+m+AeCClASAkWDunMBbmNeWcL7oGQAoSUYOGQ6tbkNi49UnAPhBQgpAQLh1S1IbexvlpwD4QUIKQEC4c0tTm30XyE4B4IKUBICRYO6aKa32Yun6n5G8E9EFKAkBIsHNKjqcO/8Wz3s984PPULwT0QUoCQEizv50g3VSpf5U2SeyCkACElWP4PZH+1pn5W/WW/Ft0DIQUIKcF4ZYMeIcEYIekREowRkh4hwVheSA8tqquuzBDcAyEFCCnBwiHdlVITjp+XIbgHQgoQUoKFQ1pQ+YNe+T0QUoCQEiwcUs1FpdgDIQUIKcHCIR3+6VLsgZAChJRg4ZBaFmiXFYGQAoSUYOGQnq+7/qD8HggpQEgJFg5pxRlq5uIVGYJ7IKQAISVY3q/j6ie4B0IKEFKChZPZ3k9wD4QUIKQE4yVCeoQEYwNCev5RqT9C0YeQAoSUYHkhPXaiUvd53r8c/5DgHggpQEgJFg7pmbHjFvsh7R17ueAeCClASAkWDmlp9a9f8UPyPsyLVn2EBGN5vyDyo142pGsmG71v7472bbe27xjiha6EFCCkBAuHNObaXEjXmvyCyO7N07M/c5qxuTtqHSEFCCnBwiHVfjIX0gdnDv2O+xaqVEPLqtUt9SnVtD9iISEFCCnBwiFdUPd2JqQHKgxeIrRBLdud3dq1RG2MWEhIAUJKsHBIP0+d87Bqf/yqqqpfDf2Osxf0BJs98+dGLCSkACElWP4viByT+aan6haDd6xe17+9tiZiISEFCCnB8l/Z8PQVC2bNW/O0yTvWLu7fXlQXsZCQAoSUYLFfa7ck1fd5a0vF0oiFhBQgpASLHVLXBNWwfmtb29b19WpiV8RCQgoQUoLFf/V3Z2Pwf15q7IxaR0gBQkqwcEhz+hm9b0fryubmla0d0asIKUBICRYOaULGGKUOm1Dk1D2fXt1nMSHlEFKCDf7S7t3/PHnRu0VOJaRCCCnBCn2PtGfa9cOYsOfP0bfzpV2AkBKs4JMNy99j8J47V5227hXviRNUxft3RK0jpAAhJVjBkFYavPr7lalKqYaX69S0SnXk6xELCSlASAlWKKSX6gw+I31eXfzQFepDs572/nyB+kLEQkIKEFKChUPalLHx44dFhpEz74gDXu9sdXt689VDF0YsJKQAISVYoV8Qeeg1Pdr1fSaem/6nRb3sb58yMWIhIQUIKcHy/tBYxt2P7jV5x0Oa0/+syb77RWMiFhJSgJASLPZLhI4+I/3P2rGZ7TN49bcJQkqw2CGdPaN/ezbfI5kgpASLHdL16g/B5nb12YiFhBQgpAQLhzQzX/Q7Hnyr79dwPfaVZyIWElKAkBIsHNLkiUqpsen/TZzsE9oDIQUIKcHCIe09Zf7de729dzecYvS8nSFCChBSgoVDWjc7+/vp9s9eV3hxLIQUIKQEC4c0/ZrcxjUzBPdASAFCSrBwSNVX5zaujvr1WsNFSAFCSrBwSMfO2pe53DfzvwnugZAChJRg4ZC+rua1vea91jZP3Si4B0IKEFKChUPqWaWU8n/Z6mqDF60aI6QAISVY/isbHlxx4swTV/xMdA+EFCCkBOOvmusREozxV831CAnG+KvmeoQEY/xVcz1CgjH+qrkeIcFYMX/V3AwhBQgpweL/VXNThBQgpASL/VfNjRFSgJASLPZfNTdGSAFCSrDYf9XcGCEFCCnBYv9Vc2OEFCCkBIv9V82NEVKAkBIsHNJj20uxB0IKEFKChUOq+Egp9kBIAUJKsHBIU5aXYg+EFCCkBAuHdNFxB0uwB0IKEFKChUP63eQr9svvgZAChJRg4ZBWnK6mnPnxFT7BPRBSgJASrNAfGksT3AMhBQgpwcLJbO8nuAdCChBSgvWF9C+/LNEeCClASAnWF5Lyvy+64Wz5PRBSgJASLD+kFSX4pUKEFCCkBCMkPUKCMULSIyQYIyQ9QoKxJIX0T7Nljbf8xCcki/SHVDVhwoQqNSFLcA8jF9Il8zaJOtzyE5+QLNIfUh7BPYxgSJafqLbPI6Qi9CXzVh7BPRDSaJlHSEVI0l+jIKTiEFIRCEnP9hOfkCxCSHq2n/iEZBFC0rP9xCckixCSnu0nPiFZhJD0bD/xCckihKRn+4lPSBYhJD3bT3xCsggh6dl+4hOSRQhJz/YTn5AsQkh6tp/4hGQRQtKz/cQnJIsQkp7tJz4hWYSQ9Gw/8QnJIoSkZ/uJT0gWISQ92098QrIIIenZfuITkkUISc/2E5+QLEJIeraf+IRkEULSs/3EJySLEJKe7Sc+IVmEkPRsP/EJySKEpGf7iU9IFiEkPdtPfEKyCCHp2X7iE5JFCEnP9hOfkCxCSHq2n/iEZBFC0rP9xCckixCSnu0nPiFZhJD0bD/xCckihKRn+4lPSBYhJD3bT3xCsggh6dl+4hOSRYoJqXdH+7Zb23f0Rq8ipNEyj5CKED+k7s3Ts39vdsbm7qh1hDRa5hFSEWKHtG+hSjW0rFrdUp9STfsjFhLSaJlHSEWIHdIGtWx3dmvXErUxYiEhjZZ5hFSE2CHNXtATbPbMnxuxkJBGyzxCKkLskKrX9W+vrYlYSEijZR4hFSF2SLWL+7cX1UUsJKTRMo+QihA7pCWpW4LNLRVLIxYS0miZR0hFiB1S1wTVsH5rW9vW9fVqYlfEQkIaLfMIqQjxf47U2ahyGjuj1hHSaJlHSEUo5pUNHa0rm5tXtnYUuOnPe/r8lJBGybzzl+6R1TP0Y5YYpXmtXVeFCnm7JPsYjJCKM0cJWzf0Y5YYJXrR6tNP9rmZz0ijZd6pt4s61aUvFXn1d8SJxbyiOPU9V/yQem5bc+V92c0bzo5YR0iuziMkEwfP878KvvBNf3tF1BRCcnUeIZm4SdV99duNasHrHiExryBCMnHymB3pL+/+XjW+SUjMK4iQTIw/NXPxTfW+fYTEvEIIyURNc/ayVZ3eTUjMK4CQTMw9ObexSX1oCSExbzBCMnFR9Ru5rb9VlYTEvMEIycQP1E3B5ipFSMwbjJBM/PnGHwebPV+7NmIhIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCEkUIbk6j5BEEZKr8whJFCG5Oo+QRBGSq/MISRQhuTqPkEQRkqvzCMlQ7472bbe27+iNXkVIrs4jJCPdm6erjBmbu6PWEZKr8wjJxL6FKtXQsmp1S31KNe2PWEhIrs4jJBMb1LLd2a1dS9TGiIWE5Oo8QjIxe0FPsNkzf27EQkJydR4hmahe17+9tiZiISG5Oo+QTNQu7t9eVBexkJBcnUdIJpakbgk2t1QsjVhISK7OIyQTXRNUw/qtbW1b19eriV0RCwnJ1XmEZKSzUeU0dkatIyRX5xGSoY7Wlc3NK1s7olcRkqvzCKlo+zZd2+diQnJ0HiEV7eVzz+xzknq7JPsYjJDsmkdIovjSztV5hDRcV8+MuJGQXJ1HSMO1ImoKIbk6j5CGi5CYVwAhmfhoyCxCYt5ghGT0jnkiFhKSq/MIycTYY+/q8wFCYt5ghGTi5MP6f1cD3yMxrwBCMnG56n+lKiExrwBCMvHjBQ/1b/N/NWfeYIQkipBcnUdIogjJ1XmEJIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkipBcnUdIogjJ1XmEJIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkipBcnUdIogjJ1XmEJIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkipBcnUdIogjJ1XmEJIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkipBcnUdIogjJ1XmEJIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkipBcnUdIogjJ1XmEJIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkSh/SrWfKOtLyE8u1eYQkSh/SJcd8UtRhlp9Yrs0jJFERIVl+IjCvOIQkipBcnUdIogjJ1XmEJL7xJOYAAAh/SURBVIqQXJ1HSKIIydV5hCSKkFydR0iiCMnVeYQkipBcnScd0t4nhe0VPDhCYl6p5kmHdJkSdpngwRES80o1TzqkS879hahzJY+PkJhXqnniIdl8fITEvFLNIyRRhOTqPEISRUiuziMkUYTk6jxCEkVIrs4jJFGE5Oo8QhJFSK7Oazp2tahjCakwQkr4vJkXiRpHSIUREvPKOY+QNGx/4Jhn1zxC0rD9gWOeXfMIScP2B455ds0jJA3bHzjm2TWPkDRsf+CYZ9c8QtKw/YFjnl3zCEnD9geOeXbNIyQN2x845tk1j5A0bH/gmGfXPELSsP2BY55d8whJw/YHjnl2zSMkDdsfOObZNY+QNGx/4Jhn1zxC0rD9gWOeXfMIScP2B455ds0jJA3bHzjm2TWPkDRsf+CYZ9c8QtKw/YFjnl3zCEnD9geOeXbNIyQN2x845tk1j5A0bH/gmGfXPGtC6t3Rvu3W9h290asIiXl2zrMkpO7N07N/92zG5u6odYTEPDvn2RHSvoUq1dCyanVLfUo17Y9YSEjMs3OeHSFtUMt2Z7d2LVEbIxYSEvPsnGdHSLMX9ASbPfPnRiwkJObZOc+OkKrX9W+vrRlw43O1k/qMV+9qRlxadZioFPOYNwxVl8Y9+QuIHVLt4v7tRXUDbuz52X197v2+bsSL98m6/XbmMW8YXox78hcQO6QlqVuCzS0VS2UOBhitYofUNUE1rN/a1rZ1fb2a2CV5SMDoE//nSJ2NKqexU/CAgNGomFc2dLSubG5e2dohdjDAaFX619oBDiAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkvSYFqzSV+4yIQEh6Sxc9mQiLknI/bP5lVYSkd4nkb+IsI+7HCCAkPasfuGHgfowAQtKz+oEbBu7HCCAkPasfuGHgfowAQtKz+oEbBu7HCCAkPasfuGHgfowAQtKz+oEbBu7HCCAkPasfuGHgfowAQtKz+oEbBu7HCCAkPasfuGHgfowAQtJbvbrcRyCD+zECCElvz55yH4EM7scIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIIKRA19K6mrkb9/dfsa3vryAcLHi7pYa4H8dlN+vKdnymBn+8e+84Y/oh77noUd3tZUVIOZ0TKxatna+auvuueXRFRqM6vfDtdhrqfhyXyrx1ZdkO0FCBj/en1YSL156Tqtiqub2sCCmnUW3xvJ4lavPAG85RP4y83TJD3Y/jakb8kGIZfD9+r6bsTl/cqY4qfHt5EVJWh6r3L3alZvTm3/B8qvadqNstM9T9GC0hFbgf96tz/YueMYcWvr28CCmrVa3PXNarHfk3bFCfjbzdMkPdD++4qi998vJ/fm2kj2u4CtyPXZW1L6Uv7lIXFL69vAgpa6XKfOXttaj2vOsPTK14Nup22wx1P3JPNoy7bcSPbHgK3Y8vqonL15035rxXNLeXFSFlNau2zOVqdWve9f+mzoy83TZD3Q/vy/e91P30FanKR0b6yIan4P247bD0fwOOu017ezkRUlbwwKxS2/KuP0v9KPJ22wx1P3I2qnNG7pjiKHQ/rq/47M79HR/MfFFn3eNBSFmaLxV+X1H3btTt1hnqfuQ8pyaP4EHFUOB+/G+1xL/oPqryeQsfD0LKCr55bcj/5vXa3NW6260z1P3I2aPGjdwxxVHgflypvpu5bFZ3Wvh4EFJWh2rwL3anpoefTn33iIrnom63z1D3I6dNzRvJoxq+AvfjMvXFzOX71T0WPh6ElNOobvG8nmXZH/BtufGPmStvV2cXvN1iQ9yPx3/l//vEkeqG8hyescH34wdq6gvp7faKv3hj4O0WIKSczgmpxesWqIWZl5zMUU9krjxD3VHwdosNcT9a1ZwzL2yoUOe/qxtgicH34+DpauxHrzxLqZsG3W4BQgp0Lamtnr1hX2Y7dwL+rmLagYK32yz6fjy16oTDx0w5a5slXxBFGHw/3vl647jK2kUPDL7dAoQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCCkxNiuVpT7EBxGSKPes+qjmUtCKidCGvUIyQaENOoRkg0IyX7pQp79m0njz/kv78UVRxzyviczVz56YV3VtGW/9byvqIxt/rI/LJl8yF/dXebDdRMh2W+7On1K0+UfUNOenVH/qfPUpNfT1303VfuJa1uqx/7Se/oG1bRt27bn0svOqJv/qY9Uph4p9wG7iJDst12pf0hfrFST1vZ63nXqq573TNXZ3emrfjXuxPCXduq69O3b1KKyHq2jCMl+29XMg+mLn6tJfjxd6mOed4V6+BXfYvV8KKSjD6QveifUlfNgXUVI9tuuLvAvdqpT/Iu3/IsFKvBYKKTFmcvjq8t1oC4jJPvlno57QZ3nXxxQCz1vlmq/L+uNQc/azass14G6jJDsVyCkeerxvpsJyQaEZL8CIa1RV/Xd/Jy6KLyMkMqCkOxXIKTOMVUP+G/s/aHnvakaw8sIqSwIyX4FQvK+N6bi7M/93aKxx6e3F6qPXb+5k5DKipDsVygkb/vyo6onHX/Zz9Kbz354UkX2lQ2Z5YRUDoQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIg4P8DFfMo+E4tg6YAAAAASUVORK5CYII=",
      "text/plain": [
       "Plot with title “Histogram of meth B4 rint”"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>-0.247468774342615</li><li>-0.89988431473858</li><li>0.028488265166309</li><li>-0.428644689178854</li><li>-0.416187607396528</li><li>1.37806954235926</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item -0.247468774342615\n",
       "\\item -0.89988431473858\n",
       "\\item 0.028488265166309\n",
       "\\item -0.428644689178854\n",
       "\\item -0.416187607396528\n",
       "\\item 1.37806954235926\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. -0.247468774342615\n",
       "2. -0.89988431473858\n",
       "3. 0.028488265166309\n",
       "4. -0.428644689178854\n",
       "5. -0.416187607396528\n",
       "6. 1.37806954235926\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] -0.24746877 -0.89988431  0.02848827 -0.42864469 -0.41618761  1.37806954"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.999355523800719"
      ],
      "text/latex": [
       "0.999355523800719"
      ],
      "text/markdown": [
       "0.999355523800719"
      ],
      "text/plain": [
       "[1] 0.9993555"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC/VBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///91tUC1AAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dC5wU1Z3o8TMDMyMZZUBEQBBYRE10RWAiYPQaom40MQRXhQis4q48dNVA1F2DxqDhZuMGE71x1SR7N6AkWd1dQxZfK4mP3ESTrAQNrmEDEddnjE/kZYSZ+tyqnq4z1TNzqk7X/Lu7ztTv+/k4XT19us+Z6vo5MzXNjPIA9Jqq9QKAvoCQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEid+imlnqj1IiL23jy1pU6pz/TqQWr/QdV+BdWQ15Ba/Wf3tI7N+f7myGCj52f83uXLl99b1bWFzlYFKUKKrDn2ML6oMMGz4dVW1WmVN0aVeqTL7Xp8w4Gtl/3KvBjjCmq3ZyuAkJJCWuy/d3FV11b0ZHCQ7jd6zKXl3zWy5riQ3h9SCOHq8HrKkAJ11xsXY1xBzfZsJRBSZ0iXL1my5PluI2v2dH/Tn/jQHanuahnSDzsiGNNevN6LkIKbDXrerR4h9Qk9hdSzmj3dN/oTn5LurpYhnVNs4CfF68FOGbOkwy+864OLwhd/JxXes6XL7R3jj7r91i9+MBh0YW9W6T5C6v6l3R+/+WeHNO036sMLvvmmN1P/H3dIMORXi48a2DDs4/+wu+Ouu754RNPwuc/uCG7fFD7C65eN7T/d89ZdecrhQ/ofcPi56zrGBmOe2Drv4OZp93jezmsPaxp18R+6LCr66Ffoif9V3x7zEJsu/dOWxhGf/tfgk0vJmgtL2n7VYU3Dz+v6aWH7fv5tn/T/W9R9p4TeDh7mJsPt4fV3Bvsb06N3i+6IcLcWLt+8clxxJV32rOsIqVtIe6bpp3h96dO999Lwyugngzu8MaFwZb9/iYb0L4f6bz7qecfqe575fjA42PrGAYV33PjGMYXLw7dHl1T66KaQenyItr+tKw4++c0eQrrrTwrXhr9Yug/+r/++1p/7bwb/sdtOCVmF5B3vb8yI3i26I6Ih/dMIvRJC6hOCY2DsFQVHdwnp64U4hvTvHtLF+oo6cKt/hzOKV5qiIQ0NrpzUEdLA5sLt1wWPHmw0dIxvDFO9Orqk0kc3hdTjQxQG1+8fvP1fe7uHtH/x6vml+2C6/64bvXH+2x902ylfLI7pFlL09jCk94b7G1+MPnR0R0RDauhcCSH1CSXfJ5eG9OngyGr39v3XrdN/7L22ba5/de62bdte8H4ZjDzuwY03BI3N9LzHguuT/v2nHZ9JwpBUvzP/5rwzPe+CGzfv8rzXLg+OlOALrsKgT939+cLnjoP+zx2H+RfjIyvq8uhvbbvavzzBn3iXHmJ6iI3+O+pW7vF+Mdp/x7dL1tyxpNZVtwWfCpr3RXfBi/X+vV70lvk3nNVtp4QHd7eQorcH14975OHvn+pfDnot+tjRHRENKbKSklW6j5C6hXSa/z/3yLcTnd8SL/S3DgxOo33Z36h7rXDLoOBLq/OiIfX7SclEeweojp/UBGMm+EfyCcHGg/73UMHl7s6BXR/dW6m6nmwwPUSwkPOCAd/1N6Z43U42jNrpeQ8HY7dEH+0GVfiE8Wv/ountrjvFOqSiw58uWWl0R0RDKlkJJxv6gJiQgs8hTScu+NrDewojO5/uD/lbFwUbvw/u8gNvsv92QXD9J9GQzivO0Hb3rCP2r+94+B95HRUER+RfBceT/ylqW/COVztX1PXRDSH19BBH+RfHnOELCuv3freQvuxfbg/G/mf00YJvsm73Ou797a47pcyQjvxt6e6N7ohoSCUrIaQ+IOZkw+9aikfHAVe+50Wf7uDc1N8Xtj7gb93qjfLf/l1w9eVoSGs6HnXnRyOd/tDrqOB+//Iy//Jk//LV4B2Rb/+7ProhpJ4eYnDJ/xRe6RZScN5wn1Kl58Gf9q/3f93fWKEKn5m8dCcbigY9VXK3yI4oCalkJYTUB8Sd/n72rKbi4RE8zZ1P9yB/66uFreDLtdu8kWFIL0VDerTjUYPvcNSE8xYvDs4xr/WK5679yyWq41U/L3YJqeujG0Lq6SEGlYS0reefI3UN6crofer+p8tOCVmctXv79uAUwgf3Ru8W2RGlp7+jKyGkPiD+JUK7fvadZZP8a0174r60Cx6j8COYx6IhFQ/WI/zNy/zLvY2WIVl+adfTQ3wocrAXWITUNrKkvr/rslNCVqe/vxYM+mb0btEfAxNSXxYTUvF/rfuCc7rPeV5wTu78wnu6ng64KLw+r4eQDij2c7+yDMnyZENPD7HIv5hSPCP36s+96JqNIf2opCN1VJedErIK6f2x/sbY6Kckq5Aiq3QfIXUL6Yozv/s/7V77fcFp6Dc873r/YsTDz217s+ME9ZT/eCo8/V04x3DcDx/r+AFQl5CCn4F+4p32Rw+xDanro5cR0obgcvYW/7C///wBy73omo0hXeBf/fCqgluCm37l2YVU/DnSFd+Pji88wprI3axCiqzSfYTULaTgMG0cUvg26Xj/vQ8W/5+9pORHpoOjP5Bt7iGkzwbv69esGhosQ+r26PYheZ8r3GtA4UUPy0vXbAhpz0D/6j8UrwTn7S737EIKzYuO33WQv3VMe+fdrEKKrNJ9hNRjSB2GP+O/d98k/XTvvSS8ZXThRPLrHS/Taf5B8Pa/vZLj543DC7f1Wz3ENqSuj15GSG2fD18iFLxWoWTNhpDuCq6GPwm9yt8e0daLkAqfXFTk3xZZhRRZpfsIqVtIL3xn4XGjBzQcfNJXOn5M+eZlhzWGT/eGxR/av+HgU28pvtpg17XjG4fP++2vgsPhLa/0+Hlr6ZiGoTMf96xD6vroZYTkeZsvnzy4X/P4T6/8TZc1G0Ka4V+bGF75aXDbQ70J6c3gk/IJnXezCqlkz7ouryGJ+rx/NBxe60WgpgipFx6Y/yP/k8fbXw9OD/zvWi8GNUVIvbBWqfqDDip8ezJhV/Jw9GGE1Atr9XcMp76aPBp9GSH1whu3nn3kgf0HHXPhj2q9EtQaIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCAkQQEiAAEICBBASIICQAAGEBAggJEAAIQECCCkTnlqfwt13p7nXU7X+WPsmQsqElg8MLF99Q4o7faCl1h9r30RImdB866byjZ6Z4k63Ntf6Y+2bCCkTCMl1hJQJhOQ6QsoEQnIdIWUCIbmOkDKBkFxHSJlASK4jpEwgJNcRUiYQkusIKRMIyXWElAmE5DpCygRCch0hZQIhuY6QMoGQXEdImUBIrutNSO2b1625c93mdrHF5BchuS59SLtXjFQFo1bsFlxQPhGS61KHtHOqqp80e+Gi2RPr1bRdkkvKI0JyXeqQrlbzXu7YemmOukZqOXlFSK5LHdK41rZws23yeJnF5BchuS51SI1LO7eXNEksJc8IyXWpQxo6s3N7xjCJpeQZIbkudUhz6u8IN1fVzZVZTH4RkutSh7S1RU1atnrt2tXLJqpBWyWXlEeE5Lr0P0faNEUVTdkkuKB8IiTX9eaVDRtWLpg1a8HKDWKLyS9Ccl1lXmv3uybVqd++iszRpxCS6yoTUvtjnb+0/Wb1x4rM0acQkut6H9KFq+Nv/xkhJSMk1/U+JHVh/O2EZIGQXJc6pGtCaqL/JmYgIVkgJNelDkmViBlISBYIyXXpQ9r/2psK1DT/TcxAQrJASK5LHdK6g0fc1/EIfI/Ue4TkuvQnG/5wpvrL7R4hiSAk1/XmrN13Djj0IUISQUiu69Xp720nqYt2EJIAQnJd736O1PbVpnGEJICQXNfbH8j+egIhCSAk1/X6lQ3te9viBxCSBUJyXeV/0yohWSAk1xFSJhCS6wgpEwjJdYSUCYTkOkLKBEJyHSFlAiG5jpAygZBcR0iZQEiuI6RMICTXEVImEJLrCCkTCMl1hJQJhOQ6QsoEQnIdIWUCIbmOkDKBkFxHSJlASK4jpEwgJNcRUiYQkusIKRMIyXWElAmE5DpCygRCch0hZQIhuY6QMoGQXEdImUBIriOkTCAk1xFSJhCS6wgpEwjJdYSUCYTkOkLKBEJyHSFlAiG5jpAygZBcR0iZQEiuI6RMICTXEVImEJLrCCkTCMl1hJQJhOQ6QsoEQnIdIWUCIbmOkDKBkFxHSJlASK4jpEwgJNcRUiYQkusIKRMIyXWElAmE5DpCygRCch0hZQIhuY6QMoGQXEdImUBIriOkTCAk1xFSJhCS6wgpEwjJdYSUCYTkOkLKBEJyHSFlAiG5jpAygZBcR0iZQEiuI6RMICTXEVImEJLrCCkTCMl1hJQJhOQ6QsoEQnIdIWUCIbmOkDKBkFxHSJlASK4jpEwgJNcRUiYQkusIKRMIyXWElAmE5DpCygRCch0hZQIhuY6QMoGQXEdImUBIriOkTCAk1xFSJhCS6wgpEwjJdYSUCYTkOkLKBEJyHSFlAiG5jpAygZBcR0iZQEiuI6RMICTXEZKwd2+6IYXGqoX0tcY067tpR633a9YRkrAH6o9KQVUtpL9Otb76B2u9X7OOkITdNyDF0b2peiFdnGp9+91f6/2adYQkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsqn3oTUvnndmjvXbW6PH0VIFgjJdelD2r1ipCoYtWJ33DhCskBIrksd0s6pqn7S7IWLZk+sV9N2xQwkJAuE5LrUIV2t5r3csfXSHHVNzEBCskBIrksd0rjWtnCzbfL4mIGEZIGQXJc6pMalndtLmmIGEpIFQnJd6pCGzuzcnjEsZiAhWSAk16UOaU79HeHmqrq5MQMJyQIhuS51SFtb1KRlq9euXb1sohq0NWYgIVkgJNel/znSpimqaMqmuHGEZIGQXNebVzZsWLlg1qwFKzfEjyIkC4Tkusq81m7X127QLiakZITkusqE9MqJrdqR6r2KzJFRhJRPvQ/prXfjb+dLOwuE5Lr0IW1bOH3p695/HqPqTtocN46QLBCS61KH9PpwpdSk3w9TI/qpQ96OGUhIFgjJdalDulb9xaOXqtPHPuO9e6b6UsxAQrJASK5LHdKxB+/12sepu/3NNwZMjRlISBYIyXWpQxr0Sf/NbPX7YPvEQTEDCckCIbkudUj7zfLfLO64+zn9YwYSkgVCcl3qkEaf7L9Z0lzYPplXf2uElE+pQzptVOf2OL5H0ggpn1KHdJ16IdzcqP42ZiAhWSAk16UOad8e/Wu4nvjKszEDCckCIbmOXxApjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsonQhJGSPlESMIIKZ8ISRgh5RMhCSOkfCIkYYSUT4QkjJDyiZCEEVI+EZIwQsqnaEhvVWQGQrJASK6LhjRg/uMVmIGQLBCS66IhHa7UhFu3S89ASBYIyXXRkNp/PLtRNV/4S9kZCMkCIbmuy8mG1244TKnJ39ohOAMhWSAk13U7a9f+0NkN6oCLNonNQEgWCMl13U9/P/+Fg5VSdee+IzQDIVkgJNd1CWnfv3+yXh36pZcemK7mCc1ASBYIyXUlIb2wfKSq+7O1+/zN9hmDhWYgJAuE5LpoSJ/qpw68fEvxypelXvNASBYIyXXRXNRxq/boKxtuEZqBkCwQkuuiIT1ZkRkIyQIhuY4XrQojpHyKhnT39BcLly9+9N8EZyAkC4TkumhIp7YWN449TXAGQrJASK6LhjRscXHjwkMEZyAkC4TkumhIDVcXN5Y1Cs5ASBYIyXXRkIbPKm7MOlhwBkKyQEiui4Z0TtNvCpfPNv254AyEZIGQXBcN6fH6A7+xZfeWbxxY/zPBGQjJAiG5ruTnSLf3U4F+t0vOQEgWCMl1pT+QfXrxxLETL/q16AyEZIGQXMcrG4QRUj4RkjBCyidCEkZI+VQS0qMzhjX2KxCcgZAsEJLroiHdW69ajj62QHAGQrJASK6LhtTa73vt8jMQkgVCcl00pKZzKjEDIVkgJNdFQzrwrysxAyFZICTXRUOa3Woc1guEZIGQXBcN6flh1+2Tn4GQLBCS66IhzT9ZjZk5v0BwBkKyQEiuK/l1XJ0EZyAkC4TkumgyGzsJzkBIFgjJdbxESBgh5VOXkJ5/XOqPUGiEZIGQXFcS0hMTlFrvef989KOCMxCSBUJyXTSkZ5v3nxmEtKP5EsEZCMkCIbkuGtLcxl+/HoTkfYoXraZGSPlU8gsiP+N1hHTlEMEZCMkCIbkuGlL/q4ohXcUviEyNkPIpGtLQvyqG9PExgjMQkgVCcl00pDOHvVcI6cd1vEQoNULKp2hIP63/xGNq3S8vb2h4WnAGQrJASK4r/QWR/QsvtGu4Q3IGQrJASK4rfWXDM5e2jj128TOiMxCSBUJyHa+1E0ZI+URIwggpnwhJGCHlUzSkwzoJzkBIFgjJddGQWgr6KzWwRXAGQrJASK7r/qXd+784fsb7gjMQkgVCcl1P3yO9NeI6wRkIyQIhua7Hkw3n/YnVfds3r1tz57rNCb/mmJAsEJLregxpgc2rv3evGNnxG4dGrdgdN46QLBCS63oK6dVhFp+Rdk5V9ZNmL1w0e2K9mrYrZiAhWSAk10VDWl5wzfkD1ZeS73i1mvdyx9ZLc9Q1MQMJyQIhua6nXxA54Mq25DuOa9WD2iaPjxlISBYIyXUlf2is4P7Hd9jcsXFp5/aSppiBhGSBkFyX+iVCQ2d2bs8YFjOQkCwQkutShzSnXv+rpVV1c2MGEpIFQnJd6pC2tqhJy1avXbt62UQ1aGvMQEKyQEiui4Y0plTCPTdNCU9OTNkUN46QLBCS66IhDRnkV9Hs/zdoSCDxvhtWLpg1a8HKDfGjCMkCIbkuGtKOEyffv8Pbcf+kE63O28X43YDI31pS7/Xy0WrlWpVGnwypLtWuuLbWz2AVRUNaOq7jFQq7xi3tebC19kfXazc7+xnpgo/8Y/lm9MmQ1GUpdsVHLqj1M1hF0ZBGXlncuHKU4Azufml3QRUP1KyHlGZ9M/MaUuMVxY0r4n7AGmr7/uLPru/YvPG0mHGEZIGQXBcN6YixOwuXO8d8MPmO+84Ivgo+a3uwPT/uJDohWSAk10UT+Lo6du2b3ptrj1U3Jd/xdjXshtumqNa3PUKKIiQttyG1LfQ/xwS/bHWRxYtWj++/2b/HF9WU7YQURUhabkPyvIfnTxgzYf4jNnc84KOFi1vUCTsJKYKQtByHVIamWR2XK9XHdhNSJ0LS8hyS/V81H398cWO5On0OIWmEpOU3pHL+qvk5jWFyn1P9CEkjJC23IZX1V82/p24PNxcqQtIIScttSGX9VfN3b7on3Gz76lUxAwnJAiG5jr9qbkZIGiEl4a+amxGSRkhJ+KvmZoSkEVIS/qq5GSFphJSEv2puRkgaISXhr5qbEZJGSEn4q+ZmhKQRUpJoSE9srMQMhGSBkFwXDanu7ErMQEgWCMl10ZAOOq8SMxCSBUJyXTSkc47cV4EZCMkCIbkuGtJvh1wa9xfDUiIkC4TkumhI8z+mDjr1/PkBwRkIyQIhua6nPzSmYv9ZRLkIyQIhuS6azMZOgjMQkgVCcp0O6Z9/XqEZCMkCIblOh6SC74tif2VqSoRkgZBcVxpS7K8DSomQLBCS6wjJjJA0QkpCSGaEpBFSEkIyIySNkJIQkhkhaYSUpDOkhpaWlgbV0kFwBkKyQEiu6wyphOAMhGSBkFynk9lTQnAGQrJASK6rwDdFXRCSBUJyHSGZEZJGSEkIyYyQNEJKQkhmhKQRUhJCMiMkjZCSEJIZIWmElISQzAhJI6QkhGRGSBohJSEkM0LSCCkJIZkRkkZISQjJjJA0QkpCSGaEpBFSEkIyIySNkJIQkhkhaYSUhJDMCEkjpCSEZEZIGiElISQzQtIIKQkhmRGSRkhJCMmMkDRCSkJIZoSkEVISQjIjJI2QkhCSGSFphJSEkMwISSOkJIRkRkgaISUhJDNC0ggpCSGZEZJGSEkIyYyQNEJKQkhmhKQRUhJCMiMkjZCSEJIZIWmElISQzAhJI6QkhGRGSBohJSEkM0LSCCkJIZkRkkZISQjJjJA0QkpCSGaEpBFSEkIyIySNkJIQkhkhaYSUhJDMCEkjpCSEZEZIGiElISQzQtIIKQkhmRGSRkhJCMmMkDRCSkJIZoSkEVISQjIjJI2QkhCSGSFphJSEkMwISSOkJIRkRkgaISUhJDNC0ggpCSGZEZJGSEkIyYyQNEJKQkhmhKQRUhJCMiMkjZCSEJIZIWmElISQzAhJI6QkhGRGSBohJSEkM0LSCCkJIZkRkkZISQjJjJA0QkpCSGaEpBFSEkIyIySNkJIQkhkhaYSUhJDMCEkjpCSEZEZIGiElISQzQtIIKQkhmRGSRkhJCMmMkDRCSkJIZoSkEVISQjIjJI2QkhCSGSFphJSEkMwISSOkJIRkRkgaISUhJDNC0ggpCSGZEZJGSEkIyYyQNEJKQkhmhKQRUhJCMiMkjZCSEJIZIWmElISQzAhJI6QkhGRGSBohJSEkM0LSCCkJIZkRkkZISQjJjJA0QkpCSGaEpBFSEkIyIySNkJIQkhkhaYSUhJDMCEkjpCS9Cal987o1d67b3B4/ipAsEJLr0oe0e8VIVTBqxe64cYRkgZBclzqknVNV/aTZCxfNnlivpu2KGUhIFgjJdalDulrNe7lj66U56pqYgYRkgZBclzqkca1t4Wbb5PExAwnJAiG5LnVIjUs7t5c0xQwkJAuE5LrUIQ2d2bk9Y1jMQEKyQEiuSx3SnPo7ws1VdXNjBhKSBUJyXeqQtraoSctWr127etlENWhrzEBCskBIrkv/c6RNU1TRlE1x4wjJAiG5rjevbNiwcsGsWQtWbogfRUgWCMl1lXmt3e6v36BdnIWQvndDCq0ZP1CzHtL01jR7/Xu1PlbSqUxIL3+kVTtSvVeROcqy/+ijyteY8QM16yGNHpRip4/ev9bHSjo5efV3c588UPvi+m5trvWxko5ISFeMibmRkCwQUijXIc2PexRCskBIIUIyISQLhBTKXUifiRhLSFrWD9Ssry93IakSMQMJyQIhhXIXUvMR92qnEJKW9QM16+vLXUjHD+z8XQ18j9Qp6wdq1teXu5AuUZ2vVCWkTlk/ULO+vtyFdE/ro53bmf+n5oQUyvr6cheSNUKyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE0KyQEghQjIhJAuEFCIkE+GQ/rDsqhQaOFCLsr6+v29I8/wue13yGEvDuZDu6/fxFDhQQ5lfX6rnt/5+yWMsDfdCyvqBwPpCVVzffoRULkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGolu+wAAAfjSURBVGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31aYRUNkLSWJ9GSGUjJI31afkO6YFFKZzeJw8E1hdKt77+p6c5lh4QPMxrGdIFh3y8fIf1yQOB9YVSrm9yikPpkAsED/OahsSBEGJ9oSqubyYhla1PHgisTyOksnEgaKxPI6SycSBorE8jpLJxIGisTyOksnEgaKxPI6SycSBorE9zO6T2zevW3Lluc3v8KEKywPpC+Qtp94qRqmDUit1x4wjJAusL5S6knVNV/aTZCxfNnlivpu2KGUhIFlhfKHchXa3mvdyx9dIcdU3MQEKywPpCuQtpXGtbuNk2eXzMQEKywPpCuQupcWnn9pKmLjc+N3SwdoB63/AQFzYMLF9TXYo7DVQfSHGnetYX6pPra7gw7cHfg9QhDZ3ZuT1jWJcb2x5Zrz30XdNDvLI+hQe/leZe3/qPFHe6++4Ud2J9WtbXt/6VtAd/D1KHNKf+jnBzVd1cmcUArkod0tYWNWnZ6rVrVy+bqAZtlVwS4J70P0faNEUVTdkkuCDARb15ZcOGlQtmzVqwcoPYYgBXVf61dkAOEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIE1DKkaQqooWmCB3MtQ5o748lqGXBztWa6eUC1ZnpyBvuvV2ZI/vKrWoZ0geRvuozXfF+1ZrqvuVozsf96SXT/EZIwVw+EeOy/JIQkzNUDIR77LwkhCXP1QIjH/ktCSMJcPRDisf+SEJIwVw+EeOy/JIQkzNUDIR77LwkhCXP1QIjH/ktCSMJcPRDisf+SEJIwVw+EeOy/JLUMadGiqk01+KFqzfTQ4GrNxP7rJdH9V8uQ3nqralNta0seI6NtW7VmYv/1kuj+459RAAIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQEANQ9p3/emjBwyeeN2blZ9qx13nfnDAwBP+sQr/PO2eSz/SrD5T+Xm8rXOHNY2/ZlflJ6raR1TNp0n84KthSHvU8JPOOX2oOuT5ik91k2qcNuuk/urTlX+KWtXAI6px2G0aVDdjyWQ1bXfFZ6rWR1TVp0n84KthSO2Fj+GP89TCik/1b7e947/9r4PV9ys+1SNb2u+txmE3Ra3yvLY5akXFZ6rWR1TVp0n84Kv990iPqunVmuoranE1pqnGYbdBTQwuXqof1V7xuarzEXWq0tMUkDv4ah/SZWpJtaa6TX22GtNU47BbqZYVLieqzRWfq9ohVelpCsgdfLUNacnic8erCX+o0mzt09T6asxTjcNugVpduJyt1lV8riqHVK2nSfbgq21IzUqp039frdmWq7OqMk81DrtZam3hcpG6s+JzVTmkaj1NsgdfDUJquyTwu8J2+6t3jRm+oSpTebeoydurMlM1Q1qo1lR8ruqGVNGnqYTowVeDkPYW/qL0/wuvPqMmVGWqG1VrBX+hYnQmvrTrhco+TV3JHXy1P9ngjVDV2HPL1fHvVGGaQDVPNkzqYycbqvg0FYgdfLUP6d1+6t3Kz/I5NX1H5WfpUJ3T35OCi5frR/ap09/VfJoCcgdfDUN64qng7RtnqpMqPlXbQnVa5V8CEKrSD2Tv8D+ueVX4gaxXtZCq+DSJH3w1DOkratwp55w4QI34TcWn+qqqnzM/cGPFp7pn/vxT1Nj586+o8DybWupnLm1VUyt/4FXrI6rq0yR+8NUwpGevaD2oX8uU66rwHdJVqui0ik91TXGmMZWeaOucoY3jrt5Z6Wmq+BFV82kSP/hq/z0S0AcQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQEiCAkAABhAQIICRAACEBAggJEEBIgABCAgQQUp+xUc2v9RJyjJCct6X4l5IJqZYIyXmElAWE5DxCygJCyj6/kC1/PviAT/y398r8g/c74cnCOx8/a1jDiHm/Cf7QfcGaYNgLc4bs9+H7a7zcfCKk7NuoPnbQtEtOUSO2jJp48Rlq8Nv++75dP/Qvr5rd2Pxz75kb1bQ1a9Y85w87edjki8/uV/+TWi84jwgp+zYqdb1/sUANXtLueV9QN3jesw2n7fbf9fT+E6Jf2qkv+LevUTNqutqcIqTs26jG7PMvfqoGB/FsVed63qXqsdcDM9XzkZBG7/Uv2luG1XKxeUVI2bdRnRlcbFMnBhd7gotWFXoiEtLMwuXRjbVaaJ4RUvYVT8e9qM4ILvaqqZ43Vq1b3+Gdbmftju1Xq4XmGSFlXw8hHat+qW8mpCwgpOzrIaTF6nJ983PqnOgwQqoJQsq+HkLa1L/hx8GVHXd53nY1JTqMkGqCkLKvh5C8f+pfd9rn/2ZG89H+9lR17nUrNhFSTRFS9vUUkrfxvEMbBx990SP+5pZPDa7reGVDYTgh1QIhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECCAkQAAhAQIICRBASIAAQgIEEBIggJAAAYQECPj/NP/Lba5wDc0AAAAASUVORK5CYII=",
      "text/plain": [
       "Plot with title “Histogram of meth AFTER rint”"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(coloc)\n",
    "\n",
    "# read in input data\n",
    "qtls <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset1_locus_cpg12073833_baseline.txt\", header=TRUE)\n",
    "#head(qtls, 2)\n",
    "\n",
    "gscan <- read.table(\"/home/ubuntu/rti-nd/coloc_review/mqtls/input_datasets/baseline/age_of_initiation/coloc_input_dataset2_locus_cpg12073833_gscan_ai.txt\", header=TRUE)\n",
    "#head(gscan, 2)\n",
    "\n",
    "meth <- read.table(\"/home/ubuntu/rti-nd/coloc_review/expression_file/mqtl_methylation_data/LIBD_NAc_MEGA_OMIC_full_matrixeQTL_input.txt\", header=TRUE)\n",
    "\n",
    "# flip GSCAN beta if the effect alleles don't match QTL effect alleles\n",
    "qtls$mQTL_Effect_Allele<-as.character(qtls$mQTL_Effect_Allele)\n",
    "gscan$GSCAN_Effect_Allele<-as.character(gscan$GSCAN_Effect_Allele)\n",
    "gscan$GSCAN_Beta<-as.numeric(gscan$GSCAN_Beta)\n",
    "for (i in 1:length(gscan$GSCAN_Beta)) {\n",
    "    if (qtls$mQTL_Effect_Allele[i] != gscan$GSCAN_Effect_Allele[i]) {\n",
    "        gscan$GSCAN_Beta[i] = -1 * gscan$GSCAN_Beta[i]\n",
    "    }\n",
    "}\n",
    "#head(gscan, 2)\n",
    "\n",
    "\n",
    "meth <- meth[which(meth$cg_id==\"cg12073833\"),]\n",
    "# Apply a rank inverse transform to a data vector\n",
    "# \n",
    "# Args:\n",
    "#   - data: A numeric vector.\n",
    "#\n",
    "# Returns: Rank inverse transformed values of the input vector.\n",
    "rint <- function(data){\n",
    "    if(! is.vector(data)){\n",
    "        stop(\"Error: Input data is not a vector.\")\n",
    "    }\n",
    "    zscore <- scale(data)\n",
    "    rankings <- rank(zscore) - 0.5\n",
    "    scaled.rank <- rankings / (max(rankings, na.rm = T) + 0.5)\n",
    "    rint.values <- qnorm(scaled.rank)\n",
    "    return(rint.values)\n",
    "}\n",
    "\n",
    "meth <- as.numeric(meth[1,2:ncol(meth)])\n",
    "hist(meth, main=\"Histogram of meth B4 rint\")\n",
    "meth <- rint(meth)\n",
    "hist(meth, main=\"Histogram of meth AFTER rint\")\n",
    "head(meth)\n",
    "meth_sd <- sd(meth)\n",
    "meth_sd # 0.999355523800719"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 001: non-filtered, beta/se, no stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n",
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 1.48e-05  1.46e-02  4.71e-05  4.57e-02  9.40e-01 \n",
      "[1] \"PP abf for shared variant: 94%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se\n",
    "my.res1 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, \n",
    "                                   varbeta=(qtls$mQTL_SE)**2, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\"),\n",
    "                     dataset2=list(beta=gscan$GSCAN_Beta, \n",
    "                                   varbeta=(gscan$GSCAN_SE)**2, N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"),\n",
    "                    MAF=gscan$MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 002: non-filtered, p-value, no stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 9.63e-05  3.00e-02  1.69e-04  5.16e-02  9.18e-01 \n",
      "[1] \"PP abf for shared variant: 91.8%\"\n"
     ]
    }
   ],
   "source": [
    "# perform coloc with p-values\n",
    "my.res2 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\"),\n",
    "                     dataset2=list(pvalues=gscan$GSCAN_pvalue,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"), \n",
    "                    MAF=gscan$MAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 003: non-filtered, beta/se, WITH stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(d$varbeta, d$MAF, d$N):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 1.16e-05  1.47e-02  3.68e-05  4.61e-02  9.39e-01 \n",
      "[1] \"PP abf for shared variant: 93.9%\"\n"
     ]
    }
   ],
   "source": [
    "# coloc using beta/se with meth_sd\n",
    "my.res3 <- coloc.abf(dataset1=list(beta=qtls$mQTL_Beta, \n",
    "                                   varbeta=(qtls$mQTL_SE)**2, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\", \n",
    "                                   sdY=meth_sd),\n",
    "                     dataset2=list(beta=gscan$GSCAN_Beta, \n",
    "                                   varbeta=(gscan$GSCAN_SE)**2,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\", \n",
    "                                   MAF=gscan$MAF))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 004: non-filtered, p-value, with stdev-methyl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      " 9.63e-05  3.00e-02  1.69e-04  5.16e-02  9.18e-01 \n",
      "[1] \"PP abf for shared variant: 91.8%\"\n"
     ]
    }
   ],
   "source": [
    "# p-value using standard deviation of methylation data as estimate of sdY for mqtls and using maf for gscan\n",
    "my.res4 <- coloc.abf(dataset1=list(pvalues=qtls$mQTL_pvalue, \n",
    "                                   N=220, \n",
    "                                   type=\"quant\", \n",
    "                                   sdY=meth_sd),\n",
    "                     dataset2=list(pvalues=gscan$GSCAN_pvalue,\n",
    "                                   N=gscan$GSCAN_Effective_N, \n",
    "                                   type=\"quant\"), \n",
    "                    MAF=gscan$MAF)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in sdY.est(vbeta = (qtls$mQTL_SE)^2, maf = gscan$MAF, n = 220):\n",
      "“estimating sdY from maf and varbeta, please directly supply sdY if known”\n"
     ]
    }
   ],
   "source": [
    "sdY.est <- function(vbeta, maf, n) {\n",
    "    warning(\"estimating sdY from maf and varbeta, please directly supply sdY if known\")\n",
    "    oneover <- 1/vbeta\n",
    "    nvx <- 2 * n * maf * (1-maf)\n",
    "    m <- lm(nvx ~ oneover - 1)\n",
    "    cf <- coef(m)[['oneover']]\n",
    "    if(cf < 0)\n",
    "        stop(\"estimated sdY is negative - this can happen with small datasets, or those with errors.  A reasonable estimate of sdY is required to continue.\")\n",
    "    return(sqrt(cf))\n",
    "}\n",
    "\n",
    "meth_sd2 <- sdY.est(vbeta=(qtls$mQTL_SE)**2, maf=gscan$MAF, n=220) # 0.961596439610978"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.1"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7a2c4b191d1ae843dde5cb5f4d1f62fa892f6b79b0f9392a84691e890e33c5a4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
