{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NGC VIDUS FOU GWAS\n",
    "__Author:__ Jesse Marks\n",
    "\n",
    "This document logs the steps taken to perform a GWAS on merged VIDUS data using the Frequency of Use (FOA) phenotype."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Software and tools\n",
    "The software and tools used for processing these data are\n",
    "\n",
    "* [Amazon Elastic Compute Cloud(EC2)](https://aws.amazon.com/ec2/)\n",
    "* GNU bash version 4.1.2\n",
    "* [PLINK v1.9 beta 3.45](https://www.cog-genomics.org/plink/)\n",
    "* [EIGENSOFT v4.2](https://www.hsph.harvard.edu/alkes-price/software/)\n",
    "* [R v3.2.3](https://www.r-project.org/)\n",
    "* R packages: MASS, moments\n",
    "* [RVtests](https://render.githubusercontent.com/view/ipynb?commit=3bb8e661ad8b75af027ed2748133452ec251aaed&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f525449496e7465726e6174696f6e616c2f6271756163685f6e6f7465626f6f6b732f336262386536363161643862373561663032376564323734383133333435326563323531616165642f6865726f696e5f70726f6a6563742f646576656c6f702f32303138303131305f756873325f756873335f666f755f677761732e6970796e623f746f6b656e3d41664d79344e373237626e764465456f46535a697770346b48776246577964706b7335617570495a7741253344253344&nwo=RTIInternational%2Fbquach_notebooks&path=heroin_project%2Fdevelop%2F20180110_uhs2_uhs3_fou_gwas.ipynb&repository_id=105297875&repository_type=Repository)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variable information\n",
    "__FOU Variables__:\n",
    "\n",
    "`Useropioid6mfq` <br>\n",
    "Combined heroin and prescription use varirable to generate opioid use in past 6 months.\n",
    "\n",
    "`Female (sex)` <br>\n",
    "0 is male, 1 is female \n",
    "\n",
    "`Ageatint` <br>\n",
    "Age at time of interview.\n",
    "\n",
    "__Note__: remove any individual with any of these data missing."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieve FOU data\n",
    "The data have already been filtered to \n",
    "* Remove subjects with Useropioid6mfq == 0\n",
    "* Remove duplicate subjects by keeping only the more recent data record\n",
    "* Remove subjects with previously reported sex discrepancy\n",
    "\n",
    "These data are located on EC2 at `/shared/sandbox/ngc_vidus-lung_cancer_fou/phenotype/GWAS-Cohort-n938_passed_g_qc_only_opioid_FOU.csv`\n",
    "\n",
    "There are 300 subjects in this FOU file. I created a new FAM file in the genotype directory which renamed the subject IDs so that we would be able to map the phenotype data to this.\n",
    "\n",
    "I need to filter the original `.fam` file based off of the IDs in the FOU phenotype data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### EC2 console ###\n",
    "cd /shared/s3/ngc_vidus_oaall/vidus/phenotype\n",
    "\n",
    "# R console on EC2 #\n",
    "pheno <- read.table(\"GWAS-Cohort-n938_passed_g_qc_only_opioid_CASES.csv\", header=T, sep=',')\n",
    "\n",
    "pheno_size <- length(pheno[,1])\n",
    "id_list <- c()\n",
    "\n",
    "for (i in 1:pheno_size){\n",
    "    tmp_num <- sprintf(\"-%04d\", pheno[i,1]) \n",
    "    id_list <- append(id_list, tmp_num)\n",
    "}\n",
    "\n",
    "write(id_list, \"id_list.txt\", sep=\"\\n\")\n",
    "quit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# command line #\n",
    "cd /shared/sandbox/ngc_vidus-fou/phenotype\n",
    "\n",
    "# filter the fam file based on the subjects that are in the FOU phenotype data\n",
    "# this is needed to create the PLINK filtered datasets to run eigenstrat\n",
    "grep -f id_list.txt ../genotype/original/final/ea_chr_all.fam > filtered.fam\n",
    "awk '{print $1,$2 }' filtered.fam > ea_subject_ids.keep\n",
    "\n",
    "# get covariates\n",
    "awk 'BEGIN{FS=\",\"; OFS=\"\\t\"} {print $1,$2,$16,$17}' GWAS-Cohort-n938_passed_g_qc_only_opioid_FOU.csv > ea_FOU.data\n",
    "\"\"\"\n",
    "gwas_code       female  useropioid6mfq  ageatint\n",
    "2       0       1       59\n",
    "3       0       4       55\n",
    "5       0       9       40\n",
    "6       1       5       35\n",
    "11      1       5       56\n",
    "13      0       7       35\n",
    "15      0       2       66\n",
    "17      0       4       34\n",
    "18      0       5       39\n",
    "\"\"\"\n",
    "\n",
    "# map sex code and gwas_code\n",
    "awk 'NR==FNR{ map[NR]=$1;next } FNR==1{print $0} FNR>=2 {$1=map[FNR-1];print $0}' id_list.txt ea_FOU.data > ea_fou_map.data\n",
    "awk '{ if( $2==0) { $2=2} {print $0}} ' ea_fou_map.data > temp && mv temp ea_fou_map.data\n",
    "\n",
    "# make sure there are no missing data for sex, FOU, or age\n",
    "# Note that the phenotype data were already filtered. If they\n",
    "# had not been, we would check for this. I did a visual inspection for \n",
    "# a sanity check though."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phenotype visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLAAAAf4CAMAAAAOHNFhAAAAM1BMVEUAAABNTU1oaGh8fHyM\njIyampqnp6eysrK9vb3Hx8fNAADQ0NDZ2dnh4eHp6enw8PD///8nQVX2AAAACXBIWXMAABJ0\nAAASdAHeZh94AAAgAElEQVR4nO3di3rbtrKAUebandMm8fs/7fFFtkVfIQICZoC1vr1bN7bB\nIUX9FWk12W562+7+v93+79bNw8f3vwbwiQGpuAvV9rDh7WmITbOAzwzqxPbib15lAZ+LEKzt\nRrCAAv0zsd0890mwgAsMuof18LenXrmHBRTQCSANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLS\nECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3B\nAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwg\nDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQ\nLCANwQLSECwgDcEC0qgO1gbQRodg1S4AcE+wgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hD\nsIA0BAtIQ7CANAQLSEOwmEy7/+6feASLvOo6JF0J9QjW5v/+3/L/+zx98v/Crylez//HPvYd\nglW7ANx5/Xqo5NQqP/22s2142RWUYJHAiIJIV0SCRWAxkhFjCu4IFhGVJaL9JeHHX6BawwkW\nsVzyYqZzsB6+SrVGEizCyNMCl4ijCBYRpAxAyqGTEyxGO/60H3FJ+Oq7VKsnwWKkumd7hGA9\nfKtqdSJYDDPZk1y1OhAshpj1yT3rfkUhWPTW7pVImEvCFwuJ1tUIFj21fS4HDdb9YqJ1FYJF\nL8s9h1fb3x4Eix6Wi9XJort9PYLF1V3tWRv4kvBsWdFqSLC4qqs+XVME635pzWpEsLgez9Nn\njkUTgsWVeIa+5OKwnmBxBZ2emmkuCc82ollVBIvW+j0nEwbrxgutKoJFU56MRUTrIMGiHc/C\nSzhaBwgWrfR//uW8JDzfpGZdSLBoYshTL32wblwcXkiwqOc5V8fxKyZY1PEKoQkHsYxgUWFw\nrGa4JHzevGYVECyOGv8MmypYN15mFRAsjvHkuobx/xIITrA4wPPqehzajwgWFwvzlJrtkvDE\nvw7eJ1hcJtKzadJg3QT6d0I0gsUFItVqcg71mwSLYp5CfWnWa4JFmYjPnnkvCR8FPOhjCRYF\nItbqZoVghT30owgWn/KUGUuzngkWn/BsCcCDcNIqWNvJ4QWIKfQzZYVLwkdeZt1rFKzt1QcX\nLkBIwZ8kKwXrJvyj0YVg8S5PkGi8zBIs3rH8cyOmxR8WweItOf5Vvtgl4UmOx+ZK3HTntSzP\niDWDdZPnAWrP2xp4Yel/gaex6qPkFRY7iz4PMlryoXIPizPJngPLXhKeJHu4WhAsnqQ7/1cP\nVsKHrJZgcbLcuT+HxR62awbrg/taRONhSmupp5ib7tykzZVLwpOkj98B3tZA3tNdsJ6s8jKr\n4Sus9/7Q8jUOZF6LnOnzW+KBbHgPa3vnq1c4jHktcZavYoEHU7BWlv06wiXhS8kf0M8J1rry\nn9yC9Vr+R/VDgrWqyU/shU39yLrpvqapT+rlTfzoelvDimY5oV0Svif73cl3NQ3W21866ZHL\na55zWbA+MM/DfE6wVjPnecwbZnyoBWstM57DvGu+h9s9rJXMdv66JPzUeg+5YM1itnNXsIrM\n9bAL1jI8DquaKVmCtYiZTlouNc+jL1hLmOeE3XFJWGyWnROsBUyaK8G6xCQngWBNb5IzlVpT\nnAiCNbkpzlLamOBcEKypTZ4rl4QXyn8+CNbE8p+enxCsi2U/JwRrWtlPTa4j93khWJPKfVpy\nTZnPDcGaUuZT8gIuCQ/Ke34I1oxWOeSCdVjWZAnWfLKei3SV8zQRrNnkPA8ZIOOpIliTWepo\nuySslC9ZgjWVfCdgFcGqlu2MEayJZDv5iCDXWSNY83CgOSRTsgRrFpnOulZcEjaS5+QRrDnk\nOeNaEqxmspxAgjWDLGcbgeU4iQQrvxxnGuFlOI8EK7uVc+WSsK0E55JgJbf00RWs1sInS7BS\nC39+kU3wM0qwMnNoaS72vwQFK6/YZ1YPLgmvIvKJJVhpOa6CdS1xD4VgZeWwcj1hX2QJVk5h\nTygmEfQME6yUHNN7LgmvKeTxEKyEgv7Lrz/BuqqI55lgpRPxNGJO8c41wcrG4aSjaKebYOUS\n7195I7kkvL5gZ5xgZRLs5BlOsHoIddYJViKOJEMESpZgpRHorGE1YU4+wcrCYXzNJWE/QZIl\nWDkEOV2CEayeQhwfwcpArgggwmkoWAk4gsQwPlmCFd74kyQsl4TdjT5IghWdw/c+wepv8L8/\nBSs2L6+IZug5KViRyRURDTwtBSswR+4TLgkHGfdvUsEKy8urTwnWMKMOlmAFJVeENuj8FKyY\nHDSCG/OvVMGKyMurMi4JhxpxwAQrIEeskGCNNeBfrIIVjpdXpNH9XBWsaBwuEun9r1fBCsbR\nuoBLwgD6JkuwQnE5eBHBCqHnWStYkThUpNQvWYIVh5dXpNXr5BWsMByni7kkDKTP4ROsILy8\nOkCwIulyCgtWDA4S+XU4iwUrBMeIGVz/PBasAFwOHuSSMJqrn8qlwdqeXGETi3OAjhKseK58\nFC95hbUVff2BTSzNyyumct0T+oJgbR99wwevvzwhP+LoMJtrntONgrW9+uCSTazLy6sqLglj\nuuJpfdEl4fu3sATrCLmqJFhRXe1YNvopoWAd4MAwrWud3II1ipdXzOxK53ejS0I33S/lqDTg\nkjCyqxzPVj8lrNrEghyUFgQrtGu8yPK2hhFcDrKE9ue5tzUM4IiwiOb/ava2hv4ckFZcEsbX\n+Kj6KWF3jkczgpVA2xdZ1wzW4f9cemYOB6tpec57W0NfDgYLanfae1tDV45FUy4Js2j2Iqvx\n2xqObWIVLgcbE6w8Gh3dhm9r2N75rPPgxIFgYW3+dd3ybQ3vFMvz9J6XVyyuxTOg5U8JBesD\njsIVuCTMpcERFqw+HIRrEKxk6i8zLr4k/OCm+zvLOQ9cDsKD2mfCxTfdr7GJ2TkCcFL5L++m\nwfrkP4xelJdXV+OSMKOq4yxY17b47l+VYKVU86/w0goV/UGqgvXa2nsPbzn+rGj0U8KrLpCY\ny0F4w+HnhWBd08K73oVLwrSO/qu8+JLww/9csHITs1p3zzsRrMSOHW6vsK7G5SB84NATRLCu\nZdHdhmIHniNNf0p4dBMzWnOvO3NJmNzlOfEK6ypcDnYhWOldetAF6xoW3GU45MLnyoX/8fN1\nrjpns94ew1GXXY1c9p/mbEeejMs9fZfb4WFcEk7hkgMvWK25fdWPYM3hgueMYDW21t5CE8VP\nmwvvYR15+bDUU3ipnYXe/JSwJZeDfbkkXI5gNbTOngYhWMu5+Dfwcw/rPV5ewbVddA/r2JNy\nkefxIrsJI110SXjsNcQaz+Q19jIYl4TLuewelmC9w+XgEIK1HL9bQwsL7CJE4KeE9by8gk4E\nq9rs+xeYS8LlFF8S3vg93d82+e6FJljL8QqrjstB6OjiYHmFdW7mfYN4BKvGxLuWgkvC5QjW\ncS4HRxOs5QjWYbPuF8QlWEdNulsQmXe6HzTnXiXjknA53tZwzJQ7lY5gLUewjnC7HYbwuzUc\nMN8eQQ6Cdbnpdigtl4TLEayLzbY/iQnWctzDutRkuwOZXBCso29smOsZPtfeQC6XvMLazv7f\ndBOJTLUz6bkkXM4lr7Ae/7pusLydIRbBWo5gXWCePYGcLrokfLiBtew9rGl2BLLyU8Jis+zH\nRFwSLkewCrl9FZBgLefAJWH7TSQwxU5AdkduujffRHwz7APkJ1glJtiFKbkkXI5gfc7tq6gE\naznuYX0q+/wwDz8l/Ezy8WEmgvWJ3NNPziXhckqDtegfQuH2VWiCtRyvsD6SeHSYkWB9IO/k\nMCfBel/awZfhknA5gvUet6/iE6zlCNY7ck4NcxOst6UcGmYnWG/KOPOCXBIuR7DeknDkJQnW\ncgTrNbfbISjBeiXbvLAOwXop2bhLc0m4nFbB+uC/NMx1HuSadnGCtZxGwdpefXDhAkG4fQWR\nCda5RKPCigTrTJ5JueeScDmC9SzNoJwI1nLcdH/k9hWE520NJzmmhLV5hfUgxZC84JJwOe5h\n3cswI68I1nIE606CEQHBuhd/QuCOYPnxYF4uCZfjpnvw8fiAYC1n+bc1xJ4OOLf6K6zQwwF7\n17yHdfgPt+8n8mx8yiXhcta+6R54NAoI1nKWDlbcyYC3LBys0NeqwBvWvekedCwu4JJwOcu+\nrSHmVFxEsJazarBCDgV8bNFgRZwJ+Eyrm+7b+zex4sXB7fZJuCRcTrOb7rULdBRuIA4SrOU0\nuyR8/8uinQfR5gFKrXcPK9g4QLnlghVrGqq4JFxO02C9/aWhzoNQw1BJsJazVrD8eBBSWypY\ncSYBjljpHlaYQWjEJeFyFgpWlDloRrCWs06wgowBHLdMsGJMAdRYJVghhqAxl4TLWSRYEWag\nOcFazhrBCjACUG+FYHm7KExigWCN3j5X45JwOfMHy2k4L8FazvTBchbCPGYPll7BRCYPll5N\nzSXhcuYOllNwboK1nJmD5e0MMJmJgyVXMJt5g6VX83NJuJxpg+XsW4BgLWfWYDn5YEJzBsvt\ndpjSlMGSq0W4JFzOjMFy4q1CsJYzYbCcdzCr+YKlVzCt6YKlVwtxSbicyYLlx4NLEazlzBUs\npxxMbapg6RXMbaZg6dVqXBIuZ6JgOd+WI1jLmSdYTjeY3jTB0iuY3yzB0qsVuSRcziTBcq4t\nSbCWM0WwvF0U1jBDsOQKFjFBsPRqWS4Jl5M/WE6zdQnWctIHy1kG68geLL2ChSQPll4tzSXh\ncnIHyym2NsFaTupgOcNgLZmDpVewmLzB8vZ2XBIuJ22wnF0I1nqyBsvJBQtKGiy9ghXlDJZe\ncccl4XJSBsuZxT3BWk7GYDmxYFEJg6VXsKp8wdIrHrkkXE62YHm7KM8EaznJguWcgpXlCpZe\nwdJSBUuv2HFJuJxMwXJCsSdYy0kULOcTrC5PsPQKltcqWNvJ4QXajMFSXBIup1GwtlcfXLhA\nkylYjGAtJ0mwnEtAlmDpFXCTJFh6xZtcEi4nw013JxJvE6zlJHhbg/MIeBD/FZZeASfh72Hp\nFe9ySbic6MFyEvE+wVpO8GA5h4BnsYOlV8CZyDfd/XbIfMwl4XICv63B+cMnBGs5cYPl9AFe\nuGawPrhMLPjmw5sFZtXqpvv2fp0OpUev+JxLwuU0u+leu0D997AcwVpOs0vC97/swHng1AHe\nEPKmu14Bb4kYLL2ijEvC5QQMlvOGQoK1nHDB8vZ24D3R3tYgV8C7gr2tQa+4gEvC5cR6W4NT\nhksI1nJC3cNyxgAfiRQsvQI+1DRYb3+pDnElLgmXI1jkJVjLESwgjUj3sAA+JFjk5ZJwOYJF\nXoK1HMEC0hAsIA3BIi+XhMsRLPISrOUIFpCGYAFpCBZ5uSRcTo9gAbRx/WAVW/zfcHZ/ZXY/\n4FJhthSS3V+Z3Q+4VJgthWT3V2b3Ay4VZksh2f2V2f2AS4XZUkh2f2V2P+BSYbYUkt1fmd0P\nuFSYLYVk91dm9wMuFWZLIdn9ldn9gEsBXJdgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCG\nYAFpdAtW0e9/OrOF995jP3qAkdo++r0O5dZzYxEt/JT12C+9809/abdcLws/btvCO796sFZ+\n7E8EK591d371YC2+83cEK591d16wlt75OzmDtfbDtu7eC9bSO3+T7vfDOv2UYNVHbfHdF6zF\nd77t7vsdR3tZd/8Fa+mdb7v3/d6H1WtDUa17AATLzgdd7YPtbKu/e3DhnV/9oV/5sW/9zF/5\nUALJCBaQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmABaQgW\nkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFte3+B9LSDvOJK5uc5rRiDOJHpxnNOFEogPX\nhLThRKIDwaINJxIdCBZtOJG4PnfdacSJxPVtkkUbTiOu7u4kc1FIC04jIA3BAtIQLCANwQLS\nECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3B\nAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwg\nDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcFaz/bo/p++bF9e\nfebu43+/bdu3/x5+9fGzpy/7vX1//YtvbOaSoX5/f/zof1+3Lz8v+77Xm/q+/b5k62QhWOvZ\nBevf27//+/Iztx/+8/DRPzdvBOv79l/zYD199ben7ZZ/3+tN/bd9f+OLSU+w1rN7ev/Yfm4/\nXn/m97b9urn5td29UHnZpr/b1zcq1miof7fvfy/JzTvBuvm6/W0zGKEI1np2T+/bC8Iv2+vP\n/Nj+d/e3/20/X7fp193rn90v/vxyuor77+v29b+zxW6D92Pbfj79ytPn//dl2378ef7Wp1d8\n33ehuf+O73fLfr/74j8/Tt/1tO7D993+75/TBI+j/HMXXKYjWOs5D9a/t8/6n0/XhM+fOb1A\nefPF1I/bK8LdL36/z8bP+9dl2/Z49+ghWPd+PP7K4+f/d//B9+dvfQrWly//ftu+P96AeviG\nb6cv/vvl7oMvf8/WfQzWj9MET6P89/y6kYkI1nrO72Hdxef303P7+dfPe/QyWPcx23/B3VXc\ndrfa7+fVHj7z4+9dnR5/5bYk91egX243+2f3rc9fs4/ez7sK/nz4mvtv/fHwwupx3dO0P5/+\n8bTefWqZjmCt5yxYf+9/RPh0v6csWK++4PY675/7F2n3r4Aef+r42I/zr/17t8n7f/rvn2+7\nb30Rte+7TZ0C+eXh+7+cr/v8FftRmt1cIxSP6nrOnsq/Ton69fIzj/e13rokfBWsP1/vMvXv\nefBuzlu3X+Dur/99OX3h2be+/Jr9pvarvfjo+R+f1hOsOXlU13P2VP52Ssy3l5853XT/dXcR\n9nQ/6+yl04tf/PO/Hw8vfF5s5mV+Hl8hfd3++X16qfXiW79/EKzzV1i7LZxv5rSeYM3Jo7qe\n56fy442eL6drwrff1vBj+3b71z/fHm9Ofdv+7H9xe7wj9eP+btPX5828DNb3042o7XaJXw9f\n8HQz6+Frft3dnfrn/D7Y81/O72HttnD+j6f1/j5GmKkI1nqe72H9Or1B8+fpmvDsVcnP5zeO\n/nm4ftu+/Hn41P1PCc9/8cf+p4T/3Twt9jJY9/6cfph3l8mnb/3y+LPEb/ub7ud/+XP2U8LH\nz355+Fnh4z8+reenhHMSrPU8B+v+tdKtP6eXI7s3PNz9pzkP96///vx624mfp16dMnf+i+fv\nw/ry62kzr4P1++v9C7O7sNx+6/399dO3/vvl8Ub7bSq//T5b4+wvZ+/Devzs/fedbcb7sOYm\nWFzq8DsGet5W8k73OQkWFzv6XxZ3DNZvV4RzEiwu9vvgf1ncMVh+t4ZJCRaQhmABaQgWkIZg\nAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpBGWbDO/mxz\ngFGKGrS9+gCgP8EC0hAsIA3BAtJw0x1I43iDRAzorPgV1s07cRIsoJfie1jbO18tWEAvggWk\nIVhAGoIFpOGmO5BGdW8EC+hFsIA0BAtIQ7CANAQLSEOwgDQEC0hDsHr5vwKjZ4TgBKsXwYJq\ngtWLYEE1wepFsKCaYPUiWFBNsHoRLKgmWL0IFlQTrF4EC6oJVi+CBdUEqxfBgmqC1YtgQTXB\n6kWwoJpg9SJYUE2wehEsqCZYvQgWVBOsXgQLqglWL4IF1QSrF8GCaoLVi2BBNcHqRbCgmmD1\nIlhQTbB6ESyoJli9CBZUE6xeBAuqCVYvggXVBKsXwYJqgtWLYEE1wepFsKCaYPUiWFBNsHoR\nLKgmWL0IFlQTrF4EC6oJVi+CBdUEqxfBgmqC1YtgQTXB6kWwoJpg9SJYUE2wehEsqCZYvQgW\nVBOsXgQLqglWL4IF1QSrF8GCaoLVi2BBNcHqRbCgmmD1IlhQTbB6ESyoJli9CBZUE6xeBAuq\nCVYvggXVBKsXwYJqgtWLYEE1wepFsKCaYPUiWFBNsHoRLKgmWL0IFlQTrF4EC6oJVi+CBdUE\nqxfBgmqC1YtgQTXB6kWwoJpg9SJYUE2wehEsqCZYvQgWVBOsXgQLqglWL4IF1QSrF8GCaoLV\ni2BBtbLebCeHF0CwoF5Rb7ZXH1y4AIIFDQhWL4IF1QSrF8GCaoLVi2BBNTfdexEsqHa8Nx9E\njDcIFlQrfIV18xCowwsgWFCv+B7W9s5XC1YhwYJqgtWLYEE1wepFsKBa+dsanqp1YAEECxrw\ntoZeBAuq+d0aehEsqCZYvQgWVBOsXgQLqglWL4IF1QSrF8GCaoLVi2BBNcHqRbCgmmD1IlhQ\nTbB6ESyoJli9CBZUE6xeBAuqCVYvggXVBKsXwYJqgtWLYEE1wepFsKCaYPUiWFBNsHoRLKgm\nWL0IFlQTrF4EC6oJVi+CBdUEqxfBgmqC1YtgQTXB6kWwoJpg9SJYUE2wehEsqCZYvQgWVBOs\nXgQLqglWL4IF1QSrF8GCaoLVi2BBNcHqRbCgmmD1IlhQTbB6ESyoJli9CBZUE6xeBAuqCVYv\nggXVBKsXwYJqgtWLYEE1wepFsKCaYPUiWFBNsHoRLKgmWL0IFlQTrF4EC6oJVi+CBdUEqxfB\ngmqC1YtgQTXB6kWwoJpg9SJYUE2wehEsqCZYvQgWVBOsXgQLqglWL4IF1QSrF8GCaoLVi2BB\nNcHqRbCgmmD1IlhQTbB6ESyoJli9CBZUE6xeBAuqPfVmO1gewSokWFDtrDfboWYJViHBgmr7\n3myXR0uwCgkWVHv9CuvCAglWIcGCaq/vYQnWdQgWVPNTwl4EC6oJVi+CBdW23UcH6iNYhQQL\nqm37Dy7Pj2AVEiyoJli9CBZU6xisxZ+xi+8+tHB2D+vAu0ZvBKvY4rsPLXT8KeHiz9jFdx9a\nEKxeFt99aOHFJaF7WFez+O5DC9urD44u8KnFn7GL7z60UBasbXv39ZdgFVp896GFomB99EWC\nVWjx3YcWzn63hvfvYQlWA4vvPrRQ1BvBamDx3YcWBKuXxXcfWti9reHd+rjpXm/x3YcWzl88\nbRe9ueGDiL1t8Wfs4rsPLVwSrDc/2ThY8z6rBQuqFQVr295/OSVYhQQLqpX9bg0f/Hk6glVI\nsKBaaW+2915+CVYhwYJq5b155/WXYBUSLKhW9E73l1/76S++SbBm3TXoZfvgnw4s8BHBmnXX\noJftw3+8fIEPCNasuwa9CFYvggXVXtzDqljgU4I1665BL9F+T/d5n9WCBdUEqxfBgmov39Zw\n8WWhYBUSLKi27T5yD+t6BAuqRfuj6ud9VgsWVBOsXgQLqpX9bg1FC3xGsGbdNejFTwl7ESyo\nJli9CBZUK/tDKIoW+Ixgzbpr0MvxP4Ti5QKfEqxZdw16EaxeBAuqCVYvggXVvK2hF8GCan5K\n2ItgQbXt1QdHF/iUYM26a9CLYPUiWFDtkj815+MFPiVYs+4a9OIeVi+CBdUEqxfBgmpHf1eZ\nFwuUEKxZdw16eQ7WwWgJViHBgmqC1YtgQTXB6kWwoJpg9SJYUO0UrIN/xteNYBUTLKjmbQ29\nCBZUE6xeBAuqCVYvggXVBKvF1F0XgnUJVoupuy4E6xKsFlN3XQjWJVgtpu66EKxLsFpM3XUh\nWJdgtZi660KwLsFqMXXXhWBdgtVi6q4LwboEq8XUXReCdQlWi6m7LgTrEqwWU3ddCNYlWC2m\n7roQrEuwWkzddSFYl2C1mLrrQrAuwWoxddeFYF2C1WLqrgvBugSrxdRdF4J1CVaLqbsuBOsS\nrBZTd10I1iVYLabuuhCsS7BaTN11IViXYLWYuutCsC7BajF114VgXYLVYuquC8G6BKvF1F0X\ngnUJVoupuy4E6xKsFlN3XQjWJVgtpu66EKxLsFpM3XUhWJdgtZi660KwLsFqMXXXhWBd0war\n1UJFU3ddCNYlWC2m7roQrEuwWkzddSFYl2C1mLrrQrAuwWoxddeFYF2C1WLqrgvBugSrxdRd\nF4J1CVaLqbsuBOsSrBZTd10I1iVYLabuuhCsS7BaTN11IVhXWW+2k8ML3Cl6xjZ7VrdaqGjq\nrgvBuop6s7364MIF7hU9Y5s9q1stVDR114VgXYLVYuquC8G6BKvF1F0XgnUJVoupuy4E63LT\nvcXUXReCdR1/W8MHEXtb0TO22bO61UJFU3ddCNZV/grrna8WLMGCXsrvYW1vf7VgCRb0csFN\n902wQiwE67rkp4SbYEVYCNZ10dsa3rrDLlhpF4J0Cm+6v//VgpV2IUjH79bQYuqcC0E6gtVi\n6pwLQTqC1WLqnAtBOoLVYuqcC0E6gtVi6pwLQTqC1WLqnAtBOoLVYuqcC0E6gtVi6pwLxTPv\nntGIYLWYOudC8cy7ZzQiWC2mzrlQPPPuGY0IVoupcy4Uz7x7RiOC1WLqaRcqWqedeBMRjGC1\nmHrahYrWaSfeRAQjWC2mnnahonXaiTcRwQhWi6mnXahonXbiTUQwgtVi6mkXKlqnnXgTEYxg\ntZh62oWK1mkn3kQEI1gtpp52oaJ12ok3EcEIVoupp12oaJ124k1EMILVYuppFypap514ExGM\nYLWYetqFitZpJ95EBCNYLaaedqGiddqJNxHBCFaLqaddqGidduJNRDCC1WLqaRcqWqedeBMR\njGC1mHrahYrWaSfeRAQjWC2mnnahonXaiTcRwQhWi6mnXahonXbiTUQwgtVi6mkXKlqnnXgT\nEYxgtZh62oWK1mkn3kQEI1gtpp52oaJ12mk2Ubxdow3BajH1tAsVrdNOs4ni7RptCFaLqadd\nqJNxTwUAACAASURBVGiddppNFG/XaEOwWkw97UJF67TTbKJ4u0YbgtVi6mkXKlqnnWYTxds1\n2hCsFlNPu1DROu00myjertGGYLWYetqFitZpp9lE8XaNNgSrxdTTLlS0TjvNJoq3a7QhWC2m\nnnahonXaaTZRvF2jDcFqMfW0CxWt006zieLtGm0IVoupp12oaJ12mk0Ub9doQ7BaTD3tQkXr\ntNNsoni7RhuC1WLqaRcqWqedZhPF2zXaEKwWU0+7UNE6My9EMILVYuppFypaZ+aFCEawWkw9\n7UJF68y8EMEIVoupp12oaJ2ZFyIYwWox9bQLFa0z80IEI1gtpp52oaJ1Zl6IYASrxdTTLlS0\nzswLEYxgtZh62oWK1pl5IYIRrBZTT7tQ0TozL0QwgtVi6mkXKlpn5oUIRrBaTD3tQkXrzLwQ\nwQhWi6mnXahonZkXIhjBajH1tAsVrTPzQgQjWC2mnnahonVmXohgBKvF1NMuVLTOzAsRjGC1\nmHrahYrWmXkhghGsFlNPu1DROjMvRDCC1WLqaRcqWsdCdCNYLaaedqGidSxEN4LVYuppFypa\nx0J0I1gtpp52oaJ1LEQ3gtVi6mkXKlrHQnQjWC2mnnahonUs1FHBQFMnVLBaTD3tQkXrWKij\ngoEEq80CRYe62cPRaqGiqaddqGgdC3VUMJBgtVmg6FA3ezhaLVQ09bQLFa1joY4KBhKsNgsU\nHepmD0erhYqmnnahonUs1FHBQILVZoGiQ93s4Wi1UNHU0y5UtI6FOioYSLDaLFB0qJs9HK0W\nKpp62oWK1rFQRwUDCVabBYoOdbOHo9VCRVNPu1DROhbqqGAgwWqzQNGhbvZwtFqoaOppFypa\nx0IdFQwUb6KGWxOsFlNPu1DROhbqqGCgeBM13JpgtZh62oWK1rFQRwUDxZuo4dYEq8XU0y5U\ntI6FOioYKN5EDbcmWC2mnnahonUs1FHBQPEmarg1wWox9bQLFa1joSbrdF6omb4DlfVmOzm8\nwJ2iQ91s71stVDT1tAsVrWOhJusIVomi3myvPrhwgXtFh7rZ3rdaqGjqaRcqWsdCTdYRrBLH\ng/XBqy6Aa6h+hQXQi2ABaVTfdAfoRYOANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQE\nC0gjYbA2oIvRz/XXAo70mXgjm+hT4QYyUYF4E0Uc6TPxRjbRp8INZKIC8SaKONJn4o1sok+F\nG8hEBeJNFHGkz8Qb2USfCjeQiQrEmyjiSJ+JN7KJPhVuIBMViDdRxJE+E29kE30q3EAmKhBv\noogjfSbeyCb6VLiBTFQg3kQRR/pMvJFN9KlwA5moQLyJIo4E8DbBAtIQLCANwQLSECwgDcEC\n0hAsIA3BAtIQLCANwQLSyBaskL9va7CJwh2jeAPd/yXSVEEnuol3co8e4DLb018CCXSW3Ql3\njOINdDdMqKmCTnQT7uQOc3wuEW3mYI9prGlugj0Rb+5GiZaH7XmQaBMFO7mjHJ+LBJt5CzZQ\nrGluYqXhQbRg3YQL1uMg0U7uaOMUCTZztMd0u4l0K+ROqHszdwTrc4LVSrCRt2gTPd8PCSJW\nGu7EDVaYgZ6OT5yJ7gUbp0C0icM9prGeiTcBB4obrDDzCFYr0QaO9yfkxnom3gQcKGywwoxz\nczpG4U7uUIeoRMh5Yw0V65l4E3CgqMEKM82deBep94KN85mAyb8JdxDDHaJ4A93/JdJU8V7P\nCBZAHcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3B\nAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLK7k+U+tevr7zfNf4QhnD1ey7T54+rNLnXJUcPZw\nJS/+JM5NsKjn7OFKBIv2nD1cyeMtLMGiHWcPV+IVFu05e7gSwaI9Zw9XIli05+zhSl6/reHs\nvQ1wiNOHK3k+tR7fOHr/0aBpmIPzB0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CA\nNAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hD\nsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQL\nSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwKLA9uv+nL9uXV5+5+/jfb9v27b+HX3387PlXff3f\nx1v59/u2ffn5t/XwTESwKLAL1r+3f//35WduP/zn4aN/bt4L1rZ9WKyfD1/z5c919oEZCBYF\ntvPz5Mf2c/vx+jO/t+3Xzc2vbfv9VrDu/vrr7JXZa7/ve/b72/at3dzMRrAosAvWbXa+bK8/\n8+Ph9dP/tp/vBev0tz8/tu3H7cuoP/f9+rL92X3/n++/3/8aVidYFDgP1r+3Rfr5dE34/Jmv\n2/3dp7/b1/eC9b/7F09/v9xf+P29u4b85/b//9t//733vobVCRYFzu9h/dj+u718+7H/zM0+\nTu/dw7q9Zry5v578cfc67LZRv54vAM+j+N7XsDrBosBZsP7eX6Q9vRy6LFh3N9S/3H3rwyq/\nb3/pz/M2nrf33tewOsGiwFlMfp29WNp95vG+1vuXhH//uXut9PDxw1+/PLx4uk/el7NLwje+\nBm4EiyJnwfp2Cta3l5853TT/dXc193Q/68v+q+7+fvbq6dfpDRL3wTp9/++7m+5vfA3cCBZF\nnrN0f0/95ubp9dDbb2v4sX27/eufb4+3uk5f9b+7bz67P/Vl++f5nQ4P3//76/b9/a9hdYJF\nged7WL/u3xh6l5Rfp888fdHP5zeO/vmyfxPo0/f/evzc3U8Af94W6cdpvZvdG0ff/RoWJ1gU\neA7Wt9Md8D+na8LdGx7u/tOch8u3vz+/3v1nNn/23//tPnKP77H6/fI9Vnfff/pPc979GtYm\nWEAaggWkIVhAGoIFpCFYQBqCBaQhWEAaggWkIVhAGoIFpCFYQBqCBaQhWEAaggWkIVhAGoIF\npCFYQBqCBaQhWEAaggWkURassz+oHGCUogZtrz4A6E+wgDQEC0hDsIA03HQH0jjeIBEDOit+\nhXXzTpwEC+il+B7W9s5XCxbQi2ABaQgWkIZgAWm46Q6kUd0bwQJ6ESwgDcEC0hAsIA3BAtIQ\nLCANwQLSECyO+L9PjJ6PSQkWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgW\nRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQwgWRwgWQ5T1Zjs5vACTESyGKOrN9uqDCxdg\nNoLFEILFEYLFEILFEYLFEILFEYLFEG66c4RgMcTx3nwQMaYnWAxR+Arr5iFQhxdgMoLFEMX3\nsLZ3vlqwliRYDCFYHCFYDCFYHCFYDFH+toanah1YgNkIFkN4WwNHCBZD+N0aOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKwGEKwOEKw\nGEKwOEKwGEKwOEKwGEKwOEKwGOKpN9vB8gjWkgSLIc56sx1qlmAtSbAYYt+b7fJoCdaSBIsh\nXr/CurBAgrUkwWKI1/ewBIvPCRZD+CkhRwgWQwgWRwgWQ2y7jw7UR7CWJFgMse0/uDw/grUk\nwWIIweIIwWKIjsFyjk/Eg8kQZ/ewDrxr9EawFuXBZIiOPyV0jk/Eg8kQgsURHkyGeHFJ6B4W\nRTyYDLG9+uDoAp9yjk/Eg8kQZcHatndffwnWkjyYDFEUrI++SLCW5MFkiLPfreH9e1iCxQse\nTIYo6o1g8YIHkyEEiyM8mAyxe1vDu/Vx0509DyZDnL942i56c8MHEXubc3wiHkyGuCRYb35y\nsmClGDIAx4khioK1be+/nBKsJTlODFH2uzV88OfpCNaSHCeGKO3N9t7LL8FakuPEEOW9eef1\nl2AtyXFiiKJ3ur/82k9/8U0pzvEUQwbgODHE9sE/HVjgIynO8RRDBuA4McT24T9evsAHUpzj\nKYYMwHFiCMHaSzFkAI4TQ7y4h1WxwKdSnOMphgzAcWIIv6f7XoohA3CcGEKw9lIMGYDjxBAv\n39Zw8WWhYC3JcWKIbfeRe1gphgzAcWIIf1T9XoohA3CcGEKw9lIMGYDjxBBlv1tD0QKfSXGO\npxgyAMeJIfyUcC/FkAE4TgwhWHsphgzAcWKIsj+EomiBz6Q4x1MMGYDjxBDH/xCKlwt8KsU5\nnmLIABwnhhCsvRRDBuA4MYRg7aUYMgDHiSG8rWEvxZABOE4M4aeEeymGDMBxYojt1QdHF/hU\ninM8xZABOE4MIVh7KYYMwHFiiEv+1JyPF/hUinM8xZABOE4M4R7WXoohA3CcGEKw9lIMGYDj\nxBBHf1eZFwuUSHGOpxgyAMeJIZ6DdTBagrUkx4khBGsvxZABOE4MIVh7KYYMwHFiCMHaSzFk\nAI4TQ5yCdfDP+LoRrEU5TgzhbQ17KYYMwHFiCMHaSzFkAI4TQwjWXoohA3CcGEKw9lIMWa1+\nL9c4ToQjWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsp\nhqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsp\nhqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsp\nhqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsp\nhqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsphqwmWCQlWHsp\nhqwmWCQlWHsBhuwwQv0mAhwnViRYewGG7DBC/SYCHCdWJFh7AYbsMEL9JgIcJ1YkWHsBhuww\nQv0mAhwnViRYewGG7DBC/SYCHCdWJFh7AYbsMEL9JgIcJ1YkWHsBhuwwQv0mAhwnViRYewGG\n7DBC/SYCHCdWJFh7AYbsMEL9JgIcJ1YkWHsBhuwwQv0mAhwnViRYewGG7DBC/SYCHCdWVNab\n7eTwAndSnOMBhuwwQv0mAhwnVlTUm+3VBxcucC/FOR5gyA4j1G8iwHFiRYK1F2DIDiPUbyLA\ncWJFgrUXYMgOI9RvIsBxYkWCtRdgyA4j1G8iwHFiRW667wUYssMI9ZsIcJxY0fG3NXwQsbel\nOMcDDNlhhPpNBDhOrKj8FdY7Xy1YCUeo30SA48SKyu9hbW9/tWAlHKF+EwGOEyu64Kb7JliC\n1XFIeO2SnxJugiVY/YaE1y56W8Nbd9gFK+EI9ZtIMSTzKbzp/v5XC1bCEeo3kWJI5uN3a9gL\nMGSKFqQYkvkI1l6AIVO0IMWQzEew9gIMmaIFKYZkPoK1F2DIFC1IMSTzEay9AEOmaEGKIZmP\nYO0FGDJFC1IMyXwEay/AkClakGJI5iNYewGGTNGCFENeX4YZ5yJYewGG7DBC/SZSDHl9GWac\ni2DtBRiywwj1m0gx5PVlmHEugrUXYMgOI9RvIsWQ15dhxrkI1l6AITuMUL+JGYYMMSOXEay9\nAEN2GKF+EzMMGWJGLiNYewGG7DBC/SZmGDLEjFxGsPYCDNlhhPpNzDBkiBm5jGDtBRiywwj1\nm5hhyBAzchnB2gswZIcR6jcxw5AhZuQygrUXYMgOI9RvYoYhQ8zIZQRrL8CQHUao38QMQ4aY\nkcsI1l6AITuMUL+JGYYMMSOXEay9AEN2GKF+EzMMGWJGLiNYewGG7DBC/SZmGDLEjFxGsPYC\nDNlhhPpNzDBkiBm5jGDtBRiywwj1m5hhyBAzchnB2gswZIcR6jcxw5AhZuQygrUXYMgOI9Rv\nYoYhQ8zIZQRrL8CQHUao38QMQ4aYkcsI1l6AITuMUL+JGYYMMSOXEay9AEN2GKF+EzMMGWJG\nLiNYewGG7DBC/SZmGDLDjDGeFYEI1l6AITuMUL+JGYbMMGOMZ0UggrUXYMgOI9RvYoYhM8wY\n41kRiGDtBRiywwj1m5hhyAwzxnhWBCJYewGG7DBC/SZmGDLDjDGeFYEI1l6AITuMUL+JGYbM\nMGOMZ0UggrUXYMgOI9RvYoYhM8wY41kRiGDtBRiywwj1m5hhyAwzxnhWBCJYewGG7DBC/SZm\nGDLDjDGeFYEI1l6AITuMUL+JGYbMMGOMZ0UggrUXYMgOI9RvYoYhM8wY41kRiGDtBRiywwj1\nm5hhyAwzxnhWBCJYewGG7DBC/SZmGDLDjDGeFYEI1l6AITuMUL+JGYYMcRjqV1iLYO0FGLLD\nCPWbmGHIEIehfoW1CNZegCE7jFC/iRmGDHEY6ldYi2DtBRiywwj1m5hhyBCHoX6FtQjWXoAh\nO4xQv4kZhgxxGOpXWItg7QUYssMI9ZuYYcgQh6F+hbUI1l6AITuMUL+JGYYMcRjqV1iLYO0F\nGLLDCPWbmGHIEIehfoW1CNZegCE7jFC/iRmGDHEY6ldYi2DtBRiywwj1m5hhyBCHoX6FtQjW\nXoAhO4xQv4kZhgxxGOpXWItg7QUYssMI9ZuYYcgQh6F+hbUI1l6AITuMUL+JGYYMcRjqV1iL\nYO0FGLLDCPWbmGHIEIehfoW1CNZegCE7jFC/iRmGDHEY6ldYi2DtBRiywwj1m5hhyBCHoX6F\ntQjWXoAhO4xQv4kZhgxxGOpXWItg7QUYssMI9ZuYYcgQh+HTFeo3MRXB2gswZIcR6jcxw5Ah\nDsOnK9RvYiqCtRdgyA4j1G9ihiFDHIZPV6jfxFQEay/AkB1GqN/EDEOGOAyfrlC/iakI1l6A\nITuMUL+JGYYMcRg+XaF+E1MRrL0AQ3YYoX4TMwwZ4jB8ukL9JjrsRT+CtRdgyBRn6AxDhjgM\nn65Qv4kOe9GPYO0FGDLFGTrDkCEOw6cr1G+iw170I1h7AYZMcYbOMGSIw/DpCvWb6LAX/QjW\nXoAhU5yhMwwZ4jB8ukL9JjrsRT+CtRdgyBRn6AxDhjgMn65Qv4kOe9GPYO0FGDLFGTrDkCEO\nw6cr1G+iw170I1h7AYZMcYbOMGSIw/DpCvWb6LAX/QjWXoAhU5yhMwwZ4jB8ukL9JsbvRYMZ\nHwnWXoAhO4xQv4kZhgxxGD5doX4T4/eiwYyPBGsvwJAdRqjfxAxDhjgMn65Qv4nxe9FgxkeC\ntRdgyA4j1G9ihiFDHIZPV6jfxPi9aDDjI8HaCzBkhxHqNzHDkCEOw6cr1G9i/F40mPGRYO0F\nGLLDCPWbmGHIEIfh0xXqv2D8Xny+QLGy3mwnhxe4U39YOggwZIcR6jcxw5AhDsOnK9R/wfi9\n+HyBYkW92V59cOEC9+oPSwcBhuwwQv0mZhgyxGH4dIX6Lxi/F58vUOx4sD541QVwDdWvsAB6\nESwgjeqb7gC9aBCQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWn0C9YGzKlbRXoG\nq9uWJudANuJAtiJYvM+BbMSBbEWweJ8D2YgD2Ypg8T4HshEHshXB4n0OZCMOZCuCxfscyEYc\nyFYEi/c5kI04kK0IFu9zIBtxIFuZMlgAlQQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQE\nC0ijR7CettH3N1Odz+5AOpJHnR08x7HKiyPZNSZX3MTTPnXa4KyeTwgHscLZaeiMrHJ++Hod\nxetvZ7sRrCa2G8FqRrBamS5YN4LVyvbi7xwmWK2cHcl+m+uzDadHpd73CyYmWK2cjmSvU1Kw\nEnEgWzm/jnEga/Q+koKVyPbuP3ARZ2Qr3U9JwUpEsNrY//DCgTxu++CfemzxuttwelRyIJvY\n9h85kIcNOJLdtnH3F/eK6ziQLWynG8QOZK0RR9LDBaQhWEAaggWkIVhAGoIFpCFYQBqCBaQh\nWEAaggWkIVhAGoIFpCFYQBqCBaQhWEAaggWkIVhAGoIFpCFYQBqCBaQhWEAaggWkIVhAGoLF\nxbbHP9/pg6/4fJHztR4/ev6EE5O3OC844uPzpuSs2nZ/O/9DF4uXYD3OC45oHqznfxQs3ue8\n4IjT66Ht/kru4cPH67rzv9x/ycPnzj5//w+CxQHOC454zM7jRdzpw8d/2v/6m58/v4d1I1iU\ncV5wxNnt8Rclev0LZ/9/8av7NW5uBItPOC844jlYD6+QnqK0vZ2mpxdcu6++ESwu47zgiN3P\n9F4l6MNXWDeCxWHOC444D9b2qk8HLgnPX7PdOC95hxODI84vCXdh2t746Obl52+e33f6fAKe\n/VrB+05ZkzMDSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOw\ngDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtI\nQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQE\nC0hDsIA0BAtIQ7AosT26/6cv25dXn7n7+N9v2/btv4dfffzs66+Cw5xBlNgF69/bv//78jO3\nH/7z8NE/N+8G60fvuZmMYFFi99rox/bzOT3Pn/m9bb9ubn5t2+/Xwbrz6+x1GRwiWJTYBes2\nPF+215/5sf3v7m//236+Hawv23+PH/68vaj8effBf7cXkd//3H7w58ft668/99/x75evt5/4\nun39b/elIFiUOQ/Pv7dF+vl0Tfj8ma/b37u//d2+vhmsH8+vyr7fXx/+vH9R9nCh+PfL3Qdf\n/j5cPH4/feL32ZfCjWBR5vwe1o/bV0q/n+rz/OvnkXodrP92N+r/3v7zdn9xedem29dRd+v9\neHht9s/N39sPfz9s4+lL4UawKHMWrL/35Tm9nCoP1tkF4e03f/3n392nv9wtd7/ydr/w/Quu\nu398+lK4ESzKnL3G+XVK1K+Xn3m8r/XmJeHjBeF93P58vavRv2effvjgOXVPGXz6UrgRLMqc\nBevbKSbfXn7mdNP9112anu5nna4Dny4IT6/G/vzvx8PLqdP3nr/CerG905fCjWBR5jkg9/fU\nb06J2X3m/G0NP7Zvt3/98+3xVteXxxdkp+/57+bP4z2s/+4WPL+HdfcVT594+lK4ESzKPN/D\n+nX/xtC7xPw6febpi34+v3H0z8NNqO3Ln/vP/Ni/0/3H/qeE/3v8+oefEt59xcMn/jv7UrgR\nLMo8B+vb9tCgP6drwt0bHu7+05yH+01/f369DdDPPy+//cHz+7C+bl/uLyTP3od18/iJX7sv\nBcECEhEsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQ\nLCANwQLSECwgDcEC0hAsII2yYL34IwQARihq0PbqA4D+BAtIQ7CANAQLSMNNdyCN4w0SMaCz\n4ldYN+/ESbCAXorvYW3vfLVgAb0IFpCGYAFpCBaQhpvuQBrVvREsoBfBAtIQLCANwQLSECwg\nDcEC0hAsIA3BotT/vWH0TCxGsCglWAwnWJQSLIYTLEoJFsMJFqUEi+EEi1KCxXCCRSnBYjjB\nopRgMZxgUUqwGE6wKCVYDCdYlBIshhMsSgkWwwkWpQSL4QSLUoLFcIJFKcFiOMGilGAxnGBR\nSrAYTrAoJVgMJ1iUEiyGEyxKCRbDCRalBIvhBItSgsVwgkUpwWI4waKUYDGcYFFKsBhOsCgl\nWAwnWJQSLIYTLEoJFsMJFqUEi+EEi1KCxXCCRSnBYjjBopRgMZxgUUqwGE6wKCVYDCdYlBIs\nhhMsSgkWwwkWpQSL4QSLUoLFcIJFKcFiOMGilGAxnGBRSrAYTrAoJVgMJ1iUEiyGEyxKCRbD\nCRalBIvhBItSgsVwgkUpwWI4waKUYDGcYFFKsBhOsCglWAwnWJQSLIYTLEoJFsMJFqUEi+EE\ni1KCxXCCRSnBYjjBopRgMZxgUUqwGE6wKCVYDCdYlBIshhMsSgkWwwkWpQSL4QSLUoLFcIJF\nKcFiOMGilGAxnGBRSrAYTrAoJVgMJ1iUEiyGK+vNdnJ4ASYgWAxX1Jvt1QcXLsAMBIvhBItS\ngsVwgkUpwWI4waKUYDGcm+6UEiyGO96bDyLGlASL4QpfYd08BOrwAkxAsBiu+B7W9s5XI9cT\nNQAADVhJREFUC9YyBIvhBItSgsVwgkUpwWK48rc1PFXrwALMQLAYztsaKCVYDOd3a6CUYDGc\nYFFKsBhOsCglWAwnWJQSLIYTLEoJFsMJFqUEi+EEi1KCxXCCRSnBYjjBopRgMZxgUUqwGE6w\nKCVYDCdYlBIshhMsSgkWwwkWpQSL4QSLUoLFcIJFKcFiOMGilGAxnGBRSrAYTrAoJVgMJ1iU\nEiyGEyxKCRbDCRalBIvhBItSgsVwgkUpwWI4waKUYDGcYFFKsBhOsCglWAwnWJQSLIYTLEoJ\nFsMJFqUEi+EEi1KCxXCCRSnBYjjBopRgMZxgUUqwGE6wKCVYDCdYlBIshhMsSgkWwwkWpQSL\n4QSLUoLFcIJFKcFiOMGilGAxnGBRSrAYTrAoJVgMJ1iUEiyGEyxKCRbDCRalBIvhBItSgsVw\ngkUpwWI4waKUYDGcYFFKsBhOsCglWAwnWJQSLIYTLEoJFsMJFqUEi+EEi1KCxXCCRSnBYjjB\nopRgMZxgUUqwGE6wKCVYDPfUm+1geQRrGYLFcGe92Q41S7CWIVgMt+/Ndnm0BGsZgsVwr19h\nXVggwVqGYDHc63tYgsXbBIvh/JSQUoLFcIJFKcFiuG330YH6CNYyBIvhtv0Hl+dHsJYhWAwn\nWJQSLIbLEixPlvE8Bgx3dg/rwLtGbwRrIR4DhsvyU0JPlvE8BgwnWJTyGDDci0tC97B4l8eA\n4bZXHxxd4Lo8WcbzGDBcWbC27d3XX4K1DI8BwxUF66MvEqxleAwY7ux3a3j/HpZgceMxIICi\n3ggWNx4DAhAsSnkMGG73toZ36+OmOx4DAjh/8bRdlJ8PInYFnizjeQwY7pJgvfnJNYMVbJw+\nltxpYikK1ra9/3JKsAKM08eSO00sZb9bwwd/no5gBRinjyV3mlhKe7O99/JLsAKM08eSO00s\n5b155/WXYAUYp48ld5pYit7p/vJrP/3F9oI9WYKN08eSO00s2wf/dGCBqwn2ZAk2Th9L7jSx\nbB/+4+ULXEuwJ0uwcfpYcqeJRbBmGKePJXeaWF7cw6pY4LqCPVmCjdPHkjtNLH5P9xnG6WPJ\nnSYWwZphnD6W3Gliefm2hosvCwUrwDh9LLnTxLLtPnIPK+c4fSy508Tij6qfYZw+ltxpYhGs\nGcbpY8mdJpay362haIGrCvZkCTZOH0vuNLH4KeEM4/Sx5E4Ti2DNME4fS+40sZT9IRRFC1xV\nsCdLsHH6WHKnieX4H0LxcoHrCvZkCTZOH0vuNLEI1gzj9LHkThOLYM0wTh9L7jSxeFvDDOP0\nseROE4ufEs4wTh9L7jSxbK8+OLrAdQV7sgQbp48ld5pYBGuGcfpYcqeJ5ZI/NefjBa4r2JMl\n2Dh9LLnTxOIe1gzj9LHkThOLYM0wTh9L7jSxHP1dZV4scHXBnizBxuljyZ0mludgHSyPYAUY\np48ld5pYBGuGcfpYcqeJRbBmGKePJXeaWARrhnH6WHKnieUUrIN/xteNYIUYp48ld5pYvK1h\nhnH6WHKniUWwZhinjyV3mlgEa4Zx+lhyp4lFsGYYp1jV2Fl3mokI1gzjFBMschOsGcYpJljk\nJlgzjFNMsMhNsGYYp5hgkZtgzTBOMcEiN8GaYZxigkVugjXDOMUEi9wEa4ZxigkWuQnWDOMU\nEyxyE6wZxikmWOQmWDOMU0ywyE2wZhinmGCRm2DNME4xwSI3wZphnGKCRW6CNcM4xQSL3ARr\nhnGKCRa5CdYM4xQTLHITrBnGKSZY5CZYM4xTTLDITbBmGKeYYJGbYM0wTjHBIjfBmmGcYoJF\nboI1wzjFBIvcBGuGcYoJFrkJ1gzjFBMschOsGcYpJljkJlgzjFNMsMhNsGYYp5hgkZtgzTBO\nMcEiN8GaYZxigkVughV4nPZbESxyE6zA4wgW7AlW4HEEC/YEK/A4ggV7ghV4HMGCPcEKPI5g\nwZ5gBR5HsGBPsAKPI1iwJ1iBxxEs2BOswOMIFuwJVuBxBAv2ynqznRxeoFqwJ4tgjX8MWFFR\nb7ZXH1y4QL1gTxbBGv8YsCLBCjyOYMGeYAUeR7BgT7ACjyNYsOeme+BxBAv2jvfmg4hdQbAn\ni2CNfwxYUfkrrHe+WrAEC3opv4e1vf3VgiVY0MsFN903weo8jmDB3iU/JdwEq+84ggV7F72t\n4a077IIlWFf/ZjgpvOn+/lcLlmBd/ZvhxO/WEHgcwYI9wQo8jmDBnmAFHkewYE+wAo8jWLAn\nWIHHESzYE6zA4wgW7AlW4HEEC/YEK/A4gtVXghGXJ1iBxxGsvhKMuDzBCjyOYPWVYMTlCVbg\ncQSrrwQjLk+wAo8jWE3mDrYVaghW4HHab6V1NQSLvgQr8Djtt9K6GoJFX4IVeJz2W2ldDcGi\nL8EKPE77rbSuhmDRl2AFHqf9VlpXQ7DoS7ACj9N+K62rIVj0JViBx2m/ldbVECz6EqzA47Tf\nSutqCBZ9CVbgcdpvpXU1BIu+BCvwOO230roagkVfghV4nPZbaV0NwaIvwQo8TvuttK6GYNGX\nYAUep/1WWldDsOhLsAKP034rrashWPQlWIHHab+V1tUQLPoSrMDjtN9K62oIFn0JVuBx2m+l\ndTUEi74EK/A47bfSuhprBqvPo89bBCvwOO230roagnXFcXiDYAUep/1WWldDsK44Dm8QrMDj\ntN9K62oI1hXH4Q2CFXic9ltpXQ3BuuI4vEGwAo/TfiutqyFYVxyHNwhW4HHab6V1NQTriuPw\nBsEKPE77rbSuhmBdcRzeIFiBx2m/ldbVEKwrjsMbBCvwOO230roagnXFcXiDYAUep/1WWldD\nsK44Dm8QrMDjtN9K62oI1hXH4Q2CFXic9ltpXQ3BuuI4vEGwAo/TfiutqxEtWBm2Qg3BCjxO\n+620fj4LVpvHhVKCFXic9ltp/XwWrDaPC6UEK/A47bfS+vksWG0eF0oJVuBx2m+l9fNZsNo8\nLpQSrMDjtN9K6+ezYLV5XCglWIHHab+V1s9nwWrzuFBKsAKP034rrZ/PgtXmcaGUYAUep/1W\nWj+fBavN40IpwQo8TvuttH4+C1abx4VSghV4nPZbaf18Fqw2jwulBCvwOO230vr5LFhtHhdK\nCVbgcdpvpfXzWbDaPC6UEqzA47TfSuvns2C1eVwoJViBx2m/ldbPZ8Fq87hQSrACj9N+K62f\nz4LV5nGhlGAFHqf9Vlo/nwWrzeNCKcEKPE77rbR+PgvWFcfhDYIVeJz2W2n9fBasK47DGwQr\n8Djtt9L6+SxYVxyHNwhW4HHab6X181mwrjgObxCswOO030rr57NgXXEc3iBYgcdpv5XWz2fB\nut44dfsyK8EKPE77rVQtOPDpV/rNwbYy8IjNSrACj9N+K1ULDnz6lX5zsK0MPGKzEqzA47Tf\nStWCA59+pd8cbCsDj9isBCvwOO23UrXgwKdf6TcH28rAIzYrwQo8TvutVC048OlX+s3BtjLw\niM1KsAKP034rVQsOfPqVfnOwrQw8YrMSrMDjtN9K1YIDn36l3xxsKwOP2KwEK/A47bdSteDA\np1/pNwfbSoIjlo1gBR6n/VZaPweiPf2CbSXBEctGsAKP034rrZ8D0Z5+wbaS4IhlI1iBx2m/\nldbPgWhPv2BbSXDEshGswOO030rr50C0p1+wrSQ4YtkIVuBx2m+l9XMg2tMv2FZqvlCw3lTW\nm+3k8ALV6h69pOO030rr54BgXe0LBetNRb3ZXn1w4QL16h69pOO030rr54BgXe0LBetNx4P1\nwasugGuofoUF0ItgAWlU33QH6EWDgDQEC0hDsIA0BAtIQ7CANAQLSEOwgDQEC0hDsIA0BAtI\no1+wNmBNDTPSbqkwW7o++xKTfYlJsAazLzHZl5gEazD7EpN9iUmwBrMvMdmXmARrMPsSk32J\nSbAGsy8x2ZeYBGsw+xKTfYlJsAazLzHZl5hSBgugkmABaQgWkIZgAWkIFpCGYAFpCBaQhmAB\naQgWkIZgAWl0C1bb3yh1tFl2xaMS00yPS87fInnrubGrm+Vs8qjENNPj0nhfuh6VSR6C2/2Y\nZE9memLM86iczLQ3gjXYJHsyVbBm2pM7M+2NYA02yZ4IVmDT7E3Ke1i9t3Vlk+yKYMVlZ668\n0gfb2OZ5Ysy0L4IV2Ez7ki1YvbfUwSQ7I1hRTbQrdxIGa65HYJK9Eayg5tmTrG9r2LaZ3gs3\nzQk11YMyzaMy17Ml5xtHAaoJFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmABaQgW\nkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmDR2FR/RhXBOLG4AqcV1+HM\noj1nFVfi1KK9p7PqdG24+9t292kXjRzitKG57fyD7dXfnj+ECzlraG578cHZ385i5dTjcs4a\nWju//NvO0nQeLD9I5BBnDa1tL//p7VdYcDlnDo1t+w9f38O6ES2OctbQ1nb+xtG3f0r4/Ctw\nGacNkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCGYAFpCBaQhmABaQgWkIZgAWkIFpCG\nYAFpCBaQhmABaQgWkIZgAWn8P/73/wm/PDAQAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"EA FOU Z-scores\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### R console ###\n",
    "library(MASS)\n",
    "options(repr.plot.width=10, repr.plot.height=17)\n",
    "# note that I copied over the phenotype data to my local machine to produce the plots\n",
    "setwd('C:/Users/jmarks/Desktop/VIDUS/pheno/')\n",
    "\n",
    "ea.fou.data <- read.table(\"ea_fou_map.data\", header = T)\n",
    "#ea.fou.data\n",
    "par(mfrow=c(4,1))\n",
    "# Perform boxcox transform\n",
    "#ea.boxcox <- boxcox(totopioid_tot_30d+1 ~ age + sex_selfreport, data=ea.fou.data)\n",
    "ea.boxcox <- boxcox(useropioid6mfq ~ ageatint + female, data=ea.fou.data)\n",
    "ea.lambda <- ea.boxcox$x[which(ea.boxcox$y==max(ea.boxcox$y))]\n",
    "\n",
    "# Convert to Z-scores\n",
    "ea.zscore <- scale((ea.fou.data$useropioid6mfq)^ea.lambda, center=T, scale=T)\n",
    "\n",
    "hist(ea.fou.data$useropioid6mfq, breaks=30, col=\"red3\", border=\"white\", main=\"EA FOU (Use in past 6 months)\", xlab=\"FOU\")\n",
    "\n",
    "hist((ea.fou.data$useropioid6mfq)^ea.lambda, breaks=30, col=\"red3\", border=\"white\", main=\"EA FOU Box-Cox\", xlab=\"Transformed FOU\")\n",
    "#\n",
    "hist(ea.zscore, breaks=30, col=\"red3\", border=\"white\", main=\"EA FOU Z-scores\", xlab=\"Z-score\", xlim=c(-2,3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=9, repr.plot.height=4)\n",
    "\n",
    "# Plot age\n",
    "hist(ea.fou.data$ageatint, breaks=30, col=\"red3\", border=\"white\", main=\"EA Age Distribution\", xlab=\"Age in Years\")\n",
    "\n",
    "# Plot sex\n",
    "table(ea.fou.data$female)\n",
    "#ea.zscore"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EIGENSTRAT\n",
    "To obtain principal component covariates to use in the GWAS statistical model, EIGENSTRAT is run on LD-pruned observed genotypes for each ancestry group. Usually a GRCh37 plus strand check is implemented, as well as a monomorphic SNP filter and discordant allele flip. Since this was already done for data in preparation for haplotype phasing, the haplotype phasing input PLINK files will be used. Note: In addition to these aforementioned data processing steps, ambiguous SNPs identified by reference panel frequency differences in the discordant allele checks were also removed prior to phasing.\n",
    "\n",
    "### Construct subject-filtered PLINK file sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line #\n",
    "mkdir -p /shared/sandbox/ngc_vidus-fou/eigenstrat\n",
    "cd /shared/sandbox/ngc_vidus-fou/eigenstrat\n",
    "\n",
    "# Remove subjects by phenotype criteria\n",
    "ancestry=\"ea\"\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile ../genotype/original/final/ea_chr_all \\\n",
    "    --keep ../phenotype/ea_subject_ids.keep \\\n",
    "    --make-bed \\\n",
    "    --out ${ancestry}_pheno_filter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove high-LD region variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/eigenstrat\n",
    "\n",
    "# Remove high-LD region variants\n",
    "ancestry=\"ea\"\n",
    "# Generate list of variants in known high-LD regions\n",
    "perl -lane 'if (($F[0]==5 && $F[3] >= 43964243 && $F[3] <= 51464243) || ($F[0]==6 && $F[3] >= 24892021 && $F[3] <= 33392022) || ($F[0]==8 && $F[3] >= 7962590 && $F[3] <= 11962591) || ($F[0]==11 && $F[3] >= 45043424 && $F[3] <= 57243424)) { print $F[1].\"\\n\"; }' ${ancestry}_pheno_filter.bim \\\n",
    "    > ${ancestry}.high_ld_regions.remove\n",
    "\n",
    "# Remove SNPs in known high-LD regions\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --bfile ${ancestry}_pheno_filter \\\n",
    "    --exclude ${ancestry}.high_ld_regions.remove \\\n",
    "    --make-bed \\\n",
    "    --out ${ancestry}_high_ld_regions_removed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linkage disequilibrium pruning\n",
    "Linkage disequilibrium (LD) pruning eliminates a large degree of redundancy in the data and reduces the influence of chromosomal artifacts. The objective of LD pruning is to select a subset of variants based off of LD such that the variants in the subset are indepdendent. This filtering will not carry forward to the final processed results, but this step improves the quality of EIGENSTRAT calculations. Consequently, the LD pruned data will be used as input for those calculations.\n",
    "\n",
    "LD pruning is implemented using PLINK `--indep-pairwise.`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/eigenstrat\n",
    "\n",
    "# Remove high-LD region variants\n",
    "ancestry=\"ea\"\n",
    "# Run per chromosome LD pruning for each chr\n",
    "for chr in {1..23}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${ancestry}_${chr}_ld_prune \\\n",
    "        --script_prefix ${ancestry}_${chr}_ld_prune \\\n",
    "        --mem 3.5 \\\n",
    "        --nslots 1 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 3500 \\\n",
    "            --bfile ${ancestry}_high_ld_regions_removed \\\n",
    "            --indep-pairwise 1500 150 0.2 \\\n",
    "            --chr ${chr} \\\n",
    "            --out ${ancestry}_chr${chr}_ld_pruned\n",
    "done\n",
    "\n",
    "#Create LD pruned PLINK file sets\n",
    "ancestry=\"ea\"\n",
    "# Merge *prune.in files\n",
    "cat ${ancestry}_chr*_ld_pruned.prune.in > ${ancestry}_chr_all_ld_pruned.prune.in\n",
    "\n",
    "# Create new PLINK filesets with only lD pruned variants\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile ${ancestry}_high_ld_regions_removed \\\n",
    "    --extract ${ancestry}_chr_all_ld_pruned.prune.in \\\n",
    "    --make-bed \\\n",
    "    --out ${ancestry}_ld_pruned\n",
    "\n",
    "\n",
    "# Clean up\n",
    "rm *ld_pruned.{prune.in,prune.out,log}\n",
    "rm *qsub*\n",
    "rm *high_ld_regions*\n",
    "rm *nosex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rename BIM/FAM file IDs\n",
    "BIM and FAM file IDs have the potential issue of being too long if encoded using the 1000 Genomes Phase 3 IMPUTE2 format. This will cause `smartpca` to throw an error. To resolve this the IDs are replaced by numeric values if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/eigenstrat\n",
    "\n",
    "# Count ID lengths greater than 39 characters\n",
    "ancestry=\"ea\"\n",
    "echo -e \"\\n\\n======== ${ancestry} ========\\n\\n\"\n",
    "awk '{ if(length($1)+length($2)>39){print $2} }' ${ancestry}_ld_pruned.fam | wc -l\n",
    "awk '{ if(length($2)>39){print $2} }' ${ancestry}_ld_pruned.bim | wc -l\n",
    "\n",
    "# Rename BIM file IDs\n",
    "# Make new BIM\n",
    "awk '{$2=\"ID_\"NR; print $0}' ${ancestry}_ld_pruned.bim \\\n",
    "    > ${ancestry}_ld_pruned_id_renamed.bim\n",
    "\n",
    "# Rename FAM file IDs\n",
    "# Make new FAM\n",
    "awk '{$1=\"ID_\"NR; print $0}' ${ancestry}_ld_pruned.fam \\\n",
    "    > ${ancestry}_ld_pruned_id_renamed.fam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run EIGENSTRAT\n",
    "__NOTE__: For `smartpca` to run, the BIM and FAM file IDs need to be less than 39 characters in length. Often this is not a problem, but with a BIM file that has IDs renamed to the 1000 Genomes Phase 3 format, the IDs for indels can exceed the character limit. Also, the EIGENSOFT bin directory should be in ~/.bashrc as an addition to the PATH variable. If it is not, it can be added as following\n",
    "\n",
    "`export PATH=$PATH:/shared/bioinformatics/software/third_party/EIG-6.1.4/bin/` # Add EIGENSOFT bin to path\n",
    "\n",
    "The `smartpca.pl` program documentation is shown below:\n",
    "\n",
    "\n",
    "DOCUMENTATION of `smartpca.pl` program:\n",
    "\n",
    "This program calls the `smartpca` program (see ../POPGEN/README). \n",
    "For this to work, the bin directory containing `smartpca` MUST be in your path. \n",
    "See `./example.perl` for a toy example.\n",
    "```\n",
    "../bin/smartpca.pl \n",
    "-i example.geno  : genotype file in any format (see ../CONVERTF/README)\n",
    "-a example.snp   : snp file in any format (see ../CONVERTF/README)\n",
    "-b example.ind   : indiv file in any format (see ../CONVERTF/README)\n",
    "-k k             : (Default is 10) number of principal components to output\n",
    "-o example.pca   : output file of principal components.  Individuals removed\n",
    "                   as outliers will have all values set to 0.0 in this file.\n",
    "-p example.plot  : prefix of output plot files of top 2 principal components.\n",
    "                   (labeling individuals according to labels in indiv file)\n",
    "-e example.eval  : output file of all eigenvalues\n",
    "-l example.log   : output logfile\n",
    "-m maxiter       : (Default is 5) maximum number of outlier removal iterations.\n",
    "                   To turn off outlier removal, set -m 0.\n",
    "-t topk          : (Default is 10) number of principal components along which \n",
    "                   to remove outliers during each outlier removal iteration.\n",
    "-s sigma         : (Default is 6.0) number of standard deviations which an\n",
    "                   individual must exceed, along one of topk top principal\n",
    "           components, in order to be removed as an outlier.\n",
    "\n",
    "OPTIONAL FLAGS:\n",
    "-w poplist       : compute eigenvectors using populations in poplist only,\n",
    "                   where poplist is an ASCII file with one population per line\n",
    "-y plotlist      : output plot will include populations in plotlist only, \n",
    "                   where plotlist is an ASCII file with one population per line\n",
    "-z badsnpname    : list of SNPs which should be excluded from the analysis\n",
    "-q YES/NO        : If set to YES, assume that there is a single population and\n",
    "                   the population field contains real-valued phenotypes.\n",
    "           (Corresponds to qtmode parameter in smartpca program.)\n",
    "           The default value for this parameter is NO.\n",
    "\n",
    "Estimated running time of the smartpca program is \n",
    "  2.5e-12 * nSNP * NSAMPLES^2 hours            if not removing outliers.\n",
    "  2.5e-12 * nSNP * NSAMPLES^2 hours * (1+m)    if m outlier removal iterations.\n",
    "Thus, under the default of up to 5 outlier removal iterations, running time is \n",
    "  up to 1.5e-11 * nSNP * NSAMPLES^2 hours.```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/eigenstrat\n",
    "mkdir results\n",
    "\n",
    "# Run EIGENSTRAT smartpca \n",
    "ancestry=\"ea\"\n",
    "# Run smartpca\n",
    "sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ${ancestry}_smartpca \\\n",
    "    --script_prefix results/smartpca_${ancestry} \\\n",
    "    --mem 13.5 \\\n",
    "    --nslots 1 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/third_party/EIG-6.1.4/bin/smartpca.perl \\\n",
    "        -i ${ancestry}_ld_pruned.bed \\\n",
    "        -a ${ancestry}_ld_pruned_id_renamed.bim \\\n",
    "        -b ${ancestry}_ld_pruned_id_renamed.fam \\\n",
    "        -o results/${ancestry}_ld_pruned.pca \\\n",
    "        -p results/${ancestry}_ld_pruned.plot \\\n",
    "        -e results/${ancestry}_ld_pruned.eval \\\n",
    "        -l results/${ancestry}_ld_pruned.pca.log \\\n",
    "        -m 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## local machine ##\n",
    "mkdir /cygdrive/c/Users/jmarks/Desktop/VIDUS/eigenstrat\n",
    "cd /cygdrive/c/Users/jmarks/Desktop/VIDUS/eigenstrat\n",
    "\n",
    "scp -i ~/.ssh/gwas_rsa ec2-user@35.171.207.199:/shared/sandbox/ngc_vidus-fou/eigenstrat/results/ea_ld_pruned.pca.evec ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load PCA data\n",
    "options(stringsAsFactors=F)\n",
    "ea.evec <- read.table(\"C:/Users/jmarks/Desktop/VIDUS/eigenstrat/ea_ld_pruned.pca.evec\",\n",
    "                      comment.char=\"#\", sep=\"\", row.names=1)\n",
    "ea.evec[1:5,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### R console ###\n",
    "\n",
    "# Function for plotting off-diagonal scatterplots\n",
    "plot.offdiag <- function(x, y, colors=label.colors){\n",
    "    #Make x and y axes equal\n",
    "    limits <- c(min(x,y), max(x,y)) * 1.3\n",
    "    par(usr = c(limits, limits))\n",
    "    points(x,y, col=colors, pch=16)\n",
    "}\n",
    "\n",
    "# Function for plotting diagonal density plots\n",
    "plot.diag <- function(x, colors=label.colors){\n",
    "    d <- density(x, bw=0.01, kernel=\"gaussian\")\n",
    "    #Update plot axes limits\n",
    "    par(usr = c(range(d$x), range(d$y)*1.3))\n",
    "    points(d$x,d$y, type=\"l\")\n",
    "    #Add PC values\n",
    "    points(x=x, y=jitter(rep(mean(d$y),times=d$n),factor=5), col=label.colors, pch=16)\n",
    "}\n",
    "\n",
    "#Run PCA and get % variance\n",
    "ea.dspace <- ea.evec[1:10] #transformed dataspace\n",
    "\n",
    "#Color label samples\n",
    "label.colors <- rgb(0,80,190,80, maxColorValue=255) #blue\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=8)\n",
    "#Redefine outer margin\n",
    "par(oma = c(4, 1, 2, 1))\n",
    "pairs(x=ea.dspace[,6:10], panel=plot.offdiag, diag.panel=plot.diag, \n",
    "      labels=paste0(\"PC\",6:10), label.pos=0.9, cex.labels=1.8)\n",
    "par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 1.5, 0), new=TRUE)\n",
    "plot(0, 0, type = \"n\", bty = \"n\", xaxt = \"n\", yaxt = \"n\", main=\"VIDUS EA\", cex.main=1)\n",
    "\n",
    "\n",
    "par(oma = c(4, 1, 2, 1))\n",
    "pairs(x=ea.dspace[,1:5], panel=plot.offdiag, diag.panel=plot.diag, \n",
    "      labels=paste0(\"PC\",1:5), label.pos=0.9, cex.labels=1.8)\n",
    "par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 1.5, 0), new=TRUE)\n",
    "plot(0, 0, type = \"n\", bty = \"n\", xaxt = \"n\", yaxt = \"n\", main=\"VIDUS EA\", cex.main=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract eigenvectors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/eigenstrat/results\n",
    "\n",
    "# Grab top 10 eigenvectors from PCA \n",
    "ancestry=\"ea\"\n",
    "# Extract eigenvectors\n",
    "echo \"FID IID EV1 EV2 EV3 EV4 EV5 EV6 EV7 EV8 EV9 EV10\" > ${ancestry}_ld_pruned_top10_eigenvecs.txt\n",
    "tail -n +2 ${ancestry}_ld_pruned.pca.evec | \\\n",
    "    perl -lne 's/:/ /; @F=split; print join(\" \",$F[0],$F[1],$F[2],$F[3],$F[4],$F[5],$F[6],$F[7],$F[8],$F[9],$F[10],$F[11]);' \\\n",
    "    >> ${ancestry}_ld_pruned_top10_eigenvecs.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phenotype processing\n",
    "### Retrieve covariates and generate mapping file\n",
    "The covariates we want to include are age (ageatint) and sex (female). Sex is encoded as 0=Male and 1=Female. For the NGC protocol these should be coded as 1 for female and 2 for male.\n",
    "\n",
    "__Note__: some preprocessing was done do make these conversions easier. Also, these steps were processed on a local machine for convenience. Need to `scp` the results over once finished.\n",
    "\n",
    "#### Create ped file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/sandbox/ngc_vidus-fou/phenotype\n",
    "\n",
    "### R ###\n",
    "pheno.data <- read.csv(\"ea_fou_map.data\", sep = \"\\t\", header = T, stringsAsFactors = F, \n",
    "                       colClasses=c(\"character\", \"integer\", \"integer\", \"integer\"))\n",
    "\n",
    "pca.data <- read.csv(\"../eigenstrat/results/ea_ld_pruned_top10_eigenvecs.txt\", sep = \" \" , \n",
    "                    colClasses=c(rep(\"character\",times=2), rep(\"double\",times = 10)))\n",
    "#\n",
    "ped.file <- data.frame(matrix(ncol = 7, nrow = 300))\n",
    "colnames(ped.file) <- c(\"fid\", \"iid\", \"patid\", \"matid\", \"sex\",  \"fou\", \"age\")\n",
    "ped.file[,1] <- pca.data[,2]\n",
    "ped.file[,2] <- pca.data[,2]\n",
    "\n",
    "# map loop (goal to fill in sex and fou on ped.file)\n",
    "for (i in 1:length(ped.file[,1])){\n",
    "    a <- nchar(ped.file[i,1])\n",
    "    key <- substr(ped.file[i,1], a-4, a)\n",
    "    \n",
    "    my_match <- match(key, pheno.data[,1])\n",
    "    ped.file[i,5] <- pheno.data[my_match,2]\n",
    "    ped.file[i,6] <- pheno.data[my_match,4]\n",
    "    ped.file[i,7] <- pheno.data[my_match,3]\n",
    "    \n",
    "} # end i loop\n",
    "\n",
    "write.table(ped.file[,1:6], \"VIDUS_NGCW1_EUR_phen.ped\", sep = \"\\t\", row.names = F, quote=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create covariate file\n",
    "Note, these are processed on local machine. Need to `scp` them over once finished."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "covar.data <- data.frame(matrix(ncol = 16, nrow = 300))\n",
    "colnames(covar.data) <- c(\"fid\", \"iid\", \"patid\", \"matid\", \"sex\", \"age\", \n",
    "                        \"PC1\", \"PC2\",\"PC3\",\"PC4\",\"PC5\",\"PC6\",\"PC7\",\"PC8\",\"PC9\",\"PC10\")\n",
    "covar.data[,1:5] <- ped.file[,1:5]\n",
    "covar.data[,6] <- ped.file[,7]\n",
    "\n",
    "covar.data[,7:length(covar.data)] <- pca.data[,3:length(pca.data)]\n",
    "\n",
    "write.table(covar.data, \"VIDUS_NGCW1_EUR_cov.ped\", quote=FALSE, row.names = FALSE, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Alter IID\n",
    "The imputation files have a unique ID that we must match with our IID. Specifically, the ID\n",
    "in the imputation files is of the for: <br>\n",
    "`iid_iid` <br>\n",
    "so we match our phen and cov ped files to match this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "awk '{OFS=\"_\"} { print $2,$2 } ' VIDUS_NGCW1_EUR_phen.ped > new_id\n",
    "awk 'NR==FNR{ map[NR]=$1; next} FNR>=2{$2=map[FNR]} { print $0}' \\\n",
    "    new_id VIDUS_NGCW1_EUR_phen.ped > phen.tmp && \\\n",
    "    mv phen.tmp VIDUS_NGCW1_EUR_phen.ped\n",
    "awk 'BEGIN{OFS=\"\\t\"} NR==FNR{ map[NR]=$1; next} FNR>=2{$2=map[FNR] } {print $0}' \\\n",
    "    new_id VIDUS_NGCW1_EUR_cov.ped > cov.tmp && \\\n",
    "    mv cov.tmp VIDUS_NGCW1_EUR_cov.ped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### local machine ###\n",
    "cd /cygdrive/c/Users/jmarks/Desktop/VIDUS/eigenstrat\n",
    "\n",
    "scp -i ~/.ssh/gwas_rsa ec2-user@35.171.207.199:/shared/sandbox/ngc_vidus-fou/phenotype/VIDUS_NGCW1_EUR_* ../pheno/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Descriptive Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EUR Box-Cox Lambda:  0.3434343 \n",
      "\n",
      "======== sex - EUR ========\n"
     ]
    },
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "    \n",
       "[1,]\n",
       "[2,]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "========  age  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  4.746667 \n",
      "Standard Deviation:  2.935623 \n",
      "Minimum:  1 \n",
      "Maximum:  15 \n",
      "Median:  5 \n",
      "Q1:  2 \n",
      "Q3:  6 \n",
      "Skewness:  0.84182 \n",
      "Pearson's Kurtosis:  3.534802 \n",
      "\n",
      "========  PC1  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  -3.333333e-07 \n",
      "Standard Deviation:  0.05782844 \n",
      "Minimum:  -0.0361 \n",
      "Maximum:  0.4098 \n",
      "Median:  -0.0171 \n",
      "Q1:  -0.02355 \n",
      "Q3:  -0.0064 \n",
      "Skewness:  3.985033 \n",
      "Pearson's Kurtosis:  20.88465 \n",
      "\n",
      "========  PC2  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  6.666667e-07 \n",
      "Standard Deviation:  0.05783424 \n",
      "Minimum:  -0.1994 \n",
      "Maximum:  0.8167 \n",
      "Median:  -0.0011 \n",
      "Q1:  -0.012 \n",
      "Q3:  0.007225 \n",
      "Skewness:  10.27765 \n",
      "Pearson's Kurtosis:  141.2769 \n",
      "\n",
      "========  PC3  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  3.333333e-07 \n",
      "Standard Deviation:  0.05783303 \n",
      "Minimum:  -0.8359 \n",
      "Maximum:  0.4372 \n",
      "Median:  0.003 \n",
      "Q1:  -0.008725 \n",
      "Q3:  0.01295 \n",
      "Skewness:  -8.678705 \n",
      "Pearson's Kurtosis:  157.498 \n",
      "\n",
      "========  PC4  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  -3e-06 \n",
      "Standard Deviation:  0.05783163 \n",
      "Minimum:  -0.5949 \n",
      "Maximum:  0.386 \n",
      "Median:  0.0031 \n",
      "Q1:  -0.0112 \n",
      "Q3:  0.017675 \n",
      "Skewness:  -3.437779 \n",
      "Pearson's Kurtosis:  49.88775 \n",
      "\n",
      "========  PC5  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  6.666667e-07 \n",
      "Standard Deviation:  0.05783021 \n",
      "Minimum:  -0.4475 \n",
      "Maximum:  0.3995 \n",
      "Median:  -0.0074 \n",
      "Q1:  -0.02495 \n",
      "Q3:  0.014075 \n",
      "Skewness:  0.75998 \n",
      "Pearson's Kurtosis:  25.34053 \n",
      "\n",
      "========  PC6  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  -1e-06 \n",
      "Standard Deviation:  0.05783209 \n",
      "Minimum:  -0.6026 \n",
      "Maximum:  0.3712 \n",
      "Median:  -0.00365 \n",
      "Q1:  -0.02365 \n",
      "Q3:  0.019 \n",
      "Skewness:  -1.969968 \n",
      "Pearson's Kurtosis:  49.83788 \n",
      "\n",
      "========  PC7  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  -1.333333e-06 \n",
      "Standard Deviation:  0.05783322 \n",
      "Minimum:  -0.2235 \n",
      "Maximum:  0.3789 \n",
      "Median:  0.0078 \n",
      "Q1:  -0.01995 \n",
      "Q3:  0.028925 \n",
      "Skewness:  0.3941122 \n",
      "Pearson's Kurtosis:  10.70724 \n",
      "\n",
      "========  PC8  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  -6.666667e-07 \n",
      "Standard Deviation:  0.0578322 \n",
      "Minimum:  -0.4147 \n",
      "Maximum:  0.3193 \n",
      "Median:  0.0014 \n",
      "Q1:  -0.0172 \n",
      "Q3:  0.0202 \n",
      "Skewness:  -1.53328 \n",
      "Pearson's Kurtosis:  20.65077 \n",
      "\n",
      "========  PC9  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  1.666667e-06 \n",
      "Standard Deviation:  0.05782957 \n",
      "Minimum:  -0.5 \n",
      "Maximum:  0.3225 \n",
      "Median:  -0.0013 \n",
      "Q1:  -0.016325 \n",
      "Q3:  0.017625 \n",
      "Skewness:  -3.054494 \n",
      "Pearson's Kurtosis:  37.06673 \n",
      "\n",
      "========  PC10  - EUR ========\n",
      "Number of observations:  300 \n",
      "Mean:  2.333333e-06 \n",
      "Standard Deviation:  0.05783301 \n",
      "Minimum:  -0.6567 \n",
      "Maximum:  0.1852 \n",
      "Median:  0.0024 \n",
      "Q1:  -0.01665 \n",
      "Q3:  0.021675 \n",
      "Skewness:  -5.046009 \n",
      "Pearson's Kurtosis:  59.05891 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "### R console ###\n",
    "library(moments)\n",
    "library(MASS)\n",
    "\n",
    "setwd(\"C://Users/jmarks/Desktop/VIDUS/pheno\")\n",
    "ea.fou.data <- read.table(\"ea_fou_map.data\", header = T)\n",
    "# Plot FOU distributions\n",
    "png(\"fou_distributions.png\", units=\"in\", width=10, height=12, res=200)\n",
    "par(mfrow=c(3,1))\n",
    "hist(ea.fou.data$useropioid6mfq, breaks=30, col=\"cornflowerblue\", border=\"white\",\n",
    "     main=\"EA FOU (Use in past 6 months)\", xlab=\"FOU\")\n",
    "\n",
    "hist((ea.fou.data$useropioid6mfq)^ea.lambda, breaks=30, col=\"cornflowerblue\", \n",
    "     border=\"white\", main=\"EA FOU Box-Cox\", xlab=\"Transformed FOU\")\n",
    "\n",
    "hist(ea.zscore, breaks=30, col=\"cornflowerblue\", border=\"white\", \n",
    "     main=\"EA FOU Z-scores\", xlab=\"Z-score\", xlim=c(-2.5,3))\n",
    "dev.off()\n",
    "\n",
    "# Print Box-Cox lambdas\n",
    "cat(\"EUR Box-Cox Lambda: \", ea.lambda, \"\\n\")\n",
    "\n",
    "# Print categorical variable stats\n",
    "cat(\"\\n======== sex - EUR ========\\n\")\n",
    "rbind(table(ea.fou.data$sex),\n",
    "      round(table(ea.fou.data$sex)/sum(table(ea.fou.data$sex))*100, 4))\n",
    "\n",
    "# Print continuous variable stats\n",
    "for(covar in c(\"age\",paste0(\"PC\",1:10))){\n",
    "    stats <- summary(covar.data[,covar])\n",
    "    cat(\"======== \", covar, \" - EUR ========\\n\")\n",
    "    cat(\"Number of observations: \", length(covar.data[,covar]), \"\\n\")\n",
    "    cat(\"Mean: \", stats[4], \"\\n\")\n",
    "    cat(\"Standard Deviation: \", sd(covar.data[,covar]), \"\\n\")\n",
    "    cat(\"Minimum: \", stats[1], \"\\n\")\n",
    "    cat(\"Maximum: \", stats[6], \"\\n\")\n",
    "    cat(\"Median: \", stats[3], \"\\n\")\n",
    "    cat(\"Q1: \", stats[2], \"\\n\")\n",
    "    cat(\"Q3: \", stats[5], \"\\n\")\n",
    "    cat(\"Skewness: \", skewness(covar.data[,covar]), \"\\n\")\n",
    "    cat(\"Pearson's Kurtosis: \", kurtosis(covar.data[,covar]), \"\\n\")\n",
    "    cat(\"\\n\")\n",
    "}\n",
    "\n",
    "# Make covariate distribution plots\n",
    "for(covar in c(\"age\",paste0(\"PC\",1:10))){\n",
    "    png(paste0(covar, \"_distributions.png\"), units=\"in\", width=10, height=5, res=200)\n",
    "    hist(covar.data[,covar], col=\"cornflowerblue\", border=\"white\", main=paste0(\"EUR - \", covar,\" Distribution\"), xlab=covar)\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run GWAS\n",
    "### Autosomes\n",
    "\n",
    "Need to find out which PCs contribute ~75% of the variance explained. This is in another notebook titled:\n",
    "\n",
    "`20180326_ngc_vidus_fou_ev_selection`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "mkdir -p /shared/sandbox/ngc_vidus-fou/data/assoc_tests\n",
    "cd /shared/sandbox/ngc_vidus-fou/data\n",
    "\n",
    "\n",
    "# Run RVtest\n",
    "ancestry=\"ea\"\n",
    "for chr in {1..22}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${ancestry}_${chr}_rvtest \\\n",
    "        --script_prefix assoc_tests2/${ancestry}_${chr}_rvtest \\\n",
    "        --mem 7.5 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/third_party/rvtests/executable/rvtest \\\n",
    "            --inVcf ../genotype/imputed/vidus_imputed/chr${chr}.dose.vcf.gz \\\n",
    "            --pheno ../phenotype/VIDUS_NGCW1_EUR_phen.ped \\\n",
    "            --pheno-name fou \\\n",
    "            --covar ../phenotype/VIDUS_NGCW1_EUR_cov.ped \\\n",
    "            --covar-name sex,age,PC5,PC6,PC7 \\\n",
    "            --xLabel X \\\n",
    "            --meta score \\\n",
    "            --useResidualAsPhenotype \\\n",
    "            --inverseNormal \\\n",
    "            --qtl \\\n",
    "            --dosage DS \\\n",
    "            --out assoc_tests2/VIDUS_EUR_1000G_fou_chr${chr}\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ChrX  TODO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results processing\n",
    "\n",
    "### Unfiltered results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/data/assoc_tests\n",
    "\n",
    "# Concatenate chr results together\n",
    "ancestry=\"EUR\"\n",
    "echo -e \"\\n\\n ======== ${ancestry} ======== \\n\\n\"\n",
    "for chr in {1..22}; do\n",
    "    zgrep -P \"^${chr}\\s\" VIDUS_${ancestry}_1000G_fou_chr${chr}.MetaScore.assoc.gz\n",
    "done |\n",
    "    gzip -c > VIDUS_${ancestry}_1000G_fou.gz\n",
    "\n",
    "# Make results table\n",
    "ancestry=\"EUR\"\n",
    "outfile=VIDUS_${ancestry}_1000G_fou.assoc.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outfile\n",
    "infile=VIDUS_${ancestry}_1000G_fou.gz\n",
    "echo \"Processing ${infile}\"\n",
    "zcat ${infile} | \\\n",
    "    perl -lane 'if (($F[2] eq \"A\" || $F[2] eq \"C\" || $F[2] eq \"G\" || $F[2] eq \"T\") && (($F[3] eq \"A\" || $F[3] eq \"C\" || $F[3] eq \"G\" || $F[3] eq \"T\"))) {\n",
    "                    print join(\"\\t\",$F[0].\":\".$F[1],$F[0],$F[1],$F[15],\"snp\");\n",
    "                } else {\n",
    "                    print join(\"\\t\",$F[0].\":\".$F[1],$F[0],$F[1],$F[15],\"indel\");\n",
    "                }' >> $outfile\n",
    "\n",
    "# Make Q-Q and manhattan plots\n",
    "ancestry=EUR\n",
    "# Plot all chromosomes\n",
    "sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots_${ancestry} \\\n",
    "    --script_prefix VIDUS_${ancestry}_1000G_fou.assoc.plot \\\n",
    "    --mem 30.6 \\\n",
    "    --priority 0 \\\n",
    "    --program Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "        --in VIDUS_${ancestry}_1000G_fou.assoc.table \\\n",
    "        --in_chromosomes autosomal_nonPAR \\\n",
    "        --in_header \\\n",
    "        --out VIDUS_${ancestry}_1000G_fou.assoc.plot.all_chr \\\n",
    "        --col_id VARIANT_ID \\\n",
    "        --col_chromosome CHR \\\n",
    "        --col_position POSITION \\\n",
    "        --col_p P \\\n",
    "        --col_variant_type TYPE \\\n",
    "        --generate_snp_indel_manhattan_plot \\\n",
    "        --manhattan_odd_chr_color red3 \\\n",
    "        --manhattan_even_chr_color dodgerblue3 \\\n",
    "        --manhattan_points_cex 1.5 \\\n",
    "        --generate_snp_indel_qq_plot \\\n",
    "        --qq_lines \\\n",
    "        --qq_points_bg black \\\n",
    "        --qq_lambda"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MAF > 0.01 and imputation quality > 0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Command line # \n",
    "cd /shared/sandbox/ngc_vidus-fou/data/assoc_tests\n",
    "\n",
    "# Get list of variants to filter by imputation quality\n",
    "ancestry=\"ea\"\n",
    "for chr in {1..22}; do\n",
    "    echo -e \"${ancestry} chr${chr}...\"\n",
    "    zcat ../../genotype/imputed/vidus_imputed/chr${chr}.info.gz | \\\n",
    "        tail -n +2 | \\\n",
    "        awk '{ if($5 > 0.01 && $7 > 0.3){ print $1\":\"$2\":\"$3 } }' \\\n",
    "        > ${ancestry}_chr${chr}_variants_maf_gt_0.01_rsq_gt_0.3.keep\n",
    "done\n",
    "\n",
    "# Filter results files\n",
    "ancestry=EUR\n",
    "for chr in {1..22}; do\n",
    "    echo \"chr${chr}...\"\n",
    "    awk '{ if(NR==FNR){ map[$1]=1; next } { if(map[$1\":\"$2\":\"$3\":\"$4]==1){ print $0 } } }' \\\n",
    "        ea_chr${chr}_variants_maf_gt_0.01_rsq_gt_0.3.keep \\\n",
    "        <(zgrep -P \"^${chr}\\s\" VIDUS_${ancestry}_1000G_fou_chr${chr}.MetaScore.assoc.gz) \\\n",
    "        > VIDUS_${ancestry}_1000G_fou_chr${chr}.tmp\n",
    "done \n",
    "cat VIDUS_${ancestry}_1000G_fou_chr*tmp | gzip -c > VIDUS_${ancestry}_1000G_fou_maf_gt_0.01_rsq_gt_0.3.gz\n",
    "\n",
    "# Clean up\n",
    "rm *tmp\n",
    "\n",
    "# Make results table\n",
    "ancestry=\"EUR\"\n",
    "outfile=VIDUS_${ancestry}_1000G_fou_maf_gt_0.01_rsq_gt_0.3.assoc.table\n",
    "echo -e \"VARIANT_ID\\tCHR\\tPOSITION\\tP\\tTYPE\" > $outfile\n",
    "infile=VIDUS_${ancestry}_1000G_fou_maf_gt_0.01_rsq_gt_0.3.gz\n",
    "echo \"Processing ${infile}\"\n",
    "zcat ${infile} | \\\n",
    "    perl -lane 'if (($F[2] eq \"A\" || $F[2] eq \"C\" || $F[2] eq \"G\" || $F[2] eq \"T\") && (($F[3] eq \"A\" || $F[3] eq \"C\" || $F[3] eq \"G\" || $F[3] eq \"T\"))) {\n",
    "                    print join(\"\\t\",$F[0].\":\".$F[1],$F[0],$F[1],$F[15],\"snp\");\n",
    "                } else {\n",
    "                    print join(\"\\t\",$F[0].\":\".$F[1],$F[0],$F[1],$F[15],\"indel\");\n",
    "                }' >> $outfile\n",
    "\n",
    "# Make Q-Q and manhattan plots\n",
    "ancestry=\"EUR\"\n",
    "# Plot all chromosomes\n",
    "sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name gwas_plots_${ancestry} \\\n",
    "    --script_prefix VIDUS_${ancestry}_1000G_fou_maf_gt_0.01_rsq_gt_0.3.assoc.plot \\\n",
    "    --mem 30.6 \\\n",
    "    --priority 0 \\\n",
    "    --program Rscript /shared/bioinformatics/software/R/generate_gwas_plots.R \\\n",
    "        --in VIDUS_${ancestry}_1000G_fou_maf_gt_0.01_rsq_gt_0.3.assoc.table \\\n",
    "        --in_chromosomes autosomal_nonPAR \\\n",
    "        --in_header \\\n",
    "        --out VIDUS_${ancestry}_1000G_fou_maf_gt_0.01_rsq_gt_0.3.assoc.plot.all_chr \\\n",
    "        --col_id VARIANT_ID \\\n",
    "        --col_chromosome CHR \\\n",
    "        --col_position POSITION \\\n",
    "        --col_p P \\\n",
    "        --col_variant_type TYPE \\\n",
    "        --generate_snp_indel_manhattan_plot \\\n",
    "        --manhattan_odd_chr_color red3 \\\n",
    "        --manhattan_even_chr_color dodgerblue3 \\\n",
    "        --manhattan_points_cex 1.5 \\\n",
    "        --generate_snp_indel_qq_plot \\\n",
    "        --qq_lines \\\n",
    "        --qq_points_bg black \\\n",
    "        --qq_lambda"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PCA Plots\n",
    "The NGC protocol calls for the submission of the samples projected onto 1000G superpopulation PCA space (EUR, AFR, EAS)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mkdir -p /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g/1000g\n",
    "cd /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g\n",
    "\n",
    "cp ../eigenstrat/ea_ld_pruned.bim .\n",
    "awk '{print $2}' ea_ld_pruned.bim > all_pruned_variant_ids\n",
    "\n",
    "cd 1000g\n",
    "# Rename 1000G autosome variant IDs\n",
    "ancestry=ea\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name recode_to_1000g_${chr} \\\n",
    "        --script_prefix ${ancestry}_chr${chr}_id_rename \\\n",
    "        --mem 8 \\\n",
    "        --nslots 1 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/perl/id_conversion/convert_to_1000g_p3_ids.pl \\\n",
    "        --file_in /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr}.bim \\\n",
    "        --file_out chr${chr}_renamed.bim \\\n",
    "        --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.legend.gz \\\n",
    "        --file_in_header 0 \\\n",
    "        --file_in_id_col 1 \\\n",
    "        --file_in_chr_col 0 \\\n",
    "        --file_in_pos_col 3 \\\n",
    "        --file_in_a1_col 4 \\\n",
    "        --file_in_a2_col 5 \\\n",
    "        --chr ${chr}\n",
    "done\n",
    "\n",
    "# get list of all variant IDs\n",
    "touch all_1000g_variants_renamed\n",
    "for chr in {1..22};do\n",
    "awk '{ print $2 }' chr${chr}_renamed.bim >> all_1000g_variants_renamed\n",
    "done\n",
    "\n",
    "sort all_1000g_variants_renamed > all_1000g_variants_renamed_sorted\n",
    "sort ../all_study_variants_renamed > all_study_variants_renamed_sorted\n",
    "\n",
    "# get overlap between 1000g variants and study variants (pruned)\n",
    "comm ../all_study_variants_renamed_sorted all_1000g_variants_renamed_sorted > ../variant_overlap\n",
    "\n",
    "# Get subject IDs by ancestry\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"African\"){print $1,$1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > 1000g_subject_IDs\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"East Asian\"){print $1,$1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    >> 1000g_subject_IDs\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"European\"){print $1,$1} }' /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    >> 1000g_subject_IDs\n",
    "\n",
    "\n",
    "\n",
    "# create binary filesets with ld pruned snps\n",
    "for chr in {1..22}; do\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name chr${chr}_filter \\\n",
    "        --script_prefix ancestry_partition_chr${chr} \\\n",
    "        --mem 8 \\\n",
    "        --priority 0 \\\n",
    "        --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink  \\\n",
    "            --noweb \\\n",
    "            --memory 10000 \\\n",
    "            --bim chr${chr}_renamed.bim \\\n",
    "            --fam /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr}.fam \\\n",
    "            --bed /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr}.bed \\\n",
    "            --bfile /shared/data/ref_panels/1000G/2013.05/plink/ALL.chr${chr} \\\n",
    "            --keep 1000g_subject_IDs \\\n",
    "            --extract ../variant_overlap \\\n",
    "            --make-bed \\\n",
    "            --out 1000g_chr${chr}_ld_pruned\n",
    "done\n",
    "\n",
    "touch autosome_merge_list.txt\n",
    "for chr in {1..22};do\n",
    "   echo \"1000g_chr${chr}_ld_pruned\" >> autosome_merge_list.txt\n",
    "done\n",
    "\n",
    "# merge 1000g autosomes\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name merge_plink_filesets \\\n",
    "    --script_prefix merge_plink_filesets \\\n",
    "    --mem 4 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 4000 \\\n",
    "        --merge-list autosome_merge_list.txt \\\n",
    "        --snps-only just-acgt \\\n",
    "        --make-bed \\\n",
    "        --out 1000g_all_auto_ld_pruned\n",
    "\n",
    "cd /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g\n",
    "\n",
    "# filter study data to the variants that overlap with 1000g\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile ea_ld_pruned \\\n",
    "    --extract variant_overlap \\\n",
    "    --exclude study_and_1000g_combined_ld_pruned-merge.missnp \\\n",
    "    --make-bed \\\n",
    "    --out ea_ld_pruned_1000g_overlap\n",
    "\n",
    "# Attempt merge 1000g with study data. produced 8 snps that were ambiguous (will remove)\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 4000 \\\n",
    "    --bfile 1000g/1000g_all_auto_ld_pruned \\\n",
    "    --bmerge ea_ld_pruned_1000g_overlap \\\n",
    "    --snps-only just-acgt \\\n",
    "    --make-bed \\\n",
    "    --out study_and_1000g_combined_ld_pruned\n",
    "\n",
    "\n",
    "# Exclude ambiguous snps (polymorphic) from study data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile ea_ld_pruned_ld_pruned_1000g_overlap \\\n",
    "    --exclude study_and_1000g_combined_ld_pruned-merge.missnp \\\n",
    "    --make-bed \\\n",
    "    --out ea_ld_pruned_1000g_overlap_removed_ambig\n",
    "\n",
    "\n",
    "# Exclude ambiguous snps (polymorphic) from 1000g data\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 2048 \\\n",
    "    --bfile 1000g/1000g_all_auto_ld_pruned \\\n",
    "    --exclude study_and_1000g_combined_ld_pruned-merge.missnp \\\n",
    "    --make-bed \\\n",
    "    --out 1000g/1000g_all_auto_ld_pruned_removed_ambig\n",
    "\n",
    "# Attempt merge 1000g with study data.\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "    --noweb \\\n",
    "    --memory 4000 \\\n",
    "    --bfile 1000g/1000g_all_auto_ld_pruned_removed_ambig \\\n",
    "    --bmerge ea_ld_pruned_1000g_overlap_removed_ambig \\\n",
    "    --snps-only just-acgt \\\n",
    "    --make-bed \\\n",
    "    --out study_and_1000g_combined_ld_pruned_removed_ambig"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Eigenstrat on LD pruned snps - study & 1000g merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mkdir -p /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g/results\n",
    "cd /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g\n",
    "\n",
    "# make not a case nor control in fam file\n",
    "awk '{$6= \"-9\"} {print $0}' study_and_1000g_combined_ld_pruned_removed_ambig.fam > final.fam\n",
    "\n",
    "# rename study IIDs\n",
    "awk 'NR<=300 { $2=\"ID_\"NR; } {print $0}' final.fam > final_renamed.fam\n",
    "\n",
    "# Run smartpca\n",
    "/shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name study_vs_1000g_smartpca \\\n",
    "    --script_prefix smartpca_study_vs_1000g_ \\\n",
    "    --mem 8 \\\n",
    "    --nslots 2 \\\n",
    "    --priority 0 \\\n",
    "    --program /shared/bioinformatics/software/third_party/EIG-6.1.4/bin/smartpca.perl \\\n",
    "        -i study_and_1000g_combined_ld_pruned_removed_ambig.bed \\\n",
    "        -a study_and_1000g_combined_ld_pruned_removed_ambig.bim \\\n",
    "        -b final_renamed.fam \\\n",
    "        -o results/study_and_1000g_ld_snps.pca \\\n",
    "        -p results/study_and_1000g_ld_snps.plot \\\n",
    "        -e results/study_and_1000g_ld_snps.eval \\\n",
    "        -l results/study_and_1000g_ld_snps.pca.log \\\n",
    "        -m 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g/results\n",
    " \n",
    "# Get subject IDs by ancestry\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"African\"){print $1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > AFR_1000g_subject_IDs\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"East Asian\"){print $1} }'  /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > EAS_1000g_subject_IDs\n",
    "awk 'BEGIN { FS=\"\\t\"; OFS=\"\\t\" } { if($7==\"European\"){print $1} }' /shared/data/ref_panels/1000G/igsr_samples.tsv \\\n",
    "    > EUR_1000g_subject_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /shared/sandbox/ngc_vidus-fou/cohort_projected_on_1000g/results\n",
    "# Load PCA data\n",
    "options(stringsAsFactors=F)\n",
    "study.vs_1000 <- read.table(\"study_and_1000g_ld_snps.pca.evec\", comment.char=\"#\", sep=\"\", row.names=1)\n",
    "\n",
    "# Function for plotting off-diagonal scatterplots\n",
    "plot.offdiag <- function(x, y, colors=label.colors){\n",
    "    #Make x and y axes equal\n",
    "    limits <- c(min(x,y), max(x,y)) * 1.3\n",
    "    par(usr = c(limits, limits))\n",
    "    points(x,y, col=colors, pch=16)\n",
    "}\n",
    "\n",
    "# Function for plotting diagonal density plots\n",
    "plot.diag <- function(x, colors=label.colors){\n",
    "    d <- density(x, bw=0.01, kernel=\"gaussian\")\n",
    "    #Update plot axes limits\n",
    "    par(usr = c(range(d$x), range(d$y)*1.3))\n",
    "    points(d$x,d$y, type=\"l\")\n",
    "    #Add PC values\n",
    "    points(x=x, y=jitter(rep(mean(d$y),times=d$n),factor=5), col=label.colors, pch=16)\n",
    "}\n",
    "\n",
    "#Run PCA and get % variance\n",
    "dspace <- study.vs_1000[1:10] #transformed dataspace\n",
    "\n",
    "AFR = scan(file=\"AFR_1000g_subject_IDs\", what=character())\n",
    "EAS = scan(file=\"EAS_1000g_subject_IDs\", what=character())\n",
    "EUR = scan(file=\"EUR_1000g_subject_IDs\", what=character())\n",
    "\n",
    "#Color label samples\n",
    "sample.names <- row.names(study.vs_1000)\n",
    "label.colors <- rep(NA, length(sample.names))\n",
    "label.colors[grepl(x=sample.names, pattern=\"80059|_93-\", ignore.case=F)] <- rgb(255,0,0,80, maxColorValue=255) #red\n",
    "label.colors[grepl(x=sample.names, pattern=paste(EUR,collapse=\"|\"), ignore.case=F)] <- rgb(0,0,255,80, maxColorValue=255) #blue\n",
    "label.colors[grepl(x=sample.names, pattern=paste(EAS,collapse=\"|\"), ignore.case=F)] <- rgb(0,255,0,80, maxColorValue=255) #green\n",
    "label.colors[grepl(x=sample.names, pattern=paste(AFR,collapse=\"|\"), ignore.case=F)] <- rgb(255,215,0,80, maxColorValue=255) #gold\n",
    "legend.cols <- c(rgb(255,215,0,255, maxColorValue=255), rgb(0,255,0,255, maxColorValue=255),\n",
    "                 rgb(0,0,255,255, maxColorValue=255), rgb(255,0,0,255, maxColorValue=255))\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=8)\n",
    "#Redefine outer margin\n",
    "par(oma = c(4, 1, 2, 1))\n",
    "pairs(x=dspace[,1:4], panel=plot.offdiag, diag.panel=plot.diag, \n",
    "      labels=paste0(\"PC\",1:4), label.pos=0.9, cex.labels=1.8)\n",
    "par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 1.5, 0), new=TRUE)\n",
    "plot(0, 0, type = \"n\", bty = \"n\", xaxt = \"n\", yaxt = \"n\", main=\"Lung Cancer in Never Smokers vs 1000G\", cex.main=1)\n",
    "legend(\"bottom\", c(\"AFR\", \"EAS\", \"EUR\", \"LCNS\"), \n",
    "       pch=19, col=legend.cols,\n",
    "       xpd=TRUE, horiz=TRUE, inset=c(0,-0.02), \n",
    "       bty=\"n\", cex=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## S3 data transfer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Copy phenotype data\n",
    "cd /shared/sandbox/ngc_vidus-fou/phenotype\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/data/phenotype --recursive --exclude=\"*\" --include=\"*ped.gz\" --quiet &\n",
    "\n",
    "cd /shared/sandbox/ngc_vidus-fou/data/assoc_tests\n",
    "# Copy association test results\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/results/rvtest/ --recursive --exclude=\"*\" --include=\"*MetaScore*gz*\" --quiet &\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/results/figures/ --recursive --exclude=\"*\" --include=\"*.png.gz\" --quiet &\n",
    "\n",
    "\n",
    "# copy imputation files\n",
    "cd /shared/sandbox/ngc_vidus-fou/genotype/imputed/vidus_imputed\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/data/genotype/imputed/${ancestry}/ --recursive --exclude \"*\" --include \"*.log\"\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/data/genotype/imputed/${ancestry}/ --recursive --exclude \"*\" --include \"*.info.gz\" \n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/data/genotype/imputed/${ancestry}/ --recursive  --quiet --exclude \"*\" --include \"*dose.vcf.gz\"\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/data/genotype/imputed/${ancestry}/ --recursive --quiet --exclude \"*\" --include \"*dose.vcf.gz.tbi\" &\n",
    "\n",
    "\n",
    "# copy original genotype data\n",
    "cd /shared/sandbox/ngc_vidus-fou/genotype/original/final\n",
    "aws s3 cp ./ s3://rti-heroin/ngc_vidus_fou/data/genotype/original/ea/ --recursive --quiet &"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.1"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
