{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UHS1-3 chrX imputation\n",
    "__Author__: Jesse Marks\n",
    "\n",
    "**Date:** September 19, 2018\n",
    "\n",
    "**GitHub Issue:** [Issue #112](https://github.com/RTIInternational/bioinformatics/issues/112)\n",
    "\n",
    "This document logs the steps taken to perform pre-imputation procedures on the merged datasets of UHS1, UHS2, and UHS3 both EA and AA. The starting point for this analysis is after quality control of observed genotypes. The quality controlled genotypes are oriented on the GRCh37 plus strand. Based on findings from [Johnson et al.](https://link.springer.com/article/10.1007/s00439-013-1266-7), the intersection set of variants for will be used for imputation. \n",
    "\n",
    "## Software and tools\n",
    "The software and tools used for porcessing these data are\n",
    "* [Michigan Imputation Server](https://imputationserver.sph.umich.edu/index.html) (MIS)\n",
    "* [Amazon Web Services (AWS) - Cloud Computing Services](https://aws.amazon.com/)\n",
    "    * Linux AMI\n",
    "* [PLINK v1.90 beta 4.10](https://www.cog-genomics.org/plink/)\n",
    "* [bgzip](http://www.htslib.org/doc/tabix.html)\n",
    "* [BCF Tools](http://www.htslib.org/doc/bcftools.html)\n",
    "* Windows 10 with [Cygwin](https://cygwin.com/) installed\n",
    "* GNU bash version 4.2.46\n",
    "\n",
    "## Data retrieval and organization\n",
    "PLINK binary filesets will be obtained from AWS S3 storage.\n",
    "\n",
    "Nathan Gaddis performed the QC and stored the observed genotype data at:\n",
    "\n",
    "```\n",
    "UHS1\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs1.aa.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs1.ea.chr23\n",
    "\n",
    "UHS2\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs2.aa.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs2.ea.chr23\n",
    "\n",
    "UHS3\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs3.aa.V1-2.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs3.ea.V1-2.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs3.aa.V1-3.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs3.ea.V1-3.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs3.aa.merged.chr23\n",
    "https://rti-midas-data.s3.amazonaws.com/studies/hiv/observed/final/uhs3.ea.merged.chr23\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 ## \n",
    "\n",
    "mkdir -p /home/ec2-user/jmarks/hiv/uhs123_impute/UHS{1,2}/genotype/observed/{ea,aa}\n",
    "mkdir -p /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-{2,3}/genotype/observed/{ea,aa}\n",
    "# download genotype to respective directories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# chrX Statistics Breakdown \n",
    "This table includes the initial number of variants in each study as well as the final number of variants in the intersection set. The `Variants Post-Filtering` is in referral to the filtering steps (1) remove discordant alleles & (2) removal of monomorphic variants.\n",
    "\n",
    "## EA\n",
    "| Data Set      | Initial Variants | Variants Post-Filtering  | Intersection     |\n",
    "|---------------|------------------|--------------------------|------------------|\n",
    "| UHS1          | 17,408           |                          | NA               |\n",
    "| UHS2          | 45,850           |                          | NA               |\n",
    "| UHS3_V1.2     | 24,202           |                          | NA               |\n",
    "| UHS3_V1.3     | 35,440           |                          | NA               |\n",
    "| Merged        | NA               |                          |                  |\n",
    "\n",
    "\n",
    "## AA\n",
    "| Data Set      | Initial Variants | Variants Post-Filtering  | Intersection     |\n",
    "|---------------|------------------|--------------------------|------------------|\n",
    "| UHS1          | 17,396           |                          | NA               |\n",
    "| UHS2          | 39,502           |                          | NA               |\n",
    "| UHS3_V1.2     | 41,884           |                          | NA               |\n",
    "| UHS3_V1.3     | 44,706           |                          | NA               |\n",
    "| Merged        | NA               |                          |                  |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data processing\n",
    "### GRCh37 strand and allele discordance check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "\n",
    "mkdir 1000g\n",
    "\n",
    "for study in {UHS1,UHS2,UHS3_v1-2,UHS3_v1-3}; do\n",
    "    mkdir -p ${base_dir}/${study}/genotype/strand_check\n",
    "    for ancestry in {ea,aa};do\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed ${base_dir}/${study}/genotype/observed/${ancestry}/*bed \\\n",
    "            --bim ${base_dir}/${study}/genotype/observed/${ancestry}/*bim \\\n",
    "            --fam ${base_dir}/${study}/genotype/observed/${ancestry}/*fam \\\n",
    "            --freq \\\n",
    "            --out ${base_dir}/${study}/genotype/strand_check/${ancestry}_chrX\n",
    "    done\n",
    "done\n",
    "\n",
    "\n",
    "# Get list of variants from all studies\n",
    "for ancestry in {ea,aa};do\n",
    "    cat ${base_dir}/{UHS1,UHS2,UHS3_v1-2,UHS3_v1-3}/genotype/observed/${ancestry}/*bim | \\\n",
    "            perl -lane 'if (($F[0]+0) <= 23) { print $F[1]; }' | \\\n",
    "            sort -u > ${base_dir}/${ancestry}_chrX_sorted_variants.txt\n",
    "done\n",
    "\n",
    " wc -l *_chrX_sorted_variants.txt\n",
    "\"\"\"\n",
    "55030 aa_chrX_sorted_variants.txt\n",
    "54454 ea_chrX_sorted_variants.txt\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "# Calculate chrX MAFs for 1000G EUR and AFR\n",
    "chr=23\n",
    "for ancestry in {ea,aa};do\n",
    "    if [ $ancestry == 'aa' ]\n",
    "    then\n",
    "        pop=\"AFR\"\n",
    "        echo $pop\n",
    "    else\n",
    "        pop=\"EUR\"\n",
    "        echo $pop\n",
    "    fi\n",
    "\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${pop}_${chr} \\\n",
    "        --script_prefix ${base_dir}/${pop}_chr${chr}.maf \\\n",
    "        --mem 7 \\\n",
    "        --priority 0 \\\n",
    "        --program perl /shared/bioinformatics/software/perl/stats/calculate_maf_from_impute2_hap_file.pl \\\n",
    "            --hap /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.hap.gz\\\n",
    "            --legend /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3_chrX_NONPAR.legend.gz \\\n",
    "            --sample /shared/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "            --chr ${chr} \\\n",
    "            --extract ${base_dir}/${ancestry}_chrX_sorted_variants.txt \\\n",
    "            --out ${base_dir}/1000g/${pop}_chr${chr}.maf \\\n",
    "            --keep_groups ${pop}\n",
    "done\n",
    "\n",
    "\"\"\"\n",
    "wc -l 1000g/*\n",
    "  51167 1000g/AFR_chr23.maf\n",
    "  50855 1000g/EUR_chr23.maf\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "# # Merge per chr MAFs for EUR\n",
    "# for pop in {EUR,AFR]};do\n",
    "#     head -n 1 ${base_dir}/1000g/${pop}_chr1.maf > ${base_dir}/1000g/${pop}_chr_all.maf\n",
    "#     tail -q -n +2 ${base_dir}/1000g/${pop}_chr{1..23}.maf \\\n",
    "#         >> ${base_dir}/1000g/${pop}_chr_all.maf\n",
    "# done\n",
    "\n",
    "# Run discordance checks for EA and AA groups\n",
    "for study in {UHS1,UHS2,UHS3_v1-2,UHS3_v1-3}; do\n",
    "    for ancestry in {ea,aa};do\n",
    "        if [ $ancestry == \"ea\" ]\n",
    "        then\n",
    "            pop=\"EUR\"\n",
    "        else\n",
    "            pop=\"AFR\"\n",
    "        fi\n",
    "    /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "        --job_name ${ancestry}_${study}_crosscheck \\\n",
    "        --script_prefix ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance_check \\\n",
    "        --mem 6 \\\n",
    "        --priority 0 \\\n",
    "        --program \"Rscript /shared/bioinformatics/software/R/check_study_data_against_1000G.R\n",
    "            --study_bim_file ${base_dir}/${study}/genotype/observed/${ancestry}/*.bim\n",
    "            --study_frq_file ${base_dir}/${study}/genotype/strand_check/${ancestry}_chrX.frq\n",
    "            --ref_maf_file ${base_dir}/1000g/${pop}_chr23.maf\n",
    "            --out_prefix ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Resolving allele discordances\n",
    "The allele discordances will be resolved by\n",
    "* Flipping allele discordances that are fixed by flipping\n",
    "* Removing SNPs with discordant names\n",
    "* Removing SNPs with discordant positions\n",
    "* Removing allele discordances that are not resolved by flipping\n",
    "* Removing alleles with large deviations from the reference population allele frequencies\n",
    "\n",
    "Given that the allele discordance check was done using a union set of SNPs across all studies within an ancestry group, some of the SNPs logged as discordant for a given study may not actually be in the study. Fortunately, if they are not in a given study they will not interfere with the filtering procedures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "\n",
    "for study in {UHS1,UHS2,UHS3_v1-2,UHS3_v1-3}; do\n",
    "    for ancestry in {ea,aa};do\n",
    "        echo -e \"\\n===============\\nProcessing ${study}\\n\"\n",
    "        # Create remove list\n",
    "        echo \"Making remove list\"\n",
    "        cat <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.at_cg_snps_freq_diff_gt_0.2 | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.discordant_names | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.discordant_positions | tail -n +2) \\\n",
    "            <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.discordant_alleles_polymorphic_in_study_not_fixed_by_strand_flip | tail -n +2) | \\\n",
    "            sort -u > ${base_dir}/${study}/genotype/${ancestry}_snps.remove\n",
    "\n",
    "        # Create flip list\n",
    "        echo \"Making flip list\"\n",
    "        comm -23 <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.discordant_alleles | tail -n +2 | sort -u) \\\n",
    "            <(cut -f2,2 ${base_dir}/${study}/genotype/strand_check/${ancestry}_allele_discordance.discordant_alleles_not_fixed_by_strand_flip | tail -n +2 | sort -u) \\\n",
    "            > ${base_dir}/${study}/genotype/${ancestry}_snps.flip\n",
    "\n",
    "        # Apply filters\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bed ${base_dir}/${study}/genotype/observed/${ancestry}/*bed \\\n",
    "            --bim ${base_dir}/${study}/genotype/observed/${ancestry}/*bim \\\n",
    "            --fam ${base_dir}/${study}/genotype/observed/${ancestry}/*fam \\\n",
    "            --exclude ${base_dir}/${study}/genotype/${ancestry}_snps.remove \\\n",
    "            --flip ${base_dir}/${study}/genotype/${ancestry}_snps.flip \\\n",
    "            --make-bed \\\n",
    "            --out ${base_dir}/${study}/genotype/observed/$ancestry/${ancestry}_filtered\n",
    "    done\n",
    "done\n",
    "\n",
    "\"\"\"\n",
    " wc -l */genotype/ob*/*/*bim\n",
    " wc -l */genotype/ob*/{aa,ea}/*bim\n",
    "   17244 UHS1/genotype/observed/aa/aa_filtered.bim\n",
    "   17396 UHS1/genotype/observed/aa/uhs1.aa.chr23.bim\n",
    "   39367 UHS2/genotype/observed/aa/aa_filtered.bim\n",
    "   39502 UHS2/genotype/observed/aa/uhs2.aa.chr23.bim\n",
    "   40158 UHS3_v1-2/genotype/observed/aa/aa_filtered.bim\n",
    "   41884 UHS3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bim\n",
    "   43426 UHS3_v1-3/genotype/observed/aa/aa_filtered.bim\n",
    "   44706 UHS3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bim\n",
    "   17255 UHS1/genotype/observed/ea/ea_filtered.bim\n",
    "   17408 UHS1/genotype/observed/ea/uhs1.ea.chr23.bim\n",
    "   45738 UHS2/genotype/observed/ea/ea_filtered.bim\n",
    "   45850 UHS2/genotype/observed/ea/uhs2.ea.chr23.bim\n",
    "   23834 UHS3_v1-2/genotype/observed/ea/ea_filtered.bim\n",
    "   24202 UHS3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bim\n",
    "   34753 UHS3_v1-3/genotype/observed/ea/ea_filtered.bim\n",
    "   35440 UHS3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bim\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove monomorphic variants\n",
    "Monomorphic variants prevent MIS from accepting the genotype data. In this case, an arbitrarily small MAF is set that is smaller than the lower bound for these data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "\n",
    "# Apply filters\n",
    "for study in {UHS1,UHS2,UHS3_v1-2,UHS3_v1-3}; do\n",
    "    for ancestry in {ea,aa};do\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --memory 2048 \\\n",
    "            --bfile ${base_dir}/${study}/genotype/observed/$ancestry/${ancestry}_filtered \\\n",
    "            --maf 0.000001 \\\n",
    "            --make-bed \\\n",
    "            --out ${base_dir}/${study}/genotype/observed/$ancestry/${ancestry}_filtered_mono\n",
    "    done\n",
    "done\n",
    "\n",
    "\"\"\"\n",
    "wc -l ${base_dir}/*/genotype/observed/{ea,aa}/*bim\n",
    "   17255 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS1/genotype/observed/ea/ea_filtered.bim\n",
    "   15532 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS1/genotype/observed/ea/ea_filtered_mono.bim\n",
    "   17408 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS1/genotype/observed/ea/uhs1.ea.chr23.bim\n",
    "   45738 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS2/genotype/observed/ea/ea_filtered.bim\n",
    "   34665 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS2/genotype/observed/ea/ea_filtered_mono.bim\n",
    "   45850 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS2/genotype/observed/ea/uhs2.ea.chr23.bim\n",
    "   23834 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-2/genotype/observed/ea/ea_filtered.bim\n",
    "   19704 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-2/genotype/observed/ea/ea_filtered_mono.bim\n",
    "   24202 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-2/genotype/observed/ea/uhs3.ea.V1-2.chr23.bim\n",
    "   34753 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-3/genotype/observed/ea/ea_filtered.bim\n",
    "   26381 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-3/genotype/observed/ea/ea_filtered_mono.bim\n",
    "   35440 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-3/genotype/observed/ea/uhs3.ea.V1-3.chr23.bim\n",
    "   17244 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS1/genotype/observed/aa/aa_filtered.bim\n",
    "   16395 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS1/genotype/observed/aa/aa_filtered_mono.bim\n",
    "   17396 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS1/genotype/observed/aa/uhs1.aa.chr23.bim\n",
    "   39367 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS2/genotype/observed/aa/aa_filtered.bim\n",
    "   35951 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS2/genotype/observed/aa/aa_filtered_mono.bim\n",
    "   39502 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS2/genotype/observed/aa/uhs2.aa.chr23.bim\n",
    "   40158 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-2/genotype/observed/aa/aa_filtered.bim\n",
    "   37063 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-2/genotype/observed/aa/aa_filtered_mono.bim\n",
    "   41884 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-2/genotype/observed/aa/uhs3.aa.V1-2.chr23.bim\n",
    "   43426 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-3/genotype/observed/aa/aa_filtered.bim\n",
    "   39760 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-3/genotype/observed/aa/aa_filtered_mono.bim\n",
    "   44706 /home/ec2-user/jmarks/hiv/uhs123_impute/UHS3_v1-3/genotype/observed/aa/uhs3.aa.V1-3.chr23.bim\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "\n",
    "mkdir ${base_dir}/intersect\n",
    "studies=(UHS1 UHS2 UHS3_v1-2 UHS3_v1-3) # array of study names\n",
    "num=${#studies[@]}\n",
    "    \n",
    "# Get intersection set\n",
    "for ancestry in {ea,aa};do\n",
    "    file1=${base_dir}/${studies[0]}/genotype/observed/$ancestry/${ancestry}_filtered_mono.bim\n",
    "    file2=${base_dir}/${studies[1]}/genotype/observed/$ancestry/${ancestry}_filtered_mono.bim\n",
    "    file3=${base_dir}/${studies[2]}/genotype/observed/$ancestry/${ancestry}_filtered_mono.bim\n",
    "    file4=${base_dir}/${studies[3]}/genotype/observed/$ancestry/${ancestry}_filtered_mono.bim\n",
    "    echo -e \"\\nCalculating intersection between ${file1}, ${file2}, ${file3}, and ${file4}...\\n\"\n",
    "    #comm -12 <(cut -f 2,2 $file1 | sort -u) <(cut -f 2,2 $file2 | sort -u) \\\n",
    "    sort $file{1,2,3,4} | uniq -dc | awk -v num=$num '$1 == num {print $3}' \\\n",
    "        > ${base_dir}/intersect/${ancestry}_variant_intersection.txt\n",
    "\n",
    "    # Make new PLINK binary file sets\n",
    "    for study in ${studies[@]}; do\n",
    "        /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "            --noweb \\\n",
    "            --bfile ${base_dir}/${study}/genotype/observed/$ancestry/${ancestry}_filtered_mono \\\n",
    "            --extract ${base_dir}/intersect/${ancestry}_variant_intersection.txt \\\n",
    "            --make-bed \\\n",
    "            --out ${base_dir}/intersect/${study}_${ancestry}_filtered_snp_intersection\n",
    "    done\n",
    "done\n",
    "\n",
    "\"\"\"\n",
    " wc -l ${base_dir}/intersect/*txt\n",
    "  7788 /home/ec2-user/jmarks/hiv/uhs123_impute/intersect/aa_variant_intersection.txt\n",
    "  4567 /home/ec2-user/jmarks/hiv/uhs123_impute/intersect/ea_variant_intersection.txt\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge test\n",
    "As a final check to confirm that our data sets are all compatible, a PLINK file set merge is conducted. If any issues persist then an error will be raised."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "\n",
    "\n",
    "for ancestry in {ea,aa};do\n",
    "    echo \"Creating $ancestry merge-list\"\n",
    "    touch ${base_dir}/intersect/${ancestry}_merge_list.txt\n",
    "\n",
    "    for study in {UHS1,UHS2,UHS3_v1-2,UHS3_v1-3};do\n",
    "        echo ${base_dir}/intersect/${study}_${ancestry}_filtered_snp_intersection >>\\\n",
    "            ${base_dir}/intersect/${ancestry}_merge_list.txt\n",
    "    done\n",
    "done\n",
    "\n",
    "# Merge file sets\n",
    "for ancestry in {ea,aa};do\n",
    "    echo -e \"\\n\\n======== ${ancestry} ========\\n\\n\"\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 4000 \\\n",
    "        --merge-list ${base_dir}/intersect/${ancestry}_merge_list.txt \\\n",
    "        --make-bed \\\n",
    "        --out ${base_dir}/intersect/${ancestry}_studies_merged\n",
    "done\n",
    "\n",
    "\"\"\"\n",
    " wc -l ${base_dir}/intersect/*merged.bim\n",
    "  7788 /home/ec2-user/jmarks/hiv/uhs123_impute/intersect/aa_studies_merged.bim\n",
    "  4567 /home/ec2-user/jmarks/hiv/uhs123_impute/intersect/ea_studies_merged.bim\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imputation preparation for Michigan Imputation Server\n",
    "Visit the [MIS Getting Started Webpage](https://imputationserver.sph.umich.edu/start.html#!pages/help) for more information about the preparing the data for upload to MIS.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VCF File Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute\n",
    "\n",
    "mkdir ${base_dir}/phase_prep\n",
    "\n",
    "# Split by chr and remove any individuals with missing data for whole chr\n",
    "chr=23\n",
    "for ancestry in {ea,aa};do\n",
    "    # Remove SNPs\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 4000 \\\n",
    "        --bfile ${base_dir}/intersect/${ancestry}_studies_merged \\\n",
    "        --chr ${chr} \\\n",
    "        --mind 0.99 \\\n",
    "        --make-bed \\\n",
    "        --out ${base_dir}/phase_prep/${ancestry}_chr${chr}_for_phasing \n",
    "done > ${base_dir}/phase_prep/chr_splitting.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Note__: No subjects were removed.\n",
    "\n",
    "```\n",
    "wc -l ${base_dir}/phase*/*bim\n",
    "  7788 /home/ec2-user/jmarks/hiv/uhs123_impute/phase_prep/aa_chr23_for_phasing.bim\n",
    "  4567 /home/ec2-user/jmarks/hiv/uhs123_impute/phase_prep/ea_chr23_for_phasing.bim\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# EC2 command line #\n",
    "base_dir=/home/ec2-user/jmarks/hiv/uhs123_impute/phase_prep\n",
    "\n",
    "chr=23\n",
    "for ancestry in {ea,aa};do\n",
    "    mkdir ${base_dir}/${ancestry}\n",
    "    /shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 5000 \\\n",
    "        --bfile ${base_dir}/${ancestry}_chr${chr}_for_phasing \\\n",
    "        --output-chr M \\\n",
    "        --set-hh-missing \\\n",
    "        --recode vcf bgz \\\n",
    "        --out ${base_dir}/${ancestry}/${ancestry}_chr${chr}_final\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transfer the *.vcf.gz files to local machine (per chromosome) and then upload to MIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/jmarks/Desktop/Data_transfer/MIS/UHS1-3_merged/ea\n",
    "scp -i ~/.ssh/gwas_rsa ec2-user@18.215.115.45:/home/ec2-user/jmarks/hiv/uhs123_impute/phase_prep/ea/* .\n",
    "\n",
    "cd /home/jmarks/Desktop/Data_transfer/MIS/UHS1-3_merged/aa\n",
    "scp -i ~/.ssh/gwas_rsa ec2-user@18.215.115.45:/home/ec2-user/jmarks/hiv/uhs123_impute/phase_prep/aa/* ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to Michigan Imputation Server (MIS)\n",
    "\n",
    "### Uploading parameters\n",
    "These are the parameters that were selected on MIS.\n",
    "\n",
    "__Name__: UHS1-3_merged_chrX_EA\n",
    "\n",
    "__Reference Panel__ 1000G Phase 3 v5\n",
    "\n",
    "__Input Files__ File Upload <br>\n",
    "\n",
    "* Select Files - select VCF (.gz) files that were downloaded to local machine from cloud. <br>\n",
    "\n",
    "__Phasing__: ShapeIT v2.r790 (unphased) \n",
    "\n",
    "__Population__: EUR\n",
    "\n",
    "__Mode__: Quality Control & Imputation\n",
    "\n",
    "* I will not attempt to re-identify or contact research participants.\n",
    "* I will report any inadvertent data release, security breach or other data management incident of which I become aware."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Troubleshooting\n",
    "\n",
    "```\n",
    "Chromosome X check failed! \n",
    "java.io.IOException: Found haplotype 0/1 at pos 2788707 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 2825403 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 2862721 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 2985672 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3002687 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3028289 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3296294 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3542110 for male proband 8002221030_HHG6599_AS93-4798_8002221030_HHG6599_AS93-4798\n",
    "Found haplotype 0/1 at pos 3627414 for male proband 8002221030_HHG6599\n",
    "Error during manifest file creation.\n",
    "```\n",
    "\n",
    "**We are going to remove subject listed above in the MIS error message from the study.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/ec2-user/jmarks/hiv/uhs123_impute/phase_prep\n",
    "mkdir remove_subject\n",
    "ancestry=aa\n",
    "chr=23\n",
    "echo 002221030_HHG6599_AS93-4798 002221030_HHG6599_AS93-4798 > remove_subject/remove_list.tsv\n",
    "\n",
    "\n",
    "/shared/bioinformatics/software/third_party/plink-1.90-beta-4.10-x86_64/plink \\\n",
    "        --noweb \\\n",
    "        --memory 5000 \\\n",
    "        --bfile ${ancestry}_chr${chr}_for_phasing \\\n",
    "        --output-chr M \\\n",
    "        --set-hh-missing \\\n",
    "        --remove-fam remove_subject/remove_list.tsv \\\n",
    "        --recode vcf bgz \\\n",
    "        --out remove_subject/${ancestry}_chr${chr}_final_v02"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Imputed Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 (share drive)\n",
    "# Note: jmarks is a symlink; you will have to save results\n",
    "# to the share drive because they are very large)\n",
    "\n",
    "# create directory structure\n",
    "cd ~/jmarks/heroin/VIDUS_and_IMP_BIOBANK_imputation/imputed_results\n",
    "\n",
    "mkdir chr{1..23}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Autosomes\n",
    "Job name -> `vidus_imp_biobank_1-22`\n",
    "\n",
    "* The password for the imputation results is: HRwTl0U?1RevTh\n",
    "This password was sent in email from MIS once the data were generated.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2 share drive (note jmarks is a symlink; you will have to save results\n",
    "# to the share drive because they are very large)\n",
    "cd ~/jmarks/heroin/VIDUS_and_IMP_BIOBANK_imputation/imputed_results\n",
    "\n",
    "# QC-results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/15d7cfd33016ad1c6c23ff2c3bb6dec6/qcreport.html -a chr1-22.wget.log.report &\n",
    "\n",
    "# SNP Statistics\n",
    "wget https://imputationserver.sph.umich.edu/share/results/f51679f8cccf17e2be83d6a795f42015/statistics.txt -a chr1-22.wget.log.report &\n",
    "\n",
    "# Logs\n",
    "wget https://imputationserver.sph.umich.edu/share/results/528bf38a57fda3e8aa47065a2047fae9/chr_1.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/e2f9049ed6e4b5647ec703d0059cc6b3/chr_10.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/20b29d8fe17575c7909072b703c0722e/chr_11.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/55809bf8a338994e91a68bfc1f4568ce/chr_12.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/360fa6e89d7ca4b973ffd355077adc8a/chr_13.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/76ecab86c7e480fa8f304b8f866e5f45/chr_14.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/f9e496bf617f0db1896bc14d29f16118/chr_15.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/cda963af9e91f362f946d91ca93c09e3/chr_16.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/468211956caeb7cdd138c19a1b9b5957/chr_17.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/eb287d1a1a4d26abdcd6e9668036f565/chr_18.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/e499b5f438cf514b21256cd42de758f7/chr_19.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/1cf1b8686eaa6533d9e147bb5c95ffd6/chr_2.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/c83776228f3274bc9aaef8ab07aea037/chr_20.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/9f74833822d8e09ede6da972886c27f0/chr_21.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/73ad60a766e296c4a27e466676fe99c/chr_22.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/eee2aadd5cef19a3cb2c5da22ac48667/chr_3.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/d072ae837d55315b9ff7620e3a36e205/chr_4.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/828fc3514965c82d4dfbd4e7b3804d89/chr_5.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/d6e6708291b5f88ba191ab5b9acb886d/chr_6.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/26c2b80f626e675de969055e181e16c8/chr_7.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/25b90f904871a19bba2926fd8dc74f81/chr_8.log -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/2e924e3a952243540265cde07798aa67/chr_9.log -a chr1-22.wget.log.report &\n",
    "\n",
    "# Imputation Results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/fb0dafc0c4fdcbc1384cbadc60a4fc30/chr_1.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/77761f1da4ce9729dc41cfa79db3a9dd/chr_10.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/e18e2100d4616d321a7de408c5ce2ea7/chr_11.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/eb5e9fb78dd8a3393f2554fee0d93ea6/chr_12.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/4e6d8d03ec1e0e960f49244c9cb5258e/chr_13.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/73e41057d80cb04aa7a4d4611ddd66e0/chr_14.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/c4250472e5539c2de225c7277e27655/chr_15.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/9a52b28d8585fcbaa0f00d899e8f5e7e/chr_16.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/b14b7583f279fb2fce85e34dde72d2c9/chr_17.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/b28ce2c6203d1ad104241992f089f68a/chr_18.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/cd42b1972ba52a67115523e3c977f4e5/chr_19.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/d45eb247e6dafb2029e263aea9f4c92f/chr_2.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/3ebef28238bf58ae1b5a1411dfe075ee/chr_20.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/577019b3c64f8e7bebe1c3cc3f32713a/chr_21.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/5ef11f7837c55b0bbc5eb2934e80706c/chr_22.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/bec79ea98b2429dc5dc6e4c3ed9b7508/chr_3.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/c33a5a67883a10571200769d610f988f/chr_4.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/f7feed686b38725e053ef069f3dbd2f6/chr_5.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/9b8ca295ae22e783bba30206dbe03df3/chr_6.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/1cca901bc248b284acdd62efe52174d/chr_7.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/15a01d8ed114a7130f9d2fb244db84a0/chr_8.zip -a chr1-22.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/aec349a81e6d98e818ad4fe2c83e702d/chr_9.zip -a chr1-22.wget.log.report &\n",
    "\n",
    "\n",
    "# move files to respective directories\n",
    "for chr in {1..22};do mv chr_$chr.* chr$chr; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inflate imputation results\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EC2\n",
    "cd /home/ec2-user/jmarks/heroin/VIDUS_and_IMP_BIOBANK_imputation/imputed_results\n",
    "\n",
    "# inflate chr results\n",
    "for chr in {1..22};do\n",
    "echo '#!/bin/bash' > chr$chr/chr_$chr.sh\n",
    "echo '' >> chr$chr/chr_$chr.sh\n",
    "echo 'unzip -P \"HRwTl0U?1RevTh\"' chr$chr/chr_$chr.zip >> chr$chr/chr_$chr.sh\n",
    "done\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    sh /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "    --job_name ea.inflate_chr${chr} \\\n",
    "    --script_prefix chr$chr/chr${chr}_inflation \\\n",
    "    --mem 5 \\\n",
    "    --priority 0 \\\n",
    "    --nslots 1 \\\n",
    "    --program bash chr$chr/chr_${chr}.sh\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ChrX\n",
    "Job name -> `vidus_imp_biobank_23`\n",
    "\n",
    "* the password for the imputation results is: zt>3G0BtmrQGY\n",
    "This password was sent in email from MIS once the data were generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# QC-results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/5e39060127c4d3c8d440c6d9d42fb279/qcreport.html -a chr23.wget.log.report &\n",
    "\n",
    "# SNP Statistics\n",
    "wget -o chrX.snp_stats.download.log -- -x https://imputationserver.sph.umich.edu/share/results/6a0e50bb9aa337b0558796ca220cb023/statistics.txt -a chr23.wget.log.report &\n",
    "\n",
    "\n",
    "# Logs\n",
    "wget https://imputationserver.sph.umich.edu/share/results/4370b38b5c563aa06c6b52343dc6356e/chr_X.no.auto_female.log -a chr23.wget.log.report &\n",
    "wget https://imputationserver.sph.umich.edu/share/results/cce234154e59153ccbb37ad21f9d615c/chr_X.no.auto_male.log -a chr23.wget.log.report &\n",
    "\n",
    "\n",
    "# Imputation Results\n",
    "wget https://imputationserver.sph.umich.edu/share/results/53f90217f0abc8b697297c38bf53d1a1/chr_X.no.auto_female.zip -a chr23.wget.log.report &\n",
    "\n",
    "mv chr_X* chr23"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inflate imputation results\n",
    "The zip files from Michigan Imputation Server (MIS) need to be inflated before you can begin working with them. They require a passcode that is sent by MIS to email."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inflate chr results\n",
    "cd /home/ec2-user/jmarks/heroin/VIDUS_and_IMP_BIOBANK_imputation/imputed_results/chr23\n",
    "unzip -P \"zt>3G0BtmrQGY\" *zip "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /home/ec2-user/jmarks/heroin/VIDUS_and_IMP_BIOBANK_imputation/imputed_results\n",
    "\n",
    "\n",
    "# mv inflated results to respective directories and upload to s3\n",
    "for chr in {1..22};do\n",
    "    mv chr$chr.* chr$chr/\n",
    "    aws s3 sync chr$chr/ s3://rti-heroin/imp_biobank-vidus_merged/data/genotype/imputed/chr$chr --quiet &\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
