{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "84d383ba-a342-4351-bc3a-52ad0bb7191d",
   "metadata": {},
   "source": [
    "# GÃ—Vit D in UK Biobank: calculate LDSC intercept\n",
    "[GitHub Issue 109](https://github.com/RTIInternational/bioinformatics/issues/109#issuecomment-840591260)\n",
    "Investigate inflated lambda with LD intercepts: due to residual bias or large amount of positive signal? The [Bulik-Sullivan 2015 Nat. Genet](https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.ncbi.nlm.nih.gov%2Fpmc%2Farticles%2FPMC4495769%2F&data=04%7C01%7Cjmarks%40rti.org%7Cb2d84e99d0fc4744238b08d8efc7a822%7C2ffc2ede4d4449948082487341fa43fb%7C0%7C0%7C637522988106115849%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Wm2EKtjtrfxWp442dd8xmodHY00maD41tUDgTZ8oq0o%3D&reserved=0) paper describes how the LDSC intercept can reveal a bit more information about you summary statitics than with only the genomic inflaction factor (lambda). If the lambda value is inflated and the LDSC intercept is ~1 then you can reason that the inflated lambda is due to polygenicity and not cryptic relatedness, population stratification, or sample overlap. Because as sample sizes and power increases, so does lambda.\n",
    "\n",
    "**FVC**\n",
    "\n",
    "UK Biobank\n",
    "Filtered: `s3://rti-pulmonary/gwas/ukbiobank/results/fvc/0003/chr{1..22}/chr{1..22}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz`\n",
    "        \n",
    "**FEV1**\n",
    "\n",
    "UK Biobank\n",
    "Filtered: `s3://rti-pulmonary/gwas/ukbiobank/results/fev1/0003/chr{1..22}/chr{1..22}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz`\n",
    "\n",
    "\n",
    "<br>\n",
    "\n",
    "**Description**: \n",
    "```\n",
    "@jaamarks It would be great if you could compute the LDSR intercept, and it is needed only for the UK Biobank filtered results. Note that we have one set of results for FEV1 and one for FVC; within each phenotype, we want to know the intercept for the 1df SNP main effect term, the 1df SNPxVitD interaction term, and the 2df result.\n",
    "\n",
    "@ngaddis the filtered UK Biobank results with Rsq>0.8 are provided above. Can you provide the Rsq>0.3 results? Also, while Jesse is working with these results, would it be helpful for him to run the QQ plot for the 1df SNP main effect term? (Jesse, FYI: the QQ and Manhattan plots are already generated for the 1df interaction and 2df results).\n",
    "    \n",
    "@danahancock @jaamarks The UK Biobank rsq>0.3 results are in the same S3 location as the rsq>0.8 results. Yeah, it would be great if Jesse could run the QQ plot for the 1df main effect term. The one catch is that the p-value for this term is not provided in the results file. It does provide the beta and SE though, so the p-value can be calculated.    \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfa97812-82b4-4316-acc5-0be2a6246e78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# setup environment\n",
    "fvcD=/shared/rti-pulmonary/gwas/ukbiobank/results/fvc/0003/\n",
    "fev1D=/shared/rti-pulmonary/gwas/ukbiobank/results/fev1/0003/\n",
    "mkdir -p $fvcD\n",
    "mkdir -p $fev1D\n",
    "\n",
    "# restore objects\n",
    "for chr in {1..22}; do\n",
    "    #FVC\n",
    "    aws s3api restore-object \\\n",
    "        --bucket rti-pulmonary \\\n",
    "        --key gwas/ukbiobank/results/fvc/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz \\\n",
    "        --restore-request '{\"GlacierJobParameters\":{\"Tier\":\"Standard\"}}' \n",
    "     \n",
    "    aws s3api restore-object \\\n",
    "        --bucket rti-pulmonary \\\n",
    "        --key gwas/ukbiobank/results/fvc/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz \\\n",
    "        --restore-request '{\"GlacierJobParameters\":{\"Tier\":\"Standard\"}}' \n",
    "    \n",
    "    #FEV1\n",
    "    aws s3api restore-object \\\n",
    "        --bucket rti-pulmonary \\\n",
    "        --key gwas/ukbiobank/results/fev1/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz \\\n",
    "        --restore-request '{\"GlacierJobParameters\":{\"Tier\":\"Standard\"}}' \n",
    "     \n",
    "    aws s3api restore-object \\\n",
    "        --bucket rti-pulmonary \\\n",
    "        --key gwas/ukbiobank/results/fev1/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz \\\n",
    "        --restore-request '{\"GlacierJobParameters\": {\"Tier\": \"Expedited\"}}'\n",
    "done\n",
    "\n",
    "aws s3api head-object \\\n",
    "    --bucket rti-pulmonary \\\n",
    "    --key gwas/ukbiobank/results/fev1/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz\n",
    "\n",
    "\n",
    "cd $fvcD\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-pulmonary/gwas/ukbiobank/results/fvc/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz .\n",
    "    aws s3 cp s3://rti-pulmonary/gwas/ukbiobank/results/fvc/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz .\n",
    "done\n",
    "\n",
    "cd $fev1D\n",
    "for chr in {1..22}; do \n",
    "    aws s3 cp s3://rti-pulmonary/gwas/ukbiobank/results/fev1/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz .\n",
    "    aws s3 cp s3://rti-pulmonary/gwas/ukbiobank/results/fev1/0003/chr$chr/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz .\n",
    "done "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "313e2703-15be-4a80-ac28-8f8284cc1c56",
   "metadata": {},
   "outputs": [],
   "source": [
    "## pseudocode\n",
    "tval <- Est.G / SE.G\n",
    "\n",
    "# The function pt returns the value of the cumulative density function (cdf) of the Student t distribution\n",
    "pvalue <- 2 * pt( abs(tval) , df=n.obs-1, lower.tail=FALSE ) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2498fd9-86b5-4d2a-b7e9-8a98dc7bd205",
   "metadata": {
    "toc-hr-collapsed": true
   },
   "source": [
    "# P-Value\n",
    "The P-value for the 1df main effect term was not included in the GWAS summary statistics. We will therefore have to calculate these P-values."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66e2e381-2af2-4684-bc22-7c887b0c68f9",
   "metadata": {},
   "source": [
    "## Calculate P-value for the 1df main effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3687b72-8c8e-4c8e-aa0f-a824903e228e",
   "metadata": {},
   "source": [
    "### Rscript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76fb9501-22bd-48cc-b33a-132ed0d71af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/share/apps/R/bin/Rscript\n",
    "\n",
    "args <- commandArgs(TRUE)\n",
    "\n",
    "loop = TRUE\n",
    "fileIn = \"\"\n",
    "fileOut = \"\"\n",
    "colBeta = \"\"\n",
    "colSE = \"\"\n",
    "colN = \"\"\n",
    "fileInHeader = TRUE\n",
    "#chiDf=1\n",
    "\n",
    "while (loop) {\n",
    "\n",
    "    if (args[1] == \"--in\") {\n",
    "        fileIn = args[2]\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--in_header\") {\n",
    "        fileInHeader = TRUE\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--out\") {\n",
    "        fileOut = args[2]\n",
    "    }\n",
    "    \n",
    "    if (args[1] == \"--col_p\") {\n",
    "        colP = gsub(\"-\",\".\",args[2])\n",
    "    }\n",
    "\n",
    "    if (args[1] == \"--col_beta\") {\n",
    "        colBeta = gsub(\"-\",\".\",args[2])\n",
    "    }\n",
    "\n",
    "    if (args[1] == \"--col_se\") {\n",
    "        colSE = gsub(\"-\",\".\",args[2])\n",
    "    }\n",
    "\n",
    "    if (args[1] == \"--col_n\") {\n",
    "        colN = gsub(\"-\",\".\",args[2])\n",
    "    }\n",
    "\n",
    "    if (length(args) > 1) {\n",
    "        args = args[2:length(args)]\n",
    "    } else {\n",
    "        loop=FALSE\n",
    "    }\n",
    "}\n",
    "\n",
    "if (fileIn == \"\") {\n",
    "    stop(\"No input file specified\")\n",
    "} else if (fileOut == \"\") {\n",
    "    stop(\"No output file specified\")\n",
    "} \n",
    "\n",
    "cat(\"Reading \", fileIn, \"...\\n\", sep = \"\")\n",
    "inputData = read.table(fileIn, header = fileInHeader)\n",
    "\n",
    "cat(\"Calculating t-values...\\n\")\n",
    "tvalues = inputData[, colBeta] / inputData[, colSE]\n",
    "#print(tvalues)\n",
    "\n",
    "cat(\"Calculating p-values...\\n\")\n",
    "pvalues <- 2 * pt( abs(tvalues) , df=inputData[, colN]-1, lower.tail=FALSE ) \n",
    "#print(pvalues)\n",
    "\n",
    "inputData$Main.pval <- pvalues\n",
    "\n",
    "cat(paste(\"Writing file\",fileOut,\"...\\n\"))\n",
    "write.table(inputData, file = fileOut, row.names = FALSE, quote = FALSE, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "357270e5-9220-4c83-9197-a48b2dcef8a7",
   "metadata": {},
   "source": [
    "### Submit jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf72ade-c28d-479f-85f7-013dbdadc07c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno=fev1 # fvc\n",
    "workingD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003\n",
    "mkdir -p $workingD/pvalue_added/logs/\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    for rsq in {0.3,0.8}; do\n",
    "        /shared/bioinformatics/software/scripts/qsub_job.sh \\\n",
    "            --job_name ${pheno}_add_pvalue_chr${chr}_rsq${rsq} \\\n",
    "            --script_prefix $workingD/pvalue_added/logs/${pheno}_chr${chr}_rsq${rsq}_add_pvalue \\\n",
    "            --nslots 1 \\\n",
    "            --program  Rscript ~/bin/calculate_1dfpvalue.R \\\n",
    "                --in $workingD/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}.tsv.gz \\\n",
    "                --out $workingD/pvalue_added/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}.tsv \\\n",
    "                --col_beta Est.G \\\n",
    "                --col_se SE.G \\\n",
    "                --col_n n.obs\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ad3fda-db58-4cdd-adf2-fdc5d8d43582",
   "metadata": {},
   "source": [
    "## Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc44567-9a7e-411a-85e5-1e6220a54440",
   "metadata": {},
   "outputs": [],
   "source": [
    "## gzip first\n",
    "\n",
    "pheno=fev1 # fvc\n",
    "workingD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    for rsq in {0.3,0.8}; do\n",
    "        infile=$workingD/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}.tsv.gz\n",
    "        aws s3 cp $infile s3://rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/chr$chr/1df_pvalue_added/\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11fa6a11-a853-4e56-b836-c78688d8cb28",
   "metadata": {},
   "source": [
    "## QQ Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "226b66dc-8412-4673-8053-0d0623a9459d",
   "metadata": {},
   "source": [
    "### combine chromosome results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f054d57-5613-4eb7-8ff6-c14572d36da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno=fvc # and other pheno\n",
    "workingD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added\n",
    "mkdir -p $workingD/qqplot/wf_input/\n",
    "\n",
    "cd $workingD\n",
    "\n",
    "# get header\n",
    "zcat $workingD/chr1_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz | head -1 > \\\n",
    "    $workingD/qqplot/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv\n",
    "zcat chr1_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv.gz | head -1 > \\\n",
    "    $workingD/qqplot/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.8.tsv\n",
    "\n",
    "for rsq in {0.3,0.8}; do\n",
    "    for chr in {1..22}; do\n",
    "        infile=$workingD/chr${chr}_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}.tsv.gz\n",
    "        zcat $infile | tail -n +2 >> $workingD/qqplot/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}.tsv\n",
    "    done\n",
    "done &"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b3a1894-aecf-4a20-88ec-3698c414f17d",
   "metadata": {},
   "source": [
    "### prepare plotting workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f6eeccb-432a-47c5-82ed-9705d277c6d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno=fvc # and other pheno\n",
    "workingD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added/qqplot\n",
    "\n",
    "cd /shared/biocloud_wdl_tools/\n",
    "git pull\n",
    "#git submodule update --init --recursive\n",
    "cd /shared\n",
    "cp -r biocloud_wdl_tools/generate_gwas_plots/* $workingD/wf_input/\n",
    "cd $workingD/wf_input\n",
    "\n",
    "## edit workflow and configuration file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc1ed53-069f-42d3-99be-73bb6e6cdc9c",
   "metadata": {},
   "source": [
    "## Submit jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3abd0ea-ad00-4dba-80cd-06bafcb9f503",
   "metadata": {},
   "outputs": [],
   "source": [
    "## FVC\n",
    "java -jar ~/bin/cromwell/cromwell-54.jar \\\n",
    "    run /shared/rti-pulmonary/gwas/ukbiobank/results/fvc/0003/pvalue_added/qqplot/wf_input/test_generate_gwas_plots.wdl \\\n",
    "    --inputs /shared/rti-pulmonary/gwas/ukbiobank/results/fvc/0003/pvalue_added/qqplot/wf_input/fvc_rsq_0.8_input.json \n",
    "\n",
    "java -jar ~/bin/cromwell/cromwell-54.jar \\\n",
    "    run /shared/rti-pulmonary/gwas/ukbiobank/results/fvc/0003/pvalue_added/qqplot/wf_input/test_generate_gwas_plots.wdl \\\n",
    "    --inputs /shared/rti-pulmonary/gwas/ukbiobank/results/fvc/0003/pvalue_added/qqplot/wf_input/fvc_rsq_0.3_input.json \n",
    "\n",
    "## FEV1\n",
    "java -jar ~/bin/cromwell/cromwell-54.jar \\\n",
    "    run /shared/rti-pulmonary/gwas/ukbiobank/results/fev1/0003/pvalue_added/qqplot/wf_input/test_generate_gwas_plots.wdl \\\n",
    "    --inputs /shared/rti-pulmonary/gwas/ukbiobank/results/fev1/0003/pvalue_added/qqplot/wf_input/fev1_rsq_0.8_input.json \n",
    "\n",
    "java -jar ~/bin/cromwell/cromwell-54.jar \\\n",
    "    run /shared/rti-pulmonary/gwas/ukbiobank/results/fev1/0003/pvalue_added/qqplot/wf_input/test_generate_gwas_plots.wdl \\\n",
    "    --inputs /shared/rti-pulmonary/gwas/ukbiobank/results/fev1/0003/pvalue_added/qqplot/wf_input/fev1_rsq_0.3_input.json \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4effec55-738a-4083-9112-4257ab867808",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open session in terminal 1\n",
    "#ssh -i ~/.ssh/gwas_rsa -L localhost:8000:localhost:8000 ec2-user@54.174.185.7\n",
    "#\n",
    "## Submit jobs in terminal 2\n",
    "#\n",
    "## AFR relatedness\n",
    "#curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "#    -F \"workflowSource=@/shared/biocloud_gwas_workflows/genotype_array_qc/relatedness/relatedness_wf.wdl\" \\\n",
    "#    -F \"workflowInputs=@/shared/bioinformatics/methods/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0003/relatedness/uhs1234_afr_relatedness_wf.json\" \\\n",
    "#    -F \"workflowDependencies=@/shared/rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0003/wf_input/biocloud_gwas_workflows.zip\"\n",
    "#echo \"\"\n",
    "## Monitor job in terminal 1\n",
    "##tail -f /tmp/cromwell-server.log\n",
    "#\n",
    "## check job status in terminal 2\n",
    "#for job in {$afr_relate,$afr_sex,$eur_relate,$eur_sex}; do\n",
    "#    curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"   \n",
    "#    echo \"\"\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53f8131e-1e48-4e6c-99bd-668c1fec36e8",
   "metadata": {},
   "source": [
    "# LDSC Intercept"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d7f3ea9-d07c-420d-b5b1-b44099dbf9c0",
   "metadata": {},
   "source": [
    "## FVC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28748be1-68a5-4b84-bbb8-8ef6962f3cc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno=fvc\n",
    "workD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added/ldsc\n",
    "mkdir -p $workD/{sumstats,1df_main,1df_interaction,2df_joint}\n",
    "cd $workD\n",
    "\n",
    "# Download data\n",
    "wget https://data.broadinstitute.org/alkesgroup/LDSCORE/eur_w_ld_chr.tar.bz2\n",
    "wget https://data.broadinstitute.org/alkesgroup/LDSCORE/w_hm3.snplist.bz2\n",
    "\n",
    "tar -jxvf eur_w_ld_chr.tar.bz2\n",
    "bunzip2 w_hm3.snplist.bz2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "655cceab-0c7c-4384-b981-daa2ed17321d",
   "metadata": {},
   "source": [
    "### calculate joint 2df BETA\n",
    "Simply add the main effect beta and the interaction beta. See GitHub issue comment [here](https://github.com/RTIInternational/bioinformatics/issues/109/#issuecomment-846119728)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17bb279f-4f71-4ad1-97c5-cef486adf1a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-pulmonary/gwas/ukbiobank/results/fvc/0003/pvalue_added/qqplot/\n",
    "\n",
    "#variant.id      chr     pos     ref     alt     n.obs   freq    MAC     Est.G   Est.G.VITD      SE.G    SE.G.VITD       GxE.Stat        Joint.Stat      GxE.pval       Joint.pval      info    Main.pval\n",
    "#rs367896724:10177:A:AC  1       10177   AC      A       205475  0.397744        163453  0.00524484      -0.000179483    0.00674009      0.000123214     1.45668 2.0751  0.145206        0.116132        0.467935        0.43647807738792\n",
    "\n",
    "### python3\n",
    "import gzip\n",
    "\n",
    "infile = \"chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz\"\n",
    "outfile = \"../ldsc/sumstats/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3_joint_2df_beta_rsid_only.tsv\"\n",
    "\n",
    "\n",
    "with gzip.open(infile, 'rt') as inF, open(outfile, 'w') as outF:\n",
    "    head = inF.readline()\n",
    "    head = head.split()\n",
    "    head.append(\"Est.joint\")\n",
    "    head = \"\\t\".join(head) + \"\\n\"\n",
    "    outF.write(head)\n",
    "\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        rs = sl[0].split(\":\")[0] # only the rsID portion\n",
    "        sl[0] = rs\n",
    "        main = sl[8]\n",
    "        interaction = sl[9]\n",
    "        joint = float(main) + float(interaction)\n",
    "        sl.append(str(joint))\n",
    "        outline = \"\\t\".join(sl) + \"\\n\"\n",
    "        outF.write(outline)\n",
    "        line = inF.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b7b8990-e376-44ce-a908-fd6312542515",
   "metadata": {},
   "outputs": [],
   "source": [
    "## start interactive session\n",
    "\n",
    "docker run -i -t  \\\n",
    "    -v $workD:$workD \\\n",
    "    rticode/ldsc:7618f4943d8f31a37cbd207f867ba5742d03373f /bin/bash\n",
    "\n",
    "pheno=fvc \n",
    "workD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added/ldsc\n",
    "\n",
    "\n",
    "# Munge data\n",
    "#for rsq in {0.3,0.8}; do\n",
    "for rsq in 0.3; do\n",
    "    for df in {\"1df_main\",\"1df_interaction\",\"2df_joint\"}; do\n",
    "    case $df in \n",
    "\n",
    "        \"1df_main\") pval=\"Main.pval\"; sign=\"Est.G\" ;;\n",
    "        \"1df_interaction\") pval=\"GxE.pval\"; sign=\"Est.G.VITD\" ;;\n",
    "        \"2df_joint\") pval=\"Joint.pval\"; sign=\"Est.Joint\" ;;\n",
    "    esac\n",
    "\n",
    "    /opt/ldsc/munge_sumstats.py \\\n",
    "        --sumstats $workD/sumstats/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}_joint_2df_beta_rsid_only.tsv \\\n",
    "        --snp variant.id \\\n",
    "        --N-col n.obs \\\n",
    "        --a1 alt \\\n",
    "        --a2 ref \\\n",
    "        --p $pval \\\n",
    "        --signed-sumstats ${sign},0 \\\n",
    "        --out $workD/$df/${pheno}_${df} \\\n",
    "        --merge-alleles $workD/w_hm3.snplist\n",
    "\n",
    "    /opt/ldsc/ldsc.py \\\n",
    "        --h2 $workD/$df/${pheno}_${df}.sumstats.gz \\\n",
    "        --ref-ld-chr eur_w_ld_chr/ \\\n",
    "        --w-ld-chr eur_w_ld_chr/ \\\n",
    "        --out $workD/$df/${pheno}_${df}_h2\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa7e458-b930-4d9b-bd76-f75b5de4dd2b",
   "metadata": {},
   "source": [
    "## FEV1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2bc6d5-8f63-4898-ac1d-b9052b488fb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno=fev1\n",
    "workD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added/ldsc\n",
    "mkdir -p $workD/{sumstats,1df_main,1df_interaction,2df_joint}\n",
    "cd $workD\n",
    "\n",
    "# Download data\n",
    "wget https://data.broadinstitute.org/alkesgroup/LDSCORE/eur_w_ld_chr.tar.bz2\n",
    "wget https://data.broadinstitute.org/alkesgroup/LDSCORE/w_hm3.snplist.bz2\n",
    "\n",
    "tar -jxvf eur_w_ld_chr.tar.bz2\n",
    "bunzip2 w_hm3.snplist.bz2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "885d7f6f-ac36-4ea7-b4d3-08b60bf71f5a",
   "metadata": {},
   "source": [
    "### calculate joint 2df BETA\n",
    "Simply add the main effect beta and the interaction beta. See GitHub issue comment [here](https://github.com/RTIInternational/bioinformatics/issues/109/#issuecomment-846119728)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a25957d-ba50-46af-ab0f-0fb8a94376a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added/qqplot/\n",
    "\n",
    "#variant.id      chr     pos     ref     alt     n.obs   freq    MAC     Est.G   Est.G.VITD      SE.G    SE.G.VITD       GxE.Stat        Joint.Stat      GxE.pval       Joint.pval      info    Main.pval\n",
    "#rs367896724:10177:A:AC  1       10177   AC      A       205475  0.397744        163453  0.00524484      -0.000179483    0.00674009      0.000123214     1.45668 2.0751  0.145206        0.116132        0.467935        0.43647807738792\n",
    "\n",
    "### python3\n",
    "import gzip\n",
    "\n",
    "infile = \"chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3.tsv.gz\"\n",
    "outfile = \"../ldsc/sumstats/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_0.3_joint_2df_beta_rsid_only.tsv\"\n",
    "\n",
    "\n",
    "with gzip.open(infile, 'rt') as inF, open(outfile, 'w') as outF:\n",
    "    head = inF.readline()\n",
    "    head = head.split()\n",
    "    head.append(\"Est.joint\")\n",
    "    head = \"\\t\".join(head) + \"\\n\"\n",
    "    outF.write(head)\n",
    "\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        rs = sl[0].split(\":\")[0] # only the rsID portion\n",
    "        sl[0] = rs\n",
    "        main = sl[8]\n",
    "        interaction = sl[9]\n",
    "        joint = float(main) + float(interaction)\n",
    "        sl.append(str(joint))\n",
    "        outline = \"\\t\".join(sl) + \"\\n\"\n",
    "        outF.write(outline)\n",
    "        line = inF.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4488e453-3b02-41c2-92e7-e6d59413819b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## start interactive session\n",
    "\n",
    "docker run -i -t  \\\n",
    "    -v $workD:$workD \\\n",
    "    rticode/ldsc:7618f4943d8f31a37cbd207f867ba5742d03373f /bin/bash\n",
    "\n",
    "pheno=fev1\n",
    "workD=/shared/rti-pulmonary/gwas/ukbiobank/results/$pheno/0003/pvalue_added/ldsc\n",
    "\n",
    "\n",
    "# Munge data\n",
    "#for rsq in {0.3,0.8}; do\n",
    "for rsq in 0.3; do\n",
    "    for df in {\"1df_main\",\"1df_interaction\",\"2df_joint\"}; do\n",
    "    case $df in \n",
    "\n",
    "        \"1df_main\") pval=\"Main.pval\"; sign=\"Est.G\" ;;\n",
    "        \"1df_interaction\") pval=\"GxE.pval\"; sign=\"Est.G.VITD\" ;;\n",
    "        \"2df_joint\") pval=\"Joint.pval\"; sign=\"Est.Joint\" ;;\n",
    "    esac\n",
    "\n",
    "    /opt/ldsc/munge_sumstats.py \\\n",
    "        --sumstats $workD/sumstats/chr_all_2dfgrch37_dbsnp_b153_maf_gt_0.01_rsq_gt_${rsq}_joint_2df_beta_rsid_only.tsv \\\n",
    "        --snp variant.id \\\n",
    "        --N-col n.obs \\\n",
    "        --a1 alt \\\n",
    "        --a2 ref \\\n",
    "        --p $pval \\\n",
    "        --signed-sumstats ${sign},0 \\\n",
    "        --out $workD/$df/${pheno}_${df} \\\n",
    "        --merge-alleles $workD/w_hm3.snplist\n",
    "\n",
    "    /opt/ldsc/ldsc.py \\\n",
    "        --h2 $workD/$df/${pheno}_${df}.sumstats.gz \\\n",
    "        --ref-ld-chr eur_w_ld_chr/ \\\n",
    "        --w-ld-chr eur_w_ld_chr/ \\\n",
    "        --out $workD/$df/${pheno}_${df}_h2\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14b92363-c703-4aa0-9b5c-9af738166f7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cf6c83f-9dae-4dc7-88b1-a782728a7bcd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19eeeecd-b29c-4c8d-863f-878f7e061ac6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "055e0030-2b3e-42c5-9b50-9b7a63d3f060",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
