{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MVP Partitioned Heritability Estimate\n",
    "Author Jesse Marks\n",
    "\n",
    "Olivia Corradin wanted us to use LDSC to compute the partitioned heritability estimate for the MVP opioid use disorder GWAS results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Olivias Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## activate ldsc environment \n",
    "## Files listed as ldsc/ directory are from the ldsc wiki/tutorial \n",
    "\n",
    "$name=analysis_identifier\n",
    "\n",
    "for j in {1..22}\n",
    "do\n",
    "python make_annot.py --bed-file \"$peaks\" --bimfile ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.\"$j\".bim --annot-file \"$name\"_\"$peaks\".\"$j\".annot.gz\n",
    "python ldsc.py --l2 --bfile ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.\"$j\" --ld-wind-cm 1 --annot \"$name\"_\"$peaks\".\"$j\".annot.gz --thin-annot --out \"$name\"_\"$peaks\".\"$j\" --print-snps ldsc/1000G_EUR_Phase3_baseline/print_snps.txt  \n",
    "done\n",
    "\n",
    "mkdir LDSC_\"$name\"\n",
    "mv \"$name\"* LDSC_\"$name\"/\n",
    "\n",
    "for d in disease_sumstats\n",
    "do\n",
    "python ldsc.py --h2 \"$d\".sumstats --w-ld-chr ldsc/weights_hm3_no_hla/weights. --ref-ld-chr LDSC_\"$name\"/\"$name\"_\"$peak\".,/lab/corradin_data/ldsc/1000G_EUR_Phase3_baseline/baseline. --overlap-annot --out OUT_\"$name\"_\"$d\"_\"$peak\" --print-coefficients --frqfile-chr /lab/corradin_data/ldsc/1000G_Phase3_frq/1000G.EUR.QC.\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Our Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Wrangle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/mvp/0001/\n",
    "\n",
    "aws s3 sync s3://rti-common/dbGaP/phs001672_mvp/RootStudyConsentSet_phs001672.MVP.v4.p1.c1.HMB-MDS/AnalysisFiles/ .\n",
    "for file in *tar.gz; do\n",
    "    tar -xvzf $file\n",
    "done\n",
    "\n",
    "cp /shared/rti-heroin/ldsc_genetic_correlation/data/mvp/0001/release/Submissions/sub20200313/OUD_EA_MVP1_MVP2_Mar12.txt.gz .\n",
    "cp /shared/rti-heroin/ldsc_genetic_correlation/data/mvp/0001/release/Submissions/sub20200313/OUD_EA_MVP1_MVP2_YP_SAGE_Mar12.txt.gz .\n",
    "\n",
    "rm -rf release/ *tar.gz\n",
    "\n",
    "zcat OUD_EA_MVP1_MVP2_YP_SAGE_Mar12.txt.gz  | head\n",
    "#SNP_ID  Chromosome      Position        Allele1 Allele2 EA      EAF     Effective_N     Original_N      Zscore  PValue  Direction\n",
    "#rs3115860       1       753405  A       C       A       0.8592  30228.7 78187   -0.941  0.3466  --???+\n",
    "#rs3131970       1       753425  T       C       T       0.1412  29680   77702   0.862   0.3888  ++????\n",
    "#rs3115858       1       755890  A       T       A       0.1382  30974.6 80209   0.821   0.4118  ++???+\n",
    "#rs3131962       1       756604  A       G       A       0.1386  30969.5 80219   0.808   0.4189  ++???+\n",
    "#rs4951929       1       757734  T       C       T       0.8617  30977.7 80201   -0.818  0.4135  --???-\n",
    "\n",
    "zcat OUD_EA_MVP1_MVP2_Mar12.txt.gz  | head\n",
    "#SNP_ID  Chromosome      Position        Allele1 Allele2 EA      EAF     SampleSize      N_case  N_control       Effect  SE      PValue  Direction\n",
    "#rs3115860       1       753405  A       C       A       0.8592  77622   8310    69312   -0.0239 0.0242  0.3241  --\n",
    "#rs3131970       1       753425  T       C       T       0.1412  77702   8314    69388   0.021   0.0242  0.3862  ++\n",
    "#rs3115858       1       755890  A       T       A       0.1382  79649   8519    71130   0.0199  0.0241  0.4083  ++\n",
    "#rs3131962       1       756604  A       G       A       0.1384  79668   8520    71148   0.0192  0.0241  0.4257  ++\n",
    "#rs4951929       1       757734  T       C       T       0.8617  79634   8518    71116   -0.0199 0.0241  0.409   --\n",
    "\n",
    "gunzip *z\n",
    "\n",
    "# format file1\n",
    "file=OUD_EA_MVP1_MVP2_YP_SAGE_Mar12\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$file.txt \\\n",
    "        --out /data/$file.csv \\\n",
    "        --force \\\n",
    "        --snp SNP_ID \\\n",
    "        --a1 Allele1 \\\n",
    "        --a2 Allele2 \\\n",
    "        --z Zscore \\\n",
    "        --n Original_N \\\n",
    "        --auto \\\n",
    "        --head 5 \n",
    "\n",
    "# format file2\n",
    "file=OUD_EA_MVP1_MVP2_Mar12\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$file.txt \\\n",
    "        --out /data/$file.csv \\\n",
    "        --force \\\n",
    "        --snp SNP_ID \\\n",
    "        --a1 Allele1 \\\n",
    "        --a2 Allele2 \\\n",
    "        --beta Effect \\\n",
    "        --n SampleSize \\\n",
    "        --auto \\\n",
    "        --head 5 \n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "#awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" OUD_EA_MVP1_MVP2_YP_SAGE_Mar12.csv \\\n",
    "#    > mvp_oud_eur_p1p2_yp_sage_ldsc_ready.txt\n",
    "#awk '{print $1, $2, $3, $5,$6,$10,$4,$7}' OFS=\"\\t\" OUD_EA_MVP1_MVP2_Mar12.csv \\\n",
    "#    > mvp_oud_eur_p1p2_ldsc_ready.txt\n",
    "gzip *ldsc_ready*\n",
    "\n",
    "## upload to S3\n",
    "#aws s3 cp mvp_oud_eur_p1p2_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/mvp_oud/\n",
    "#aws s3 cp mvp_oud_eur_p1p2_yp_sage_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/mvp_oud/\n",
    "\n",
    "# munge sumstats\n",
    "wget https://data.broadinstitute.org/alkesgroup/LDSCORE/w_hm3.snplist.bz2\n",
    "bunzip2 w_hm3.snplist.bz2\n",
    "    \n",
    "/opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats /data/mvp/OUD_EA_MVP1_MVP2_Mar12.csv \\\n",
    "    --signed-sumstats BETA,0 \\\n",
    "    --N-col N \\\n",
    "    --out /data/mvp/mvp_oud_eur_p1p2 \\\n",
    "    --merge-alleles /data/w_hm3.snplist\n",
    "\n",
    "/opt/ldsc/munge_sumstats.py \\\n",
    "    --sumstats /data/mvp/OUD_EA_MVP1_MVP2_YP_SAGE_Mar12.csv \\\n",
    "    --signed-sumstats Z,0 \\\n",
    "    --N-col N \\\n",
    "    --out /data/mvp/mvp_oud_eur_p1p2_yp_sage \\\n",
    "    --merge-alleles /data/w_hm3.snplist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# upload peaks files\n",
    "cd /home/jesse/Projects/heroin/ldsc/mvp # local\n",
    "scp -i ~/.ssh/gwas_rsa * ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/data/mvp/0001/peaks/\n",
    "#GVEL_ALL_FDR05\n",
    "#LVEL_ALL_FDR05\n",
    "#linear_regression_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/mvp/0001/ldsc/\n",
    "\n",
    "# download ldsc files\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_baseline_ldscores.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_plinkfiles.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/1000G_Phase3_frq.tgz\n",
    "wget https://storage.googleapis.com/broad-alkesgroup-public/LDSCORE/weights_hm3_no_hla.tgz\n",
    "\n",
    "# decompress\n",
    "for file in *tgz; do\n",
    "    tar xvzf $file\n",
    "    rm $file \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LDSC\n",
    "1. Create annotation file based off of a BED formatted file (file containing chr, chr-start, chr-end)\n",
    "2. Compute the annotation-specific (partitioned) LD scores.\n",
    "3. Compute the partitioned heritability estimate. \n",
    "\n",
    "MVP_p1p2_YP_SAGE: N=82,707 (10,544 cases and 72,163 controls) <br>\n",
    "MVP_p1p2: 79,729 (8,529 cases and 71,200 controls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# interactive docker container for LDSC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "  rticode/ldsc:7618f4943d8f31a37cbd207f867ba5742d03373f bash \n",
    "    \n",
    "# loop through both MVP metas\n",
    "for name in {mvp_p1p2_yp_sage,mvp_p1p2}; do\n",
    "\n",
    "    # store processing files for each meta in separate dir\n",
    "    mkdir /data/LDSC_${name}/\n",
    "    \n",
    "    # use sumstats file that corresponds to the meta name for the h2 estimate\n",
    "    if [[ $name == \"mvp_p1p2_yp_sage\" ]]; then\n",
    "        stats=mvp_oud_eur_p1p2_yp_sage.sumstats.gz\n",
    "    else\n",
    "        stats=mvp_oud_eur_p1p2.sumstats.gz\n",
    "    fi\n",
    "    \n",
    "    # loop through each BED file\n",
    "    for peaks in {GVEL_ALL_FDR05,LVEL_ALL_FDR05,linear_regression_peaks}; do\n",
    "        # loop through each chromosome\n",
    "        for j in {1..22}; do\n",
    "        \n",
    "            # create annotation files\n",
    "            python /opt/ldsc/make_annot.py \\\n",
    "                --bed-file \"/data/peaks/$peaks\" \\\n",
    "                --bimfile \"/data/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim\" \\\n",
    "                --annot-file \"/data/LDSC_${name}/${name}_${peaks}.$j.annot.gz\"\n",
    "\n",
    "            # compute LD scores\n",
    "            python /opt/ldsc/ldsc.py \\\n",
    "                --l2 \\\n",
    "                --bfile \"/data/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j\" \\\n",
    "                --ld-wind-cm 1 \\\n",
    "                --annot \"/data/LDSC_${name}/${name}_${peaks}.$j.annot.gz\" \\\n",
    "                --thin-annot \\\n",
    "                --out \"/data/LDSC_${name}/${name}_${peaks}.$j\" \\\n",
    "                --print-snps \"/data/ldsc/1000G_EUR_Phase3_baseline/print_snps.txt\"\n",
    "        done # end chr loop\n",
    "        \n",
    "    # computed partitioned heritability estimate\n",
    "    python /opt/ldsc/ldsc.py \\\n",
    "        --h2 \"/data/mvp/$stats\" \\\n",
    "        --w-ld-chr \"/data/ldsc/weights_hm3_no_hla/weights.\" \\\n",
    "        --ref-ld-chr \"/data/LDSC_${name}/${name}_${peaks}.,/data/ldsc/1000G_EUR_Phase3_baseline/baseline.\" \\\n",
    "        --overlap-annot \\\n",
    "        --out \"/data/results/OUT_${name}_${peaks}\" \\\n",
    "        --print-coefficients \\\n",
    "        --frqfile-chr \"/data/ldsc/1000G_Phase3_frq/1000G.EUR.QC.\"\n",
    "    \n",
    "    done # end BED file loop\n",
    "done # end meta file loop\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 002\n",
    "New Request by Olivia to run this code on one more list of peaks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/mvp/0002/ # EC2\n",
    "\n",
    "# interactive docker container for LDSC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "  rticode/ldsc:7618f4943d8f31a37cbd207f867ba5742d03373f bash \n",
    "    \n",
    "# loop through both MVP metas\n",
    "for name in {mvp_p1p2_yp_sage,mvp_p1p2}; do\n",
    "\n",
    "    # store processing files for each meta in separate dir\n",
    "    mkdir /data/LDSC_${name}/\n",
    "    \n",
    "    # use sumstats file that corresponds to the meta name for the h2 estimate\n",
    "    if [[ $name == \"mvp_p1p2_yp_sage\" ]]; then\n",
    "        stats=mvp_oud_eur_p1p2_yp_sage.sumstats.gz\n",
    "    else\n",
    "        stats=mvp_oud_eur_p1p2.sumstats.gz\n",
    "    fi\n",
    "    \n",
    "    # loop through each BED file (peaks)\n",
    "    for peaks in {GVEL_ALL_FDR05,LVEL_ALL_FDR05,linear_regression_peaks,FDR05_LVEL_targeting_shared_genes_gr1}; do\n",
    "        # loop through each chromosome\n",
    "        for j in {1..22}; do\n",
    "        \n",
    "            # create annotation files\n",
    "            python /opt/ldsc/make_annot.py \\\n",
    "                --bed-file \"/data/peaks/$peaks\" \\\n",
    "                --bimfile \"/data/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim\" \\\n",
    "                --annot-file \"/data/LDSC_${name}/${name}_${peaks}.$j.annot.gz\"\n",
    "\n",
    "            # compute LD scores\n",
    "            python /opt/ldsc/ldsc.py \\\n",
    "                --l2 \\\n",
    "                --bfile \"/data/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j\" \\\n",
    "                --ld-wind-cm 1 \\\n",
    "                --annot \"/data/LDSC_${name}/${name}_${peaks}.$j.annot.gz\" \\\n",
    "                --thin-annot \\\n",
    "                --out \"/data/LDSC_${name}/${name}_${peaks}.$j\" \\\n",
    "                --print-snps \"/data/ldsc/1000G_EUR_Phase3_baseline/print_snps.txt\"\n",
    "        done # end chr loop\n",
    "        \n",
    "    # computed partitioned heritability estimate\n",
    "    python /opt/ldsc/ldsc.py \\\n",
    "        --h2 \"/data/mvp/$stats\" \\\n",
    "        --w-ld-chr \"/data/ldsc/weights_hm3_no_hla/weights.\" \\\n",
    "        --ref-ld-chr \"/data/LDSC_${name}/${name}_${peaks}.,/data/ldsc/1000G_EUR_Phase3_baseline/baseline.\" \\\n",
    "        --overlap-annot \\\n",
    "        --out \"/data/results/OUT_${name}_${peaks}\" \\\n",
    "        --print-coefficients \\\n",
    "        --frqfile-chr \"/data/ldsc/1000G_Phase3_frq/1000G.EUR.QC.\"\n",
    "    \n",
    "    done # end BED file loop\n",
    "done # end meta file loop\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sandbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks=GVEL_ALL_FDR05.bed\n",
    "j=22\n",
    "\n",
    "\n",
    "python /opt/ldsc/make_annot.py \\\n",
    "    --bed-file \"/data/peaks/$peaks\" \\\n",
    "    --bimfile \"/data/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j.bim\" \\\n",
    "    --annot-file \"/data/t0/${name}_${peaks}.$j.annot.gz\"\n",
    "\n",
    "    \n",
    "python /opt/ldsc/ldsc.py \\\n",
    "    --l2 \\\n",
    "    --bfile \"/data/ldsc/1000G_EUR_Phase3_plink/1000G.EUR.QC.$j\" \\\n",
    "    --ld-wind-cm 1 \\\n",
    "    --annot \"/data/t0/${name}_${peaks}.$j.annot.gz\" \\\n",
    "    --thin-annot \\\n",
    "    --out \"/data/t1/${name}_${peaks}.$j\" \\\n",
    "    --print-snps \"/data/ldsc/1000G_EUR_Phase3_baseline/print_snps.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"/data/LDSC_${name}/${name}_${peaks}.,/data/ldsc/1000G_EUR_Phase3_baseline/baseline.\"\n",
    "\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "  rticode/ldsc:7618f4943d8f31a37cbd207f867ba5742d03373f bash \n",
    "\n",
    "GVEL_ALL_FDR05,LVEL_ALL_FDR05,linear_regression_peaks\n",
    "name=mvp_p1p1\n",
    "stats=mvp_oud_eur_p1p2.sumstats.gz\n",
    "#ref_string=/data/ldsc/1000G_EUR_Phase3_baseline/baseline.,/data/LDSC_mvp_p1p1/mvp_p1p1_GVEL_ALL_FDR05.,/data/LDSC_mvp_p1p1/mvp_p1p1_LVEL_ALL_FDR05.\n",
    "ref_string=/data/ldsc/1000G_EUR_Phase3_baseline/baseline.,/data/t1/mvp_p1p1_GVEL_ALL_FDR05.\n",
    "\n",
    "# computed partitioned heritability estimate\n",
    "python /opt/ldsc/ldsc.py \\\n",
    "    --h2 \"/data/mvp/$stats\" \\\n",
    "    --w-ld-chr \"/data/ldsc/weights_hm3_no_hla/weights.\" \\\n",
    "    --ref-ld-chr $ref_string \\\n",
    "    --overlap-annot \\\n",
    "    --out \"/data/results/OUT_${name}_${peaks}\" \\\n",
    "    --print-coefficients \\\n",
    "    --frqfile-chr \"/data/ldsc/1000G_Phase3_frq/1000G.EUR.QC.\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
