{
 "cells": [
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAA48AAAAlCAYAAADiBtwaAAAb/klEQVR4Ae1dPc4zxw1WlcKNm3Tuco29gvvUOUCABXKHFDGQC6h1Gh9C8CUCpPgO4FtsQHI48/BnV6vfT9JLG7Z254d8+JDDGWolvYel/ikGioFi4M0YOCyP/ffN6Ci4xUAxUAwUA8VAMVAMPIWBw1O0lJJioBgoBu7IwGNLx0qLd3RViSoGioFioBgoBoqBD2Lgu56Svh2n5TAdl2+G0G/LcTos09G2miEfepPz8aHGfk+zvh2X6TAt7xtip2U+HJb59HgSXy8mxfbnFY/P4/rx3ryXhtfJ0af5sByesRDuRV3JKQaKgWKgGCgG3pyB71g8rhxAHnCw5wPwyx8wVvh4oQB7xkHtGTqeUxA9sOg4zcvhMC+Prx1fMCab7U8rHp/G9Qst9HNQHpCjvUqTs1lfFu8PXGMe0BvfPyffWYLupdPEgVVx37vVGLuvmsWsHcmvjzma0Np4lTdIX3SdPsvndw6hZ4ujM9nbPMxhnx6Ww+GNMD/boXfSd5fikQ/8B3HYYW/CWlm45zeLmBR5zsZh+rzMO7HJYmRDuPjd8BU+EFm0Q5LyvRb2OR6fUdjt1xHjALlav963Yd+eMHHDvDImVoy4HduKYN+8Iyb9lL3319qg80bx+Kflt7//Zfnj55+W3/43WvvV6c/LHz//Zfnjlx/MtyT/9QvNsf/9fuqzuhmqrze820XbTO95QI156P6kGB1rcUiFffjkymVY3t6/O8xlLm/kaYcaM+ReOk0cGA13vlmLsTur+czi8dz+hnvhvQm9Qd6zfH4DxFeYenWObHvPrTn6Eg4I68Vn71TBtefLVNhHNt5ePNIGDicTdt6OjSrfFPYc7C93aq7rtfy5B+OeMY+06n4Lcx3lfh2XxwFr3blhEI7bivJHbZhX2r1O+WrPI+PtOn6H7VzmcWH40/Lb6cflv6F4bEXlLz8sXChmxaNrG6WjpsWhb5WkL9exJ0ffToqJvZU1e10MWWwXy7hDwWoR3HC3wouXyFzu2JP9vFvuL9eZx5WJg1sAnZu7g8u7YGE9+kQwtzlCvSYPXTNnaL54XYypyRVhudPXLHb4KQGQN91TVq7hI1qviQWaQ0XjKf1q2uNouQZrjua29ZPL/KxWPSXtsoqTZ3/CmH2MaFkWt7nKHE2WqmYlafrF3ApTDkR6DP2Pv/HjaHokLf81DE6natFXTfoqh+ZCvUugObl1udBJc6fjib+HaeaRzo7DFhomgJsNyF0sSjwfhGdeTqpj+ufyz0ltxkfyfp5arK9NDvPa5vMhAu0FP3oeVT/ZOR2Xo3tXB/m07y4Jrs4P8KnI+us1OnAO+2DYsI5JNGospPr5qTlyI5wxfObQxjHrAttQ93Q8mg2T+tDvGA/+aT3Ksbzm64vk4hyAJOsR4pR9ogMwLsInBlxs0ViKHeSe5KAMd0DNbVzhdyFRE6yp4VPjq6YDC73D/7LicYy4ungkW8Emwue5Hj4lvqbleKLv01LcDPzoG+9rw5/xgVtDgMPOgVyGvumyiO8Rt4RlPqHs0cc8GxktZ2i80AD297AtmzM4oV7rb9uHOOw7xhwLqjfTyXK3sANGY5POsbgwtxt/Ie9sjo0Jtp/+h+sA+F4WZyPEhUzDmBe+laOz6yHlxWOZl6M/wBk+IH7YGMuLuoC6DC/Gxlt1MhO8x6K+zk9vFGzCD/GarTeLH9/YjjZAjDgu2VbnexOTJEz/Mb63cez9Px/x+/YSG/MJMWt8qnB6pX7fLnM6NewC/A2JNqfnI9nDxy9JoE67N7FGt19JfDsMHEfCIfGlcavzdf+3e6HYjGP5zAjrYjPunZ+6LiSirTnVsSovyLKc+NghXeMfGjsvJ+N7iCf1R997kbtkLsYa8TqfzH6otgz9cGUwWD9s51+QwZfWfqVUfDvOwHYPa7ZoTkE7lAPXFrSu7EmYbxSLwMSzt/JqsVuMXuO4Rx0ar/O/Ub7dy8fMutpfPIaFlpEXE4MsXHVwm7MiKyTnFpB24VCQOHk0biNANXloAMq9LnSPWYLQjnUbLOMCDG3x6hyTSL0Nfi5REvhoC8HZFPhpSVL1Ro94Oe2+8ye29yRpeLQ8SIIfGyNhGX6xHEacEZm0XK8j21C3MIk+wTn4Iv3gR4BpfEjt7CM7lhOPCmO/akyNg9bohqTuxjJf6mvjAwDULj0ungtvhsi94rD2cZ8Caof6fkv29Ru1V+Xo/fD/OCzrGPFlj4ktG9tBtI8l22g86DfcJraP0vCwXF084sdW//7j8lf4YKtAAp8xBXLQ7zBNTLS1BIcglYG5KfoHcgv4wNqP68vHcCMnPWAyAhPjLBdinu819pwMGy+iJ7Sxn2FdMCdqE2H1feOedHcuXTwaPSxT46zZ69eJ8YXEE8cXtXf7aBnjAVvWaYhDGK/s9levlztcHiOdbFiLiWGkFGAq39lFfHQsO9ZD3DcIDMYKrFujc/jAHtzdXPJJm2fjpPHYY93NY7ugWOH7NZ3KrMgAqrhjxIHLL70ox7hwOLKYUh5aDuo5B3zhbe0IKXYiQH6zaDRbnDbeBN842Or9sCHXTTKBPwbkYk7zU7dP+kfusbhMrKmB8Br7Pbdtj2uGm/GcE5xNfY+KckwMgh8AzrjM+p0+s1dvrSMjy+OK/A4QdOX5tXwYmy70jcx1uSD4X9E4nGyTxgrZpNeaC+BeRfBrYj+ufdBvYzTygGJt/GPPuGZ5Xj58Z5FlaK4h+3qM78jnQ028otjossR+u47XuIqivmLLhcUjbvaDLnF+8i71GGKuePzwUutzi4BajXNVBI1zTk3H6fgWYKgPF5hJIDJnbFjJ3Lbp9E2+qUGb6Lr3B2zRTpwr4hIbNQGhHbuKR8uVwaYydQEhVrxObGxN/QVtMIu9j0gubtCRFY9eA2LiPtKnSUga+Olg9xUI8DyZDamNQ/lhvDu4YD9eiyjwN2O0PuuwMG5bI8YqN+EYH9vIN153BeMCbRtyx4FAD6kYjjhn08aVNTS0x03B839r8WjmLz8sv7vvRXp9hC1wbQ7qfvOhGeDXbhys/1UfJPP6WNETDrEt3mIsW1nBLyRX18RWvDB+wN7sCfKabyM+5SPfQzSeFL/h2uMyvAsQM75hS1/Q3iwOud+vP7Fb35nGV8bbfeM0BtzuAOfmbdnAfZynm//7E42279I9LcZE55i7HsO8jpO5YpGNodHWfJnM26uTx2W2tJjk/vmY/AJ7st7WcIhx8ckmjm/X4SkttwPHgFXojgXlsH0No8bXdj/nU9DX447jIK7FoZc8FH3G8piLVujovq9hy7Gf2Sp4o/yxljEP4HWIlWTtmkKr8d1gNmRnYt7ttRIz+c/JGRsS/+OsLTkZv2hHzoH6fds38dwbfa0ui2PFt5pH+zi+2JCDXNhJ/IaXkcdxomeBaAtON3xjB1wHrlxezPbiPt1g2bK9z+gX3r8Wx7ZdXcgXvthfPBJJkEhNMAGB7BBf3EG/LLqRdHqXCwJu90EkjZvFIwWATbLZhgmBwXphjibrlr18gPlDjuLHcSYIgw1+w8gWNOBTBRuHV4YKviH7BX6UY7CpTN1EACvbo+0NA3Pbs3qS0Htf45y5bAkzwXebjmib+qb7n/QDJotfiRX+eY4ba2Kc8Wvyl7lDnnBhxrsNbfCe8IY8kWiISYAUnprQUIy7hgrWh+Vo4G3znH9lPgMwH7nldrZfNwxq8XHcDiQM+LyNgw/QimuXOAF8PlZM8XfFx1bNfHriSN+fhKePXh9bnDx5GHZEPvJNb4zLdAyuk5zU+QB+exvNzGI5xoCJU+PXOBbXDx6OxGuCw8jzMenW/chNDi+vgfGmm4lrg5EVhD+1g7Et2Mb/WVaTL7lhxPHw3xi/tv54BORInbHqR17HQ5eMB46NXZFLtglxG18rD05+ohPxBZlNPi/bZC5jZpw295n1n8zbrVNJTPIJdaHvMBca/SqDcSTrRoyzX03pvDb+ME6tIpWe5FrMeX2Y5G/2laxHI85wOXLBmA3x0RvX2nSfl4HIeVbccD+AGbHgYih7U4VUIHbiGuJxrKMYx5qXRHXS7+MHfOFzi2CIeIdtkdNhZ4sNxc16mqyt2AE8I38lPunyBIPkGoxHXUNx7sDf9n/FyK6NNmlo8Lwey0NX581hH/hVQnv1PoDu4dvW2O2k+2gLTE3PLNjPEvBhizSY2DJx53LCrnzuFeq9iWGyA9fTtl0q4iu/XlY8KlMtICEPaU8/wOR92cKQqSFAqdk4V1UkTk3H6fgzB2yzEMYcvTKLujVmWDlBNaNNf8DmEkHoJyWJjbpoDLFOloLur1GOwaYyNVEhFrwGu/UwifYyYvd9SIUgyS0me+6/SUdum+KLmOJ4xSiv1D8Or54nn8C8/DB+tXiUA0dP7haEvSN++hsxua9jfIKdYeMAPyTcd+VZX1gnEQ/GROSjS+cL3892aBz6uEwOlqb4u0Px+Nf//GSKR/pbs2apKSbTKByILyMf+TqWOGMxGc/CDrwBYHmzd6ITY176MZYhHjiljhjnsejXrXjJ5m60CSawlZXhPXJHnfbexDViTP2Q5XhhIhS8vKbGOvBx2GbJC4/VAx/Igzjl1jU/Otw8ltuaTJY/Dnx4GN9eDw3LqvxhH41EWYbXJqa/ZPK408aQjAdfJvN26xzK45NBxU6LJfgiWW8Jji4+ySGjjxUtEz/tFLuy/Jxxt9rGMbKGUWPqXL8iPMN/G4acZ7knw0pTOW+7mF5bF9qurx0hFAC+T7BoPrXrnOezb23Mcjv7U+c1TWs+1nZ97cPtx9QNRzgWr9vc7ZfEJ2BH5AClxbnGNyTH+COJExUXxmoHvUosjy3L38PYDfuDLWZstAWkmtyD7Xgd5HubWB/mTIgV4JxkBlmoyF0z51h4D6Iad7pO3cS6ZQauKx6huCEHWM7xwNs2rn4AXnPuSgD6IGLIyQJIxw0Pm4XZZWhgiLxss6Chca7aBAHMwT2SnAnggM0mAjO2Q875YCx7k0qwU4R7fUYmYnU29Xfj2dl+AxAO4yGWV7NLhN1IItJ+X4T002LepcPHwRlMaBtAGJfRL9YeFycGa4wJ4hnf5UPemXP9qOAAEK+YnxZneA0jY3xC7BBGszjNRMs9yW9jEWufEfRbvmgc29xknLMRx/q5Wkj0DTToXuDbidd85/GH5Xd4ynhoH1v973/+JHK/HdpBslvPF2LTWOf2QBv56HaADwwvbBfI6z4QWau+A1ipr8xBGeIh21yR2614gZwP6iU/YiyjTXytebblgv7urlu/bZ7mYRPXiNHYBkh4PVpdJIvlQM7kuAO8Pg5BYstPkOepk/SAPB6PNlMD3bPPXc7QNdLmEza11+jVcT1uWjykeh2+dmDschu2jtnzZBR7vPqdxxiPJo5v0qkAsvXT8mqaU7LxHr/Kllf2tedQh2CMeX+2MSYmdV7g02KwOhuP/Tyk98OHdnxXkryZ1OZqjHg/J+s1xb9yxiEcPYYUBr1y/E/pk38db2Oj7Qt93bf77gfhq390HnUlNsiZYfA1hiuXen6QHuZTOWq5o68F9LmP4SF45arh7nZYf3gOrBCaC7nK+yDkGJGtZpgYWYlV1sd9oIdjdew3Rk6wX9e++Et9O+SqD6ItaCvz0DmSHs9NiDVvP9jh5bENLp8brAjGXdNc5dR1UaC7J5FxxFdv2V88tsCTx8SYWNoi6hU8BKsuir5QVgLNB4t6ZaVdgo8O5y2AV8apGB5vosTj8DYM++JckTow2CKBes1iCNgwEXgcinirXfTJAkFZOhdfoxyDTf2ji9tjpXv1q/76l/LYkpb0T8s8j+9+yIIWnOPJGeKC6yt1kIThgxYHZzCFpGLG201HC1ssAHsBTZzoz1ArH3rga3zpL1pqd8q7csuvYsOwycbVZhyqEqbV+tz6wn4U1Nhzbo0yV7phMPvhSQHrAizelr5exXntV0k1gRNujZnDMs/jkJ7Zzk8e9W844g/f9D/ZoX//0f4tR/17j/7vPPbCkcrHo4sF5lUPsvL9K10Xw9y1tSjtOt5wQHIx/rsPBr9jnmKyPNmD0OBvFJ4xHsw6cH5djRefGxonjJQ+zou+gy8OmRiY5mXGJ7rG9nmZ4cBqfI4Y8RowRBwaq5b/aZ7tGwMsb6w1gxcOvE5VvDW24B644i+WYLExhz2g7DxcD1G5awGbOD687wxWsh3w4lxjv8eq/DbdOI/2k0t0Ovh4a+JAcyzvV4Kn09UnWd6IU4z31fhm/GATcwS8dPnJheMT9fU3kHh9kDxcj2QD/mIs+QIwJKpME3JO8wznqEdmDS7P+NIo8TeNXz0ztG7iFe1Gnv1eqAdzyRlqf7N7k0uPxd7r2rUxYePBrCPv81YwYC5Dm6y2xu8RzkiOE8Uz5Klvt3zT9gQjy8Y6c4v9Jg5GLiO8BoPLv+fkKI/et7aAj7aQXvR/t79hFkzKhTsvy2T7Bh3bp2vRxq7P5wGrdZq7s7FBONHfg7uB1Qn40rf7i8d70GSS2xB4mcPHvLe/WuHj7e16OQPyBPdyMFcB2c1jdZjr4OSHm0xL6pgg3RQpZtycMOapDbnt5mOr9jnk7XdY4ICtzKfuqND+KZdb8fIqOfqTfMAHLBNPeax/SnyVHcXAQxngwvNZB/13P1M81BNvIFxyrT0LlU8vcdxTi8f8APJ1HZbzcYn7auwuBt69SA/vkO6y2nyMVGZkCdPKermYXLH9YcXjN3nnFh6edYI+qXDpRsHFWjEzHX8NH7OCaU+8/KTiKluLtBfqk/gn0lqqioEPYCDmr0ca9XXPrY9k9Xmyk1xrnnA+D8m7anpq8fiuJBXuYuA9GZADav/YiPtYxnvaJKgfVjy255YZN59ePNqP2EkRbd+ZzViptusZkAMMrk/zIPJ6wTWzGPhiDCTFwEMZqOLxofTeQTi/mQBfp8A8yx/n52JR9jnp04/GRuW896/JuuSj5lH027ZU8fi2rivgxcDXZeB7FI9fl+2yvBgoBoqBYqAYKAaKAWGgiseKhGKgGHg7Bqp4fDuXFeBioBgoBoqBYqAY+AAGqnj8ACeWCcVAMVAMFAPFQDFQDBQDxUAxUAw8moEqHh/NcMkvBoqBYqAYKAaKgWKgGCgGioFi4AMYqOLxA5xYJhQDxUAxUAwUA8VAMVAMFAPFQDHwaAaqeHw0w68qf+VPIDwU7vfQedagZ/9K21lANaAYKAaKgWKgGCgGioFioBh4SQYuLB7dT4tv/jFx/2cC4h9v1Z/Sff2fJ7/iZ5n5D9aOnwHe+sl7/zPA9+BDZA7OiWuD4XsUco/QefPfcLygeLxRV/DBS6aEAlUMFAPFQDFQDBQDxUAxUAzkDFxQPPo/atyKw5VKhwtD6ON7LTa5iKA/hnzBwT3H/6TWC4tHtg/+Zky3N4FLfcrLsiy+6EtmXNUUChfGNIrLq4ReOukROm8s6Jblghi8UVfwwaX81fhioBgoBoqBYqAYKAaKgWLgOzKwv3jMDv50mE7/QGZSbPF8KKjY6B0Hd5o3HZfTcVr4D3m2gpSLUf2jnVB8Yft0PC3HiYrUPQz7J6X01HBajr+SjeMJYm6vlc8FoFPKuFybndXuMp5717QcUEbgH/jshY60oQ0sQvWwjGYfyk7BQSPOMzHgeYQCdYfOraewxKG+4UC++cff0C8Hyw1A9Zc2Ro7LfMAYsXzp01qcg3G4sE0DR06hlUnzZZzjihpZnlsnwc/DIs8XySbMoR3WyJhdV8VAMVAMFAPFQDFQDBQDxcB+BvYXj70YAeHZQZe6tUiAofkTHih2zFi40cM5nsozLDTFHbL1wI9TQbK5tAUf4cIDvL83U8MN6dWiQztZ/o4DPGNeA+x4/XaclwmLY+x3HAVMntd2v6Za7eBXHgv8nOZmb3wazfao3Wd0SsEDxSb7c+hhWcYvzecq34BcudmMEcI/9IVCznEqMe3xwr2D4H1gY04HC4cYP2x35hjmc+jz8kVilKea6rUYKAaKgWKgGCgGioFioBi4hIHdxWNa/PgiQjW7A7o0Z4Vi1qZC2qs7IHOrKyp0Rjw875DfJtu5cuAe5/VLisf8sJ7y101sT1XpCedmIYT2kJ55OfXCrX3kVUG7QsfalxX4OW7lFl/zoieTqW2tIAu+RJ1yrfBFH/bTewOxKOc3DDY5Q+SZDOTUjg1veDhO/ehQbLoBHj/zaJ7aygQbKxv4HJ7UL26Mg1S3xUAxUAwUA8VAMVAMFAPFwG4GdheP6SE9FANNb9qeFWAbB2M1IZWlTxnl44JScNhCQ6bvkN8hw0dCWSc8geLvxeG9gstffZFAo9KDfTadDvv9Y41xwJB9WmYynLBy8eSKL1c0jHlNZsJrGBPVcwuNMx+f1XGMfTwJk2bw+6bO3FeICa9VZRqXvdNf7IgRxjg+hmp84Thl6c1fNE7+W4+TDL8UkDQX5jGGdp/pVLMMnzts03n1WgwUA8VAMVAMFAPFQDFQDFzBwGXFo39KsnawxcOvgjIHXW3MCwbt5dd0Hozgw7sctOPhfIf8JormjgLAP+GCAghUr12yLPsILX9qlgrIigAYqJz3J440nuwnjFC46bg2NXCT8BrGgFq8XC2EE5nmaVzSP3SK3ZY2y8UYC2icndCTXkYZGCN4TdPdvdfF9qwUfYn2qHsMkiJy+E/H6usYCVeEpxet8ak1zU2LfBBRl8VAMVAMFAPFQDFQDBQDxcBeBvYXj+0gPb6LZQ/28n1DPUhLHx5c84OsO5wzajc3KTiMcdCfHcDpcG0LEjO73ZBOxZ71Zzizca2ND/UgzxUZzIV+1LIXgQrluEyA2dskBc20zPPU7SJ58zzbQsEVOoF/4E0toTHDv9qavHr7ug3CE8owtp7RacaSWqcnxcdjRtGVoDVNnk/WqXw7P4l+iB+vy92L7OF3bw/frwUjydKYUNunaZnwiSRTMopE0odcG0MdNtNXN8VAMVAMFAPFQDFQDBQDxcAVDFxQPOr31+AJHR6E+bA6Ds761KY/GcGDMR/SQU57eiLizhePckgf8y0MbL+g6Eswodyhc1+hMsYLHpRli4pmLzxBwrEix+qUggfamHsocigQfDEC9rF8vgcZrTBZLUZccFn7UI5wvu53HBu/gyi2qQ8xnuJYgQT8IXEOL96ijvkk83WqsWualxl/kGiJulDWYZ7NDy1x30rckz4z1xWJff3gfFc8LohH46cZYmUTn5Z35KOui4FioBgoBoqBYqAYKAaKgT0MXFY87pH4UmP2Fo/JuKS4einTCsyXZ4ALRK16mQ1bCH95goqAYqAYKAaKgWKgGCgGioG7MlDFI9HJhaJ9yiUfWaynNXeNthJ2RwakULRPipM3Qe6osUQVA8VAMVAMFAPFQDFQDHxtBr5M8Rg/xqcfjWzfIWsf/ewft9z4mN9ZWW8eU+ajm/pxyP76ygU1fKy04wU/m6d0b+4khu8+Iqzf3fwE08qGYqAYKAaKgWKgGCgGioGXY+DDi8eX47sAFQPFQDFQDBQDxUAxUAwUA8VAMfCWDFTx+JZuK9DFQDFQDBQDxUAxUAwUA8VAMVAMPJeBKh6fy3dpKwaKgWKgGCgGioFioBgoBoqBYuAtGfg/1rRh4bwRZoMAAAAASUVORK5CYII="
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LDSC NGC OAall \n",
    "**Author**: Jesse Marks <br>\n",
    "**GitHub Issue:** [140](https://github.com/RTIInternational/bioinformatics/issues/140)\n",
    "**GWAS location:** <br>\n",
    "* `s3://rti-heroin/rti-midas-data/studies/ngc/meta/113/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz` \n",
    "* `s3://rti-heroin/rti-midas-data/studies/ngc/meta/115/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr{1..22}.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz`<br>\n",
    "\n",
    "**Results Location:** \n",
    "\n",
    "**Description**:\n",
    "Recalculating the SNP-based heritability for the new OAall metas with 113 and 115. These meta-analyses include an updated version of Yale-Penn Perhaps correcting Yale-Penn gives us better estimates. We'd also like to see if the big difference between including decode OAall and OAexp persists. \n",
    "\n",
    "\n",
    "We are going to utilize the [LD score regression pipeline](https://github.com/RTIInternational/ld-regression-pipeline) that Alex Waldrop developed to perform LD score regression. \n",
    "\n",
    "**Note:** The h^2 ended up being greater than 1. After researching this issue it was decided that the sample size was too small. LDSC software calls for ~5K samples for an accurate calculation. Here was an order of maginitude off. ![image.png](attachment:image.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data wrangling\n",
    "Format the summary stats for input into cromwell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### meta128"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_128 ##\n",
    "mkdir -p /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0004/eur/meta128\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0004/eur/meta128/\n",
    "prefix=oaall_meta128\n",
    "N=22373\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/128/processing/oaall/cats+coga+decode+kreek+odb+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.tsv.gz . \n",
    "done \n",
    "# combine into one file\n",
    "zcat  cats+coga+decode+kreek+odb+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.tsv.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat  cats+coga+decode+kreek+odb+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.tsv.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt \n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta127"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_127 ##\n",
    "mkdir -p /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0004/eur/meta127\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0004/eur/meta127/\n",
    "prefix=oaall_meta127\n",
    "N=23987\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/127/processing/oaall/cats+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.tsv.gz . \n",
    "done \n",
    "# combine into one file\n",
    "zcat cats+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.tsv.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat cats+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.tsv.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt \n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta125"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_125 ##\n",
    "mkdir -p /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0003/eur/meta115\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0003/eur/meta115/\n",
    "prefix=oaall_meta125\n",
    "N=29037\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/125/processing/oaall/cats+coga+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.tsv.gz . \n",
    "done \n",
    "# combine into one file\n",
    "zcat cats+coga+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.tsv.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat cats+coga+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.tsv.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt \n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta115"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_115 ##\n",
    "mkdir -p /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta115\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta115/\n",
    "prefix=oaall_meta115\n",
    "N=31618\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/115/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz . \n",
    "done \n",
    "# combine into one file\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt \n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta113"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_113 ##\n",
    "mkdir -p /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta113/\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta113/\n",
    "prefix=oaall_meta113\n",
    "N=304505\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/113/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz .\n",
    "done \n",
    "\n",
    "# combine into one file\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.tsv.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt \n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/\n",
    "\n",
    "## split by chrosome\n",
    "awk '\n",
    "    NR==1{ h = $0}\n",
    "    NR>1 { print (!a[$2]++ ? h ORS $0: $0) > \"oaall_meta113_ldsc_ready_chr\"$2\".txt\"}    \n",
    "'  <(zcat ${prefix}_ldsc_ready.txt.gz) \n",
    "\n",
    "gzip ${prefix}_ldsc_ready_chr*.txt\n",
    "\n",
    "for i in {1..22}; do\n",
    "    aws s3 cp ${prefix}_ldsc_ready_chr${i}.txt.gz \\\n",
    "    s3://rti-shared/ldsc/data/oaall/meta113/\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta107"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_107 ##\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta107/\n",
    "prefix=oaall_meta107\n",
    "N=30037\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/107/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz . --quiet \n",
    "done \n",
    "\n",
    "# combine into one file\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus.ea.chr1.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat cats+coga+decode+kreek+odb+uhs+vidus.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt \n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "## convert to Zscore\n",
    "#docker run -it -v \"$PWD:/data/\" \\\n",
    "#               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "#    continuumio/anaconda3 python /data/python_convert/sumstats.py zscore \\\n",
    "#        --sumstats /data/oaall_meta115.csv \\\n",
    "#        --out /data/oaall_meta115_z.csv \n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta094"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_094 ##\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta094/\n",
    "prefix=oaall_meta094\n",
    "N=31620\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/094/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz . --quiet \n",
    "done \n",
    "\n",
    "# combine into one file\n",
    "zcat  cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz \\\n",
    " > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt -5 | cut -f1-15\n",
    "wc -l $prefix.txt &\n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "## convert to Zscore\n",
    "#docker run -it -v \"$PWD:/data/\" \\\n",
    "#               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "#    continuumio/anaconda3 python /data/python_convert/sumstats.py zscore \\\n",
    "#        --sumstats /data/oaall_meta115.csv \\\n",
    "#        --out /data/oaall_meta115_z.csv \n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "#### meta089"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process meta_089 ##\n",
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0002/eur/meta089/\n",
    "prefix=oaall_meta089\n",
    "N=304507\n",
    "\n",
    "# download meta results\n",
    "for chr in {1..22};do \n",
    "    aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/meta/089/processing/oaall/cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz . \n",
    "done \n",
    "\n",
    "# combine into one file\n",
    "zcat cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr1.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz \\\n",
    "    > $prefix.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    zcat  cats+coga+decode+kreek+odb+uhs+vidus+yale-penn.ea.chr$chr.maf_gt_0.01.rsq_gt_0.3.with_cohort_data.gz | \\\n",
    "      tail -n +2  >> $prefix.txt\n",
    "done \n",
    "\n",
    "head $prefix.txt\n",
    "wc -l $prefix.txt\n",
    "\n",
    "# QC\n",
    "docker run -it -v \"$PWD:/data/\" \\\n",
    "               -v \"/shared/jmarks/bin/python_convert/:/data/python_convert/\" \\\n",
    "    continuumio/anaconda3 python  /data/python_convert/sumstats.py csv \\\n",
    "        --sumstats /data/$prefix.txt \\\n",
    "        --out /data/$prefix.csv \\\n",
    "        --force \\\n",
    "        --a1 Allele2 \\\n",
    "        --a2 Allele1 \\\n",
    "        --auto \\\n",
    "        --head 5 \\\n",
    "        --n-val $N\n",
    "\n",
    "# SNP CHR BP A1 A2 BETA  PVAL  N \n",
    "awk '{print $1, $2, $3, $5,$6,$8,$4,$7}' OFS=\"\\t\" $prefix.csv \\\n",
    "    > ${prefix}_ldsc_ready.txt \n",
    "\n",
    "gzip ${prefix}_ldsc_ready.txt\n",
    "\n",
    "## upload to S3\n",
    "aws s3 cp ${prefix}_ldsc_ready.txt.gz s3://rti-shared/ldsc/data/oaall/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "# 0001 \n",
    "Version B is what is on GitHub."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "## A (with FTND)\n",
    "FTND as ref vs meta113 and meta115\n",
    "\n",
    "WF Job IDs:<br>\n",
    "`4c0bec40-10bc-449c-96ba-d24671ee02c5`\n",
    "\n",
    "`2f52f9f9-6719-4ef4-a56a-48c923cd8288`\n",
    "\n",
    "`42a9e7b5-932f-4e4c-a9bc-eb3f0b761991`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create WorkFlow inputs\n",
    "Here is an example entry in the Excel Phenotype File:\n",
    "\n",
    "**trait\tplot_label\tsumstats_path\tpmid\tcategory\tsample_size\tid_col\tchr_col\tpos_col\teffect_allele_col\tref_allele_col\teffect_col\tpvalue_col\tsample_size_col\teffect_type\tw_ld_chr**\n",
    "```\n",
    "COPDGWAS Hobbs et al.\tCOPD\ts3://rti-nd/LDSC/COPDGWAS_HobbsEtAl/modGcNoOtherMinMissSorted.withchrpos.txt.gz\t28166215\tRespiratory\t51772\t3\t1\t2\t4\t5\t10\t12\t\tbeta\ts3://clustername--files/eur_w_ld_chr.tar.bz2\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20200728_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0001/eur\n",
    "mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200728_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0001/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "## create config file\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\n",
    "\n",
    "\n",
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "## copy cromwell config file from S3 to EC2 instance\n",
    "cd /shared/jmarks/bin/cromwell\n",
    "#aws s3 cp s3://rti-cromwell-output/cromwell-config/cromwell_default_genomics_queue.conf .\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Workflow\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "## B \n",
    "just meta113 & meta115\n",
    "`84aea61b-2233-432d-b7ea-b57abde69e0f`<br>\n",
    "\n",
    "`05872e14-00ec-43ef-9d81-18ebf7f2e8c2`\n",
    "\n",
    "`b75d3427-0f02-43de-943e-6e5aec99750d`\n",
    "\n",
    "`6e30d8d2-a938-4eff-833b-8f27073cf4dd`\n",
    "\n",
    "`2160bbb5-ca37-4ff6-8729-a9d6de2816a2`\n",
    "\n",
    "`2723da3f-2eb9-4723-a83d-1379ff1532ba`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20200728_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0002/eur\n",
    "mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200728_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0001/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "## copy cromwell config file from S3 to EC2 instance\n",
    "cd /shared/jmarks/bin/cromwell\n",
    "#aws s3 cp s3://rti-cromwell-output/cromwell-config/cromwell_default_genomics_queue.conf .\n",
    "\n",
    "## Run workflow—Navigate to cromwell directory\n",
    "#java -Dconfig.file=/shared/jmarks/bin/cromwell/cromwell_default_genomics_queue.conf \\\n",
    "#    -jar ~/bin/cromwell/cromwell-51.jar \\\n",
    "#    run $procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl \\\n",
    "#    -i $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json \\\n",
    "#    -p $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "# 0002\n",
    "3f93045f-8606-4ef8-95c7-4691fc3c0e74"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20200730_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0002/eur\n",
    "mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200728_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0001/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "## copy cromwell config file from S3 to EC2 instance\n",
    "cd /shared/jmarks/bin/cromwell\n",
    "#aws s3 cp s3://rti-cromwell-output/cromwell-config/cromwell_default_genomics_queue.conf .\n",
    "\n",
    "## Run workflow—Navigate to cromwell directory\n",
    "#java -Dconfig.file=/shared/jmarks/bin/cromwell/cromwell_default_genomics_queue.conf \\\n",
    "#    -jar ~/bin/cromwell/cromwell-51.jar \\\n",
    "#    run $procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl \\\n",
    "#    -i $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json \\\n",
    "#    -p $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-ld_regression/shard-0/LDSC.single_ld_regression_wf/25a3e2a6-a2bd-486e-aaa0-9c1f97dab120/call-ld_regression/meta113_by_meta115.ldsc_regression.log to shard-0/LDSC.single_ld_regression_wf/25a3e2a6-a2bd-486e-aaa0-9c1f97dab120/call-ld_regression/meta113_by_meta115.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-ld_regression/shard-2/LDSC.single_ld_regression_wf/7847f217-ddf8-460f-ac0f-02bde4d1acad/call-ld_regression/meta113_by_meta094.ldsc_regression.log to shard-2/LDSC.single_ld_regression_wf/7847f217-ddf8-460f-ac0f-02bde4d1acad/call-ld_regression/meta113_by_meta094.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-ld_regression/shard-1/LDSC.single_ld_regression_wf/14ecd43a-922c-4af3-94aa-01f286371941/call-ld_regression/meta113_by_meta107.ldsc_regression.log to shard-1/LDSC.single_ld_regression_wf/14ecd43a-922c-4af3-94aa-01f286371941/call-ld_regression/meta113_by_meta107.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-ld_regression/shard-3/LDSC.single_ld_regression_wf/ea00d508-d1bb-4c2b-bac1-6f4940f7d912/call-ld_regression/meta113_by_meta089.ldsc_regression.log to shard-3/LDSC.single_ld_regression_wf/ea00d508-d1bb-4c2b-bac1-6f4940f7d912/call-ld_regression/meta113_by_meta089.ldsc_regression.log\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/6888827e-2b33-4bce-ace0-86b5c897ccae/call-merge_munge_output/test.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/test.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-munge_pheno/shard-0/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/5ea0f20a-d3dc-4f83-a913-207b7b02ffec/call-merge_munge_output/oaall_meta115_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta115_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-munge_pheno/shard-1/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/f7eb4d4f-b993-49cf-b3f7-3edd4d95a308/call-merge_munge_output/oaall_meta107_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta107_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-munge_pheno/shard-2/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/d9a15282-335c-45bd-8217-7f1962820b6d/call-merge_munge_output/oaall_meta094_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta094_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-munge_pheno/shard-3/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/86975cc9-45b6-45e0-890e-34f5e1e1dd4b/call-merge_munge_output/oaall_meta089_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta089_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-plot_ld/PLOT.plot_ld_regression_wf/0e5444d6-2cb4-45ad-962c-733e6674bfc6/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_test.ld_regression_results.adj_pvalue.csv\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/3f93045f-8606-4ef8-95c7-4691fc3c0e74/call-plot_ld/PLOT.plot_ld_regression_wf/0e5444d6-2cb4-45ad-962c-733e6674bfc6/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_test.ld_regression_results.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  2116  100  2116    0     0    743      0  0:00:02  0:00:02 --:--:--   743\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    45  100    45    0     0     22      0  0:00:02  0:00:01  0:00:01    22\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "job=\"3f93045f-8606-4ef8-95c7-4691fc3c0e74\"\n",
    "phen=\"oaall\"\n",
    "project=\"heroin\"\n",
    "version=\"0002\"\n",
    "\n",
    "\n",
    "mkdir -p ~/Projects/$project/ldsc/$phen/$version/processing/output\n",
    "cd ~/Projects/$project/ldsc/$phen/$version/processing/output\n",
    "    \n",
    "#Download output from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/outputs\" -H \"accept: application/json\" \\\n",
    "    > final_outputs_${job}.json\n",
    "\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/logs\" -H \"accept: application/json\" \\\n",
    "    > final_logs_${job}.json\n",
    "\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-ld_regression/ \\\n",
    "    . --recursive --exclude \"*\" --include \"*.ldsc_regression.log\"\n",
    "    \n",
    "mv shard-*/LDSC.single_ld_regression_wf/*/call-ld_regression/*.ldsc_regression.log .\n",
    "rm -rf shard-*\n",
    "\n",
    "# note that I had to unindent here to get it to run\n",
    "python - <<EOF\n",
    "import json\n",
    "import os\n",
    "\n",
    "\n",
    "def traverse(o, tree_types=(list, tuple)):\n",
    "    if isinstance(o, tree_types):\n",
    "        for value in o:\n",
    "            for subvalue in traverse(value, tree_types):\n",
    "                yield subvalue\n",
    "    else:\n",
    "        yield o\n",
    "\n",
    "with open('final_outputs_' + \"$job\" + '.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for key in outputs:\n",
    "        if (type(outputs[key]) == list):\n",
    "            for value in traverse(outputs[key]):\n",
    "                if (str(value)[0:2] == \"s3\"):\n",
    "                    message = \"aws s3 cp {} s3://rti-heroin/ldsc/results/results/oaall/eur/\".format(value)\n",
    "                    os.system(message)\n",
    "        else:\n",
    "            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "                message = \"aws s3 cp {} s3://rti-heroin/ldsc/results/results/oaall/eur/\".format(outputs[key])\n",
    "                os.system(message)\n",
    "EOF\n",
    "\n",
    "# Rename files\n",
    "#for file in $(aws s3 ls s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/ | grep ${study} | perl -lne '/(\\S+)$/; print $1;'); do\n",
    "#    finalFile=$(echo $file | perl -spe 's/$study.+MetaAssoc/MetaScore/; s/.rsq./_rsq_gt_/; s/.sampleMAF./_sample_maf_gt_/; s/.popMAF./_pop_maf_gt_/; s/\\.p\\./_p_lt_/; s/.manhattan/_manhattan/; s/.qq/_qq/;' -- -study=$study)\n",
    "#\n",
    "#    aws s3 mv s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/$file \\\n",
    "#              s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/$finalFile\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "# 0003\n",
    "meta113 vs 125,115,107,094,089\n",
    "\n",
    "`{\"id\":\"4f7cb309-4a2d-428b-8771-f926145087aa\",\"status\":\"Submitted\"}`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20200804_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0003/eur\n",
    "mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200728_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0001/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "cd /shared/jmarks/bin/cromwell\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-ld_regression/shard-0/LDSC.single_ld_regression_wf/82ebfafc-6c10-4302-88f0-b803ccaa07cd/call-ld_regression/meta113_by_meta125.ldsc_regression.log to shard-0/LDSC.single_ld_regression_wf/82ebfafc-6c10-4302-88f0-b803ccaa07cd/call-ld_regression/meta113_by_meta125.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-ld_regression/shard-4/LDSC.single_ld_regression_wf/f388a073-c5a6-4aab-b836-7215a0f8d6c5/call-ld_regression/meta113_by_meta089.ldsc_regression.log to shard-4/LDSC.single_ld_regression_wf/f388a073-c5a6-4aab-b836-7215a0f8d6c5/call-ld_regression/meta113_by_meta089.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-ld_regression/shard-1/LDSC.single_ld_regression_wf/bcd4b6a2-8e72-499f-883e-f5a34965eab4/call-ld_regression/meta113_by_meta115.ldsc_regression.log to shard-1/LDSC.single_ld_regression_wf/bcd4b6a2-8e72-499f-883e-f5a34965eab4/call-ld_regression/meta113_by_meta115.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-ld_regression/shard-3/LDSC.single_ld_regression_wf/cc1ffb7f-334b-40b7-81a9-bd66d9f5e9b4/call-ld_regression/meta113_by_meta094.ldsc_regression.log to shard-3/LDSC.single_ld_regression_wf/cc1ffb7f-334b-40b7-81a9-bd66d9f5e9b4/call-ld_regression/meta113_by_meta094.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-ld_regression/shard-2/LDSC.single_ld_regression_wf/2deaa98c-eb45-45c2-ad3b-3247aa6f2d6e/call-ld_regression/meta113_by_meta107.ldsc_regression.log to shard-2/LDSC.single_ld_regression_wf/2deaa98c-eb45-45c2-ad3b-3247aa6f2d6e/call-ld_regression/meta113_by_meta107.ldsc_regression.log\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/23f27976-e550-4b9e-afdb-c3ee1f2720aa/call-merge_munge_output/test.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/test.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-munge_pheno/shard-0/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/22848e21-0dca-46c4-b477-d8a69e832ec9/call-merge_munge_output/oaall_meta125_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta125_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-munge_pheno/shard-1/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/7f66280f-0e56-4fa1-ab92-3e051b1d92f4/call-merge_munge_output/oaall_meta115_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta115_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-munge_pheno/shard-2/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/ec4a807d-ee05-4619-a3f2-403e275f9741/call-merge_munge_output/oaall_meta107_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta107_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-munge_pheno/shard-3/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/86d2e77f-e8db-4b52-b8a1-800763d26b12/call-merge_munge_output/oaall_meta094_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta094_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-munge_pheno/shard-4/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/0b3a391e-039a-452d-8e75-7d94b95307ce/call-merge_munge_output/oaall_meta089_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_meta089_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-plot_ld/PLOT.plot_ld_regression_wf/eeec7707-724f-465d-a654-78e051746dcc/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_test.ld_regression_results.pdf\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/4f7cb309-4a2d-428b-8771-f926145087aa/call-plot_ld/PLOT.plot_ld_regression_wf/eeec7707-724f-465d-a654-78e051746dcc/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv to s3://rti-heroin/ldsc/results/results/oaall/eur/oaall_test.ld_regression_results.adj_pvalue.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  2397  100  2397    0     0  12229      0 --:--:-- --:--:-- --:--:-- 12229\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    45  100    45    0     0    266      0 --:--:-- --:--:-- --:--:--   264\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "job=\"4f7cb309-4a2d-428b-8771-f926145087aa\"\n",
    "phen=\"oaall\"\n",
    "project=\"heroin\"\n",
    "version=\"0003\"\n",
    "\n",
    "\n",
    "mkdir -p ~/Projects/$project/ldsc/$phen/$version/processing/output/logs/\n",
    "cd ~/Projects/$project/ldsc/$phen/$version/processing/output\n",
    "    \n",
    "# Download output JSON from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/outputs\" -H \"accept: application/json\" \\\n",
    "    > final_outputs_${job}.json\n",
    "\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/logs\" -H \"accept: application/json\" \\\n",
    "    > final_logs_${job}.json\n",
    "\n",
    "# Download logs that contain h2\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-ld_regression/ \\\n",
    "    logs/ --recursive --exclude \"*\" --include \"*.ldsc_regression.log\" \n",
    "    \n",
    "mv logs/shard-*/LDSC.single_ld_regression_wf/*/call-ld_regression/*.ldsc_regression.log logs/\n",
    "rm -rf logs/shard-*\n",
    "\n",
    "\n",
    "# Download plot\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-plot_ld/ \\\n",
    "    . --recursive --exclude \"*\" --include \"*.ld_regression_results.*\"\n",
    "\n",
    "\n",
    "# note that I had to unindent here to get it to run\n",
    "python - <<EOF\n",
    "import json\n",
    "import os\n",
    "\n",
    "\n",
    "def traverse(o, tree_types=(list, tuple)):\n",
    "    if isinstance(o, tree_types):\n",
    "        for value in o:\n",
    "            for subvalue in traverse(value, tree_types):\n",
    "                yield subvalue\n",
    "    else:\n",
    "        yield o\n",
    "\n",
    "with open('final_outputs_' + \"$job\" + '.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for key in outputs:\n",
    "        if (type(outputs[key]) == list):\n",
    "            for value in traverse(outputs[key]):\n",
    "                if (str(value)[0:2] == \"s3\"):\n",
    "                    message = \"aws s3 cp {} s3://rti-heroin/ldsc/results/oaall/eur/$version/\".format(value)\n",
    "                    os.system(message)\n",
    "        else:\n",
    "            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "                message = \"aws s3 cp {} s3://rti-heroin/ldsc/results/oaall/eur/$version/\".format(outputs[key])\n",
    "                os.system(message)\n",
    "EOF\n",
    "\n",
    "# Rename files\n",
    "#for file in $(aws s3 ls s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/ | grep ${study} | perl -lne '/(\\S+)$/; print $1;'); do\n",
    "#    finalFile=$(echo $file | perl -spe 's/$study.+MetaAssoc/MetaScore/; s/.rsq./_rsq_gt_/; s/.sampleMAF./_sample_maf_gt_/; s/.popMAF./_pop_maf_gt_/; s/\\.p\\./_p_lt_/; s/.manhattan/_manhattan/; s/.qq/_qq/;' -- -study=$study)\n",
    "#\n",
    "#    aws s3 mv s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/$file \\\n",
    "#              s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/$finalFile\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": true,
    "toc-nb-collapsed": true
   },
   "source": [
    "# 0004\n",
    "meta113 vs 128,127,125,115,107,094,089\n",
    "\n",
    "`6b161cb8-d225-4de6-8323-592a8f60d63e`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20200810_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0004/eur\n",
    "mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200810_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0004/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "#cd /shared/jmarks/bin/cromwell\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-3/LDSC.single_ld_regression_wf/29efe478-80b3-4af7-b6fc-f2ab221bb86c/call-ld_regression/meta113_by_meta115.ldsc_regression.log to logs/shard-3/LDSC.single_ld_regression_wf/29efe478-80b3-4af7-b6fc-f2ab221bb86c/call-ld_regression/meta113_by_meta115.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-2/LDSC.single_ld_regression_wf/f9dabe0d-eba1-4cd6-8bcf-1925fa55ce66/call-ld_regression/meta113_by_meta125.ldsc_regression.log to logs/shard-2/LDSC.single_ld_regression_wf/f9dabe0d-eba1-4cd6-8bcf-1925fa55ce66/call-ld_regression/meta113_by_meta125.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-1/LDSC.single_ld_regression_wf/7f4ec65c-ff39-48cb-8d61-7b92e826c144/call-ld_regression/meta113_by_meta127.ldsc_regression.log to logs/shard-1/LDSC.single_ld_regression_wf/7f4ec65c-ff39-48cb-8d61-7b92e826c144/call-ld_regression/meta113_by_meta127.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-0/LDSC.single_ld_regression_wf/785412dc-0bc0-4fb0-928c-affbec6fa18d/call-ld_regression/meta113_by_meta128.ldsc_regression.log to logs/shard-0/LDSC.single_ld_regression_wf/785412dc-0bc0-4fb0-928c-affbec6fa18d/call-ld_regression/meta113_by_meta128.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-6/LDSC.single_ld_regression_wf/3d02b970-ec9c-410b-a1c3-6d17d14e2255/call-ld_regression/meta113_by_meta089.ldsc_regression.log to logs/shard-6/LDSC.single_ld_regression_wf/3d02b970-ec9c-410b-a1c3-6d17d14e2255/call-ld_regression/meta113_by_meta089.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-5/LDSC.single_ld_regression_wf/aec8a28a-6cd1-46ca-984a-4ecf13826209/call-ld_regression/meta113_by_meta094.ldsc_regression.log to logs/shard-5/LDSC.single_ld_regression_wf/aec8a28a-6cd1-46ca-984a-4ecf13826209/call-ld_regression/meta113_by_meta094.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-ld_regression/shard-4/LDSC.single_ld_regression_wf/7c668ccc-3c35-4384-9b63-c2f275f11d08/call-ld_regression/meta113_by_meta107.ldsc_regression.log to logs/shard-4/LDSC.single_ld_regression_wf/7c668ccc-3c35-4384-9b63-c2f275f11d08/call-ld_regression/meta113_by_meta107.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-plot_ld/PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf to PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-plot_ld/PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv to PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-plot_ld/PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-make_plot_table/oaall_test.ld_regression_results.tsv to PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-make_plot_table/oaall_test.ld_regression_results.tsv\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/f47dfbd1-6244-4124-b1ca-bb061b5a9e07/call-merge_munge_output/test.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/test.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-plot_ld/PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_test.ld_regression_results.pdf\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-plot_ld/PLOT.plot_ld_regression_wf/863c6d16-0a77-4544-b2a0-b7439b0b27e6/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_test.ld_regression_results.adj_pvalue.csv\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-0/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/07f1dedc-a135-477e-8808-086b26bf6228/call-merge_munge_output/oaall_meta128_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta128_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-1/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/dbb8c83a-3a4b-45e1-8945-485f5f4699f1/call-merge_munge_output/oaall_meta127_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta127_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-2/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/aad0b45b-c023-46ae-9731-13ff6d536d66/call-merge_munge_output/oaall_meta125_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta125_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-3/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/93785101-0f1b-4ae0-8387-9c22cd3ccc49/call-merge_munge_output/oaall_meta115_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta115_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-4/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/fbf5fb92-05d5-498e-a591-ba702d84edb6/call-merge_munge_output/oaall_meta107_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta107_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-5/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/30711f82-6dc3-411c-bace-bcfc72cc3003/call-merge_munge_output/oaall_meta094_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta094_ldsc_ready.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/6b161cb8-d225-4de6-8323-592a8f60d63e/call-munge_pheno/shard-6/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/24c1fd5c-c8ce-430d-b900-601f9af69e09/call-merge_munge_output/oaall_meta089_ldsc_ready.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0004/oaall_meta089_ldsc_ready.txt.munged.merged.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  2959  100  2959    0     0  20129      0 --:--:-- --:--:-- --:--:-- 20129\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    45  100    45    0     0    319      0 --:--:-- --:--:-- --:--:--   319\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "job=\"6b161cb8-d225-4de6-8323-592a8f60d63e\"\n",
    "phen=\"oaall\"\n",
    "project=\"heroin\"\n",
    "version=\"0004\"\n",
    "\n",
    "\n",
    "mkdir -p ~/Projects/$project/ldsc/$phen/$version/processing/output/logs/\n",
    "cd ~/Projects/$project/ldsc/$phen/$version/processing/output\n",
    "    \n",
    "# Download output JSON from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/outputs\" -H \"accept: application/json\" \\\n",
    "    > final_outputs_${job}.json\n",
    "\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/logs\" -H \"accept: application/json\" \\\n",
    "    > final_logs_${job}.json\n",
    "\n",
    "# Download logs that contain h2\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-ld_regression/ \\\n",
    "    logs/ --recursive --exclude \"*\" --include \"*.ldsc_regression.log\" \n",
    "    \n",
    "mv logs/shard-*/LDSC.single_ld_regression_wf/*/call-ld_regression/*.ldsc_regression.log logs/\n",
    "rm -rf logs/shard-*\n",
    "\n",
    "\n",
    "# Download plot\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-plot_ld/ \\\n",
    "    . --recursive --exclude \"*\" --include \"*.ld_regression_results.*\"\n",
    "    \n",
    "mv PLOT.plot_ld_regression_wf/*/call-*/*ld_regression_results* .\n",
    "\n",
    "\n",
    "Projects note that I had to unindent here to get it to run\n",
    "python - <<EOF\n",
    "import json\n",
    "import os\n",
    "\n",
    "\n",
    "def traverse(o, tree_types=(list, tuple)):\n",
    "    if isinstance(o, tree_types):\n",
    "        for value in o:\n",
    "            for subvalue in traverse(value, tree_types):\n",
    "                yield subvalue\n",
    "    else:\n",
    "        yield o\n",
    "\n",
    "with open('final_outputs_' + \"$job\" + '.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for key in outputs:\n",
    "        if (type(outputs[key]) == list):\n",
    "            for value in traverse(outputs[key]):\n",
    "                if (str(value)[0:2] == \"s3\"):\n",
    "                    message = \"aws s3 cp {} s3://rti-heroin/ldsc/results/oaall/eur/$version/\".format(value)\n",
    "                    os.system(message)\n",
    "        else:\n",
    "            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "                message = \"aws s3 cp {} s3://rti-heroin/ldsc/results/oaall/eur/$version/\".format(outputs[key])\n",
    "                os.system(message)\n",
    "EOF\n",
    "\n",
    "# Rename files\n",
    "#for file in $(aws s3 ls s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/ | grep ${study} | perl -lne '/(\\S+)$/; print $1;'); do\n",
    "#    finalFile=$(echo $file | perl -spe 's/$study.+MetaAssoc/MetaScore/; s/.rsq./_rsq_gt_/; s/.sampleMAF./_sample_maf_gt_/; s/.popMAF./_pop_maf_gt_/; s/\\.p\\./_p_lt_/; s/.manhattan/_manhattan/; s/.qq/_qq/;' -- -study=$study)\n",
    "#\n",
    "#    aws s3 mv s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/$file \\\n",
    "#              s3://rti-${project}/ldsc/${study}/results/$phen/$version/$ancestry/$finalFile\n",
    "#done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Final Plot\n",
    "Merge the output tables from cromwell and LDHub. The merged table should have the header:\n",
    "```\n",
    "trait2\tTrait_Label\tTrait_Group\trg\tse\tz\tp\th2_obs\th2_obs_se\th2_int\th2_int_se\tgcov_int\tgcov_int_se\n",
    "```\n",
    "\n",
    "**Note**: upload the plot table to EC2 instance to run docker and create the plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# send plot file from local to EC2\n",
    "# scp -r -i ~/.ssh/gwas_rsa oaall_test.ld_regression_results.csv   ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0003/eur/plot/\n",
    "\n",
    "# enter interactive mode\n",
    "docker run -it -v\"/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0004/eur/plot/:/data/\" \\\n",
    "    rticode/plot_ld_regression_results:172bbfcc46b857dc95b4d0a080ccd092fd3a4bac  /bin/bash\n",
    "    \n",
    "Rscript /opt/plot_ld_regression/plot_ld_regression_results.R  \\\n",
    "    --input_file oaall_test.ld_regression_results.csv \\\n",
    "    --output_file oaall_test.ld_regression_results.pdf  \\\n",
    "    --comma_delimited \\\n",
    "    --xmin -0.025 \\\n",
    "    --title \"OAall Cohorts: Meta113 as Reference\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0005\n",
    "Running 0004 with Neff instead of N."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/rti-heroin/ldsc_genetic_correlation/data/oaall/0005/\n",
    "\n",
    "aws s3 sync s3://rti-shared/ldsc/data/oaall/ .\n",
    "    \n",
    "# meta089    \n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 27772; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta089_ldsc_ready.txt.gz) > oaall_meta089_ldsc_ready_neff.txt &\n",
    "    \n",
    "# meta094\n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 22034; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta094_ldsc_ready.txt.gz) > oaall_meta094_ldsc_ready_neff.txt &\n",
    "    \n",
    "# meta107\n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 19478; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta107_ldsc_ready.txt.gz) > oaall_meta107_ldsc_ready_neff.txt &\n",
    "    \n",
    "    \n",
    "# meta 113    \n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 27765; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta113_ldsc_ready.txt.gz) > oaall_meta113_ldsc_ready_neff.txt &\n",
    "    \n",
    "cd meta113/\n",
    "\n",
    "for chr in {1..22}; do\n",
    "    awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "      NR>1 { $8 = 27765; {print $0}}' \\\n",
    "      OFS=\"\\t\" <(zcat oaall_meta113_ldsc_ready_chr${chr}.txt.gz) > oaall_meta113_ldsc_ready_chr${chr}_neff.txt\n",
    "done\n",
    "\n",
    "\n",
    "# meta115\n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 22029; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta115_ldsc_ready.txt.gz) > oaall_meta115_ldsc_ready_neff.txt &\n",
    "    \n",
    "# meta125\n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 19357; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta125_ldsc_ready.txt.gz) > oaall_meta125_ldsc_ready_neff.txt &\n",
    "    \n",
    "# meta127\n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 19430; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta127_ldsc_ready.txt.gz) > oaall_meta127_ldsc_ready_neff.txt &\n",
    "    \n",
    "# meta128\n",
    "awk 'NR==1 {print \"SNP\\tCHR\\tBP\\tA1\\tA2\\tBETA\\tPVAL\\tNEFF\"}\n",
    "  NR>1 { $8 = 18212; {print $0}}' \\\n",
    "  OFS=\"\\t\" <(zcat oaall_meta128_ldsc_ready.txt.gz) > oaall_meta128_ldsc_ready_neff.txt &\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20200812_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0005/eur\n",
    "mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200812_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0005/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "#cd /shared/jmarks/bin/cromwell\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "{\"id\":\"49d1d140-ddca-40e7-a694-d13b4b32d3f9\",\"status\":\"Submitted\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-0/LDSC.single_ld_regression_wf/1a722c09-131f-4936-b900-a7259a37b109/call-ld_regression/meta113_by_meta128.ldsc_regression.log to logs/shard-0/LDSC.single_ld_regression_wf/1a722c09-131f-4936-b900-a7259a37b109/call-ld_regression/meta113_by_meta128.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-1/LDSC.single_ld_regression_wf/766db178-a089-427f-8f5b-72737722d194/call-ld_regression/meta113_by_meta127.ldsc_regression.log to logs/shard-1/LDSC.single_ld_regression_wf/766db178-a089-427f-8f5b-72737722d194/call-ld_regression/meta113_by_meta127.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-2/LDSC.single_ld_regression_wf/280dd98a-7b44-466a-85e5-a7113df3ed3d/call-ld_regression/meta113_by_meta125.ldsc_regression.log to logs/shard-2/LDSC.single_ld_regression_wf/280dd98a-7b44-466a-85e5-a7113df3ed3d/call-ld_regression/meta113_by_meta125.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-3/LDSC.single_ld_regression_wf/6c25494c-9c71-47b7-98de-2fc736242c39/call-ld_regression/meta113_by_meta115.ldsc_regression.log to logs/shard-3/LDSC.single_ld_regression_wf/6c25494c-9c71-47b7-98de-2fc736242c39/call-ld_regression/meta113_by_meta115.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-4/LDSC.single_ld_regression_wf/f5a4b19d-37a4-489c-aaa1-1ac54aa31b94/call-ld_regression/meta113_by_meta107.ldsc_regression.log to logs/shard-4/LDSC.single_ld_regression_wf/f5a4b19d-37a4-489c-aaa1-1ac54aa31b94/call-ld_regression/meta113_by_meta107.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-5/LDSC.single_ld_regression_wf/a064af14-ebed-4ab0-b163-f4bf34e6c536/call-ld_regression/meta113_by_meta094.ldsc_regression.log to logs/shard-5/LDSC.single_ld_regression_wf/a064af14-ebed-4ab0-b163-f4bf34e6c536/call-ld_regression/meta113_by_meta094.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-ld_regression/shard-6/LDSC.single_ld_regression_wf/dbe6e404-093d-4d02-a82f-5e77cd74167b/call-ld_regression/meta113_by_meta089.ldsc_regression.log to logs/shard-6/LDSC.single_ld_regression_wf/dbe6e404-093d-4d02-a82f-5e77cd74167b/call-ld_regression/meta113_by_meta089.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-plot_ld/PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv to PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-plot_ld/PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf to PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-plot_ld/PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-make_plot_table/oaall_test.ld_regression_results.tsv to PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-make_plot_table/oaall_test.ld_regression_results.tsv\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-plot_ld/PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-adj_pvalues/oaall_test.ld_regression_results.adj_pvalue.csv to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_test.ld_regression_results.adj_pvalue.csv\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-0/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/d2cb9a81-a7cc-40c3-81c6-be6c4f9acfe2/call-merge_munge_output/oaall_meta128_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta128_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-1/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/6b582cdb-9ac7-4a5c-b4fa-d4972cdd7b4d/call-merge_munge_output/oaall_meta127_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta127_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-2/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/b1dc994b-a752-4d84-9ce8-dbc6e28dd97b/call-merge_munge_output/oaall_meta125_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta125_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-3/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/805722cb-b85b-489d-b3dd-2d3ffcd22078/call-merge_munge_output/oaall_meta115_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta115_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-4/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/0cf9df7d-e63e-49c0-afbd-88d3ae6ac7ea/call-merge_munge_output/oaall_meta107_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta107_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-5/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/d16bc8c2-058f-4397-8ece-e0740329727a/call-merge_munge_output/oaall_meta094_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta094_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_pheno/shard-6/MUNGE_TRAIT_WF.munge_phenotype_sumstats_wf/b884c9a4-03fa-4739-8470-768ac14fadbe/call-merge_munge_output/oaall_meta089_ldsc_ready_neff.txt.munged.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_meta089_ldsc_ready_neff.txt.munged.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/b47f0a59-8af2-4366-9c14-37ad8d11859a/call-merge_munge_output/test.merged.txt to s3://rti-heroin/ldsc/results/oaall/eur/0005/test.merged.txt\n",
      "copy: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/49d1d140-ddca-40e7-a694-d13b4b32d3f9/call-plot_ld/PLOT.plot_ld_regression_wf/154d44bd-359f-4b86-96f9-4814c6ed5f0d/call-plot_ld_regression_results/oaall_test.ld_regression_results.pdf to s3://rti-heroin/ldsc/results/oaall/eur/0005/oaall_test.ld_regression_results.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  2994  100  2994    0     0  14821      0 --:--:-- --:--:-- --:--:-- 14895\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    45  100    45    0     0    378      0 --:--:-- --:--:-- --:--:--   378\n",
      "bash: line 33: Projects: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "job=\"49d1d140-ddca-40e7-a694-d13b4b32d3f9\"\n",
    "phen=\"oaall\"\n",
    "project=\"heroin\"\n",
    "version=\"0005\"\n",
    "aws_path=\"s3://rti-heroin/ldsc/results/$phen/eur/$version/\"\"\n",
    "\n",
    "\n",
    "mkdir -p ~/Projects/$project/ldsc/$phen/$version/processing/output/logs/\n",
    "cd ~/Projects/$project/ldsc/$phen/$version/processing/output\n",
    "    \n",
    "# Download output JSON from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/outputs\" -H \"accept: application/json\" \\\n",
    "    > final_outputs_${job}.json\n",
    "\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/logs\" -H \"accept: application/json\" \\\n",
    "    > final_logs_${job}.json\n",
    "\n",
    "# Download logs that contain h2\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-ld_regression/ \\\n",
    "    logs/ --recursive --exclude \"*\" --include \"*.ldsc_regression.log\" \n",
    "    \n",
    "mv logs/shard-*/LDSC.single_ld_regression_wf/*/call-ld_regression/*.ldsc_regression.log logs/\n",
    "aws s3 sync logs/ $aws_path\n",
    "rm -rf logs/shard-*\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Download plot\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-plot_ld/ \\\n",
    "    . --recursive --exclude \"*\" --include \"*.ld_regression_results.*\"\n",
    "    \n",
    "mv PLOT.plot_ld_regression_wf/*/call-*/*ld_regression_results* .\n",
    "\n",
    "\n",
    "# Projects note that I had to unindent here to get it to run\n",
    "python - <<EOF\n",
    "import json\n",
    "import os\n",
    "\n",
    "\n",
    "def traverse(o, tree_types=(list, tuple)):\n",
    "    if isinstance(o, tree_types):\n",
    "        for value in o:\n",
    "            for subvalue in traverse(value, tree_types):\n",
    "                yield subvalue\n",
    "    else:\n",
    "        yield o\n",
    "\n",
    "with open('final_outputs_' + \"$job\" + '.json') as f:\n",
    "    outputs = json.load(f)\n",
    "    outputs = outputs[\"outputs\"]\n",
    "    for key in outputs:\n",
    "        if (type(outputs[key]) == list):\n",
    "            for value in traverse(outputs[key]):\n",
    "                if (str(value)[0:2] == \"s3\"):\n",
    "                    message = \"aws s3 cp {} $aws_path\".format(value)\n",
    "                    os.system(message)\n",
    "        else:\n",
    "            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "                message = \"aws s3 cp {} $aws_path\".format(outputs[key])\n",
    "                os.system(message)\n",
    "EOF\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {
    "height": "714.716px",
    "left": "0px",
    "right": "1299.83px",
    "top": "110.284px",
    "width": "211.989px"
   },
   "toc_section_display": "block",
   "toc_window_display": true
  },
  "toc-autonumbering": false,
  "toc-showcode": true,
  "toc-showmarkdowntxt": true
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
