{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LDSC on Opioid Addiction GWAS\n",
    "**author**: Jesse Marks<br>\n",
    "\n",
    "We want to perform LDSC to obtain the genetic correlation of our NGC opioid addiction phenotype with several other phenotypes. Specifically, we will create a plot analogous to [Figure 3](https://www.nature.com/articles/s41467-020-19265-z/figures/3) in the paper *Expanding the genetic architecture of nicotine dependence and its shared genetics with multiple traits*. One difference is that we will exclude the groups Cardiometabolic and Respiratory.\n",
    "\n",
    "We will use the automated WDL/cromwell [ld-regression-pipeline](https://github.com/RTIInternational/ld-regression-pipeline) workflow created by Alex Waldrop that implements [LDSC](https://github.com/bulik/ldsc). We will also need to use [LD-Hub](http://ldsc.broadinstitute.org/ldhub/) to get the genetic correlation of phenotypes that we don't have in-house."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Phenotype Table\n",
    "Perform LDSC on the following phenotypes:\n",
    "\n",
    "**Cigarette Smoking**\n",
    "* Heaviness of smoking (UKB): `s3://rti-shared/ldsc/data/ukb_hsi/`\n",
    "* Cigaretts per day (GSCAN): `s3://rti-shared/ldsc/data/gscan_liu2019/`\n",
    "* Smoking cessation (GSCAN): `s3://rti-shared/ldsc/data/gscan_liu2019/`\n",
    "* Cotinine levels: `s3://rti-shared/ldsc/data/cotinine_ware2016/`\n",
    "* Smoking initiation (GSCAN): `s3://rti-shared/ldsc/data/gscan_liu2019/`\n",
    "* Age of initiation (GSCAN): `s3://rti-shared/ldsc/data/gscan_liu2019/`\n",
    "\n",
    "**Drug and alcohol use**\n",
    "* Alcohol dependence: `s3://rti-shared/ldsc/data/alcdep_walters2018/`\n",
    "* Cannabis use disorder: `s3://rti-shared/ldsc/data/cud_demontis2019/`\n",
    "* Alcohol drinks per week (GSCAN): `s3://rti-shared/ldsc/data/gscan_liu2019/`\n",
    "* Lifetime cannabis use (ever vs never): `s3://rti-shared/ldsc/data/cannabis_icc_ukb/`\n",
    "    - Note I need to process these data probably. I should rename it to cannabis_ever_vs_never_pasman2018. This is without 23andMe results, so the N=162,082. Need to make a note of this in the S3 directory. (https://www.nature.com/articles/s41593-018-0206-1)\n",
    "    \n",
    "~**Cancer**~\n",
    "* ~Squamous cell lung cancer~\n",
    "* ~Lung cancer~\n",
    "* ~Lung adenocarcinoma~\n",
    "* ~Small cell carcinoma~\n",
    "    - We don't actually have the results to these cancer phenotypes. The genetic correlation results with ND were provide to us by Michael Bray and Laura Bierut at Wash U. See [this](https://github.com/RTIInternational/bioinformatics/issues/103#issuecomment-477357675) GitHub comment. \n",
    "\n",
    "~**Cardiometabolic**~\n",
    "- Don't include these results.\n",
    "\n",
    "~**Respiratory**~\n",
    "- Don't include these results.\n",
    "\n",
    "**Neurological**\n",
    "* Parkinsons disease: [PMID 19915575](https://pubmed.ncbi.nlm.nih.gov/19915575/) on LDHub\n",
    "* Amyotrophic lateral sclerosis: [PMID 27455348](https://pubmed.ncbi.nlm.nih.gov/27455348/) on LDHub\n",
    "* Alzheimers disease: [PMID 24162737](https://pubmed.ncbi.nlm.nih.gov/24162737/) on LDHub\n",
    "\n",
    "**Cognitive/education**\n",
    "* Intelligence: [PMID 28530673](https://pubmed.ncbi.nlm.nih.gov/28530673/) on LDHub\n",
    "* Childhood IQ:  [PMID 23358156](https://pubmed.ncbi.nlm.nih.gov/23358156/) on LDHub\n",
    "* College completion: [PMID 23722424](https://pubmed.ncbi.nlm.nih.gov/23722424/) on LDHub\n",
    "* Years of schooling: [PMID 25201988](https://pubmed.ncbi.nlm.nih.gov/25201988/) on LDHub\n",
    "\n",
    "**Personality**\n",
    "* Neuroticism: [PMID 24828478](https://pubmed.ncbi.nlm.nih.gov/24828478/) on LDHub\n",
    "* Conscientiousness: [PMID 21173776](https://pubmed.ncbi.nlm.nih.gov/21173776/) on LDHub\n",
    "* Openness to experience: [PMID 21173776](https://pubmed.ncbi.nlm.nih.gov/21173776/) on LDHub\n",
    "\n",
    "\n",
    "**Psychiatric**\n",
    "* Post traumatic stress disorder: `s3://rti-shared/ldsc/data/ptsd_nievergelt2019/`\n",
    "* Attention deficit hyperactivity disorder: `s3://rti-shared/ldsc/data/adhd_demontis2018/`\n",
    "* Depressive symptoms: `s3://rti-shared/ldsc/data/depressive_symptoms_okbay2016/`\n",
    "* Major depressive disorder: `s3://rti-shared/ldsc/data/mdd_howard2019/`\n",
    "* Bipolar disorder: `s3://rti-shared/ldsc/data/bipolar_stahl2019/`\n",
    "* Autism spectrum disorder: `s3://rti-shared/ldsc/data/autism_groves2019/`\n",
    "* Schizophrenia: `s3://rti-shared/ldsc/data/schizophrenia_ripke2014/`\n",
    "* Psychiatric cross-disorder: [PMID 23453885](https://pubmed.ncbi.nlm.nih.gov/23453885/) on LDHub\n",
    "* Anorexia nervosa: `s3://rti-shared/ldsc/data/anorexia_watson2019/`\n",
    "* Subjective well being: [PMID 27089181](https://pubmed.ncbi.nlm.nih.gov/27089181/) on LDHub\n",
    "\n",
    "**Brain volume**\n",
    "* Putamen volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "* Accumbens volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "* Pallidum volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "* Caudate volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "* Thalamus volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "* Hippocampus volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "* Intracranial volume: [PMID 25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/) on LDHub\n",
    "\n",
    "<br>\n",
    "\n",
    "___\n",
    "\n",
    "Note that we should recreate the Nicotine Dependence plot to verify the allele directions, then run it again for opioid addiction."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See /mnt/c/Users/jmarks/OneDrive - Research Triangle Institute/Projects/heroin/ldsc/oaall/0006/processing/input/20210226_heroin_ldsc_phenotypes_local.xlsx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Process OA results\n",
    "aws s3 cp s3://rti-heroin/rti-midas-data/studies/ngc/GenomicSEM/results/29/gSEM/final/genomicSEM_GWAS.oaALL.MVP1_MVP2_YP_SAGE.PGC.Song.table .\n",
    "\n",
    "# A2 is the coded allele. est = effect\n",
    "#SNP     CHR     BP      MAF     A1      A2      est     SE      Z_Estimate      Pval_Estimate\n",
    "awk 'NR==1{h = $0} NR>1{ print (!a[$2]++ ? h ORS $0: $0) > \"genomic_sem_gwas_chr\"$2\".txt\"}' \\\n",
    "    genomicSEM_GWAS.oaALL.MVP1_MVP2_YP_SAGE.PGC.Song.table\n",
    "\n",
    "for file in genomic_sem*; do\n",
    "    aws s3 cp $file s3://rti-heroin/rti-midas-data/studies/ngc/GenomicSEM/results/29/gSEM/final/ldsc_ready/\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. upload Excel phenotype file to EC2 instance\n",
    "## 2. then edit full_ld_regression_wf_template.json to include the reference data of choice\n",
    "## 3. lastly use dockerized tool to finish filling out the json file that will be input for workflow\n",
    "\n",
    "## login to a larger compute node\n",
    "qrsh\n",
    "\n",
    "phen=20210226_heroin_ldsc_phenotypes_local.xlsx\n",
    "procD=/shared/rti-heroin/ldsc/results/oaall/eur/0006/ecr_test2\n",
    "#mkdir -p $procD/{ldhub,plot} # for later processing\n",
    "git clone https://github.com/RTIInternational/ld-regression-pipeline/ $procD/ld-regression-pipeline\n",
    "mkdir $procD/ld-regression-pipeline/workflow_inputs\n",
    "\n",
    "# create final workflow input (a json file) \n",
    "cp $procD/ld-regression-pipeline/json_input/full_ld_regression_wf_template.json \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/\n",
    "# edit this file\n",
    "\n",
    "## upload files to */workflow_inputs/\n",
    "#scp -i ~/.ssh/gwas_rsa 20200812_heroin_ldsc_phenotypes_local.xlsx     ec2-user@34.195.174.206:/shared/rti-heroin/ldsc_genetic_correlation/results/oaall/0005/eur/ld-regression-pipeline/workflow_inputs/\n",
    "\n",
    "docker run -v $procD/ld-regression-pipeline/workflow_inputs/:/data/ \\\n",
    "    rticode/generate_ld_regression_input_json:1ddbd682cb1e44dab6d11ee571add34bd1d06e21 \\\n",
    "    --json-input /data/full_ld_regression_wf_template.json \\\n",
    "    --pheno-file /data/$phen >\\\n",
    "        $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## zip appropriate files \n",
    "# Change to directory immediately above metaxcan-pipeline repo\n",
    "cd $procD/ld-regression-pipeline\n",
    "cd ..\n",
    "# Make zipped copy of repo somewhere\n",
    "zip --exclude=*var/* --exclude=*.git/* -r \\\n",
    "    $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip \\\n",
    "    ld-regression-pipeline\n",
    "\n",
    "#cd /shared/jmarks/bin/cromwell\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\" "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "19952256-d8d3-4c02-bf05-8c63dae8adde"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### alternative job submission\n",
    "#\n",
    "### run ldsc workflow on AWS EC2 instance\n",
    "#java -Dconfig.file=/shared/jmarks/bin/cromwell/cromwell-server-new.conf \\\n",
    "#    -jar ~/bin/cromwell/cromwell-54.jar \\\n",
    "#    run $procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl \\\n",
    "#    -i $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json \\\n",
    "#    -p $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip #> $procD/output.log 2>&1\n",
    "#\n",
    "#java -jar ~/bin/cromwell/cromwell-54.jar \\\n",
    "#    run $procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl \\\n",
    "#    -i $procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs_local.json \\\n",
    "#    -p $procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip > $procD/output.log 2>&1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-0/LDSC.single_ld_regression_wf/a3babf3c-e1ca-4c4b-a772-f9d0386b7fd8/call-ld_regression/oa_by_hsi.ldsc_regression.log to logs/shard-0/LDSC.single_ld_regression_wf/a3babf3c-e1ca-4c4b-a772-f9d0386b7fd8/call-ld_regression/oa_by_hsi.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-1/LDSC.single_ld_regression_wf/0b072573-40a8-492f-81c5-aca0e6f8d2ce/call-ld_regression/oa_by_cpd.ldsc_regression.log to logs/shard-1/LDSC.single_ld_regression_wf/0b072573-40a8-492f-81c5-aca0e6f8d2ce/call-ld_regression/oa_by_cpd.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-17/LDSC.single_ld_regression_wf/b4c0e86c-c682-484d-a4e5-f86c23edb388/call-ld_regression/oa_by_anorexia_nervosa.ldsc_regression.log to logs/shard-17/LDSC.single_ld_regression_wf/b4c0e86c-c682-484d-a4e5-f86c23edb388/call-ld_regression/oa_by_anorexia_nervosa.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-2/LDSC.single_ld_regression_wf/223b2166-07f4-411c-ad15-7bc711b1043e/call-ld_regression/oa_by_smoking_cessation.ldsc_regression.log to logs/shard-2/LDSC.single_ld_regression_wf/223b2166-07f4-411c-ad15-7bc711b1043e/call-ld_regression/oa_by_smoking_cessation.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-13/LDSC.single_ld_regression_wf/200fcaba-00c0-40bb-8ba5-63105ffbfd52/call-ld_regression/oa_by_mdd.ldsc_regression.log to logs/shard-13/LDSC.single_ld_regression_wf/200fcaba-00c0-40bb-8ba5-63105ffbfd52/call-ld_regression/oa_by_mdd.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-11/LDSC.single_ld_regression_wf/7fb2d103-d5cf-4036-8cc0-07b82b8b8351/call-ld_regression/oa_by_adhd.ldsc_regression.log to logs/shard-11/LDSC.single_ld_regression_wf/7fb2d103-d5cf-4036-8cc0-07b82b8b8351/call-ld_regression/oa_by_adhd.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-10/LDSC.single_ld_regression_wf/ae98b0f7-0e08-470d-b09b-eafb08aae654/call-ld_regression/oa_by_ptsd.ldsc_regression.log to logs/shard-10/LDSC.single_ld_regression_wf/ae98b0f7-0e08-470d-b09b-eafb08aae654/call-ld_regression/oa_by_ptsd.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-3/LDSC.single_ld_regression_wf/a52e484a-48ea-468d-9351-fc32edc8747a/call-ld_regression/oa_by_cotinine_levels.ldsc_regression.log to logs/shard-3/LDSC.single_ld_regression_wf/a52e484a-48ea-468d-9351-fc32edc8747a/call-ld_regression/oa_by_cotinine_levels.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-12/LDSC.single_ld_regression_wf/d738fc9f-622e-47ec-b49a-e85e38ba5cfe/call-ld_regression/oa_by_depressive_symptoms.ldsc_regression.log to logs/shard-12/LDSC.single_ld_regression_wf/d738fc9f-622e-47ec-b49a-e85e38ba5cfe/call-ld_regression/oa_by_depressive_symptoms.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-4/LDSC.single_ld_regression_wf/2b55eb9a-a2ee-4995-abd5-b691666b5cb1/call-ld_regression/oa_by_smoking_initiation.ldsc_regression.log to logs/shard-4/LDSC.single_ld_regression_wf/2b55eb9a-a2ee-4995-abd5-b691666b5cb1/call-ld_regression/oa_by_smoking_initiation.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-5/LDSC.single_ld_regression_wf/e0d46ded-0994-4103-a632-56e3acb1fb5f/call-ld_regression/oa_by_age_of_initiation.ldsc_regression.log to logs/shard-5/LDSC.single_ld_regression_wf/e0d46ded-0994-4103-a632-56e3acb1fb5f/call-ld_regression/oa_by_age_of_initiation.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-6/LDSC.single_ld_regression_wf/04792a39-1fe2-48a8-9443-b8fff28a9e2b/call-ld_regression/oa_by_alcohol_dependence_.ldsc_regression.log to logs/shard-6/LDSC.single_ld_regression_wf/04792a39-1fe2-48a8-9443-b8fff28a9e2b/call-ld_regression/oa_by_alcohol_dependence_.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-7/LDSC.single_ld_regression_wf/b3cd8be4-c773-4588-a891-536f95171460/call-ld_regression/oa_by_cud.ldsc_regression.log to logs/shard-7/LDSC.single_ld_regression_wf/b3cd8be4-c773-4588-a891-536f95171460/call-ld_regression/oa_by_cud.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-14/LDSC.single_ld_regression_wf/b4368eb6-dfd6-47d2-9e81-b382aee7e72d/call-ld_regression/oa_by_bipolar_disorder.ldsc_regression.log to logs/shard-14/LDSC.single_ld_regression_wf/b4368eb6-dfd6-47d2-9e81-b382aee7e72d/call-ld_regression/oa_by_bipolar_disorder.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-8/LDSC.single_ld_regression_wf/8f262bc7-c2fb-4dc1-891a-2474e931d8ef/call-ld_regression/oa_by_alcohol_dpw.ldsc_regression.log to logs/shard-8/LDSC.single_ld_regression_wf/8f262bc7-c2fb-4dc1-891a-2474e931d8ef/call-ld_regression/oa_by_alcohol_dpw.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-9/LDSC.single_ld_regression_wf/62d49f42-23a2-4c17-a5e0-e0300345c7d7/call-ld_regression/oa_by_lifetime_cannabis_use_ever_vs_never.ldsc_regression.log to logs/shard-9/LDSC.single_ld_regression_wf/62d49f42-23a2-4c17-a5e0-e0300345c7d7/call-ld_regression/oa_by_lifetime_cannabis_use_ever_vs_never.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-16/LDSC.single_ld_regression_wf/8902d590-9f39-4008-b6ed-0b04f311ff7f/call-ld_regression/oa_by_schizophrenia.ldsc_regression.log to logs/shard-16/LDSC.single_ld_regression_wf/8902d590-9f39-4008-b6ed-0b04f311ff7f/call-ld_regression/oa_by_schizophrenia.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-ld_regression/shard-15/LDSC.single_ld_regression_wf/483bbccc-975d-41ca-8874-a0011f430677/call-ld_regression/oa_by_asd.ldsc_regression.log to logs/shard-15/LDSC.single_ld_regression_wf/483bbccc-975d-41ca-8874-a0011f430677/call-ld_regression/oa_by_asd.ldsc_regression.log\n",
      "upload: logs/oa_by_adhd.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_adhd.ldsc_regression.log\n",
      "upload: logs/oa_by_age_of_initiation.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_age_of_initiation.ldsc_regression.log\n",
      "upload: logs/oa_by_hsi.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_hsi.ldsc_regression.log\n",
      "upload: logs/oa_by_smoking_cessation.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_smoking_cessation.ldsc_regression.log\n",
      "upload: logs/oa_by_cpd.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_cpd.ldsc_regression.log\n",
      "upload: logs/oa_by_ptsd.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_ptsd.ldsc_regression.log\n",
      "upload: logs/oa_by_alcohol_dependence_.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_alcohol_dependence_.ldsc_regression.log\n",
      "upload: logs/oa_by_asd.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_asd.ldsc_regression.log\n",
      "upload: logs/oa_by_bipolar_disorder.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_bipolar_disorder.ldsc_regression.log\n",
      "upload: logs/oa_by_anorexia_nervosa.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_anorexia_nervosa.ldsc_regression.log\n",
      "upload: logs/oa_by_lifetime_cannabis_use_ever_vs_never.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_lifetime_cannabis_use_ever_vs_never.ldsc_regression.log\n",
      "upload: logs/oa_by_alcohol_dpw.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_alcohol_dpw.ldsc_regression.log\n",
      "upload: logs/oa_by_cud.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_cud.ldsc_regression.log\n",
      "upload: logs/oa_by_smoking_initiation.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_smoking_initiation.ldsc_regression.log\n",
      "upload: logs/oa_by_depressive_symptoms.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_depressive_symptoms.ldsc_regression.log\n",
      "upload: logs/oa_by_schizophrenia.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_schizophrenia.ldsc_regression.log\n",
      "upload: logs/oa_by_cotinine_levels.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_cotinine_levels.ldsc_regression.log\n",
      "upload: logs/oa_by_mdd.ldsc_regression.log to s3://rti-heroin/ldsc/results/oaall/eur/0006/output/cromwell/oa_by_mdd.ldsc_regression.log\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-plot_ld/PLOT.plot_ld_regression_wf/7282c57b-f5de-44eb-8706-2e35621b770b/call-adj_pvalues/oa_ngc_gsem.ld_regression_results.adj_pvalue.csv to PLOT.plot_ld_regression_wf/7282c57b-f5de-44eb-8706-2e35621b770b/call-adj_pvalues/oa_ngc_gsem.ld_regression_results.adj_pvalue.csv\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-plot_ld/PLOT.plot_ld_regression_wf/7282c57b-f5de-44eb-8706-2e35621b770b/call-make_plot_table/oa_ngc_gsem.ld_regression_results.tsv to PLOT.plot_ld_regression_wf/7282c57b-f5de-44eb-8706-2e35621b770b/call-make_plot_table/oa_ngc_gsem.ld_regression_results.tsv\n",
      "download: s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/19952256-d8d3-4c02-bf05-8c63dae8adde/call-plot_ld/PLOT.plot_ld_regression_wf/7282c57b-f5de-44eb-8706-2e35621b770b/call-plot_ld_regression_results/oa_ngc_gsem.ld_regression_results.pdf to PLOT.plot_ld_regression_wf/7282c57b-f5de-44eb-8706-2e35621b770b/call-plot_ld_regression_results/oa_ngc_gsem.ld_regression_results.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  6089  100  6089    0     0  21216      0 --:--:-- --:--:-- --:--:-- 21216\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    45  100    45    0     0    226      0 --:--:-- --:--:-- --:--:--   226\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "\n",
    "job=\"19952256-d8d3-4c02-bf05-8c63dae8adde\"\n",
    "phen=\"oaall\"\n",
    "project=\"heroin\"\n",
    "version=\"0006\"\n",
    "aws_path=\"s3://rti-heroin/ldsc/results/$phen/eur/$version/output/cromwell/\"\n",
    "\n",
    "\n",
    "mkdir -p ~/Projects/$project/ldsc/$phen/$version/processing/output/cromwell/logs/\n",
    "cd ~/Projects/$project/ldsc/$phen/$version/processing/output/cromwell\n",
    "    \n",
    "# Download output JSON from Swagger UI.\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/outputs\" -H \"accept: application/json\" \\\n",
    "    > final_outputs_${job}.json\n",
    "\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/$job/logs\" -H \"accept: application/json\" \\\n",
    "    > final_logs_${job}.json\n",
    "\n",
    "# Download logs that contain h2\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-ld_regression/ \\\n",
    "    logs/ --recursive --exclude \"*\" --include \"*.ldsc_regression.log\" \n",
    "    \n",
    "mv logs/shard-*/LDSC.single_ld_regression_wf/*/call-ld_regression/*.ldsc_regression.log logs/\n",
    "aws s3 sync logs/ $aws_path\n",
    "rm -rf logs/shard-*\n",
    "\n",
    "\n",
    "# Download plot\n",
    "aws s3 cp s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/$job/call-plot_ld/ \\\n",
    "    . --recursive --exclude \"*\" --include \"*.ld_regression_results.*\"\n",
    "    \n",
    "mv PLOT.plot_ld_regression_wf/*/call-*/*ld_regression_results* .\n",
    "\n",
    "\n",
    "## upload to S3\n",
    "# Projects note that I had to unindent here to get it to run\n",
    "#python - <<EOF\n",
    "#import json\n",
    "#import os\n",
    "#\n",
    "#\n",
    "#def traverse(o, tree_types=(list, tuple)):\n",
    "#    if isinstance(o, tree_types):\n",
    "#        for value in o:\n",
    "#            for subvalue in traverse(value, tree_types):\n",
    "#                yield subvalue\n",
    "#    else:\n",
    "#        yield o\n",
    "#\n",
    "#with open('final_outputs_' + \"$job\" + '.json') as f:\n",
    "#    outputs = json.load(f)\n",
    "#    outputs = outputs[\"outputs\"]\n",
    "#    for key in outputs:\n",
    "#        if (type(outputs[key]) == list):\n",
    "#            for value in traverse(outputs[key]):\n",
    "#                if (str(value)[0:2] == \"s3\"):\n",
    "#                    message = \"aws s3 cp {} $aws_path\".format(value)\n",
    "#                    os.system(message)\n",
    "#        else:\n",
    "#            if (str(outputs[key])[0:2] == \"s3\"):\n",
    "#                message = \"aws s3 cp {} $aws_path\".format(outputs[key])\n",
    "#                os.system(message)\n",
    "#EOF\n",
    "\n",
    "rm -rf PLOT.plot_ld_regression_wf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LD Hub\n",
    "\n",
    "How to prepare/format data for LDHub.\n",
    "\n",
    "```\n",
    "Important notes for your uploaded file:\n",
    "\n",
    "1. To save the uploading time, LD Hub only accepts zipped files as input (e.g. mydata.zip).\n",
    "\n",
    "2. Please check that there is ONLY ONE plain TXT file (e.g. mydata.txt) in your zipped file.\n",
    "\n",
    "3. Please make sure you do NOT zip any folder together with the plain txt file (e.g. /myfolder/mydata.txt), otherwise you will get an error: [Errno 2] No such file or directory\n",
    "\n",
    "4. Please do NOT zip multiple files (e.g. zip mydata.zip file1.txt file2.txt ..) or zip a file with in a folder (e.g. zip mydata.zip /path/to/my/file/mydata.txt).\n",
    "\n",
    "5. Please keep the file name of your plain txt file short (less than 50 characters), otherwise you may get an error: [Errno 2] No such file or directory\n",
    "\n",
    "6. Please zip your plain txt file using following command (ONE file at a time):\n",
    "\n",
    "For Windows system: 1) Locate the file that you want to compress. 2) Right-click the file, point to Send to, and then click Compressed (zipped) folder.\n",
    "\n",
    "For Linux and Mac OS system: zip mydata.zip mydata.txt\n",
    "\n",
    "Reminder: for Mac OS system, please do NOT zip you file by right click mouse and click \"Compress\" to zip your file, this will automatically create a folder called \"__MACOS\". You will get an error: [Errno 2] No such file or directory.\n",
    "\n",
    "Upload the trait of interest\n",
    "To save your upload time, we highly recommend you to use the SNP list we used in LD Hub to reduce the number of SNPs in your uploaded file. Click here to download our SNP list (w_hm3.noMHC.snplist.zip).\n",
    "\n",
    "Please upload the zipped file you just created. Click here to download an input example.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare input file\n",
    "Create LD Hub input file.\n",
    "\n",
    "```css\n",
    "MarkerName      A1      A2      BETA    P    N\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Download outputs for each ref chr from rftm_sumstats step\n",
    "procD=/shared/rti-heroin/ldsc/results/oaall/eur/0006/ \n",
    "mkdir -p $procD/ldhub/\n",
    "cd ldhub\n",
    "\n",
    "aws s3 sync s3://rti-cromwell-output/cromwell-execution/full_ld_regression_wf/1b601278-9a63-496c-a3e5-4859ea396abd/call-munge_ref/MUNGE_REF_WF.munge_sumstats_wf/1686181f-a5f8-4a6c-9b92-bcd227f89af9/call-munge_chr_wf/ .\n",
    "    \n",
    "mv shard-*/MUNGE_CHR.munge_sumstats_chr_wf/*/call-rfmt_sumstats/genomic_sem_gwas_chr*.txt.standardized.phase3ID.munge_ready.txt  .\n",
    "\n",
    "## Concat into single file\n",
    "cp genomic_sem_gwas_chr1.txt.standardized.phase3ID.munge_ready.txt \\\n",
    "oa_ngc_gsem_0006_ld_hub_with_pvalues.txt\n",
    "\n",
    "for chr in {2..22}; do\n",
    "    tail -n +2   genomic_sem_gwas_chr${chr}.txt.standardized.phase3ID.munge_ready.txt >>\\\n",
    "    oa_ngc_gsem_0006_ld_hub_with_pvalues.txt\n",
    "done\n",
    "\n",
    "\n",
    "## Remove unnecessary columns (need snpID, A1, A2, Beta, Pvalue)\n",
    "cat oa_ngc_gsem_0006_ld_hub_with_pvalues.txt |\\\n",
    "cut -f 1,4,5,6,7 > tmp && mv tmp oa_ngc_gsem_0006_ld_hub_with_pvalues.txt\n",
    "\n",
    "## Add sample size column \n",
    "cat oa_ngc_gsem_0006_ld_hub_with_pvalues.txt |\\\n",
    "awk   '{print $1,$2,$3,$4,\"88114.00\",$5}' OFS=\"\\t\" >\\\n",
    "oa_ngc_gsem_0006_ld_hub_final.txt\n",
    "\n",
    "## Use vim to change column names to be:\n",
    "#SNP A1 A2 BETA N P-value\n",
    "\n",
    "## zip file\n",
    "zip oa_ngc_gsem_0006_ld_hub_final.txt.zip oa_ngc_gsem_0006_ld_hub_final.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final Plot\n",
    "We combine the Cromwell/WDL results with the LD Hub updated results with the previous set of genetic correlation results.\n",
    "\n",
    "Upload the plot table to EC2 instance to run docker and create the plot.\n",
    "\n",
    "**Example:**\n",
    "```\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## enter interactive mode ##\n",
    "# note that the image tag corresponds to the latest tag for this image\n",
    "\n",
    "cd /shared/rti-heroin/ldsc/results/oaall/eur/0006/ecr_test2/plot/\n",
    "\n",
    "$(aws ecr get-login --no-include-email --region us-east-1) # docker login\n",
    "docker run -it -v\"$PWD:/data/\" \\\n",
    "    404545384114.dkr.ecr.us-east-1.amazonaws.com/plot_ld_regression_results:v1.0_0f1f25f /bin/bash\n",
    "    \n",
    "inf=20210301_oa_ngc_gsem_ldhub_genetic_correlation.csv\n",
    "outf=20210301_oa_ngc_gsem_ldhub_genetic_correlation.pdf\n",
    "plot_order=oa_ngc_gsem_rg_plot_order.csv\n",
    "\n",
    "\n",
    "Rscript /opt/plot_ld_regression/plot_ld_regression_results.R  \\\n",
    "    --input_file $inf \\\n",
    "    --output_file $outf \\\n",
    "    --group_order_file $plot_order \\\n",
    "    --comma_delimited \\\n",
    "    --vertical_rg 1 \\\n",
    "    --colorblind\n",
    "\n",
    "    #--title \"\" \\"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
