{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a672fdf7-1c48-4a26-9148-1e83d76858a4",
   "metadata": {},
   "source": [
    "# LDSC-SEG on TWAS Meta-Analysis Results for OUD\n",
    "\n",
    "**Author**: Jesse Marks<br>\n",
    "**NIH Project**: [Harnessing Knowledge of Gene Function in Brain Tissue for Discovering Biology Underlying Heroin Addiction](https://reporter.nih.gov/search/RC99reuHhEW0n_3WuFPU6g/project-details/10116351) <br>\n",
    "**Charge Code**: 0215889.001.001<br>\n",
    "**GitHub Issue**:  [Opioid Use Disorder TWAS Meta-analysis (Uniform Processing) #183](https://github.com/RTIInternational/bioinformatics/issues/183)<br>\n",
    "\n",
    "\n",
    "**Description**:<br>\n",
    "This notebook outlines the steps taken to test our TWAS meta-analysis significantly differentially expressed genes for heritability enrichment for 39 phenotypes.\n",
    "\n",
    "perform LD Score Regression (LDSC) Analyses. More specifically, we use the approach described in [Finucane et al. 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5896795/), which is LD score regression applied to specifically expressed genes (LDSC-SEG). LDSC-SEG uses stratified LD score regression to test whether disease heritability is enriched in regions surrounding genes with the highest specific expression in a given tissue. This approach is a powerful way to leverage gene expression data to help interpret GWAS signal. The gene sets we use are from a TWAS meta-analysis of 4 OUD case/control datasets:\n",
    "\n",
    "- [Corradin et al. (2022) Molecular Psychiatry](https://doi.org/10.1038/s41380-022-01477-y)\n",
    "- [Mendez et al. (2021) Molecular Psychiatry](https://doi.org/10.1038/s41380-021-01259-y)\n",
    "- [Seney et al. (2021) Biological Psychiatry](https://doi.org/10.1016/j.biopsych.2021.06.007)\n",
    "- [Sosnowski et al. (2022) Drug and Alcohol Dependence Reports](https://doi.org/10.1016/j.dadr.2022.100040)\n",
    "\n",
    "These meta-analysis results are published in [THIS](https://github.com/RTIInternational/bioinformatics/issues/183#issuecomment-1198311459) GitHub comment. We will use the results that have had singletons removed: genes that only appear in one study. The file name is `meta_analysis_sumstats_no_singletons_20220727.tsv.gz` and is under the Results section.\n",
    "\n",
    "We filter down those results and get two sets of significantly expressed genes. The two sets are:\n",
    "* List of genes with Benjamini-Hochberg FDR <0.05\n",
    "* List of genes with Benjamini-Hochberg FDR <0.10\n",
    "\n",
    "The column named `wfisher_adj_pvalue` is the one we use to create these two gene lists. This column contains the [weighted Fisher's p-value](https://www.nature.com/articles/s41598-021-86465-y) of interest that we will apply the Benjamini-Hochberg FDR thresholds.\n",
    "\n",
    "The GWAS results we use in these LDSC-SEG analyses are the same set used in the paper [Multi-trait genome-wide association study of opioid addiction: OPRM1 and Beyond](https://www.medrxiv.org/content/10.1101/2021.09.13.21263503v1). In particular, we attempted to determine if the significantly expressed genes from the opioid use disorder meta-analysis (including the GENOA GWAS meta-analysis) results showed enrichment in heritability in any of the following EUR-specific studies:\n",
    "\n",
    "___\n",
    "\n",
    "<br><br>\n",
    "\n",
    "<details>\n",
    "    <summary>phenotype list</summary>\n",
    "    \n",
    "* Age of Initiation  (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Alcohol Dependence (Walters et al., 2018 Nat Neurosci [30482948](https://pubmed.ncbi.nlm.nih.gov/30482948))\n",
    "* Alcohol Drinks per Week (DPW) (Liu et al., 2019 Nat Genet [30643251]())\n",
    "* Alzheimer's Disease (Lambert et al., 2013 Nat Genet [24162737](https://pubmed.ncbi.nlm.nih.gov/24162737))\n",
    "* Amyotrophic Lateral Sclerosis (Rheenen et al., 2016 Nat Genet [27455348](https://pubmed.ncbi.nlm.nih.gov/27455348))\n",
    "* Anorexia Nervosa (Watson et al., 2019 Nat Genet [31308545](https://pubmed.ncbi.nlm.nih.gov/31308545))\n",
    "* Attention Deficit Hyperactivity Disorder (Demontis et al., 2019 Nat Genet [30478444]())\n",
    "* Autism Spectrum Disorders (Grove et al., 2019 Nat Genet [30804558](https://pubmed.ncbi.nlm.nih.gov/30804558))\n",
    "* Bipolar Disorder (Stahl et al., 2019 Nat Genet [31043756](https://pubmed.ncbi.nlm.nih.gov/31043756))\n",
    "* Cannabis Use Disorder (CUD) (Demontis et al., 2019 Nat Neurosci [31209380](https://pubmed.ncbi.nlm.nih.gov/31209380))\n",
    "* Childhood IQ (Benyamin et al., 2014 Mol Psychiatry [23358156](https://pubmed.ncbi.nlm.nih.gov/23358156))\n",
    "* Cigarettes Per Day (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* College Completion (Rietveld et al., 2013 Science [23722424](https://pubmed.ncbi.nlm.nih.gov/23722424))\n",
    "* Cotinine Levels Ware et al., 2016 Sci Rep [26833182](https://pubmed.ncbi.nlm.nih.gov/26833182/)\n",
    "* Fagerstrom Test for Nicotine Dependence (FTND) (Quach et al., 2020 Nat Commun [33144568](https://pubmed.ncbi.nlm.nih.gov/33144568/))\n",
    "* Heaviness of Smoking Index (HSI) (Quach et al., 2020 Nat Commun [33144568](https://pubmed.ncbi.nlm.nih.gov/33144568/))\n",
    "* Intelligence (Sniekers et al., 2017 Nat Genet [28530673](https://pubmed.ncbi.nlm.nih.gov/28530673))\n",
    "* Lifetime Cannabis Use (Ever vs. Never) (Pasman et al., 2018 Nat Neurosci [30150663](https://pubmed.ncbi.nlm.nih.gov/30150663))\n",
    "* Major Depressive Disorder (Howard et al., 2018 Nat Commun [29662059](https://pubmed.ncbi.nlm.nih.gov/29662059))\n",
    "* Mean Accumbens Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Caudate Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Hippocampus Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Pallidum Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Putamen Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Mean Thalamus Volume (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Neo-conscientiousness (de Moor et al., 2012 Mol Psychiatry [21173776](https://pubmed.ncbi.nlm.nih.gov/21173776))\n",
    "* Neo-openness to Experience (de Moor et al., 2012 Mol Psychiatry [21173776](https://pubmed.ncbi.nlm.nih.gov/21173776))\n",
    "* Neuroticism (Okbay et al., 2016 Nat Genet [27089181]())\n",
    "* Opioid Addiction: GENOA GWAS meta-analysis\n",
    "* Opioid Addiction: gSEM OA GWAS meta-analysis (i.e., GENOA, MVP-SAGE-YP, PGC-SUD, and Partners Health)\n",
    "* Parkinson's Disease (Sanchez et al., 2009 Nat Genet [19915575](https://pubmed.ncbi.nlm.nih.gov/19915575))\n",
    "* Post-traumatic Stress Disorder (Nievergelt et al., 2019 Nat Commun [31594949](https://pubmed.ncbi.nlm.nih.gov/31594949))\n",
    "* Psychiatric Genetics Consortium Cross-disorder GWAS (Schizophrenia, Bipolar Disorder, MDD, ASD and ADHD) (Cross-Disorder Group of the Psychiatric Genomics Consortium, 2013 Lancet [23453885](https://pubmed.ncbi.nlm.nih.gov/23453885))\n",
    "* Schizophrenia (Ripke et al., 2014 Nature [25056061](https://pubmed.ncbi.nlm.nih.gov/25056061))\n",
    "* Smoking Cessation (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Smoking Initiation (Liu et al., 2019 Nat Genet [30643251](https://pubmed.ncbi.nlm.nih.gov/30643251/))\n",
    "* Subjective Well Being (Okbay et al., 2016 Nat Genet [27089181](https://pubmed.ncbi.nlm.nih.gov/27089181))\n",
    "* Total Intracranial Volume (ICV) (Hibar et al., 2015 Nature [25607358](https://pubmed.ncbi.nlm.nih.gov/25607358/))\n",
    "* Years of Schooling (Okbay et al., 2016 Nature [27225129](https://pubmed.ncbi.nlm.nih.gov/27225129))\n",
    "</details><br><br>\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e67c8a81-412e-452e-996b-4643635e2e89",
   "metadata": {},
   "source": [
    "# Genomic start-end locations for GENCODE v30 IDS\n",
    "Retrieve the genomic start and end locations for the GENCODE v30 IDs for meta-analysis significant genes at FDR <0.05 when singletons are excluded. Meta-analysis summary statistics without singletons are posted in this [comment](https://github.com/RTIInternational/bioinformatics/issues/183#issuecomment-1198311459) as file `meta_analysis_sumstats_no_singletons_20220727.tsv.gz`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "975c6590-0f2c-4fe9-93d1-a7c9075c08ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gencode_id\tgene_name\tbase_mean_expression_corradin\tbase_mean_log2cpm_corradin\tfold_change_corradin\tlog2_fold_change_corradin\tlog2_fold_change_se_corradin\ttest_statistic_corradin\tb_statistic_corradin\tpvalue_corradin\tadjusted_pvalue_corradin\tbase_mean_expression_mendez\tbase_mean_log2cpm_mendez\tfold_change_mendez\tlog2_fold_change_mendez\tlog2_fold_change_se_mendez\ttest_statistic_mendez\tb_statistic_mendez\tpvalue_mendez\tadjusted_pvalue_mendez\tbase_mean_expression_seney\tbase_mean_log2cpm_seney\tfold_change_seney\tlog2_fold_change_seney\tlog2_fold_change_se_seney\ttest_statistic_seney\tb_statistic_seney\tpvalue_seney\tadjusted_pvalue_seney\tbase_mean_expression_sosnowski\tbase_mean_log2cpm_sosnowski\tfold_change_sosnowski\tlog2_fold_change_sosnowski\tlog2_fold_change_se_sosnowski\ttest_statistic_sosnowski\tb_statistic_sosnowski\tpvalue_sosnowski\tadjusted_pvalue_sosnowski\tnum_datasets\tfc_sign_corradin\tfc_sign_mendez\tfc_sign_seney\tfc_sign_sosnowski\twfisher_fc_sign\twfisher_pvalue\twfisher_adj_pvalue\n",
      "ENSG00000000003.14\tTSPAN6\t280.590909090909\t2.15463993595764\t0.972324915595254\t-0.0404896043186799\t0.115514949361746\t-0.350513977129339\t-5.56185774421871\t0.728294376660881\t0.941113912764389\t171.515675\t2.87339596889899\t1.29794088555831\t0.376224677594399\t0.158993917816571\t2.36628345763794\t-3.38865413716201\t0.02332628916534\t0.192483045460433\t151.4161\t3.00917901371786\t1.05677694215658\t0.0796708938403902\t0.0723683795231596\t1.10090752847234\t-5.62032119746489\t0.278135223021147\t0.592633128020453\t176.66272972973\t2.6468452333018\t1.02623331886532\t0.0373587715905427\t0.0519246798255103\t0.719480056806985\t-5.43229896328812\t0.473151309979134\t0.887644640704241\t4\t-1\t1\t1\t1\t1\t0.0913917604863265\t0.32213328944865\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | head -2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dfb81a3f-32f9-4109-9f43-0329ff7c45fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wfisher_adj_pvalue\n",
      "0.32213328944865\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# verify the last column is the wfisher_adj_pvalue column (the Benjamini-Hochberg pvalue)\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | head -2 | awk \\\n",
    "'{print $NF}'\n",
    "\n",
    "# create FDR filtered file: 0.05\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | \\\n",
    "  head -1 > meta_analysis_sumstats_no_singletons_20220727_fdr0.05.tsv\n",
    "\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | tail -n +2 | awk \\\n",
    "'$NF < 0.05 ' >> meta_analysis_sumstats_no_singletons_20220727_fdr0.05.tsv\n",
    "\n",
    "\n",
    "# create FDR filtered file: 0.10\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | \\\n",
    "  head -1 > meta_analysis_sumstats_no_singletons_20220727_fdr0.10.tsv\n",
    "\n",
    "gunzip --to-stdout meta_analysis_sumstats_no_singletons_20220727.tsv.gz | tail -n +2 | awk \\\n",
    "'$NF < 0.10 ' >> meta_analysis_sumstats_no_singletons_20220727_fdr0.10.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f0d69cd-2805-48a0-b118-7dac9b096862",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# download gencode v30 annotation file\n",
    "wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_30/GRCh37_mapping/gencode.v30lift37.annotation.gtf.gz\n",
    "\n",
    "gunzip --to-stdout gencode.v30lift37.annotation.gtf.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3d10ced6-e212-411a-8eff-2f3f3c72b3dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#description: evidence-based annotation of the human genome, version 30 (Ensembl 96), mapped to GRCh37 with gencode-backmap\n",
      "#provider: GENCODE\n",
      "#contact: gencode-help@ebi.ac.uk\n",
      "#format: gtf\n",
      "#date: 2019-04-02\n",
      "chr1\tHAVANA\tgene\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; level 2; havana_gene \"OTTHUMG00000000961.2_2\"; remap_status \"full_contig\"; remap_num_mappings 1; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\ttranscript\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_num_mappings 1; remap_status \"full_contig\"; remap_target_status \"overlap\";\n",
      "chr1\tHAVANA\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 1; exon_id \"ENSE00002234944.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:11869-12227\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t12613\t12721\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 2; exon_id \"ENSE00003582793.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:12613-12721\"; remap_status \"full_contig\";\n",
      "chr1\tHAVANA\texon\t13221\t14409\t.\t+\t.\tgene_id \"ENSG00000223972.5_2\"; transcript_id \"ENST00000456328.2_1\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_name \"DDX11L1-202\"; exon_number 3; exon_id \"ENSE00002312635.1_1\"; level 2; transcript_support_level 1; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2_2\"; havana_transcript \"OTTHUMT00000362751.1_1\"; remap_original_location \"chr1:+:13221-14409\"; remap_status \"full_contig\";\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f2accd5e-b7d4-42a6-89dc-1c5c977009d4",
   "metadata": {},
   "source": [
    "<br><br>\n",
    "\n",
    "get the start and end positions for each gene\n",
    "https://www.gencodegenes.org/pages/data_format.html\n",
    "\n",
    "| column-number   | content                | values/format           |\n",
    "|-----------------|------------------------|-------------------------|\n",
    "| 4               | genomic start location | integer-value (1-based) |\n",
    "| 5               | genomic end location   | integer-value           |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f0e41733-37d5-48cb-b39f-c1fe021f448e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gencode_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>base_mean_expression_corradin</th>\n",
       "      <th>base_mean_log2cpm_corradin</th>\n",
       "      <th>fold_change_corradin</th>\n",
       "      <th>log2_fold_change_corradin</th>\n",
       "      <th>log2_fold_change_se_corradin</th>\n",
       "      <th>test_statistic_corradin</th>\n",
       "      <th>b_statistic_corradin</th>\n",
       "      <th>pvalue_corradin</th>\n",
       "      <th>...</th>\n",
       "      <th>pvalue_sosnowski</th>\n",
       "      <th>adjusted_pvalue_sosnowski</th>\n",
       "      <th>num_datasets</th>\n",
       "      <th>fc_sign_corradin</th>\n",
       "      <th>fc_sign_mendez</th>\n",
       "      <th>fc_sign_seney</th>\n",
       "      <th>fc_sign_sosnowski</th>\n",
       "      <th>wfisher_fc_sign</th>\n",
       "      <th>wfisher_pvalue</th>\n",
       "      <th>wfisher_adj_pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>280.590909</td>\n",
       "      <td>2.154640</td>\n",
       "      <td>0.972325</td>\n",
       "      <td>-0.040490</td>\n",
       "      <td>0.115515</td>\n",
       "      <td>-0.350514</td>\n",
       "      <td>-5.561858</td>\n",
       "      <td>0.728294</td>\n",
       "      <td>...</td>\n",
       "      <td>0.473151</td>\n",
       "      <td>0.887645</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.091392</td>\n",
       "      <td>0.322133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>393.045477</td>\n",
       "      <td>2.667048</td>\n",
       "      <td>0.888755</td>\n",
       "      <td>-0.170143</td>\n",
       "      <td>0.092914</td>\n",
       "      <td>-1.831176</td>\n",
       "      <td>-4.277967</td>\n",
       "      <td>0.076581</td>\n",
       "      <td>...</td>\n",
       "      <td>0.762672</td>\n",
       "      <td>0.959600</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.363189</td>\n",
       "      <td>0.609168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>543.378159</td>\n",
       "      <td>3.128846</td>\n",
       "      <td>0.934389</td>\n",
       "      <td>-0.097905</td>\n",
       "      <td>0.064771</td>\n",
       "      <td>-1.511541</td>\n",
       "      <td>-4.832444</td>\n",
       "      <td>0.140654</td>\n",
       "      <td>...</td>\n",
       "      <td>0.439635</td>\n",
       "      <td>0.876792</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.403248</td>\n",
       "      <td>0.637638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>155.691750</td>\n",
       "      <td>1.329310</td>\n",
       "      <td>0.975023</td>\n",
       "      <td>-0.036492</td>\n",
       "      <td>0.127130</td>\n",
       "      <td>-0.287042</td>\n",
       "      <td>-5.374356</td>\n",
       "      <td>0.775967</td>\n",
       "      <td>...</td>\n",
       "      <td>0.401802</td>\n",
       "      <td>0.865274</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.832877</td>\n",
       "      <td>0.935349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000938.13</td>\n",
       "      <td>FGR</td>\n",
       "      <td>44.590909</td>\n",
       "      <td>-0.829102</td>\n",
       "      <td>0.824932</td>\n",
       "      <td>-0.277653</td>\n",
       "      <td>0.283458</td>\n",
       "      <td>-0.979521</td>\n",
       "      <td>-4.768122</td>\n",
       "      <td>0.334812</td>\n",
       "      <td>...</td>\n",
       "      <td>0.706047</td>\n",
       "      <td>0.949224</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.266319</td>\n",
       "      <td>0.525174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gencode_id gene_name  base_mean_expression_corradin  \\\n",
       "0  ENSG00000000003.14    TSPAN6                     280.590909   \n",
       "1  ENSG00000000419.12      DPM1                     393.045477   \n",
       "2  ENSG00000000457.14     SCYL3                     543.378159   \n",
       "3  ENSG00000000460.17  C1orf112                     155.691750   \n",
       "4  ENSG00000000938.13       FGR                      44.590909   \n",
       "\n",
       "   base_mean_log2cpm_corradin  fold_change_corradin  \\\n",
       "0                    2.154640              0.972325   \n",
       "1                    2.667048              0.888755   \n",
       "2                    3.128846              0.934389   \n",
       "3                    1.329310              0.975023   \n",
       "4                   -0.829102              0.824932   \n",
       "\n",
       "   log2_fold_change_corradin  log2_fold_change_se_corradin  \\\n",
       "0                  -0.040490                      0.115515   \n",
       "1                  -0.170143                      0.092914   \n",
       "2                  -0.097905                      0.064771   \n",
       "3                  -0.036492                      0.127130   \n",
       "4                  -0.277653                      0.283458   \n",
       "\n",
       "   test_statistic_corradin  b_statistic_corradin  pvalue_corradin  ...  \\\n",
       "0                -0.350514             -5.561858         0.728294  ...   \n",
       "1                -1.831176             -4.277967         0.076581  ...   \n",
       "2                -1.511541             -4.832444         0.140654  ...   \n",
       "3                -0.287042             -5.374356         0.775967  ...   \n",
       "4                -0.979521             -4.768122         0.334812  ...   \n",
       "\n",
       "   pvalue_sosnowski  adjusted_pvalue_sosnowski  num_datasets  \\\n",
       "0          0.473151                   0.887645             4   \n",
       "1          0.762672                   0.959600             4   \n",
       "2          0.439635                   0.876792             4   \n",
       "3          0.401802                   0.865274             4   \n",
       "4          0.706047                   0.949224             4   \n",
       "\n",
       "   fc_sign_corradin  fc_sign_mendez  fc_sign_seney  fc_sign_sosnowski  \\\n",
       "0              -1.0             1.0            1.0                1.0   \n",
       "1              -1.0            -1.0            1.0               -1.0   \n",
       "2              -1.0            -1.0            1.0                1.0   \n",
       "3              -1.0             1.0           -1.0               -1.0   \n",
       "4              -1.0             1.0            1.0                1.0   \n",
       "\n",
       "   wfisher_fc_sign  wfisher_pvalue  wfisher_adj_pvalue  \n",
       "0                1        0.091392            0.322133  \n",
       "1               -1        0.363189            0.609168  \n",
       "2               -1        0.403248            0.637638  \n",
       "3               -1        0.832877            0.935349  \n",
       "4                1        0.266319            0.525174  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>annotation_source</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>genomic_start_location</th>\n",
       "      <th>genomic_end_location</th>\n",
       "      <th>score(not_used)</th>\n",
       "      <th>genomic_strand</th>\n",
       "      <th>genomic_phase</th>\n",
       "      <th>additional_information</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; gene_type \"tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5_2\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome_name annotation_source feature_type  genomic_start_location  \\\n",
       "0            chr1            HAVANA         gene                   11869   \n",
       "1            chr1            HAVANA   transcript                   11869   \n",
       "2            chr1            HAVANA         exon                   11869   \n",
       "3            chr1            HAVANA         exon                   12613   \n",
       "4            chr1            HAVANA         exon                   13221   \n",
       "\n",
       "   genomic_end_location score(not_used) genomic_strand genomic_phase  \\\n",
       "0                 14409               .              +             .   \n",
       "1                 14409               .              +             .   \n",
       "2                 12227               .              +             .   \n",
       "3                 12721               .              +             .   \n",
       "4                 14409               .              +             .   \n",
       "\n",
       "                              additional_information  \n",
       "0  gene_id \"ENSG00000223972.5_2\"; gene_type \"tran...  \n",
       "1  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  \n",
       "2  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  \n",
       "3  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  \n",
       "4  gene_id \"ENSG00000223972.5_2\"; transcript_id \"...  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#import pandas as pd\n",
    "\n",
    "#df = pd.read_csv(\"meta_analysis_sumstats_no_singletons_20220727.tsv.gz\", compression=\"gzip\", sep=\"\\t\")\n",
    "#ann = pd.read_csv(\"gencode.v30lift37.annotation.gtf.gz\", sep=\"\\t\", compression=\"gzip\", skiprows=5, header=None)\n",
    "#ann.columns = [\"chromosome_name\", \"annotation_source\", \"feature_type\",\"genomic_start_location\",\"genomic_end_location\",\"score(not_used)\",\"genomic_strand\",\"genomic_phase\", \"additional_information\"]\n",
    "\n",
    "df.head()\n",
    "ann.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "2d96fcc9-fa7d-4118-acaa-5a5eeeca8e79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ENSG00000223972.5': ['chr1', '11869', '14409'], 'ENSG00000227232.5': ['chr1', '14404', '29570']}\n",
      "ENSG00000198521.11\n",
      "ENSG00000277209.1\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "\n",
    "fdr = \"0.05\"\n",
    "in1 = \"meta_analysis_sumstats_no_singletons_20220727.tsv.gz\"\n",
    "out1 = f\"meta_analysis_sumstats_no_singletons_20220727_fdr{fdr}_startend_pos.tsv\"\n",
    "annfile = \"gencode.v30lift37.annotation.gtf.gz\"\n",
    "\n",
    "with gzip.open(in1, 'rt') as inF, gzip.open(annfile, 'rt') as annF, open(out1, 'w') as outF:\n",
    "#with gzip.open(in1, 'rt') as inF, open(out1, 'w') as outF:\n",
    "    for _ in range(5):\n",
    "        next(annF)\n",
    "    line = annF.readline()\n",
    "    \n",
    "    gencode_dic  = {}\n",
    "    while line:\n",
    "        sl = line.split(\"\\t\")\n",
    "        if sl[2] == \"gene\":\n",
    "            gencode = sl[8].split(\";\")[0] # remove all additional_info except \"gene_id <ENSG...>\"\n",
    "            gencode = gencode.split(\" \")[1].strip('\"') # remove \"gene_id\" portion, and double quotes\n",
    "            gencode = gencode.split(\"_\")[0] # remove trailing underscore+number from <ENSG...>\n",
    "            gencode_dic[gencode] = [sl[0], sl[3], sl[4]] # chr, start-, and end-genomic position\n",
    "        line = annF.readline()\n",
    "    \n",
    "    print(dict(list(gencode_dic.items())[0:2]))\n",
    "\n",
    "    outF.write(\"chr\\tgencode_id\\tgenomic_start_location\\tgenomic_end_location\\n\")\n",
    "    next(inF)\n",
    "    line = inF.readline()\n",
    "    while line:\n",
    "        sl = line.split()\n",
    "        gencode = sl[0]\n",
    "        wfisher_adj_p = sl[-1]\n",
    "        if float(wfisher_adj_p) < float(fdr):\n",
    "            if gencode in gencode_dic:\n",
    "                chrom = gencode_dic[gencode][0]\n",
    "                start = gencode_dic[gencode][1]\n",
    "                end = gencode_dic[gencode][2]\n",
    "                outline = f\"{chrom}\\t{gencode}\\t{start}\\t{end}\\n\"\n",
    "                outF.write(outline)\n",
    "            else:\n",
    "                print(gencode)\n",
    "        line = inF.readline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2e1a227e-bc4a-4f85-994c-18df2da48cec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "yes\n"
     ]
    }
   ],
   "source": [
    "dict(list(gencode_dic.items())[0:2])\n",
    "if '\"ENSG00000223972.5' in gencode_dic.keys():\n",
    "    print(\"yes\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
