{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a672fdf7-1c48-4a26-9148-1e83d76858a4",
   "metadata": {},
   "source": [
    "# LDSC-SEG on TWAS Meta-Analysis Results for OUD\n",
    "\n",
    "**Author**: Jesse Marks<br>\n",
    "**NIH Project**: [Harnessing Knowledge of Gene Function in Brain Tissue for Discovering Biology Underlying Heroin Addiction](https://reporter.nih.gov/search/RC99reuHhEW0n_3WuFPU6g/project-details/10116351) <br>\n",
    "**Charge Code**: 0215889.001.001<br>\n",
    "**GitHub Issue**:  [Opioid Use Disorder TWAS Meta-analysis (Uniform Processing) #183](https://github.com/RTIInternational/bioinformatics/issues/183)<br>\n",
    "\n",
    "\n",
    "**Description**:<br>\n",
    "This notebook outlines the steps taken to perform LD Score Regression (LDSC) Analyses. More specifically, we use the approach described in [Finucane et al. 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5896795/), which is LD score regression applied to specifically expressed genes (LDSC-SEG). LDSC-SEG uses stratified LD score regression to test whether disease heritability is enriched in regions surrounding genes with the highest specific expression in a given tissue. This approach is a powerful way to leverage gene expression data to help interpret GWAS signal. The gene sets we use are from a TWAS meta-analysis of 4 OUD case/control datasets:\n",
    "\n",
    "- [Corradin et al. (2022) Molecular Psychiatry](https://doi.org/10.1038/s41380-022-01477-y)\n",
    "- [Mendez et al. (2021) Molecular Psychiatry](https://doi.org/10.1038/s41380-021-01259-y)\n",
    "- [Seney et al. (2021) Biological Psychiatry](https://doi.org/10.1016/j.biopsych.2021.06.007)\n",
    "- [Sosnowski et al. (2022) Drug and Alcohol Dependence Reports](https://doi.org/10.1016/j.dadr.2022.100040)\n",
    "\n",
    "These meta-analysis results are published in [THIS](https://github.com/RTIInternational/bioinformatics/issues/183#issuecomment-1198311459) GitHub comment. We will use the results that have had singletons removed: genes that only appear in one study. The file name is `meta_analysis_sumstats_no_singletons_20220727.tsv.gz` and is under the Results section.\n",
    "\n",
    "We filter down those results and get two sets of significantly expressed genes. The two sets are:\n",
    "* List of genes with Benjamini-Hochberg FDR <0.05\n",
    "* List of genes with Benjamini-Hochberg FDR <0.10\n",
    "\n",
    "The column named `wfisher_adj_pvalue` is the one we use to create these two gene lists. This column contains the [weighted Fisher's p-value](https://www.nature.com/articles/s41598-021-86465-y) of interest that we will apply the Benjamini-Hochberg FDR thresholds.\n",
    "\n",
    "The GWAS results we use in these LDSC-SEG analyses are the same set used in the paper [Multi-trait genome-wide association study of opioid addiction: OPRM1 and Beyond](https://www.medrxiv.org/content/10.1101/2021.09.13.21263503v1). In particular, we attempted to determine if the significantly expressed genes from the opioid use disorder meta-analysis showed enrichment in heritability in any of the following studies:\n",
    "\n",
    "intracranial volume\n",
    "Mean Accumbens \n",
    "Mean Caudate \n",
    "Mean Hippocampus \n",
    "Mean Pallidum \n",
    "Mean Putamen \n",
    "Mean Thalamus \n",
    "HSI \n",
    "CPD \n",
    "Smoking cessation \n",
    "Cotinine levels \n",
    "Smoking initiation \n",
    "Age of initiation \n",
    "FTND \n",
    "Childhood IQ \n",
    "Years of schooling \n",
    "Intelligence \n",
    "College completion \n",
    "Alcohol dependence  \n",
    "CUD \n",
    "Alcohol DPW \n",
    "Lifetime cannabis use (ever vs never) \n",
    "Alzheimers disease \n",
    "Parkinsons disease \n",
    "Amyotrophic lateral sclerosis \n",
    "Neo-conscientiousness \n",
    "Neo-openness to experience \n",
    "Neuroticism \n",
    "PTSD \n",
    "ADHD \n",
    "Depressive Symptoms \n",
    "MDD \n",
    "Bipolar disorder \n",
    "ASD \n",
    "Schizophrenia \n",
    "Anorexia nervosa \n",
    "PGC cross-disorder analysis \n",
    "Subjective well being\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4b00db8-4336-407f-a2df-51088554478f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R-jesse",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
