{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bb81f4b-3e4c-4cde-a669-6b8a61a1cfb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget https://repo.anaconda.com/archive/Anaconda3-2020.02-Linux-x86_64.sh\n",
    "bash Anaconda3-2020.02-Linux-x86_64.sh\n",
    "# after installation completes, allow the installer to initialize Anaconda3\n",
    "source ~/.bashrc\n",
    "conda --version # to verify installation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e1e7e10-2e40-4cb5-b9f5-501a12f754d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "./deploy-cloudformation-batch.sh \\\n",
    "    --projectNumber 0217761.000.001 \\\n",
    "    --projectShortName \"Cannibis-Fang\" \\\n",
    "    --cpuMax 500 \\\n",
    "    --stackName cromwell-64-7b7bfb4\n",
    "\n",
    "\n",
    "./deploy-cloudformation-batch.sh \\\n",
    "    --projectNumber 0160470.000.043 \\\n",
    "    --projectShortName \"EOJ-fellows-fund\" \\\n",
    "    --cpuMax 500 \\\n",
    "    --stackName cromwell-64-7b7bfb4\n",
    "\n",
    "\n",
    "bash ./deploy-cloudformation-batch.sh \\\n",
    "    --projectNumber 0217653.001.001 \\\n",
    "    --projectShortName \"hiv-gnetii\" \\\n",
    "    --cpuMax 500 \\\n",
    "    --stackName cromwell-64-7b7bfb4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4adc55da-0f5d-4729-be4b-517565d7e782",
   "metadata": {},
   "outputs": [],
   "source": [
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/shared/jmarks/myWorkflow.wdl\" \\\n",
    "    -F \"workflowOptions=@//home/ubuntu/bin/cromwell/cannabis_charge_code.json\"\n",
    "\n",
    "job=230954e7-3ded-4aad-88ff-4e2b0f5c0b74\n",
    "job=36255589-5af7-4816-af9c-8a057d619988\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9adf6299-c40e-4f2f-abcd-22c901a4623b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## run ldsc wf through cromwell API call to cromwell server\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@$procD/ld-regression-pipeline/workflow/full_ld_regression_wf.wdl\" \\\n",
    "    -F \"workflowInputs=@$procD/ld-regression-pipeline/workflow_inputs/final_wf_inputs.json\" \\\n",
    "    -F \"workflowDependencies=@$procD/ld-regression-pipeline/workflow_inputs/ld-regression-pipeline.zip\" \\\n",
    "    -F \"workflowOptions=@//home/ubuntu/bin/cromwell/cannabis_charge_code.json\"\n",
    "\n",
    "#job=995cf759-372b-4201-9ec2-b0f92174273a\n",
    "#job=2f0541e5-a71e-4352-8d45-108d1b8d7cf9\n",
    "job=8fc04851-7b8e-4847-a6c7-76f3fe8eefcc\n",
    "curl -X GET \"http://localhost:8000/api/workflows/v1/${job}/status\"   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a86006da-364a-4645-9e48-05556fbd696a",
   "metadata": {},
   "outputs": [],
   "source": [
    "java -jar ~/bin/cromwell/cromwell-54.jar run ewas_meta_preprocessing inputs inputs.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c461170-b6ce-4785-87b3-cb3419af0989",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /shared/jmarks/\n",
    "zip --exclude=*var/* --exclude=*.git/* -r biocloud_gwas_workflows/meta_analysis/metal/ewas/biocloud_gwas_workflows.zip biocloud_gwas_workflows/\n",
    "\n",
    "curl -X POST \"http://localhost:8000/api/workflows/v1\" -H \"accept: application/json\" \\\n",
    "    -F \"workflowSource=@/shared/jmarks/biocloud_gwas_workflows/meta_analysis/metal/ewas/full_ewas_meta.wdl\" \\\n",
    "    -F \"workflowInputs=@/home/ubuntu/inputs.json\" \\\n",
    "    -F \"workflowDependencies=@/shared/jmarks/biocloud_gwas_workflows/meta_analysis/metal/ewas/biocloud_gwas_workflows.zip\" \\\n",
    "    -F \"workflowOptions=@/home/ubuntu/bin/cromwell/cannabis_charge_code.json\"\n",
    "\n",
    "job=0f5e61fc-7501-4a45-a74d-8e10276a4ab6 # failed\n",
    "job2=e4aba847-9784-4d1b-bc22-18f0105b6bfd # failed\n",
    "job3=121cecfd-046c-417f-95d3-90225dc833c5 # no charge code: succeeded\n",
    "job4=6453dc41-d8df-45b0-9ca9-0ab978210b18 # failed, but made progress\n",
    "job5=838896f8-3826-4d41-8f7a-cbe1b5052146 # failed\n",
    "job=010dc3e5-f49f-4169-9e87-f0d679fec1fb # running\n",
    "\n",
    "for ff in {$job}; do\n",
    "    echo \n",
    "    curl -X GET \"http://localhost:8000/api/workflows/v1/$ff/status\"\n",
    "    echo \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3dac3ae-65fa-4c41-8f1c-f0a428a15116",
   "metadata": {},
   "outputs": [],
   "source": [
    "HIV GNetii R01: 0217653.001.001\n",
    "    \n",
    "./deploy-cloudformation-batch.sh \\\n",
    "    --projectNumber 0217653.001.001 \\\n",
    "    --projectShortName \"HIV GNetii\" \\\n",
    "    --cpuMax 500 \\\n",
    "    --stackName cromwell-64-7b7bfb4\n",
    "\n",
    "pcluster create \\\n",
    "  -c ~/.parallelcluster/config \\\n",
    "  -t default_spot \\\n",
    "  -g '{ \"project-number\" : \"0217653.001.001\" }' \\\n",
    "      j-test2-sge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d7c42df-e742-43e0-8cf0-8a63f4f41f75",
   "metadata": {},
   "source": [
    "# UHS1â€“4 data share\n",
    "use presign urls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4ffaafd-8317-439d-b5b2-175bf9d0e493",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EUR\n",
    "touch uhs1234-eur-presigned-urls.txt\n",
    "echo -e \"## UHS1234 phenotype data\" | tee uhs1234-eur-presigned-urls.txt >/dev/null\n",
    "#aws s3 presign s3://rti-hiv/gwas/uhs1234/data/acquisition/0010/phenotype/uhs1234_eur_hivstat_gwassex_age_pcs.txt \\\n",
    "aws s3 presign s3://rti-shared/shared_data/post_qc/uhs1234/phenotype/0010/dbGaP_phenotypeDD_20210714.xlsx \\\n",
    "    --expires-in 604800 | tee --append uhs1234-eur-presigned-urls.txt >/dev/null\n",
    "aws s3 presign s3://rti-shared/shared_data/post_qc/uhs1234/phenotype/0010/dbGaP_phenotypeDS_20210714.txt \\\n",
    "    --expires-in 604800 | tee --append uhs1234-eur-presigned-urls.txt >/dev/null\n",
    "    \n",
    "echo -e \"\\n\\n## raw genotype array data\" | tee --append uhs1234-eur-presigned-urls.txt >/dev/null\n",
    "aws s3 presign s3://rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0010/final/eur/uhs1234_eur_all_chromosomes.zip \\\n",
    "    --expires-in 604800 | tee --append uhs1234-eur-presigned-urls.txt >/dev/null\n",
    "    \n",
    "echo -e \"\\n\\n## imputed genotype data\" | tee --append uhs1234-eur-presigned-urls.txt >/dev/null\n",
    "path=s3://rti-shared/shared_data/post_qc/uhs1234/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0010/eur/\n",
    "for file in $(aws s3 ls $path | awk -v mypath=$path '{print mypath$4}'); do\n",
    "    aws s3 presign $file --expires-in 604800 | tee --append uhs1234-eur-presigned-urls.txt > /dev/null\n",
    "done\n",
    "\n",
    "\n",
    "## AFR\n",
    "touch uhs1234-afr-presigned-urls.txt\n",
    "echo -e \"## UHS1234 phenotype data\" | tee uhs1234-afr-presigned-urls.txt >/dev/null\n",
    "#aws s3 presign s3://rti-hiv/gwas/uhs1234/data/acquisition/0010/phenotype/uhs1234_afr_hivstat_gwassex_age_pcs.txt \\\n",
    "aws s3 presign s3://rti-shared/shared_data/post_qc/uhs1234/phenotype/0010/dbGaP_phenotypeDD_20210714.xlsx \\\n",
    "    --expires-in 604800 | tee --append uhs1234-afr-presigned-urls.txt >/dev/null\n",
    "aws s3 presign s3://rti-shared/shared_data/post_qc/uhs1234/phenotype/0010/dbGaP_phenotypeDS_20210714.txt \\\n",
    "    --expires-in 604800 | tee --append uhs1234-afr-presigned-urls.txt >/dev/null\n",
    "    \n",
    "echo -e \"\\n\\n## raw genotype array data\" | tee --append uhs1234-afr-presigned-urls.txt >/dev/null\n",
    "aws s3 presign s3://rti-shared/shared_data/post_qc/uhs1234/genotype/array/observed/0010/final/afr/uhs1234_afr_all_chromosomes.zip \\\n",
    "    --expires-in 604800 | tee --append uhs1234-afr-presigned-urls.txt >/dev/null\n",
    "    \n",
    "echo -e \"\\n\\n## imputed genotype data\" | tee --append uhs1234-afr-presigned-urls.txt >/dev/null\n",
    "path=s3://rti-shared/shared_data/post_qc/uhs1234/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0010/afr/\n",
    "for file in $(aws s3 ls $path | awk -v mypath=$path '{print mypath$4}'); do\n",
    "    aws s3 presign $file --expires-in 604800 | tee --append uhs1234-afr-presigned-urls.txt > /dev/null\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf87f2b9-78ff-420b-800a-c9d27213f44b",
   "metadata": {},
   "source": [
    "# McLaren"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf2e32a2-acfc-436e-8b46-7b2da9a80f10",
   "metadata": {},
   "outputs": [],
   "source": [
    "touch mclaren-eur-presigned-urls.txt\n",
    "path=s3://rti-hiv/gwas/mclaren/original/\n",
    "    \n",
    "echo -e \"## McLaren results\" | tee mclaren-eur-presigned-urls.txt >/dev/null\n",
    "    \n",
    "for file in $(aws s3 ls $path | awk -v mypath=$path '{print mypath$4}'); do\n",
    "    aws s3 presign $File --expires-in 604800 | tee --append mclaren-eur-presigned-urls.txt > /dev/null\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6aa507-d2da-4953-8c5d-05cce7f5e5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "file=s3://rti-hic/shared_data/post_qc/cohort_0001/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0003/amr/output/casewestern_amr_imputed_genotypes.tgz\n",
    "aws s3 presign $file --expires-in 604800\n",
    "# https://rti-hic.s3.amazonaws.com/shared_data/post_qc/cohort_0001/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0003/amr/output/casewestern_amr_imputed_genotypes.tgz?AWSAccessKeyId=AKIAJONBCJHOJSW2PFJA&Expires=1635176060&Signature=y%2BIVlvt3Cj1RgbUeV61Q7KVXIpw%3D\n",
    "\n",
    "file=s3://rti-hic/shared_data/post_qc/cohort_0001/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0003/afr/output/casewestern_afr_imputed_genotypes.tgz\n",
    "aws s3 presign $file --expires-in 604800\n",
    "# https://rti-hic.s3.amazonaws.com/shared_data/post_qc/cohort_0001/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0003/afr/output/casewestern_afr_imputed_genotypes.tgz?AWSAccessKeyId=AKIAJONBCJHOJSW2PFJA&Expires=1635176117&Signature=ArEWw2nc66yV%2Fv%2BrEDCk1nsc6cM%3D\n",
    "\n",
    "file=s3://rti-hic/shared_data/post_qc/cohort_0001/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0003/eur/output/casewestern_eur_imputed_genotypes.tgz\n",
    "aws s3 presign $file --expires-in 604800\n",
    "# https://rti-hic.s3.amazonaws.com/shared_data/post_qc/cohort_0001/genotype/array/imputed/minimac4_eagle2.4/topmed_r2/0003/eur/output/casewestern_eur_imputed_genotypes.tgz?AWSAccessKeyId=AKIAJONBCJHOJSW2PFJA&Expires=1635176143&Signature=OaSrG9LRSJeJiYeCaKAqJ8qm14U%3D"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
