{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Procedures for downloading data from dbGaP and moving to S3\n",
    "** Author: ** Jesse Marks \n",
    "\n",
    "This notebook outlines the data \n",
    "Note: be sure to remove studies from local machine once they have been uploaded to s3. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Directory Structure for Amazon Web Services (AWS) S3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    rti-common\n",
    "    |\n",
    "    |-- dbgap_studies\n",
    "    |    |\n",
    "    |     ` -- <dbGaP study name>\n",
    "    |    |    |\n",
    "    |    |     ` -- meta\n",
    "    |    |    |\n",
    "    |    |     ` -- genotype\n",
    "    |    |    |    |    \n",
    "    |    |    |     ` -- original\n",
    "    |    |    |    |     |    \n",
    "    |    |    |    |      ` -- unprocessed\n",
    "    |    |    |    |     |    \n",
    "    |    |    |    |      ` -- processing\n",
    "    |    |    |    |     |    \n",
    "    |    |    |    |      ` -- final\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- imputed\n",
    "    |    |    |\n",
    "    |    |     ` -- phenotype\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- unprocessed\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- processing\n",
    "    |    |    |    |\n",
    "    |    |    |     ` -- final "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Install Amazon Web Services Command Line Interface (AWS CLI) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Install pip, a package manager for python applications on local machine\n",
    "curl \"https://bootstrap.pypa.io/get-pip.py\" -o \"get-pip.py\"\n",
    "python get-pip.py\n",
    "\n",
    "# Install awscli via pip\n",
    "pip install awscli\n",
    "\n",
    "# Verify installation - should see something similar to the string below\n",
    "aws --version\n",
    "\"aws-cli/1.11.178 Python/2.7.13 CYGWIN_NT-10.0/2.9.0(0.318/5/3) botocore/1.7.36\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configure AWS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Local Machine\n",
    "aws configure\n",
    "\n",
    "AWS Access Key ID [None]: AKIAJONBCJHOJSW2PFJA\n",
    "AWS Secret Access Key [None]:  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\n",
    "Default region name [None]: us-east-1\n",
    "Default output format [None]: text  # could be json, text, or table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Population Controls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000206_pancreatic \n",
    "\n",
    "Genome-wide association study of heroin abuse using public controls\n",
    "CGEMS Pancreatic Cancer (PanScan) (phs000206.v5.p3)\n",
    "\n",
    "* Download 58549"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Local machine\n",
    "\n",
    "# Decrypt (must be executed within project dir) \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58549\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58527/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure (on local)\n",
    "mkdir -p s3/phs000206_pancreatic/{genotype,phenotype,meta}\n",
    "mkdir -p s3/phs000206_pancreatic/genotype/original/{final,processing,unprocessed}\n",
    "mkdir s3/phs000206_pancreatic/phenotype/{final,processing,unprocessed}\n",
    "\n",
    "# Move files/data to appropriate \"unprocessed\" folders\n",
    "cd /58549/PhenoGenotypeFiles/RootStudyConsentSet_phs000206.CGEMS_PanScan.v5.p3.c1.CADM/\n",
    "mv /PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/phenotype/unprocessed/\n",
    "mv /GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/genotype/original/unprocessed/\n",
    "mv /Study\\ MetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/meta/\n",
    "\n",
    "# Start processing the files\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000206_pancreatic/\n",
    "cp phenotype/unprocessed/* phenotype/processing\n",
    "cp genotype/original/unprocessed/* genotype/original/processing/\n",
    "\n",
    "# unzip and rename\n",
    "cd phenotype/processing/\n",
    "for i in *.gz; do gunzip $i; done\n",
    "for i in *.gz; do mv $i $(echo $i|perl -pi -e s/phs0.+CAN_//g); done\n",
    "for i in *.xml; do mv $i $(echo $i|perl -pi -e s/phs000.+CAN_//g); done\n",
    "\n",
    "cd ../../genotype/original/processing/\n",
    "for i in *.gz; do tar -xzvf $i; done\n",
    "rm *.gz\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000206_pancreatic/ s3://rti-common/dbGaP/phs000206_pancreatic/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000199_chop-ctrl\n",
    "CHOP Control CNV Study (phs000199.v1.p1)\n",
    "* Download 58599"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Local machine\n",
    "\n",
    "# Decrypt (must be executed within project dir) \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58599\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58599/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000199_chop-ctrl/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000199_chop-ctrl/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000199_chop-ctrl/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "\n",
    "# Move files/data to appropriate \"unprocessed\" folders\n",
    "/cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58599/NIGMS/CHOP_CNV/phs000199v1/p1/\n",
    "for i in *.pdf; do mv $i /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/meta/; done\n",
    "for i in phs*; do mv $i /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/phenotype/unprocessed/; done\n",
    "for i in phg000059v1/*; do mv $i /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/genotype/original/unprocessed/; done\n",
    "\n",
    "# unzip & rename genotype\n",
    "cd genotype/original/\n",
    "cp unprocessed/ processing/\n",
    "for i in processing/*.gz; do tar -xzvf $i; done\n",
    "rm *.gz\n",
    "mv phg000059.v1.p1.CHOP_CNV.individual-info.MULTI.txt individual-info.MULTI.txt\n",
    "\n",
    "# unzip & rename phenotype\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000199_chop-ctrl/phenotype\n",
    "cp unprocessed/* processing/\n",
    "cd processing/\n",
    "for i in *.txt.gz; do gunzip $i; done\n",
    "for i in *.gz; do tar -xzvf $i; done$ \n",
    "rm *.gz\n",
    "for i in *; do mv $i $(echo $i | perl -pi -e s/phs0.+CNV_//g); done\n",
    "\n",
    "# upload to S3\n",
    "aws s3 mv phs000199_chop-ctrl s3://rti-common/dbGaP/phs000199_chop-ctrl/ --recursive\n",
    "rm -rf phs000199_chop-ctrl/*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000209_MESA-c1\n",
    "MESA Cohort (phs000209.v13.p3.c1)\n",
    "* Download 58601"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Decrypt (must be executed within project dir) \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58601\n",
    "\n",
    "\n",
    "if [ $(find 58601/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000209_MESA/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000209_MESA/genotype/original/{final,processing,unprocessed}\n",
    "mkdir -p s3/phs000209_MESA/phenotype/{final,processing,unprocessed}\n",
    "\n",
    "# mv files to appropriate directories\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58601/PhenoGenotypeFiles/\\\n",
    "  RootStudyConsentSet_phs000209.MESA.v13.p3.c1.HMB/\n",
    "mv StudyFiles* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/meta/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "\n",
    "cd ../ChildStudyConsentSet_phs000420.MESA.v6.p3.c1.HMB/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/genotype/original/unprocessed\n",
    "\n",
    "cd ../ChildStudyConsentSet_phs000403.MESA.v3.p3.c1.HMB/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/genotype/original/unprocessed\n",
    "\n",
    "cd ../ChildStudyConsentSet_phs000283.MESA.v7.p3.c1.HMB/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/phenotype/unprocessed\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA/genotype/original/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "aws s3 mv phs000209_MESA/ s3://rti-common/dbGaP/phs000209_MESA/ --recursive\n",
    "    \n",
    "    \n",
    "## NOTE: had to rename this study to \"phs000209_MESA-c1\" from \"phs000209_MESA\"\n",
    "## here is the command to perform this action\n",
    "aws s3 cp s3://rti-common/dbGaP/phs000209_MESA/ s3://rti-common/dbGaP/phs000209_MESA-c1/ --recursive\n",
    "aws s3 rm s3://rti-common/dbGaP/phs000209_MESA/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000209_MESA-c2\n",
    "\n",
    "Multi-Ethnic Study of Atherosclerosis (MESA) Cohort \n",
    "(phs000209.v13.p3.c2)\n",
    "\n",
    "* Download 58619"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Decrypt (must be executed within project dir) \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58619\n",
    "\n",
    "if [ $(find 58619/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000209_MESA-c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000209_MESA-c2/genotype/original/{final,processing,unprocessed}\n",
    "mkdir -p s3/phs000209_MESA-c2/phenotype/{final,processing,unprocessed}\n",
    "\n",
    "# mv files to appropriate directories\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58619/PhenoGenotypeFiles/\\\n",
    "  RootStudyConsentSet_phs000209.MESA.v13.p3.c2.MB/\n",
    "mv StudyFiles* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/meta/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "\n",
    "cd ../ChildStudyConsentSet_phs000420.MESA.v6.p3.c2.HMB-NPU/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/genotype/original/unprocessed\n",
    "\n",
    "cd ../ChildStudyConsentSet_phs000403.MESA.v3.p3.c2.HMB-NPU/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/genotype/original/unprocessed\n",
    "\n",
    "cd ../ChildStudyConsentSet_phs000283.MESA.v7.p3.c2.HMB-NPU/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/phenotype/unprocessed\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000209_MESA-c2/genotype/original/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3\n",
    "aws s3 cp phs000209_MESA-c2/ s3://rti-common/dbGaP/phs000209_MESA-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000169_visc-adiposity\n",
    "Whole Genome Association Study of Visceral Adiposity in the HABC Study(phs000169.v1.p1)\n",
    "* Download 58623"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58623\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58601/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# resolve problematic decryption\n",
    "for i in $(find ./58623/ -name *ncbi_enc -print)\n",
    "do      \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\"\n",
    "mv ${i} ${j}_tmp.ncbi_enc\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\"\n",
    "mv ${j}_tmp ${i/.ncbi_enc/}\n",
    "let j=$j+1\n",
    "done\n",
    "\n",
    "# double check for successful decryption\n",
    "if [ $(find 58601/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000169_visc-adiposity/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000169_visc-adiposity/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000169_visc-adiposity/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# copy files \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58623/JAAMH/CIDR_VisceralAdiposity/phs000169v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000169_visc-adiposity/meta/\n",
    "mv *.gz /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000169_visc-adiposity/phenotype/unprocessed/\n",
    "mv phg000052v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000169_visc-adiposity/genotype/original/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 cp phs000169_visc-adiposity s3://rti-common/dbGaP/phs000169_visc-adiposity/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000170_cataract-lowHDL\n",
    "eMERGE Genome Wide Association Study of Cataract and Low HDL cholesterol(phs000170.v2.p1)\n",
    "* Download 58626\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58626\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58626/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000170_cataract-lowHDL/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000170_cataract-lowHDL/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000170_cataract-lowHDL/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# copy files\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58626/PhenoGenotypeFiles/\\\n",
    "RootStudyConsentSet_phs000170.eMERGE_Cataracts.v2.p1.c1.HMB-GSO\n",
    "mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000170_cataract-lowHDL/meta/\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000170_cataract-lowHDL/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000170_cataract-lowHDL/phenotype/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP/s3/\n",
    "aws s3 cp phs000170_cataract-lowHDL/ s3://rti-common/dbGaP/phs000170_cataract-lowHDL/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### phs000287_cardio-health_c1\n",
    "Cardiovascular Health Study (CHS) (phs000287.v6.p1.c2)\n",
    "* Download 58627"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58627\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58627/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000287_cardio-health/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000287_cardio-health/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000287_cardio-health/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "cd 58627/PhenoGenotypeFiles/\n",
    "\n",
    "# copy files - meta\n",
    "cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/StudyMetaFiles/* ../../s3/phs000287_cardio-health/meta/\n",
    "\n",
    "# copy files - phenotype\n",
    "cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c1.HMB-IRB-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health/phenotype/unprocessed/\n",
    "\n",
    "# copy files - genotype\n",
    "cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/       \n",
    "cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c1.HMB-IRB-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 cp phs000287_cardio-health/ s3://rti-common/dbGaP/phs000287_cardio-health/ --recursive\n",
    "    \n",
    "# had to rename this folder on s3: appended \"_c1\"\n",
    "aws s3 mv s3://rti-common/dbGaP/phs000287_cardio-health/ s3://rti-common/dbGaP/phs000287_cardio-health_c1/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000287_cardio-health_c2\n",
    "Cardiovascular Health Study (CHS) (phs000287.v6.p1.c2)\n",
    "* Download 58634"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58634\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58634/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# resolve problematic decryption\n",
    "for i in $(find ./58634/ -name *ncbi_enc -print); do\n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\"; mv ${i} ${j}_tmp.ncbi_enc; \n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc;\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\";\n",
    "mv ${j}_tmp ${i/.ncbi_enc/};\n",
    "let j=$j+1;\n",
    "done\n",
    "\n",
    "# verify the decryption resolution was successful\n",
    "if [ $(find 58634/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000287_cardio-health_c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000287_cardio-health_c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000287_cardio-health_c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "\n",
    "# copy meta files \n",
    "cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/StudyMetaFiles/* ../../s3/phs000287_cardio-health_c2/meta/\n",
    "\n",
    "# copy phenotype files\n",
    "cp RootStudyConsentSet_phs000287.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c2.HMB-IRB-NPU-MDS/PhenotypeFiles/* ../../s3/phs000287_cardio-health_c2/phenotype/unprocessed/\n",
    "\n",
    "# copy genotype files\n",
    "cp ChildStudyConsentSet_phs000226.CHS_CHS.v6.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000301.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000377.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000400.CHS_CHS.v5.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "cp ChildStudyConsentSet_phs000667.CHS_CHS.v3.p1.c2.HMB-IRB-NPU-MDS/GenotypeFiles/* ../../s3/phs000287_cardio-health_c2/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd ../../s3/\n",
    "aws s3 mv phs000287_cardio-health_c2/ s3://rti-common/dbGaP/phs000287_cardio-health_c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000289_thrombosis\n",
    "GENEVA GWAS of Venous Thrombosis (phs000289.v2.p1)\n",
    "* Download 58637"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58637\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58637/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000289_thrombosis/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000289_thrombosis/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000289_thrombosis/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move data to the directory structure that was created\n",
    "cd 58637/NHGRI/VenousThrombosis/phs000289v2/p1/\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000289_thrombosis/meta/\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000289_thrombosis/phenotype/unprocessed/\n",
    "mv genotype/phg000088v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000289_thrombosis/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000289_thrombosis/ s3://rti-common/dbGaP/phs000289_thrombosis/ --recursive\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000285_CARDIA-c1\n",
    "CARDIA_Cohort (phs000285.v3.p2.c1)\n",
    "* Download 58673"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58673\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58673/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000285_CARDIA-c1/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000285_CARDIA-c1/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000285_CARDIA-c1/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv data to the directory structure that was created\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58673/PhenoGenotypeFiles/\n",
    "mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c1.HMB-IRB/StudyFiles/* ../../s3/phs000285_CARDIA-c1/meta/\n",
    "mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000285_CARDIA-c1/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c1/phenotype/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000285_CARDIA-c1/ s3://rti-common/dbGaP/phs000285_CARDIA-c1/ --recursive "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000285_CARDIA-c2\n",
    "CARDIA_Cohort (phs000285.v3.p2.c2)\n",
    "* Download 58657"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58657\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58657/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000285_CARDIA-c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000285_CARDIA-c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000285_CARDIA-c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58657/PhenoGenotypeFiles/\n",
    "mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c2.HMB-IRB-NPU/StudyFiles/* ../../s3/phs000285_CARDIA-c2/meta/\n",
    "mv RootStudyConsentSet_phs000285.CARDIA.v3.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000236.CARDIA.v2.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000309.CARDIA.v3.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000399.CARDIA.v1.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c2.HMB-IRB-NPU/GenotypeFiles/* ../../s3/phs000285_CARDIA-c2/genotype/original/unprocessed\n",
    "mv ChildStudyConsentSet_phs000613.CARDIA.v1.p2.c2.HMB-IRB-NPU/PhenotypeFiles/* ../../s3/phs000285_CARDIA-c2/phenotype/unprocessed\n",
    "\n",
    "cd /cygdrive/c/User/jmarks/ncbi/dbGaP/s3/\n",
    "aws s3 mv phs000285_CARDIA-c2 s3://rti-common/dbGaP/phs000285_CARDIA-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000280_atherosclerosis\n",
    "Atherosclerosis Risk in Communities (ARIC) Cohort (phs000280.v4.p1)\n",
    "* Download 58680"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58680\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58680/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# resolve problematic decryption\n",
    "for i in $(find ./58680/ -name *ncbi_enc -print)\n",
    "do      \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\"\n",
    "mv ${i} ${j}_tmp.ncbi_enc\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\"\n",
    "mv ${j}_tmp ${i/.ncbi_enc/}\n",
    "let j=$j+1\n",
    "done\n",
    "\n",
    "# Double check to make sure all files were decrypted\n",
    "if [ $(find 58680/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000280_atherosclerosis/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000280_atherosclerosis/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000280_atherosclerosis/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "cd 58680/PhenoGenotypeFiles/\n",
    "mv RootStudyConsentSet_phs000280.RootStudy.v4.p1.c1.HMB-IRB/StudyMetaFiles/* ../../s3/phs000280_atherosclerosis/meta\n",
    "mv RootStudyConsentSet_phs000280.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000090.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000223.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000398.RootStudy.v4.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000557.RootStudy.v3.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000668.RootStudy.v2.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000860.RootStudy.v2.p1.c1.HMB-IRB/PhenotypeFiles/* ../../s3/phs000280_atherosclerosis/phenotype/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000860.RootStudy.v2.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000668.RootStudy.v2.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000557.RootStudy.v3.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000398.RootStudy.v4.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000223.RootStudy.v4.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "mv ChildStudyConsentSet_phs000090.RootStudy.v4.p1.c1.HMB-IRB/GenotypeFiles/* ../../s3/phs000280_atherosclerosis/genotype/original/unprocessed  \n",
    "\n",
    "# upload to S3\n",
    "cd ../../s3\n",
    "aws s3 mv phs000280_atherosclerosis/ s3://rti-common/dbGaP/phs000280_atherosclerosis/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000103_preterm\n",
    "GENEVA Study of Preterm Delivery (DNBC/Iowa) (phs000103.v1.p1)\n",
    "* Download 58702"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58702\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58702/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000103_preterm/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000103_preterm/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000103_preterm/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58702/NHGRI/GEI_PretermBirth_Murray/phs000103v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000103_preterm/meta/\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000103_preterm/phenotype/unprocessed/\n",
    "mv genotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000103_preterm/genotype/original/unprocessed/\n",
    "\n",
    "# upload to S3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000103_preterm/ s3://rti-common/dbGaP/phs000103_preterm/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000203_arterial-disease\n",
    "A Genome-Wide Association Study of Peripheral Arterial Disease (phs000203.v1.p1)\n",
    "* Download 58703"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58703\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58703/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000203_arterial-disease/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000203_arterial-disease/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000203_arterial-disease/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58703/NHGRI/eMERGE_PAD/phs000203v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000203_arterial-disease/meta/\n",
    "mv *.gz /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000203_arterial-disease/phenotype/unprocessed/\n",
    "mv phg000097v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000203_arterial-disease/genotype/original/unprocessed/\n",
    "\n",
    "# upload to S3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000203_arterial-disease/ s3://rti-common/dbGaP/phs000203_arterial-disease/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000308_glaucoma-genes\n",
    "The Glaucoma Genes and Environment (GLAUGEN) Study (phs000308.v1.p1)\n",
    "* Download 58706"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58706\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58706/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000308_glaucoma-genes/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000308_glaucoma-genes/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000308_glaucoma-genes/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58706/NHGRI/Glaucoma_Pasquale/phs000308v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000308_glaucoma-genes/meta/\n",
    "mv genotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000308_glaucoma-genes/genotype/original/unprocessed\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000308_glaucoma-genes/phenotype/unprocessed\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000308_glaucoma-genes/ s3://rti-common/dbGaP/phs000308_glaucoma-genes/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000179_COPD\n",
    "Genetic Epidemiology of COPD (phs000179.v5.p2)\n",
    "* Download 58729"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58729\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58729/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000179_COPD/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000179_COPD/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000179_COPD/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58729/PhenoGenotypeFiles\n",
    "mv RootStudyConsentSet_phs000179.COPDGene.v5.p2.c1.HMB/StudyMetaFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000179_COPD/meta/\n",
    "mv RootStudyConsentSet_phs000179.COPDGene.v5.p2.c1.HMB/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000179_COPD/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000296.COPDGene.v4.p2.c1.HMB/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000179_COPD/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000765.COPDGene.v2.p2.c1.HMB/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000179_COPD/phenotype/unprocessed\n",
    "mv ChildStudyConsentSet_phs000296.COPDGene.v4.p2.c1.HMB/GenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000179_COPD/genotype/original/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000765.COPDGene.v2.p2.c1.HMB/GenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000179_COPD/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000179_COPD/ s3://rti-common/dbGaP/phs000179_COPD/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000388_IPM\n",
    "IPM BioBank GWAS (phs000388.v1.p1)\n",
    "* Download 58711"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58711\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58711/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000388_IPM/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000388_IPM/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000388_IPM/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58711/NHGRI/eMERGE_IPM_CardiovascularDiseasesT2DM/phs000388v1/p1\n",
    "\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000388_IPM/meta/ \n",
    "\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000388_IPM/phenotype/unprocessed/\n",
    "\n",
    "mv genotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000388_IPM/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000388_IPM/ s3://rti-common/dbGaP/phs000388_IPM/ --recursive\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000428_retirement\n",
    "Genetics Resource with the Health and Retirement Study (phs000428.v2.p2)\n",
    "* Download 58712"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58712\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58712/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000428_retirement/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000428_retirement/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000428_retirement/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58712/PhenoGenotypeFiles/RootStudyConsentSet_phs000428.CIDR_Aging_Omni1.v2.p2.c1.NPR\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000428_retirement/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000428_retirement/phenotype/unprocessed/\n",
    "mv StudyMetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000428_retirement/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000428_retirement/ s3://rti-common/dbGaP/phs000428_retirement/ --recursive\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000524_renal-insufficiency\n",
    "Chronic Renal Insufficiency Cohort (CRIC) GWAS (phs000524.v1.p1)\n",
    "* Download 58713"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58713\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58713/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000524_renal-insufficiency/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000524_renal-insufficiency/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000524_renal-insufficiency/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58713/PhenoGenotypeFiles/RootStudyConsentSet_phs000524.ChronicRenalInsufficiency.v1.p1.c1.AHC\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000524_renal-insufficiency/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000524_renal-insufficiency/phenotype/unprocessed/\n",
    "mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000524_renal-insufficiency/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000524_renal-insufficiency/ s3://rti-common/dbGaP/phs000524_renal-insufficiency/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000801_lymphoma\n",
    "Genome-Wide Association Study of Non-Hodgkin Lymphoma (NHL) (phs000801.v2.p1)\n",
    "* Download 58730"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58730\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58730/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# solve problematic decryption\n",
    "for i in $(find ./58730/ -name *ncbi_enc -print); do     \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\";\n",
    "mv ${i} ${j}_tmp.ncbi_enc;\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc;\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\";\n",
    "mv ${j}_tmp ${i/.ncbi_enc/};\n",
    "let j=$j+1;\n",
    "done\n",
    "\n",
    "# double check to make sure successful decryption\n",
    "if [ $(find 58730/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000801_lymphoma/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000801_lymphoma/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000801_lymphoma/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58730/PhenoGenotypeFiles\n",
    "mv RootStudyConsentSet_phs000801.Non_HodgkinLymphoma.v2.p1.c3.GRU-NPU-MDS/StudyMetaFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/meta/\n",
    "mv RootStudyConsentSet_phs000801.Non_HodgkinLymphoma.v2.p1.c3.GRU-NPU-MDS/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000802.Non_HodgkinLymphoma.v2.p1.c3.GRU-NPU-MDS/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000818.Non_HodgkinLymphoma.v2.p1.c3.GRU-NPU-MDS/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000889.Non_HodgkinLymphoma.v1.p1.c3.GRU-NPU-MDS/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000890.Non_HodgkinLymphoma.v1.p1.c3.GRU-NPU-MDS/PhenotypeFiles/* \\\n",
    "   /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000891.Non_HodgkinLymphoma.v1.p1.c3.GRU-NPU-MDS/PhenotypeFiles/* \\ \n",
    "  /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000802.Non_HodgkinLymphoma.v2.p1.c3.GRU-NPU-MDS/GenotypeFiles/* \\\n",
    "  /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/genotype/original/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000818.Non_HodgkinLymphoma.v2.p1.c3.GRU-NPU-MDS/GenotypeFiles/* \\\n",
    "  /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/genotype/original/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000889.Non_HodgkinLymphoma.v1.p1.c3.GRU-NPU-MDS/GenotypeFiles/* \\\n",
    "  /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/genotype/original/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000890.Non_HodgkinLymphoma.v1.p1.c3.GRU-NPU-MDS/GenotypeFiles/* \\\n",
    "  /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/genotype/original/unprocessed/\n",
    "mv ChildStudyConsentSet_phs000891.Non_HodgkinLymphoma.v1.p1.c3.GRU-NPU-MDS/GenotypeFiles/* \\\n",
    "  /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000801_lymphoma/ s3://rti-common/dbGaP/phs000801_lymphoma/ --recursive\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000982_psoriasis\n",
    "Genetic Analysis of Psoriasis and Psoriatic Arthritis (phs000982.v1.p1)\n",
    "* Download 58731"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58731\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58731/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000982_psoriasis/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000982_psoriasis/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000982_psoriasis/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58731/PhenoGenotypeFiles/RootStudyConsentSet_phs000982.Psoriasis_Arthritis.v1.p1.c1.GRU\n",
    "mv Study\\ MetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/meta/ \n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000801_lymphoma/phenotype/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000982_psoriasis/ s3://rti-common/dbGaP/phs000982_psoriasis/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000634_never-smoker\n",
    "A genome wide study of lung cancer in never smokers (phs000634.v1.p1)\n",
    "* Download 58732"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58732\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58732/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# solve problematic decryption\n",
    "for i in $(find ./58732/ -name *ncbi_enc -print); do     \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\";\n",
    "mv ${i} ${j}_tmp.ncbi_enc;\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc;\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\";\n",
    "mv ${j}_tmp ${i/.ncbi_enc/};\n",
    "let j=$j+1;\n",
    "done\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000634_never-smoker/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000634_never-smoker/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000634_never-smoker/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to the directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58732/PhenoGenotypeFiles/RootStudyConsentSet_phs000634.NeverSmokers_LungCancer.v1.p1.c1.GRU-MDS\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000634_never-smoker/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000634_never-smoker/phenotype/unprocessed/\n",
    "mv Study\\ MetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000634_never-smoker/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000634_never-smoker/ s3://rti-common/dbGaP/phs000634_never-smoker/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heroin Cohorts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000952_Eur-AA\n",
    "Substance Dependence GWAS in European- and African Americans (phs000952.v1.p1)\n",
    "* Download 58734"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:3:74: unexpected numeric constant\n2: cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n3: ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58734\n                                                                            ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:3:74: unexpected numeric constant\n2: cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n3: ../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58734\n                                                                            ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58734\n",
    "\n",
    "# Check for successful decryption\n",
    "if [ $(find 58734/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# solve problematic decryption\n",
    "for i in $(find ./58734/ -name *ncbi_enc -print); do     \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\";\n",
    "mv ${i} ${j}_tmp.ncbi_enc;\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc;\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\";\n",
    "mv ${j}_tmp ${i/.ncbi_enc/};\n",
    "let j=$j+1;\n",
    "done\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000952_Eur-AA/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000952_Eur-AA/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000952_Eur-AA/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# mv files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58734/PhenoGenotypeFiles/RootStudyConsentSet_phs000952.SubstanceDependenceEU_AA.v1.p1.c1.DS-ADX-IRB-RD\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000952_Eur-AA/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000952_Eur-AA/phenotype/unprocessed/\n",
    "mv Study\\ MetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000952_Eur-AA/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000952_Eur-AA/ s3://rti-common/dbGaP/phs000952_Eur-AA/ --recursive\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs001109_polymorphisms\n",
    "Addictions: Genotypes, Polymorphisms, and Function/Human Genetic Correlates of Addictive Diseases (phs001109.v1.p1)\n",
    "* Download 58740"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58740\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58740/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# solve problematic decryption\n",
    "for i in $(find ./58740/ -name *ncbi_enc -print); do     \n",
    "echo -e \"\\nmv ${i} ${j}_tmp.ncbi_enc\";\n",
    "mv ${i} ${j}_tmp.ncbi_enc;\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe ${j}_tmp.ncbi_enc;\n",
    "echo \"mv ${j}_tmp ${i/.ncbi_enc/}\";\n",
    "mv ${j}_tmp ${i/.ncbi_enc/};\n",
    "let j=$j+1;\n",
    "done\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs001109_polymorphisms/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs001109_polymorphisms/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs001109_polymorphisms/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58740/PhenoGenotypeFiles/RootStudyConsentSet_phs001109.AddictiveDiseases.v1.p1.c1.DS-ADX-IRB-NPU\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001109_polymorphisms/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001109_polymorphisms/phenotype/unprocessed/\n",
    "mv StudyMetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001109_polymorphisms/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs001109_polymorphisms/ s3://rti-common/dbGaP/phs001109_polymorphisms/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs001213_china\n",
    "Molecular Genetics of Heroin Dependence in China (phs001213.v1.p1)\n",
    "* Download 58742"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58742\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58742/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs001213_china/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs001213_china/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs001213_china/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58742/PhenoGenotypeFiles/RootStudyConsentSet_phs001213.HeroinDependence_China.v1.p1.c1.GRU\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001213_china/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001213_china/phenotype/unprocessed/\n",
    "mv StudyMetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001213_china/meta/\n",
    "\n",
    "# upload to s3 \n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs001213_china/ s3://rti-common/dbGaP/phs001213_china/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000277_GWAS-heroin-c1\n",
    "GWAS of Heroin Dependence (phs000277.v1.p1)\n",
    "Substance Use Dependence Only (phs000277.v1.p1.c1), JAAMH\n",
    "* Download 58743"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58743\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58743/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c1/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c1/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000277_GWAS-heroin-c1/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58743/PhenoGenotypeFiles/RootStudyConsentSet_phs000277.CIDR_HeroinDependence.v1.p1.c1.SUD\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c1/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c1/phenotype/unprocessed/\n",
    "mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c1/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000277_GWAS-heroin-c1/ s3://rti-common/dbGaP/phs000277_GWAS-heroin-c1/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000277_GWAS-heroin-c2\n",
    "GWAS of Heroin Dependence (phs000277.v1.p1)\n",
    "Research on Alcohol and Tobacco Related Phenotypes (phs000277.v1.p1.c2), JAAMH\n",
    "* Download 58746"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58746\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58746/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000277_GWAS-heroin-c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd 58746/PhenoGenotypeFiles/RootStudyConsentSet_phs000277.CIDR_HeroinDependence.v1.p1.c2.RATRP/\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c2/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c2/phenotype/unprocessed/\n",
    "mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c2/meta/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000277_GWAS-heroin-c2/ s3://rti-common/dbGaP/phs000277_GWAS-heroin-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000277_GWAS-heroin-c3\n",
    "GWAS of Heroin Dependence (phs000277.v1.p1)\n",
    "Substance Dependence, Childhood Trauma and Related Disorders(phs000277.v1.p1.c3), JAAMH\n",
    "* Download 58755"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58755\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58755/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c3/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c3/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000277_GWAS-heroin-c3/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58755/PhenoGenotypeFiles/RootStudyConsentSet_phs000277.CIDR_HeroinDependence.v1.p1.c3.SUDCT\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c3/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c3/phenotype/unprocessed/\n",
    "mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c3/meta/ \n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000277_GWAS-heroin-c3/ s3://rti-common/dbGaP/phs000277_GWAS-heroin-c3/ --recursive\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000277_GWAS-heroin-c4\n",
    "GWAS of Heroin Dependence (phs000277.v1.p1)\n",
    "Alcohol Related Conditions (phs000277.v1.p1.c4), JAAMH\n",
    "* Download 58756"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58756\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58756/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c4/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000277_GWAS-heroin-c4/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000277_GWAS-heroin-c4/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58756/PhenoGenotypeFiles/RootStudyConsentSet_phs000277.CIDR_HeroinDependence.v1.p1.c4.ARC\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c4/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c4/phenotype/unprocessed/\n",
    "mv StudyFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000277_GWAS-heroin-c4/meta/ \n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000277_GWAS-heroin-c4/ s3://rti-common/dbGaP/phs000277_GWAS-heroin-c4/ --recursive\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  phs000425_EuAA-alcohol\n",
    "Alcohol Dependence GWAS in European- and African Americans (phs000425.v1.p1)\n",
    "* Download 58757"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58757\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58757/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000425_EuAA-alcohol/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000425_EuAA-alcohol/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000425_EuAA-alcohol/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move data to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58757/JAAMH/CIDR_AlcoholDependence/phs000425v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000425_EuAA-alcohol/meta/\n",
    "mv genotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000425_EuAA-alcohol/genotype/original/unprocessed/\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000425_EuAA-alcohol/phenotype/unprocessed/\n",
    " \n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000425_EuAA-alcohol/ s3://rti-common/dbGap/phs000425_EuAA-alcohol/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000125_COGA-c1\n",
    "CIDR: Collaborative Study on the Genetics of Alcoholism (COGA) (phs000125.v1.p1.c1)\n",
    "* Download 58793"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58793\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58793/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000125_COGA-c1/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000125_COGA-c1/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000125_COGA-c1/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58793/JAAMH/CIDR_COGA/phs000125v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000125_COGA-c1/meta/\n",
    "mv *.gz /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000125_COGA-c1/phenotype/unprocessed/\n",
    "mv phg000024v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000125_COGA-c1/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000125_COGA-c1/ s3://rti-common/dbGaP/phs000125_COGA-c1/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000125_COGA-c2\n",
    "CIDR: Collaborative Study on the Genetics of Alcoholism (COGA) (phs000125.v1.p1.c2)\n",
    "* Download 58794"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58794\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58794/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000125_COGA-c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000125_COGA-c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000125_COGA-c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58794/JAAMH/CIDR_COGA/phs000125v1/p1\n",
    "mv *.pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000125_COGA-c2/meta/\n",
    "mv *.gz /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000125_COGA-c2/phenotype/unprocessed/\n",
    "mv phg000024v1/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000125_COGA-c2/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000125_COGA-c2 s3://rti-common/dbGaP/phs000125_COGA-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000092_SAGE-c1\n",
    "GENEVA Study of Addiction: Genetics and Environment (SAGE) (phs000092.v1.p1.c1)\n",
    "* Download 58795"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58795\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58795/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000092_SAGE-c1/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000092_SAGE-c1/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000092_SAGE-c1/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58795/NHGRI/GENEVA_SAGE/phs000092v1/p1\n",
    "mv *pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000092_SAGE-c1/meta/\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000092_SAGE-c1/phenotype/unprocessed/\n",
    "mv genotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000092_SAGE-c1/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000092_SAGE-c1/ s3://rti-common/dbGaP/phs000092_SAGE-c1/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs000092_SAGE-c2\n",
    "GENEVA Study of Addiction: Genetics and Environment (SAGE) (phs000092.v1.p1.c2)\n",
    "* Download 58799"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58799\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58799/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs000092_SAGE-c2/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs000092_SAGE-c2/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs000092_SAGE-c2/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58799/NHGRI/GENEVA_SAGE/phs000092v1/p1\n",
    "mv *pdf /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000092_SAGE-c2/meta/\n",
    "mv phenotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000092_SAGE-c2/phenotype/unprocessed/\n",
    "mv genotype/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs000092_SAGE-c2/genotype/original/unprocessed/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs000092_SAGE-c2/ s3://rti-common/dbGaP/phs000092_SAGE-c2/ --recursive"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### phs001135_START\n",
    "Starting Treatment with Agonist Replacement Therapy (START) (phs001135.v1.p1)\n",
    "* Download 58817"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# decrypt\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556\n",
    "../../Desktop/sratoolkit/sratoolkit.2.8.2-1-win64/bin/vdb-decrypt.exe -v 58817\n",
    "\n",
    "# check for successful decryption\n",
    "if [ $(find 58817/ -name *ncbi_enc -print | wc -l) = 0 ]; then echo \"Success!\"; else echo \"Failed!\"; fi\n",
    "\n",
    "# create directory structure\n",
    "mkdir -p s3/phs001135_START/{meta,genotype,phenotype}\n",
    "mkdir -p s3/phs001135_START/genotype/original/{final,unprocessed,processing}\n",
    "mkdir s3/phs001135_START/phenotype/{final,unprocessed,processing}\n",
    "\n",
    "# move files to directory structure\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/58817/PhenoGenotypeFiles/RootStudyConsentSet_phs001135.START.v1.p1.c2.GRU\n",
    "mv StudyMetaFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001135_START/meta/\n",
    "mv GenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001135_START/genotype/original/unprocessed/\n",
    "mv PhenotypeFiles/* /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/phs001135_START/phenotype/\n",
    "\n",
    "# upload to s3\n",
    "cd /cygdrive/c/Users/jmarks/ncbi/dbGaP-2556/s3/\n",
    "aws s3 mv phs001135_START/ s3://rti-common/dbGaP/phs001135_START/ --recursive"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  },
  "toc": {
   "nav_menu": {
    "height": "12px",
    "width": "160px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
