{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Applying Nyholt correction to snp lookup for GSCAN\n",
    "__Author:__ Jesse Marks\n",
    "\n",
    "This is documentation is relevant to [GitHub Issue #59](https://github.com/RTIInternational/bioinformatics/issues/59#issuecomment-355282624). In order to make corrections for multiple testing in the data, one can apply the [Nyholt Correction Method](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1181954/). The Nyholt Correction Method is a simple correction for multiple testing for single-nucleotide polymorphisms in linkage disequilibrium with each other.\n",
    "\n",
    "## Nyholt Correction Script\n",
    "This script was aquired from Nathan Gaddis. The script is currently on MIDAS but will need to be transfered to EC2 at some point in the near future."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Copy list from \\\\rcdcollaboration01.rti.ns\\johnson_share\\1.HIV GWAS II\\technical\\Meta\\Meta001_GoGeneCandidates_ListForNyholtCorrection_05.26.16.xlsx\n",
    "\n",
    "# Copy list from \\\\rcdcollaboration01.rti.ns\\johnson_share\\1.HIV GWAS II\\technical\\Meta\\Meta001_GoGeneCandidates_ListForNyholtCorrection_05.26.16.xlsx\n",
    "\n",
    "inFile=/home/jmarks/nyholt/044_snps.txt\n",
    "outFile=/home/jmarks/nyholt/044_snps.final\n",
    "# Copy variants with phase 3 IDs to final file\n",
    "perl -lane 'if ($F[0] =~ /\\:/) { print $F[0]; }' $inFile >> $outFile\n",
    "# Convert non-phase 3 variant IDs to phase 3 IDs based on name\n",
    "for variant in $(perl -lane 'if ($F[0] !~ /\\:/) { print $F[0]; }' $inFile); do\n",
    "    chr=$(grep -P \"$variant\\s\" $inFile | perl -lane 'print $F[1];')\n",
    "    gunzip -c /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz|\n",
    "        grep \"$variant:\" |\n",
    "        perl -lane 'print $F[0];'\n",
    "done > $outFile\n",
    "# Convert non-phase 3 variant IDs to phase 3 IDs based on position\n",
    "for position in $(perl -lane 'if ($F[0] !~ /\\:/) { print $F[2]; }' $inFile); do\n",
    "    chr=$(grep -P \"$position\" $inFile | grep -v \":\" | perl -lane 'print $F[1];')\n",
    "    gunzip -c /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr${chr}.legend.gz|\n",
    "        grep \":$position:\"\n",
    "done > /home/jmarks/nyholt/044_variant_list.nonphase3.legend\n",
    "perl -lane 'BEGIN { $lastPosition = 0 } if ($F[1] != $lastPosition) { print $F[0]; $lastPosition = $F[1]; }' /home/jmarks/nyholt/044_variant_list.nonphase3.legend >> $outFile\n",
    "\n",
    "\n",
    "### START Extract replication SNPs from 1000G ###\n",
    "\n",
    "# Extract variants from 1000G panel\n",
    "for (( chr=1; chr<23; chr++ )); do\n",
    "    /share/nas03/bioinformatics_group/software/scripts/qsub_job.sh \\\n",
    "      --job_name ALL_1000G \\\n",
    "      --script_prefix /home/jmarks/nyholt/044/1000G_ALL.chr$chr.extracted_variants \\\n",
    "      --mem 14 \\\n",
    "      --priority 0 \\\n",
    "      --program /share/nas03/bioinformatics_group/software/perl/convert_reference_panels.v2.pl \\\n",
    "      --impute2_hap /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.hap.gz \\\n",
    "      --impute2_legend /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/1000GP_Phase3_chr$chr.legend.gz \\\n",
    "      --impute2_sample /share/nas03/bioinformatics_group/data/ref_panels/1000G/2014.10/1000GP_Phase3.sample \\\n",
    "      --extract variant_list.final \\\n",
    "      --out 1000G_ALL.chr$chr.extracted_variants \\\n",
    "      --generate_plink_ped_file \\\n",
    "      --generate_plink_map_file \\\n",
    "      --chr $chr\n",
    "done\n",
    "\n",
    "### END Extract replication SNPs from 1000G ###\n",
    "\n",
    "\n",
    "\n",
    "### START Generate correlation matrix ###\n",
    "\n",
    "for (( chr=1; chr<23; chr++ )); do\n",
    "    /share/nas03/bioinformatics_group/software/plink \\\n",
    "        --noweb \\\n",
    "        --file 1000G_ALL.chr$chr.extracted_variants.plink \\\n",
    "        --r \\\n",
    "        --matrix \\\n",
    "        --out 1000G_ALL.chr$chr.extracted_variants.r\n",
    "done\n",
    "\n",
    "# for chr 2, rs568413:21235475:T:C appears to be causing trouble so remove it and recalculate matrix\n",
    "#echo rs568413:21235475:T:C > /data/common/data/studies/hiv/imputed/nyholt/009/rs568413\n",
    "#/data/common/software/plink \\\n",
    "#    --noweb \\\n",
    "#    --file /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr2.extracted_variants.plink \\\n",
    "#    --exclude /data/common/data/studies/hiv/imputed/nyholt/009/rs568413 \\\n",
    "#    --recode \\\n",
    "#    --out /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr2.extracted_variants.minus_rs568413.plink\n",
    "#/data/common/software/plink \\\n",
    "#    --noweb \\\n",
    "#    --file /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr2.extracted_variants.minus_rs568413.plink \\\n",
    "#    --r \\\n",
    "#    --matrix \\\n",
    "#    --out /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr2.extracted_variants.minus_rs568413.r\n",
    "\n",
    " ### END Generate correlation matrix ###\n",
    "\n",
    "\n",
    "### START Run matSpD analysis ###\n",
    "\n",
    "cd /data/common/software/matSpD/\n",
    "for chr in 1 2 4 6 8 10 11 12 17 19; do\n",
    "    echo $chr\n",
    "    cp /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr$chr.extracted_variants.r.ld \\\n",
    "        /data/common/software/matSpD/correlation.matrix\n",
    "    R CMD BATCH matSpDlite.R\n",
    "    mv /data/common/software/matSpD/matSpDlite.out /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr$chr.extracted_variants.matspdlite\n",
    "done\n",
    "\n",
    "cp /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr2.extracted_variants.minus_rs568413.r.ld \\\n",
    "    /data/common/software/matSpD/correlation.matrix\n",
    "R CMD BATCH matSpDlite.R\n",
    "mv /data/common/software/matSpD/matSpDlite.out /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr2.extracted_variants.matspdlite\n",
    "\n",
    "veffLi=0\n",
    "for chr in 1 2 4 6 8 10 11 12 17 19; do\n",
    "    chrVeffLi=$(grep -A 2 Equation /data/common/data/studies/hiv/imputed/nyholt/009/1000G_ALL.chr$chr.extracted_variants.matspdlite | tail -n +3 | perl -pe 's/^\\s+//; s/\\s+$//;')\n",
    "    echo $chr\n",
    "    echo $chrVeffLi\n",
    "    veffLi=`echo $veffLi + $chrVeffLi | bc`\n",
    "done\n",
    "echo $veffLi\n",
    "\n",
    "### END Run matSpD analysis ###\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.2"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
