
% Default to the notebook output style

    


% Inherit from the specified cell style.




    
\documentclass[11pt]{article}

    
    
    \usepackage[T1]{fontenc}
    % Nicer default font (+ math font) than Computer Modern for most use cases
    \usepackage{mathpazo}

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % We will generate all images so they have a width \maxwidth. This means
    % that they will get their normal width if they fit onto the page, but
    % are scaled down if they would overflow the margins.
    \makeatletter
    \def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth
    \else\Gin@nat@width\fi}
    \makeatother
    \let\Oldincludegraphics\includegraphics
    % Set max figure width to be 80% of text width, for now hardcoded.
    \renewcommand{\includegraphics}[1]{\Oldincludegraphics[width=.8\maxwidth]{#1}}
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionLabelFormat{nolabel}{}
    \captionsetup{labelformat=nolabel}

    \usepackage{adjustbox} % Used to constrain images to a maximum size 
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage[utf8x]{inputenc} % Allow utf-8 characters in the tex document
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics 
                         % to support a larger range 
    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    

    
    
    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}
    
    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    
    
    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatability definitions
    \def\gt{>}
    \def\lt{<}
    % Document parameters
    \title{20190530\_nicotine\_ftnd\_gscan\_snp\_lookup}
    
    
    

    % Pygments definitions
    
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % Exact colors from NB
    \definecolor{incolor}{rgb}{0.0, 0.0, 0.5}
    \definecolor{outcolor}{rgb}{0.545, 0.0, 0.0}



    
    % Prevent overflowing lines due to hard-to-break entities
    \sloppy 
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults
    
    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}
    
    

    \begin{document}
    
    
    \maketitle
    
    

    
    \hypertarget{extension-of-gscan-results-to-nicotine-dependence-issue-59}{%
\section{Extension of GSCAN results to nicotine dependence (issue
\#59)}\label{extension-of-gscan-results-to-nicotine-dependence-issue-59}}

\textbf{Author:} Jesse Marks

We have preliminary results from a very large-scale genome-wide study
for cigarette smoking phenotypes that relate to our nicotine dependence
GWAS results. See the supplemental Tables S7-S9 here:

\texttt{//RTPNFIL02/dhancock/Analysis/GSCAN/shared\ MS\ version\ 1/Supplementary\_Tables\_S6-S12\_Loci.xlsx}.

The phenotypes of interest to us include:

\begin{enumerate}
\def\labelenumi{\arabic{enumi})}
\item
  \textbf{Age of smoking initiation (AI)} - supplemental table 6
\item
  \textbf{Cigarettes per day (CPD)} - supplemental table 7
\item
  \textbf{Smoking cessation (SC)} - supplemental table 8
\item
  \textbf{smoking initiation (SI)} - supplemental table 9
\end{enumerate}

We're interested in seeing whether these associations extend over to
nicotine dependence. I use the SNP look-up script to extend Tables S6-S9
with our FTND meta-analysis results on S3 at: *
\texttt{s3://rti-nd/META/1df/results/cross/final} *
\texttt{s3://rti-nd/META/1df/results/ea/final} *
\texttt{s3://rti-nd/META/1df/results/aa}

\textbf{Note} that because the meta-analysis results files are quite
large, we will not perform the search locally.

    \hypertarget{download-data}{%
\subsection{Download Data}\label{download-data}}

Create the directory structure and download the meta-analysis results to
EC2. Also, create a list of SNPs for the lookup from the GSCAN Excel
sheet. Copy the SNPs from each of the four Excel sheets (supplemental
tables) into separate files and then combine them (no duplicates).

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PY{c+c1}{\PYZsh{}\PYZsh{} EC2 \PYZsh{}\PYZsh{}}
        
        \PY{c+c1}{\PYZsh{} Create directory }
        \PY{n}{mkdir} \PY{o}{\PYZhy{}}\PY{n}{p} \PY{o}{/}\PY{n}{shared}\PY{o}{/}\PY{n}{jmarks}\PY{o}{/}\PY{n}{nicotine}\PY{o}{/}\PY{n}{gscan}\PY{o}{/}\PY{n}{lookup}\PY{o}{/}\PY{l+m+mi}{20190530}\PY{o}{/}\PY{p}{\PYZob{}}\PY{n}{aa}\PY{p}{,}\PY{n}{ea}\PY{p}{,}\PY{n}{cross}\PY{p}{\PYZcb{}}
        \PY{n}{cd} \PY{o}{/}\PY{n}{shared}\PY{o}{/}\PY{n}{jmarks}\PY{o}{/}\PY{n}{nicotine}\PY{o}{/}\PY{n}{gscan}\PY{o}{/}\PY{n}{lookup}\PY{o}{/}\PY{l+m+mi}{20190530}
        
        \PY{c+c1}{\PYZsh{} populate these files with the SNPs from respective Excel sheets}
        \PY{n}{touch} \PY{n}{age\PYZus{}of\PYZus{}initiation}\PY{o}{.}\PY{n}{tsv}  \PY{n}{cig\PYZus{}per\PYZus{}day}\PY{o}{.}\PY{n}{tsv}  \PY{n}{smoking\PYZus{}cessation}\PY{o}{.}\PY{n}{tsv}  \PY{n}{smoking\PYZus{}initiation}\PY{o}{.}\PY{n}{tsv}
        \PY{n}{wc} \PY{o}{\PYZhy{}}\PY{n}{l} \PY{o}{*} \PY{c+c1}{\PYZsh{} note these are with headers}
        \PY{l+s+sd}{\PYZdq{}\PYZdq{}\PYZdq{}}
        \PY{l+s+sd}{   11 age\PYZus{}of\PYZus{}initiation}
        \PY{l+s+sd}{   56 cig\PYZus{}per\PYZus{}day}
        \PY{l+s+sd}{   25 smoking\PYZus{}cessation}
        \PY{l+s+sd}{  377 smoking\PYZus{}initiation}
        \PY{l+s+sd}{\PYZdq{}\PYZdq{}\PYZdq{}}
        
        \PY{n}{head} \PY{n}{age\PYZus{}of\PYZus{}initiation}
        \PY{l+s+sd}{\PYZdq{}\PYZdq{}\PYZdq{}}
        \PY{l+s+sd}{Chr     Pos     rsID    Reference Allele        Alternate Allele}
        \PY{l+s+sd}{2       145638766       rs72853300      C       T}
        \PY{l+s+sd}{2       225353649       rs12611472      T       C}
        \PY{l+s+sd}{2       63622309        rs7559982       T       A}
        \PY{l+s+sd}{3       85699040        rs11915747      C       G}
        \PY{l+s+sd}{4       140908755       rs13136239      G       A}
        \PY{l+s+sd}{\PYZdq{}\PYZdq{}\PYZdq{}}
        
        \PY{c+c1}{\PYZsh{} combine SNPs into one file, make sure SNPs are not listed twice}
        \PY{n}{head} \PY{o}{\PYZhy{}}\PY{l+m+mi}{1} \PY{n}{age\PYZus{}of\PYZus{}initiation}\PY{o}{.}\PY{n}{tsv} \PY{o}{\PYZgt{}} \PY{n}{combined\PYZus{}snp\PYZus{}list}\PY{o}{.}\PY{n}{tsv}
        \PY{k}{for} \PY{n}{file} \PY{o+ow}{in} \PY{n}{age\PYZus{}of\PYZus{}initiation}\PY{o}{.}\PY{n}{tsv} \PY{n}{cig\PYZus{}per\PYZus{}day}\PY{o}{.}\PY{n}{tsv} \PY{n}{smoking\PYZus{}cessation}\PY{o}{.}\PY{n}{tsv} \PY{n}{smoking\PYZus{}initiation}\PY{o}{.}\PY{n}{tsv}\PY{p}{;} \PY{n}{do}
            \PY{n}{tail} \PY{o}{\PYZhy{}}\PY{n}{n} \PY{o}{+}\PY{l+m+mi}{2} \PY{err}{\PYZdl{}}\PY{n}{file} \PY{o}{\PYZgt{}\PYZgt{}} \PY{n}{combined\PYZus{}snp\PYZus{}list}\PY{o}{.}\PY{n}{tsv}
        \PY{n}{done}
        
        \PY{c+c1}{\PYZsh{} filter so that there are no duplicated SNPs (no header either)}
        \PY{n}{tail} \PY{o}{\PYZhy{}}\PY{n}{n} \PY{o}{+}\PY{l+m+mi}{2} \PY{n}{combined\PYZus{}snp\PYZus{}list}\PY{o}{.}\PY{n}{tsv} \PY{o}{|} \PY{n}{sort} \PY{o}{\PYZhy{}}\PY{n}{u} \PY{o}{\PYZgt{}} \PY{n}{combined\PYZus{}snp\PYZus{}list\PYZus{}filtered}\PY{o}{.}\PY{n}{tsv}
        
        \PY{c+c1}{\PYZsh{} convert to 1000g\PYZus{}p3 format}
        \PY{n}{awk} \PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{\PYZob{}}\PY{l+s+s1}{print \PYZdl{}3}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{:}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{\PYZdl{}2}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{:}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{\PYZdl{}4}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{:}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{\PYZdl{}5}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+se}{\PYZbs{}t}\PY{l+s+s1}{\PYZdq{}}\PY{l+s+s1}{\PYZdl{}1\PYZcb{}}\PY{l+s+s1}{\PYZsq{}} \PY{n}{combined\PYZus{}snp\PYZus{}list\PYZus{}filtered}\PY{o}{.}\PY{n}{tsv} \PY{o}{\PYZgt{}} \PY{n}{combined\PYZus{}snp\PYZus{}list\PYZus{}filtered\PYZus{}1000g\PYZus{}p3\PYZus{}chr}\PY{o}{.}\PY{n}{tsv}
\end{Verbatim}


    \begin{verbatim}
wc -l combined_snp_list_filtered_1000g_p3_chr.tsv
460 combined_snp_list_filtered_1000g_p3_chr.tsv

 head combined_snp_list_filtered_1000g_p3_chr.tsv
rs7920501:10043159:T:A  10
rs7901883:103186838:G:A 10
rs11594623:103960351:T:C        10
rs11191269:104120522:C:G        10
rs28408682:104403310:A:G        10
rs12244388:104640052:G:A        10
rs111842178:104852121:A:G       10
rs34970111:106078937:C:T        10
rs9787523:106460460:T:C 10
rs11192347:106929313:G:A        10
\end{verbatim}

    \hypertarget{snp-lookup}{%
\subsection{SNP lookup}\label{snp-lookup}}

I need to create a dictionary of the GSCAN SNPs and then see if each SNP
in the meta-analysis is in the dictionary. I think this makes more sense
rather than vice-versa; in particular, creating a dictionary for each
SNP in the meta-analysis and then searching to see if the GSCAN SNPs are
in the dictionary. This latter strategy would require a large amount of
memory to create the Python dictionary. The former strategy makes more
sense when comsidering computation expense.

Also note that there are some SNPs in the meta-analysis which have the
format of \texttt{chr:position:a1:a2} instead of
\texttt{rsid:position:a1:a2}. I think the reason is that these SNPs of
the former format did not have an associated rsID available. If a GSCAN
SNP is not found in the lookup, then we need to output the SNPs that
were not found and deal with those later. It might be the case that we
have to convert them from \texttt{rsid:position:a1:a2} format to
\texttt{chr:position:a1:a2} and then perform the search again with just
these SNPs.

    \hypertarget{aa}{%
\subsubsection{AA}\label{aa}}

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PY{c+c1}{\PYZsh{}\PYZsh{}\PYZsh{} Python3 \PYZsh{}\PYZsh{}\PYZsh{}}
        \PY{l+s+sd}{\PYZdq{}\PYZdq{}\PYZdq{}}
        \PY{l+s+sd}{*SNP lookup*}
        
        \PY{l+s+sd}{    Make sure the IDs are of the same format for the snp\PYZhy{}list}
        \PY{l+s+sd}{    and the IDs in the meta\PYZhy{}analysis results. e.g. 1000g\PYZus{}p3 or rsID only}
        \PY{l+s+sd}{\PYZdq{}\PYZdq{}\PYZdq{}}
        \PY{k+kn}{import} \PY{n+nn}{gzip}
        
        \PY{c+c1}{\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}}
        \PY{n}{date} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{20190530}\PY{l+s+s2}{\PYZdq{}} \PY{c+c1}{\PYZsh{} date the results were generated (in results files name)}
        \PY{n}{ancestry} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{aa}\PY{l+s+s2}{\PYZdq{}}
        
        \PY{k}{if} \PY{n}{ancestry}\PY{o}{==}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{aa}\PY{l+s+s2}{\PYZdq{}}\PY{p}{:}
            \PY{n}{pop} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{afr}\PY{l+s+s2}{\PYZdq{}}
        \PY{k}{elif} \PY{n}{ancestry}\PY{o}{==}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{ea}\PY{l+s+s2}{\PYZdq{}}\PY{p}{:}
            \PY{n}{pop} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{eur}\PY{l+s+s2}{\PYZdq{}}
        \PY{k}{else}\PY{p}{:}
            \PY{n}{pop} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{afr+eur}\PY{l+s+s2}{\PYZdq{}}
        
        \PY{c+c1}{\PYZsh{}\PYZsh{} dict to hold gscan snps and the number of times they were found.}
        \PY{c+c1}{\PYZsh{}\PYZsh{} we can tell which SNPs did not show up in any of the meta files}
        \PY{n}{gscan\PYZus{}dict} \PY{o}{=}  \PY{p}{\PYZob{}}\PY{p}{\PYZcb{}}
        \PY{n}{base\PYZus{}dir} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{/shared/jmarks/nicotine/gscan/lookup/20190530}\PY{l+s+s2}{\PYZdq{}}
        \PY{n}{snp\PYZus{}list} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{/combined\PYZus{}snp\PYZus{}list\PYZus{}filtered\PYZus{}1000g\PYZus{}p3\PYZus{}chr.tsv}\PY{l+s+s2}{\PYZdq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{base\PYZus{}dir}\PY{p}{)}
        
        \PY{k}{for} \PY{n}{chrom} \PY{o+ow}{in} \PY{n+nb}{range}\PY{p}{(}\PY{l+m+mi}{1}\PY{p}{,}\PY{l+m+mi}{23}\PY{p}{)}\PY{p}{:}
            \PY{n}{progress} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Prosessing }\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{ }\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZdq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{ancestry}\PY{p}{,} \PY{n}{chrom}\PY{p}{)}
            \PY{c+c1}{\PYZsh{}print(progress)}
            \PY{n}{out\PYZus{}file} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{/}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{/}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZus{}}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZus{}chr}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup.txt}\PY{l+s+s2}{\PYZdq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{base\PYZus{}dir}\PY{p}{,} \PY{n}{ancestry}\PY{p}{,} \PY{n}{date}\PY{p}{,} \PY{n}{pop}\PY{p}{,} \PY{n}{chrom}\PY{p}{)}
            \PY{n}{results} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{20181106\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3.afr.chr}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{.exclude\PYZus{}singletons.1df.gz}\PY{l+s+s2}{\PYZdq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{chrom}\PY{p}{)}
            \PY{n}{meta} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{/shared/jmarks/nicotine/meta/results/}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{/}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZdq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{ancestry}\PY{p}{,} \PY{n}{results}\PY{p}{)}
            \PY{n}{not\PYZus{}found} \PY{o}{=} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{/}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{/}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZus{}}\PY{l+s+si}{\PYZob{}\PYZcb{}}\PY{l+s+s2}{\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}snps\PYZus{}not\PYZus{}found}\PY{l+s+s2}{\PYZdq{}}\PY{o}{.}\PY{n}{format}\PY{p}{(}\PY{n}{base\PYZus{}dir}\PY{p}{,} \PY{n}{ancestry}\PY{p}{,} \PY{n}{date}\PY{p}{,} \PY{n}{pop}\PY{p}{)}
        \PY{c+c1}{\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}\PYZsh{}}
        
            \PY{k}{with} \PY{n}{gzip}\PY{o}{.}\PY{n}{open}\PY{p}{(}\PY{n}{meta}\PY{p}{)} \PY{k}{as} \PY{n}{metF}\PY{p}{,} \PY{n+nb}{open}\PY{p}{(}\PY{n}{snp\PYZus{}list}\PY{p}{)} \PY{k}{as} \PY{n}{gscanF}\PY{p}{,} \PY{n+nb}{open}\PY{p}{(}\PY{n}{out\PYZus{}file}\PY{p}{,} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{w}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)} \PY{k}{as} \PY{n}{outF}\PY{p}{:}
                \PY{n}{gscan\PYZus{}line} \PY{o}{=} \PY{n}{gscanF}\PY{o}{.}\PY{n}{readline}\PY{p}{(}\PY{p}{)}
                \PY{n}{met\PYZus{}head} \PY{o}{=} \PY{n}{metF}\PY{o}{.}\PY{n}{readline}\PY{p}{(}\PY{p}{)}
                \PY{n}{met\PYZus{}line} \PY{o}{=} \PY{n}{metF}\PY{o}{.}\PY{n}{readline}\PY{p}{(}\PY{p}{)}
        
                \PY{n}{outF}\PY{o}{.}\PY{n}{write}\PY{p}{(}\PY{n}{met\PYZus{}head}\PY{p}{)}
        
                \PY{c+c1}{\PYZsh{}\PYZsh{} create a dictionary containing the gscan snps (just 1000G\PYZus{}p3 portion)}
                \PY{k}{if} \PY{n+nb}{len}\PY{p}{(}\PY{n}{gscan\PYZus{}dict}\PY{p}{)} \PY{o}{==} \PY{l+m+mi}{0}\PY{p}{:}
                    \PY{k}{while} \PY{n}{gscan\PYZus{}line}\PY{p}{:}
                        \PY{n}{key} \PY{o}{=} \PY{n}{gscan\PYZus{}line}\PY{o}{.}\PY{n}{split}\PY{p}{(}\PY{p}{)}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]} \PY{c+c1}{\PYZsh{} SNP id}
                        \PY{n}{value} \PY{o}{=} \PY{n}{gscan\PYZus{}line}\PY{o}{.}\PY{n}{split}\PY{p}{(}\PY{p}{)}\PY{p}{[}\PY{l+m+mi}{1}\PY{p}{]} \PY{c+c1}{\PYZsh{}chromosome}
                        \PY{n}{gscan\PYZus{}dict}\PY{p}{[}\PY{n}{key}\PY{p}{]} \PY{o}{=} \PY{p}{[}\PY{l+m+mi}{0}\PY{p}{,} \PY{n}{value}\PY{p}{]}
                        \PY{n}{gscan\PYZus{}line} \PY{o}{=} \PY{n}{gscanF}\PY{o}{.}\PY{n}{readline}\PY{p}{(}\PY{p}{)}
        
                \PY{k}{while} \PY{n}{met\PYZus{}line}\PY{p}{:}
                    \PY{n}{met\PYZus{}id} \PY{o}{=} \PY{n}{met\PYZus{}line}\PY{o}{.}\PY{n}{split}\PY{p}{(}\PY{p}{)}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]} \PY{c+c1}{\PYZsh{} the 1000g\PYZus{}p3 ID in the meta\PYZhy{}analysis}
                    \PY{k}{if} \PY{n}{met\PYZus{}id} \PY{o+ow}{in} \PY{n}{gscan\PYZus{}dict}\PY{p}{:}
                        \PY{n}{gscan\PYZus{}dict}\PY{p}{[}\PY{n}{met\PYZus{}id}\PY{p}{]}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]} \PY{o}{+}\PY{o}{=} \PY{l+m+mi}{1}
                        \PY{n}{outF}\PY{o}{.}\PY{n}{write}\PY{p}{(}\PY{n}{met\PYZus{}line}\PY{p}{)}
                    \PY{n}{met\PYZus{}line} \PY{o}{=} \PY{n}{metF}\PY{o}{.}\PY{n}{readline}\PY{p}{(}\PY{p}{)}
        
        \PY{k}{with} \PY{n+nb}{open}\PY{p}{(}\PY{n}{not\PYZus{}found}\PY{p}{,} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{w}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)} \PY{k}{as} \PY{n}{notF}\PY{p}{:}
            \PY{n}{notF}\PY{o}{.}\PY{n}{write}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{id}\PY{l+s+se}{\PYZbs{}t}\PY{l+s+s2}{chr}\PY{l+s+se}{\PYZbs{}n}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
            \PY{k}{for} \PY{n}{key}\PY{p}{,} \PY{n}{value} \PY{o+ow}{in} \PY{n}{gscan\PYZus{}dict}\PY{o}{.}\PY{n}{items}\PY{p}{(}\PY{p}{)}\PY{p}{:}
                \PY{k}{if} \PY{n}{value}\PY{p}{[}\PY{l+m+mi}{0}\PY{p}{]}\PY{o}{==}\PY{l+m+mi}{0}\PY{p}{:}
                    \PY{n}{notF}\PY{o}{.}\PY{n}{write}\PY{p}{(}\PY{n}{key} \PY{o}{+} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+se}{\PYZbs{}t}\PY{l+s+s2}{\PYZdq{}} \PY{o}{+}  \PY{n}{value}\PY{p}{[}\PY{l+m+mi}{1}\PY{p}{]} \PY{o}{+} \PY{l+s+s2}{\PYZdq{}}\PY{l+s+se}{\PYZbs{}n}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}


    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PY{c+c1}{\PYZsh{}\PYZsh{}\PYZsh{} bash \PYZsh{}\PYZsh{}\PYZsh{}}
        
        \PY{c+c1}{\PYZsh{}\PYZsh{} merge results}
        \PY{n}{cd} \PY{o}{/}\PY{n}{shared}\PY{o}{/}\PY{n}{jmarks}\PY{o}{/}\PY{n}{nicotine}\PY{o}{/}\PY{n}{gscan}\PY{o}{/}\PY{n}{lookup}\PY{o}{/}\PY{l+m+mi}{20190530}\PY{o}{/}\PY{n}{aa}
        \PY{n}{head} \PY{o}{\PYZhy{}}\PY{l+m+mi}{1} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr\PYZus{}chr6\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup}\PY{o}{.}\PY{n}{txt} \PY{o}{\PYZgt{}}\PYZbs{}
            \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr\PYZus{}ftnd\PYZus{}1df\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}lookup\PYZus{}merged\PYZus{}results}\PY{o}{.}\PY{n}{txt}
            
        \PY{k}{for} \PY{n}{file} \PY{o+ow}{in} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr\PYZus{}chr}\PY{p}{\PYZob{}}\PY{l+m+mf}{1.}\PY{o}{.}\PY{l+m+mi}{22}\PY{p}{\PYZcb{}}\PY{n}{\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup}\PY{o}{.}\PY{n}{txt}\PY{p}{;} \PY{n}{do}
            \PY{n}{tail} \PY{o}{\PYZhy{}}\PY{n}{n} \PY{o}{+}\PY{l+m+mi}{2} \PY{err}{\PYZdl{}}\PY{n}{file} \PY{o}{\PYZgt{}\PYZgt{}} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr\PYZus{}ftnd\PYZus{}1df\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}lookup\PYZus{}merged\PYZus{}results}\PY{o}{.}\PY{n}{txt} 
        \PY{n}{done} 
\end{Verbatim}


    \hypertarget{ea}{%
\subsubsection{EA}\label{ea}}

simply change the \texttt{ancestry,\ pop,} and \texttt{date} variables
at the top of the script.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PY{c+c1}{\PYZsh{}\PYZsh{}\PYZsh{} bash \PYZsh{}\PYZsh{}\PYZsh{} }
        
        \PY{c+c1}{\PYZsh{}\PYZsh{} merge results}
        \PY{n}{cd} \PY{o}{/}\PY{n}{shared}\PY{o}{/}\PY{n}{jmarks}\PY{o}{/}\PY{n}{nicotine}\PY{o}{/}\PY{n}{gscan}\PY{o}{/}\PY{n}{lookup}\PY{o}{/}\PY{l+m+mi}{20190530}\PY{o}{/}\PY{n}{ea}
        \PY{n}{head} \PY{o}{\PYZhy{}}\PY{l+m+mi}{1} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}eur\PYZus{}chr8\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup}\PY{o}{.}\PY{n}{txt} \PY{o}{\PYZgt{}}\PYZbs{}
            \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}eur\PYZus{}ftnd\PYZus{}1df\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}lookup\PYZus{}merged\PYZus{}results}\PY{o}{.}\PY{n}{txt}
        \PY{k}{for} \PY{n}{file} \PY{o+ow}{in} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}eur\PYZus{}chr}\PY{p}{\PYZob{}}\PY{l+m+mf}{1.}\PY{o}{.}\PY{l+m+mi}{22}\PY{p}{\PYZcb{}}\PY{n}{\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup}\PY{o}{.}\PY{n}{txt}\PY{p}{;}\PY{n}{do}
            \PY{n}{tail} \PY{o}{\PYZhy{}}\PY{n}{n} \PY{o}{+}\PY{l+m+mi}{2} \PY{err}{\PYZdl{}}\PY{n}{file} \PY{o}{\PYZgt{}\PYZgt{}} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}eur\PYZus{}ftnd\PYZus{}1df\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}lookup\PYZus{}merged\PYZus{}results}\PY{o}{.}\PY{n}{txt}
        \PY{n}{done}
\end{Verbatim}


    \hypertarget{cross-ancestry}{%
\subsubsection{Cross-Ancestry}\label{cross-ancestry}}

simply change the \texttt{ancestry,\ pop,} and \texttt{date} variables
at the top of the script.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PY{c+c1}{\PYZsh{}\PYZsh{}\PYZsh{} bash \PYZsh{}\PYZsh{}\PYZsh{} }
        
        \PY{c+c1}{\PYZsh{}\PYZsh{} merge results}
        \PY{n}{cd} \PY{o}{/}\PY{n}{shared}\PY{o}{/}\PY{n}{jmarks}\PY{o}{/}\PY{n}{nicotine}\PY{o}{/}\PY{n}{gscan}\PY{o}{/}\PY{n}{lookup}\PY{o}{/}\PY{n}{cross}
        \PY{n}{head} \PY{o}{\PYZhy{}}\PY{l+m+mi}{1} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr}\PY{o}{+}\PY{n}{eur\PYZus{}chr8\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup}\PY{o}{.}\PY{n}{txt} \PY{o}{\PYZgt{}}\PYZbs{}
            \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr}\PY{o}{+}\PY{n}{eur\PYZus{}ftnd\PYZus{}1df\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}lookup\PYZus{}merged\PYZus{}results}\PY{o}{.}\PY{n}{txt}
        \PY{k}{for} \PY{n}{file} \PY{o+ow}{in} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr}\PY{o}{+}\PY{n}{eur\PYZus{}chr}\PY{p}{\PYZob{}}\PY{l+m+mf}{1.}\PY{o}{.}\PY{l+m+mi}{22}\PY{p}{\PYZcb{}}\PY{n}{\PYZus{}ftnd\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}1df\PYZus{}gscan\PYZus{}lookup}\PY{o}{.}\PY{n}{txt}\PY{p}{;}\PY{n}{do}
            \PY{n}{tail} \PY{o}{\PYZhy{}}\PY{n}{n} \PY{o}{+}\PY{l+m+mi}{2} \PY{err}{\PYZdl{}}\PY{n}{file} \PY{o}{\PYZgt{}\PYZgt{}} \PY{l+m+mi}{20190530}\PY{n}{\PYZus{}afr}\PY{o}{+}\PY{n}{eur\PYZus{}ftnd\PYZus{}1df\PYZus{}meta\PYZus{}analysis\PYZus{}wave3\PYZus{}gscan\PYZus{}lookup\PYZus{}merged\PYZus{}results}\PY{o}{.}\PY{n}{txt}
        \PY{n}{done}
\end{Verbatim}


    \hypertarget{retrieve-missing-snps}{%
\section{Retrieve Missing SNPs}\label{retrieve-missing-snps}}

Some SNPs were not retrieved. * 43 not found for AA * 10 not found for
EA * 10 not found for cross-ancestry


    % Add a bibliography block to the postdoc
    
    
    
    \end{document}
